{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Tests\n",
    "\n",
    "This notebook is intended for running different kinds of analyses that would validate our reconstruction.\n",
    "\n",
    "[1. Subsystem Overview and Analysis](#subsystems) <br>\n",
    "[2. Context-specific Model Generation](#context_specific) <br>\n",
    "[3. tSNE Comparison of Models](#tSNE) <br>\n",
    "[4. Biomass prediction using exp. data](#biomass) <br>\n",
    "[5. Flux Sampling Analysis](#fsa) <br>\n",
    "[6. Pathway Enrichment Analysis](#pea) <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Subsystem Overview and Analysis <a id='subsystems'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from google_sheet import GoogleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate Subsystems datasets from Google Sheet ----- #####\n",
    "\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_subsystems = 'Summary Systems'\n",
    "sheet_reactions = 'Rxns'\n",
    "subsystems = sheet.read_google_sheet(sheet_subsystems)\n",
    "reactions_rec = sheet.read_google_sheet(sheet_reactions)\n",
    "\n",
    "# Remove the total count\n",
    "subsystems = subsystems.iloc[:-1, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Pie Chart of the Amount of Reactions per System --- ####\n",
    "\n",
    "subsystems[\"Number of Reactions\"] = pd.to_numeric(subsystems[\"Number of Reactions\"])\n",
    "\n",
    "# Aggregate data by \"System\" and sum \"Number of Reactions\"\n",
    "system_reactions = subsystems.groupby(\"System\")[\"Number of Reactions\"].sum()\n",
    "\n",
    "# Set the figure and axes for more control\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "\n",
    "# Generate the pie chart again with the custom labels\n",
    "patches, texts, autotexts = ax.pie(system_reactions, labels=system_reactions.index, autopct=lambda p: '{:.1f}%'.format(p) if p > 0 else '', startangle=140, colors=plt.cm.tab20.colors)\n",
    "\n",
    "# Improve aesthetics\n",
    "for text in texts + autotexts:\n",
    "    text.set_color('black')\n",
    "ax.set_title(\"Pie Chart of Number of Reactions by System\", pad=30)  # Move the title upwards by increasing pad\n",
    "\n",
    "# Set equal aspect ratio\n",
    "ax.axis('equal')\n",
    "\n",
    "# Save the plot\n",
    "plt.savefig('../Plots/pie_chart_reactions_per_system.png')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the Sunburst Plot\n",
    "\n",
    "fig = px.sunburst(subsystems, path=['System', 'Subsystems', 'Kegg Pathway'],\n",
    "                  color='System')\n",
    "\n",
    "# Convert to a Graph Objects figure\n",
    "fig_go = go.Figure(fig)\n",
    "\n",
    "# Update layout\n",
    "fig_go.update_layout(width=1200, height=1000)\n",
    "\n",
    "# Define font sizes\n",
    "font_size_last_layer = 10\n",
    "font_size_previous_layer = 15\n",
    "default_font_size = 12\n",
    "\n",
    "# Create a list to store font sizes\n",
    "font_sizes = []\n",
    "\n",
    "# Update font size for each level\n",
    "for trace in fig_go.data:\n",
    "    if isinstance(trace, go.Sunburst):\n",
    "        for id in trace.ids:\n",
    "            level = id.count(\"/\")  # Determine level by the number of slashes in the id\n",
    "            if level == 2:  # Last layer (Kegg Pathway)\n",
    "                font_sizes.append(font_size_last_layer)\n",
    "            elif level == 1:  # Previous layer (Subsystems)\n",
    "                font_sizes.append(font_size_previous_layer)\n",
    "            else:\n",
    "                font_sizes.append(default_font_size)  # Default size for other layers\n",
    "\n",
    "# Apply the font sizes to the figure\n",
    "fig_go.update_traces(insidetextfont=dict(size=font_sizes))\n",
    "\n",
    "# Save the figure\n",
    "fig_go.write_html(\"../Plots/sunburst_subsystems.html\")  # Save as interactive HTML file\n",
    "fig_go.write_image(\"../Plots/sunburst_subsystems.png\", width=1200, height=1000, scale=2)  # Increase resolution by setting scale parameter\n",
    "\n",
    "# Show the plot\n",
    "fig_go.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a susbsystem dictionary\n",
    "\n",
    "subsystems = subsystems[['System', 'Subsystems']]\n",
    "subsystems_dict = subsystems.groupby('System')['Subsystems'].apply(list).to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Auxotrophies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.io import load_json_model\n",
    "\n",
    "iCHO_path = \"iCHO3595.json\"\n",
    "iCHO = load_json_model(iCHO_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids = {\n",
    "    \"arginine\": [\"EX_arg_L_e\", \"EX_arg_D_e\"],\n",
    "    \"asparagine\": [\"EX_asn_L_e\", \"EX_asn_D_e\"],\n",
    "    \"cysteine\": [\"EX_cys_L_e\", \"EX_cys_D_e\", \"EX_Lcystin_e\"],\n",
    "    \"histidine\": [\"EX_his_L_e\", \"EX_his_D_e\"],\n",
    "    \"isoleucine\": [\"EX_ile_L_e\", \"EX_ile_D_e\"],\n",
    "    \"leucine\": [\"EX_leu_L_e\", \"EX_leu_D_e\"],\n",
    "    \"lysine\": [\"EX_lys_L_e\", \"EX_lys_D_e\"],\n",
    "    \"methionine\": [\"EX_met_L_e\", \"EX_met_D_e\"],\n",
    "    \"phenylalanine\": [\"EX_phe_L_e\", \"EX_phe_D_e\"],\n",
    "    \"proline\": [\"EX_pro_L_e\", \"EX_pro_D_e\"],\n",
    "    \"threonine\": [\"EX_thr_L_e\", \"EX_thr_D_e\"],\n",
    "    \"tryptophan\": [\"EX_trp_L_e\", \"EX_trp_D_e\"],\n",
    "    \"valine\": [\"EX_val_L_e\", \"EX_val_D_e\"]\n",
    "}\n",
    "for amino_acid in amino_acids:\n",
    "    # ----- Setup initial bounds -----\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        exchange_reaction.bounds = -10, 10\n",
    "\n",
    "    # Arginine\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00557').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00558').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GLYAMDTRc').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ARGSL').bounds = 0, 0\n",
    "\n",
    "    # Asparigine\n",
    "    iCHO.reactions.get_by_id('ASNS1').bounds = 0, 0\n",
    "\n",
    "    # Cysteine\n",
    "    iCHO.reactions.get_by_id('r0129').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_cgly_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECG').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECGe').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('CYSTGL').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('EX_HC00250_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_sfcys_e').bounds = 0, 0\n",
    "\n",
    "    # Histidine\n",
    "    iCHO.reactions.get_by_id('VALTRPVALr').bounds = 0,0 \n",
    "\n",
    "    # Isoleucine - DONE\n",
    "    iCHO.reactions.get_by_id('EX_CE2916_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2915_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('ILETA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ILETAm').bounds = 0, 10\n",
    "\n",
    "    # Leucine\n",
    "    iCHO.reactions.get_by_id('LEULEULAPc').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_leugly_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_glyleu_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('LEUTA').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('LEUTAm').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('EX_CE5797_e').bounds = 0, 0\n",
    "\n",
    "    # Lysine\n",
    "    iCHO.reactions.get_by_id('EX_biocyt_e').bounds = 0,0 \n",
    "\n",
    "    # Methionine\n",
    "    iCHO.reactions.get_by_id('METS').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('BHMT').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('GAUGE-R00648').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('UNK2').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('UNK3').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('TYRA').bounds = -10, 0 \n",
    "    #iCHO.reactions.get_by_id('GAUGE-R06895').bounds = 0, 0 # Curated by MR with 1 score / Erased from the reconstruction\n",
    "\n",
    "    # Phenylalanine\n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_pheleu_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_glyphe_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2917_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5789_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_phpyr_e').bounds = 0, 0 \n",
    "\n",
    "    # Proline\n",
    "    iCHO.reactions.get_by_id('EX_glypro_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_progly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('P5CR').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRxm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRx').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('r1453').bounds = 0, 10\n",
    "\n",
    "    # Threonine\n",
    "    #iCHO.reactions.get_by_id('THRS').bounds = 0,0  # Erased from the reconstruction\n",
    "\n",
    "\n",
    "    # Tryptophan\n",
    "    iCHO.reactions.get_by_id('NBAHH_ir').bounds = 0,0  # Histidine hydrolase\n",
    "\n",
    "    # Valine\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_vallystyr_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTAm').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_valval_e').bounds = 0, 10\n",
    "    # -----------------------------------------\n",
    "\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        if exchange_reaction.id in amino_acids[amino_acid]:\n",
    "            exchange_reaction.bounds = 0, 10\n",
    "            # print(amino_acid, exchange_reaction.id, amino_acids[amino_acid])\n",
    "    sol = iCHO.optimize()\n",
    "    print(amino_acid, sol.objective_value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCHO.metabolites.leu_L_c.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Deletion test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cobra.io import load_json_model, read_sbml_model\n",
    "from cobra.flux_analysis import single_gene_deletion\n",
    "\n",
    "#iCHO = load_json_model('iCHO3595_unblocked.json')\n",
    "#iCHO1766 = read_sbml_model('../Data/Reconciliation/models/iCHOv1_final.xml')\n",
    "iCHO2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "#iCHO2441 = read_sbml_model('../Data/Reconciliation/models/iCHO2441.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_deletion_results = single_gene_deletion(iCHO2291)\n",
    "essentiality_threshold = 0.01\n",
    "essential_genes = gene_deletion_results[gene_deletion_results['growth'] < essentiality_threshold]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the essential genes\n",
    "print(f\"Number of essential genes for biomass: {len(essential_genes)}\")\n",
    "print(\"Essential genes for biomass production:\")\n",
    "for gene_idx in essential_genes.index:\n",
    "    gene = iCHO2291.genes[gene_idx]\n",
    "    print(f\"{gene.id}: {gene.name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import cobra\n",
    "import pandas as pd\n",
    "from cobra.flux_analysis.variability import find_essential_genes\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = load_json_model('iCHO3595.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_dict = uptsec_intrvl_zela\n",
    "time = 'P2 to P4'\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "model.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "for rxn in model.boundary:\n",
    "    \n",
    "    # Models that are forced to secrete ethanol are not feasible\n",
    "    if rxn.id == 'EX_etoh_e':\n",
    "        rxn.bounds = (-0.1,0.1)\n",
    "        continue\n",
    "\n",
    "    # Replace the lower and upper bound with experimental data\n",
    "    if rxn.id in temp_dict.keys():\n",
    "        rxn.bounds = temp_dict[rxn.id][time]\n",
    "        continue\n",
    "    \n",
    "    # Keep boundaries open for essential metabolites\n",
    "    if rxn.id == 'EX_h2o_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_h_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_o2_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_hco3_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_so4_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_pi_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "\n",
    "    # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "    if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Tyr_ggn_c':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Ser_Thr_g':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_pre_prot_r':\n",
    "        rxn.bounds = (-0.0006,1000)\n",
    "        continue\n",
    "\n",
    "    # Close uptake rates for the rest of the boundaries\n",
    "    if rxn.id.startswith(\"EX_\"):\n",
    "        rxn.bounds = (0,1000) \n",
    "    if rxn.id.startswith(\"SK_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    if rxn.id.startswith(\"DM_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    \n",
    "pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(pfba_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_essential_genes = find_essential_genes(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(sim_essential_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset for the experimentally validated essential genes\n",
    "\n",
    "exp_essential_genes = pd.read_csv('../Data/Gene_Essentiality/cho_essential_genes.csv', sep='\\t')\n",
    "exp_essential_genes['gene'] = exp_essential_genes['gene'].str.replace('__1$', '', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_essential_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract all gene names\n",
    "all_gene_names = set()\n",
    "for names in exp_essential_genes['gene']:\n",
    "    for name in names.split('|'): #Extract the names of the genes with two aliases\n",
    "        all_gene_names.add(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "essential_genes_model = []\n",
    "for g in model.genes:\n",
    "    if g.name in all_gene_names:\n",
    "        essential_genes_model.append(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(essential_genes_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set1 = set(sim_essential_genes)\n",
    "set2 = set(essential_genes_model)\n",
    "\n",
    "# Find shared elements\n",
    "shared_elements = set1.intersection(set2)\n",
    "\n",
    "# Find unique elements\n",
    "unique_in_list1 = set1.difference(set2)\n",
    "unique_in_list2 = set2.difference(set1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Shared genes between Exp Validated Essential Genes and Simulated Essential Genes:\", len(shared_elements))\n",
    "print(\"Unique genes in Simulated Essential Genes:\", len(unique_in_list1))\n",
    "print(\"Unique genes in Exp Validated Essentail Genes:\", len(unique_in_list2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#iCHO3597 Total essential genes: 451 Sim essential genes: 54\n",
    "#iCHO1766 Total essential genes: 229 Sim essential genes: 34\n",
    "#iCHO2101 Total essential genes: 254 Sim essential genes: 51\n",
    "#iCHO2291 Total essential genes: 340 Sim essential genes: 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Essential for growth genes and comparison to (An optimised genome-wide, virus free CRISPR screen for mammalian cells)\n",
    "https://pubmed.ncbi.nlm.nih.gov/34935002/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Context-specific Model Generation <a id='context_specific'></a>\n",
    "\n",
    "Here we use a matrix generated with rmf_CADRE to generate each context_specific model for each one of the conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from scipy.io import loadmat\n",
    "from time import process_time\n",
    "\n",
    "import pymCADRE\n",
    "from pymCADRE.rank import rank_reactions\n",
    "from pymCADRE.check import check_model_function\n",
    "from pymCADRE.prune import prune_model\n",
    "\n",
    "import cobra\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Intervals for Context-specific model generation\n",
    "Here we define the set of constraints in the uptake and secretion rates that we want to incorporate to our context specific models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Model Loading and Constraining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load generic model\n",
    "model = load_json_model('iCHO3595_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_dict = uptsec_intrvl_zela\n",
    "time = 'P2 to P4'\n",
    "\n",
    "with model as model:\n",
    "    model.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "    for rxn in model.boundary:\n",
    "        \n",
    "        # Models that are forced to secrete ethanol are not feasible\n",
    "        if rxn.id == 'EX_etoh_e':\n",
    "            rxn.bounds = (-0.1,0.1)\n",
    "            continue\n",
    "\n",
    "        # Replace the lower and upper bound with experimental data\n",
    "        if rxn.id in temp_dict.keys():\n",
    "            rxn.bounds = temp_dict[rxn.id][time]\n",
    "            continue\n",
    "        \n",
    "        # Keep boundaries open for essential metabolites\n",
    "        if rxn.id == 'EX_h2o_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_h_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_o2_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_hco3_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_so4_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_pi_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "\n",
    "        # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "        if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Tyr_ggn_c':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Ser_Thr_g':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_pre_prot_r':\n",
    "            rxn.bounds = (-0.0006,1000)\n",
    "            continue\n",
    "\n",
    "        # Close uptake rates for the rest of the boundaries\n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        \n",
    "    pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "    print(pfba_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading Ubiquity and Confidence Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the .mat file\n",
    "mat_data = loadmat('../Data/Context_specific_models/ubiData.mat')\n",
    "\n",
    "# Access the ubiData structure\n",
    "ubiData = mat_data['ubiData']\n",
    "\n",
    "# Accessing individual fields within ubiData\n",
    "ubiScores = ubiData['ubiScores'][0, 0]\n",
    "uScore = ubiData['uScore'][0, 0]\n",
    "rxns = ubiData['rxns'][0, 0]\n",
    "Condition = ubiData['Condition'][0, 0]\n",
    "\n",
    "# Print or process the data as needed\n",
    "print(ubiScores.shape)\n",
    "print(uScore.shape)\n",
    "print(rxns.shape)\n",
    "print(Condition.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,cond in enumerate(Condition[0]):\n",
    "    if \"WT_P4\" in cond[0]:\n",
    "        print(cond[0], i)\n",
    "    elif \"ZeLa_P4\" in cond[0]:\n",
    "        print(cond[0], i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Condition[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ubiquity score list\n",
    "U = list(ubiScores[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conf Score List\n",
    "conf_scores = pd.read_csv('../Data/Context_specific_models/confidence_scores.csv', header=None).iloc[:, 0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Protected Reations\n",
    "high_confidence_reactions = ['biomass_cho_s', 'DNAsyn', 'LipidSyn_cho_s', 'PROTsyn_cho_s', 'RNAsyn_cho_s', 'EX_bhb_e', 'EX_nh4_e', 'EX_ac_e', 'EX_ala_L_e',\n",
    "    'EX_arg_L_e', 'EX_asn_L_e', 'EX_asp_L_e', 'EX_2hb_e', 'EX_cit_e', 'EX_cys_L_e', 'EX_etoh_e', 'EX_for_e', 'EX_fum_e', 'EX_glc_e', 'EX_glu_L_e',\n",
    "    'EX_gln_L_e', 'EX_gly_e', 'EX_his_L_e', 'EX_4hpro_e', 'EX_ile_L_e', 'EX_lac_L_e', 'EX_leu_L_e', 'EX_lys_L_e', 'EX_mal_L_e', 'EX_met_L_e',\n",
    "    'EX_phe_L_e', 'EX_pro_L_e', 'EX_5oxpro_e', 'EX_pyr_e', 'EX_ser_L_e', 'EX_thr_L_e', 'EX_trp_L_e', 'EX_tyr_L_e', 'EX_val_L_e', 'EX_h2o_e',\n",
    "    'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e', 'SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reduced model generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################\n",
    "# Rank reactions\n",
    "##############################################\n",
    "\n",
    "print('Processing inputs and ranking reactions...')\n",
    "GM, C, NC, P, Z, model_C = rank_reactions(model, U, conf_scores, high_confidence_reactions, method=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output the results\n",
    "print(\"Generic Model:\", GM)\n",
    "print(\"Core Reactions:\", len(C))\n",
    "print(\"Non-Core Reactions:\", len(NC))\n",
    "print(\"Ranked Non-Core Reactions:\", len(P))\n",
    "print(\"Zero-Expression Reactions:\", len(Z))\n",
    "print(\"Core Reactions in Original Model:\", len(model_C))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.util import create_stoichiometric_matrix\n",
    "\n",
    "def retrieve_precursor_metabolites(model, protected_rxns):\n",
    "    \"\"\"\n",
    "    Retrieve precursor metabolites from a specified list of reactions in a metabolic model.\n",
    "\n",
    "    Args:\n",
    "        model: cobra.Model - The metabolic model.\n",
    "        protected_rxns: list - A list of reaction IDs for which precursor metabolites need to be retrieved.\n",
    "\n",
    "    Returns:\n",
    "        precursor_mets: list - A list of precursor metabolites.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a list of all the reactions in the model\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Ensure that the protected reactions list is not empty\n",
    "    if protected_rxns:\n",
    "        \n",
    "        #Define a list to store the indices of the protected reactions\n",
    "        rxn_indices = []\n",
    "        \n",
    "        for rxn in protected_rxns:\n",
    "            # Ensure that the protected reaction is included in the model\n",
    "            if rxn in model_reactions:\n",
    "                # Get the indices of the protected reactions in the model, only if the reaction exists\n",
    "                rxn_indices.append(model.reactions.index(model.reactions.get_by_id(rxn)))\n",
    "\n",
    "        # Create stoichiometric matrix S\n",
    "        S = create_stoichiometric_matrix(model)\n",
    "        \n",
    "        # Retrieve precursor metabolites: metabolites with a negative coefficient in the stoichiometric matrix\n",
    "        precursor_mets = [\n",
    "            model.metabolites[i] \n",
    "            for i in range(len(model.metabolites)) \n",
    "            if any(S[i, idx] < 0 for idx in rxn_indices)\n",
    "        ]\n",
    "\n",
    "        return precursor_mets\n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precursor_mets = retrieve_precursor_metabolites(model, high_confidence_reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# Define inputs to the model pruning step\n",
    "##################################################\n",
    "\n",
    "# Define core vs. non-core ratio threshold for removing reactions\n",
    "eta = 1 / 3\n",
    "\n",
    "# Check functionality of generic model\n",
    "genericStatus = check_model_function(GM, 'required_mets', precursor_mets)[0]\n",
    "print(genericStatus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constraint the model with exp data\n",
    "\n",
    "time = 'P2 to P4'\n",
    "\n",
    "GM.solver = 'gurobi'\n",
    "objective = 'biomass_cho_s' # 'biomass_cho'\n",
    "GM.objective = objective\n",
    "\n",
    "GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "for rxn in GM.boundary:\n",
    "    \n",
    "    # Models that are forced to secrete ethanol are not feasible\n",
    "    if rxn.id == 'EX_etoh_e':\n",
    "        rxn.bounds = (-0.1,0.1)\n",
    "        continue\n",
    "\n",
    "    # Replace the lower and upper bound with experimental data\n",
    "    if rxn.id in temp_dict.keys():\n",
    "        rxn.bounds = temp_dict[rxn.id][time]\n",
    "        continue\n",
    "    \n",
    "    # Keep boundaries open for essential metabolites\n",
    "    if rxn.id == 'EX_h2o_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_h_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_o2_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_hco3_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_so4_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_pi_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "\n",
    "    # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "    if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Tyr_ggn_c':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Ser_Thr_g':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_pre_prot_r':\n",
    "        rxn.bounds = (-0.0006,1000)\n",
    "        continue\n",
    "\n",
    "    # Close uptake rates for the rest of the boundaries\n",
    "    if rxn.id.startswith(\"EX_\"):\n",
    "        rxn.bounds = (0,1000) \n",
    "    if rxn.id.startswith(\"SK_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    if rxn.id.startswith(\"DM_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    \n",
    "pfba_solution = cobra.flux_analysis.pfba(GM)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(pfba_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if genericStatus:\n",
    "    print('Generic model passed precursor metabolites test')\n",
    "\n",
    "    ##############################################################\n",
    "    # If generic functionality test is passed, prune reactions\n",
    "    ###############################################################\n",
    "    print('Pruning reactions...')\n",
    "    t0 = process_time()\n",
    "    PM, cRes = prune_model(GM, P, C, Z, eta, precursor_mets, salvage_check=0, method=1)\n",
    "    # Stop the stopwatch / counter\n",
    "    t_stop = process_time()\n",
    "    # compute elapsed time\n",
    "    pruneTime = t_stop - t0\n",
    "\n",
    "else:\n",
    "    print('Generic model failed precursor metabolites test!!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ATP Generation loops removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import chain\n",
    "import pickle\n",
    "import numpy as np\n",
    "from optlang.symbolics import Zero\n",
    "import pandas as pd\n",
    "\n",
    "import cobra\n",
    "from cobra.util import create_stoichiometric_matrix\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading and Constraining Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = load_json_model('iCHO3595_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ATP loops removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# First, define your list of reaction IDs\n",
    "atp_loop_reactions = [\n",
    "    'SCP22x','TMNDNCCOAtx','OCCOAtx','r0391','BiGGRxn67','r2247','r2280',\n",
    "    'r2246','r2279','r2245','r2305','r2317','r2335','HMR_0293','HMR_7741',\n",
    "    'r0509','r1453','HMR_4343','ACONTm','PDHm','r0426','r0383','r0555',\n",
    "    'r1393','NICRNS','GAUGE-R00648','GAUGE-R03326','GapFill-R08726','RE2915M',\n",
    "    'HMR_3288','HMR_1325','HMR_7599','r1431','r1433','RE2439C','r0082','r0791',\n",
    "    'r1450','GAUGE-R00270','GAUGE-R02285','GAUGE-R04283','GAUGE-R06127','GAUGE-R06128',\n",
    "    'GAUGE-R06238','GAUGE-R00524','RE3477C','AAPSAS','RE3347C','HMR_0960','HMR_0980',\n",
    "    'RE3476C','r0708','r0777','r0084','r0424','r0698','3HDH260p','HMR_3272','ACOAD183n3m',\n",
    "    'HMR_1996','GapFill-R01463','GapFill-R04807','r1468','r2435','r0655','r0603','r0541',\n",
    "    'RE0383C','HMR_1329','TYRA','NRPPHRt_2H','GAUGE-R07364','GapFill-R03599','ARD',\n",
    "    'RE3095C','RE3104C','RE3104R','ACONT','ICDHxm','ICDHy','AKGDm','r0083',\n",
    "    'r0425','r0556','NH4t4r','PROPAT4te','r0085','r0156','r0464','ABUTDm',\n",
    "    'OIVD1m','OIVD2m','OIVD3m','r2194','r2202','HMR_9617','r2197','r2195',\n",
    "    '2OXOADOXm','r2328','r0386','r0451','FAS100COA','FAS120COA','FAS140COA',\n",
    "    'FAS80COA_L','r0604','r0670','r2334','r0193','r0595','r0795','GLYCLm',\n",
    "    'MACACI','r2193','r0779','r0669','UDCHOLt','r2146','r2139'\n",
    "]\n",
    "\n",
    "# Function to knock out reactions in the model\n",
    "def knock_out_reactions(model, reactions_to_knockout):\n",
    "    for rxn_id in reactions_to_knockout:\n",
    "        try:\n",
    "            rxn = model.reactions.get_by_id(rxn_id)\n",
    "            rxn.knock_out()\n",
    "        except KeyError:\n",
    "            print(f\"Reaction {rxn_id} not found in the model.\")\n",
    "\n",
    "# Set up the boundary conditions as per your code\n",
    "def set_boundary_conditions(model, exp_boundaries, batch, timepoint):\n",
    "    for rxn in model.boundary:\n",
    "        if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "            rxn.bounds = (-1000, 1000)\n",
    "        elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "            rxn.bounds = (-0.001, 1000)\n",
    "        elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "            rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "    \n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "            continue\n",
    "        for r,v in uptsec_zela.items():\n",
    "            if reaction.id == r:\n",
    "                reaction.upper_bound = 1000\n",
    "                reaction.lower_bound = v[(batch, timepoint)]\n",
    "\n",
    "# Function to run pFBA and return ATPM and growth rate\n",
    "def run_pfba(model, objective_id='biomass_cho_s'):\n",
    "    model.objective = objective_id\n",
    "    solution = cobra.flux_analysis.pfba(model)\n",
    "    growth_rate = solution.fluxes.get(objective_id, None)\n",
    "    return growth_rate\n",
    "\n",
    "# Function to find the ATPM lower bound where growth rate changes\n",
    "def find_atpm_threshold(model):\n",
    "    # Start from 0 and incrementally increase ATPM lower bound\n",
    "    for atpm_bound in range(30, -1, -1):\n",
    "        try:\n",
    "            with model as test_model:\n",
    "                baseline_growth_rate = run_pfba(test_model)\n",
    "                test_model.reactions.get_by_id('ATPM').lower_bound = atpm_bound\n",
    "                growth_rate = run_pfba(test_model)\n",
    "                \n",
    "                # Check if growth rate changes compared to the baseline\n",
    "                if round(growth_rate, 6) != round(baseline_growth_rate, 6):\n",
    "                    print(f\"Growth rate altered at ATPM lower bound: {atpm_bound}\")\n",
    "                    return atpm_bound\n",
    "        except Exception as e:\n",
    "            print(f\"Infeasible solution for bound: {atpm_bound}, {e}\")\n",
    "    return 0  # If no change, default to 0\n",
    "\n",
    "\n",
    "# Initialize an empty DataFrame to store results\n",
    "results_df = pd.DataFrame(columns=['Batch', 'Timepoint', 'Reaction', 'Growth Rate', 'Growth Changed'])\n",
    "\n",
    "timepoints = ['P0 to P2','P2 to P4','P4 to P6','P6 to P8']\n",
    "batches = ['U4','U5','U6','U7','U8']\n",
    "\n",
    "for timepoint in timepoints:\n",
    "    for batch in batches:\n",
    "        print(f'Analising batch {batch} at timepoint {timepoint}')\n",
    "        try:\n",
    "            # Establish baseline with all reactions knocked out\n",
    "            with model as base_model:\n",
    "                set_boundary_conditions(base_model, uptsec_zela, batch, timepoint)\n",
    "                knock_out_reactions(base_model, atp_loop_reactions)\n",
    "    \n",
    "                # Find the ATPM lower bound where the growth rate changes\n",
    "                atpm_threshold = find_atpm_threshold(base_model)\n",
    "                \n",
    "                # Add ATP Maintenance Cost\n",
    "                base_model.reactions.get_by_id('ATPM').lower_bound = atpm_threshold\n",
    "                \n",
    "                # Run pFBA\n",
    "                base_growth_rate = run_pfba(base_model)\n",
    "                print(f'Baseline Growth Rate: {base_growth_rate}')\n",
    "                \n",
    "            # Now, iteratively \"knock in\" each reaction and test\n",
    "            results = []\n",
    "            for rxn_id in atp_loop_reactions:\n",
    "                with model as test_model:\n",
    "                    set_boundary_conditions(test_model, uptsec_zela, batch, timepoint)\n",
    "                    # Knock out all reactions except the current one\n",
    "                    reactions_to_knockout = [rid for rid in atp_loop_reactions if rid != rxn_id]\n",
    "                    knock_out_reactions(test_model, reactions_to_knockout)\n",
    "                    \n",
    "                    # Add ATP Maintenance Cost\n",
    "                    test_model.reactions.get_by_id('ATPM').lower_bound = atpm_threshold\n",
    "            \n",
    "                    # Run pFBA\n",
    "                    growth_rate = run_pfba(test_model)\n",
    "                    \n",
    "                    # Check if there's a change compared to the baseline\n",
    "                    growth_changed = round(growth_rate,6) != round(base_growth_rate,6)\n",
    "            \n",
    "                    # Create a new row and concatenate with the existing DataFrame\n",
    "                    new_row = pd.DataFrame([{\n",
    "                        'Batch': batch,\n",
    "                        'Timepoint': timepoint,\n",
    "                        'Reaction': rxn_id,\n",
    "                        'Growth Rate': growth_rate,\n",
    "                        'Growth Changed': growth_changed\n",
    "                    }])\n",
    "                    \n",
    "                    results_df = pd.concat([results_df, new_row], ignore_index=True)\n",
    "                    \n",
    "                    print(f'Tested reaction: {rxn_id}')\n",
    "                    print(f'Growth Rate: {growth_rate}')\n",
    "                    print(f'Growth Changed: {growth_changed}')\n",
    "                    print('---')\n",
    "        except Exception as e:\n",
    "            print(f\"Error with batch {batch} and timepoint {timepoint}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Group by Reaction and check if all Growth Changed values are False\n",
    "consistent_non_affecting_reactions = results_df.groupby('Reaction')['Growth Changed'].apply(lambda x: (x == False).all())\n",
    "# Filter reactions that are consistently False across all batches and timepoints\n",
    "reactions_never_affecting_growth = consistent_non_affecting_reactions[consistent_non_affecting_reactions].index.tolist()\n",
    "# Removing the specified reactions\n",
    "filtered_atp_loop_reactions = [rxn for rxn in atp_loop_reactions if rxn not in reactions_never_affecting_growth]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_atp_loop_reactions = [\n",
    "    'SCP22x','TMNDNCCOAtx','OCCOAtx','r0391','BiGGRxn67','r2247','r2246','r2245','r2317','HMR_0293',\n",
    "    'HMR_7741','r1453','r1393','NICRNS','GapFill-R08726','RE2915M','HMR_3288','HMR_1325','RE2439C',\n",
    " 'r1450','RE3477C','AAPSAS','r0698','3HDH260p','HMR_3272','ACOAD183n3m','GapFill-R01463','r1468',\n",
    " 'r0655','r0603','r0541','HMR_1329','GapFill-R03599','AKGDm','r0556','OIVD1m','OIVD3m','2OXOADOXm',\n",
    " 'r0386','r0451','GLYCLm','MACACI'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#########################################\n",
    "# Pablo's modification\n",
    "#########################################\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "def get_atp_related_reactions(model, atp_mets):\n",
    "    \"\"\" Get reactions associated with ATP and related metabolites like GTP, GDP, etc. \"\"\"\n",
    "    atp_rxns_indices = set()  # Use set to avoid duplicates\n",
    "    for met in atp_mets:\n",
    "        for reac in met.reactions:\n",
    "            if any(x in reac.reactants + reac.products for x in atp_mets):\n",
    "                atp_rxns_indices.add(reac.id)\n",
    "    return list(atp_rxns_indices)\n",
    "\n",
    "def find_low_flux_reactions(model, atp_rxns_indices, atp_met_indices, stoich_matrix, consumption_cutoff=2e-12):\n",
    "    \"\"\" Identify reactions with low ATP or related metabolite consumption flux. \"\"\"\n",
    "    atp_consumption_idx = []\n",
    "    for atp_rxn in atp_rxns_indices:\n",
    "        reaction = model.reactions.get_by_id(atp_rxn)\n",
    "        for met in atp_met_indices:\n",
    "            # Check for both ATP and GTP-related consumption (stoichiometric coefficient < 0)\n",
    "            if stoich_matrix[model.metabolites.index(met), model.reactions.index(reaction)] < 0:\n",
    "                atp_consumption_idx.append(atp_rxn)\n",
    "                break\n",
    "    return atp_consumption_idx\n",
    "\n",
    "def perform_fba_for_max_atp(model, oxpho_rxns):\n",
    "    \"\"\" Set FBA objective to maximize ATP production. \"\"\"\n",
    "    reaction_variables = ((rxn.forward_variable, rxn.reverse_variable) for rxn in oxpho_rxns)\n",
    "    variables = chain(*reaction_variables)\n",
    "    model.objective = model.problem.Objective(Zero, direction=\"max\", sloppy=True, name=\"maxATP\")\n",
    "    model.objective.set_linear_coefficients({v: 1.0 for v in variables})\n",
    "    return model.optimize()\n",
    "\n",
    "def remove_cyclic_atp_reactions(model, atp_rxns_to_remove, biomass_rxn):\n",
    "    \"\"\" Remove reactions consuming ATP in a cyclic fashion and check for safety. \"\"\"\n",
    "    \n",
    "    safe_rxn_irrev = []\n",
    "\n",
    "    # Copy the original model\n",
    "    temp_model = model.copy()\n",
    "\n",
    "    # Set the objective for the temporary model\n",
    "    temp_model.objective = biomass_rxn\n",
    "    sol_bio_control = temp_model.optimize()\n",
    "\n",
    "    # Cumulative list of safe reactions to remove or modify\n",
    "    candidate_safe_rmv = []\n",
    "    candidate_safe_irrev = []\n",
    "\n",
    "    for rxn_id in tqdm(atp_rxns_to_remove, desc=\"Processing reactions\"):\n",
    "        with temp_model:\n",
    "            # Try to make the reaction irreversible\n",
    "            if temp_model.reactions.get_by_id(rxn_id).lower_bound != 0:\n",
    "                temp_model.reactions.get_by_id(rxn_id).bounds = (0, 1000)\n",
    "                sol_bio_temp = temp_model.optimize()\n",
    "                \n",
    "                if (round(sol_bio_temp.objective_value, 5) == round(sol_bio_control.objective_value, 5)):\n",
    "                    candidate_safe_irrev.append(rxn_id)\n",
    "                    \n",
    "                    # Test cumulative effect of all candidate modifications\n",
    "                    for rxn in candidate_safe_irrev:\n",
    "                        temp_model.reactions.get_by_id(rxn).bounds = (0, 1000)\n",
    "                    sol_bio_cumulative = temp_model.optimize()\n",
    "\n",
    "                    # If the cumulative effect is still safe, add to the final modification list\n",
    "                    if round(sol_bio_cumulative.objective_value, 5) == round(sol_bio_control.objective_value, 5):\n",
    "                        safe_rxn_irrev = list(candidate_safe_irrev)\n",
    "                    else:\n",
    "                        # If not safe, undo the last added irreversible modification\n",
    "                        candidate_safe_irrev.remove(rxn_id)\n",
    "                        print(f\"Making {rxn_id} irreversible is not safe, reverting.\")\n",
    "\n",
    "    # Make irreversible reactions that can be safely modified\n",
    "    for rxn in safe_rxn_irrev:\n",
    "        if model.reactions.get_by_id(rxn).lower_bound != 0:\n",
    "            model.reactions.get_by_id(rxn).bounds = (0, 1000)\n",
    "    \n",
    "    return model, safe_rxn_irrev\n",
    "\n",
    "# Set objectives\n",
    "biomass_rxn = 'biomass_cho_s'\n",
    "model.objective = biomass_rxn\n",
    "biomass_control = model.optimize('maximize')\n",
    "\n",
    "# Find ATP related reactions\n",
    "atp_related_mets = ['atp_c', 'atp_n', 'atp_m', 'atp_x', 'atp_l', 'atp_g', 'atp_r', 'atp_e', 'datp_c', 'dadp_c', 'gtp_c', 'gdp_c',\n",
    "                    'utp_c', 'udp_c', 'ctp_c', 'cdp_c', 'ump_c', 'mlthf_c', 'mlthf_m', 'dgtp_c', 'dgdp_c', 'adxrd_c', 'adxrd_m',\n",
    "                    'adxo_2_2_m', 'adxo_2_2_c', 'udpg_c', 'udpgal_c', 'succ_m', 'succoa_m', '4met2obut_c', 'q10h2_m', 'q10_m']\n",
    "atp_met_indices = [model.metabolites.get_by_id(met) for met in atp_related_mets]\n",
    "atp_rxns_indices = get_atp_related_reactions(model, atp_met_indices)\n",
    "\n",
    "# Create stoichiometric matrix\n",
    "stoich_matrix = create_stoichiometric_matrix(model)\n",
    "atp_consumption_idx = find_low_flux_reactions(model, atp_rxns_indices, atp_met_indices, stoich_matrix)\n",
    "\n",
    "# Maximize ATP production\n",
    "oxpho_rxns = [r for r in model.reactions if 'OXIDATIVE PHOSPHORYLATION' in r.subsystem]\n",
    "solution_maxATP = perform_fba_for_max_atp(model, oxpho_rxns)\n",
    "\n",
    "# Check flux of ATP consuming reactions\n",
    "atp_loop_rxn_rmv = []\n",
    "\n",
    "model.objective = biomass_rxn\n",
    "sol = model.optimize('maximize')\n",
    "atp_rate = np.vstack([sol[atp_consumption_idx], solution_maxATP[atp_consumption_idx]])\n",
    "\n",
    "# Find low flux ATP reactions\n",
    "atp_rxn_lowFlux = [rxn_id for rxn_id, rate in zip(atp_consumption_idx, np.sum(atp_rate, axis=0)) if rate <= 2e-12]\n",
    "\n",
    "# Filter protected subsystems and exchanges\n",
    "protected_subsystems = ['BIOMASS', 'OXIDATIVE PHOSPHORYLATION']\n",
    "atp_loop_rxn_rmv += [rxn for rxn in atp_rxn_lowFlux if model.reactions.get_by_id(rxn).subsystem not in protected_subsystems]\n",
    "\n",
    "# Ensure ATPM and exchange reactions are not removed\n",
    "atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id != 'ATPM' and not r_id.startswith(\"EX_\")]\n",
    "\n",
    "# Remove duplicates in the list\n",
    "atp_loop_rxn_rmv = set(atp_loop_rxn_rmv)\n",
    "\n",
    "# Remove cyclic ATP reactions and check safety\n",
    "pruned_model, safe_rxn_irrev = remove_cyclic_atp_reactions(model, atp_loop_rxn_rmv, biomass_rxn)\n",
    "\n",
    "print(f\"Reactions consuming ATP in a cyclic fashion and safe to make irreversible: {len(safe_rxn_irrev)}\")\n",
    "\n",
    "# Final biomass\n",
    "pruned_model.objective = biomass_rxn\n",
    "sol_bio = pruned_model.optimize('maximize')\n",
    "\n",
    "print(f\"Old Bio: {biomass_control.objective_value:.3f}, New Bio: {sol_bio.objective_value:.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. tSNE Comparison of Models <a id='tSNE'></a>\n",
    "Here we visualize the relationships across the reaction structures of individual cell lines to reveal of similar cell types group together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from scipy.stats import mstats\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.metrics import pairwise_distances\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "from collections import defaultdict\n",
    "from cobra.io import load_matlab_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions for model comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_multiple_cobra_models(models, print_results=False, plot_results=False, group_vector=None, func_compare=False, task_file=None, subsystem_mapping=None,\n",
    "                                  subsystem_filter=None, condition_filter=None, save_path=None, cluster_rows=False, cluster_cols=False):\n",
    "    \"\"\"\n",
    "    Compares two or more condition-specific COBRA models in terms of reactions and subsystems.\n",
    "    \n",
    "    Parameters:\n",
    "        models (list): List of `cobra.Model` objects to compare.\n",
    "        print_results (bool): If True, prints the results. Default is False.\n",
    "        plot_results (bool): If True, generates plots. Default is False.\n",
    "        group_vector (list): List or array that groups similar models. Default is None.\n",
    "        func_compare (bool): If True, performs functional comparison. Default is False.\n",
    "        task_file (str): Name of the task file for functional comparison. Default is None.\n",
    "        subsystem_mapping (dict): Mapping from subsystem names to grouped system names.\n",
    "    \n",
    "    Returns:\n",
    "        comp_struct (dict): A dictionary containing the comparison results.\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if there are at least two models\n",
    "    if len(models) <= 1:\n",
    "        raise ValueError(\"Cannot compare only one model. Provide at least two models.\")\n",
    "\n",
    "    # Initialize comp_struct to store the comparison results\n",
    "    comp_struct = {\n",
    "        'modelIDs': [],\n",
    "        'reactions': {},\n",
    "        'subsystems': {},\n",
    "        'structComp': None,\n",
    "        'structCompMap': None,\n",
    "        'funcComp': None\n",
    "    }\n",
    "\n",
    "    # Get model IDs\n",
    "    comp_struct['modelIDs'] = [model.id for model in models]\n",
    "\n",
    "    # Assign labels based on the model names\n",
    "    labels = assign_labels_based_on_name(comp_struct['modelIDs'])\n",
    "\n",
    "    # Group models by WT and ZeLa\n",
    "    wt_models = [model for model in models if 'WT' in model.id]\n",
    "    zela_models = [model for model in models if 'ZeLa' in model.id]\n",
    "\n",
    "    # Compare subsystems with grouping\n",
    "    print(\"\\nComparing subsystem utilization\")\n",
    "    subsystems, subsystem_matrix = compare_cobra_field(models, 'subsystem', normalize=True, subsystem_mapping=subsystem_mapping)\n",
    "    comp_struct['subsystems'] = {'ID': subsystems, 'matrix': subsystem_matrix}\n",
    "\n",
    "    if print_results:\n",
    "        print(\"\\nComparison of grouped subsystem populations:\")\n",
    "        df = pd.DataFrame(subsystem_matrix, index=subsystems, columns=comp_struct['modelIDs'])\n",
    "        print(df.head(15))\n",
    "\n",
    "    if plot_results:\n",
    "        # Compute the average usage of subsystems for WT and ZeLa groups\n",
    "        wt_indices = [i for i, model_id in enumerate(comp_struct['modelIDs']) if 'WT' in model_id]\n",
    "        zela_indices = [i for i, model_id in enumerate(comp_struct['modelIDs']) if 'ZeLa' in model_id]\n",
    "\n",
    "        wt_avg = np.mean(subsystem_matrix[:, wt_indices], axis=1)\n",
    "        zela_avg = np.mean(subsystem_matrix[:, zela_indices], axis=1)\n",
    "\n",
    "        # Calculate the absolute differences between WT and ZeLa\n",
    "        differences = np.abs(wt_avg - zela_avg)\n",
    "\n",
    "        # Sort subsystems by the biggest differences\n",
    "        sorted_indices = np.argsort(differences)[::-1]  # Sort in descending order of differences\n",
    "        sorted_subsystems = [subsystems[i] for i in sorted_indices]\n",
    "        sorted_matrix = subsystem_matrix[sorted_indices, :]\n",
    "\n",
    "        # Plot heatmap for the sorted matrix\n",
    "        plot_heatmap(sorted_matrix, sorted_subsystems, comp_struct['modelIDs'], \"Systems Differences: WT vs ZeLa\", save_path=save_path, subsystem_filter=subsystem_filter, condition_filter=condition_filter,\n",
    "                     cluster_rows=cluster_rows, cluster_cols=cluster_cols)\n",
    "\n",
    "    # Compare model structures\n",
    "    print(\"\\nComparing model structures\")\n",
    "    reactions, reaction_matrix = compare_cobra_field(models, 'reaction')\n",
    "    comp_struct['reactions'] = {'IDs': reactions, 'matrix': reaction_matrix}\n",
    "\n",
    "    # t-SNE projection for structure comparison\n",
    "    n_models = reaction_matrix.T.shape[0]  # Get the number of models\n",
    "\n",
    "    # Set perplexity to be less than the number of models\n",
    "    perplexity = min(n_models - 1, 2)  # Ensure that perplexity is valid\n",
    "\n",
    "    tsne = TSNE(metric='hamming', n_components=2, perplexity=perplexity, random_state=42)\n",
    "    tsne_projection = tsne.fit_transform(reaction_matrix.T)\n",
    "    comp_struct['structCompMap'] = tsne_projection\n",
    "\n",
    "    if plot_results:\n",
    "        plot_tsne(tsne_projection, labels=labels, model_ids=comp_struct['modelIDs'], title=\"Network Topology t-SNE Projection\", save_path=save_path)\n",
    "\n",
    "    # Functional comparison (if required)\n",
    "    if func_compare and task_file:\n",
    "        print(\"\\nChecking model performance on specified tasks.\")\n",
    "        task_report = perform_functional_comparison(models, task_file)\n",
    "        comp_struct['funcComp'] = task_report\n",
    "        if plot_results:\n",
    "            task_matrix = np.array([report['ok'] for report in task_report])\n",
    "            plot_heatmap(task_matrix, comp_struct['modelIDs'], [task['description'] for task in task_report[0]['tasks']], \"Functional Comparison\")\n",
    "\n",
    "    return comp_struct\n",
    "\n",
    "\n",
    "def compare_cobra_field(models, field, normalize=False, subsystem_mapping=None):\n",
    "    \"\"\"\n",
    "    Compares COBRA models by a specified field (e.g., reactions, subsystems) and returns\n",
    "    a matrix indicating the number of reactions per field (e.g., subsystems) for each model.\n",
    "\n",
    "    Parameters:\n",
    "        models (list): List of `cobra.Model` objects.\n",
    "        field (str): The field to compare ('reaction', 'subsystem').\n",
    "        normalize (bool): If True and field is 'subsystem', applies Z-score normalization.\n",
    "        subsystem_mapping (dict): Mapping from subsystem names to grouped system names.\n",
    "\n",
    "    Returns:\n",
    "        id_list (list): List of unique field entries (reactions or subsystems).\n",
    "        comp_matrix (numpy array): Comparison matrix for the field.\n",
    "    \"\"\"\n",
    "    if field == 'reaction':\n",
    "        # For reactions, collect all unique reaction IDs\n",
    "        field_entries = [set(rxn.id for rxn in model.reactions) for model in models]\n",
    "    elif field == 'subsystem':\n",
    "        # For subsystems, collect all unique subsystems\n",
    "        if subsystem_mapping is not None:\n",
    "            # Reverse the mapping to get a mapping from subsystem names to group names\n",
    "            reverse_mapping = {}\n",
    "            for group, subsystems in subsystem_mapping.items():\n",
    "                for subsystem in subsystems:\n",
    "                    reverse_mapping[subsystem.upper()] = group  # Uppercase for case-insensitive matching\n",
    "            field_entries = []\n",
    "            for model in models:\n",
    "                mapped_subsystems = set()\n",
    "                for rxn in model.reactions:\n",
    "                    if rxn.subsystem:\n",
    "                        subsystem_name = rxn.subsystem.upper()\n",
    "                        group = reverse_mapping.get(subsystem_name)\n",
    "                        if group:\n",
    "                            mapped_subsystems.add(group)\n",
    "                field_entries.append(mapped_subsystems)\n",
    "        else:\n",
    "            field_entries = [set(rxn.subsystem for rxn in model.reactions if rxn.subsystem) for model in models]\n",
    "    else:\n",
    "        raise ValueError(\"Invalid field. Supported fields are 'reaction' and 'subsystem'.\")\n",
    "\n",
    "    # Create a sorted list of all unique field entries\n",
    "    id_list = sorted(set.union(*field_entries))\n",
    "\n",
    "    # Initialize a matrix to store the number of reactions per field (subsystem or reaction)\n",
    "    comp_matrix = np.zeros((len(id_list), len(models)), dtype=int)\n",
    "\n",
    "    # Fill the comparison matrix\n",
    "    for i, model in enumerate(models):\n",
    "        if field == 'reaction':\n",
    "            # Create a set of reaction IDs in the model\n",
    "            reaction_ids = set(rxn.id for rxn in model.reactions)\n",
    "            for j, entry in enumerate(id_list):\n",
    "                if entry in reaction_ids:\n",
    "                    comp_matrix[j, i] = 1  # Presence of the reaction\n",
    "        elif field == 'subsystem':\n",
    "            # Create a dictionary to count reactions per subsystem\n",
    "            subsystem_counts = {subsystem: 0 for subsystem in id_list}\n",
    "            for rxn in model.reactions:\n",
    "                if rxn.subsystem:\n",
    "                    if subsystem_mapping is not None:\n",
    "                        subsystem_name = rxn.subsystem.upper()\n",
    "                        group = reverse_mapping.get(subsystem_name)\n",
    "                        if group:\n",
    "                            subsystem_counts[group] += 1\n",
    "                    else:\n",
    "                        subsystem_counts[rxn.subsystem] += 1\n",
    "            # Update the matrix with subsystem counts\n",
    "            for j, entry in enumerate(id_list):\n",
    "                comp_matrix[j, i] = subsystem_counts.get(entry, 0)\n",
    "\n",
    "    # Apply Z-score normalization only for subsystems if requested\n",
    "    if field == 'subsystem' and normalize:\n",
    "        comp_matrix = z_score_normalize(comp_matrix)\n",
    "\n",
    "    return id_list, comp_matrix\n",
    "\n",
    "\n",
    "def z_score_normalize(matrix):\n",
    "    \"\"\"\n",
    "    Applies Z-score normalization to each row of the matrix (i.e., for each subsystem).\n",
    "\n",
    "    Parameters:\n",
    "        matrix (numpy array): The comparison matrix to normalize.\n",
    "\n",
    "    Returns:\n",
    "        normalized_matrix (numpy array): Z-score normalized matrix.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Compute Z-scores along the rows (axis=1)\n",
    "    mean_values = np.mean(matrix, axis=1, keepdims=True)\n",
    "    std_values = np.std(matrix, axis=1, keepdims=True)\n",
    "\n",
    "    # Avoid division by zero in case any row has zero standard deviation\n",
    "    std_values[std_values == 0] = 1\n",
    "\n",
    "    normalized_matrix = (matrix - mean_values) / std_values\n",
    "    return normalized_matrix\n",
    "\n",
    "\n",
    "def plot_heatmap(data, row_labels, col_labels, title, vmin=-2, vmax=2, save_path=None, \n",
    "                           subsystem_filter=None, condition_filter=None, method='average', metric='euclidean',\n",
    "                           standardize=False, cluster_rows=True, cluster_cols=False):\n",
    "    \"\"\"\n",
    "    Generates a clustered heatmap for the given data, optionally filtering by subsystems and conditions.\n",
    "\n",
    "    Parameters:\n",
    "        data (numpy array): Data matrix to plot.\n",
    "        row_labels (list): Labels for rows (subsystems).\n",
    "        col_labels (list): Labels for columns (conditions).\n",
    "        title (str): Title of the plot.\n",
    "        vmin (float): Minimum value for the color scale. Default is -2.\n",
    "        vmax (float): Maximum value for the color scale. Default is 2.\n",
    "        save_path (str): Path to save the plot. Default is None (no saving).\n",
    "        subsystem_filter (list): List of subsystem names to include in the plot. Default is None (all subsystems).\n",
    "        condition_filter (list): List of condition names (column labels) to include. Default is None (all conditions).\n",
    "        method (str): Linkage method for clustering ('single', 'complete', 'average', 'ward'). Default is 'average'.\n",
    "        metric (str): Distance metric for clustering ('euclidean', 'cityblock', 'cosine', etc.). Default is 'euclidean'.\n",
    "        standardize (bool): Whether to standardize the data before clustering. Default is False.\n",
    "    \"\"\"\n",
    "\n",
    "    # Filter rows by specified subsystems if a filter is provided\n",
    "    if subsystem_filter:\n",
    "        subsystem_indices = [i for i, label in enumerate(row_labels) if label in subsystem_filter]\n",
    "        data = data[subsystem_indices, :]\n",
    "        row_labels = [row_labels[i] for i in subsystem_indices]\n",
    "\n",
    "    # Filter columns by specified conditions if a filter is provided\n",
    "    if condition_filter:\n",
    "        condition_indices = [i for i, label in enumerate(col_labels) if label in condition_filter]\n",
    "        data = data[:, condition_indices]\n",
    "        col_labels = [col_labels[i] for i in condition_indices]\n",
    "\n",
    "    # Check if the filtered data is valid for plotting\n",
    "    if data.shape[0] < 2 or data.shape[1] < 2:\n",
    "        print(f\"Insufficient data for clustered heatmap in '{title}'. Need at least 2 subsystems and 2 conditions.\")\n",
    "        return  # Skip plotting if data is too small\n",
    "\n",
    "    # Optional standardization\n",
    "    if standardize:\n",
    "        scaler = StandardScaler()\n",
    "        data = scaler.fit_transform(data)\n",
    "\n",
    "    # Create a pandas DataFrame for easier handling in seaborn\n",
    "    df = pd.DataFrame(data, index=row_labels, columns=col_labels)\n",
    "\n",
    "    # **Remove rows where all values are zero, and update row_labels\n",
    "    df = df.loc[(df != 0).any(axis=1)]\n",
    "    row_labels = df.index.tolist()\n",
    "    \n",
    "    # Plotting the clustered heatmap\n",
    "    g = sns.clustermap(df, cmap=\"coolwarm\", vmin=vmin, vmax=vmax, method=method, metric=metric,\n",
    "                       figsize=(12, 10), row_cluster=cluster_rows, col_cluster=cluster_cols, dendrogram_ratio=(0.2, 0.2), cbar_kws={'label': 'Value'})\n",
    "\n",
    "    # Adjust the title\n",
    "    plt.title(title, y=1.05)\n",
    "\n",
    "    # Save the plot if save_path is provided\n",
    "    if save_path:\n",
    "        # Ensure the directory exists\n",
    "        os.makedirs(save_path, exist_ok=True)\n",
    "\n",
    "        # Replace spaces and special characters in the title to create a valid filename\n",
    "        sanitized_title = title.replace(' ', '_').replace('-', '_').replace(':', '_')\n",
    "\n",
    "        # Construct the full file path with the .svg extension\n",
    "        filename = f\"{sanitized_title}.svg\"\n",
    "        full_path = os.path.join(save_path, filename)\n",
    "\n",
    "        plt.savefig(full_path, format='svg', dpi=300, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "    \n",
    "\n",
    "def plot_tsne(data_matrix, labels=None, model_ids=None, title=\"t-SNE Projection\", save_path=None):\n",
    "    \"\"\"\n",
    "    Generates a 2D t-SNE projection of the input data with a specific color for each condition.\n",
    "\n",
    "    Parameters:\n",
    "        data_matrix (numpy array): High-dimensional data to project.\n",
    "        labels (list): Labels to use for coloring the plot (e.g., 'WT_P2', 'ZeLa_P4', etc.).\n",
    "        model_ids (list): Specific model IDs to annotate each point.\n",
    "        title (str): Title of the plot.\n",
    "    \"\"\"\n",
    "    if data_matrix.shape[1] < 2:\n",
    "        print(f\"Insufficient data for t-SNE in '{title}'. Only one model or not enough models found.\")\n",
    "        return  # Skip t-SNE if there are not enough models\n",
    "\n",
    "    # Define a color palette to map each condition to a specific color\n",
    "    unique_conditions = sorted(set(labels))  # Get unique conditions\n",
    "    color_palette = sns.color_palette(\"hsv\", len(unique_conditions))  # Generate distinct colors\n",
    "    condition_to_color = {condition: color_palette[i] for i, condition in enumerate(unique_conditions)}\n",
    "\n",
    "    # Assign colors based on the condition of each label\n",
    "    colors = [condition_to_color[label] for label in labels]\n",
    "\n",
    "    # Plotting the t-SNE results\n",
    "    plt.figure(figsize=(10, 8))\n",
    "    plt.scatter(data_matrix[:, 0], data_matrix[:, 1], c=colors, s=100)\n",
    "\n",
    "    # Annotating the points with the model IDs\n",
    "    if model_ids is None:\n",
    "        model_ids = labels  # Default to labels if model_ids not provided\n",
    "    for i, model_id in enumerate(model_ids):\n",
    "        plt.annotate(model_id, (data_matrix[i, 0], data_matrix[i, 1]), fontsize=11)\n",
    "\n",
    "    # Create a legend mapping each condition to its color\n",
    "    handles = [plt.Line2D([0], [0], marker='o', color=color, linestyle='', markersize=10)\n",
    "               for condition, color in condition_to_color.items()]\n",
    "    plt.legend(handles, unique_conditions, title=\"Conditions\", loc=\"best\")\n",
    "\n",
    "    # Plot title and labels\n",
    "    plt.title(title)\n",
    "    plt.xlabel('t-SNE 1')\n",
    "    plt.ylabel('t-SNE 2')\n",
    "    plt.grid(True)\n",
    "\n",
    "    # Save the plot if save_path is provided\n",
    "    if save_path:\n",
    "        # Ensure the directory exists\n",
    "        os.makedirs(save_path, exist_ok=True)\n",
    "\n",
    "        # Replace spaces and special characters in the title to create a valid filename\n",
    "        sanitized_title = title.replace(' ', '_').replace('-', '_').replace(':', '_')\n",
    "\n",
    "        # Construct the full file path with the .svg extension\n",
    "        filename = f\"{sanitized_title}.svg\"\n",
    "        full_path = os.path.join(save_path, filename)\n",
    "\n",
    "        plt.savefig(full_path, format='svg', dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def perform_functional_comparison(models, task_file):\n",
    "    \"\"\"\n",
    "    Performs functional comparison for the models using a task file.\n",
    "\n",
    "    Parameters:\n",
    "        models (list): List of models.\n",
    "        task_file (str): Path to the task file.\n",
    "\n",
    "    Returns:\n",
    "        task_report (list): A list of task performance reports.\n",
    "    \"\"\"\n",
    "    task_report = []\n",
    "    for model in models:\n",
    "        # Placeholder for actual task-checking logic\n",
    "        report = {\n",
    "            'ok': np.random.randint(0, 2, size=10),  # Random binary result (pass/fail)\n",
    "            'description': [f'Task {i}' for i in range(10)]  # Example task descriptions\n",
    "        }\n",
    "        task_report.append(report)\n",
    "    return task_report\n",
    "\n",
    "def assign_labels_based_on_name(model_ids):\n",
    "    \"\"\"\n",
    "    Assigns labels to models based on their names (e.g., 'WT_P2', 'WT_P4', 'ZeLa_P2', etc.).\n",
    "\n",
    "    Parameters:\n",
    "        model_ids (list): List of model IDs.\n",
    "\n",
    "    Returns:\n",
    "        labels (list): List of labels corresponding to each model based on its ID.\n",
    "    \"\"\"\n",
    "    labels = []\n",
    "    for model_id in model_ids:\n",
    "        if 'WT_P2' in model_id:\n",
    "            labels.append('WT_P2')\n",
    "        elif 'WT_P4' in model_id:\n",
    "            labels.append('WT_P4')\n",
    "        elif 'WT_P6' in model_id:\n",
    "            labels.append('WT_P6')\n",
    "        elif 'WT_P8' in model_id:\n",
    "            labels.append('WT_P8')\n",
    "        elif 'ZeLa_P2' in model_id:\n",
    "            labels.append('ZeLa_P2')\n",
    "        elif 'ZeLa_P4' in model_id:\n",
    "            labels.append('ZeLa_P4')\n",
    "        elif 'ZeLa_P6' in model_id:\n",
    "            labels.append('ZeLa_P6')\n",
    "        elif 'ZeLa_P8' in model_id:\n",
    "            labels.append('ZeLa_P8')\n",
    "        else:\n",
    "            labels.append('Unknown')  # Fallback for unexpected names\n",
    "    return labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory containing context-specific models\n",
    "model_directory = '../Data/Context_specific_models/'\n",
    "\n",
    "# Model Names\n",
    "wt_model_names = ['WT_P4_Bio141', 'WT_P4_Bio142', 'WT_P4_Bio143',\n",
    "                  'WT_P6_Bio141', 'WT_P6_Bio142', 'WT_P6_Bio143']\n",
    "\n",
    "zela_model_names = ['ZeLa_P2_Bio145', 'ZeLa_P2_Bio146', 'ZeLa_P2_Bio147', 'ZeLa_P2_Bio148',\n",
    "                    'ZeLa_P4_Bio144', 'ZeLa_P4_Bio145', 'ZeLa_P4_Bio146', 'ZeLa_P4_Bio147', 'ZeLa_P4_Bio148',\n",
    "                    'ZeLa_P6_Bio144', 'ZeLa_P6_Bio146', 'ZeLa_P6_Bio147', 'ZeLa_P6_Bio148',\n",
    "                    'ZeLa_P8_Bio144', 'ZeLa_P8_Bio145', 'ZeLa_P8_Bio146', 'ZeLa_P8_Bio147', 'ZeLa_P8_Bio148']\n",
    "\n",
    "# List to store the loaded models\n",
    "models_list = []\n",
    "\n",
    "# List all .mat files in the directory\n",
    "model_files = [f for f in os.listdir(model_directory) if f.endswith('.mat')]\n",
    "\n",
    "# Function to parse and extract category, time point, and batch for sorting\n",
    "def parse_model_name(model_name):\n",
    "    parts = model_name.split('_')\n",
    "    if len(parts) == 3:\n",
    "        category = parts[0]  # WT or ZeLa\n",
    "        time_point = parts[1]  # P2, P4, P6, P8\n",
    "        batch = parts[2][3:]  # Extract the number from 'Bio141', 'Bio142', etc.\n",
    "        return category, time_point, batch\n",
    "    return None, None, None\n",
    "\n",
    "# Iterate through the .mat files and load only the models that match predefined strings\n",
    "for model_file in model_files:\n",
    "    # Full path to the model file\n",
    "    model_path = os.path.join(model_directory, model_file)\n",
    "    \n",
    "    # Check if the file corresponds to a WT or ZeLa model\n",
    "    for model_name in wt_model_names + zela_model_names:\n",
    "        if model_name in model_file:\n",
    "            # Load the model\n",
    "            model = load_matlab_model(model_path)\n",
    "            model.id = model_name\n",
    "\n",
    "            # Append the model to the list\n",
    "            models_list.append(model)\n",
    "            print(f\"Loaded model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed\n",
    "\n",
    "# Function to generate a sorting key for models\n",
    "def model_sort_key(model):\n",
    "    category, time_point, batch = parse_model_name(model.id)\n",
    "    return (category, time_point, int(batch))\n",
    "\n",
    "# Sort the list of models by category (WT, ZeLa), time point (P2, P4, P6), and batch (numeric)\n",
    "models_list.sort(key=model_sort_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acid_subsystems = subsystems_dict['AMINO ACID METABOLISM']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison of multiple models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_conditions = ['WT_P4_Bio141', 'WT_P4_Bio142', 'WT_P4_Bio143', 'ZeLa_P4_Bio144', 'ZeLa_P4_Bio145', 'ZeLa_P4_Bio146', 'ZeLa_P4_Bio147', 'ZeLa_P4_Bio148']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Call the comparison function\n",
    "save_path='../Analyses/tSNE/'\n",
    "\n",
    "comp_struct = compare_multiple_cobra_models(\n",
    "    models_list,\n",
    "    print_results=False,\n",
    "    plot_results=True,\n",
    "    subsystem_mapping=subsystems_dict,\n",
    "    #subsystem_filter=amino_acid_subsystems,\n",
    "    #condition_filter=selected_conditions,\n",
    "    save_path=save_path,\n",
    "    cluster_rows=True,\n",
    "    cluster_cols=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison of transcriptomics datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq = pd.read_csv('../Data/ZeLa Data/20200307_Bio141-148_merged.tpm.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the mapping from 'S' numbers to sample conditions\n",
    "sample_conditions = {\n",
    "    'S1':  'WT_P2_Bio141', 'S7':  'WT_P2_Bio142', 'S13': 'WT_P2_Bio143', 'S2':  'WT_P4_Bio141', 'S8':  'WT_P4_Bio142', 'S14': 'WT_P4_Bio143', 'S3':  'WT_P6_Bio141',\n",
    "    'S9':  'WT_P6_Bio142', 'S15': 'WT_P6_Bio143', 'S4':  'WT_P8_Bio141', 'S10': 'WT_P8_Bio142', 'S16': 'WT_P8_Bio143', 'S5':  'WT_P12_Bio141', 'S11': 'WT_P12_Bio142',\n",
    "    'S6':  'WT_P14_Bio141', 'S12': 'WT_P14_Bio142', 'S17': 'WT_P14_Bio143', 'S18': 'ZeLa_P4_Bio144', 'S23': 'ZeLa_P4_Bio145', 'S28': 'ZeLa_P4_Bio146',\n",
    "    'S34': 'ZeLa_P4_Bio147', 'S39': 'ZeLa_P4_Bio148', 'S19': 'ZeLa_P6_Bio144', 'S29': 'ZeLa_P6_Bio146', 'S35': 'ZeLa_P6_Bio147', 'S40': 'ZeLa_P6_Bio148',\n",
    "    'S20': 'ZeLa_P8_Bio144', 'S24': 'ZeLa_P8_Bio145', 'S30': 'ZeLa_P8_Bio146', 'S36': 'ZeLa_P8_Bio147', 'S41': 'ZeLa_P8_Bio148', 'S25': 'ZeLa_P12_Bio145',\n",
    "    'S31': 'ZeLa_P12_Bio146', 'S42': 'ZeLa_P12_Bio148', 'S21': 'ZeLa_P14_Bio144', 'S26': 'ZeLa_P14_Bio145', 'S32': 'ZeLa_P14_Bio146', 'S37': 'ZeLa_P14_Bio147',\n",
    "    'S43': 'ZeLa_P14_Bio148', 'S22': 'ZeLa_P2_Bio145',  'S27': 'ZeLa_P2_Bio146', 'S33': 'ZeLa_P2_Bio147', 'S38': 'ZeLa_P2_Bio148'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map the columns in 'rna_seq' to sample conditions\n",
    "sample_columns = rna_seq.columns.tolist()[1:]  # Exclude 'Gene' column\n",
    "column_to_condition = {col: sample_conditions.get(col, 'Unknown') for col in sample_columns}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the conditions to include\n",
    "include_conditions = set(wt_model_names + zela_model_names)\n",
    "\n",
    "# Create a list of columns to include\n",
    "columns_to_include = ['Gene'] + [col for col in sample_columns if column_to_condition[col] in include_conditions]\n",
    "\n",
    "# Subset the data\n",
    "rna_seq_subset = rna_seq[columns_to_include]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the sample conditions for the subset columns (excluding 'Gene')\n",
    "subset_columns = rna_seq_subset.columns.tolist()[1:]\n",
    "subset_sample_conditions = [column_to_condition[col] for col in subset_columns]\n",
    "\n",
    "# Define a function to assign labels\n",
    "def assign_sample_labels(sample_conditions):\n",
    "    labels = []\n",
    "    for condition in sample_conditions:\n",
    "        if 'WT_P2' in condition:\n",
    "            labels.append('WT_P2')\n",
    "        elif 'WT_P4' in condition:\n",
    "            labels.append('WT_P4')\n",
    "        elif 'WT_P6' in condition:\n",
    "            labels.append('WT_P6')\n",
    "        elif 'WT_P8' in condition:\n",
    "            labels.append('WT_P8')\n",
    "        elif 'ZeLa_P2' in condition:\n",
    "            labels.append('ZeLa_P2')\n",
    "        elif 'ZeLa_P4' in condition:\n",
    "            labels.append('ZeLa_P4')\n",
    "        elif 'ZeLa_P6' in condition:\n",
    "            labels.append('ZeLa_P6')\n",
    "        elif 'ZeLa_P8' in condition:\n",
    "            labels.append('ZeLa_P8')\n",
    "        else:\n",
    "            labels.append('Unknown')\n",
    "    return labels\n",
    "\n",
    "# Assign labels\n",
    "labels = assign_sample_labels(subset_sample_conditions)\n",
    "\n",
    "# Verify the labels\n",
    "print(\"Sample Labels:\")\n",
    "for sample, label in zip(subset_sample_conditions, labels):\n",
    "    print(f\"{sample}: {label}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set 'Gene' as the index\n",
    "rna_seq_subset.set_index('Gene', inplace=True)\n",
    "\n",
    "# Perform log2 transformation\n",
    "data_log = np.log2(rna_seq_subset + 1)\n",
    "\n",
    "# Transpose the data to have samples as rows\n",
    "data_matrix = data_log.T.values  # Shape: (samples, genes)\n",
    "\n",
    "# Check for NaNs or infinite values\n",
    "if np.isnan(data_matrix).any():\n",
    "    print(\"Data contains NaNs, handling missing values.\")\n",
    "    # Fill NaNs with the mean expression of each gene\n",
    "    data_matrix = np.nan_to_num(data_matrix, nan=np.nanmean(data_matrix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine perplexity\n",
    "n_samples = data_matrix.shape[0]\n",
    "perplexity = min(n_samples - 1, 30)\n",
    "\n",
    "# Perform t-SNE\n",
    "tsne = TSNE(metric='euclidean', n_components=3, perplexity=perplexity, random_state=42)\n",
    "tsne_results = tsne.fit_transform(data_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tsne_transcriptomics(tsne_results, labels, sample_names, title=\"Transcriptomics t-SNE Projection\", save_path=None):\n",
    "    unique_conditions = sorted(set(labels))\n",
    "    color_palette = sns.color_palette(\"hsv\", len(unique_conditions))\n",
    "    condition_to_color = {condition: color_palette[i] for i, condition in enumerate(unique_conditions)}\n",
    "    colors = [condition_to_color[label] for label in labels]\n",
    "    plt.figure(figsize=(10, 8))\n",
    "    plt.scatter(tsne_results[:, 0], tsne_results[:, 1], c=colors, s=100)\n",
    "    for i, sample_name in enumerate(sample_names):\n",
    "        plt.annotate(sample_name, (tsne_results[i, 0], tsne_results[i, 1]), fontsize=11)\n",
    "    handles = [plt.Line2D([0], [0], marker='o', color=color, linestyle='', markersize=10)\n",
    "               for condition, color in condition_to_color.items()]\n",
    "    plt.legend(handles, unique_conditions, title=\"Conditions\", loc=\"best\")\n",
    "    plt.title(title)\n",
    "    plt.xlabel('t-SNE 1')\n",
    "    plt.ylabel('t-SNE 2')\n",
    "    plt.grid(True)\n",
    "\n",
    "    # Save the plot if save_path is provided\n",
    "    if save_path:\n",
    "        # Ensure the directory exists\n",
    "        os.makedirs(save_path, exist_ok=True)\n",
    "\n",
    "        # Replace spaces and special characters in the title to create a valid filename\n",
    "        sanitized_title = title.replace(' ', '_').replace('-', '_').replace(':', '_')\n",
    "\n",
    "        # Construct the full file path with the .svg extension\n",
    "        filename = f\"{sanitized_title}.svg\"\n",
    "        full_path = os.path.join(save_path, filename)\n",
    "\n",
    "        plt.savefig(full_path, format='svg', bbox_inches='tight')\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "# Call the plotting function\n",
    "save_path='../Analyses/tSNE/'\n",
    "sample_names = subset_sample_conditions  # Sample conditions corresponding to the samples\n",
    "plot_tsne_transcriptomics(tsne_results, labels, sample_names, save_path=save_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Biomass prediction using exp. data <a id='biomass'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "import cobra\n",
    "from cobra import Reaction\n",
    "from cobra.io import load_json_model, read_sbml_model, load_matlab_model\n",
    "from cobra.exceptions import Infeasible\n",
    "from cobra.sampling import ACHRSampler\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mode and Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory containing context-specific models models\n",
    "model_directory = '../Data/Context_specific_models/'\n",
    "\n",
    "# Model Names\n",
    "wt_model_names = ['WT_P4_Bio141', 'WT_P4_Bio142', 'WT_P4_Bio143',\n",
    "                  'WT_P6_Bio141', 'WT_P6_Bio142', 'WT_P6_Bio143']\n",
    "\n",
    "zela_model_names = ['ZeLa_P2_Bio145', 'ZeLa_P2_Bio146', 'ZeLa_P2_Bio147', 'ZeLa_P2_Bio148',\n",
    "                    'ZeLa_P4_Bio144', 'ZeLa_P4_Bio145', 'ZeLa_P4_Bio146', 'ZeLa_P4_Bio147', 'ZeLa_P4_Bio148',\n",
    "                    'ZeLa_P6_Bio144', 'ZeLa_P6_Bio146', 'ZeLa_P6_Bio147', 'ZeLa_P6_Bio148',\n",
    "                    'ZeLa_P8_Bio144', 'ZeLa_P8_Bio145', 'ZeLa_P8_Bio146', 'ZeLa_P8_Bio147', 'ZeLa_P8_Bio148']\n",
    "\n",
    "# Dictionaries to store the loaded models\n",
    "wt_models = {}\n",
    "zela_models = {}\n",
    "\n",
    "# List all .mat files in the directory\n",
    "model_files = [f for f in os.listdir(model_directory) if f.endswith('.mat')]\n",
    "\n",
    "# Iterate through the .mat files and load only the models that match predefined strings\n",
    "for model_file in model_files:\n",
    "    # Full path to the model file\n",
    "    model_path = os.path.join(model_directory, model_file)\n",
    "    \n",
    "    # Check if the file corresponds to a WT model\n",
    "    for model_name in wt_model_names:\n",
    "        if model_name in model_file:\n",
    "            wt_models[model_name] = load_matlab_model(model_path)\n",
    "            wt_models[model_name].id = model_name\n",
    "            print(f\"Loaded WT model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed\n",
    "    \n",
    "    # Check if the file corresponds to a ZeLa model\n",
    "    for model_name in zela_model_names:\n",
    "        if model_name in model_file:\n",
    "            zela_models[model_name] = load_matlab_model(model_path)\n",
    "            zela_models[model_name].id = model_name\n",
    "            print(f\"Loaded ZeLa model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATP generating reactions loop\n",
    "\n",
    "atp_loop_reactions = [\n",
    " 'SCP22x','TMNDNCCOAtx','OCCOAtx','r0391','BiGGRxn67','r2247','r2246','r2245','r2317','HMR_0293',\n",
    " 'HMR_7741','r1453','r1393','NICRNS','GapFill-R08726','RE2915M','HMR_3288','HMR_1325','RE2439C',\n",
    " 'r1450','RE3477C','AAPSAS','r0698','3HDH260p','HMR_3272','ACOAD183n3m','GapFill-R01463','r1468',\n",
    " 'r0655','r0603','r0541','HMR_1329','GapFill-R03599','AKGDm','r0556','OIVD1m','OIVD3m','2OXOADOXm',\n",
    " 'r0386','r0451','GLYCLm','MACACI'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Sanity Check. Make sure all the exchange reactions in the experimental data are present in the model and that the models ar growing\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('')\n",
    "    print(f'GR1: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    \n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "\n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_loop_reactions:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            new_gr = model.slim_optimize()\n",
    "            print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "            \n",
    "            # Check if the model's growth rate is infeasible or close to zero\n",
    "            if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                reaction.bounds = (0,1000)  # Set reaction to irreversible\n",
    "\n",
    "                new_gr_3 = model.slim_optimize() # Optimize model with irreversible reaction\n",
    "                print(f'GR3 (After making reaction irreversible {atp_reaction}): {new_gr_3}')\n",
    "\n",
    "                # Check if the model's growth rate 3 is infeasible or close to zero\n",
    "                if math.isnan(new_gr_3) or new_gr_3 < 1e-6:\n",
    "                    print(f'{atp_reaction} in its irreversible format causes infeasibility.')\n",
    "                    reaction.bounds = original_bounds\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully transformed to reversible without infeasibility.')\n",
    "                    \n",
    "            else:\n",
    "                print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "\n",
    "                \n",
    "    print('')\n",
    "    print('Missing reactions from exp data:')\n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')\n",
    "    print('')\n",
    "\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('---------------------------------------')\n",
    "    print(f'GR: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "\n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_loop_reactions:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            new_gr = model.slim_optimize()\n",
    "            print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "            \n",
    "            # Check if the model's growth rate is infeasible or close to zero\n",
    "            if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                reaction.bounds = (0,1000)  # Set reaction to irreversible\n",
    "\n",
    "                new_gr_3 = model.slim_optimize() # Optimize model with irreversible reaction\n",
    "                print(f'GR3 (After making reaction irreversible {atp_reaction}): {new_gr_3}')\n",
    "\n",
    "                # Check if the model's growth rate 3 is infeasible or close to zero\n",
    "                if math.isnan(new_gr_3) or new_gr_3 < 1e-6:\n",
    "                    print(f'{atp_reaction} in its irreversible format causes infeasibility.')\n",
    "                    reaction.bounds = original_bounds\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully transformed to reversible without infeasibility.')\n",
    "                    \n",
    "            else:\n",
    "                print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "    \n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Growth Rate Predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfba_solutions_fluxes = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# WT\n",
    "\n",
    "results = []\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "replicates = {'U1':'Bio141', 'U2':'Bio142', 'U3':'Bio143'}\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "\n",
    "                     with model as modified_model:\n",
    "\n",
    "                        # Open the bounds for the biomass reaction \n",
    "                        modified_model.reactions.biomass_cho_s.bounds = (0,1000)\n",
    "                         \n",
    "                        # Create a copy of the modified_model before making changes\n",
    "                        pre_modification_model = modified_model.copy()\n",
    "                         \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "                        \n",
    "                        print(f'Calculating Growth Rate for WT Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_wt['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                reaction.lower_bound = 200\n",
    "                            for r,v in uptsec_wt.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "            \n",
    "                            # If pFBA succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = modified_model\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')\n",
    "                            print('Reverting to the original model and retrying pFBA...')\n",
    "\n",
    "                            for rxn in pre_modification_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                                if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                    rxn.bounds = (-1000, 1000)\n",
    "                                elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                    rxn.bounds = (-0.001, 1000)\n",
    "                            \n",
    "                            pfba_solution = cobra.flux_analysis.pfba(pre_modification_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate with original model is: {pfba_obj}')\n",
    "\n",
    "                            # If pFBA with original model succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = pre_modification_model\n",
    "\n",
    "                        results.append([rep_suffix, interval_key, exp_gr, pfba_obj])\n",
    "                        # Save the solution with metadata\n",
    "                        pfba_solutions_fluxes.append({\n",
    "                            \"model\": pre_modification_model.id,\n",
    "                            \"condition\": interval_key,\n",
    "                            \"fluxes\": pfba_solution.fluxes\n",
    "                        })\n",
    "\n",
    "                        '''\n",
    "                         # Perform flux sampling using ACHRSampler\n",
    "                        print(\"Performing flux sampling...\")\n",
    "                        achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                        samples = 600 \n",
    "                        sampled_fluxes = achr_sampler.sample(samples)\n",
    "                        \n",
    "                        # Validate and save the sampled fluxes\n",
    "                        if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                            print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                        else:\n",
    "                            f_name = f\"../Analyses/growth_rate_pred/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                            with open(f_name, 'wb') as file:\n",
    "                                pickle.dump(sampled_fluxes, file)\n",
    "                       ''' \n",
    "# Creating a DataFrame\n",
    "df_wt = pd.DataFrame(results, columns=['Hue', 'Category', 'X Axis', 'Y Axis'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df_wt\n",
    "# Define markers for categories\n",
    "markers = {\n",
    "    'P0 to P2': 'p',  # Circle\n",
    "    'P2 to P4': 'o',  # Circle\n",
    "    'P4 to P6': 's',   # Square\n",
    "}\n",
    "\n",
    "# Set up the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# Plot each group with different colors for hue and different shapes for categories\n",
    "hues = df['Hue'].unique()\n",
    "colors = plt.colormaps.get_cmap('tab10')\n",
    "\n",
    "for hue_idx, hue in enumerate(hues):\n",
    "    hue_subset = df[df['Hue'] == hue]\n",
    "    for category, marker in markers.items():\n",
    "        subset = hue_subset[hue_subset['Category'] == category]\n",
    "        plt.scatter(subset['X Axis'], subset['Y Axis'], label=f'{hue} - {category}', marker=marker, s=100, color=colors(hue_idx))\n",
    "\n",
    "# Add the identity line y = x\n",
    "plt.plot([0.00, 0.055], [0.00, 0.055], color='red', linestyle='--')\n",
    "\n",
    "# Set evenly distributed ticks from 0.01 to 0.08\n",
    "ticks = [i / 100.0 for i in range(1, 11)]\n",
    "plt.xticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "plt.yticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "\n",
    "# Set limits for both axes\n",
    "plt.xlim(0.0, 0.055)\n",
    "plt.ylim(0.0, 0.055)\n",
    "\n",
    "# Setting the same scale for both axes\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Adding labels and title\n",
    "plt.xlabel('Experimental Growth Rate')\n",
    "plt.ylabel('Simulated Growth Rate')\n",
    "plt.title('WT')\n",
    "plt.legend(title='Hue - Category', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Analyses/growth_rate_pred/bar_plot_growth_rate_wt.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ZeLa\n",
    "\n",
    "results = []\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "replicates = {'U4':'Bio144', 'U5':'Bio145', 'U6':'Bio146', 'U7':'Bio147', 'U8':'Bio148'}\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "                    \n",
    "                     with model as modified_model:\n",
    "\n",
    "                        # Open the bounds for the biomass reaction \n",
    "                        modified_model.reactions.biomass_cho_s.bounds = (0,1000)\n",
    "                         \n",
    "                        # Create a copy of the modified_model before making changes\n",
    "                        pre_modification_model = modified_model.copy()\n",
    " \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "                        print(f'Calculating Growth Rate for ZeLa Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_zela['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                reaction.lower_bound = 6\n",
    "                            for r,v in uptsec_zela.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')\n",
    "                            print('Reverting to the original model and retrying pFBA...')\n",
    "\n",
    "                            for rxn in pre_modification_model.reactions:\n",
    "                                if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                    rxn.bounds = (-1000, 1000)\n",
    "                                elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                    rxn.bounds = (-0.001, 1000)\n",
    "                                elif rxn.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                    rxn.lower_bound = 0\n",
    "                                \n",
    "                            pfba_solution = cobra.flux_analysis.pfba(pre_modification_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate with original model is: {pfba_obj}')\n",
    "                            \n",
    "                        results.append([rep_suffix, interval_key, exp_gr, pfba_obj])\n",
    "\n",
    "                        # Save the solution with metadata\n",
    "                        pfba_solutions_fluxes.append({\n",
    "                            \"model\": pre_modification_model.id,\n",
    "                            \"condition\": interval_key,\n",
    "                            \"fluxes\": pfba_solution.fluxes\n",
    "                        })\n",
    "                        ''' \n",
    "                        if (interval_key == 'P2 to P4') or (interval_key == 'P4 to P6'):\n",
    "                            # Perform flux sampling using ACHRSampler\n",
    "                            print(\"Performing flux sampling...\")\n",
    "                            achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                            samples = 600 \n",
    "                            sampled_fluxes = achr_sampler.sample(samples)\n",
    "                            \n",
    "                            # Validate and save the sampled fluxes\n",
    "                            if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                                print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                            else:\n",
    "                                f_name = f\"../Simulations/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                                with open(f_name, 'wb') as file:\n",
    "                                    pickle.dump(sampled_fluxes, file)\n",
    "                        '''\n",
    "    \n",
    "# Creating a DataFrame\n",
    "df_zela = pd.DataFrame(results, columns=['Hue', 'Category', 'X Axis', 'Y Axis'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df_zela\n",
    "# Define markers for categories\n",
    "markers = {\n",
    "    'P0 to P2': 'p',\n",
    "    'P2 to P4': 'o',  # Circle\n",
    "    'P4 to P6': 's',   # Square\n",
    "    'P6 to P8': 'v',\n",
    "}\n",
    "\n",
    "# Set up the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# Plot each group with different colors for hue and different shapes for categories\n",
    "hues = df['Hue'].unique()\n",
    "colors = plt.colormaps.get_cmap('tab10')\n",
    "\n",
    "for hue_idx, hue in enumerate(hues):\n",
    "    hue_subset = df[df['Hue'] == hue]\n",
    "    for category, marker in markers.items():\n",
    "        subset = hue_subset[hue_subset['Category'] == category]\n",
    "        plt.scatter(subset['X Axis'], subset['Y Axis'], label=f'{hue} - {category}', marker=marker, s=100, color=colors(hue_idx))\n",
    "\n",
    "# Add the identity line y = x\n",
    "plt.plot([0.00, 0.055], [0.00, 0.055], color='red', linestyle='--')\n",
    "\n",
    "# Set evenly distributed ticks from 0.01 to 0.08\n",
    "ticks = [i / 100.0 for i in range(1,11)]\n",
    "plt.xticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "plt.yticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "\n",
    "# Set limits for both axes\n",
    "plt.xlim(0.0, 0.055)\n",
    "plt.ylim(0.0, 0.055)\n",
    "\n",
    "# Setting the same scale for both axes\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Adding labels and title\n",
    "plt.xlabel('Experimental Growth Rate')\n",
    "plt.ylabel('Simulated Growth Rate')\n",
    "plt.title('ZeLa')\n",
    "plt.legend(title='Hue - Category', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Analyses/growth_rate_pred/bar_plot_growth_rate_zela.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for reaction, values in uptsec_zela.items():\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    \n",
    "    # Organize data for plotting\n",
    "    batches = sorted(set(k[0] for k in values.keys()))\n",
    "    timepoints = ['P0 to P2', 'P2 to P4', 'P4 to P6', 'P6 to P8', 'P8 to P12', 'P12 to P14']\n",
    "    \n",
    "    for batch in batches:\n",
    "        batch_data = [values.get((batch, tp), 0) for tp in timepoints]\n",
    "        plt.plot(timepoints, batch_data, marker='o', label=batch)\n",
    "    \n",
    "    plt.title(f'Changes in {reaction} over time')\n",
    "    plt.xlabel('Timepoints')\n",
    "    plt.ylabel('Values')\n",
    "    plt.legend()\n",
    "    plt.xticks(rotation=45)\n",
    "    plt.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for reaction, values in uptsec_wt.items():\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    \n",
    "    # Organize data for plotting\n",
    "    batches = sorted(set(k[0] for k in values.keys()))\n",
    "    timepoints = ['P0 to P2', 'P2 to P4', 'P4 to P6', 'P6 to P8', 'P8 to P12', 'P12 to P14']\n",
    "    \n",
    "    for batch in batches:\n",
    "        batch_data = [values.get((batch, tp), 0) for tp in timepoints]\n",
    "        plt.plot(timepoints, batch_data, marker='o', label=batch)\n",
    "    \n",
    "    plt.title(f'Changes in {reaction} over time')\n",
    "    plt.xlabel('Timepoints')\n",
    "    plt.ylabel('Values')\n",
    "    plt.legend()\n",
    "    plt.xticks(rotation=45)\n",
    "    plt.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Flux Sampling Analysis <a id='fsa'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "import cobra\n",
    "from cobra import Reaction\n",
    "from cobra.io import load_json_model, read_sbml_model, load_matlab_model\n",
    "from cobra.flux_analysis.loopless import add_loopless\n",
    "from cobra.exceptions import Infeasible\n",
    "from cobra.sampling import ACHRSampler\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model and Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Loaded ZeLa model: ZeLa_P6_Bio144\n",
      "Loaded WT model: WT_P6_Bio143\n",
      "Loaded ZeLa model: ZeLa_P6_Bio146\n",
      "Loaded ZeLa model: ZeLa_P6_Bio148\n",
      "Loaded ZeLa model: ZeLa_P4_Bio145\n",
      "Loaded WT model: WT_P4_Bio141\n",
      "Loaded ZeLa model: ZeLa_P4_Bio144\n",
      "Loaded ZeLa model: ZeLa_P4_Bio146\n",
      "Loaded WT model: WT_P4_Bio143\n",
      "Loaded ZeLa model: ZeLa_P4_Bio148\n",
      "Loaded WT model: WT_P4_Bio142\n",
      "Loaded ZeLa model: ZeLa_P4_Bio147\n",
      "Loaded WT model: WT_P6_Bio141\n",
      "Loaded WT model: WT_P6_Bio142\n",
      "Loaded ZeLa model: ZeLa_P6_Bio147\n"
     ]
    }
   ],
   "source": [
    "# Directory containing context-specific models models\n",
    "model_directory = '../Data/Context_specific_models/'\n",
    "\n",
    "# Model Names\n",
    "wt_model_names = ['WT_P4_Bio141', 'WT_P4_Bio142', 'WT_P4_Bio143',\n",
    "                  'WT_P6_Bio141', 'WT_P6_Bio142', 'WT_P6_Bio143']\n",
    "\n",
    "zela_model_names = ['ZeLa_P4_Bio144', 'ZeLa_P4_Bio145', 'ZeLa_P4_Bio146', 'ZeLa_P4_Bio147', 'ZeLa_P4_Bio148',\n",
    "                    'ZeLa_P6_Bio144', 'ZeLa_P6_Bio146', 'ZeLa_P6_Bio147', 'ZeLa_P6_Bio148']\n",
    "\n",
    "# Dictionaries to store the loaded models\n",
    "wt_models = {}\n",
    "zela_models = {}\n",
    "\n",
    "# List all .mat files in the directory\n",
    "model_files = [f for f in os.listdir(model_directory) if f.endswith('.mat')]\n",
    "\n",
    "# Iterate through the .mat files and load only the models that match predefined strings\n",
    "for model_file in model_files:\n",
    "    # Full path to the model file\n",
    "    model_path = os.path.join(model_directory, model_file)\n",
    "    \n",
    "    # Check if the file corresponds to a WT model\n",
    "    for model_name in wt_model_names:\n",
    "        if model_name in model_file:\n",
    "            wt_models[model_name] = load_matlab_model(model_path)\n",
    "            wt_models[model_name].id = model_name\n",
    "            print(f\"Loaded WT model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed\n",
    "    \n",
    "    # Check if the file corresponds to a ZeLa model\n",
    "    for model_name in zela_model_names:\n",
    "        if model_name in model_file:\n",
    "            zela_models[model_name] = load_matlab_model(model_path)\n",
    "            zela_models[model_name].id = model_name\n",
    "            print(f\"Loaded ZeLa model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATP generating reactions loop\n",
    "\n",
    "atp_loop_reactions = [\n",
    " 'SCP22x','TMNDNCCOAtx','OCCOAtx','r0391','BiGGRxn67','r2247','r2246','r2245','r2317','HMR_0293',\n",
    " 'HMR_7741','r1453','r1393','NICRNS','GapFill-R08726','RE2915M','HMR_3288','HMR_1325','RE2439C',\n",
    " 'r1450','RE3477C','AAPSAS','r0698','3HDH260p','HMR_3272','ACOAD183n3m','GapFill-R01463','r1468',\n",
    " 'r0655','r0603','r0541','HMR_1329','GapFill-R03599','AKGDm','r0556','OIVD1m','OIVD3m','2OXOADOXm',\n",
    " 'r0386','r0451','GLYCLm','MACACI'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------------------------\n",
      "------------ZeLa_P6_Bio144------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.018563977057166503\n",
      "GR2 (After knocking out SCP22x): 0.018563977057166503\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.018563977057166503\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): nan\n",
      "HMR_7741 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible HMR_7741): 0.018563977057166503\n",
      "HMR_7741 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r1453): 0.018563977057166503\n",
      "r1453 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.018563977057166503\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): 0.018563977057166503\n",
      "NICRNS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.018563977057166503\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.018563977057166503\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.018563977057166503\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): nan\n",
      "AAPSAS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible AAPSAS): 0.018563977057166503\n",
      "AAPSAS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.018563977057166503\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.018563977057166503\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): nan\n",
      "HMR_3272 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible HMR_3272): 0.018563977057166503\n",
      "HMR_3272 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.018563977057166503\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.018563977057166503\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.018563977057166503\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.018563977057166503\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): nan\n",
      "r0556 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0556): nan\n",
      "r0556 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.018181899161166637\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.018181899161166637\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.018181899161166703\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.018181899161166703\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P6_Bio146------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.018088924148261774\n",
      "GR2 (After knocking out SCP22x): 0.018088924148261774\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.018088924148261774\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.018088924148261774\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1453): 0.018088924148261767\n",
      "r1453 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.018088924148261864\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.01808892414826175\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out GapFill-R08726): nan\n",
      "GapFill-R08726 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible GapFill-R08726): nan\n",
      "GapFill-R08726 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.018088924148261784\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.01808892414826179\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.01808892414826179\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.01808892414826179\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.01808892414826179\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.018088924148261774\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.018088924148261774\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.01808892414826181\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.018088924148261788\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.018088924148261788\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.018088924148261774\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.01808892414826182\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.01808892414826182\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.01808892414826182\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): nan\n",
      "OIVD1m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD1m): 0.0180889241482618\n",
      "OIVD1m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.0180889241482618\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.0180889241482618\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): nan\n",
      "r0451 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0451): 0.01808892414826182\n",
      "r0451 successfully transformed to reversible without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P6_Bio148------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.014319558260879004\n",
      "GR2 (After knocking out SCP22x): 0.014319558260879004\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.014319558260879004\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.014319558260879004\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.014319558260879004\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): 0.014319558260879004\n",
      "NICRNS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.014319558260879002\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.014319558260879005\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.014319558260879005\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): nan\n",
      "AAPSAS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible AAPSAS): 0.014319558260879005\n",
      "AAPSAS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.014319558260879005\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.014319558260879007\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): nan\n",
      "HMR_3272 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible HMR_3272): 0.014319558260879\n",
      "HMR_3272 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.014319558260879002\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.014319558260879007\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.014319558260879007\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.014319558260879004\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.014319558260879004\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.014319558260879004\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): nan\n",
      "OIVD1m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD1m): 0.014319558260879002\n",
      "OIVD1m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.014319558260879002\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.014319558260879007\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.014319558260879002\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P4_Bio145------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.03725896271642408\n",
      "GR2 (After knocking out SCP22x): 0.03725896271642408\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.03725896271642408\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.03725896271642408\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.03725896271642408\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.03725896271642408\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.03725896271642408\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.03725896271642408\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.03725896271642408\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.03725896271642408\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.03725896271642408\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.03725896271642408\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.03725896271642408\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.03725896271642408\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.03725896271642408\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.03725896271642408\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.03725896271642408\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.03725896271642408\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.03725896271642408\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.0341238620496739\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.0341238620496739\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.0341238620496739\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.0341238620496739\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_lac_L_e\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P4_Bio144------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.03725896271642408\n",
      "GR2 (After knocking out SCP22x): 0.03725896271642408\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.03725896271642408\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): nan\n",
      "HMR_7741 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible HMR_7741): 0.03725896271642408\n",
      "HMR_7741 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r1453): 0.03725896271642408\n",
      "r1453 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.03725896271642408\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.03725896271642408\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.03725896271642408\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.03725896271642408\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.03725896271642408\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.03725896271642408\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.03725896271642408\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.03725896271642408\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.03725896271642408\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.03725896271642408\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.03725896271642408\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.03725896271642408\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.03725896271642408\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.03725896271642408\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.03725896271642408\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.03725896271642408\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.03725896271642408\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.03725896271642408\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_lac_L_e\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P4_Bio146------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.03725896271642408\n",
      "GR2 (After knocking out SCP22x): 0.03725896271642408\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.03725896271642408\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.03725896271642408\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.03725896271642408\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.03725896271642408\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.03725896271642408\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.03725896271642408\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.03725896271642408\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.03725896271642408\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.03725896271642408\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.03725896271642408\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.03725896271642408\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.03725896271642408\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.03725896271642408\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.03725896271642408\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.03725896271642408\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.03725896271642408\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.03725896271642408\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.0341238620496739\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.0341238620496739\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.03412386204967391\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.03412386204967391\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_lac_L_e\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P4_Bio148------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.03725896271642408\n",
      "GR2 (After knocking out SCP22x): 0.03725896271642408\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.03725896271642408\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.03725896271642408\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.03725896271642408\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.03725896271642408\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.03725896271642408\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.03725896271642408\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.03725896271642408\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.03725896271642408\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.03725896271642408\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.03725896271642408\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.03725896271642408\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.03725896271642408\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.03725896271642408\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.03725896271642408\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.03725896271642408\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.03725896271642408\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.03725896271642408\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.03412386204967391\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.03412386204967391\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.0341238620496739\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.03412386204967391\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_lac_L_e\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P4_Bio147------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.03725896271642408\n",
      "GR2 (After knocking out SCP22x): 0.03725896271642408\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.03725896271642408\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.03725896271642408\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.03725896271642408\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.03725896271642408\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.03725896271642408\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.03725896271642408\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.03725896271642408\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.03725896271642408\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.03725896271642408\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.03725896271642408\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.03725896271642408\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.03725896271642408\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.03725896271642408\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.03725896271642408\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.03725896271642408\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.03725896271642408\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.03725896271642408\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.03725896271642408\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.03725896271642408\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.03725896271642408\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_lac_L_e\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------ZeLa_P6_Bio147------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "\n",
      "GR1: 0.018563977057166503\n",
      "GR2 (After knocking out SCP22x): 0.018563977057166503\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.018563977057166503\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): nan\n",
      "HMR_7741 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible HMR_7741): 0.018563977057166503\n",
      "HMR_7741 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.018563977057166503\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.018563977057166503\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.018563977057166503\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.018563977057166503\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.018563977057166503\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.018563977057166503\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.018563977057166503\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.018563977057166503\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.018563977057166503\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.018563977057166503\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.018563977057166503\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.018563977057166503\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.018563977057166503\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.018563977057166503\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.018563977057166503\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.018563977057166503\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.018563977057166503\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.018563977057166503\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "\n",
      "Missing reactions from exp data:\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "\n",
      "---------------------------------------\n",
      "------------WT_P6_Bio143------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "---------------------------------------\n",
      "GR: 0.018010251694238397\n",
      "GR2 (After knocking out SCP22x): 0.018010251694238397\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.018010251694238397\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.018010251694238397\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1453): 0.018010251694238397\n",
      "r1453 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.018010251694238397\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.018010251694238397\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out GapFill-R08726): nan\n",
      "GapFill-R08726 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible GapFill-R08726): nan\n",
      "GapFill-R08726 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.018010251694238397\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.018010251694238397\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.018010251694238397\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.018010251694238397\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.018010251694238397\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.018010251694238397\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.018010251694238397\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.018010251694238397\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.018010251694238397\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): nan\n",
      "r0655 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0655): nan\n",
      "r0655 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.018010251694238397\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.018010251694238397\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): nan\n",
      "OIVD1m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD1m): 0.018010251694238397\n",
      "OIVD1m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.018010251694238397\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.018010251694238397\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "---------------------------------------\n",
      "------------WT_P4_Bio141------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "---------------------------------------\n",
      "GR: 0.035864607666101986\n",
      "GR2 (After knocking out SCP22x): 0.035864607666101986\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.035864607666101986\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.035864607666101986\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1453): 0.035864607666101986\n",
      "r1453 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.035864607666101986\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.035864607666101986\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out GapFill-R08726): nan\n",
      "GapFill-R08726 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible GapFill-R08726): nan\n",
      "GapFill-R08726 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.035864607666101986\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.035864607666101986\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.035864607666101986\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.035864607666101986\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.035864607666101986\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.035864607666101986\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.035864607666101986\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.035864607666101986\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.035864607666101986\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.035864607666101986\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.035864607666101986\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.035864607666101986\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.035864607666101986\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.035864607666101986\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.035864607666101986\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "---------------------------------------\n",
      "------------WT_P4_Bio143------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "---------------------------------------\n",
      "GR: 0.033931420343137106\n",
      "GR2 (After knocking out SCP22x): 0.03393142034313709\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.03393142034313709\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.03393142034313709\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.0339314203431371\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.0339314203431371\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out GapFill-R08726): nan\n",
      "GapFill-R08726 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible GapFill-R08726): nan\n",
      "GapFill-R08726 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.033931420343137106\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.0339314203431371\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.03393142034313709\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.0339314203431371\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.0339314203431371\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.03393142034313709\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.0339314203431371\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.03393142034313709\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.03393142034313709\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.033931420343137106\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.0339314203431371\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.03393142034313709\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.03393142034313709\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): nan\n",
      "OIVD1m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD1m): 0.03393142034313709\n",
      "OIVD1m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.03393142034313709\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.03393142034313709\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.033931420343137106\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "---------------------------------------\n",
      "------------WT_P4_Bio142------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "---------------------------------------\n",
      "GR: 0.035864607666101986\n",
      "GR2 (After knocking out SCP22x): 0.035864607666101986\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.035864607666101986\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.035864607666101986\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1453): 0.035864607666101986\n",
      "r1453 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.035864607666101986\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.035864607666101986\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.035864607666101986\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.035864607666101986\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.035864607666101986\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE2439C): nan\n",
      "RE2439C causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible RE2439C): nan\n",
      "RE2439C in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.035864607666101986\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.035864607666101986\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.035864607666101986\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.035864607666101986\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.035864607666101986\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.035864607666101986\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.035864607666101986\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.035864607666101986\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.035864607666101986\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.035864607666101986\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.035864607666101986\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): nan\n",
      "OIVD3m causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible OIVD3m): 0.035864607666101986\n",
      "OIVD3m successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.035864607666101986\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.035864607666101986\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "---------------------------------------\n",
      "------------WT_P6_Bio141------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "---------------------------------------\n",
      "GR: 0.01710810443493256\n",
      "GR2 (After knocking out SCP22x): 0.01710810443493256\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.01710810443493256\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.01710810443493256\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.01710810443493256\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.01710810443493255\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out GapFill-R08726): nan\n",
      "GapFill-R08726 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible GapFill-R08726): nan\n",
      "GapFill-R08726 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.017108104434932776\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.01710810443493277\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.01710810443493277\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1450): 0.01710810443493273\n",
      "r1450 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.01710810443493273\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.017108104434932693\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.017108104434932544\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.017108104434932544\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.017108104434932558\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.017108104434932558\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.017108104434932558\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.01710810443493256\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.017108104434932544\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.017108104434932544\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.017108104434932544\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.017108104434932544\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): 0.017108104434932544\n",
      "OIVD3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.017108104434932544\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0386): 0.01710810443493251\n",
      "r0386 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0451): 0.01710810443493254\n",
      "r0451 successfully knocked out without infeasibility.\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n",
      "---------------------------------------\n",
      "------------WT_P6_Bio142------------\n",
      "\n",
      "Objective:Maximize\n",
      "1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e\n",
      "---------------------------------------\n",
      "GR: 0.018010251694238397\n",
      "GR2 (After knocking out SCP22x): 0.018010251694238397\n",
      "SCP22x successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_0293): 0.018010251694238397\n",
      "HMR_0293 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_7741): 0.018010251694238397\n",
      "HMR_7741 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1393): 0.018010251694238397\n",
      "r1393 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out NICRNS): nan\n",
      "NICRNS causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible NICRNS): 0.018010251694238397\n",
      "NICRNS successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out GapFill-R08726): nan\n",
      "GapFill-R08726 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible GapFill-R08726): nan\n",
      "GapFill-R08726 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out RE2915M): 0.018010251694238397\n",
      "RE2915M successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3288): 0.018010251694238397\n",
      "HMR_3288 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_1325): 0.018010251694238397\n",
      "HMR_1325 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r1450): 0.018010251694238397\n",
      "r1450 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out RE3477C): 0.018010251694238397\n",
      "RE3477C successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AAPSAS): 0.018010251694238397\n",
      "AAPSAS successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0698): nan\n",
      "r0698 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0698): 0.018010251694238397\n",
      "r0698 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out 3HDH260p): 0.018010251694238397\n",
      "3HDH260p successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out HMR_3272): 0.018010251694238397\n",
      "HMR_3272 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out ACOAD183n3m): 0.018010251694238397\n",
      "ACOAD183n3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0655): 0.018010251694238397\n",
      "r0655 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0603): 0.018010251694238397\n",
      "r0603 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0541): nan\n",
      "r0541 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0541): nan\n",
      "r0541 in its irreversible format causes infeasibility.\n",
      "GR2 (After knocking out GapFill-R03599): 0.018010251694238397\n",
      "GapFill-R03599 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out AKGDm): 0.018010251694238397\n",
      "AKGDm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0556): 0.018010251694238397\n",
      "r0556 successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD1m): 0.018010251694238397\n",
      "OIVD1m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out OIVD3m): 0.018010251694238397\n",
      "OIVD3m successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out 2OXOADOXm): 0.018010251694238397\n",
      "2OXOADOXm successfully knocked out without infeasibility.\n",
      "GR2 (After knocking out r0386): nan\n",
      "r0386 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0386): 0.018010251694238397\n",
      "r0386 successfully transformed to reversible without infeasibility.\n",
      "GR2 (After knocking out r0451): nan\n",
      "r0451 causes infeasibility. Reverting knockout.\n",
      "GR3 (After making reaction irreversible r0451): 0.018010251694238397\n",
      "r0451 successfully transformed to reversible without infeasibility.\n",
      "EX_mal_L_e\n",
      "exp_growth_rate\n",
      "---------------------------------------\n",
      "---------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Sanity Check. Make sure all the exchange reactions in the experimental data are present in the model and that the models ar growing\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('')\n",
    "    print(f'GR1: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    \n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "\n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_loop_reactions:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            new_gr = model.slim_optimize()\n",
    "            print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "            \n",
    "            # Check if the model's growth rate is infeasible or close to zero\n",
    "            if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                reaction.bounds = (0,1000)  # Set reaction to irreversible\n",
    "\n",
    "                new_gr_3 = model.slim_optimize() # Optimize model with irreversible reaction\n",
    "                print(f'GR3 (After making reaction irreversible {atp_reaction}): {new_gr_3}')\n",
    "\n",
    "                # Check if the model's growth rate 3 is infeasible or close to zero\n",
    "                if math.isnan(new_gr_3) or new_gr_3 < 1e-6:\n",
    "                    print(f'{atp_reaction} in its irreversible format causes infeasibility.')\n",
    "                    reaction.bounds = original_bounds\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully transformed to reversible without infeasibility.')\n",
    "                    \n",
    "            else:\n",
    "                print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "\n",
    "                \n",
    "    print('')\n",
    "    print('Missing reactions from exp data:')\n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')\n",
    "    print('')\n",
    "\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('---------------------------------------')\n",
    "    print(f'GR: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "\n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_loop_reactions:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            new_gr = model.slim_optimize()\n",
    "            print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "            \n",
    "            # Check if the model's growth rate is infeasible or close to zero\n",
    "            if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                reaction.bounds = (0,1000)  # Set reaction to irreversible\n",
    "\n",
    "                new_gr_3 = model.slim_optimize() # Optimize model with irreversible reaction\n",
    "                print(f'GR3 (After making reaction irreversible {atp_reaction}): {new_gr_3}')\n",
    "\n",
    "                # Check if the model's growth rate 3 is infeasible or close to zero\n",
    "                if math.isnan(new_gr_3) or new_gr_3 < 1e-6:\n",
    "                    print(f'{atp_reaction} in its irreversible format causes infeasibility.')\n",
    "                    reaction.bounds = original_bounds\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully transformed to reversible without infeasibility.')\n",
    "                    \n",
    "            else:\n",
    "                print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "    \n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Flux Sampling Simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing model: WT_P6_Bio143\n",
      "Calculating Growth Rate for WT Condition:('U3', 'P4 to P6')\n",
      "Experimental growth rate is: 0.01611849093438101\n",
      "Infeasible solution for replicate U3 interval P4 to P6\n",
      "Processing model: WT_P4_Bio141\n",
      "Calculating Growth Rate for WT Condition:('U1', 'P2 to P4')\n",
      "Experimental growth rate is: 0.03278062118952783\n",
      "Simulated growth rate is: 0.035864607666101986\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpcnuglf4v.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 3021 rows, 8990 columns, 40850 nonzeros\n",
      "Processing model: WT_P4_Bio143\n",
      "Calculating Growth Rate for WT Condition:('U3', 'P2 to P4')\n",
      "Experimental growth rate is: 0.035978739841978656\n",
      "Infeasible solution for replicate U3 interval P2 to P4\n",
      "Processing model: WT_P4_Bio142\n",
      "Calculating Growth Rate for WT Condition:('U2', 'P2 to P4')\n",
      "Experimental growth rate is: 0.033990046589398894\n",
      "Simulated growth rate is: 0.035864607666101986\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp_ph_dsbo.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2981 rows, 8914 columns, 40146 nonzeros\n",
      "Processing model: WT_P6_Bio141\n",
      "Calculating Growth Rate for WT Condition:('U1', 'P4 to P6')\n",
      "Experimental growth rate is: 0.016358716703023954\n",
      "Simulated growth rate is: 0.018010251694238397\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpqwumulev.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3200 rows, 9648 columns, 43810 nonzeros\n",
      "Processing model: WT_P6_Bio142\n",
      "Calculating Growth Rate for WT Condition:('U2', 'P4 to P6')\n",
      "Experimental growth rate is: 0.018176950468208428\n",
      "Infeasible solution for replicate U2 interval P4 to P6\n"
     ]
    }
   ],
   "source": [
    "# WT\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P2 to P4':'P4', 'P4 to P6':'P6'}\n",
    "replicates = {'U1':'Bio141', 'U2':'Bio142', 'U3':'Bio143'}\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "\n",
    "                     with model as modified_model:\n",
    "                         \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "                        \n",
    "                        print(f'Calculating Growth Rate for WT Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_wt['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            for r,v in uptsec_wt.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "            \n",
    "                            # If pFBA succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = modified_model\n",
    "\n",
    "                            # Perform flux sampling using ACHRSampler\n",
    "                            print(\"Performing flux sampling...\")\n",
    "                            achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                            samples = 600 \n",
    "                            sampled_fluxes = achr_sampler.sample(samples)\n",
    "                            \n",
    "                            # Validate and save the sampled fluxes\n",
    "                            if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                                print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                            else:\n",
    "                                f_name = f\"../Analyses/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                                with open(f_name, 'wb') as file:\n",
    "                                    pickle.dump(sampled_fluxes, file)\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing model: ZeLa_P6_Bio144\n",
      "Calculating Growth Rate for WT Condition:('U4', 'P4 to P6')\n",
      "Experimental growth rate is: 0.012617780920061862\n",
      "Simulated growth rate is: 0.01296261076963557\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmplhuvs5sk.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 3113 rows, 9264 columns, 41688 nonzeros\n",
      "Processing model: ZeLa_P6_Bio146\n",
      "Calculating Growth Rate for WT Condition:('U6', 'P4 to P6')\n",
      "Experimental growth rate is: 0.014713581054016853\n",
      "Simulated growth rate is: 0.015744603050483905\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpm65dyimx.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3191 rows, 9602 columns, 43442 nonzeros\n",
      "Processing model: ZeLa_P6_Bio148\n",
      "Calculating Growth Rate for WT Condition:('U8', 'P4 to P6')\n",
      "Experimental growth rate is: 0.020921091667455156\n",
      "Simulated growth rate is: 0.014279302629349113\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpt3_p2fmz.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3179 rows, 9484 columns, 42904 nonzeros\n",
      "Processing model: ZeLa_P4_Bio145\n",
      "Calculating Growth Rate for WT Condition:('U5', 'P2 to P4')\n",
      "Experimental growth rate is: 0.03634059690486983\n",
      "Infeasible solution for replicate U5 interval P2 to P4\n",
      "Processing model: ZeLa_P4_Bio144\n",
      "Calculating Growth Rate for WT Condition:('U4', 'P2 to P4')\n",
      "Experimental growth rate is: 0.036151474706364185\n",
      "Simulated growth rate is: 0.03725896271642408\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpvab_1c99.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2778 rows, 7554 columns, 33692 nonzeros\n",
      "Processing model: ZeLa_P4_Bio146\n",
      "Calculating Growth Rate for WT Condition:('U6', 'P2 to P4')\n",
      "Experimental growth rate is: 0.031590486533636726\n",
      "Infeasible solution for replicate U6 interval P2 to P4\n",
      "Processing model: ZeLa_P4_Bio148\n",
      "Calculating Growth Rate for WT Condition:('U8', 'P2 to P4')\n",
      "Experimental growth rate is: 0.03665450181291928\n",
      "Simulated growth rate is: 0.03677572172477863\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpobkaeedw.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3027 rows, 9064 columns, 41074 nonzeros\n",
      "Processing model: ZeLa_P4_Bio147\n",
      "Calculating Growth Rate for WT Condition:('U7', 'P2 to P4')\n",
      "Experimental growth rate is: 0.03514420417641855\n",
      "Simulated growth rate is: 0.03725896271642408\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpqc3oxbkn.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3100 rows, 9246 columns, 41640 nonzeros\n",
      "Processing model: ZeLa_P6_Bio147\n",
      "Calculating Growth Rate for WT Condition:('U7', 'P4 to P6')\n",
      "Experimental growth rate is: 0.013577413497102097\n",
      "Simulated growth rate is: 0.018563977057166503\n",
      "Performing flux sampling...\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp3e_sfiu6.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 3013 rows, 8920 columns, 40224 nonzeros\n"
     ]
    }
   ],
   "source": [
    "# ZeLa\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P2 to P4':'P4', 'P4 to P6':'P6'}\n",
    "replicates = {'U4':'Bio144', 'U5':'Bio145', 'U6':'Bio146', 'U7':'Bio147', 'U8':'Bio148'}\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "                    \n",
    "                     with model as modified_model:\n",
    "                         \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "                        \n",
    "                        print(f'Calculating Growth Rate for WT Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_zela['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            for r,v in uptsec_zela.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "            \n",
    "                            # If pFBA succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = modified_model\n",
    "                            \n",
    "                            # Perform flux sampling using ACHRSampler\n",
    "                            print(\"Performing flux sampling...\")\n",
    "                            achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                            samples = 600 \n",
    "                            sampled_fluxes = achr_sampler.sample(samples)\n",
    "                            \n",
    "                            # Validate and save the sampled fluxes\n",
    "                            if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                                print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                            else:\n",
    "                                f_name = f\"../Analyses/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                                with open(f_name, 'wb') as file:\n",
    "                                    pickle.dump(sampled_fluxes, file)\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Pathway Enrichment Analysis <a id='pea'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 1\n",
    "# Use the fluxes from the growth rate calculations to filter reactions with fluxes\n",
    "# Define a cut off for the flux ??? # Plot to visualization of the flux distribution\n",
    "# Run Flux Enrichment Analysis on those reactions to see pathways enriched when optimized by biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 2\n",
    "# Transcriptomic data from cell batches / biorreactiors\n",
    "# Overlay this into the recons and extract the reactions associated to the genes\n",
    "# Extract a reaction vector / \n",
    "# Run Flux Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import hypergeom\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def flux_enrichment_analysis(generic_model, rxn_list, attribute='subsystem'):\n",
    "    \n",
    "    # Check if the attribute exists in the first reaction as a proxy for all\n",
    "    if not hasattr(model.reactions[0], attribute):\n",
    "        raise ValueError(f'Attribute {attribute} not found in model reactions')\n",
    "\n",
    "    # Extract attribute information for all reactions\n",
    "    attribute_values = [getattr(rxn, attribute, 'None') for rxn in model.reactions]\n",
    "    unique_attributes = set(attribute_values)\n",
    "    \n",
    "    # Count occurrences in the model and in the reaction set\n",
    "    model_counts = {attr: attribute_values.count(attr) for attr in unique_attributes}\n",
    "    rxn_set_counts = {attr: 0 for attr in unique_attributes}\n",
    "    for rxn in rxn_list:\n",
    "        rxn_attr = getattr(model.reactions.get_by_id(rxn), attribute, 'None')\n",
    "        rxn_set_counts[rxn_attr] += 1\n",
    "\n",
    "    # Calculate p-values using hypergeometric test\n",
    "    M = len(model.reactions)  # Total number of reactions\n",
    "    n = len(rxn_list)  # Size of reaction set\n",
    "    p_values = []\n",
    "    for attr in unique_attributes:\n",
    "        N = model_counts[attr]  # Total reactions in group\n",
    "        x = rxn_set_counts[attr]  # Reactions in group and in set\n",
    "        p_value = hypergeom.sf(x-1, M, N, n)\n",
    "        p_values.append(p_value)\n",
    "\n",
    "    # Adjust p-values for multiple testing\n",
    "    _, adj_p_values, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "    # Compile results\n",
    "    results = pd.DataFrame({\n",
    "        'Group': list(unique_attributes),\n",
    "        'P-value': p_values,\n",
    "        'Adjusted P-value': adj_p_values,\n",
    "        'Enriched set size': [rxn_set_counts[attr] for attr in unique_attributes],\n",
    "        'Total set size': [model_counts[attr] for attr in unique_attributes],\n",
    "    }).sort_values(by='Adjusted P-value')\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test 1: FEA on reactions active during growth rate optimization\n",
    "Use the fluxes from the growth rate calculations to filter reactions with fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate a vector of active reactions for each one of the conditions\n",
    "\n",
    "active_reactions = []\n",
    "\n",
    "for sol in pfba_solutions_fluxes:\n",
    "    rxns_fluxes = []\n",
    "    for n,f in sol['fluxes'].items():\n",
    "        if f != 0:\n",
    "            rxns_fluxes.append(n)\n",
    "    \n",
    "    active_reactions.append({\n",
    "        \"batch\": sol['model'],\n",
    "        \"condition\": sol['condition'],\n",
    "        \"flux_vector\": rxns_fluxes\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load generic model\n",
    "model = load_json_model('iCHO3595_unblocked.json')\n",
    "\n",
    "# Add atp demand reaction\n",
    "DM_atp = Reaction('DM_atp_c')\n",
    "model.add_reactions([DM_atp])\n",
    "model.reactions.DM_atp_c.build_reaction_from_string('atp_c -->')\n",
    "model.reactions.DM_atp_c.bounds = (0.01, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate FEA results for each one of the vectors generated for each condition\n",
    "\n",
    "fea_results = []\n",
    "\n",
    "# Load generic model\n",
    "#model = load_json_model('iCHO3595_unblocked.json')\n",
    "\n",
    "for fluxes in active_reactions:\n",
    "    results = flux_enrichment_analysis(model, fluxes['flux_vector'], 'subsystem')\n",
    "    fea_results.append({\n",
    "        \"Batch\": fluxes['batch'],\n",
    "        \"Condition\": fluxes['condition'],\n",
    "        \"Results\": results\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Transform p-values to -log10\n",
    "\n",
    "all_data = pd.DataFrame()\n",
    "\n",
    "for result in fea_results:\n",
    "    batch = result['Batch']\n",
    "    temp_df = result['Results'][['Group', 'P-value']].copy()\n",
    "    temp_df.columns = ['Group', f'P-value_{batch}']\n",
    "    if all_data.empty:\n",
    "        all_data = temp_df\n",
    "    else:\n",
    "        all_data = pd.merge(all_data, temp_df, on='Group', how='outer')\n",
    "\n",
    "# Remove groups with a 0 value in all conditions (assuming a '0' value indicates non-significance)\n",
    "significant_filter = (all_data.drop(columns='Group') < 0.05).any(axis=1)\n",
    "filtered_data = all_data[significant_filter]\n",
    "filtered_data.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# Transform the p-values\n",
    "for col in filtered_data.columns:\n",
    "    if col.startswith('P-value'):\n",
    "        filtered_data[f'-log10_{col}'] = -np.log10(filtered_data[col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import zscore\n",
    "\n",
    "# Extract relevant columns for heatmap\n",
    "heatmap_data = filtered_data[[col for col in filtered_data.columns if col.startswith('-log10')]]\n",
    "\n",
    "# Clean column names\n",
    "heatmap_data.columns = heatmap_data.columns.str.replace(r\"-log10_P-value_\", \"\", regex=True)\n",
    "\n",
    "# Normalize the data (z-score)\n",
    "normalized_data = heatmap_data.apply(zscore, axis=1)\n",
    "\n",
    "# Prepare heatmap data\n",
    "normalized_data.index = filtered_data['Group']\n",
    "\n",
    "# Reorder columns: Group P2, P4, P6, and P8 together\n",
    "columns_order = []\n",
    "for phase in ['P2', 'P4', 'P6', 'P8']:\n",
    "    columns_order.extend([col for col in normalized_data.columns if phase in col])\n",
    "\n",
    "# Reorder the DataFrame\n",
    "normalized_data = normalized_data[columns_order]\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(20, 20))\n",
    "sns.heatmap(normalized_data, annot=True, cmap=\"viridis\", cbar_kws={'label': 'Z-score of -log10(P-value)'}, vmin=-1.5, vmax=1.5)\n",
    "plt.title('Heatmap of Z-score Normalized -log10(P-values) Across Conditions', fontsize=20)\n",
    "plt.xlabel('Conditions', fontsize=15)\n",
    "plt.ylabel('Groups', fontsize=15)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Simulations/flux_enrichment_analysis/FEA_heatmap_all.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the columns based on the presence of \"P2 to P4\" or \"P4 to P6\"\n",
    "subset_df = heatmap_data.filter(regex='_P4_|_P6_')\n",
    "\n",
    "# Normalize the data (z-score)\n",
    "normalized_data = subset_df.apply(zscore, axis=1)\n",
    "\n",
    "# Prepare heatmap data\n",
    "normalized_data.index = filtered_data['Group']\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(20, 20))\n",
    "sns.heatmap(normalized_data, annot=True, cmap=\"viridis\", cbar_kws={'label': 'Z-score of -log10(P-value)'}, vmin=-1.5, vmax=1.5)\n",
    "plt.title('Heatmap of Z-score Normalized -log10(P-values) Across Conditions', fontsize=20)\n",
    "plt.xlabel('Conditions', fontsize=15)\n",
    "plt.ylabel('Groups', fontsize=15)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Simulations/flux_enrichment_analysis/FEA_heatmap_p2_p6.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
