{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Tests\n",
    "\n",
    "This notebook is intended for running different kinds of analyses that would validate our reconstruction.\n",
    "\n",
    "[1. Subsystem Overview and Analysis](#subsystems) <br>\n",
    "[2. Context-specific Model Generation](#context_specific) <br>\n",
    "[3. Biomass prediction using exp. data](#biomass) <br>\n",
    "[4. Flux Enrichment Analysis](#fea) <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Subsystem Overview and Analysis <a id='subsystems'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from google_sheet import GoogleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate Subsystems datasets from Google Sheet ----- #####\n",
    "\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_subsystems = 'Summary Systems'\n",
    "sheet_reactions = 'Rxns'\n",
    "subsystems = sheet.read_google_sheet(sheet_subsystems)\n",
    "reactions_rec = sheet.read_google_sheet(sheet_reactions)\n",
    "\n",
    "# Remove the total count\n",
    "subsystems = subsystems.iloc[:-1, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Pie Chart of the Amount of Reactions per System --- ####\n",
    "\n",
    "subsystems[\"Number of Reactions\"] = pd.to_numeric(subsystems[\"Number of Reactions\"])\n",
    "\n",
    "# Aggregate data by \"System\" and sum \"Number of Reactions\"\n",
    "system_reactions = subsystems.groupby(\"System\")[\"Number of Reactions\"].sum()\n",
    "\n",
    "# Set the figure and axes for more control\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "\n",
    "# Generate the pie chart again with the custom labels\n",
    "patches, texts, autotexts = ax.pie(system_reactions, labels=system_reactions.index, autopct=lambda p: '{:.1f}%'.format(p) if p > 0 else '', startangle=140, colors=plt.cm.tab20.colors)\n",
    "\n",
    "# Improve aesthetics\n",
    "for text in texts + autotexts:\n",
    "    text.set_color('black')\n",
    "ax.set_title(\"Pie Chart of Number of Reactions by System\", pad=30)  # Move the title upwards by increasing pad\n",
    "\n",
    "# Set equal aspect ratio\n",
    "ax.axis('equal')\n",
    "\n",
    "# Save the plot\n",
    "plt.savefig('../Plots/pie_chart_reactions_per_system.png')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the Sunburst Plot\n",
    "\n",
    "fig = px.sunburst(subsystems, path=['System', 'Subsystems', 'Kegg Pathway'],\n",
    "                  color='System')\n",
    "\n",
    "# Convert to a Graph Objects figure\n",
    "fig_go = go.Figure(fig)\n",
    "\n",
    "# Update layout\n",
    "fig_go.update_layout(width=1200, height=1000)\n",
    "\n",
    "# Define font sizes\n",
    "font_size_last_layer = 10\n",
    "font_size_previous_layer = 15\n",
    "default_font_size = 12\n",
    "\n",
    "# Create a list to store font sizes\n",
    "font_sizes = []\n",
    "\n",
    "# Update font size for each level\n",
    "for trace in fig_go.data:\n",
    "    if isinstance(trace, go.Sunburst):\n",
    "        for id in trace.ids:\n",
    "            level = id.count(\"/\")  # Determine level by the number of slashes in the id\n",
    "            if level == 2:  # Last layer (Kegg Pathway)\n",
    "                font_sizes.append(font_size_last_layer)\n",
    "            elif level == 1:  # Previous layer (Subsystems)\n",
    "                font_sizes.append(font_size_previous_layer)\n",
    "            else:\n",
    "                font_sizes.append(default_font_size)  # Default size for other layers\n",
    "\n",
    "# Apply the font sizes to the figure\n",
    "fig_go.update_traces(insidetextfont=dict(size=font_sizes))\n",
    "\n",
    "# Save the figure\n",
    "fig_go.write_html(\"../Plots/sunburst_subsystems.html\")  # Save as interactive HTML file\n",
    "fig_go.write_image(\"../Plots/sunburst_subsystems.png\", width=1200, height=1000, scale=2)  # Increase resolution by setting scale parameter\n",
    "\n",
    "# Show the plot\n",
    "fig_go.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Auxotrophies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.io import load_json_model\n",
    "\n",
    "iCHO_path = \"iCHO3595.json\"\n",
    "iCHO = load_json_model(iCHO_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids = {\n",
    "    \"arginine\": [\"EX_arg_L_e\", \"EX_arg_D_e\"],\n",
    "    \"asparagine\": [\"EX_asn_L_e\", \"EX_asn_D_e\"],\n",
    "    \"cysteine\": [\"EX_cys_L_e\", \"EX_cys_D_e\", \"EX_Lcystin_e\"],\n",
    "    \"histidine\": [\"EX_his_L_e\", \"EX_his_D_e\"],\n",
    "    \"isoleucine\": [\"EX_ile_L_e\", \"EX_ile_D_e\"],\n",
    "    \"leucine\": [\"EX_leu_L_e\", \"EX_leu_D_e\"],\n",
    "    \"lysine\": [\"EX_lys_L_e\", \"EX_lys_D_e\"],\n",
    "    \"methionine\": [\"EX_met_L_e\", \"EX_met_D_e\"],\n",
    "    \"phenylalanine\": [\"EX_phe_L_e\", \"EX_phe_D_e\"],\n",
    "    \"proline\": [\"EX_pro_L_e\", \"EX_pro_D_e\"],\n",
    "    \"threonine\": [\"EX_thr_L_e\", \"EX_thr_D_e\"],\n",
    "    \"tryptophan\": [\"EX_trp_L_e\", \"EX_trp_D_e\"],\n",
    "    \"valine\": [\"EX_val_L_e\", \"EX_val_D_e\"]\n",
    "}\n",
    "for amino_acid in amino_acids:\n",
    "    # ----- Setup initial bounds -----\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        exchange_reaction.bounds = -10, 10\n",
    "\n",
    "    # Arginine\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00557').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00558').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GLYAMDTRc').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ARGSL').bounds = 0, 0\n",
    "\n",
    "    # Asparigine\n",
    "    iCHO.reactions.get_by_id('ASNS1').bounds = 0, 0\n",
    "\n",
    "    # Cysteine\n",
    "    iCHO.reactions.get_by_id('r0129').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_cgly_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECG').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECGe').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('CYSTGL').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('EX_HC00250_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_sfcys_e').bounds = 0, 0\n",
    "\n",
    "    # Histidine\n",
    "    iCHO.reactions.get_by_id('VALTRPVALr').bounds = 0,0 \n",
    "\n",
    "    # Isoleucine - DONE\n",
    "    iCHO.reactions.get_by_id('EX_CE2916_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2915_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('ILETA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ILETAm').bounds = 0, 10\n",
    "\n",
    "    # Leucine\n",
    "    iCHO.reactions.get_by_id('LEULEULAPc').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_leugly_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_glyleu_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('LEUTA').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('LEUTAm').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('EX_CE5797_e').bounds = 0, 0\n",
    "\n",
    "    # Lysine\n",
    "    iCHO.reactions.get_by_id('EX_biocyt_e').bounds = 0,0 \n",
    "\n",
    "    # Methionine\n",
    "    iCHO.reactions.get_by_id('METS').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('BHMT').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('GAUGE-R00648').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('UNK2').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('UNK3').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('TYRA').bounds = -10, 0 \n",
    "    #iCHO.reactions.get_by_id('GAUGE-R06895').bounds = 0, 0 # Curated by MR with 1 score / Erased from the reconstruction\n",
    "\n",
    "    # Phenylalanine\n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_pheleu_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_glyphe_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2917_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5789_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_phpyr_e').bounds = 0, 0 \n",
    "\n",
    "    # Proline\n",
    "    iCHO.reactions.get_by_id('EX_glypro_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_progly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('P5CR').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRxm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRx').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('r1453').bounds = 0, 10\n",
    "\n",
    "    # Threonine\n",
    "    #iCHO.reactions.get_by_id('THRS').bounds = 0,0  # Erased from the reconstruction\n",
    "\n",
    "\n",
    "    # Tryptophan\n",
    "    iCHO.reactions.get_by_id('NBAHH_ir').bounds = 0,0  # Histidine hydrolase\n",
    "\n",
    "    # Valine\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_vallystyr_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTAm').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_valval_e').bounds = 0, 10\n",
    "    # -----------------------------------------\n",
    "\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        if exchange_reaction.id in amino_acids[amino_acid]:\n",
    "            exchange_reaction.bounds = 0, 10\n",
    "            # print(amino_acid, exchange_reaction.id, amino_acids[amino_acid])\n",
    "    sol = iCHO.optimize()\n",
    "    print(amino_acid, sol.objective_value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCHO.metabolites.leu_L_c.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Deletion test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cobra.io import load_json_model, read_sbml_model\n",
    "from cobra.flux_analysis import single_gene_deletion\n",
    "\n",
    "#iCHO = load_json_model('iCHO3595_unblocked.json')\n",
    "#iCHO1766 = read_sbml_model('../Data/Reconciliation/models/iCHOv1_final.xml')\n",
    "iCHO2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "#iCHO2441 = read_sbml_model('../Data/Reconciliation/models/iCHO2441.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_deletion_results = single_gene_deletion(iCHO2291)\n",
    "essentiality_threshold = 0.01\n",
    "essential_genes = gene_deletion_results[gene_deletion_results['growth'] < essentiality_threshold]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the essential genes\n",
    "print(f\"Number of essential genes for biomass: {len(essential_genes)}\")\n",
    "print(\"Essential genes for biomass production:\")\n",
    "for gene_idx in essential_genes.index:\n",
    "    gene = iCHO2291.genes[gene_idx]\n",
    "    print(f\"{gene.id}: {gene.name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.flux_analysis.variability import find_essential_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_essential_genes = find_essential_genes(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset for the experimentally validated essential genes\n",
    "\n",
    "exp_essential_genes = pd.read_csv('../Data/Gene_Essentiality/cho_essential_genes.csv', sep='\\t')\n",
    "exp_essential_genes['gene'] = exp_essential_genes['gene'].str.replace('__1$', '', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_essential_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract all gene names\n",
    "all_gene_names = set()\n",
    "for names in exp_essential_genes['gene']:\n",
    "    for name in names.split('|'): #Extract the names of the genes with two aliases\n",
    "        all_gene_names.add(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "essential_genes_model = []\n",
    "for g in model.genes:\n",
    "    if g.name in all_gene_names:\n",
    "        essential_genes_model.append(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set1 = set(sim_essential_genes)\n",
    "set2 = set(essential_genes_model)\n",
    "\n",
    "# Find shared elements\n",
    "shared_elements = set1.intersection(set2)\n",
    "\n",
    "# Find unique elements\n",
    "unique_in_list1 = set1.difference(set2)\n",
    "unique_in_list2 = set2.difference(set1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Shared genes between Exp Validated Essential Genes and Simulated Essential Genes:\", len(shared_elements))\n",
    "print(\"Unique genes in Simulated Essential Genes:\", len(unique_in_list1))\n",
    "print(\"Unique genes in Exp Validated Essentail Genes:\", len(unique_in_list2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Essential for growth genes and comparison to (An optimised genome-wide, virus free CRISPR screen for mammalian cells)\n",
    "https://pubmed.ncbi.nlm.nih.gov/34935002/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Context-specific Model Generation <a id='context_specific'></a>\n",
    "\n",
    "Here we use a matrix generated with rmf_CADRE to generate each context_specific model for each one of the conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from scipy.io import loadmat\n",
    "from time import process_time\n",
    "\n",
    "import pymCADRE\n",
    "from pymCADRE.rank import rank_reactions\n",
    "from pymCADRE.check import check_model_function\n",
    "from pymCADRE.prune import prune_model\n",
    "\n",
    "import cobra\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Intervals for Context-specific model generation\n",
    "Here we define the set of constraints in the uptake and secretion rates that we want to incorporate to our context specific models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Model Loading and Constraining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load generic model\n",
    "model = load_json_model('iCHO3595_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_dict = uptsec_intrvl_zela\n",
    "time = 'P2 to P4'\n",
    "\n",
    "with model as model:\n",
    "    model.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "    for rxn in model.boundary:\n",
    "        \n",
    "        # Models that are forced to secrete ethanol are not feasible\n",
    "        if rxn.id == 'EX_etoh_e':\n",
    "            rxn.bounds = (-0.1,0.1)\n",
    "            continue\n",
    "\n",
    "        # Replace the lower and upper bound with experimental data\n",
    "        if rxn.id in temp_dict.keys():\n",
    "            rxn.bounds = temp_dict[rxn.id][time]\n",
    "            continue\n",
    "        \n",
    "        # Keep boundaries open for essential metabolites\n",
    "        if rxn.id == 'EX_h2o_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_h_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_o2_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_hco3_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_so4_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_pi_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "\n",
    "        # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "        if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Tyr_ggn_c':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Ser_Thr_g':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_pre_prot_r':\n",
    "            rxn.bounds = (-0.0006,1000)\n",
    "            continue\n",
    "\n",
    "        # Close uptake rates for the rest of the boundaries\n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        \n",
    "    pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "    print(pfba_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading Ubiquity and Confidence Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the .mat file\n",
    "mat_data = loadmat('../Data/Context_specific_models/ubiData.mat')\n",
    "\n",
    "# Access the ubiData structure\n",
    "ubiData = mat_data['ubiData']\n",
    "\n",
    "# Accessing individual fields within ubiData\n",
    "ubiScores = ubiData['ubiScores'][0, 0]\n",
    "uScore = ubiData['uScore'][0, 0]\n",
    "rxns = ubiData['rxns'][0, 0]\n",
    "Condition = ubiData['Condition'][0, 0]\n",
    "\n",
    "# Print or process the data as needed\n",
    "print(ubiScores.shape)\n",
    "print(uScore.shape)\n",
    "print(rxns.shape)\n",
    "print(Condition.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,cond in enumerate(Condition[0]):\n",
    "    if \"WT_P4\" in cond[0]:\n",
    "        print(cond[0], i)\n",
    "    elif \"ZeLa_P4\" in cond[0]:\n",
    "        print(cond[0], i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Condition[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ubiquity score list\n",
    "U = list(ubiScores[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conf Score List\n",
    "conf_scores = pd.read_csv('../Data/Context_specific_models/confidence_scores.csv', header=None).iloc[:, 0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Protected Reations\n",
    "high_confidence_reactions = ['biomass_cho_s', 'DNAsyn', 'LipidSyn_cho_s', 'PROTsyn_cho_s', 'RNAsyn_cho_s', 'EX_bhb_e', 'EX_nh4_e', 'EX_ac_e', 'EX_ala_L_e',\n",
    "    'EX_arg_L_e', 'EX_asn_L_e', 'EX_asp_L_e', 'EX_2hb_e', 'EX_cit_e', 'EX_cys_L_e', 'EX_etoh_e', 'EX_for_e', 'EX_fum_e', 'EX_glc_e', 'EX_glu_L_e',\n",
    "    'EX_gln_L_e', 'EX_gly_e', 'EX_his_L_e', 'EX_4hpro_e', 'EX_ile_L_e', 'EX_lac_L_e', 'EX_leu_L_e', 'EX_lys_L_e', 'EX_mal_L_e', 'EX_met_L_e',\n",
    "    'EX_phe_L_e', 'EX_pro_L_e', 'EX_5oxpro_e', 'EX_pyr_e', 'EX_ser_L_e', 'EX_thr_L_e', 'EX_trp_L_e', 'EX_tyr_L_e', 'EX_val_L_e', 'EX_h2o_e',\n",
    "    'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e', 'SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reduced model generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################\n",
    "# Rank reactions\n",
    "##############################################\n",
    "\n",
    "print('Processing inputs and ranking reactions...')\n",
    "GM, C, NC, P, Z, model_C = rank_reactions(model, U, conf_scores, high_confidence_reactions, method=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output the results\n",
    "print(\"Generic Model:\", GM)\n",
    "print(\"Core Reactions:\", len(C))\n",
    "print(\"Non-Core Reactions:\", len(NC))\n",
    "print(\"Ranked Non-Core Reactions:\", len(P))\n",
    "print(\"Zero-Expression Reactions:\", len(Z))\n",
    "print(\"Core Reactions in Original Model:\", len(model_C))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.util import create_stoichiometric_matrix\n",
    "\n",
    "def retrieve_precursor_metabolites(model, protected_rxns):\n",
    "    \"\"\"\n",
    "    Retrieve precursor metabolites from a specified list of reactions in a metabolic model.\n",
    "\n",
    "    Args:\n",
    "        model: cobra.Model - The metabolic model.\n",
    "        protected_rxns: list - A list of reaction IDs for which precursor metabolites need to be retrieved.\n",
    "\n",
    "    Returns:\n",
    "        precursor_mets: list - A list of precursor metabolites.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a list of all the reactions in the model\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Ensure that the protected reactions list is not empty\n",
    "    if protected_rxns:\n",
    "        \n",
    "        #Define a list to store the indices of the protected reactions\n",
    "        rxn_indices = []\n",
    "        \n",
    "        for rxn in protected_rxns:\n",
    "            # Ensure that the protected reaction is included in the model\n",
    "            if rxn in model_reactions:\n",
    "                # Get the indices of the protected reactions in the model, only if the reaction exists\n",
    "                rxn_indices.append(model.reactions.index(model.reactions.get_by_id(rxn)))\n",
    "\n",
    "        # Create stoichiometric matrix S\n",
    "        S = create_stoichiometric_matrix(model)\n",
    "        \n",
    "        # Retrieve precursor metabolites: metabolites with a negative coefficient in the stoichiometric matrix\n",
    "        precursor_mets = [\n",
    "            model.metabolites[i] \n",
    "            for i in range(len(model.metabolites)) \n",
    "            if any(S[i, idx] < 0 for idx in rxn_indices)\n",
    "        ]\n",
    "\n",
    "        return precursor_mets\n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precursor_mets = retrieve_precursor_metabolites(model, high_confidence_reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# Define inputs to the model pruning step\n",
    "##################################################\n",
    "\n",
    "# Define core vs. non-core ratio threshold for removing reactions\n",
    "eta = 1 / 3\n",
    "\n",
    "# Check functionality of generic model\n",
    "genericStatus = check_model_function(GM, 'required_mets', precursor_mets)[0]\n",
    "print(genericStatus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constraint the model with exp data\n",
    "\n",
    "time = 'P2 to P4'\n",
    "\n",
    "GM.solver = 'gurobi'\n",
    "objective = 'biomass_cho_s' # 'biomass_cho'\n",
    "GM.objective = objective\n",
    "\n",
    "GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "for rxn in GM.boundary:\n",
    "    \n",
    "    # Models that are forced to secrete ethanol are not feasible\n",
    "    if rxn.id == 'EX_etoh_e':\n",
    "        rxn.bounds = (-0.1,0.1)\n",
    "        continue\n",
    "\n",
    "    # Replace the lower and upper bound with experimental data\n",
    "    if rxn.id in temp_dict.keys():\n",
    "        rxn.bounds = temp_dict[rxn.id][time]\n",
    "        continue\n",
    "    \n",
    "    # Keep boundaries open for essential metabolites\n",
    "    if rxn.id == 'EX_h2o_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_h_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_o2_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_hco3_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_so4_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_pi_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "\n",
    "    # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "    if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Tyr_ggn_c':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Ser_Thr_g':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_pre_prot_r':\n",
    "        rxn.bounds = (-0.0006,1000)\n",
    "        continue\n",
    "\n",
    "    # Close uptake rates for the rest of the boundaries\n",
    "    if rxn.id.startswith(\"EX_\"):\n",
    "        rxn.bounds = (0,1000) \n",
    "    if rxn.id.startswith(\"SK_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    if rxn.id.startswith(\"DM_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    \n",
    "pfba_solution = cobra.flux_analysis.pfba(GM)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(pfba_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if genericStatus:\n",
    "    print('Generic model passed precursor metabolites test')\n",
    "\n",
    "    ##############################################################\n",
    "    # If generic functionality test is passed, prune reactions\n",
    "    ###############################################################\n",
    "    print('Pruning reactions...')\n",
    "    t0 = process_time()\n",
    "    PM, cRes = prune_model(GM, P, C, Z, eta, precursor_mets, salvage_check=0, method=1)\n",
    "    # Stop the stopwatch / counter\n",
    "    t_stop = process_time()\n",
    "    # compute elapsed time\n",
    "    pruneTime = t_stop - t0\n",
    "\n",
    "else:\n",
    "    print('Generic model failed precursor metabolites test!!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ATP Generation loops removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import chain\n",
    "import pickle\n",
    "import numpy as np\n",
    "from optlang.symbolics import Zero\n",
    "\n",
    "import cobra\n",
    "from cobra.util import create_stoichiometric_matrix\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading and Constraining Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n"
     ]
    }
   ],
   "source": [
    "model = load_json_model('iCHO3595_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)\n",
    "\n",
    "for rxn in model.boundary:\n",
    "    if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "        rxn.bounds = (-1000, 1000)\n",
    "    elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "        rxn.bounds = (-0.001, 1000)\n",
    "    elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "        rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "for reaction in model.reactions:\n",
    "    if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "        continue\n",
    "    elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "        reaction.lower_bound = 0\n",
    "    for r,v in uptsec_zela.items():\n",
    "        if reaction.id == r:\n",
    "            reaction.upper_bound = 1000\n",
    "            reaction.lower_bound = v[('U6','P2 to P4')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ATP loops removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################\n",
    "# Thanasis' original version\n",
    "#########################################\n",
    "\n",
    "\n",
    "# Get the maximum flux through the oxidatitve phosphoryaltion pathway. Get the maximum flux of the objectives, biomass and IgG. \n",
    "# If a reaction that consumes ATP (negative stoichiometric coefficient), has a flux value below a cutoff in both maximisation of biomass\n",
    "# IgG and ox. pho. pathway, then the reaction doesn't provide any useable ATP (cyclic)\n",
    "\n",
    "step6_safe_rxn_rmv = []\n",
    "\n",
    "protected_subsystems = [ # Why protecting these subsystems\n",
    "    'BIOMASS', 'Product Formation', 'OXIDATIVE PHOSPHORYLATION', 'CITRIC ACID CYCLE', 'GLYCOLYSIS/GLUCONEOGENESIS',\n",
    "    'PYRUVATE METABOLISM', 'PENTOSE PHOSPHATE PATHWAY', 'GLUTAMATE METABOLISM', 'NAD METABOLISM',\n",
    "    'TRANSPORT, MITOCHONDRIAL', 'TRANSPORT, ENDOPLASMIC RETICULAR', 'ARGININE AND PROLINE METABOLISM',\n",
    "    'UBIQUINONE SYNTHESIS', 'VALINE, LEUCINE, AND ISOLEUCINE METABOLISM', 'TRYPTOPHAN METABOLISM',\n",
    "    'METHIONINE AND CYSTEINE METABOLISM', 'TYROSINE METABOLISM', 'GLYCINE, SERINE, AND THREONINE METABOLISM',\n",
    "    'HISTIDINE METABOLISM', 'ALANINE AND ASPARTATE METABOLISM', 'PHENYLALANINE METABOLISM',\n",
    "    'LYSINE METABOLISM', 'TYROSINE METABOLISM', 'N-GLYCAN BIOSYNTHESIS', 'PROTEIN PRODUCTION', 'SECRETORY PATHWAY'\n",
    "]\n",
    "\n",
    "\n",
    "# biomass and igg reactions of the model; we want to ensure their production after removals of any ATP cycles\n",
    "biomass_rxn = 'biomass_cho_s'\n",
    "igg_rxn = 'igg_formation'\n",
    "\n",
    "model.objective = biomass_rxn\n",
    "biomass_control = model.optimize('maximise')\n",
    "model.objective = igg_rxn\n",
    "igg_control = model.optimize('maximise')\n",
    "print(\"Original: Biomass: \", round(biomass_control.objective_value, 5), \"IgG: \", round(igg_control.objective_value, 5))\n",
    "\n",
    "model_temp = model.copy()\n",
    "atp_related_mets = ['atp_c', 'atp_n', 'atp_m', 'atp_x'] # What about atp_l, atp_g, atp_r, atp_e\n",
    "atp_met_indices = [model_temp.metabolites.get_by_id(met) for met in atp_related_mets]\n",
    "\n",
    "# Find the reactions associated with these metabolites\n",
    "atp_rxns_indices = []\n",
    "for met in atp_met_indices:\n",
    "    for reac in met.reactions:\n",
    "        atp_rxns_indices.append(reac.id)\n",
    "atp_rxns_indices = list(set(atp_rxns_indices))\n",
    "\n",
    "# Determine which reactions are consuming ATP (negative stoichiometry for ATP)\n",
    "atp_consumption_idx = []\n",
    "stoich_matrix = create_stoichiometric_matrix(model_temp)\n",
    "for atp_rxn in atp_rxns_indices:\n",
    "    reaction = model_temp.reactions.get_by_id(atp_rxn)\n",
    "    if not reaction.reversibility:  # Only irreversible reactions\n",
    "        for met in atp_met_indices:\n",
    "            if stoich_matrix[model_temp.metabolites.index(met), model_temp.reactions.index(reaction)] < 0:\n",
    "                atp_consumption_idx.append(atp_rxn)\n",
    "                break\n",
    "\n",
    "# Perform FBA to get maximum ATP production\n",
    "model_temp.objective = 0\n",
    "oxpho_rxns = [r for r in model_temp.reactions if 'OXIDATIVE PHOSPHORYLATION' in r.subsystem]\n",
    "reaction_variables = ((rxn.forward_variable, rxn.reverse_variable) for rxn in oxpho_rxns)\n",
    "variables = chain(*reaction_variables)\n",
    "model_temp.objective = model_temp.problem.Objective( Zero, direction=\"max\", sloppy=True, name=\"maxATP\" )\n",
    "model_temp.objective.set_linear_coefficients({v: 1.0 for v in variables})\n",
    "print(\"objective function\", model_temp.objective.expression)\n",
    "solution_maxATP = model_temp.optimize()\n",
    "\n",
    "for obj in [biomass_rxn, igg_rxn]:\n",
    "    model_temp.objective = obj\n",
    "    sol = model_temp.optimize('maximize')\n",
    "    atp_rate = np.vstack([sol[atp_consumption_idx], solution_maxATP[atp_consumption_idx]])\n",
    "\n",
    "    # Find ATP consuming reactions with low rates\n",
    "    atp_rxn_lowFlux = [id for id, rate in zip(atp_consumption_idx, np.sum(atp_rate, axis=0)) if rate <= 2e-12]\n",
    "\n",
    "    atp_loop_rxn_rmv = []\n",
    "    for rxn in atp_rxn_lowFlux:\n",
    "        if model_temp.reactions.get_by_id(rxn).subsystem not in protected_subsystems:\n",
    "            atp_loop_rxn_rmv.append(rxn)\n",
    "\n",
    "    # Filter out exchange reactions\n",
    "    exchange_reactions = [r.id for r in model_temp.exchanges]\n",
    "    atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id not in exchange_reactions]\n",
    "\n",
    "    # Ensure ATP maintenance reaction is not deleted\n",
    "    id_atp_maint = model_temp.reactions.get_by_id('ATPM').id\n",
    "    atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id != id_atp_maint]\n",
    "\n",
    "    if obj == biomass_rxn:\n",
    "        atp_loop_rxns_bio = atp_loop_rxn_rmv\n",
    "    elif obj == igg_rxn:\n",
    "        atp_loop_rxns_igg = atp_loop_rxn_rmv\n",
    "\n",
    "# Print the reactions to be deleted\n",
    "atp_loop_rxns = list(set(atp_loop_rxns_igg).intersection(atp_loop_rxns_bio))\n",
    "print(\"Reactions to be deleted:\", len(atp_loop_rxns))\n",
    "\n",
    "# Remove reactions from the model and check that Biomass and IgG are produced\n",
    "model_temp.objective = biomass_rxn\n",
    "sol_bio_control = model_temp.optimize('maximize')\n",
    "model_temp.objective = igg_rxn\n",
    "sol_igg_control = model_temp.optimize('maximize')\n",
    "model_temp_temp = model_temp.copy()\n",
    "\n",
    "for rxn_to_rmv in atp_loop_rxns:\n",
    "    model_temp_temp.remove_reactions(rxn_to_rmv)\n",
    "    model_temp_temp.objective = biomass_rxn\n",
    "    sol_bio = model_temp_temp.optimize('maximize')\n",
    "    model_temp_temp.objective = igg_rxn\n",
    "    sol_igg = model_temp_temp.optimize('maximize')\n",
    "    if round(sol_bio.objective_value, 5) == round(sol_bio_control.objective_value, 5) and round(sol_igg.objective_value, 5) == round(sol_igg_control.objective_value, 5):\n",
    "        step6_safe_rxn_rmv.append(rxn_to_rmv)\n",
    "\n",
    "model_temp.remove_reactions(step6_safe_rxn_rmv)\n",
    "print(\"Reactions consuming ATP in a cyclic fashion AND are safe to remove: \", len(step6_safe_rxn_rmv))\n",
    "print(\"Model size after the removal of ATP cyclic reactions: Reactions: \", len(model_temp.reactions), \"Metabolites: \", len(model_temp.metabolites))\n",
    "model_temp.objective = biomass_rxn\n",
    "sol_bio = model_temp.optimize('maximize')\n",
    "model_temp.objective = igg_rxn\n",
    "sol_igg = model_temp.optimize('maximize')\n",
    "print(\"Old Bio: \", round(sol_bio_control.objective_value, 3), \"New Bio: \", round(sol_bio.objective_value, 3), \"Old IgG: \", round(sol_igg_control.objective_value, 5), \"New IgG: \", round(sol_igg.objective_value, 5))\n",
    "model = model_temp.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpm5doitsd.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Original: Biomass: 0.04670, IgG: 0.00016\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ACHVESSEC not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r2517 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: R_group_phosphotase_1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FPGS6m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: TUDCA3Sabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CE7085te not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: SADT not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_6566 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RNMK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: 1MNCAMti not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HC02195te not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HEX2 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0427 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CTPS2 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ATAx not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FPGS5m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL180i not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0408 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: SPHK21n not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PNTK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL205x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1029 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL181i not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1253 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL203x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FPGS9m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DNDPt13m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0679 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NRVNCCOAtxc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PI3P4K not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_7755 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL226 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: 7DHCDCHOLabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1262 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DGNSKm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CE2211te not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: SPHK21c not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PI4P5K not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NMNATm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PROSTGE1t3 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: MEPIVESSte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ADNK1n not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: TXB2t not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1529 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL170 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CARNS1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NRVNCABCt not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL1832x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0253 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL1831x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RE2637C not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ISOCHOLabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PIK4 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PFK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: LEUKABCtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HEXDTRte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CYTDK2m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0423 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ESTRSABCtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_4261 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PSFLIP not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FPGS8 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0371 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GCA3Sabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: BALAVECSEC not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: COPROSTabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HDCA24Gte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GSNKm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PYDAMK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: OCDCAFATPc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL224 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: MINOHPKn not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DOCOSACT not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0321 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: XOL7AONEATP not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL120i not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: VLCSp not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NAXE1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL181x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ATAn not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RE1441G not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: LCA3Gte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DNDPt56m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL140i not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GLNS not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PI5P4K not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CGMPt not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ASCBSVCTtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PI3P4Kn not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: 5EIPENCte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DURIK1m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0299 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GCDCA3Sabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: LINOFATPtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GLCSGLT1le not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FAOXC16Brx not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL40im not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FAOXC16DCr not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DNDPt32m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DNDPt31m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GTHETCabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: R_group_phosphotase_3 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PIK3n not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ARACHCOAtcx not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DCA3Gte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: OLEICFATPtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CBPSam not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1094 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1341 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: C04483t1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DOCO13EFATP not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0680 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RE0124C not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PCSm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CA3Sabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0207 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PYDXK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CE7096te not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0239 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NAXE2 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL2251 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0303 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL244_1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL1812x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r2505 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1012 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: KHK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HYOCHOLabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CHOLK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL204x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL246_1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL180x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PALFATPtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HEX7 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: THYOCHOLabc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RBK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FPGSm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_7656 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: VLCSr not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: TCHOLABCtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: POMKr not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ACCOAL not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DNDPt20m not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL240x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r2499 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL100im not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PHYQt not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0435 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: MI13456PK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PPCOACm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: 5ADTSTSTERONEGLCte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RBFK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: EANDRSTRNtr not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL2042 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CAATPS not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ACCOACm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CAATPaser not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1536 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: GALK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL80x not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: CUATPase not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ESTRONEGLCt not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0387 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DURIK1 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RTOTAL2FATPc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NNATr not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: SPHK11 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL241 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0924 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: DTMPKm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: MALCL not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ADSKn not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: AK2LGCHOLABCt not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1515 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: LGNCCOAtcx not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL160i not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: 15HPETATP not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FOLABCCte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: TRIOK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1100 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: TSTSTERONEGLCte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1251 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PSFLIPm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: PCm not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r1254 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: XYLK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: NADK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r2518 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: AHANDROSTANtr not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0859 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HEX10 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0813 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_0291 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: r0033 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: XOLDIOLONEATP not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HMR_4278 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RETABCtc not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: HC02194te not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ETHAK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: RE1957R not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: ADSK not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: FACOAL1822 not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n",
      "/Users/pablodigiusto/miniforge3/envs/mncho3.0/lib/python3.12/site-packages/cobra/core/model.py:787: UserWarning: 11DOCRTSTRNte not in iCHO3595\n",
      "  warn(f\"{reaction} not in {self}\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 0.000 at 0x166392630>\n",
      "Reactions consuming ATP in a cyclic fashion and safe to remove: 203\n",
      "Old Bio: 0.047, New Bio: 0.039, Old IgG: 0.00016, New IgG: 0.00000\n"
     ]
    }
   ],
   "source": [
    "#########################################\n",
    "# Pablo's modification\n",
    "#########################################\n",
    "\n",
    "\n",
    "\n",
    "def get_atp_related_reactions(model, atp_mets):\n",
    "    \"\"\" Get reactions associated with ATP metabolites. \"\"\"\n",
    "    atp_rxns_indices = []\n",
    "    for met in atp_mets:\n",
    "        for reac in met.reactions:\n",
    "            atp_rxns_indices.append(reac.id)\n",
    "    return list(set(atp_rxns_indices))\n",
    "\n",
    "def find_low_flux_reactions(model, atp_rxns_indices, atp_met_indices, stoich_matrix, consumption_cutoff=2e-12):\n",
    "    \"\"\" Identify reactions with low ATP consumption flux. \"\"\"\n",
    "    atp_consumption_idx = []\n",
    "    for atp_rxn in atp_rxns_indices:\n",
    "        reaction = model.reactions.get_by_id(atp_rxn)\n",
    "        if not reaction.reversibility:  # Only irreversible reactions\n",
    "            for met in atp_met_indices:\n",
    "                if stoich_matrix[model.metabolites.index(met), model.reactions.index(reaction)] < 0:\n",
    "                    atp_consumption_idx.append(atp_rxn)\n",
    "                    break\n",
    "    return atp_consumption_idx\n",
    "\n",
    "def perform_fba_for_max_atp(model, oxpho_rxns):\n",
    "    \"\"\" Set FBA objective to maximize ATP production. \"\"\"\n",
    "    reaction_variables = ((rxn.forward_variable, rxn.reverse_variable) for rxn in oxpho_rxns)\n",
    "    variables = chain(*reaction_variables)\n",
    "    model.objective = model.problem.Objective(Zero, direction=\"max\", sloppy=True, name=\"maxATP\")\n",
    "    model.objective.set_linear_coefficients({v: 1.0 for v in variables})\n",
    "    return model.optimize()\n",
    "\n",
    "def remove_cyclic_atp_reactions(model, atp_rxns_to_remove, biomass_rxn, igg_rxn):\n",
    "    \"\"\" Remove reactions consuming ATP in a cyclic fashion and check for safety. \"\"\"\n",
    "    step6_safe_rxn_rmv = []\n",
    "\n",
    "    temp_model.objective = biomass_rxn\n",
    "    sol_bio_control = model.optimize('maximize')\n",
    "    temp_model.objective = igg_rxn\n",
    "    sol_igg_control = model.optimize('maximize')\n",
    "\n",
    "    for rxn_id in atp_rxns_to_remove:\n",
    "        model.remove_reactions([rxn_id])\n",
    "        temp_model.objective = biomass_rxn\n",
    "        sol_bio = model.optimize('maximize')\n",
    "        temp_model.objective = igg_rxn\n",
    "        sol_igg = model.optimize('maximize')\n",
    "        \n",
    "        if (round(sol_bio.objective_value, 5) == round(sol_bio_control.objective_value, 5) and \n",
    "            round(sol_igg.objective_value, 5) == round(sol_igg_control.objective_value, 5)):\n",
    "            step6_safe_rxn_rmv.append(rxn_id)\n",
    "\n",
    "    model.remove_reactions(step6_safe_rxn_rmv)\n",
    "    return step6_safe_rxn_rmv\n",
    "\n",
    "# Set objectives\n",
    "biomass_rxn = 'biomass_cho_s'\n",
    "igg_rxn = 'igg_formation'\n",
    "\n",
    "# Copy the original model\n",
    "temp_model = model.copy()\n",
    "\n",
    "# Get initial biomass and IgG flux\n",
    "temp_model.objective = biomass_rxn\n",
    "biomass_control = temp_model.optimize('maximize')\n",
    "temp_model.objective = igg_rxn\n",
    "igg_control = temp_model.optimize('maximize')\n",
    "print(f\"Original: Biomass: {biomass_control.objective_value:.5f}, IgG: {igg_control.objective_value:.5f}\")\n",
    "\n",
    "# Find ATP related reactions\n",
    "atp_related_mets = ['atp_c', 'atp_n', 'atp_m', 'atp_x', 'atp_l', 'atp_g', 'atp_r', 'atp_e']\n",
    "atp_met_indices = [temp_model.metabolites.get_by_id(met) for met in atp_related_mets]\n",
    "atp_rxns_indices = get_atp_related_reactions(temp_model, atp_met_indices)\n",
    "\n",
    "# Create stoichiometric matrix\n",
    "stoich_matrix = create_stoichiometric_matrix(temp_model)\n",
    "atp_consumption_idx = find_low_flux_reactions(temp_model, atp_rxns_indices, atp_met_indices, stoich_matrix)\n",
    "\n",
    "# Maximize ATP production\n",
    "oxpho_rxns = [r for r in temp_model.reactions if 'OXIDATIVE PHOSPHORYLATION' in r.subsystem]\n",
    "solution_maxATP = perform_fba_for_max_atp(temp_model, oxpho_rxns)\n",
    "\n",
    "# Check flux of ATP consuming reactions\n",
    "atp_loop_rxn_rmv = []\n",
    "for obj in [biomass_rxn, igg_rxn]:\n",
    "    temp_model.objective = obj\n",
    "    sol = temp_model.optimize('maximize')\n",
    "    atp_rate = np.vstack([sol[atp_consumption_idx], solution_maxATP[atp_consumption_idx]])\n",
    "\n",
    "    # Find low flux ATP reactions\n",
    "    atp_rxn_lowFlux = [rxn_id for rxn_id, rate in zip(atp_consumption_idx, np.sum(atp_rate, axis=0)) if rate <= 2e-12]\n",
    "\n",
    "    # Filter protected subsystems and exchanges\n",
    "    protected_subsystems = ['BIOMASS', 'OXIDATIVE PHOSPHORYLATION']\n",
    "    atp_loop_rxn_rmv += [rxn for rxn in atp_rxn_lowFlux if temp_model.reactions.get_by_id(rxn).subsystem not in protected_subsystems]\n",
    "\n",
    "# Ensure ATPM and exchange reactions are not removed\n",
    "atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id != 'ATPM' and not r_id.startswith(\"EX_\")]\n",
    "\n",
    "# Remove duplicates in the list\n",
    "atp_loop_rxn_rmv = set(atp_loop_rxn_rmv)\n",
    "\n",
    "# Remove cyclic ATP reactions and check safety\n",
    "safe_rxns_to_remove = remove_cyclic_atp_reactions(temp_model, atp_loop_rxn_rmv, biomass_rxn, igg_rxn)\n",
    "print(f\"Reactions consuming ATP in a cyclic fashion and safe to remove: {len(safe_rxns_to_remove)}\")\n",
    "\n",
    "# Final biomass and IgG\n",
    "temp_model.objective = biomass_rxn\n",
    "sol_bio = temp_model.optimize('maximize')\n",
    "temp_model.objective = igg_rxn\n",
    "sol_igg = temp_model.optimize('maximize')\n",
    "print(f\"Old Bio: {biomass_control.objective_value:.3f}, New Bio: {sol_bio.objective_value:.3f}, Old IgG: {igg_control.objective_value:.5f}, New IgG: {sol_igg.objective_value:.5f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulated ATP Maintencance is: 500.0\n",
      "Simulated growth rate is: 0.03893171377404033\n",
      "ATPM 500.0\n",
      "NACPRT -500.39396043463955\n",
      "r0391 -500.39396043463955\n"
     ]
    }
   ],
   "source": [
    "with temp_model as model:\n",
    "    objective = 'biomass_cho_s'\n",
    "    model.objective = objective\n",
    "    '''\n",
    "    model.reactions.SCP22x.knock_out()\n",
    "    model.reactions.TMNDNCCOAtx.knock_out()\n",
    "    model.reactions.OCCOAtx.knock_out()\n",
    "    model.reactions.TMNDNCCOAtx.knock_out()\n",
    "    model.reactions.r0391.knock_out()\n",
    "    model.reactions.BiGGRxn67.knock_out()\n",
    "    model.reactions.r2247.knock_out()\n",
    "    model.reactions.r2280.knock_out()\n",
    "    model.reactions.r2246.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.r2279.knock_out()\n",
    "    model.reactions.r2245.knock_out()\n",
    "    model.reactions.r2305.knock_out()\n",
    "    model.reactions.r2317.knock_out()\n",
    "    model.reactions.r2335.knock_out()\n",
    "    model.reactions.HMR_0293.knock_out()\n",
    "    model.reactions.HMR_7741.knock_out()\n",
    "    model.reactions.r0509.knock_out()\n",
    "    model.reactions.r1453.knock_out()\n",
    "    model.reactions.HMR_4343.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.ACONTm.knock_out()\n",
    "    model.reactions.PDHm.knock_out()\n",
    "    model.reactions.r0426.knock_out()\n",
    "    model.reactions.r0383.knock_out()\n",
    "    model.reactions.r0555.knock_out()\n",
    "    model.reactions.r1393.knock_out()\n",
    "    model.reactions.NICRNS.knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R00648').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R03326').knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R08726').knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.RE2915M.knock_out()\n",
    "    model.reactions.HMR_3288.knock_out()\n",
    "    model.reactions.HMR_1325.knock_out()\n",
    "    model.reactions.HMR_7599.knock_out()\n",
    "    model.reactions.r1431.knock_out()\n",
    "    model.reactions.r1433.knock_out()\n",
    "    model.reactions.RE2439C.knock_out()\n",
    "    model.reactions.r0082.knock_out()\n",
    "    model.reactions.r0791.knock_out()\n",
    "    model.reactions.r1450.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.get_by_id('GAUGE-R00270').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R02285').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R04283').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R06127').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R06128').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R06238').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R00524').knock_out()\n",
    "    model.reactions.RE3477C.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.AAPSAS.knock_out()\n",
    "    model.reactions.RE3347C.knock_out()\n",
    "    model.reactions.HMR_0960.knock_out()\n",
    "    model.reactions.HMR_0980.knock_out()\n",
    "    model.reactions.RE3476C.knock_out()\n",
    "    model.reactions.r0708.knock_out()\n",
    "    model.reactions.r0777.knock_out()\n",
    "    model.reactions.r0084.knock_out()\n",
    "    model.reactions.r0424.knock_out()\n",
    "    model.reactions.r0698.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.get_by_id('3HDH260p').knock_out()\n",
    "    model.reactions.HMR_3272.knock_out()\n",
    "    model.reactions.ACOAD183n3m.knock_out()\n",
    "    model.reactions.HMR_1996.knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R01463').knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R04807').knock_out()\n",
    "    model.reactions.r1468.knock_out()\n",
    "    model.reactions.r2435.knock_out()\n",
    "    model.reactions.r0655.knock_out()\n",
    "    model.reactions.r0603.knock_out()\n",
    "\n",
    "    model.reactions.r0541.knock_out()\n",
    "    model.reactions.RE0383C.knock_out()\n",
    "    model.reactions.HMR_1329.knock_out()\n",
    "    model.reactions.TYRA.knock_out()\n",
    "    model.reactions.NRPPHRt_2H.knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R07364').knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R03599').knock_out()\n",
    "    model.reactions.ARD.knock_out()\n",
    "    model.reactions.RE3095C.knock_out()\n",
    "    model.reactions.RE3104C.knock_out()\n",
    "    model.reactions.RE3104R.knock_out()\n",
    "\n",
    "    model.reactions.ACONT.knock_out()\n",
    "    model.reactions.ACONTm.knock_out()\n",
    "    model.reactions.ICDHxm.knock_out()\n",
    "    model.reactions.ICDHy.knock_out()\n",
    "    model.reactions.AKGDm.knock_out()\n",
    "    model.reactions.r0083.knock_out()\n",
    "    model.reactions.r0425.knock_out()\n",
    "    model.reactions.r0556.knock_out()\n",
    "    model.reactions.NH4t4r.knock_out()\n",
    "    model.reactions.PROPAT4te.knock_out()\n",
    "    \n",
    "    model.reactions.r0085.knock_out()\n",
    "    model.reactions.r0156.knock_out()\n",
    "    model.reactions.r0464.knock_out()\n",
    "    model.reactions.ABUTDm.knock_out()\n",
    "    model.reactions.OIVD1m.knock_out()\n",
    "    model.reactions.OIVD2m.knock_out()\n",
    "    model.reactions.OIVD3m.knock_out()\n",
    "    model.reactions.r2194.knock_out()\n",
    "    model.reactions.r2202.knock_out()\n",
    "    model.reactions.HMR_9617.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.r2197.knock_out()\n",
    "    model.reactions.r2195.knock_out()\n",
    "    model.reactions.get_by_id('2OXOADOXm').knock_out()\n",
    "    model.reactions.r2328.knock_out()\n",
    "    model.reactions.r0386.knock_out()\n",
    "    model.reactions.r0451.knock_out()\n",
    "    model.reactions.FAS100COA.knock_out()\n",
    "    model.reactions.FAS120COA.knock_out()\n",
    "    model.reactions.FAS140COA.knock_out()\n",
    "    model.reactions.FAS80COA_L.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.r0604.knock_out()\n",
    "    model.reactions.r0670.knock_out()\n",
    "    model.reactions.r2334.knock_out()\n",
    "    model.reactions.r0193.knock_out()\n",
    "    model.reactions.r0595.knock_out()\n",
    "    model.reactions.r0795.knock_out()\n",
    "    model.reactions.GLYCLm.knock_out()\n",
    "    model.reactions.MACACI.knock_out()\n",
    "    model.reactions.r2193.knock_out()\n",
    "    model.reactions.r0779.knock_out()\n",
    "    model.reactions.r0669.knock_out()\n",
    "    model.reactions.UDCHOLt.knock_out()\n",
    "    model.reactions.r2146.knock_out()\n",
    "    model.reactions.r2139.knock_out()\n",
    "    model.reactions.r0391.knock_out()\n",
    "    '''\n",
    "    for rxn in model.boundary:\n",
    "        if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "            rxn.bounds = (-1000, 1000)\n",
    "        elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "            rxn.bounds = (-0.001, 1000)\n",
    "        elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "            rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "            continue\n",
    "        elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "            reaction.lower_bound = 500\n",
    "        for r,v in uptsec_zela.items():\n",
    "            if reaction.id == r:\n",
    "                reaction.upper_bound = 1000\n",
    "                reaction.lower_bound = v[('U7','P2 to P4')]\n",
    "    pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "    pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "    print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "    print(f'Simulated growth rate is: {pfba_obj}')\n",
    "    \n",
    "    for i,f in pfba_solution.fluxes.items():\n",
    "        #if (reactions_rec['iCHO2291'][reactions_rec['Reaction'] == i] == '-').any():\n",
    "        if abs(f) > 1:\n",
    "            print(i,f)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Biomass prediction using exp. data <a id='biomass'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "import cobra\n",
    "from cobra import Reaction\n",
    "from cobra.io import load_json_model, read_sbml_model, load_matlab_model\n",
    "from cobra.exceptions import Infeasible\n",
    "from cobra.sampling import ACHRSampler\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory containing context-specific models models\n",
    "model_directory = '../Data/Context_specific_models/'\n",
    "\n",
    "# Model Names\n",
    "wt_model_names = ['WT_P4_Bio141', 'WT_P4_Bio142', 'WT_P4_Bio143',\n",
    "                  'WT_P6_Bio141', 'WT_P6_Bio142', 'WT_P6_Bio143']\n",
    "\n",
    "zela_model_names = ['ZeLa_P2_Bio145', 'ZeLa_P2_Bio146', 'ZeLa_P2_Bio147', 'ZeLa_P2_Bio148',\n",
    "                    'ZeLa_P4_Bio144', 'ZeLa_P4_Bio145', 'ZeLa_P4_Bio146', 'ZeLa_P4_Bio147', 'ZeLa_P4_Bio148',\n",
    "                    'ZeLa_P6_Bio144', 'ZeLa_P6_Bio146', 'ZeLa_P6_Bio147', 'ZeLa_P6_Bio148',\n",
    "                    'ZeLa_P8_Bio144', 'ZeLa_P8_Bio145', 'ZeLa_P8_Bio146', 'ZeLa_P8_Bio147', 'ZeLa_P8_Bio148']\n",
    "\n",
    "# Dictionaries to store the loaded models\n",
    "wt_models = {}\n",
    "zela_models = {}\n",
    "\n",
    "# List all .mat files in the directory\n",
    "model_files = [f for f in os.listdir(model_directory) if f.endswith('.mat')]\n",
    "\n",
    "# Iterate through the .mat files and load only the models that match predefined strings\n",
    "for model_file in model_files:\n",
    "    # Full path to the model file\n",
    "    model_path = os.path.join(model_directory, model_file)\n",
    "    \n",
    "    # Check if the file corresponds to a WT model\n",
    "    for model_name in wt_model_names:\n",
    "        if model_name in model_file:\n",
    "            wt_models[model_name] = load_matlab_model(model_path)\n",
    "            wt_models[model_name].id = model_name\n",
    "            print(f\"Loaded WT model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed\n",
    "    \n",
    "    # Check if the file corresponds to a ZeLa model\n",
    "    for model_name in zela_model_names:\n",
    "        if model_name in model_file:\n",
    "            zela_models[model_name] = load_matlab_model(model_path)\n",
    "            zela_models[model_name].id = model_name\n",
    "            print(f\"Loaded ZeLa model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATP loop generating reactions\n",
    "\n",
    "atp_reactions_list = [\n",
    "    'SCP22x', 'TMNDNCCOAtx', 'OCCOAtx', 'r0391', 'BiGGRxn67', 'r2247', 'r2280', 'r2246', 'r2279', 'r2245', \n",
    "    'r2305', 'r2317', 'r2335', 'HMR_0293', 'HMR_7741', 'r0509', 'r1453', 'HMR_4343', 'ACONTm', 'PDHm', \n",
    "    'r0426', 'r0383', 'r0555', 'r1393', 'NICRNS', 'GAUGE-R00648', 'GapFill-R08726', \n",
    "    'RE2915M', 'HMR_3288', 'HMR_1325', 'HMR_7599', 'r1431', 'r1433', 'RE2439C', 'r0082', 'r0791', 'r1450', \n",
    "    'GAUGE-R00270', 'GAUGE-R02285', 'GAUGE-R04283', 'GAUGE-R06127', 'GAUGE-R06128', 'GAUGE-R06238', \n",
    "    'GAUGE-R00524', 'GAUGE-R03326', 'RE3477C', 'AAPSAS', 'RE3347C', 'HMR_0960', 'HMR_0980', 'RE3476C', \n",
    "    'r0708', 'r0777', 'r0084', 'r0424', 'r0698', '3HDH260p', 'HMR_3272', 'ACOAD183n3m', 'HMR_1996', \n",
    "    'GapFill-R01463', 'GapFill-R04807', 'r1468', 'r2435', 'r0655', 'r0603', 'r0541', 'RE0383C', 'HMR_1329', \n",
    "    'TYRA', 'NRPPHRt_2H', 'GAUGE-R07364', 'GapFill-R03599', 'ARD', 'RE3095C', 'RE3104C', 'RE3104R', 'ACONT', \n",
    "    'ACONTm', 'ICDHxm', 'ICDHy', 'AKGDm', 'r0083', 'r0425', 'r0556', 'NH4t4r', 'PROPAT4te', 'r0085', \n",
    "    'r0156', 'r0464', 'ABUTDm', 'OIVD1m', 'OIVD2m', 'OIVD3m', 'r2194', 'r2202', 'HMR_9617', 'r2197', \n",
    "    'r2195', '2OXOADOXm', 'r2328', 'r0386', 'r0451', 'FAS100COA', 'FAS120COA', 'FAS140COA', 'FAS80COA_L', \n",
    "    'r0604', 'r0670', 'r2334', 'r0193', 'r0595', 'r0795', 'GLYCLm', 'MACACI', 'r2193', 'r0779', 'r0669', \n",
    "    'UDCHOLt', 'r2146', 'r2139'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Sanity Check. Make sure all the exchange reactions in the experimental data are present in the model and that the models ar growing\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('')\n",
    "    print(f'GR1: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    \n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_reactions_list:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            try:\n",
    "                new_gr = model.slim_optimize()\n",
    "                print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "                \n",
    "                # Check if the model's growth rate is infeasible or close to zero\n",
    "                if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                    print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                    reaction.bounds = original_bounds  # Revert the bounds if infeasible\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f'Error optimizing model after knocking out {atp_reaction}: {e}')\n",
    "                # Revert the reaction bounds if an error occurs\n",
    "                reaction.bounds = original_bounds\n",
    "\n",
    "    print('')\n",
    "    print('Missing reactions from exp data:')\n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')\n",
    "    print('')\n",
    "\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('---------------------------------------')\n",
    "    print(f'GR: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_reactions_list:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            try:\n",
    "                new_gr = model.slim_optimize()\n",
    "                print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "                \n",
    "                # Check if the model's growth rate is infeasible or close to zero\n",
    "                if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                    print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                    reaction.bounds = original_bounds  # Revert the bounds if infeasible\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f'Error optimizing model after knocking out {atp_reaction}: {e}')\n",
    "                # Revert the reaction bounds if an error occurs\n",
    "                reaction.bounds = original_bounds\n",
    "\n",
    "    \n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfba_solutions_fluxes = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# WT\n",
    "\n",
    "results = []\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "replicates = {'U1':'Bio141', 'U2':'Bio142', 'U3':'Bio143'}\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "\n",
    "                     with model as modified_model:\n",
    "\n",
    "                        # Open the bounds for the biomass reaction \n",
    "                        modified_model.reactions.biomass_cho_s.bounds = (0,1000)\n",
    "                         \n",
    "                        # Create a copy of the modified_model before making changes\n",
    "                        pre_modification_model = modified_model.copy()\n",
    "                         \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "                        \n",
    "                        print(f'Calculating Growth Rate for WT Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_wt['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                reaction.lower_bound = 8\n",
    "                            for r,v in uptsec_wt.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "            \n",
    "                            # If pFBA succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = modified_model\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')\n",
    "                            print('Reverting to the original model and retrying pFBA...')\n",
    "\n",
    "                            for rxn in pre_modification_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                                if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                    rxn.bounds = (-1000, 1000)\n",
    "                                elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                    rxn.bounds = (-0.001, 1000)\n",
    "                            \n",
    "                            pfba_solution = cobra.flux_analysis.pfba(pre_modification_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate with original model is: {pfba_obj}')\n",
    "\n",
    "                            # If pFBA with original model succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = pre_modification_model\n",
    "\n",
    "                        results.append([rep_suffix, interval_key, exp_gr, pfba_obj])\n",
    "                        # Save the solution with metadata\n",
    "                        pfba_solutions_fluxes.append({\n",
    "                            \"model\": pre_modification_model.id,\n",
    "                            \"condition\": interval_key,\n",
    "                            \"fluxes\": pfba_solution.fluxes\n",
    "                        })\n",
    "\n",
    "                        \n",
    "                         # Perform flux sampling using ACHRSampler\n",
    "                        #print(\"Performing flux sampling...\")\n",
    "                        #achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                        #samples = 5000 \n",
    "                        #sampled_fluxes = achr_sampler.sample(samples)\n",
    "                        \n",
    "                        # Validate and save the sampled fluxes\n",
    "                        #if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                        #    print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                        #else:\n",
    "                        #    f_name = f\"../Simulations/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                        #    with open(f_name, 'wb') as file:\n",
    "                        #        pickle.dump(sampled_fluxes, file)\n",
    "                        \n",
    "# Creating a DataFrame\n",
    "df_wt = pd.DataFrame(results, columns=['Hue', 'Category', 'X Axis', 'Y Axis'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df_wt\n",
    "# Define markers for categories\n",
    "markers = {\n",
    "    'P0 to P2': 'p',  # Circle\n",
    "    'P2 to P4': 'o',  # Circle\n",
    "    'P4 to P6': 's',   # Square\n",
    "}\n",
    "\n",
    "# Set up the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# Plot each group with different colors for hue and different shapes for categories\n",
    "hues = df['Hue'].unique()\n",
    "colors = plt.colormaps.get_cmap('tab10')\n",
    "\n",
    "for hue_idx, hue in enumerate(hues):\n",
    "    hue_subset = df[df['Hue'] == hue]\n",
    "    for category, marker in markers.items():\n",
    "        subset = hue_subset[hue_subset['Category'] == category]\n",
    "        plt.scatter(subset['X Axis'], subset['Y Axis'], label=f'{hue} - {category}', marker=marker, s=100, color=colors(hue_idx))\n",
    "\n",
    "# Add the identity line y = x\n",
    "plt.plot([0.00, 0.055], [0.00, 0.055], color='red', linestyle='--')\n",
    "\n",
    "# Set evenly distributed ticks from 0.01 to 0.08\n",
    "ticks = [i / 100.0 for i in range(1, 11)]\n",
    "plt.xticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "plt.yticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "\n",
    "# Set limits for both axes\n",
    "plt.xlim(0.0, 0.055)\n",
    "plt.ylim(0.0, 0.055)\n",
    "\n",
    "# Setting the same scale for both axes\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Adding labels and title\n",
    "plt.xlabel('Experimental Growth Rate')\n",
    "plt.ylabel('Simulated Growth Rate')\n",
    "plt.title('WT')\n",
    "plt.legend(title='Hue - Category', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Simulations/growth_rate_pred/bar_plot_growth_rate_wt.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ZeLa\n",
    "\n",
    "results = []\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "replicates = {'U4':'Bio144', 'U5':'Bio145', 'U6':'Bio146', 'U7':'Bio147', 'U8':'Bio148'}\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "                    \n",
    "                     with model as modified_model:\n",
    "\n",
    "                        # Open the bounds for the biomass reaction \n",
    "                        modified_model.reactions.biomass_cho_s.bounds = (0,1000)\n",
    "                         \n",
    "                        # Create a copy of the modified_model before making changes\n",
    "                        pre_modification_model = modified_model.copy()\n",
    " \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "                        print(f'Calculating Growth Rate for ZeLa Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_zela['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                reaction.lower_bound = 8\n",
    "                            for r,v in uptsec_zela.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')\n",
    "                            print('Reverting to the original model and retrying pFBA...')\n",
    "\n",
    "                            for rxn in pre_modification_model.reactions:\n",
    "                                if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                    rxn.bounds = (-1000, 1000)\n",
    "                                elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                    rxn.bounds = (-0.001, 1000)\n",
    "                                elif rxn.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                    rxn.lower_bound = 8\n",
    "                                \n",
    "                            pfba_solution = cobra.flux_analysis.pfba(pre_modification_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate with original model is: {pfba_obj}')\n",
    "                            \n",
    "                        results.append([rep_suffix, interval_key, exp_gr, pfba_obj])\n",
    "\n",
    "                        # Save the solution with metadata\n",
    "                        pfba_solutions_fluxes.append({\n",
    "                            \"model\": pre_modification_model.id,\n",
    "                            \"condition\": interval_key,\n",
    "                            \"fluxes\": pfba_solution.fluxes\n",
    "                        })\n",
    "                         \n",
    "                        '''\n",
    "                        # Perform flux sampling using ACHRSampler\n",
    "                        print(\"Performing flux sampling...\")\n",
    "                        achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                        samples = 1 \n",
    "                        sampled_fluxes = achr_sampler.sample(samples)\n",
    "                        \n",
    "                        # Validate and save the sampled fluxes\n",
    "                        if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                            print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                        else:\n",
    "                            f_name = f\"../Simulations/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                            with open(f_name, 'wb') as file:\n",
    "                                pickle.dump(sampled_fluxes, file)\n",
    "                        '''\n",
    "    \n",
    "# Creating a DataFrame\n",
    "df_zela = pd.DataFrame(results, columns=['Hue', 'Category', 'X Axis', 'Y Axis'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df_zela\n",
    "# Define markers for categories\n",
    "markers = {\n",
    "    'P0 to P2': 'p',\n",
    "    'P2 to P4': 'o',  # Circle\n",
    "    'P4 to P6': 's',   # Square\n",
    "    'P6 to P8': 'v',\n",
    "}\n",
    "\n",
    "# Set up the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# Plot each group with different colors for hue and different shapes for categories\n",
    "hues = df['Hue'].unique()\n",
    "colors = plt.colormaps.get_cmap('tab10')\n",
    "\n",
    "for hue_idx, hue in enumerate(hues):\n",
    "    hue_subset = df[df['Hue'] == hue]\n",
    "    for category, marker in markers.items():\n",
    "        subset = hue_subset[hue_subset['Category'] == category]\n",
    "        plt.scatter(subset['X Axis'], subset['Y Axis'], label=f'{hue} - {category}', marker=marker, s=100, color=colors(hue_idx))\n",
    "\n",
    "# Add the identity line y = x\n",
    "plt.plot([0.00, 0.055], [0.00, 0.055], color='red', linestyle='--')\n",
    "\n",
    "# Set evenly distributed ticks from 0.01 to 0.08\n",
    "ticks = [i / 100.0 for i in range(1,11)]\n",
    "plt.xticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "plt.yticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "\n",
    "# Set limits for both axes\n",
    "plt.xlim(0.0, 0.055)\n",
    "plt.ylim(0.0, 0.055)\n",
    "\n",
    "# Setting the same scale for both axes\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Adding labels and title\n",
    "plt.xlabel('Experimental Growth Rate')\n",
    "plt.ylabel('Simulated Growth Rate')\n",
    "plt.title('ZeLa')\n",
    "plt.legend(title='Hue - Category', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Simulations/growth_rate_pred/bar_plot_growth_rate_zela.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for reaction, values in uptsec_zela.items():\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    \n",
    "    # Organize data for plotting\n",
    "    batches = sorted(set(k[0] for k in values.keys()))\n",
    "    timepoints = ['P0 to P2', 'P2 to P4', 'P4 to P6', 'P6 to P8', 'P8 to P12', 'P12 to P14']\n",
    "    \n",
    "    for batch in batches:\n",
    "        batch_data = [values.get((batch, tp), 0) for tp in timepoints]\n",
    "        plt.plot(timepoints, batch_data, marker='o', label=batch)\n",
    "    \n",
    "    plt.title(f'Changes in {reaction} over time')\n",
    "    plt.xlabel('Timepoints')\n",
    "    plt.ylabel('Values')\n",
    "    plt.legend()\n",
    "    plt.xticks(rotation=45)\n",
    "    plt.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for reaction, values in uptsec_wt.items():\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    \n",
    "    # Organize data for plotting\n",
    "    batches = sorted(set(k[0] for k in values.keys()))\n",
    "    timepoints = ['P0 to P2', 'P2 to P4', 'P4 to P6', 'P6 to P8', 'P8 to P12', 'P12 to P14']\n",
    "    \n",
    "    for batch in batches:\n",
    "        batch_data = [values.get((batch, tp), 0) for tp in timepoints]\n",
    "        plt.plot(timepoints, batch_data, marker='o', label=batch)\n",
    "    \n",
    "    plt.title(f'Changes in {reaction} over time')\n",
    "    plt.xlabel('Timepoints')\n",
    "    plt.ylabel('Values')\n",
    "    plt.legend()\n",
    "    plt.xticks(rotation=45)\n",
    "    plt.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Flux Enrichment Analysis <a id='fea'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 1\n",
    "# Use the fluxes from the growth rate calculations to filter reactions with fluxes\n",
    "# Define a cut off for the flux ??? # Plot to visualization of the flux distribution\n",
    "# Run Flux Enrichment Analysis on those reactions to see pathways enriched when optimized by biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 2\n",
    "# Transcriptomic data from cell batches / biorreactiors\n",
    "# Overlay this into the recons and extract the reactions associated to the genes\n",
    "# Extract a reaction vector / \n",
    "# Run Flux Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import hypergeom\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def flux_enrichment_analysis(generic_model, rxn_list, attribute='subsystem'):\n",
    "    \n",
    "    # Check if the attribute exists in the first reaction as a proxy for all\n",
    "    if not hasattr(model.reactions[0], attribute):\n",
    "        raise ValueError(f'Attribute {attribute} not found in model reactions')\n",
    "\n",
    "    # Extract attribute information for all reactions\n",
    "    attribute_values = [getattr(rxn, attribute, 'None') for rxn in model.reactions]\n",
    "    unique_attributes = set(attribute_values)\n",
    "    \n",
    "    # Count occurrences in the model and in the reaction set\n",
    "    model_counts = {attr: attribute_values.count(attr) for attr in unique_attributes}\n",
    "    rxn_set_counts = {attr: 0 for attr in unique_attributes}\n",
    "    for rxn in rxn_list:\n",
    "        rxn_attr = getattr(model.reactions.get_by_id(rxn), attribute, 'None')\n",
    "        rxn_set_counts[rxn_attr] += 1\n",
    "\n",
    "    # Calculate p-values using hypergeometric test\n",
    "    M = len(model.reactions)  # Total number of reactions\n",
    "    n = len(rxn_list)  # Size of reaction set\n",
    "    p_values = []\n",
    "    for attr in unique_attributes:\n",
    "        N = model_counts[attr]  # Total reactions in group\n",
    "        x = rxn_set_counts[attr]  # Reactions in group and in set\n",
    "        p_value = hypergeom.sf(x-1, M, N, n)\n",
    "        p_values.append(p_value)\n",
    "\n",
    "    # Adjust p-values for multiple testing\n",
    "    _, adj_p_values, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "    # Compile results\n",
    "    results = pd.DataFrame({\n",
    "        'Group': list(unique_attributes),\n",
    "        'P-value': p_values,\n",
    "        'Adjusted P-value': adj_p_values,\n",
    "        'Enriched set size': [rxn_set_counts[attr] for attr in unique_attributes],\n",
    "        'Total set size': [model_counts[attr] for attr in unique_attributes],\n",
    "    }).sort_values(by='Adjusted P-value')\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test 1: FEA on reactions active during growth rate optimization\n",
    "Use the fluxes from the growth rate calculations to filter reactions with fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate a vector of active reactions for each one of the conditions\n",
    "\n",
    "active_reactions = []\n",
    "\n",
    "for sol in pfba_solutions_fluxes:\n",
    "    rxns_fluxes = []\n",
    "    for n,f in sol['fluxes'].items():\n",
    "        if f != 0:\n",
    "            rxns_fluxes.append(n)\n",
    "    \n",
    "    active_reactions.append({\n",
    "        \"batch\": sol['model'],\n",
    "        \"condition\": sol['condition'],\n",
    "        \"flux_vector\": rxns_fluxes\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load generic model\n",
    "model = load_json_model('iCHO3595_unblocked.json')\n",
    "\n",
    "# Add atp demand reaction\n",
    "DM_atp = Reaction('DM_atp_c')\n",
    "model.add_reactions([DM_atp])\n",
    "model.reactions.DM_atp_c.build_reaction_from_string('atp_c -->')\n",
    "model.reactions.DM_atp_c.bounds = (0.01, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate FEA results for each one of the vectors generated for each condition\n",
    "\n",
    "fea_results = []\n",
    "\n",
    "# Load generic model\n",
    "#model = load_json_model('iCHO3595_unblocked.json')\n",
    "\n",
    "for fluxes in active_reactions:\n",
    "    results = flux_enrichment_analysis(model, fluxes['flux_vector'], 'subsystem')\n",
    "    fea_results.append({\n",
    "        \"Batch\": fluxes['batch'],\n",
    "        \"Condition\": fluxes['condition'],\n",
    "        \"Results\": results\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Transform p-values to -log10\n",
    "\n",
    "all_data = pd.DataFrame()\n",
    "\n",
    "for result in fea_results:\n",
    "    batch = result['Batch']\n",
    "    temp_df = result['Results'][['Group', 'P-value']].copy()\n",
    "    temp_df.columns = ['Group', f'P-value_{batch}']\n",
    "    if all_data.empty:\n",
    "        all_data = temp_df\n",
    "    else:\n",
    "        all_data = pd.merge(all_data, temp_df, on='Group', how='outer')\n",
    "\n",
    "# Remove groups with a 0 value in all conditions (assuming a '0' value indicates non-significance)\n",
    "significant_filter = (all_data.drop(columns='Group') < 0.05).any(axis=1)\n",
    "filtered_data = all_data[significant_filter]\n",
    "filtered_data.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# Transform the p-values\n",
    "for col in filtered_data.columns:\n",
    "    if col.startswith('P-value'):\n",
    "        filtered_data[f'-log10_{col}'] = -np.log10(filtered_data[col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import zscore\n",
    "\n",
    "# Extract relevant columns for heatmap\n",
    "heatmap_data = filtered_data[[col for col in filtered_data.columns if col.startswith('-log10')]]\n",
    "\n",
    "# Clean column names\n",
    "heatmap_data.columns = heatmap_data.columns.str.replace(r\"-log10_P-value_\", \"\", regex=True)\n",
    "\n",
    "# Normalize the data (z-score)\n",
    "normalized_data = heatmap_data.apply(zscore, axis=1)\n",
    "\n",
    "# Prepare heatmap data\n",
    "normalized_data.index = filtered_data['Group']\n",
    "\n",
    "# Reorder columns: Group P2, P4, P6, and P8 together\n",
    "columns_order = []\n",
    "for phase in ['P2', 'P4', 'P6', 'P8']:\n",
    "    columns_order.extend([col for col in normalized_data.columns if phase in col])\n",
    "\n",
    "# Reorder the DataFrame\n",
    "normalized_data = normalized_data[columns_order]\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(20, 20))\n",
    "sns.heatmap(normalized_data, annot=True, cmap=\"viridis\", cbar_kws={'label': 'Z-score of -log10(P-value)'}, vmin=-1.5, vmax=1.5)\n",
    "plt.title('Heatmap of Z-score Normalized -log10(P-values) Across Conditions', fontsize=20)\n",
    "plt.xlabel('Conditions', fontsize=15)\n",
    "plt.ylabel('Groups', fontsize=15)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Simulations/flux_enrichment_analysis/FEA_heatmap_all.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the columns based on the presence of \"P2 to P4\" or \"P4 to P6\"\n",
    "subset_df = heatmap_data.filter(regex='_P4_|_P6_')\n",
    "\n",
    "# Normalize the data (z-score)\n",
    "normalized_data = subset_df.apply(zscore, axis=1)\n",
    "\n",
    "# Prepare heatmap data\n",
    "normalized_data.index = filtered_data['Group']\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(20, 20))\n",
    "sns.heatmap(normalized_data, annot=True, cmap=\"viridis\", cbar_kws={'label': 'Z-score of -log10(P-value)'}, vmin=-1.5, vmax=1.5)\n",
    "plt.title('Heatmap of Z-score Normalized -log10(P-values) Across Conditions', fontsize=20)\n",
    "plt.xlabel('Conditions', fontsize=15)\n",
    "plt.ylabel('Groups', fontsize=15)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Simulations/flux_enrichment_analysis/FEA_heatmap_p2_p6.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
