{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Tests\n",
    "\n",
    "This notebook is intended for running different kinds of analyses that would validate our reconstruction.\n",
    "\n",
    "[1. Subsystem Overview and Analysis](#subsystems) <br>\n",
    "[2. Context-specific Model Generation](#context_specific) <br>\n",
    "[3. Biomass prediction using exp. data](#biomass) <br>\n",
    "[4. Flux Enrichment Analysis](#fea) <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Subsystem Overview and Analysis <a id='subsystems'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from google_sheet import GoogleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate Subsystems datasets from Google Sheet ----- #####\n",
    "\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_subsystems = 'Summary Systems'\n",
    "sheet_reactions = 'Rxns'\n",
    "subsystems = sheet.read_google_sheet(sheet_subsystems)\n",
    "reactions_rec = sheet.read_google_sheet(sheet_reactions)\n",
    "\n",
    "# Remove the total count\n",
    "subsystems = subsystems.iloc[:-1, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Pie Chart of the Amount of Reactions per System --- ####\n",
    "\n",
    "subsystems[\"Number of Reactions\"] = pd.to_numeric(subsystems[\"Number of Reactions\"])\n",
    "\n",
    "# Aggregate data by \"System\" and sum \"Number of Reactions\"\n",
    "system_reactions = subsystems.groupby(\"System\")[\"Number of Reactions\"].sum()\n",
    "\n",
    "# Set the figure and axes for more control\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "\n",
    "# Generate the pie chart again with the custom labels\n",
    "patches, texts, autotexts = ax.pie(system_reactions, labels=system_reactions.index, autopct=lambda p: '{:.1f}%'.format(p) if p > 0 else '', startangle=140, colors=plt.cm.tab20.colors)\n",
    "\n",
    "# Improve aesthetics\n",
    "for text in texts + autotexts:\n",
    "    text.set_color('black')\n",
    "ax.set_title(\"Pie Chart of Number of Reactions by System\", pad=30)  # Move the title upwards by increasing pad\n",
    "\n",
    "# Set equal aspect ratio\n",
    "ax.axis('equal')\n",
    "\n",
    "# Save the plot\n",
    "plt.savefig('../Plots/pie_chart_reactions_per_system.png')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the Sunburst Plot\n",
    "\n",
    "fig = px.sunburst(subsystems, path=['System', 'Subsystems', 'Kegg Pathway'],\n",
    "                  color='System')\n",
    "\n",
    "# Convert to a Graph Objects figure\n",
    "fig_go = go.Figure(fig)\n",
    "\n",
    "# Update layout\n",
    "fig_go.update_layout(width=1200, height=1000)\n",
    "\n",
    "# Define font sizes\n",
    "font_size_last_layer = 10\n",
    "font_size_previous_layer = 15\n",
    "default_font_size = 12\n",
    "\n",
    "# Create a list to store font sizes\n",
    "font_sizes = []\n",
    "\n",
    "# Update font size for each level\n",
    "for trace in fig_go.data:\n",
    "    if isinstance(trace, go.Sunburst):\n",
    "        for id in trace.ids:\n",
    "            level = id.count(\"/\")  # Determine level by the number of slashes in the id\n",
    "            if level == 2:  # Last layer (Kegg Pathway)\n",
    "                font_sizes.append(font_size_last_layer)\n",
    "            elif level == 1:  # Previous layer (Subsystems)\n",
    "                font_sizes.append(font_size_previous_layer)\n",
    "            else:\n",
    "                font_sizes.append(default_font_size)  # Default size for other layers\n",
    "\n",
    "# Apply the font sizes to the figure\n",
    "fig_go.update_traces(insidetextfont=dict(size=font_sizes))\n",
    "\n",
    "# Save the figure\n",
    "fig_go.write_html(\"../Plots/sunburst_subsystems.html\")  # Save as interactive HTML file\n",
    "fig_go.write_image(\"../Plots/sunburst_subsystems.png\", width=1200, height=1000, scale=2)  # Increase resolution by setting scale parameter\n",
    "\n",
    "# Show the plot\n",
    "fig_go.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Auxotrophies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.io import load_json_model\n",
    "\n",
    "iCHO_path = \"iCHO3595.json\"\n",
    "iCHO = load_json_model(iCHO_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids = {\n",
    "    \"arginine\": [\"EX_arg_L_e\", \"EX_arg_D_e\"],\n",
    "    \"asparagine\": [\"EX_asn_L_e\", \"EX_asn_D_e\"],\n",
    "    \"cysteine\": [\"EX_cys_L_e\", \"EX_cys_D_e\", \"EX_Lcystin_e\"],\n",
    "    \"histidine\": [\"EX_his_L_e\", \"EX_his_D_e\"],\n",
    "    \"isoleucine\": [\"EX_ile_L_e\", \"EX_ile_D_e\"],\n",
    "    \"leucine\": [\"EX_leu_L_e\", \"EX_leu_D_e\"],\n",
    "    \"lysine\": [\"EX_lys_L_e\", \"EX_lys_D_e\"],\n",
    "    \"methionine\": [\"EX_met_L_e\", \"EX_met_D_e\"],\n",
    "    \"phenylalanine\": [\"EX_phe_L_e\", \"EX_phe_D_e\"],\n",
    "    \"proline\": [\"EX_pro_L_e\", \"EX_pro_D_e\"],\n",
    "    \"threonine\": [\"EX_thr_L_e\", \"EX_thr_D_e\"],\n",
    "    \"tryptophan\": [\"EX_trp_L_e\", \"EX_trp_D_e\"],\n",
    "    \"valine\": [\"EX_val_L_e\", \"EX_val_D_e\"]\n",
    "}\n",
    "for amino_acid in amino_acids:\n",
    "    # ----- Setup initial bounds -----\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        exchange_reaction.bounds = -10, 10\n",
    "\n",
    "    # Arginine\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00557').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00558').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GLYAMDTRc').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ARGSL').bounds = 0, 0\n",
    "\n",
    "    # Asparigine\n",
    "    iCHO.reactions.get_by_id('ASNS1').bounds = 0, 0\n",
    "\n",
    "    # Cysteine\n",
    "    iCHO.reactions.get_by_id('r0129').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_cgly_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECG').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECGe').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('CYSTGL').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('EX_HC00250_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_sfcys_e').bounds = 0, 0\n",
    "\n",
    "    # Histidine\n",
    "    iCHO.reactions.get_by_id('VALTRPVALr').bounds = 0,0 \n",
    "\n",
    "    # Isoleucine - DONE\n",
    "    iCHO.reactions.get_by_id('EX_CE2916_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2915_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('ILETA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ILETAm').bounds = 0, 10\n",
    "\n",
    "    # Leucine\n",
    "    iCHO.reactions.get_by_id('LEULEULAPc').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_leugly_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_glyleu_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('LEUTA').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('LEUTAm').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('EX_CE5797_e').bounds = 0, 0\n",
    "\n",
    "    # Lysine\n",
    "    iCHO.reactions.get_by_id('EX_biocyt_e').bounds = 0,0 \n",
    "\n",
    "    # Methionine\n",
    "    iCHO.reactions.get_by_id('METS').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('BHMT').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('GAUGE-R00648').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('UNK2').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('UNK3').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('TYRA').bounds = -10, 0 \n",
    "    #iCHO.reactions.get_by_id('GAUGE-R06895').bounds = 0, 0 # Curated by MR with 1 score / Erased from the reconstruction\n",
    "\n",
    "    # Phenylalanine\n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_pheleu_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_glyphe_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2917_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5789_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_phpyr_e').bounds = 0, 0 \n",
    "\n",
    "    # Proline\n",
    "    iCHO.reactions.get_by_id('EX_glypro_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_progly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('P5CR').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRxm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRx').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('r1453').bounds = 0, 10\n",
    "\n",
    "    # Threonine\n",
    "    #iCHO.reactions.get_by_id('THRS').bounds = 0,0  # Erased from the reconstruction\n",
    "\n",
    "\n",
    "    # Tryptophan\n",
    "    iCHO.reactions.get_by_id('NBAHH_ir').bounds = 0,0  # Histidine hydrolase\n",
    "\n",
    "    # Valine\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_vallystyr_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTAm').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_valval_e').bounds = 0, 10\n",
    "    # -----------------------------------------\n",
    "\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        if exchange_reaction.id in amino_acids[amino_acid]:\n",
    "            exchange_reaction.bounds = 0, 10\n",
    "            # print(amino_acid, exchange_reaction.id, amino_acids[amino_acid])\n",
    "    sol = iCHO.optimize()\n",
    "    print(amino_acid, sol.objective_value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCHO.metabolites.leu_L_c.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Deletion test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from scipy.io import loadmat\n",
    "from time import process_time\n",
    "import cobra\n",
    "from cobra.io import load_json_model, read_sbml_model\n",
    "\n",
    "\n",
    "model = load_json_model('iCHO3595_unblocked.json')\n",
    "iCHO1766 = read_sbml_model('../Data/Reconciliation/models/iCHOv1_final.xml')\n",
    "iCHO2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "iCHO2441 = read_sbml_model('../Data/Reconciliation/models/iCHO2441.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset for the experimentally validated essential genes\n",
    "\n",
    "exp_essential_genes = pd.read_csv('../Data/Gene_Essentiality/cho_essential_genes.csv', sep='\\t')\n",
    "exp_essential_genes['gene'] = exp_essential_genes['gene'].str.replace('__1$', '', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract all gene names\n",
    "all_gene_names = set()\n",
    "for names in exp_essential_genes['gene']:\n",
    "    for name in names.split('|'): #Extract the names of the genes with two aliases\n",
    "        all_gene_names.add(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n# Generation gene ID of exp_essential_genes with biopython\\nfrom Bio import Entrez\\nimport pandas as pd\\n\\n# Set your email (this is required by NCBI for accessing their services)\\nEntrez.email = \"dh.choi@orcid\"\\n\\n# Your dataset with gene symbols (replace with actual dataframe)\\ngene_symbols = exp_essential_genes.gene  # Example gene symbols\\n\\n# Function to fetch gene ID using Entrez API\\ndef get_gene_id(gene_symbol):\\n    search_handle = Entrez.esearch(db=\"gene\", term=f\"{gene_symbol}[Gene] AND Cricetulus griseus[Organism]\")\\n    record = Entrez.read(search_handle)\\n    search_handle.close()\\n    if record[\"IdList\"]:\\n        return record[\"IdList\"][0]  # Return the first gene ID found\\n    else:\\n        return None\\n\\n# Create a DataFrame to store gene symbols and their corresponding gene IDs\\ndf = pd.DataFrame({\\'gene\\': gene_symbols})\\ndf[\\'entrez_gene_id\\'] = df[\\'gene\\'].apply(get_gene_id)\\n\\n# Display the result\\nprint(df)\\n\\n# Save the DataFrame to a pickle file\\nwith open(\"gene_id_mapping.pkl\", \"wb\") as f:\\n    pickle.dump(df, f)\\n\\n# Save the result to a CSV file as well, if needed\\ndf.to_csv(\"gene_id_mapping.csv\", index=False)\\n\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "# Generation gene ID of exp_essential_genes with biopython\n",
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "\n",
    "# Set your email (this is required by NCBI for accessing their services)\n",
    "Entrez.email = \"dh.choi@orcid\"\n",
    "\n",
    "# Your dataset with gene symbols (replace with actual dataframe)\n",
    "gene_symbols = exp_essential_genes.gene  # Example gene symbols\n",
    "\n",
    "# Function to fetch gene ID using Entrez API\n",
    "def get_gene_id(gene_symbol):\n",
    "    search_handle = Entrez.esearch(db=\"gene\", term=f\"{gene_symbol}[Gene] AND Cricetulus griseus[Organism]\")\n",
    "    record = Entrez.read(search_handle)\n",
    "    search_handle.close()\n",
    "    if record[\"IdList\"]:\n",
    "        return record[\"IdList\"][0]  # Return the first gene ID found\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Create a DataFrame to store gene symbols and their corresponding gene IDs\n",
    "df = pd.DataFrame({'gene': gene_symbols})\n",
    "df['entrez_gene_id'] = df['gene'].apply(get_gene_id)\n",
    "\n",
    "# Display the result\n",
    "print(df)\n",
    "\n",
    "# Save the DataFrame to a pickle file\n",
    "with open(\"gene_id_mapping.pkl\", \"wb\") as f:\n",
    "    pickle.dump(df, f)\n",
    "\n",
    "# Save the result to a CSV file as well, if needed\n",
    "df.to_csv(\"gene_id_mapping.csv\", index=False)\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          gene entrez_gene_id\n",
      "0          Ubc      100689267\n",
      "1       Ncaph2      100756862\n",
      "2        Huwe1      100757403\n",
      "3      Isg20l2      100752909\n",
      "4        Coq8b      100770066\n",
      "...        ...            ...\n",
      "1975     Sprtn      100764217\n",
      "1976     Perm1      100773289\n",
      "1977   Tmem147      100751744\n",
      "1978    Rnf216      100758416\n",
      "1979  Rnaseh2b      100751948\n",
      "\n",
      "[1980 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load the DataFrame from the pickle file\n",
    "with open(\"gene_id_mapping.pkl\", \"rb\") as f:\n",
    "    df_loaded = pickle.load(f)\n",
    "\n",
    "# Display the loaded DataFrame\n",
    "print(df_loaded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)\n",
    "    \n",
    "temp_dict_zela = uptsec_intrvl_zela\n",
    "temp_dict_wt = uptsec_intrvl_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GM = model\n",
    "# Constraint the model with exp data\n",
    "temp_dict = temp_dict_zela\n",
    "time = 'P2 to P4'\n",
    "\n",
    "GM.solver = 'gurobi'\n",
    "objective = 'biomass_cho_s' # 'biomass_cho'\n",
    "GM.objective = objective\n",
    "\n",
    "GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "for rxn in GM.boundary:\n",
    "    \n",
    "    # Models that are forced to secrete ethanol are not feasible\n",
    "    if rxn.id == 'EX_etoh_e':\n",
    "        rxn.bounds = (-0.1,0.1)\n",
    "        continue\n",
    "\n",
    "    # Replace the lower and upper bound with experimental data\n",
    "    if rxn.id in temp_dict.keys():\n",
    "        rxn.bounds = temp_dict[rxn.id][time]\n",
    "        continue\n",
    "    \n",
    "    # Keep boundaries open for essential metabolites\n",
    "    if rxn.id == 'EX_h2o_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_h_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_o2_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_hco3_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_so4_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_pi_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "\n",
    "    # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "    if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Tyr_ggn_c':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Ser_Thr_g':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_pre_prot_r':\n",
    "        rxn.bounds = (-0.0006,1000)\n",
    "        continue\n",
    "\n",
    "    # Close uptake rates for the rest of the boundaries\n",
    "    if rxn.id.startswith(\"EX_\"):\n",
    "        rxn.bounds = (0,1000) \n",
    "    if rxn.id.startswith(\"SK_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    if rxn.id.startswith(\"DM_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    \n",
    "pfba_solution = cobra.flux_analysis.pfba(GM)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(pfba_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.flux_analysis.variability import find_essential_genes\n",
    "\n",
    "# Find the essential genes for the given model\n",
    "essential_genes = find_essential_genes(model)\n",
    "\n",
    "# Print the number of essential genes for biomass\n",
    "print(f\"Number of essential genes for biomass: {len(essential_genes)}\")\n",
    "print(\"Essential genes for biomass production:\")\n",
    "\n",
    "# Iterate over the essential genes and print them (essential_genes is a set)\n",
    "for gene in essential_genes:\n",
    "    print(f\"{gene.id}: {gene.name}\")\n",
    "\n",
    "# Create a reverse lookup dictionary where the key is the Entrez Gene ID and the value is the gene symbol\n",
    "entrez_dict = df_loaded.set_index('entrez_gene_id')['gene'].to_dict()\n",
    "\n",
    "# Now, match essential genes with Entrez Gene IDs and print the results\n",
    "print(\"Essential genes for biomass production with matching Entrez Gene IDs:\")\n",
    "\n",
    "# Iterate over the essential genes and match them with Entrez Gene IDs\n",
    "for gene in essential_genes:\n",
    "    # Check if the gene.name matches any Entrez Gene ID in the dictionary\n",
    "    entrez_id = entrez_dict.get(gene.name, \"No Match\")  # You can also try gene.id if it's closer to Entrez IDs\n",
    "    \n",
    "    # Print the gene along with its corresponding Entrez Gene ID (if matched)\n",
    "    print(f\"{gene.id}: {gene.name}, Entrez Gene ID: {entrez_id}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list to store the results\n",
    "matched_genes = []\n",
    "\n",
    "# Iterate over the essential genes\n",
    "for gene in essential_genes:\n",
    "    gene_exp_ess = model.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "    # Check if the gene.id matches any Entrez Gene ID in the reverse dictionary\n",
    "    if gene_exp_ess.id in entrez_dict:\n",
    "        # If it matches, retrieve the corresponding gene symbol\n",
    "        entrez_gene_symbol = entrez_dict[gene_exp_ess.id]\n",
    "        print(f\"{gene_exp_ess.id}: {gene_exp_ess.name}, Entrez Gene Symbol: {entrez_gene_symbol}\")\n",
    "\n",
    "        # Append the matched result to the list\n",
    "        matched_genes.append((gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol))\n",
    "    else:\n",
    "        print(f\"{gene_exp_ess.id}: {gene_exp_ess.name}, No Entrez Gene Match\")\n",
    "\n",
    "# Convert the matched list into a DataFrame if needed\n",
    "matched_genes_essential_df = pd.DataFrame(matched_genes, columns=['gene_exp_ess.id', 'gene_exp_ess.name', 'entrez_gene_symbol'])\n",
    "\n",
    "# Save the result to a CSV file\n",
    "matched_genes_essential_df.to_csv(\"matched_gene_entrez_ids_feg.csv\", index=False)\n",
    "\n",
    "# Optional: Print the DataFrame\n",
    "print(matched_genes_essential_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find rows where entrez_gene_id is None (indicating no match was found)\n",
    "no_match_df = df_loaded[df_loaded['entrez_gene_id'].isnull()]\n",
    "len(no_match_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(matched_genes_essential_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Iterating Condition-Specific Constraints and Identifyng Matched Essential Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from scipy.io import loadmat\n",
    "from time import process_time\n",
    "import cobra\n",
    "from cobra.io import load_json_model, read_sbml_model\n",
    "from cobra.flux_analysis.variability import find_essential_genes\n",
    "\n",
    "# model = load_json_model('iCHO3595_unblocked.json')\n",
    "iCHO1766 = read_sbml_model('../Data/Reconciliation/models/iCHOv1_final.xml')\n",
    "iCHO2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "iCHO2101 = load_json_model('../Data/Reconciliation/models/iCHO2101.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biomass_cho: 0.44909 ala_L[c] + 0.24693 arg_L[c] + 0.20889 asn_L[c] + 0.25757 asp_L[c] + 29.9348 atp[c] + 0.086971 chsterol[c] + 0.0066321 clpn_cho[c] + 0.097355 ctp[c] + 0.083084 cys_L[c] + 0.035778 datp[c] + 0.025173 dctp[c] + 0.025173 dgtp[c] + 0.035778 dttp[c] + 0.35321 gln_L[c] + 0.30325 glu_L[c] + 0.47669 gly[c] + 0.025686 glygn2[c] + 0.10294 gtp[c] + 24.7485 h2o[c] + 0.096216 his_L[c] + 0.22017 ile_L[c] + 0.42729 leu_L[c] + 0.38764 lys_L[c] + 0.10558 met_L[c] + 0.0017194 pa_cho[c] + 0.018177 pail_cho[c] + 0.13019 pchol_cho[c] + 0.049864 pe_cho[c] + 0.0014738 pglyc_cho[c] + 0.18672 phe_L[c] + 0.22206 pro_L[c] + 0.016457 ps_cho[c] + 0.323 ser_L[c] + 0.02137 sphmyln_cho[c] + 0.010837 tag_cho[c] + 0.25537 thr_L[c] + 0.025173 trp_L[c] + 0.13125 tyr_L[c] + 0.085376 utp[c] + 0.32263 val_L[c] + 0.055982 xolest_cho[c] --> 29.8303 adp[c] + 29.8303 h[c] + 29.8303 pi[c] + 0.5121 ppi[c]\n"
     ]
    }
   ],
   "source": [
    "objective = 'biomass_cho'\n",
    "pfba_solution = cobra.flux_analysis.pfba(iCHO2101)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "tmp = iCHO2101.reactions.get_by_id(objective)\n",
    "print(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset for the experimentally validated essential genes\n",
    "\n",
    "exp_essential_genes = pd.read_csv('../Data/Gene_Essentiality/cho_essential_genes.csv', sep='\\t')\n",
    "exp_essential_genes['gene'] = exp_essential_genes['gene'].str.replace('__1$', '', regex=True)\n",
    "# Extract all gene names\n",
    "all_gene_names = set()\n",
    "for names in exp_essential_genes['gene']:\n",
    "    for name in names.split('|'): #Extract the names of the genes with two aliases\n",
    "        all_gene_names.add(name)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          gene entrez_gene_id\n",
      "0          Ubc      100689267\n",
      "1       Ncaph2      100756862\n",
      "2        Huwe1      100757403\n",
      "3      Isg20l2      100752909\n",
      "4        Coq8b      100770066\n",
      "...        ...            ...\n",
      "1975     Sprtn      100764217\n",
      "1976     Perm1      100773289\n",
      "1977   Tmem147      100751744\n",
      "1978    Rnf216      100758416\n",
      "1979  Rnaseh2b      100751948\n",
      "\n",
      "[1980 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load the DataFrame from the pickle file\n",
    "with open(\"gene_id_mapping.pkl\", \"rb\") as f:\n",
    "    df_loaded = pickle.load(f)\n",
    "\n",
    "# Display the loaded DataFrame\n",
    "print(df_loaded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)\n",
    "    \n",
    "temp_dict_zela = uptsec_intrvl_zela\n",
    "temp_dict_wt = uptsec_intrvl_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CHOOSE ONE!!!! (Recommended version of code)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def constraint_model_feg(model, time, exp_dict, objective=None, solver=None, make_GM=None):\n",
    "    \"\"\"\n",
    "    Constrain the model with experimental data and calculate the objective function value using pFBA.\n",
    "    \n",
    "    Parameters:\n",
    "    - model: COBRA model to be constrained.\n",
    "    - time: Time interval for applying the experimental constraints. Default is 'P2 to P4'.\n",
    "    - objective: The objective reaction. Default is 'biomass_cho_s'.\n",
    "    - solver: Solver to be used. Default is 'gurobi'.\n",
    "    \n",
    "    Returns:\n",
    "    - pfba_obj: The optimized objective function value (growth rate or other objective).\n",
    "    - pfba_solution: The full pFBA solution (flux distribution).\n",
    "    \"\"\"\n",
    "    \n",
    "    GM = model.copy()  # Copy the model to avoid modifying the original\n",
    "    temp_dict = exp_dict\n",
    "    \n",
    "    # Default values if no argument is provided\n",
    "    if objective is None:\n",
    "        objective = 'biomass_cho_s'\n",
    "    if solver is None:\n",
    "        solver = 'gurobi'\n",
    "\n",
    "    # Set the solver and the objective reaction\n",
    "    GM.solver = solver\n",
    "    GM.objective = objective\n",
    "\n",
    "    # Constraint the model with experimental data\n",
    "    if objective is None:\n",
    "        GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "    \n",
    "    for rxn in GM.boundary:\n",
    "        # Models that are forced to secrete ethanol are not feasible\n",
    "        if rxn.id == 'EX_etoh_e':\n",
    "            rxn.bounds = (-0.1,0.1)\n",
    "            continue\n",
    "\n",
    "        # Replace the lower and upper bound with experimental data\n",
    "        if rxn.id in temp_dict.keys():\n",
    "            rxn.bounds = temp_dict[rxn.id][time]\n",
    "            continue\n",
    "\n",
    "        # Keep boundaries open for essential metabolites\n",
    "        if rxn.id == 'EX_h2o_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_h_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_o2_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_hco3_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_so4_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_pi_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "\n",
    "        # Boundaries from Sink reactions on iCHO_v1\n",
    "        if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Tyr_ggn_c':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Ser_Thr_g':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_pre_prot_r':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "\n",
    "        # Close uptake rates for the rest of the boundaries\n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "    \n",
    "    if make_GM == True:\n",
    "        for rxn in GM.boundary:\n",
    "            if rxn.id in temp_dict.keys():\n",
    "                rxn.bounds = (-1000,1000)\n",
    "                continue\n",
    "        pfba_solution_GM = cobra.flux_analysis.pfba(GM)\n",
    "        pfba_obj_GM = pfba_solution_GM.fluxes[objective]        \n",
    "            \n",
    "            \n",
    "    model_exp_constrained = GM    \n",
    "    # Perform pFBA (parsimonious FBA) to minimize the total flux\n",
    "    pfba_solution = cobra.flux_analysis.pfba(GM)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "\n",
    "    # Print and return the objective value and the full pFBA solution\n",
    "    print(f\"Optimized objective (growth rate): {pfba_obj}\")\n",
    "    \n",
    "    # Return the objective value and the full pFBA solution\n",
    "    return model_exp_constrained, pfba_obj, pfba_solution, pfba_obj_GM, GM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CHOOSE ONE!!!!\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def constraint_model_feg(model, time, exp_dict, objective=None, solver=None):\n",
    "    \"\"\"\n",
    "    Constrain the model with experimental data and calculate the objective function value using pFBA.\n",
    "    \n",
    "    Parameters:\n",
    "    - model: COBRA model to be constrained.\n",
    "    - time: Time interval for applying the experimental constraints. Default is 'P2 to P4'.\n",
    "    - objective: The objective reaction. Default is 'biomass_cho_s'.\n",
    "    - solver: Solver to be used. Default is 'gurobi'.\n",
    "    \n",
    "    Returns:\n",
    "    - pfba_obj: The optimized objective function value (growth rate or other objective).\n",
    "    - pfba_solution: The full pFBA solution (flux distribution).\n",
    "    \"\"\"\n",
    "    \n",
    "    GM = model.copy()  # Copy the model to avoid modifying the original\n",
    "    temp_dict = exp_dict\n",
    "    \n",
    "    # Default values if no argument is provided\n",
    "    if objective is None:\n",
    "        objective = 'biomass_cho_s'\n",
    "    if solver is None:\n",
    "        solver = 'gurobi'\n",
    "\n",
    "    # Set the solver and the objective reaction\n",
    "    GM.solver = solver\n",
    "    GM.objective = objective\n",
    "\n",
    "    # Constraint the model with experimental data\n",
    "    if objective is None:\n",
    "        GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "    \n",
    "    for rxn in GM.boundary:\n",
    "        # Models that are forced to secrete ethanol are not feasible\n",
    "        if rxn.id == 'EX_etoh_e':\n",
    "            rxn.bounds = (-0.1,0.1)\n",
    "            continue\n",
    "\n",
    "        # Replace the lower and upper bound with experimental data\n",
    "        if rxn.id in temp_dict.keys():\n",
    "            rxn.bounds = temp_dict[rxn.id][time]\n",
    "            continue\n",
    "\n",
    "        # Keep boundaries open for essential metabolites\n",
    "        if rxn.id == 'EX_h2o_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_h_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_o2_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_hco3_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_so4_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_pi_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "\n",
    "        # Boundaries from Sink reactions on iCHO_v1\n",
    "        if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Tyr_ggn_c':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Ser_Thr_g':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_pre_prot_r':\n",
    "            rxn.bounds = (-0.1,1000)\n",
    "            continue\n",
    "\n",
    "        # Close uptake rates for the rest of the boundaries\n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "    \n",
    "#     if make_GM == True:\n",
    "#         for rxn in GM.boundary:\n",
    "#             if rxn.id in temp_dict.keys():\n",
    "#                 rxn.bounds = (-1000,1000)\n",
    "#                 continue\n",
    "# #         pfba_solution_GM = cobra.flux_analysis.pfba(GM)\n",
    "# #         pfba_obj_GM = pfba_solution_GM.fluxes[objective]        \n",
    "            \n",
    "            \n",
    "    model_exp_constrained = GM   \n",
    "    # Perform pFBA (parsimonious FBA) to minimize the total flux\n",
    "    pfba_solution = cobra.flux_analysis.pfba(GM)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "\n",
    "    # Print and return the objective value and the full pFBA solution\n",
    "    print(f\"Optimized objective (growth rate): {pfba_obj}\")\n",
    "    \n",
    "    # Return the objective value and the full pFBA solution\n",
    "    return model_exp_constrained, pfba_obj, pfba_solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CHOOSE ONE!!!!(Recommended)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def find_and_match_essential_genes(model, df_exp_essential, save_csv=True, print_df=False, model_GM = None):\n",
    "    \"\"\"\n",
    "    This function finds and matches essential genes in the model with experimental Entrez gene data.\n",
    "    It returns the number of total matched genes (len_mat) and the number of essential matched genes (len_mat_ess).\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if the necessary columns 'entrez_gene_id' and 'gene' exist in the dataframe\n",
    "    required_columns = ['entrez_gene_id', 'gene']\n",
    "    for col in required_columns:\n",
    "        if col not in df_exp_essential.columns:\n",
    "            raise KeyError(f\"Column '{col}' is missing from the dataframe.\")\n",
    "\n",
    "    # Create a reverse lookup dictionary where the key is the Entrez Gene ID and the value is the gene symbol\n",
    "    entrez_dict = df_exp_essential.set_index('entrez_gene_id')['gene'].to_dict()\n",
    "\n",
    "    # Matching metabolic genes\n",
    "    tmp_gene = []    \n",
    "    for gene in model.genes:\n",
    "        # Check if the gene.id matches any Entrez Gene ID in the reverse dictionary\n",
    "        if gene.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene.id]\n",
    "            tmp_gene.append((gene.id, gene.name, entrez_gene_symbol))\n",
    "    \n",
    "    # Number of total matched genes\n",
    "    len_mat = len(tmp_gene)\n",
    "\n",
    "    # Find the essential genes for the given model\n",
    "    essential_genes = find_essential_genes(model)\n",
    "    \n",
    "    # Create a list to store the matched essential genes\n",
    "    matched_genes = []\n",
    "\n",
    "    # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "    for gene in essential_genes:\n",
    "        gene_exp_ess = model.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "        if gene_exp_ess.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene_exp_ess.id]\n",
    "            matched_genes.append((gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol))\n",
    "\n",
    "    # Convert the matched essential genes into a DataFrame\n",
    "    matched_genes_essential_df = pd.DataFrame(matched_genes, columns=['gene_exp_ess.id', 'gene_exp_ess.name', 'entrez_gene_symbol'])\n",
    "\n",
    "    # Save the result to a CSV file if save_csv is True\n",
    "    if save_csv:\n",
    "        matched_genes_essential_df.to_csv(\"matched_gene_entrez_ids_feg.csv\", index=False)\n",
    "\n",
    "    # Print the DataFrame if print_df is True\n",
    "    if print_df:\n",
    "        print(matched_genes_essential_df)\n",
    "\n",
    "    # Number of essential matched genes\n",
    "    len_mat_ess = len(matched_genes_essential_df)\n",
    "    \n",
    "    if model_GM is None:\n",
    "        # Print the summary information\n",
    "        print(f\"Total matched metabolic genes: {len_mat}\")\n",
    "        print(f\"Matched essential genes: {len_mat_ess}/{len_mat}\")\n",
    "    else: \n",
    "        \n",
    "        essential_genes = find_essential_genes(model_GM)\n",
    "\n",
    "        # Create a list to store the matched essential genes\n",
    "        matched_genes_GM = []\n",
    "\n",
    "        # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "        for gene in essential_genes:\n",
    "            gene_exp_ess_GM = model_GM.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "            if gene_exp_ess_GM.id in entrez_dict:\n",
    "                entrez_gene_symbol = entrez_dict[gene_exp_ess_GM.id]\n",
    "                matched_genes_GM.append((gene_exp_ess_GM.id, gene_exp_ess_GM.name, entrez_gene_symbol))\n",
    "\n",
    "        # Convert the matched essential genes into a DataFrame\n",
    "        matched_genes_essential_GM_df = pd.DataFrame(matched_genes_GM, columns=['gene_exp_ess_GM.id', 'gene_exp_ess_GM.name', 'entrez_gene_symbol'])\n",
    "        len_mat_GM_ess = len(matched_genes_essential_GM_df)\n",
    "        # Print the summary information\n",
    "        print(f\"Total matched metabolic genes_GM: {len_mat}\")\n",
    "        print(f\"Matched essential genes_GM: {len_mat_GM_ess}/{len_mat}\")\n",
    "\n",
    "\n",
    "\n",
    "    # Return the matched DataFrame and gene counts\n",
    "    return matched_genes_essential_df, len_mat, len_mat_ess, len_mat_GM_ess\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CHOOSE ONE!!!!\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def find_and_match_essential_genes(model, df_exp_essential, save_csv=True, print_df=False):\n",
    "    \"\"\"\n",
    "    This function finds and matches essential genes in the model with experimental Entrez gene data.\n",
    "    It returns the number of total matched genes (len_mat) and the number of essential matched genes (len_mat_ess).\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if the necessary columns 'entrez_gene_id' and 'gene' exist in the dataframe\n",
    "    required_columns = ['entrez_gene_id', 'gene']\n",
    "    for col in required_columns:\n",
    "        if col not in df_exp_essential.columns:\n",
    "            raise KeyError(f\"Column '{col}' is missing from the dataframe.\")\n",
    "\n",
    "    # Create a reverse lookup dictionary where the key is the Entrez Gene ID and the value is the gene symbol\n",
    "    entrez_dict = df_exp_essential.set_index('entrez_gene_id')['gene'].to_dict()\n",
    "\n",
    "    # Matching metabolic genes\n",
    "    tmp_gene = []    \n",
    "    for gene in model.genes:\n",
    "        # Check if the gene.id matches any Entrez Gene ID in the reverse dictionary\n",
    "        if gene.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene.id]\n",
    "            tmp_gene.append((gene.id, gene.name, entrez_gene_symbol))\n",
    "    \n",
    "    # Number of total matched genes\n",
    "    len_mat = len(tmp_gene)\n",
    "\n",
    "    # Find the essential genes for the given model\n",
    "    essential_genes = find_essential_genes(model)\n",
    "    \n",
    "    # Create a list to store the matched essential genes\n",
    "    matched_genes = []\n",
    "\n",
    "    # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "    for gene in essential_genes:\n",
    "        gene_exp_ess = model.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "        if gene_exp_ess.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene_exp_ess.id]\n",
    "            matched_genes.append((gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol))\n",
    "\n",
    "    # Convert the matched essential genes into a DataFrame\n",
    "    matched_genes_essential_df = pd.DataFrame(matched_genes, columns=['gene_exp_ess.id', 'gene_exp_ess.name', 'entrez_gene_symbol'])\n",
    "\n",
    "    # Save the result to a CSV file if save_csv is True\n",
    "    if save_csv:\n",
    "        matched_genes_essential_df.to_csv(\"matched_gene_entrez_ids_feg.csv\", index=False)\n",
    "\n",
    "    # Print the DataFrame if print_df is True\n",
    "    if print_df:\n",
    "        print(matched_genes_essential_df)\n",
    "\n",
    "    # Number of essential matched genes\n",
    "    len_mat_ess = len(matched_genes_essential_df)\n",
    "    \n",
    "#     if model_GM is None:\n",
    "        \n",
    "        \n",
    "        \n",
    "    # Print the summary information\n",
    "    print(f\"Total matched metabolic genes: {len_mat}\")\n",
    "    print(f\"Matched essential genes: {len_mat_ess}/{len_mat}\")\n",
    "\n",
    "    # Return the matched DataFrame and gene counts\n",
    "    return matched_genes_essential_df, len_mat, len_mat_ess\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_reaction_ids_specific(model):\n",
    "    # Iterate over each reaction in the model\n",
    "    for rxn in model.reactions:\n",
    "        # Check if the reaction ID contains \"(e)\"\n",
    "        if \"(e)\" in rxn.id:\n",
    "            # Replace \"(e)\" with \"_e\"\n",
    "            new_id = rxn.id.replace(\"(e)\", \"_e\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(c)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(c)\", \"_c\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(x)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(x)\", \"_x\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(r)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(r)\", \"_r\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(g)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(g)\", \"_g\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_e_\" in rxn.id:\n",
    "            # Replace \"(e)\" with \"_e\"\n",
    "            new_id = rxn.id.replace(\"_e_\", \"_e\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_c_\" in rxn.id:\n",
    "            # Replace \"_c_\" with \"_c\"\n",
    "            new_id = rxn.id.replace(\"_c_\", \"_c\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_x_\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"_x_\", \"_x\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_r_\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"_r_\", \"_r\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_g_\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"_g_\", \"_g\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_10fthf(e) -> EX_10fthf_e\n",
      "Updating reaction ID: EX_10fthf5glu(e) -> EX_10fthf5glu_e\n",
      "Updating reaction ID: EX_10fthf6glu(e) -> EX_10fthf6glu_e\n",
      "Updating reaction ID: EX_10fthf7glu(e) -> EX_10fthf7glu_e\n",
      "Updating reaction ID: EX_11_cis_retfa(e) -> EX_11_cis_retfa_e\n",
      "Updating reaction ID: EX_13_cis_retnglc(e) -> EX_13_cis_retnglc_e\n",
      "Updating reaction ID: EX_1glyc_cho(e) -> EX_1glyc_cho_e\n",
      "Updating reaction ID: EX_1mncam(e) -> EX_1mncam_e\n",
      "Updating reaction ID: EX_23cump(e) -> EX_23cump_e\n",
      "Updating reaction ID: EX_2425dhvitd2(e) -> EX_2425dhvitd2_e\n",
      "Updating reaction ID: EX_2425dhvitd3(e) -> EX_2425dhvitd3_e\n",
      "Updating reaction ID: EX_24nph(e) -> EX_24nph_e\n",
      "Updating reaction ID: EX_25hvitd2(e) -> EX_25hvitd2_e\n",
      "Updating reaction ID: EX_25hvitd3(e) -> EX_25hvitd3_e\n",
      "Updating reaction ID: EX_2hb(e) -> EX_2hb_e\n",
      "Updating reaction ID: EX_2mcit(e) -> EX_2mcit_e\n",
      "Updating reaction ID: EX_34dhoxpeg(e) -> EX_34dhoxpeg_e\n",
      "Updating reaction ID: EX_34dhphe(e) -> EX_34dhphe_e\n",
      "Updating reaction ID: EX_35cgmp(e) -> EX_35cgmp_e\n",
      "Updating reaction ID: EX_3aib_D(e) -> EX_3aib_D_e\n",
      "Updating reaction ID: EX_3aib(e) -> EX_3aib_e\n",
      "Updating reaction ID: EX_3mlda(e) -> EX_3mlda_e\n",
      "Updating reaction ID: EX_3mob(e) -> EX_3mob_e\n",
      "Updating reaction ID: EX_3mop(e) -> EX_3mop_e\n",
      "Updating reaction ID: EX_3ump(e) -> EX_3ump_e\n",
      "Updating reaction ID: EX_4abut(e) -> EX_4abut_e\n",
      "Updating reaction ID: EX_4abutn(e) -> EX_4abutn_e\n",
      "Updating reaction ID: EX_4hdebrisoquine(e) -> EX_4hdebrisoquine_e\n",
      "Updating reaction ID: EX_4hphac(e) -> EX_4hphac_e\n",
      "Updating reaction ID: EX_4hpro(e) -> EX_4hpro_e\n",
      "Updating reaction ID: EX_4mop(e) -> EX_4mop_e\n",
      "Updating reaction ID: EX_4mptnl(e) -> EX_4mptnl_e\n",
      "Updating reaction ID: EX_4mtolbutamide(e) -> EX_4mtolbutamide_e\n",
      "Updating reaction ID: EX_4nph(e) -> EX_4nph_e\n",
      "Updating reaction ID: EX_4nphsf(e) -> EX_4nphsf_e\n",
      "Updating reaction ID: EX_4pyrdx(e) -> EX_4pyrdx_e\n",
      "Updating reaction ID: EX_5adtststerone(e) -> EX_5adtststerone_e\n",
      "Updating reaction ID: EX_5adtststeroneglc(e) -> EX_5adtststeroneglc_e\n",
      "Updating reaction ID: EX_5adtststerones(e) -> EX_5adtststerones_e\n",
      "Updating reaction ID: EX_5dhf(e) -> EX_5dhf_e\n",
      "Updating reaction ID: EX_5fthf(e) -> EX_5fthf_e\n",
      "Updating reaction ID: EX_5homeprazole(e) -> EX_5homeprazole_e\n",
      "Updating reaction ID: EX_5htrp(e) -> EX_5htrp_e\n",
      "Updating reaction ID: EX_5mta(e) -> EX_5mta_e\n",
      "Updating reaction ID: EX_5mthf(e) -> EX_5mthf_e\n",
      "Updating reaction ID: EX_5oxpro(e) -> EX_5oxpro_e\n",
      "Updating reaction ID: EX_5thf(e) -> EX_5thf_e\n",
      "Updating reaction ID: EX_6dhf(e) -> EX_6dhf_e\n",
      "Updating reaction ID: EX_6htststerone(e) -> EX_6htststerone_e\n",
      "Updating reaction ID: EX_6thf(e) -> EX_6thf_e\n",
      "Updating reaction ID: EX_7dhf(e) -> EX_7dhf_e\n",
      "Updating reaction ID: EX_7thf(e) -> EX_7thf_e\n",
      "Updating reaction ID: EX_9_cis_retfa(e) -> EX_9_cis_retfa_e\n",
      "Updating reaction ID: EX_abt(e) -> EX_abt_e\n",
      "Updating reaction ID: EX_ac(e) -> EX_ac_e\n",
      "Updating reaction ID: EX_acac(e) -> EX_acac_e\n",
      "Updating reaction ID: EX_acald(e) -> EX_acald_e\n",
      "Updating reaction ID: EX_acetone(e) -> EX_acetone_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho(e) -> EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho(e) -> EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgam(e) -> EX_acgam_e\n",
      "Updating reaction ID: EX_ach(e) -> EX_ach_e\n",
      "Updating reaction ID: EX_acmana(e) -> EX_acmana_e\n",
      "Updating reaction ID: EX_acn13acngalgbside_cho(e) -> EX_acn13acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acn23acngalgbside_cho(e) -> EX_acn23acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acnacngal14acglcgalgluside_cho(e) -> EX_acnacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acnacngalgbside_cho(e) -> EX_acnacngalgbside_cho_e\n",
      "Updating reaction ID: EX_acngalacglcgal14acglcgalgluside_cho(e) -> EX_acngalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_ade(e) -> EX_ade_e\n",
      "Updating reaction ID: EX_adn(e) -> EX_adn_e\n",
      "Updating reaction ID: EX_adpcbl(e) -> EX_adpcbl_e\n",
      "Updating reaction ID: EX_adprbp(e) -> EX_adprbp_e\n",
      "Updating reaction ID: EX_adprib(e) -> EX_adprib_e\n",
      "Updating reaction ID: EX_adrn(e) -> EX_adrn_e\n",
      "Updating reaction ID: EX_adrnl(e) -> EX_adrnl_e\n",
      "Updating reaction ID: EX_aflatoxin(e) -> EX_aflatoxin_e\n",
      "Updating reaction ID: EX_ahandrostanglc(e) -> EX_ahandrostanglc_e\n",
      "Updating reaction ID: EX_ahcys(e) -> EX_ahcys_e\n",
      "Updating reaction ID: EX_ahdt(e) -> EX_ahdt_e\n",
      "Updating reaction ID: EX_aicar(e) -> EX_aicar_e\n",
      "Updating reaction ID: EX_ak2lgchol_cho(e) -> EX_ak2lgchol_cho_e\n",
      "Updating reaction ID: EX_akg(e) -> EX_akg_e\n",
      "Updating reaction ID: EX_ala_B(e) -> EX_ala_B_e\n",
      "Updating reaction ID: EX_ala_D(e) -> EX_ala_D_e\n",
      "Updating reaction ID: EX_ala_L(e) -> EX_ala_L_e\n",
      "Updating reaction ID: EX_alaala(e) -> EX_alaala_e\n",
      "Updating reaction ID: EX_aldstrn(e) -> EX_aldstrn_e\n",
      "Updating reaction ID: EX_amp(e) -> EX_amp_e\n",
      "Updating reaction ID: EX_andrstrn(e) -> EX_andrstrn_e\n",
      "Updating reaction ID: EX_andrstrnglc(e) -> EX_andrstrnglc_e\n",
      "Updating reaction ID: EX_anth(e) -> EX_anth_e\n",
      "Updating reaction ID: EX_antipyrene(e) -> EX_antipyrene_e\n",
      "Updating reaction ID: EX_apnnox(e) -> EX_apnnox_e\n",
      "Updating reaction ID: EX_appnn(e) -> EX_appnn_e\n",
      "Updating reaction ID: EX_aprgstrn(e) -> EX_aprgstrn_e\n",
      "Updating reaction ID: EX_aqcobal(e) -> EX_aqcobal_e\n",
      "Updating reaction ID: EX_arab_L(e) -> EX_arab_L_e\n",
      "Updating reaction ID: EX_arach(e) -> EX_arach_e\n",
      "Updating reaction ID: EX_arachd(e) -> EX_arachd_e\n",
      "Updating reaction ID: EX_arg_L(e) -> EX_arg_L_e\n",
      "Updating reaction ID: EX_ascb_L(e) -> EX_ascb_L_e\n",
      "Updating reaction ID: EX_asn_L(e) -> EX_asn_L_e\n",
      "Updating reaction ID: EX_asp_D(e) -> EX_asp_D_e\n",
      "Updating reaction ID: EX_asp_L(e) -> EX_asp_L_e\n",
      "Updating reaction ID: EX_atp(e) -> EX_atp_e\n",
      "Updating reaction ID: EX_avite1(e) -> EX_avite1_e\n",
      "Updating reaction ID: EX_bhb(e) -> EX_bhb_e\n",
      "Updating reaction ID: EX_bildglcur(e) -> EX_bildglcur_e\n",
      "Updating reaction ID: EX_bilglcur(e) -> EX_bilglcur_e\n",
      "Updating reaction ID: EX_bilirub(e) -> EX_bilirub_e\n",
      "Updating reaction ID: EX_biocyt(e) -> EX_biocyt_e\n",
      "Updating reaction ID: EX_btn(e) -> EX_btn_e\n",
      "Updating reaction ID: EX_but(e) -> EX_but_e\n",
      "Updating reaction ID: EX_bz(e) -> EX_bz_e\n",
      "Updating reaction ID: EX_xanta(e) -> EX_xanta_e\n",
      "Updating reaction ID: EX_cdchac(e) -> EX_cdchac_e\n",
      "Updating reaction ID: EX_ca2(e) -> EX_ca2_e\n",
      "Updating reaction ID: EX_camp(e) -> EX_camp_e\n",
      "Updating reaction ID: EX_carn(e) -> EX_carn_e\n",
      "Updating reaction ID: EX_caro(e) -> EX_caro_e\n",
      "Updating reaction ID: EX_carveol(e) -> EX_carveol_e\n",
      "Updating reaction ID: EX_cbasp(e) -> EX_cbasp_e\n",
      "Updating reaction ID: EX_cca_d3(e) -> EX_cca_d3_e\n",
      "Updating reaction ID: EX_cdp(e) -> EX_cdp_e\n",
      "Updating reaction ID: EX_cgly(e) -> EX_cgly_e\n",
      "Updating reaction ID: EX_chol(e) -> EX_chol_e\n",
      "Updating reaction ID: EX_cholate(e) -> EX_cholate_e\n",
      "Updating reaction ID: EX_chopg(e) -> EX_chopg_e\n",
      "Updating reaction ID: EX_chsterol(e) -> EX_chsterol_e\n",
      "Updating reaction ID: EX_chtn(e) -> EX_chtn_e\n",
      "Updating reaction ID: EX_cit(e) -> EX_cit_e\n",
      "Updating reaction ID: EX_citr_L(e) -> EX_citr_L_e\n",
      "Updating reaction ID: EX_cl(e) -> EX_cl_e\n",
      "Updating reaction ID: EX_clpnd(e) -> EX_clpnd_e\n",
      "Updating reaction ID: EX_cmp(e) -> EX_cmp_e\n",
      "Updating reaction ID: EX_co(e) -> EX_co_e\n",
      "Updating reaction ID: EX_co2(e) -> EX_co2_e\n",
      "Updating reaction ID: EX_coumarin(e) -> EX_coumarin_e\n",
      "Updating reaction ID: EX_creat(e) -> EX_creat_e\n",
      "Updating reaction ID: EX_crm_cho(e) -> EX_crm_cho_e\n",
      "Updating reaction ID: EX_crmp_cho(e) -> EX_crmp_cho_e\n",
      "Updating reaction ID: EX_crn(e) -> EX_crn_e\n",
      "Updating reaction ID: EX_crtsl(e) -> EX_crtsl_e\n",
      "Updating reaction ID: EX_crtstrn(e) -> EX_crtstrn_e\n",
      "Updating reaction ID: EX_crvnc(e) -> EX_crvnc_e\n",
      "Updating reaction ID: EX_csn(e) -> EX_csn_e\n",
      "Updating reaction ID: EX_cspg_a(e) -> EX_cspg_a_e\n",
      "Updating reaction ID: EX_cspg_b(e) -> EX_cspg_b_e\n",
      "Updating reaction ID: EX_cspg_c(e) -> EX_cspg_c_e\n",
      "Updating reaction ID: EX_cspg_d(e) -> EX_cspg_d_e\n",
      "Updating reaction ID: EX_cspg(e)e_ -> EX_cspg_ee_\n",
      "Updating reaction ID: EX_ctp(e) -> EX_ctp_e\n",
      "Updating reaction ID: EX_cyan(e) -> EX_cyan_e\n",
      "Updating reaction ID: EX_cys_L(e) -> EX_cys_L_e\n",
      "Updating reaction ID: EX_cytd(e) -> EX_cytd_e\n",
      "Updating reaction ID: EX_dad_2(e) -> EX_dad_2_e\n",
      "Updating reaction ID: EX_dag_cho(e) -> EX_dag_cho_e\n",
      "Updating reaction ID: EX_dca(e) -> EX_dca_e\n",
      "Updating reaction ID: EX_dcsptn1(e) -> EX_dcsptn1_e\n",
      "Updating reaction ID: EX_dcyt(e) -> EX_dcyt_e\n",
      "Updating reaction ID: EX_debrisoquine(e) -> EX_debrisoquine_e\n",
      "Updating reaction ID: EX_dgchol(e) -> EX_dgchol_e\n",
      "Updating reaction ID: EX_dgmp(e) -> EX_dgmp_e\n",
      "Updating reaction ID: EX_dgsn(e) -> EX_dgsn_e\n",
      "Updating reaction ID: EX_dgtp(e) -> EX_dgtp_e\n",
      "Updating reaction ID: EX_dha(e) -> EX_dha_e\n",
      "Updating reaction ID: EX_dhap(e) -> EX_dhap_e\n",
      "Updating reaction ID: EX_dhdascb(e) -> EX_dhdascb_e\n",
      "Updating reaction ID: EX_dheas(e) -> EX_dheas_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_dhf(e) -> EX_dhf_e\n",
      "Updating reaction ID: EX_digalsgalside_cho(e) -> EX_digalsgalside_cho_e\n",
      "Updating reaction ID: EX_din(e) -> EX_din_e\n",
      "Updating reaction ID: EX_dlnlcg(e) -> EX_dlnlcg_e\n",
      "Updating reaction ID: EX_dmantipyrine(e) -> EX_dmantipyrine_e\n",
      "Updating reaction ID: EX_dmhptcrn(e) -> EX_dmhptcrn_e\n",
      "Updating reaction ID: EX_dopa(e) -> EX_dopa_e\n",
      "Updating reaction ID: EX_dopasf(e) -> EX_dopasf_e\n",
      "Updating reaction ID: EX_drib(e) -> EX_drib_e\n",
      "Updating reaction ID: EX_dtdp(e) -> EX_dtdp_e\n",
      "Updating reaction ID: EX_dtmp(e) -> EX_dtmp_e\n",
      "Updating reaction ID: EX_dttp(e) -> EX_dttp_e\n",
      "Updating reaction ID: EX_duri(e) -> EX_duri_e\n",
      "Updating reaction ID: EX_eaflatoxin(e) -> EX_eaflatoxin_e\n",
      "Updating reaction ID: EX_ebastine(e) -> EX_ebastine_e\n",
      "Updating reaction ID: EX_ebastineoh(e) -> EX_ebastineoh_e\n",
      "Updating reaction ID: EX_eicostet(e) -> EX_eicostet_e\n",
      "Updating reaction ID: EX_elaid(e) -> EX_elaid_e\n",
      "Updating reaction ID: EX_estradiol(e) -> EX_estradiol_e\n",
      "Updating reaction ID: EX_estradiolglc(e) -> EX_estradiolglc_e\n",
      "Updating reaction ID: EX_estriolglc(e) -> EX_estriolglc_e\n",
      "Updating reaction ID: EX_estroneglc(e) -> EX_estroneglc_e\n",
      "Updating reaction ID: EX_estrones(e) -> EX_estrones_e\n",
      "Updating reaction ID: EX_etoh(e) -> EX_etoh_e\n",
      "Updating reaction ID: EX_fad(e) -> EX_fad_e\n",
      "Updating reaction ID: EX_fald(e) -> EX_fald_e\n",
      "Updating reaction ID: EX_fe2(e) -> EX_fe2_e\n",
      "Updating reaction ID: EX_fe3(e) -> EX_fe3_e\n",
      "Updating reaction ID: EX_fmn(e) -> EX_fmn_e\n",
      "Updating reaction ID: EX_fol(e) -> EX_fol_e\n",
      "Updating reaction ID: EX_for(e) -> EX_for_e\n",
      "Updating reaction ID: EX_fru(e) -> EX_fru_e\n",
      "Updating reaction ID: EX_fuc_L(e) -> EX_fuc_L_e\n",
      "Updating reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho(e) -> EX_fuc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fuc14galacglcgalgluside_cho(e) -> EX_fuc14galacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacgalfucgalacglcgalgluside_cho(e) -> EX_fucacgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngal14acglcgalgluside_cho(e) -> EX_fucacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngalacglcgalgluside_cho(e) -> EX_fucacngalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc12gal14acglcgalgluside_cho(e) -> EX_fucfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho(e) -> EX_fucfuc132galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho(e) -> EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho(e) -> EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucgalacglcgalgluside_cho(e) -> EX_fucfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgal14acglcgalgluside_cho(e) -> EX_fucgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalfucgalacglcgalgluside_cho(e) -> EX_fucgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalgbside_cho(e) -> EX_fucgalgbside_cho_e\n",
      "Updating reaction ID: EX_g1p(e) -> EX_g1p_e\n",
      "Updating reaction ID: EX_gal(e) -> EX_gal_e\n",
      "Updating reaction ID: EX_galacglcgalgbside_cho(e) -> EX_galacglcgalgbside_cho_e\n",
      "Updating reaction ID: EX_galfuc12gal14acglcgalgluside_cho(e) -> EX_galfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho(e) -> EX_galfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho(e) -> EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalgalthcrm_cho(e) -> EX_galgalgalthcrm_cho_e\n",
      "Updating reaction ID: EX_gam(e) -> EX_gam_e\n",
      "Updating reaction ID: EX_gbside_cho(e) -> EX_gbside_cho_e\n",
      "Updating reaction ID: EX_gchola(e) -> EX_gchola_e\n",
      "Updating reaction ID: EX_gd1b2_cho(e) -> EX_gd1b2_cho_e\n",
      "Updating reaction ID: EX_gd1c_cho(e) -> EX_gd1c_cho_e\n",
      "Updating reaction ID: EX_gdp(e) -> EX_gdp_e\n",
      "Updating reaction ID: EX_glc(e) -> EX_glc_e\n",
      "Updating reaction ID: EX_glcur(e) -> EX_glcur_e\n",
      "Updating reaction ID: EX_gln_L(e) -> EX_gln_L_e\n",
      "Updating reaction ID: EX_glu_L(e) -> EX_glu_L_e\n",
      "Updating reaction ID: EX_gluala(e) -> EX_gluala_e\n",
      "Updating reaction ID: EX_gly(e) -> EX_gly_e\n",
      "Updating reaction ID: EX_glyb(e) -> EX_glyb_e\n",
      "Updating reaction ID: EX_glyc(e) -> EX_glyc_e\n",
      "Updating reaction ID: EX_glyc_S(e) -> EX_glyc_S_e\n",
      "Updating reaction ID: EX_glygly(e) -> EX_glygly_e\n",
      "Updating reaction ID: EX_glygn2(e) -> EX_glygn2_e\n",
      "Updating reaction ID: EX_glygn4(e) -> EX_glygn4_e\n",
      "Updating reaction ID: EX_glygn5(e) -> EX_glygn5_e\n",
      "Updating reaction ID: EX_glyleu(e) -> EX_glyleu_e\n",
      "Updating reaction ID: EX_glyphe(e) -> EX_glyphe_e\n",
      "Updating reaction ID: EX_glypro(e) -> EX_glypro_e\n",
      "Updating reaction ID: EX_glysar(e) -> EX_glysar_e\n",
      "Updating reaction ID: EX_gmp(e) -> EX_gmp_e\n",
      "Updating reaction ID: EX_gp1c_cho(e) -> EX_gp1c_cho_e\n",
      "Updating reaction ID: EX_gp1calpha_cho(e) -> EX_gp1calpha_cho_e\n",
      "Updating reaction ID: EX_gq1b_cho(e) -> EX_gq1b_cho_e\n",
      "Updating reaction ID: EX_gq1balpha_cho(e) -> EX_gq1balpha_cho_e\n",
      "Updating reaction ID: EX_gsn(e) -> EX_gsn_e\n",
      "Updating reaction ID: EX_gt1a_cho(e) -> EX_gt1a_cho_e\n",
      "Updating reaction ID: EX_gthox(e) -> EX_gthox_e\n",
      "Updating reaction ID: EX_gthrd(e) -> EX_gthrd_e\n",
      "Updating reaction ID: EX_gtp(e) -> EX_gtp_e\n",
      "Updating reaction ID: EX_gua(e) -> EX_gua_e\n",
      "Updating reaction ID: EX_h(e) -> EX_h_e\n",
      "Updating reaction ID: EX_h2o(e) -> EX_h2o_e\n",
      "Updating reaction ID: EX_h2o2(e) -> EX_h2o2_e\n",
      "Updating reaction ID: EX_ha(e) -> EX_ha_e\n",
      "Updating reaction ID: EX_ha_pre1(e) -> EX_ha_pre1_e\n",
      "Updating reaction ID: EX_ApoA1(e) -> EX_ApoA1_e\n",
      "Updating reaction ID: EX_isomal(e) -> EX_isomal_e\n",
      "Updating reaction ID: EX_h2s(e) -> EX_h2s_e\n",
      "Updating reaction ID: EX_chtbs(e) -> EX_chtbs_e\n",
      "Updating reaction ID: EX_cyala(e) -> EX_cyala_e\n",
      "Updating reaction ID: EX_npphos(e) -> EX_npphos_e\n",
      "Updating reaction ID: EX_dhnpt(e) -> EX_dhnpt_e\n",
      "Updating reaction ID: EX_3kgal(e) -> EX_3kgal_e\n",
      "Updating reaction ID: EX_lcts6p(e) -> EX_lcts6p_e\n",
      "Updating reaction ID: EX_ggl(e) -> EX_ggl_e\n",
      "Updating reaction ID: EX_3klcts(e) -> EX_3klcts_e\n",
      "Updating reaction ID: EX_glucyala(e) -> EX_glucyala_e\n",
      "Updating reaction ID: EX_uppg1(e) -> EX_uppg1_e\n",
      "Updating reaction ID: EX_cpppg1(e) -> EX_cpppg1_e\n",
      "Updating reaction ID: EX_gluampn(e) -> EX_gluampn_e\n",
      "Updating reaction ID: EX_lpdmd(e) -> EX_lpdmd_e\n",
      "Updating reaction ID: EX_gm4p(e) -> EX_gm4p_e\n",
      "Updating reaction ID: EX_zn2(e) -> EX_zn2_e\n",
      "Updating reaction ID: EX_ddca(e) -> EX_ddca_e\n",
      "Updating reaction ID: EX_wharachd(e) -> EX_wharachd_e\n",
      "Updating reaction ID: EX_txb2(e) -> EX_txb2_e\n",
      "Updating reaction ID: EX_335trithyr(e) -> EX_335trithyr_e\n",
      "Updating reaction ID: EX_gthleuc4(e) -> EX_gthleuc4_e\n",
      "Updating reaction ID: EX_gth24dnb(e) -> EX_gth24dnb_e\n",
      "Updating reaction ID: EX_gthetcrya(e) -> EX_gthetcrya_e\n",
      "Updating reaction ID: EX_prostga1(e) -> EX_prostga1_e\n",
      "Updating reaction ID: EX_prostga2(e) -> EX_prostga2_e\n",
      "Updating reaction ID: EX_prostgb1(e) -> EX_prostgb1_e\n",
      "Updating reaction ID: EX_prostgb2(e) -> EX_prostgb2_e\n",
      "Updating reaction ID: EX_prostgc1(e) -> EX_prostgc1_e\n",
      "Updating reaction ID: EX_prostgc2(e) -> EX_prostgc2_e\n",
      "Updating reaction ID: EX_prostge3(e) -> EX_prostge3_e\n",
      "Updating reaction ID: EX_hco3(e) -> EX_hco3_e\n",
      "Updating reaction ID: EX_hcoumarin(e) -> EX_hcoumarin_e\n",
      "Updating reaction ID: EX_hdca(e) -> EX_hdca_e\n",
      "Updating reaction ID: EX_hdcea(e) -> EX_hdcea_e\n",
      "Updating reaction ID: EX_hestratriol(e) -> EX_hestratriol_e\n",
      "Updating reaction ID: EX_hexc(e) -> EX_hexc_e\n",
      "Updating reaction ID: EX_his_L(e) -> EX_his_L_e\n",
      "Updating reaction ID: EX_hista(e) -> EX_hista_e\n",
      "Updating reaction ID: EX_hom_L(e) -> EX_hom_L_e\n",
      "Updating reaction ID: EX_hpdca(e) -> EX_hpdca_e\n",
      "Updating reaction ID: EX_htaxol(e) -> EX_htaxol_e\n",
      "Updating reaction ID: EX_hxan(e) -> EX_hxan_e\n",
      "Updating reaction ID: EX_i(e) -> EX_i_e\n",
      "Updating reaction ID: EX_idour(e) -> EX_idour_e\n",
      "Updating reaction ID: EX_idp(e) -> EX_idp_e\n",
      "Updating reaction ID: EX_ile_L(e) -> EX_ile_L_e\n",
      "Updating reaction ID: EX_imp(e) -> EX_imp_e\n",
      "Updating reaction ID: EX_inost(e) -> EX_inost_e\n",
      "Updating reaction ID: EX_ins(e) -> EX_ins_e\n",
      "Updating reaction ID: EX_k(e) -> EX_k_e\n",
      "Updating reaction ID: EX_ksi_deg1(e) -> EX_ksi_deg1_e\n",
      "Updating reaction ID: EX_ksi(e) -> EX_ksi_e\n",
      "Updating reaction ID: EX_ksii_core2(e) -> EX_ksii_core2_e\n",
      "Updating reaction ID: EX_ksii_core4(e) -> EX_ksii_core4_e\n",
      "Updating reaction ID: EX_lac_D(e) -> EX_lac_D_e\n",
      "Updating reaction ID: EX_lac_L(e) -> EX_lac_L_e\n",
      "Updating reaction ID: EX_lcts(e) -> EX_lcts_e\n",
      "Updating reaction ID: EX_Lcystin(e) -> EX_Lcystin_e\n",
      "Updating reaction ID: EX_leu_L(e) -> EX_leu_L_e\n",
      "Updating reaction ID: EX_leugly(e) -> EX_leugly_e\n",
      "Updating reaction ID: EX_leuktrA4(e) -> EX_leuktrA4_e\n",
      "Updating reaction ID: EX_leuktrB4(e) -> EX_leuktrB4_e\n",
      "Updating reaction ID: EX_leuktrC4(e) -> EX_leuktrC4_e\n",
      "Updating reaction ID: EX_leuktrD4(e) -> EX_leuktrD4_e\n",
      "Updating reaction ID: EX_leuktrE4(e) -> EX_leuktrE4_e\n",
      "Updating reaction ID: EX_leuktrF4(e) -> EX_leuktrF4_e\n",
      "Updating reaction ID: EX_leuleu(e) -> EX_leuleu_e\n",
      "Updating reaction ID: EX_lgnc(e) -> EX_lgnc_e\n",
      "Updating reaction ID: EX_limnen(e) -> EX_limnen_e\n",
      "Updating reaction ID: EX_lipoate(e) -> EX_lipoate_e\n",
      "Updating reaction ID: EX_lneldc(e) -> EX_lneldc_e\n",
      "Updating reaction ID: EX_lnlc(e) -> EX_lnlc_e\n",
      "Updating reaction ID: EX_lnlnca(e) -> EX_lnlnca_e\n",
      "Updating reaction ID: EX_lnlncg(e) -> EX_lnlncg_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_lpchol_cho(e) -> EX_lpchol_cho_e\n",
      "Updating reaction ID: EX_lys_L(e) -> EX_lys_L_e\n",
      "Updating reaction ID: EX_mag_cho(e) -> EX_mag_cho_e\n",
      "Updating reaction ID: EX_mal_L(e) -> EX_mal_L_e\n",
      "Updating reaction ID: EX_malt(e) -> EX_malt_e\n",
      "Updating reaction ID: EX_malttr(e) -> EX_malttr_e\n",
      "Updating reaction ID: EX_man(e) -> EX_man_e\n",
      "Updating reaction ID: EX_meoh(e) -> EX_meoh_e\n",
      "Updating reaction ID: EX_mepi(e) -> EX_mepi_e\n",
      "Updating reaction ID: EX_mercplaccys(e) -> EX_mercplaccys_e\n",
      "Updating reaction ID: EX_met_L(e) -> EX_met_L_e\n",
      "Updating reaction ID: EX_mthgxl(e) -> EX_mthgxl_e\n",
      "Updating reaction ID: EX_n2m2nmasn(e) -> EX_n2m2nmasn_e\n",
      "Updating reaction ID: EX_na1(e) -> EX_na1_e\n",
      "Updating reaction ID: EX_nac(e) -> EX_nac_e\n",
      "Updating reaction ID: EX_nad(e) -> EX_nad_e\n",
      "Updating reaction ID: EX_nadp(e) -> EX_nadp_e\n",
      "Updating reaction ID: EX_ncam(e) -> EX_ncam_e\n",
      "Updating reaction ID: EX_nh4(e) -> EX_nh4_e\n",
      "Updating reaction ID: EX_nifedipine(e) -> EX_nifedipine_e\n",
      "Updating reaction ID: EX_no(e) -> EX_no_e\n",
      "Updating reaction ID: EX_npthl(e) -> EX_npthl_e\n",
      "Updating reaction ID: EX_nrpphr(e) -> EX_nrpphr_e\n",
      "Updating reaction ID: EX_nrpphrsf(e) -> EX_nrpphrsf_e\n",
      "Updating reaction ID: EX_nrvnc(e) -> EX_nrvnc_e\n",
      "Updating reaction ID: EX_o2(e) -> EX_o2_e\n",
      "Updating reaction ID: EX_o2s(e) -> EX_o2s_e\n",
      "Updating reaction ID: EX_oagd3_cho(e) -> EX_oagd3_cho_e\n",
      "Updating reaction ID: EX_oagt3_cho(e) -> EX_oagt3_cho_e\n",
      "Updating reaction ID: EX_ocdca(e) -> EX_ocdca_e\n",
      "Updating reaction ID: EX_ocdcea(e) -> EX_ocdcea_e\n",
      "Updating reaction ID: EX_octa(e) -> EX_octa_e\n",
      "Updating reaction ID: EX_oh1(e) -> EX_oh1_e\n",
      "Updating reaction ID: EX_omeprazole(e) -> EX_omeprazole_e\n",
      "Updating reaction ID: EX_onpthl(e) -> EX_onpthl_e\n",
      "Updating reaction ID: EX_orn(e) -> EX_orn_e\n",
      "Updating reaction ID: EX_orot(e) -> EX_orot_e\n",
      "Updating reaction ID: EX_oxa(e) -> EX_oxa_e\n",
      "Updating reaction ID: EX_paf_cho(e) -> EX_paf_cho_e\n",
      "Updating reaction ID: EX_pchol_cho(e) -> EX_pchol_cho_e\n",
      "Updating reaction ID: EX_pe_cho(e) -> EX_pe_cho_e\n",
      "Updating reaction ID: EX_peplys(e) -> EX_peplys_e\n",
      "Updating reaction ID: EX_perillyl(e) -> EX_perillyl_e\n",
      "Updating reaction ID: EX_pglyc_cho(e) -> EX_pglyc_cho_e\n",
      "Updating reaction ID: EX_phe_L(e) -> EX_phe_L_e\n",
      "Updating reaction ID: EX_pheacgln(e) -> EX_pheacgln_e\n",
      "Updating reaction ID: EX_pheme(e) -> EX_pheme_e\n",
      "Updating reaction ID: EX_phyQ(e) -> EX_phyQ_e\n",
      "Updating reaction ID: EX_phyt(e) -> EX_phyt_e\n",
      "Updating reaction ID: EX_pi(e) -> EX_pi_e\n",
      "Updating reaction ID: EX_pnto_R(e) -> EX_pnto_R_e\n",
      "Updating reaction ID: EX_ppa(e) -> EX_ppa_e\n",
      "Updating reaction ID: EX_ppi(e) -> EX_ppi_e\n",
      "Updating reaction ID: EX_prgstrn(e) -> EX_prgstrn_e\n",
      "Updating reaction ID: EX_pro_D(e) -> EX_pro_D_e\n",
      "Updating reaction ID: EX_pro_L(e) -> EX_pro_L_e\n",
      "Updating reaction ID: EX_progly(e) -> EX_progly_e\n",
      "Updating reaction ID: EX_prostgd2(e) -> EX_prostgd2_e\n",
      "Updating reaction ID: EX_prostge1(e) -> EX_prostge1_e\n",
      "Updating reaction ID: EX_prostge2(e) -> EX_prostge2_e\n",
      "Updating reaction ID: EX_prostgf2(e) -> EX_prostgf2_e\n",
      "Updating reaction ID: EX_prostgh2(e) -> EX_prostgh2_e\n",
      "Updating reaction ID: EX_prostgi2(e) -> EX_prostgi2_e\n",
      "Updating reaction ID: EX_prpp(e) -> EX_prpp_e\n",
      "Updating reaction ID: EX_ps_cho(e) -> EX_ps_cho_e\n",
      "Updating reaction ID: EX_ptdca(e) -> EX_ptdca_e\n",
      "Updating reaction ID: EX_ptrc(e) -> EX_ptrc_e\n",
      "Updating reaction ID: EX_pydam(e) -> EX_pydam_e\n",
      "Updating reaction ID: EX_pydx(e) -> EX_pydx_e\n",
      "Updating reaction ID: EX_pydx5p(e) -> EX_pydx5p_e\n",
      "Updating reaction ID: EX_pydxn(e) -> EX_pydxn_e\n",
      "Updating reaction ID: EX_pyr(e) -> EX_pyr_e\n",
      "Updating reaction ID: EX_q10(e) -> EX_q10_e\n",
      "Updating reaction ID: EX_q10h2(e) -> EX_q10h2_e\n",
      "Updating reaction ID: EX_rbt(e) -> EX_rbt_e\n",
      "Updating reaction ID: EX_retfa(e) -> EX_retfa_e\n",
      "Updating reaction ID: EX_retinol(e) -> EX_retinol_e\n",
      "Updating reaction ID: EX_retn(e) -> EX_retn_e\n",
      "Updating reaction ID: EX_retnglc(e) -> EX_retnglc_e\n",
      "Updating reaction ID: EX_rib_D(e) -> EX_rib_D_e\n",
      "Updating reaction ID: EX_ribflv(e) -> EX_ribflv_e\n",
      "Updating reaction ID: EX_s2l2fn2m2masn(e) -> EX_s2l2fn2m2masn_e\n",
      "Updating reaction ID: EX_s2l2n2m2masn(e) -> EX_s2l2n2m2masn_e\n",
      "Updating reaction ID: EX_sarcs(e) -> EX_sarcs_e\n",
      "Updating reaction ID: EX_sbt_d(e) -> EX_sbt_d_e\n",
      "Updating reaction ID: EX_sel(e) -> EX_sel_e\n",
      "Updating reaction ID: EX_ser_D(e) -> EX_ser_D_e\n",
      "Updating reaction ID: EX_ser_L(e) -> EX_ser_L_e\n",
      "Updating reaction ID: EX_sfcys(e) -> EX_sfcys_e\n",
      "Updating reaction ID: EX_sl_L(e) -> EX_sl_L_e\n",
      "Updating reaction ID: EX_so4(e) -> EX_so4_e\n",
      "Updating reaction ID: EX_spc_cho(e) -> EX_spc_cho_e\n",
      "Updating reaction ID: EX_sph1p(e) -> EX_sph1p_e\n",
      "Updating reaction ID: EX_sphs1p(e) -> EX_sphs1p_e\n",
      "Updating reaction ID: EX_spmd(e) -> EX_spmd_e\n",
      "Updating reaction ID: EX_sprm(e) -> EX_sprm_e\n",
      "Updating reaction ID: EX_srtn(e) -> EX_srtn_e\n",
      "Updating reaction ID: EX_strch1(e) -> EX_strch1_e\n",
      "Updating reaction ID: EX_strch2(e) -> EX_strch2_e\n",
      "Updating reaction ID: EX_strdnc(e) -> EX_strdnc_e\n",
      "Updating reaction ID: EX_succ(e) -> EX_succ_e\n",
      "Updating reaction ID: EX_sucr(e) -> EX_sucr_e\n",
      "Updating reaction ID: EX_tag_cho(e) -> EX_tag_cho_e\n",
      "Updating reaction ID: EX_tagat_D(e) -> EX_tagat_D_e\n",
      "Updating reaction ID: EX_taur(e) -> EX_taur_e\n",
      "Updating reaction ID: EX_taxol(e) -> EX_taxol_e\n",
      "Updating reaction ID: EX_tchola(e) -> EX_tchola_e\n",
      "Updating reaction ID: EX_tcynt(e) -> EX_tcynt_e\n",
      "Updating reaction ID: EX_tdchola(e) -> EX_tdchola_e\n",
      "Updating reaction ID: EX_tethex3(e) -> EX_tethex3_e\n",
      "Updating reaction ID: EX_tetpent3(e) -> EX_tetpent3_e\n",
      "Updating reaction ID: EX_tetpent6(e) -> EX_tetpent6_e\n",
      "Updating reaction ID: EX_tettet6(e) -> EX_tettet6_e\n",
      "Updating reaction ID: EX_thf(e) -> EX_thf_e\n",
      "Updating reaction ID: EX_thm(e) -> EX_thm_e\n",
      "Updating reaction ID: EX_thmmp(e) -> EX_thmmp_e\n",
      "Updating reaction ID: EX_thmtp(e) -> EX_thmtp_e\n",
      "Updating reaction ID: EX_thr_L(e) -> EX_thr_L_e\n",
      "Updating reaction ID: EX_thym(e) -> EX_thym_e\n",
      "Updating reaction ID: EX_thymd(e) -> EX_thymd_e\n",
      "Updating reaction ID: EX_thyox_L(e) -> EX_thyox_L_e\n",
      "Updating reaction ID: EX_tmndnc(e) -> EX_tmndnc_e\n",
      "Updating reaction ID: EX_tolbutamide(e) -> EX_tolbutamide_e\n",
      "Updating reaction ID: EX_tre(e) -> EX_tre_e\n",
      "Updating reaction ID: EX_triodthy(e) -> EX_triodthy_e\n",
      "Updating reaction ID: EX_triodthysuf(e) -> EX_triodthysuf_e\n",
      "Updating reaction ID: EX_trp_L(e) -> EX_trp_L_e\n",
      "Updating reaction ID: EX_tststerone(e) -> EX_tststerone_e\n",
      "Updating reaction ID: EX_tststeroneglc(e) -> EX_tststeroneglc_e\n",
      "Updating reaction ID: EX_tststerones(e) -> EX_tststerones_e\n",
      "Updating reaction ID: EX_tsul(e) -> EX_tsul_e\n",
      "Updating reaction ID: EX_ttdca(e) -> EX_ttdca_e\n",
      "Updating reaction ID: EX_txa2(e) -> EX_txa2_e\n",
      "Updating reaction ID: EX_tymsf(e) -> EX_tymsf_e\n",
      "Updating reaction ID: EX_Tyr_ggn(e) -> EX_Tyr_ggn_e\n",
      "Updating reaction ID: EX_tyr_L(e) -> EX_tyr_L_e\n",
      "Updating reaction ID: EX_udp(e) -> EX_udp_e\n",
      "Updating reaction ID: EX_udpg(e) -> EX_udpg_e\n",
      "Updating reaction ID: EX_udpgal(e) -> EX_udpgal_e\n",
      "Updating reaction ID: EX_ump(e) -> EX_ump_e\n",
      "Updating reaction ID: EX_ura(e) -> EX_ura_e\n",
      "Updating reaction ID: EX_urate(e) -> EX_urate_e\n",
      "Updating reaction ID: EX_urea(e) -> EX_urea_e\n",
      "Updating reaction ID: EX_uri(e) -> EX_uri_e\n",
      "Updating reaction ID: EX_utp(e) -> EX_utp_e\n",
      "Updating reaction ID: EX_vacc(e) -> EX_vacc_e\n",
      "Updating reaction ID: EX_val_L(e) -> EX_val_L_e\n",
      "Updating reaction ID: EX_vitd2(e) -> EX_vitd2_e\n",
      "Updating reaction ID: EX_vitd3(e) -> EX_vitd3_e\n",
      "Updating reaction ID: EX_whddca(e) -> EX_whddca_e\n",
      "Updating reaction ID: EX_whhdca(e) -> EX_whhdca_e\n",
      "Updating reaction ID: EX_whtststerone(e) -> EX_whtststerone_e\n",
      "Updating reaction ID: EX_whttdca(e) -> EX_whttdca_e\n",
      "Updating reaction ID: EX_xolest2_cho(e) -> EX_xolest2_cho_e\n",
      "Updating reaction ID: EX_xoltri24(e) -> EX_xoltri24_e\n",
      "Updating reaction ID: EX_xoltri25(e) -> EX_xoltri25_e\n",
      "Updating reaction ID: EX_xoltri27(e) -> EX_xoltri27_e\n",
      "Updating reaction ID: EX_xyl_D(e) -> EX_xyl_D_e\n",
      "Updating reaction ID: EX_xylt(e) -> EX_xylt_e\n",
      "Updating reaction ID: EX_glcn(e) -> EX_glcn_e\n",
      "Updating reaction ID: EX_retinal(e) -> EX_retinal_e\n",
      "Updating reaction ID: EX_crtn(e) -> EX_crtn_e\n",
      "Updating reaction ID: EX_trypta(e) -> EX_trypta_e\n",
      "Updating reaction ID: EX_tym(e) -> EX_tym_e\n",
      "Updating reaction ID: EX_doco13ac(e) -> EX_doco13ac_e\n",
      "Updating reaction ID: EX_dhcholestanate(e) -> EX_dhcholestanate_e\n",
      "Updating reaction ID: EX_2hyoxplac(e) -> EX_2hyoxplac_e\n",
      "Updating reaction ID: EX_pheacgly(e) -> EX_pheacgly_e\n",
      "Updating reaction ID: EX_acglu(e) -> EX_acglu_e\n",
      "Updating reaction ID: EX_acorn(e) -> EX_acorn_e\n",
      "Updating reaction ID: EX_coke(e) -> EX_coke_e\n",
      "Updating reaction ID: EX_galt(e) -> EX_galt_e\n",
      "Updating reaction ID: EX_Lkynr(e) -> EX_Lkynr_e\n",
      "Updating reaction ID: EX_hLkynr(e) -> EX_hLkynr_e\n",
      "Updating reaction ID: EX_kynate(e) -> EX_kynate_e\n",
      "Updating reaction ID: EX_1a2425thvitd2(e) -> EX_1a2425thvitd2_e\n",
      "Updating reaction ID: EX_34dhoxmand(e) -> EX_34dhoxmand_e\n",
      "Updating reaction ID: EX_3mox4hoxm(e) -> EX_3mox4hoxm_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_gncore1(e) -> EX_gncore1_e\n",
      "Updating reaction ID: EX_ind3ac(e) -> EX_ind3ac_e\n",
      "Updating reaction ID: EX_12HPET(e) -> EX_12HPET_e\n",
      "Updating reaction ID: EX_15HPET(e) -> EX_15HPET_e\n",
      "Updating reaction ID: EX_arg_D(e) -> EX_arg_D_e\n",
      "Updating reaction ID: DM_oak2g_hs(c) -> DM_oak2g_hs_c\n",
      "Updating reaction ID: EX_mg2(e) -> EX_mg2_e\n",
      "Updating reaction ID: EX_cu2(e) -> EX_cu2_e\n",
      "Updating reaction ID: EX_eicostri(e) -> EX_eicostri_e\n",
      "Updating reaction ID: EX_docastri(e) -> EX_docastri_e\n",
      "Updating reaction ID: EX_Rtotal(e) -> EX_Rtotal_e\n",
      "Updating reaction ID: EX_Rtotal2(e) -> EX_Rtotal2_e\n",
      "Updating reaction ID: EX_Rtotal3(e) -> EX_Rtotal3_e\n",
      "Updating reaction ID: EX_HC01610_e_ -> EX_HC01610_e\n",
      "Updating reaction ID: EX_HC02179_e_ -> EX_HC02179_e\n",
      "Updating reaction ID: EX_C05965_e_ -> EX_C05965_e\n",
      "Updating reaction ID: EX_HC02176_e_ -> EX_HC02176_e\n",
      "Updating reaction ID: EX_10fthf_e_ -> EX_10fthf_e\n",
      "Updating reaction ID: EX_10fthf5glu_e_ -> EX_10fthf5glu_e\n",
      "Updating reaction ID: EX_10fthf6glu_e_ -> EX_10fthf6glu_e\n",
      "Updating reaction ID: EX_10fthf7glu_e_ -> EX_10fthf7glu_e\n",
      "Updating reaction ID: EX_11_cis_retfa_e_ -> EX_11_cis_retfa_e\n",
      "Updating reaction ID: EX_13_cis_retnglc_e_ -> EX_13_cis_retnglc_e\n",
      "Updating reaction ID: EX_1glyc_cho_e_ -> EX_1glyc_cho_e\n",
      "Updating reaction ID: EX_1mncam_e_ -> EX_1mncam_e\n",
      "Updating reaction ID: EX_23cump_e_ -> EX_23cump_e\n",
      "Updating reaction ID: EX_2425dhvitd2_e_ -> EX_2425dhvitd2_e\n",
      "Updating reaction ID: EX_2425dhvitd3_e_ -> EX_2425dhvitd3_e\n",
      "Updating reaction ID: EX_24nph_e_ -> EX_24nph_e\n",
      "Updating reaction ID: EX_25hvitd2_e_ -> EX_25hvitd2_e\n",
      "Updating reaction ID: EX_25hvitd3_e_ -> EX_25hvitd3_e\n",
      "Updating reaction ID: EX_2hb_e_ -> EX_2hb_e\n",
      "Updating reaction ID: EX_2mcit_e_ -> EX_2mcit_e\n",
      "Updating reaction ID: EX_34dhoxpeg_e_ -> EX_34dhoxpeg_e\n",
      "Updating reaction ID: EX_34dhphe_e_ -> EX_34dhphe_e\n",
      "Updating reaction ID: EX_35cgmp_e_ -> EX_35cgmp_e\n",
      "Updating reaction ID: EX_3aib_D_e_ -> EX_3aib_D_e\n",
      "Updating reaction ID: EX_3aib_e_ -> EX_3aib_e\n",
      "Updating reaction ID: EX_3mlda_e_ -> EX_3mlda_e\n",
      "Updating reaction ID: EX_3mob_e_ -> EX_3mob_e\n",
      "Updating reaction ID: EX_3mop_e_ -> EX_3mop_e\n",
      "Updating reaction ID: EX_3ump_e_ -> EX_3ump_e\n",
      "Updating reaction ID: EX_4abut_e_ -> EX_4abut_e\n",
      "Updating reaction ID: EX_4abutn_e_ -> EX_4abutn_e\n",
      "Updating reaction ID: EX_4hdebrisoquine_e_ -> EX_4hdebrisoquine_e\n",
      "Updating reaction ID: EX_4hphac_e_ -> EX_4hphac_e\n",
      "Updating reaction ID: EX_4hpro_e_ -> EX_4hpro_e\n",
      "Updating reaction ID: EX_4mop_e_ -> EX_4mop_e\n",
      "Updating reaction ID: EX_4mptnl_e_ -> EX_4mptnl_e\n",
      "Updating reaction ID: EX_4mtolbutamide_e_ -> EX_4mtolbutamide_e\n",
      "Updating reaction ID: EX_4nph_e_ -> EX_4nph_e\n",
      "Updating reaction ID: EX_4nphsf_e_ -> EX_4nphsf_e\n",
      "Updating reaction ID: EX_4pyrdx_e_ -> EX_4pyrdx_e\n",
      "Updating reaction ID: EX_5adtststerone_e_ -> EX_5adtststerone_e\n",
      "Updating reaction ID: EX_5adtststeroneglc_e_ -> EX_5adtststeroneglc_e\n",
      "Updating reaction ID: EX_5adtststerones_e_ -> EX_5adtststerones_e\n",
      "Updating reaction ID: EX_5dhf_e_ -> EX_5dhf_e\n",
      "Updating reaction ID: EX_5fthf_e_ -> EX_5fthf_e\n",
      "Updating reaction ID: EX_5homeprazole_e_ -> EX_5homeprazole_e\n",
      "Updating reaction ID: EX_5htrp_e_ -> EX_5htrp_e\n",
      "Updating reaction ID: EX_5mta_e_ -> EX_5mta_e\n",
      "Updating reaction ID: EX_5mthf_e_ -> EX_5mthf_e\n",
      "Updating reaction ID: EX_5oxpro_e_ -> EX_5oxpro_e\n",
      "Updating reaction ID: EX_5thf_e_ -> EX_5thf_e\n",
      "Updating reaction ID: EX_6dhf_e_ -> EX_6dhf_e\n",
      "Updating reaction ID: EX_6htststerone_e_ -> EX_6htststerone_e\n",
      "Updating reaction ID: EX_6thf_e_ -> EX_6thf_e\n",
      "Updating reaction ID: EX_7dhf_e_ -> EX_7dhf_e\n",
      "Updating reaction ID: EX_7thf_e_ -> EX_7thf_e\n",
      "Updating reaction ID: EX_9_cis_retfa_e_ -> EX_9_cis_retfa_e\n",
      "Updating reaction ID: EX_abt_e_ -> EX_abt_e\n",
      "Updating reaction ID: EX_ac_e_ -> EX_ac_e\n",
      "Updating reaction ID: EX_acac_e_ -> EX_acac_e\n",
      "Updating reaction ID: EX_acald_e_ -> EX_acald_e\n",
      "Updating reaction ID: EX_acetone_e_ -> EX_acetone_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgam_e_ -> EX_acgam_e\n",
      "Updating reaction ID: EX_ach_e_ -> EX_ach_e\n",
      "Updating reaction ID: EX_acmana_e_ -> EX_acmana_e\n",
      "Updating reaction ID: EX_acn13acngalgbside_cho_e_ -> EX_acn13acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acn23acngalgbside_cho_e_ -> EX_acn23acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acnacngal14acglcgalgluside_cho_e_ -> EX_acnacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acnacngalgbside_cho_e_ -> EX_acnacngalgbside_cho_e\n",
      "Updating reaction ID: EX_acngalacglcgal14acglcgalgluside_cho_e_ -> EX_acngalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_ade_e_ -> EX_ade_e\n",
      "Updating reaction ID: EX_adn_e_ -> EX_adn_e\n",
      "Updating reaction ID: EX_adpcbl_e_ -> EX_adpcbl_e\n",
      "Updating reaction ID: EX_adprbp_e_ -> EX_adprbp_e\n",
      "Updating reaction ID: EX_adprib_e_ -> EX_adprib_e\n",
      "Updating reaction ID: EX_adrn_e_ -> EX_adrn_e\n",
      "Updating reaction ID: EX_adrnl_e_ -> EX_adrnl_e\n",
      "Updating reaction ID: EX_aflatoxin_e_ -> EX_aflatoxin_e\n",
      "Updating reaction ID: EX_ahandrostanglc_e_ -> EX_ahandrostanglc_e\n",
      "Updating reaction ID: EX_ahcys_e_ -> EX_ahcys_e\n",
      "Updating reaction ID: EX_ahdt_e_ -> EX_ahdt_e\n",
      "Updating reaction ID: EX_aicar_e_ -> EX_aicar_e\n",
      "Updating reaction ID: EX_ak2lgchol_cho_e_ -> EX_ak2lgchol_cho_e\n",
      "Updating reaction ID: EX_akg_e_ -> EX_akg_e\n",
      "Updating reaction ID: EX_ala_B_e_ -> EX_ala_B_e\n",
      "Updating reaction ID: EX_ala_D_e_ -> EX_ala_D_e\n",
      "Updating reaction ID: EX_ala_L_e_ -> EX_ala_L_e\n",
      "Updating reaction ID: EX_alaala_e_ -> EX_alaala_e\n",
      "Updating reaction ID: EX_aldstrn_e_ -> EX_aldstrn_e\n",
      "Updating reaction ID: EX_amp_e_ -> EX_amp_e\n",
      "Updating reaction ID: EX_andrstrn_e_ -> EX_andrstrn_e\n",
      "Updating reaction ID: EX_andrstrnglc_e_ -> EX_andrstrnglc_e\n",
      "Updating reaction ID: EX_anth_e_ -> EX_anth_e\n",
      "Updating reaction ID: EX_antipyrene_e_ -> EX_antipyrene_e\n",
      "Updating reaction ID: EX_apnnox_e_ -> EX_apnnox_e\n",
      "Updating reaction ID: EX_appnn_e_ -> EX_appnn_e\n",
      "Updating reaction ID: EX_aprgstrn_e_ -> EX_aprgstrn_e\n",
      "Updating reaction ID: EX_aqcobal_e_ -> EX_aqcobal_e\n",
      "Updating reaction ID: EX_arab_L_e_ -> EX_arab_L_e\n",
      "Updating reaction ID: EX_arach_e_ -> EX_arach_e\n",
      "Updating reaction ID: EX_arachcoa_e_ -> EX_arachcoa_e\n",
      "Updating reaction ID: EX_arachd_e_ -> EX_arachd_e\n",
      "Updating reaction ID: EX_arg_L_e_ -> EX_arg_L_e\n",
      "Updating reaction ID: EX_ascb_L_e_ -> EX_ascb_L_e\n",
      "Updating reaction ID: EX_asn_L_e_ -> EX_asn_L_e\n",
      "Updating reaction ID: EX_asp_D_e_ -> EX_asp_D_e\n",
      "Updating reaction ID: EX_asp_L_e_ -> EX_asp_L_e\n",
      "Updating reaction ID: EX_atp_e_ -> EX_atp_e\n",
      "Updating reaction ID: EX_avite1_e_ -> EX_avite1_e\n",
      "Updating reaction ID: EX_avite2_e_ -> EX_avite2_e\n",
      "Updating reaction ID: EX_bglc_e_ -> EX_bglc_e\n",
      "Updating reaction ID: EX_bhb_e_ -> EX_bhb_e\n",
      "Updating reaction ID: EX_bildglcur_e_ -> EX_bildglcur_e\n",
      "Updating reaction ID: EX_bilglcur_e_ -> EX_bilglcur_e\n",
      "Updating reaction ID: EX_bilirub_e_ -> EX_bilirub_e\n",
      "Updating reaction ID: EX_biocyt_e_ -> EX_biocyt_e\n",
      "Updating reaction ID: EX_btn_e_ -> EX_btn_e\n",
      "Updating reaction ID: EX_but_e_ -> EX_but_e\n",
      "Updating reaction ID: EX_bvite_e_ -> EX_bvite_e\n",
      "Updating reaction ID: EX_bz_e_ -> EX_bz_e\n",
      "Updating reaction ID: EX_C02470_e_ -> EX_C02470_e\n",
      "Updating reaction ID: EX_C02528_e_ -> EX_C02528_e\n",
      "Updating reaction ID: EX_C04849_e_ -> EX_C04849_e\n",
      "Updating reaction ID: EX_ca2_e_ -> EX_ca2_e\n",
      "Updating reaction ID: EX_camp_e_ -> EX_camp_e\n",
      "Updating reaction ID: EX_carn_e_ -> EX_carn_e\n",
      "Updating reaction ID: EX_caro_e_ -> EX_caro_e\n",
      "Updating reaction ID: EX_carveol_e_ -> EX_carveol_e\n",
      "Updating reaction ID: EX_cbasp_e_ -> EX_cbasp_e\n",
      "Updating reaction ID: EX_cca_d3_e_ -> EX_cca_d3_e\n",
      "Updating reaction ID: EX_cdp_e_ -> EX_cdp_e\n",
      "Updating reaction ID: EX_CE0074_e_ -> EX_CE0074_e\n",
      "Updating reaction ID: EX_CE1925_e_ -> EX_CE1925_e\n",
      "Updating reaction ID: EX_CE1926_e_ -> EX_CE1926_e\n",
      "Updating reaction ID: EX_CE1935_e_ -> EX_CE1935_e\n",
      "Updating reaction ID: EX_CE1936_e_ -> EX_CE1936_e\n",
      "Updating reaction ID: EX_CE1939_e_ -> EX_CE1939_e\n",
      "Updating reaction ID: EX_CE1940_e_ -> EX_CE1940_e\n",
      "Updating reaction ID: EX_CE1943_e_ -> EX_CE1943_e\n",
      "Updating reaction ID: EX_CE2011_e_ -> EX_CE2011_e\n",
      "Updating reaction ID: EX_CE2250_e_ -> EX_CE2250_e\n",
      "Updating reaction ID: EX_CE2838_e_ -> EX_CE2838_e\n",
      "Updating reaction ID: EX_CE2839_e_ -> EX_CE2839_e\n",
      "Updating reaction ID: EX_CE2915_e_ -> EX_CE2915_e\n",
      "Updating reaction ID: EX_CE2916_e_ -> EX_CE2916_e\n",
      "Updating reaction ID: EX_CE2917_e_ -> EX_CE2917_e\n",
      "Updating reaction ID: EX_CE4633_e_ -> EX_CE4633_e\n",
      "Updating reaction ID: EX_CE4722_e_ -> EX_CE4722_e\n",
      "Updating reaction ID: EX_CE4723_e_ -> EX_CE4723_e\n",
      "Updating reaction ID: EX_CE4724_e_ -> EX_CE4724_e\n",
      "Updating reaction ID: EX_CE4881_e_ -> EX_CE4881_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_CE5786_e_ -> EX_CE5786_e\n",
      "Updating reaction ID: EX_CE5788_e_ -> EX_CE5788_e\n",
      "Updating reaction ID: EX_CE5789_e_ -> EX_CE5789_e\n",
      "Updating reaction ID: EX_CE5797_e_ -> EX_CE5797_e\n",
      "Updating reaction ID: EX_CE5798_e_ -> EX_CE5798_e\n",
      "Updating reaction ID: EX_CE5853_e_ -> EX_CE5853_e\n",
      "Updating reaction ID: EX_CE5854_e_ -> EX_CE5854_e\n",
      "Updating reaction ID: EX_CE5867_e_ -> EX_CE5867_e\n",
      "Updating reaction ID: EX_CE5868_e_ -> EX_CE5868_e\n",
      "Updating reaction ID: EX_CE5869_e_ -> EX_CE5869_e\n",
      "Updating reaction ID: EX_cgly_e_ -> EX_cgly_e\n",
      "Updating reaction ID: EX_chol_e_ -> EX_chol_e\n",
      "Updating reaction ID: EX_cholate_e_ -> EX_cholate_e\n",
      "Updating reaction ID: EX_chsterol_e_ -> EX_chsterol_e\n",
      "Updating reaction ID: EX_chtn_e_ -> EX_chtn_e\n",
      "Updating reaction ID: EX_cit_e_ -> EX_cit_e\n",
      "Updating reaction ID: EX_citr_L_e_ -> EX_citr_L_e\n",
      "Updating reaction ID: EX_cl_e_ -> EX_cl_e\n",
      "Updating reaction ID: EX_CLPND_e_ -> EX_CLPND_e\n",
      "Updating reaction ID: EX_cmp_e_ -> EX_cmp_e\n",
      "Updating reaction ID: EX_co_e_ -> EX_co_e\n",
      "Updating reaction ID: EX_co2_e_ -> EX_co2_e\n",
      "Updating reaction ID: EX_coa_e_ -> EX_coa_e\n",
      "Updating reaction ID: EX_coumarin_e_ -> EX_coumarin_e\n",
      "Updating reaction ID: EX_creat_e_ -> EX_creat_e\n",
      "Updating reaction ID: EX_crm_cho_e_ -> EX_crm_cho_e\n",
      "Updating reaction ID: EX_crmp_cho_e_ -> EX_crmp_cho_e\n",
      "Updating reaction ID: EX_crn_e_ -> EX_crn_e\n",
      "Updating reaction ID: EX_crtsl_e_ -> EX_crtsl_e\n",
      "Updating reaction ID: EX_crtstrn_e_ -> EX_crtstrn_e\n",
      "Updating reaction ID: EX_crvnc_e_ -> EX_crvnc_e\n",
      "Updating reaction ID: EX_csn_e_ -> EX_csn_e\n",
      "Updating reaction ID: EX_cspg_a_e_ -> EX_cspg_a_e\n",
      "Updating reaction ID: EX_cspg_b_e_ -> EX_cspg_b_e\n",
      "Updating reaction ID: EX_cspg_c_e_ -> EX_cspg_c_e\n",
      "Updating reaction ID: EX_cspg_d_e_ -> EX_cspg_d_e\n",
      "Updating reaction ID: EX_cspg_e_e_ -> EX_cspg_ee_\n",
      "Updating reaction ID: EX_ctp_e_ -> EX_ctp_e\n",
      "Updating reaction ID: EX_cyan_e_ -> EX_cyan_e\n",
      "Updating reaction ID: EX_cys_L_e_ -> EX_cys_L_e\n",
      "Updating reaction ID: EX_cysam_e_ -> EX_cysam_e\n",
      "Updating reaction ID: EX_cytd_e_ -> EX_cytd_e\n",
      "Updating reaction ID: EX_dad_2_e_ -> EX_dad_2_e\n",
      "Updating reaction ID: EX_dag_cho_e_ -> EX_dag_cho_e\n",
      "Updating reaction ID: EX_dca_e_ -> EX_dca_e\n",
      "Updating reaction ID: EX_dchac_e_ -> EX_dchac_e\n",
      "Updating reaction ID: EX_dcsptn1_e_ -> EX_dcsptn1_e\n",
      "Updating reaction ID: EX_dcyt_e_ -> EX_dcyt_e\n",
      "Updating reaction ID: EX_debrisoquine_e_ -> EX_debrisoquine_e\n",
      "Updating reaction ID: EX_dgchol_e_ -> EX_dgchol_e\n",
      "Updating reaction ID: EX_dgmp_e_ -> EX_dgmp_e\n",
      "Updating reaction ID: EX_dgsn_e_ -> EX_dgsn_e\n",
      "Updating reaction ID: EX_dgtp_e_ -> EX_dgtp_e\n",
      "Updating reaction ID: EX_dha_e_ -> EX_dha_e\n",
      "Updating reaction ID: EX_dhap_e_ -> EX_dhap_e\n",
      "Updating reaction ID: EX_dhdascb_e_ -> EX_dhdascb_e\n",
      "Updating reaction ID: EX_dheas_e_ -> EX_dheas_e\n",
      "Updating reaction ID: EX_dhf_e_ -> EX_dhf_e\n",
      "Updating reaction ID: EX_digalsgalside_cho_e_ -> EX_digalsgalside_cho_e\n",
      "Updating reaction ID: EX_din_e_ -> EX_din_e\n",
      "Updating reaction ID: EX_dlnlcg_e_ -> EX_dlnlcg_e\n",
      "Updating reaction ID: EX_dmantipyrine_e_ -> EX_dmantipyrine_e\n",
      "Updating reaction ID: EX_dmhptcrn_e_ -> EX_dmhptcrn_e\n",
      "Updating reaction ID: EX_dopa_e_ -> EX_dopa_e\n",
      "Updating reaction ID: EX_dopasf_e_ -> EX_dopasf_e\n",
      "Updating reaction ID: EX_dpcoa_e_ -> EX_dpcoa_e\n",
      "Updating reaction ID: EX_drib_e_ -> EX_drib_e\n",
      "Updating reaction ID: EX_dtdp_e_ -> EX_dtdp_e\n",
      "Updating reaction ID: EX_dtmp_e_ -> EX_dtmp_e\n",
      "Updating reaction ID: EX_dttp_e_ -> EX_dttp_e\n",
      "Updating reaction ID: EX_duri_e_ -> EX_duri_e\n",
      "Updating reaction ID: EX_eaflatoxin_e_ -> EX_eaflatoxin_e\n",
      "Updating reaction ID: EX_ebastine_e_ -> EX_ebastine_e\n",
      "Updating reaction ID: EX_ebastineoh_e_ -> EX_ebastineoh_e\n",
      "Updating reaction ID: EX_eicostet_e_ -> EX_eicostet_e\n",
      "Updating reaction ID: EX_elaid_e_ -> EX_elaid_e\n",
      "Updating reaction ID: EX_estradiol_e_ -> EX_estradiol_e\n",
      "Updating reaction ID: EX_estradiolglc_e_ -> EX_estradiolglc_e\n",
      "Updating reaction ID: EX_estriolglc_e_ -> EX_estriolglc_e\n",
      "Updating reaction ID: EX_estroneglc_e_ -> EX_estroneglc_e\n",
      "Updating reaction ID: EX_estrones_e_ -> EX_estrones_e\n",
      "Updating reaction ID: EX_etoh_e_ -> EX_etoh_e\n",
      "Updating reaction ID: EX_fad_e_ -> EX_fad_e\n",
      "Updating reaction ID: EX_fald_e_ -> EX_fald_e\n",
      "Updating reaction ID: EX_fe2_e_ -> EX_fe2_e\n",
      "Updating reaction ID: EX_fe3_e_ -> EX_fe3_e\n",
      "Updating reaction ID: EX_fmn_e_ -> EX_fmn_e\n",
      "Updating reaction ID: EX_fol_e_ -> EX_fol_e\n",
      "Updating reaction ID: EX_for_e_ -> EX_for_e\n",
      "Updating reaction ID: EX_fru_e_ -> EX_fru_e\n",
      "Updating reaction ID: EX_fuc_L_e_ -> EX_fuc_L_e\n",
      "Updating reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fuc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fuc14galacglcgalgluside_cho_e_ -> EX_fuc14galacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacgalfucgalacglcgalgluside_cho_e_ -> EX_fucacgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngal14acglcgalgluside_cho_e_ -> EX_fucacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngalacglcgalgluside_cho_e_ -> EX_fucacngalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc12gal14acglcgalgluside_cho_e_ -> EX_fucfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfuc132galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucgalacglcgalgluside_cho_e_ -> EX_fucfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgal14acglcgalgluside_cho_e_ -> EX_fucgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalfucgalacglcgalgluside_cho_e_ -> EX_fucgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalgbside_cho_e_ -> EX_fucgalgbside_cho_e\n",
      "Updating reaction ID: EX_g1p_e_ -> EX_g1p_e\n",
      "Updating reaction ID: EX_gal_e_ -> EX_gal_e\n",
      "Updating reaction ID: EX_galacglcgalgbside_cho_e_ -> EX_galacglcgalgbside_cho_e\n",
      "Updating reaction ID: EX_galfuc12gal14acglcgalgluside_cho_e_ -> EX_galfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_galfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e_ -> EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalgalthcrm_cho_e_ -> EX_galgalgalthcrm_cho_e\n",
      "Updating reaction ID: EX_galside_cho_e_ -> EX_galside_cho_e\n",
      "Updating reaction ID: EX_gam_e_ -> EX_gam_e\n",
      "Updating reaction ID: EX_gbside_cho_e_ -> EX_gbside_cho_e\n",
      "Updating reaction ID: EX_gchola_e_ -> EX_gchola_e\n",
      "Updating reaction ID: EX_gd1b2_cho_e_ -> EX_gd1b2_cho_e\n",
      "Updating reaction ID: EX_gd1c_cho_e_ -> EX_gd1c_cho_e\n",
      "Updating reaction ID: EX_gdp_e_ -> EX_gdp_e\n",
      "Updating reaction ID: EX_glc_e_ -> EX_glc_e\n",
      "Updating reaction ID: EX_glcur_e_ -> EX_glcur_e\n",
      "Updating reaction ID: EX_glgchlo_e_ -> EX_glgchlo_e\n",
      "Updating reaction ID: EX_gln_L_e_ -> EX_gln_L_e\n",
      "Updating reaction ID: EX_gltcho_e_ -> EX_gltcho_e\n",
      "Updating reaction ID: EX_gltdechol_e_ -> EX_gltdechol_e\n",
      "Updating reaction ID: EX_glu_L_e_ -> EX_glu_L_e\n",
      "Updating reaction ID: EX_gluala_e_ -> EX_gluala_e\n",
      "Updating reaction ID: EX_gly_e_ -> EX_gly_e\n",
      "Updating reaction ID: EX_glyb_e_ -> EX_glyb_e\n",
      "Updating reaction ID: EX_glyc_e_ -> EX_glyc_e\n",
      "Updating reaction ID: EX_glyc_S_e_ -> EX_glyc_S_e\n",
      "Updating reaction ID: EX_glygly_e_ -> EX_glygly_e\n",
      "Updating reaction ID: EX_glygn2_e_ -> EX_glygn2_e\n",
      "Updating reaction ID: EX_glygn4_e_ -> EX_glygn4_e\n",
      "Updating reaction ID: EX_glygn5_e_ -> EX_glygn5_e\n",
      "Updating reaction ID: EX_glyleu_e_ -> EX_glyleu_e\n",
      "Updating reaction ID: EX_glyphe_e_ -> EX_glyphe_e\n",
      "Updating reaction ID: EX_glypro_e_ -> EX_glypro_e\n",
      "Updating reaction ID: EX_glysar_e_ -> EX_glysar_e\n",
      "Updating reaction ID: EX_gmp_e_ -> EX_gmp_e\n",
      "Updating reaction ID: EX_gp1c_cho_e_ -> EX_gp1c_cho_e\n",
      "Updating reaction ID: EX_gp1calpha_cho_e_ -> EX_gp1calpha_cho_e\n",
      "Updating reaction ID: EX_gq1b_cho_e_ -> EX_gq1b_cho_e\n",
      "Updating reaction ID: EX_gq1balpha_cho_e_ -> EX_gq1balpha_cho_e\n",
      "Updating reaction ID: EX_gsn_e_ -> EX_gsn_e\n",
      "Updating reaction ID: EX_gt1a_cho_e_ -> EX_gt1a_cho_e\n",
      "Updating reaction ID: EX_gthox_e_ -> EX_gthox_e\n",
      "Updating reaction ID: EX_gthrd_e_ -> EX_gthrd_e\n",
      "Updating reaction ID: EX_gtp_e_ -> EX_gtp_e\n",
      "Updating reaction ID: EX_gua_e_ -> EX_gua_e\n",
      "Updating reaction ID: EX_gum_e_ -> EX_gum_e\n",
      "Updating reaction ID: EX_gumdchac_e_ -> EX_gumdchac_e\n",
      "Updating reaction ID: EX_gumgchol_e_ -> EX_gumgchol_e\n",
      "Updating reaction ID: EX_gumtchol_e_ -> EX_gumtchol_e\n",
      "Updating reaction ID: EX_h_e_ -> EX_h_e\n",
      "Updating reaction ID: EX_h2o_e_ -> EX_h2o_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_h2o2_e_ -> EX_h2o2_e\n",
      "Updating reaction ID: EX_ha_e_ -> EX_ha_e\n",
      "Updating reaction ID: EX_ha_pre1_e_ -> EX_ha_pre1_e\n",
      "Updating reaction ID: EX_HC00004_e_ -> EX_HC00004_e\n",
      "Updating reaction ID: EX_HC00229_e_ -> EX_HC00229_e\n",
      "Updating reaction ID: EX_HC00250_e_ -> EX_HC00250_e\n",
      "Updating reaction ID: EX_HC00822_e_ -> EX_HC00822_e\n",
      "Updating reaction ID: EX_HC00955_e_ -> EX_HC00955_e\n",
      "Updating reaction ID: EX_HC01104_e_ -> EX_HC01104_e\n",
      "Updating reaction ID: EX_HC01361_e_ -> EX_HC01361_e\n",
      "Updating reaction ID: EX_HC01440_e_ -> EX_HC01440_e\n",
      "Updating reaction ID: EX_HC01441_e_ -> EX_HC01441_e\n",
      "Updating reaction ID: EX_HC01444_e_ -> EX_HC01444_e\n",
      "Updating reaction ID: EX_HC01446_e_ -> EX_HC01446_e\n",
      "Updating reaction ID: EX_HC01577_e_ -> EX_HC01577_e\n",
      "Updating reaction ID: EX_HC01609_e_ -> EX_HC01609_e\n",
      "Updating reaction ID: EX_HC01700_e_ -> EX_HC01700_e\n",
      "Updating reaction ID: EX_HC01787_e_ -> EX_HC01787_e\n",
      "Updating reaction ID: EX_HC02154_e_ -> EX_HC02154_e\n",
      "Updating reaction ID: EX_HC02160_e_ -> EX_HC02160_e\n",
      "Updating reaction ID: EX_HC02161_e_ -> EX_HC02161_e\n",
      "Updating reaction ID: EX_HC02172_e_ -> EX_HC02172_e\n",
      "Updating reaction ID: EX_HC02180_e_ -> EX_HC02180_e\n",
      "Updating reaction ID: EX_HC02187_e_ -> EX_HC02187_e\n",
      "Updating reaction ID: EX_HC02191_e_ -> EX_HC02191_e\n",
      "Updating reaction ID: EX_HC02192_e_ -> EX_HC02192_e\n",
      "Updating reaction ID: EX_HC02193_e_ -> EX_HC02193_e\n",
      "Updating reaction ID: EX_HC02194_e_ -> EX_HC02194_e\n",
      "Updating reaction ID: EX_HC02195_e_ -> EX_HC02195_e\n",
      "Updating reaction ID: EX_HC02196_e_ -> EX_HC02196_e\n",
      "Updating reaction ID: EX_HC02197_e_ -> EX_HC02197_e\n",
      "Updating reaction ID: EX_HC02198_e_ -> EX_HC02198_e\n",
      "Updating reaction ID: EX_HC02199_e_ -> EX_HC02199_e\n",
      "Updating reaction ID: EX_HC02200_e_ -> EX_HC02200_e\n",
      "Updating reaction ID: EX_HC02201_e_ -> EX_HC02201_e\n",
      "Updating reaction ID: EX_HC02202_e_ -> EX_HC02202_e\n",
      "Updating reaction ID: EX_HC02203_e_ -> EX_HC02203_e\n",
      "Updating reaction ID: EX_HC02204_e_ -> EX_HC02204_e\n",
      "Updating reaction ID: EX_HC02205_e_ -> EX_HC02205_e\n",
      "Updating reaction ID: EX_HC02206_e_ -> EX_HC02206_e\n",
      "Updating reaction ID: EX_HC02207_e_ -> EX_HC02207_e\n",
      "Updating reaction ID: EX_HC02208_e_ -> EX_HC02208_e\n",
      "Updating reaction ID: EX_HC02210_e_ -> EX_HC02210_e\n",
      "Updating reaction ID: EX_HC02213_e_ -> EX_HC02213_e\n",
      "Updating reaction ID: EX_HC02214_e_ -> EX_HC02214_e\n",
      "Updating reaction ID: EX_HC02216_e_ -> EX_HC02216_e\n",
      "Updating reaction ID: EX_HC02217_e_ -> EX_HC02217_e\n",
      "Updating reaction ID: EX_HC02220_e_ -> EX_HC02220_e\n",
      "Updating reaction ID: EX_hco3_e_ -> EX_hco3_e\n",
      "Updating reaction ID: EX_hcoumarin_e_ -> EX_hcoumarin_e\n",
      "Updating reaction ID: EX_hdca_e_ -> EX_hdca_e\n",
      "Updating reaction ID: EX_hdcea_e_ -> EX_hdcea_e\n",
      "Updating reaction ID: EX_hestratriol_e_ -> EX_hestratriol_e\n",
      "Updating reaction ID: EX_hexc_e_ -> EX_hexc_e\n",
      "Updating reaction ID: EX_his_L_e_ -> EX_his_L_e\n",
      "Updating reaction ID: EX_hista_e_ -> EX_hista_e\n",
      "Updating reaction ID: EX_hom_L_e_ -> EX_hom_L_e\n",
      "Updating reaction ID: EX_hpdca_e_ -> EX_hpdca_e\n",
      "Updating reaction ID: EX_chopg_e_ -> EX_chopg_e\n",
      "Updating reaction ID: EX_htaxol_e_ -> EX_htaxol_e\n",
      "Updating reaction ID: EX_hxan_e_ -> EX_hxan_e\n",
      "Updating reaction ID: EX_hyptaur_e_ -> EX_hyptaur_e\n",
      "Updating reaction ID: EX_i_e_ -> EX_i_e\n",
      "Updating reaction ID: EX_idour_e_ -> EX_idour_e\n",
      "Updating reaction ID: EX_idp_e_ -> EX_idp_e\n",
      "Updating reaction ID: EX_ile_L_e_ -> EX_ile_L_e\n",
      "Updating reaction ID: EX_imp_e_ -> EX_imp_e\n",
      "Updating reaction ID: EX_inost_e_ -> EX_inost_e\n",
      "Updating reaction ID: EX_ins_e_ -> EX_ins_e\n",
      "Updating reaction ID: EX_k_e_ -> EX_k_e\n",
      "Updating reaction ID: EX_ksi_deg1_e_ -> EX_ksi_deg1_e\n",
      "Updating reaction ID: EX_ksi_e_ -> EX_ksi_e\n",
      "Updating reaction ID: EX_ksii_core2_e_ -> EX_ksii_core2_e\n",
      "Updating reaction ID: EX_ksii_core4_e_ -> EX_ksii_core4_e\n",
      "Updating reaction ID: EX_lac_D_e_ -> EX_lac_D_e\n",
      "Updating reaction ID: EX_lac_L_e_ -> EX_lac_L_e\n",
      "Updating reaction ID: EX_lcts_e_ -> EX_lcts_e\n",
      "Updating reaction ID: EX_Lcystin_e_ -> EX_Lcystin_e\n",
      "Updating reaction ID: EX_leu_L_e_ -> EX_leu_L_e\n",
      "Updating reaction ID: EX_leugly_e_ -> EX_leugly_e\n",
      "Updating reaction ID: EX_leuktrA4_e_ -> EX_leuktrA4_e\n",
      "Updating reaction ID: EX_leuktrB4_e_ -> EX_leuktrB4_e\n",
      "Updating reaction ID: EX_leuktrC4_e_ -> EX_leuktrC4_e\n",
      "Updating reaction ID: EX_leuktrD4_e_ -> EX_leuktrD4_e\n",
      "Updating reaction ID: EX_leuktrE4_e_ -> EX_leuktrE4_e\n",
      "Updating reaction ID: EX_leuktrF4_e_ -> EX_leuktrF4_e\n",
      "Updating reaction ID: EX_leuleu_e_ -> EX_leuleu_e\n",
      "Updating reaction ID: EX_lgnc_e_ -> EX_lgnc_e\n",
      "Updating reaction ID: EX_limnen_e_ -> EX_limnen_e\n",
      "Updating reaction ID: EX_lipoate_e_ -> EX_lipoate_e\n",
      "Updating reaction ID: EX_lneldc_e_ -> EX_lneldc_e\n",
      "Updating reaction ID: EX_lnlc_e_ -> EX_lnlc_e\n",
      "Updating reaction ID: EX_lnlnca_e_ -> EX_lnlnca_e\n",
      "Updating reaction ID: EX_lnlncg_e_ -> EX_lnlncg_e\n",
      "Updating reaction ID: EX_lpchol_cho_e_ -> EX_lpchol_cho_e\n",
      "Updating reaction ID: EX_lys_L_e_ -> EX_lys_L_e\n",
      "Updating reaction ID: EX_mag_cho_e_ -> EX_mag_cho_e\n",
      "Updating reaction ID: EX_mal_L_e_ -> EX_mal_L_e\n",
      "Updating reaction ID: EX_malcoa_e_ -> EX_malcoa_e\n",
      "Updating reaction ID: EX_malt_e_ -> EX_malt_e\n",
      "Updating reaction ID: EX_malthp_e_ -> EX_malthp_e\n",
      "Updating reaction ID: EX_malthx_e_ -> EX_malthx_e\n",
      "Updating reaction ID: EX_maltpt_e_ -> EX_maltpt_e\n",
      "Updating reaction ID: EX_malttr_e_ -> EX_malttr_e\n",
      "Updating reaction ID: EX_maltttr_e_ -> EX_maltttr_e\n",
      "Updating reaction ID: EX_man_e_ -> EX_man_e\n",
      "Updating reaction ID: EX_meoh_e_ -> EX_meoh_e\n",
      "Updating reaction ID: EX_mepi_e_ -> EX_mepi_e\n",
      "Updating reaction ID: EX_mercplaccys_e_ -> EX_mercplaccys_e\n",
      "Updating reaction ID: EX_met_L_e_ -> EX_met_L_e\n",
      "Updating reaction ID: EX_mthgxl_e_ -> EX_mthgxl_e\n",
      "Updating reaction ID: EX_n2m2nmasn_e_ -> EX_n2m2nmasn_e\n",
      "Updating reaction ID: EX_na1_e_ -> EX_na1_e\n",
      "Updating reaction ID: EX_nac_e_ -> EX_nac_e\n",
      "Updating reaction ID: EX_nad_e_ -> EX_nad_e\n",
      "Updating reaction ID: EX_nadp_e_ -> EX_nadp_e\n",
      "Updating reaction ID: EX_ncam_e_ -> EX_ncam_e\n",
      "Updating reaction ID: EX_nh4_e_ -> EX_nh4_e\n",
      "Updating reaction ID: EX_nifedipine_e_ -> EX_nifedipine_e\n",
      "Updating reaction ID: EX_no_e_ -> EX_no_e\n",
      "Updating reaction ID: EX_no2_e_ -> EX_no2_e\n",
      "Updating reaction ID: EX_npthl_e_ -> EX_npthl_e\n",
      "Updating reaction ID: EX_nrpphr_e_ -> EX_nrpphr_e\n",
      "Updating reaction ID: EX_nrpphrsf_e_ -> EX_nrpphrsf_e\n",
      "Updating reaction ID: EX_nrvnc_e_ -> EX_nrvnc_e\n",
      "Updating reaction ID: EX_o2_e_ -> EX_o2_e\n",
      "Updating reaction ID: EX_o2s_e_ -> EX_o2s_e\n",
      "Updating reaction ID: EX_oagd3_cho_e_ -> EX_oagd3_cho_e\n",
      "Updating reaction ID: EX_oagt3_cho_e_ -> EX_oagt3_cho_e\n",
      "Updating reaction ID: EX_ocdca_e_ -> EX_ocdca_e\n",
      "Updating reaction ID: EX_ocdcea_e_ -> EX_ocdcea_e\n",
      "Updating reaction ID: EX_octa_e_ -> EX_octa_e\n",
      "Updating reaction ID: EX_oh1_e_ -> EX_oh1_e\n",
      "Updating reaction ID: EX_omeprazole_e_ -> EX_omeprazole_e\n",
      "Updating reaction ID: EX_onpthl_e_ -> EX_onpthl_e\n",
      "Updating reaction ID: EX_orn_e_ -> EX_orn_e\n",
      "Updating reaction ID: EX_orot_e_ -> EX_orot_e\n",
      "Updating reaction ID: EX_oxa_e_ -> EX_oxa_e\n",
      "Updating reaction ID: EX_paf_cho_e_ -> EX_paf_cho_e\n",
      "Updating reaction ID: EX_pan4p_e_ -> EX_pan4p_e\n",
      "Updating reaction ID: EX_pchol_cho_e_ -> EX_pchol_cho_e\n",
      "Updating reaction ID: EX_pe_cho_e_ -> EX_pe_cho_e\n",
      "Updating reaction ID: EX_pect_e_ -> EX_pect_e\n",
      "Updating reaction ID: EX_pectindchac_e_ -> EX_pectindchac_e\n",
      "Updating reaction ID: EX_pectingchol_e_ -> EX_pectingchol_e\n",
      "Updating reaction ID: EX_pectintchol_e_ -> EX_pectintchol_e\n",
      "Updating reaction ID: EX_peplys_e_ -> EX_peplys_e\n",
      "Updating reaction ID: EX_perillyl_e_ -> EX_perillyl_e\n",
      "Updating reaction ID: EX_pglyc_cho_e_ -> EX_pglyc_cho_e\n",
      "Updating reaction ID: EX_phe_L_e_ -> EX_phe_L_e\n",
      "Updating reaction ID: EX_pheacgln_e_ -> EX_pheacgln_e\n",
      "Updating reaction ID: EX_pheme_e_ -> EX_pheme_e\n",
      "Updating reaction ID: EX_phyQ_e_ -> EX_phyQ_e\n",
      "Updating reaction ID: EX_phyt_e_ -> EX_phyt_e\n",
      "Updating reaction ID: EX_pi_e_ -> EX_pi_e\n",
      "Updating reaction ID: EX_pnto_R_e_ -> EX_pnto_R_e\n",
      "Updating reaction ID: EX_ppa_e_ -> EX_ppa_e\n",
      "Updating reaction ID: EX_ppi_e_ -> EX_ppi_e\n",
      "Updating reaction ID: EX_prgstrn_e_ -> EX_prgstrn_e\n",
      "Updating reaction ID: EX_pro_D_e_ -> EX_pro_D_e\n",
      "Updating reaction ID: EX_pro_L_e_ -> EX_pro_L_e\n",
      "Updating reaction ID: EX_progly_e_ -> EX_progly_e\n",
      "Updating reaction ID: EX_prostgd2_e_ -> EX_prostgd2_e\n",
      "Updating reaction ID: EX_prostge1_e_ -> EX_prostge1_e\n",
      "Updating reaction ID: EX_prostge2_e_ -> EX_prostge2_e\n",
      "Updating reaction ID: EX_prostgf2_e_ -> EX_prostgf2_e\n",
      "Updating reaction ID: EX_prostgh2_e_ -> EX_prostgh2_e\n",
      "Updating reaction ID: EX_prostgi2_e_ -> EX_prostgi2_e\n",
      "Updating reaction ID: EX_prpp_e_ -> EX_prpp_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_ps_cho_e_ -> EX_ps_cho_e\n",
      "Updating reaction ID: EX_psyl_e_ -> EX_psyl_e\n",
      "Updating reaction ID: EX_psylchol_e_ -> EX_psylchol_e\n",
      "Updating reaction ID: EX_psyltchol_e_ -> EX_psyltchol_e\n",
      "Updating reaction ID: EX_psyltdechol_e_ -> EX_psyltdechol_e\n",
      "Updating reaction ID: EX_ptdca_e_ -> EX_ptdca_e\n",
      "Updating reaction ID: EX_ptrc_e_ -> EX_ptrc_e\n",
      "Updating reaction ID: EX_ptth_e_ -> EX_ptth_e\n",
      "Updating reaction ID: EX_pydam_e_ -> EX_pydam_e\n",
      "Updating reaction ID: EX_pydx_e_ -> EX_pydx_e\n",
      "Updating reaction ID: EX_pydx5p_e_ -> EX_pydx5p_e\n",
      "Updating reaction ID: EX_pydxn_e_ -> EX_pydxn_e\n",
      "Updating reaction ID: EX_pyr_e_ -> EX_pyr_e\n",
      "Updating reaction ID: EX_q10_e_ -> EX_q10_e\n",
      "Updating reaction ID: EX_q10h2_e_ -> EX_q10h2_e\n",
      "Updating reaction ID: EX_rbt_e_ -> EX_rbt_e\n",
      "Updating reaction ID: EX_retfa_e_ -> EX_retfa_e\n",
      "Updating reaction ID: EX_retinol_9_cis_e_ -> EX_retinol_9_cis_e\n",
      "Updating reaction ID: EX_retinol_cis_11_e_ -> EX_retinol_cis_11_e\n",
      "Updating reaction ID: EX_retinol_e_ -> EX_retinol_e\n",
      "Updating reaction ID: EX_retn_e_ -> EX_retn_e\n",
      "Updating reaction ID: EX_retnglc_e_ -> EX_retnglc_e\n",
      "Updating reaction ID: EX_rib_D_e_ -> EX_rib_D_e\n",
      "Updating reaction ID: EX_ribflv_e_ -> EX_ribflv_e\n",
      "Updating reaction ID: EX_Rtotal_e_ -> EX_Rtotal_e\n",
      "Updating reaction ID: EX_Rtotal2_e_ -> EX_Rtotal2_e\n",
      "Updating reaction ID: EX_Rtotal3_e_ -> EX_Rtotal3_e\n",
      "Updating reaction ID: EX_s2l2fn2m2masn_e_ -> EX_s2l2fn2m2masn_e\n",
      "Updating reaction ID: EX_s2l2n2m2masn_e_ -> EX_s2l2n2m2masn_e\n",
      "Updating reaction ID: EX_sarcs_e_ -> EX_sarcs_e\n",
      "Updating reaction ID: EX_sbt_d_e_ -> EX_sbt_d_e\n",
      "Updating reaction ID: EX_sel_e_ -> EX_sel_e\n",
      "Updating reaction ID: EX_ser_D_e_ -> EX_ser_D_e\n",
      "Updating reaction ID: EX_ser_L_e_ -> EX_ser_L_e\n",
      "Updating reaction ID: EX_sfcys_e_ -> EX_sfcys_e\n",
      "Updating reaction ID: EX_sl_L_e_ -> EX_sl_L_e\n",
      "Updating reaction ID: EX_so4_e_ -> EX_so4_e\n",
      "Updating reaction ID: EX_spc_cho_e_ -> EX_spc_cho_e\n",
      "Updating reaction ID: EX_sph1p_e_ -> EX_sph1p_e\n",
      "Updating reaction ID: EX_sphs1p_e_ -> EX_sphs1p_e\n",
      "Updating reaction ID: EX_spmd_e_ -> EX_spmd_e\n",
      "Updating reaction ID: EX_sprm_e_ -> EX_sprm_e\n",
      "Updating reaction ID: EX_srtn_e_ -> EX_srtn_e\n",
      "Updating reaction ID: EX_strch1_e_ -> EX_strch1_e\n",
      "Updating reaction ID: EX_strch2_e_ -> EX_strch2_e\n",
      "Updating reaction ID: EX_strdnc_e_ -> EX_strdnc_e\n",
      "Updating reaction ID: EX_succ_e_ -> EX_succ_e\n",
      "Updating reaction ID: EX_sucr_e_ -> EX_sucr_e\n",
      "Updating reaction ID: EX_tag_cho_e_ -> EX_tag_cho_e\n",
      "Updating reaction ID: EX_tagat_D_e_ -> EX_tagat_D_e\n",
      "Updating reaction ID: EX_taur_e_ -> EX_taur_e\n",
      "Updating reaction ID: EX_taxol_e_ -> EX_taxol_e\n",
      "Updating reaction ID: EX_tchola_e_ -> EX_tchola_e\n",
      "Updating reaction ID: EX_tcynt_e_ -> EX_tcynt_e\n",
      "Updating reaction ID: EX_tdchola_e_ -> EX_tdchola_e\n",
      "Updating reaction ID: EX_tdechola_e_ -> EX_tdechola_e\n",
      "Updating reaction ID: EX_tethex3_e_ -> EX_tethex3_e\n",
      "Updating reaction ID: EX_tetpent3_e_ -> EX_tetpent3_e\n",
      "Updating reaction ID: EX_tetpent6_e_ -> EX_tetpent6_e\n",
      "Updating reaction ID: EX_tettet6_e_ -> EX_tettet6_e\n",
      "Updating reaction ID: EX_thf_e_ -> EX_thf_e\n",
      "Updating reaction ID: EX_thm_e_ -> EX_thm_e\n",
      "Updating reaction ID: EX_thmmp_e_ -> EX_thmmp_e\n",
      "Updating reaction ID: EX_thmtp_e_ -> EX_thmtp_e\n",
      "Updating reaction ID: EX_thr_L_e_ -> EX_thr_L_e\n",
      "Updating reaction ID: EX_thym_e_ -> EX_thym_e\n",
      "Updating reaction ID: EX_thymd_e_ -> EX_thymd_e\n",
      "Updating reaction ID: EX_thyox_L_e_ -> EX_thyox_L_e\n",
      "Updating reaction ID: EX_tmndnc_e_ -> EX_tmndnc_e\n",
      "Updating reaction ID: EX_tolbutamide_e_ -> EX_tolbutamide_e\n",
      "Updating reaction ID: EX_tre_e_ -> EX_tre_e\n",
      "Updating reaction ID: EX_triodthy_e_ -> EX_triodthy_e\n",
      "Updating reaction ID: EX_triodthysuf_e_ -> EX_triodthysuf_e\n",
      "Updating reaction ID: EX_trp_L_e_ -> EX_trp_L_e\n",
      "Updating reaction ID: EX_tststerone_e_ -> EX_tststerone_e\n",
      "Updating reaction ID: EX_tststeroneglc_e_ -> EX_tststeroneglc_e\n",
      "Updating reaction ID: EX_tststerones_e_ -> EX_tststerones_e\n",
      "Updating reaction ID: EX_tsul_e_ -> EX_tsul_e\n",
      "Updating reaction ID: EX_ttdca_e_ -> EX_ttdca_e\n",
      "Updating reaction ID: EX_txa2_e_ -> EX_txa2_e\n",
      "Updating reaction ID: EX_tymsf_e_ -> EX_tymsf_e\n",
      "Updating reaction ID: EX_Tyr_ggn_e_ -> EX_Tyr_ggn_e\n",
      "Updating reaction ID: EX_tyr_L_e_ -> EX_tyr_L_e\n",
      "Updating reaction ID: EX_udp_e_ -> EX_udp_e\n",
      "Updating reaction ID: EX_udpg_e_ -> EX_udpg_e\n",
      "Updating reaction ID: EX_udpgal_e_ -> EX_udpgal_e\n",
      "Updating reaction ID: EX_ump_e_ -> EX_ump_e\n",
      "Updating reaction ID: EX_ura_e_ -> EX_ura_e\n",
      "Updating reaction ID: EX_urate_e_ -> EX_urate_e\n",
      "Updating reaction ID: EX_urea_e_ -> EX_urea_e\n",
      "Updating reaction ID: EX_uri_e_ -> EX_uri_e\n",
      "Updating reaction ID: EX_utp_e_ -> EX_utp_e\n",
      "Updating reaction ID: EX_vacc_e_ -> EX_vacc_e\n",
      "Updating reaction ID: EX_val_L_e_ -> EX_val_L_e\n",
      "Updating reaction ID: EX_vitd2_e_ -> EX_vitd2_e\n",
      "Updating reaction ID: EX_vitd3_e_ -> EX_vitd3_e\n",
      "Updating reaction ID: EX_whddca_e_ -> EX_whddca_e\n",
      "Updating reaction ID: EX_whhdca_e_ -> EX_whhdca_e\n",
      "Updating reaction ID: EX_whtststerone_e_ -> EX_whtststerone_e\n",
      "Updating reaction ID: EX_whttdca_e_ -> EX_whttdca_e\n",
      "Updating reaction ID: EX_xolest_cho_e_ -> EX_xolest_cho_e\n",
      "Updating reaction ID: EX_xolest2_cho_e_ -> EX_xolest2_cho_e\n",
      "Updating reaction ID: EX_xoltri24_e_ -> EX_xoltri24_e\n",
      "Updating reaction ID: EX_xoltri25_e_ -> EX_xoltri25_e\n",
      "Updating reaction ID: EX_xoltri27_e_ -> EX_xoltri27_e\n",
      "Updating reaction ID: EX_xyl_D_e_ -> EX_xyl_D_e\n",
      "Updating reaction ID: EX_xylt_e_ -> EX_xylt_e\n",
      "Updating reaction ID: EX_yvite_e_ -> EX_yvite_e\n",
      "Updating reaction ID: SK_Asn_X_Ser_Thr_r_ -> SK_Asn_X_Ser_Thr_r\n",
      "Updating reaction ID: SK_Ser_Thr_g_ -> SK_Ser_Thr_g\n",
      "Updating reaction ID: SK_Tyr_ggn_c_ -> SK_Tyr_ggn_c\n",
      "Updating reaction ID: DM_epo_g_ -> DM_epo_g\n",
      "Updating reaction ID: DM_igg_g_ -> DM_igg_g\n",
      "Updating reaction ID: DM_T_antigen_g_ -> DM_T_antigen_g\n",
      "Updating reaction ID: DM_avite1_c_ -> DM_avite1_c\n",
      "Updating reaction ID: DM_avite2_c_ -> DM_avite2_c\n",
      "Updating reaction ID: DM_bvite_c_ -> DM_bvite_c\n",
      "Updating reaction ID: DM_core5_g_ -> DM_core5_g\n",
      "Updating reaction ID: DM_core7_g_ -> DM_core7_g\n",
      "Updating reaction ID: DM_core8_g_ -> DM_core8_g\n",
      "Updating reaction ID: DM_dem2emgacpail_prot_cho_r_ -> DM_dem2emgacpail_prot_cho_r\n",
      "Updating reaction ID: DM_dgpi_prot_cho_r_ -> DM_dgpi_prot_cho_r\n",
      "Updating reaction ID: DM_dsT_antigen_g_ -> DM_dsT_antigen_g\n",
      "Updating reaction ID: DM_ethamp_r_ -> DM_ethamp_r\n",
      "Updating reaction ID: DM_gncore2_g_ -> DM_gncore2_g\n",
      "Updating reaction ID: DM_kdn_c_ -> DM_kdn_c\n",
      "Updating reaction ID: DM_n5m2masn_g_ -> DM_n5m2masn_g\n",
      "Updating reaction ID: DM_sTn_antigen_g_ -> DM_sTn_antigen_g\n",
      "Updating reaction ID: DM_sprm_c_ -> DM_sprm_c\n",
      "Updating reaction ID: DM_yvite_c_ -> DM_yvite_c\n",
      "Updating reaction ID: DM_56iqcrbxlt_c_ -> DM_56iqcrbxlt_c\n",
      "Updating reaction ID: DM_atp_c_ -> DM_atp_c\n",
      "Updating reaction ID: SK_pre_prot_r_ -> SK_pre_prot_r\n",
      "Updating reaction ID: EX_HC01610_e_ -> EX_HC01610_e\n",
      "Updating reaction ID: EX_HC02179_e_ -> EX_HC02179_e\n",
      "Updating reaction ID: EX_C05965_e_ -> EX_C05965_e\n",
      "Updating reaction ID: EX_HC02176_e_ -> EX_HC02176_e\n",
      "Updating reaction ID: EX_10fthf_e_ -> EX_10fthf_e\n",
      "Updating reaction ID: EX_10fthf5glu_e_ -> EX_10fthf5glu_e\n",
      "Updating reaction ID: EX_10fthf6glu_e_ -> EX_10fthf6glu_e\n",
      "Updating reaction ID: EX_10fthf7glu_e_ -> EX_10fthf7glu_e\n",
      "Updating reaction ID: EX_11_cis_retfa_e_ -> EX_11_cis_retfa_e\n",
      "Updating reaction ID: EX_13_cis_retnglc_e_ -> EX_13_cis_retnglc_e\n",
      "Updating reaction ID: EX_1glyc_cho_e_ -> EX_1glyc_cho_e\n",
      "Updating reaction ID: EX_1mncam_e_ -> EX_1mncam_e\n",
      "Updating reaction ID: EX_23cump_e_ -> EX_23cump_e\n",
      "Updating reaction ID: EX_2425dhvitd2_e_ -> EX_2425dhvitd2_e\n",
      "Updating reaction ID: EX_2425dhvitd3_e_ -> EX_2425dhvitd3_e\n",
      "Updating reaction ID: EX_24nph_e_ -> EX_24nph_e\n",
      "Updating reaction ID: EX_25hvitd2_e_ -> EX_25hvitd2_e\n",
      "Updating reaction ID: EX_25hvitd3_e_ -> EX_25hvitd3_e\n",
      "Updating reaction ID: EX_2hb_e_ -> EX_2hb_e\n",
      "Updating reaction ID: EX_2mcit_e_ -> EX_2mcit_e\n",
      "Updating reaction ID: EX_34dhoxpeg_e_ -> EX_34dhoxpeg_e\n",
      "Updating reaction ID: EX_34dhphe_e_ -> EX_34dhphe_e\n",
      "Updating reaction ID: EX_35cgmp_e_ -> EX_35cgmp_e\n",
      "Updating reaction ID: EX_3aib_D_e_ -> EX_3aib_D_e\n",
      "Updating reaction ID: EX_3aib_e_ -> EX_3aib_e\n",
      "Updating reaction ID: EX_3mlda_e_ -> EX_3mlda_e\n",
      "Updating reaction ID: EX_3mob_e_ -> EX_3mob_e\n",
      "Updating reaction ID: EX_3mop_e_ -> EX_3mop_e\n",
      "Updating reaction ID: EX_3ump_e_ -> EX_3ump_e\n",
      "Updating reaction ID: EX_4abut_e_ -> EX_4abut_e\n",
      "Updating reaction ID: EX_4abutn_e_ -> EX_4abutn_e\n",
      "Updating reaction ID: EX_4hdebrisoquine_e_ -> EX_4hdebrisoquine_e\n",
      "Updating reaction ID: EX_4hphac_e_ -> EX_4hphac_e\n",
      "Updating reaction ID: EX_4hpro_e_ -> EX_4hpro_e\n",
      "Updating reaction ID: EX_4mop_e_ -> EX_4mop_e\n",
      "Updating reaction ID: EX_4mptnl_e_ -> EX_4mptnl_e\n",
      "Updating reaction ID: EX_4mtolbutamide_e_ -> EX_4mtolbutamide_e\n",
      "Updating reaction ID: EX_4nph_e_ -> EX_4nph_e\n",
      "Updating reaction ID: EX_4nphsf_e_ -> EX_4nphsf_e\n",
      "Updating reaction ID: EX_4pyrdx_e_ -> EX_4pyrdx_e\n",
      "Updating reaction ID: EX_5adtststerone_e_ -> EX_5adtststerone_e\n",
      "Updating reaction ID: EX_5adtststeroneglc_e_ -> EX_5adtststeroneglc_e\n",
      "Updating reaction ID: EX_5adtststerones_e_ -> EX_5adtststerones_e\n",
      "Updating reaction ID: EX_5dhf_e_ -> EX_5dhf_e\n",
      "Updating reaction ID: EX_5fthf_e_ -> EX_5fthf_e\n",
      "Updating reaction ID: EX_5homeprazole_e_ -> EX_5homeprazole_e\n",
      "Updating reaction ID: EX_5htrp_e_ -> EX_5htrp_e\n",
      "Updating reaction ID: EX_5mta_e_ -> EX_5mta_e\n",
      "Updating reaction ID: EX_5mthf_e_ -> EX_5mthf_e\n",
      "Updating reaction ID: EX_5oxpro_e_ -> EX_5oxpro_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_5thf_e_ -> EX_5thf_e\n",
      "Updating reaction ID: EX_6dhf_e_ -> EX_6dhf_e\n",
      "Updating reaction ID: EX_6htststerone_e_ -> EX_6htststerone_e\n",
      "Updating reaction ID: EX_6thf_e_ -> EX_6thf_e\n",
      "Updating reaction ID: EX_7dhf_e_ -> EX_7dhf_e\n",
      "Updating reaction ID: EX_7thf_e_ -> EX_7thf_e\n",
      "Updating reaction ID: EX_9_cis_retfa_e_ -> EX_9_cis_retfa_e\n",
      "Updating reaction ID: EX_abt_e_ -> EX_abt_e\n",
      "Updating reaction ID: EX_ac_e_ -> EX_ac_e\n",
      "Updating reaction ID: EX_acac_e_ -> EX_acac_e\n",
      "Updating reaction ID: EX_acald_e_ -> EX_acald_e\n",
      "Updating reaction ID: EX_acetone_e_ -> EX_acetone_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgam_e_ -> EX_acgam_e\n",
      "Updating reaction ID: EX_ach_e_ -> EX_ach_e\n",
      "Updating reaction ID: EX_acmana_e_ -> EX_acmana_e\n",
      "Updating reaction ID: EX_acn13acngalgbside_cho_e_ -> EX_acn13acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acn23acngalgbside_cho_e_ -> EX_acn23acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acnacngal14acglcgalgluside_cho_e_ -> EX_acnacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acnacngalgbside_cho_e_ -> EX_acnacngalgbside_cho_e\n",
      "Updating reaction ID: EX_acngalacglcgal14acglcgalgluside_cho_e_ -> EX_acngalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_ade_e_ -> EX_ade_e\n",
      "Updating reaction ID: EX_adn_e_ -> EX_adn_e\n",
      "Updating reaction ID: EX_adpcbl_e_ -> EX_adpcbl_e\n",
      "Updating reaction ID: EX_adprbp_e_ -> EX_adprbp_e\n",
      "Updating reaction ID: EX_adprib_e_ -> EX_adprib_e\n",
      "Updating reaction ID: EX_adrn_e_ -> EX_adrn_e\n",
      "Updating reaction ID: EX_adrnl_e_ -> EX_adrnl_e\n",
      "Updating reaction ID: EX_aflatoxin_e_ -> EX_aflatoxin_e\n",
      "Updating reaction ID: EX_ahandrostanglc_e_ -> EX_ahandrostanglc_e\n",
      "Updating reaction ID: EX_ahcys_e_ -> EX_ahcys_e\n",
      "Updating reaction ID: EX_ahdt_e_ -> EX_ahdt_e\n",
      "Updating reaction ID: EX_aicar_e_ -> EX_aicar_e\n",
      "Updating reaction ID: EX_ak2lgchol_cho_e_ -> EX_ak2lgchol_cho_e\n",
      "Updating reaction ID: EX_akg_e_ -> EX_akg_e\n",
      "Updating reaction ID: EX_ala_B_e_ -> EX_ala_B_e\n",
      "Updating reaction ID: EX_ala_D_e_ -> EX_ala_D_e\n",
      "Updating reaction ID: EX_ala_L_e_ -> EX_ala_L_e\n",
      "Updating reaction ID: EX_alaala_e_ -> EX_alaala_e\n",
      "Updating reaction ID: EX_aldstrn_e_ -> EX_aldstrn_e\n",
      "Updating reaction ID: EX_amp_e_ -> EX_amp_e\n",
      "Updating reaction ID: EX_andrstrn_e_ -> EX_andrstrn_e\n",
      "Updating reaction ID: EX_andrstrnglc_e_ -> EX_andrstrnglc_e\n",
      "Updating reaction ID: EX_anth_e_ -> EX_anth_e\n",
      "Updating reaction ID: EX_antipyrene_e_ -> EX_antipyrene_e\n",
      "Updating reaction ID: EX_apnnox_e_ -> EX_apnnox_e\n",
      "Updating reaction ID: EX_appnn_e_ -> EX_appnn_e\n",
      "Updating reaction ID: EX_aprgstrn_e_ -> EX_aprgstrn_e\n",
      "Updating reaction ID: EX_aqcobal_e_ -> EX_aqcobal_e\n",
      "Updating reaction ID: EX_arab_L_e_ -> EX_arab_L_e\n",
      "Updating reaction ID: EX_arach_e_ -> EX_arach_e\n",
      "Updating reaction ID: EX_arachcoa_e_ -> EX_arachcoa_e\n",
      "Updating reaction ID: EX_arachd_e_ -> EX_arachd_e\n",
      "Updating reaction ID: EX_arg_L_e_ -> EX_arg_L_e\n",
      "Updating reaction ID: EX_ascb_L_e_ -> EX_ascb_L_e\n",
      "Updating reaction ID: EX_asn_L_e_ -> EX_asn_L_e\n",
      "Updating reaction ID: EX_asp_D_e_ -> EX_asp_D_e\n",
      "Updating reaction ID: EX_asp_L_e_ -> EX_asp_L_e\n",
      "Updating reaction ID: EX_atp_e_ -> EX_atp_e\n",
      "Updating reaction ID: EX_avite1_e_ -> EX_avite1_e\n",
      "Updating reaction ID: EX_avite2_e_ -> EX_avite2_e\n",
      "Updating reaction ID: EX_bglc_e_ -> EX_bglc_e\n",
      "Updating reaction ID: EX_bhb_e_ -> EX_bhb_e\n",
      "Updating reaction ID: EX_bildglcur_e_ -> EX_bildglcur_e\n",
      "Updating reaction ID: EX_bilglcur_e_ -> EX_bilglcur_e\n",
      "Updating reaction ID: EX_bilirub_e_ -> EX_bilirub_e\n",
      "Updating reaction ID: EX_biocyt_e_ -> EX_biocyt_e\n",
      "Updating reaction ID: EX_btn_e_ -> EX_btn_e\n",
      "Updating reaction ID: EX_but_e_ -> EX_but_e\n",
      "Updating reaction ID: EX_bvite_e_ -> EX_bvite_e\n",
      "Updating reaction ID: EX_bz_e_ -> EX_bz_e\n",
      "Updating reaction ID: EX_xtra_e_ -> EX_xtra_e\n",
      "Updating reaction ID: EX_C02528_e_ -> EX_C02528_e\n",
      "Updating reaction ID: EX_12e8hdxt_e_ -> EX_12e8hdxt_e\n",
      "Updating reaction ID: EX_ca2_e_ -> EX_ca2_e\n",
      "Updating reaction ID: EX_camp_e_ -> EX_camp_e\n",
      "Updating reaction ID: EX_carn_e_ -> EX_carn_e\n",
      "Updating reaction ID: EX_caro_e_ -> EX_caro_e\n",
      "Updating reaction ID: EX_carveol_e_ -> EX_carveol_e\n",
      "Updating reaction ID: EX_cbasp_e_ -> EX_cbasp_e\n",
      "Updating reaction ID: EX_cca_d3_e_ -> EX_cca_d3_e\n",
      "Updating reaction ID: EX_cdp_e_ -> EX_cdp_e\n",
      "Updating reaction ID: EX_CE0074_e_ -> EX_CE0074_e\n",
      "Updating reaction ID: EX_CE1925_e_ -> EX_CE1925_e\n",
      "Updating reaction ID: EX_CE1926_e_ -> EX_CE1926_e\n",
      "Updating reaction ID: EX_CE1935_e_ -> EX_CE1935_e\n",
      "Updating reaction ID: EX_CE1936_e_ -> EX_CE1936_e\n",
      "Updating reaction ID: EX_CE1939_e_ -> EX_CE1939_e\n",
      "Updating reaction ID: EX_CE1940_e_ -> EX_CE1940_e\n",
      "Updating reaction ID: EX_CE1943_e_ -> EX_CE1943_e\n",
      "Updating reaction ID: EX_CE2011_e_ -> EX_CE2011_e\n",
      "Updating reaction ID: EX_CE2250_e_ -> EX_CE2250_e\n",
      "Updating reaction ID: EX_CE2838_e_ -> EX_CE2838_e\n",
      "Updating reaction ID: EX_CE2839_e_ -> EX_CE2839_e\n",
      "Updating reaction ID: EX_CE2915_e_ -> EX_CE2915_e\n",
      "Updating reaction ID: EX_CE2916_e_ -> EX_CE2916_e\n",
      "Updating reaction ID: EX_CE2917_e_ -> EX_CE2917_e\n",
      "Updating reaction ID: EX_CE4633_e_ -> EX_CE4633_e\n",
      "Updating reaction ID: EX_CE4722_e_ -> EX_CE4722_e\n",
      "Updating reaction ID: EX_CE4723_e_ -> EX_CE4723_e\n",
      "Updating reaction ID: EX_CE4724_e_ -> EX_CE4724_e\n",
      "Updating reaction ID: EX_CE4881_e_ -> EX_CE4881_e\n",
      "Updating reaction ID: EX_CE5786_e_ -> EX_CE5786_e\n",
      "Updating reaction ID: EX_CE5788_e_ -> EX_CE5788_e\n",
      "Updating reaction ID: EX_CE5789_e_ -> EX_CE5789_e\n",
      "Updating reaction ID: EX_CE5797_e_ -> EX_CE5797_e\n",
      "Updating reaction ID: EX_CE5798_e_ -> EX_CE5798_e\n",
      "Updating reaction ID: EX_CE5853_e_ -> EX_CE5853_e\n",
      "Updating reaction ID: EX_CE5854_e_ -> EX_CE5854_e\n",
      "Updating reaction ID: EX_CE5867_e_ -> EX_CE5867_e\n",
      "Updating reaction ID: EX_CE5868_e_ -> EX_CE5868_e\n",
      "Updating reaction ID: EX_CE5869_e_ -> EX_CE5869_e\n",
      "Updating reaction ID: EX_cgly_e_ -> EX_cgly_e\n",
      "Updating reaction ID: EX_chol_e_ -> EX_chol_e\n",
      "Updating reaction ID: EX_cholate_e_ -> EX_cholate_e\n",
      "Updating reaction ID: EX_chsterol_e_ -> EX_chsterol_e\n",
      "Updating reaction ID: EX_chtn_e_ -> EX_chtn_e\n",
      "Updating reaction ID: EX_cit_e_ -> EX_cit_e\n",
      "Updating reaction ID: EX_citr_L_e_ -> EX_citr_L_e\n",
      "Updating reaction ID: EX_cl_e_ -> EX_cl_e\n",
      "Updating reaction ID: EX_CLPND_e_ -> EX_CLPND_e\n",
      "Updating reaction ID: EX_cmp_e_ -> EX_cmp_e\n",
      "Updating reaction ID: EX_co_e_ -> EX_co_e\n",
      "Updating reaction ID: EX_co2_e_ -> EX_co2_e\n",
      "Updating reaction ID: EX_coa_e_ -> EX_coa_e\n",
      "Updating reaction ID: EX_coumarin_e_ -> EX_coumarin_e\n",
      "Updating reaction ID: EX_creat_e_ -> EX_creat_e\n",
      "Updating reaction ID: EX_crm_cho_e_ -> EX_crm_cho_e\n",
      "Updating reaction ID: EX_crmp_cho_e_ -> EX_crmp_cho_e\n",
      "Updating reaction ID: EX_crn_e_ -> EX_crn_e\n",
      "Updating reaction ID: EX_crtsl_e_ -> EX_crtsl_e\n",
      "Updating reaction ID: EX_crtstrn_e_ -> EX_crtstrn_e\n",
      "Updating reaction ID: EX_crvnc_e_ -> EX_crvnc_e\n",
      "Updating reaction ID: EX_csn_e_ -> EX_csn_e\n",
      "Updating reaction ID: EX_cspg_a_e_ -> EX_cspg_a_e\n",
      "Updating reaction ID: EX_cspg_b_e_ -> EX_cspg_b_e\n",
      "Updating reaction ID: EX_cspg_c_e_ -> EX_cspg_c_e\n",
      "Updating reaction ID: EX_cspg_d_e_ -> EX_cspg_d_e\n",
      "Updating reaction ID: EX_cspg_e_e_ -> EX_cspg_ee_\n",
      "Updating reaction ID: EX_ctp_e_ -> EX_ctp_e\n",
      "Updating reaction ID: EX_cyan_e_ -> EX_cyan_e\n",
      "Updating reaction ID: EX_cys_L_e_ -> EX_cys_L_e\n",
      "Updating reaction ID: EX_cysam_e_ -> EX_cysam_e\n",
      "Updating reaction ID: EX_cytd_e_ -> EX_cytd_e\n",
      "Updating reaction ID: EX_dad_2_e_ -> EX_dad_2_e\n",
      "Updating reaction ID: EX_dag_cho_e_ -> EX_dag_cho_e\n",
      "Updating reaction ID: EX_dca_e_ -> EX_dca_e\n",
      "Updating reaction ID: EX_dchac_e_ -> EX_dchac_e\n",
      "Updating reaction ID: EX_dcsptn1_e_ -> EX_dcsptn1_e\n",
      "Updating reaction ID: EX_dcyt_e_ -> EX_dcyt_e\n",
      "Updating reaction ID: EX_debrisoquine_e_ -> EX_debrisoquine_e\n",
      "Updating reaction ID: EX_dgchol_e_ -> EX_dgchol_e\n",
      "Updating reaction ID: EX_dgmp_e_ -> EX_dgmp_e\n",
      "Updating reaction ID: EX_dgsn_e_ -> EX_dgsn_e\n",
      "Updating reaction ID: EX_dgtp_e_ -> EX_dgtp_e\n",
      "Updating reaction ID: EX_dha_e_ -> EX_dha_e\n",
      "Updating reaction ID: EX_dhap_e_ -> EX_dhap_e\n",
      "Updating reaction ID: EX_dhdascb_e_ -> EX_dhdascb_e\n",
      "Updating reaction ID: EX_dheas_e_ -> EX_dheas_e\n",
      "Updating reaction ID: EX_dhf_e_ -> EX_dhf_e\n",
      "Updating reaction ID: EX_digalsgalside_cho_e_ -> EX_digalsgalside_cho_e\n",
      "Updating reaction ID: EX_din_e_ -> EX_din_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_dlnlcg_e_ -> EX_dlnlcg_e\n",
      "Updating reaction ID: EX_dmantipyrine_e_ -> EX_dmantipyrine_e\n",
      "Updating reaction ID: EX_dmhptcrn_e_ -> EX_dmhptcrn_e\n",
      "Updating reaction ID: EX_dopa_e_ -> EX_dopa_e\n",
      "Updating reaction ID: EX_dopasf_e_ -> EX_dopasf_e\n",
      "Updating reaction ID: EX_dpcoa_e_ -> EX_dpcoa_e\n",
      "Updating reaction ID: EX_drib_e_ -> EX_drib_e\n",
      "Updating reaction ID: EX_dtdp_e_ -> EX_dtdp_e\n",
      "Updating reaction ID: EX_dtmp_e_ -> EX_dtmp_e\n",
      "Updating reaction ID: EX_dttp_e_ -> EX_dttp_e\n",
      "Updating reaction ID: EX_duri_e_ -> EX_duri_e\n",
      "Updating reaction ID: EX_eaflatoxin_e_ -> EX_eaflatoxin_e\n",
      "Updating reaction ID: EX_ebastine_e_ -> EX_ebastine_e\n",
      "Updating reaction ID: EX_ebastineoh_e_ -> EX_ebastineoh_e\n",
      "Updating reaction ID: EX_eicostet_e_ -> EX_eicostet_e\n",
      "Updating reaction ID: EX_elaid_e_ -> EX_elaid_e\n",
      "Updating reaction ID: EX_estradiol_e_ -> EX_estradiol_e\n",
      "Updating reaction ID: EX_estradiolglc_e_ -> EX_estradiolglc_e\n",
      "Updating reaction ID: EX_estriolglc_e_ -> EX_estriolglc_e\n",
      "Updating reaction ID: EX_estroneglc_e_ -> EX_estroneglc_e\n",
      "Updating reaction ID: EX_estrones_e_ -> EX_estrones_e\n",
      "Updating reaction ID: EX_etoh_e_ -> EX_etoh_e\n",
      "Updating reaction ID: EX_fad_e_ -> EX_fad_e\n",
      "Updating reaction ID: EX_fald_e_ -> EX_fald_e\n",
      "Updating reaction ID: EX_fe2_e_ -> EX_fe2_e\n",
      "Updating reaction ID: EX_fe3_e_ -> EX_fe3_e\n",
      "Updating reaction ID: EX_fmn_e_ -> EX_fmn_e\n",
      "Updating reaction ID: EX_fol_e_ -> EX_fol_e\n",
      "Updating reaction ID: EX_for_e_ -> EX_for_e\n",
      "Updating reaction ID: EX_fru_e_ -> EX_fru_e\n",
      "Updating reaction ID: EX_fuc_L_e_ -> EX_fuc_L_e\n",
      "Updating reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fuc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fuc14galacglcgalgluside_cho_e_ -> EX_fuc14galacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacgalfucgalacglcgalgluside_cho_e_ -> EX_fucacgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngal14acglcgalgluside_cho_e_ -> EX_fucacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngalacglcgalgluside_cho_e_ -> EX_fucacngalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc12gal14acglcgalgluside_cho_e_ -> EX_fucfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfuc132galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucgalacglcgalgluside_cho_e_ -> EX_fucfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgal14acglcgalgluside_cho_e_ -> EX_fucgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalfucgalacglcgalgluside_cho_e_ -> EX_fucgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalgbside_cho_e_ -> EX_fucgalgbside_cho_e\n",
      "Updating reaction ID: EX_g1p_e_ -> EX_g1p_e\n",
      "Updating reaction ID: EX_gal_e_ -> EX_gal_e\n",
      "Updating reaction ID: EX_galacglcgalgbside_cho_e_ -> EX_galacglcgalgbside_cho_e\n",
      "Updating reaction ID: EX_galfuc12gal14acglcgalgluside_cho_e_ -> EX_galfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_galfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e_ -> EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalgalthcrm_cho_e_ -> EX_galgalgalthcrm_cho_e\n",
      "Updating reaction ID: EX_galside_cho_e_ -> EX_galside_cho_e\n",
      "Updating reaction ID: EX_gam_e_ -> EX_gam_e\n",
      "Updating reaction ID: EX_gbside_cho_e_ -> EX_gbside_cho_e\n",
      "Updating reaction ID: EX_gchola_e_ -> EX_gchola_e\n",
      "Updating reaction ID: EX_gd1b2_cho_e_ -> EX_gd1b2_cho_e\n",
      "Updating reaction ID: EX_gd1c_cho_e_ -> EX_gd1c_cho_e\n",
      "Updating reaction ID: EX_gdp_e_ -> EX_gdp_e\n",
      "Updating reaction ID: EX_glc_e_ -> EX_glc_e\n",
      "Updating reaction ID: EX_glcur_e_ -> EX_glcur_e\n",
      "Updating reaction ID: EX_glgchlo_e_ -> EX_glgchlo_e\n",
      "Updating reaction ID: EX_gln_L_e_ -> EX_gln_L_e\n",
      "Updating reaction ID: EX_gltcho_e_ -> EX_gltcho_e\n",
      "Updating reaction ID: EX_gltdechol_e_ -> EX_gltdechol_e\n",
      "Updating reaction ID: EX_glu_L_e_ -> EX_glu_L_e\n",
      "Updating reaction ID: EX_gluala_e_ -> EX_gluala_e\n",
      "Updating reaction ID: EX_gly_e_ -> EX_gly_e\n",
      "Updating reaction ID: EX_glyb_e_ -> EX_glyb_e\n",
      "Updating reaction ID: EX_glyc_e_ -> EX_glyc_e\n",
      "Updating reaction ID: EX_glyc_S_e_ -> EX_glyc_S_e\n",
      "Updating reaction ID: EX_glygly_e_ -> EX_glygly_e\n",
      "Updating reaction ID: EX_glygn2_e_ -> EX_glygn2_e\n",
      "Updating reaction ID: EX_glygn4_e_ -> EX_glygn4_e\n",
      "Updating reaction ID: EX_glygn5_e_ -> EX_glygn5_e\n",
      "Updating reaction ID: EX_glyleu_e_ -> EX_glyleu_e\n",
      "Updating reaction ID: EX_glyphe_e_ -> EX_glyphe_e\n",
      "Updating reaction ID: EX_glypro_e_ -> EX_glypro_e\n",
      "Updating reaction ID: EX_glysar_e_ -> EX_glysar_e\n",
      "Updating reaction ID: EX_gmp_e_ -> EX_gmp_e\n",
      "Updating reaction ID: EX_gp1c_cho_e_ -> EX_gp1c_cho_e\n",
      "Updating reaction ID: EX_gp1calpha_cho_e_ -> EX_gp1calpha_cho_e\n",
      "Updating reaction ID: EX_gq1b_cho_e_ -> EX_gq1b_cho_e\n",
      "Updating reaction ID: EX_gq1balpha_cho_e_ -> EX_gq1balpha_cho_e\n",
      "Updating reaction ID: EX_gsn_e_ -> EX_gsn_e\n",
      "Updating reaction ID: EX_gt1a_cho_e_ -> EX_gt1a_cho_e\n",
      "Updating reaction ID: EX_gthox_e_ -> EX_gthox_e\n",
      "Updating reaction ID: EX_gthrd_e_ -> EX_gthrd_e\n",
      "Updating reaction ID: EX_gtp_e_ -> EX_gtp_e\n",
      "Updating reaction ID: EX_gua_e_ -> EX_gua_e\n",
      "Updating reaction ID: EX_gum_e_ -> EX_gum_e\n",
      "Updating reaction ID: EX_gumdchac_e_ -> EX_gumdchac_e\n",
      "Updating reaction ID: EX_gumgchol_e_ -> EX_gumgchol_e\n",
      "Updating reaction ID: EX_gumtchol_e_ -> EX_gumtchol_e\n",
      "Updating reaction ID: EX_h_e_ -> EX_h_e\n",
      "Updating reaction ID: EX_h2o_e_ -> EX_h2o_e\n",
      "Updating reaction ID: EX_h2o2_e_ -> EX_h2o2_e\n",
      "Updating reaction ID: EX_ha_e_ -> EX_ha_e\n",
      "Updating reaction ID: EX_ha_pre1_e_ -> EX_ha_pre1_e\n",
      "Updating reaction ID: EX_HC00004_e_ -> EX_HC00004_e\n",
      "Updating reaction ID: EX_HC00229_e_ -> EX_HC00229_e\n",
      "Updating reaction ID: EX_HC00250_e_ -> EX_HC00250_e\n",
      "Updating reaction ID: EX_HC00822_e_ -> EX_HC00822_e\n",
      "Updating reaction ID: EX_HC00955_e_ -> EX_HC00955_e\n",
      "Updating reaction ID: EX_HC01104_e_ -> EX_HC01104_e\n",
      "Updating reaction ID: EX_HC01361_e_ -> EX_HC01361_e\n",
      "Updating reaction ID: EX_HC01440_e_ -> EX_HC01440_e\n",
      "Updating reaction ID: EX_HC01441_e_ -> EX_HC01441_e\n",
      "Updating reaction ID: EX_HC01444_e_ -> EX_HC01444_e\n",
      "Updating reaction ID: EX_HC01446_e_ -> EX_HC01446_e\n",
      "Updating reaction ID: EX_HC01577_e_ -> EX_HC01577_e\n",
      "Updating reaction ID: EX_HC01609_e_ -> EX_HC01609_e\n",
      "Updating reaction ID: EX_HC01700_e_ -> EX_HC01700_e\n",
      "Updating reaction ID: EX_HC01787_e_ -> EX_HC01787_e\n",
      "Updating reaction ID: EX_HC02154_e_ -> EX_HC02154_e\n",
      "Updating reaction ID: EX_HC02160_e_ -> EX_HC02160_e\n",
      "Updating reaction ID: EX_HC02161_e_ -> EX_HC02161_e\n",
      "Updating reaction ID: EX_HC02172_e_ -> EX_HC02172_e\n",
      "Updating reaction ID: EX_HC02180_e_ -> EX_HC02180_e\n",
      "Updating reaction ID: EX_HC02187_e_ -> EX_HC02187_e\n",
      "Updating reaction ID: EX_HC02191_e_ -> EX_HC02191_e\n",
      "Updating reaction ID: EX_HC02192_e_ -> EX_HC02192_e\n",
      "Updating reaction ID: EX_HC02193_e_ -> EX_HC02193_e\n",
      "Updating reaction ID: EX_HC02194_e_ -> EX_HC02194_e\n",
      "Updating reaction ID: EX_HC02195_e_ -> EX_HC02195_e\n",
      "Updating reaction ID: EX_HC02196_e_ -> EX_HC02196_e\n",
      "Updating reaction ID: EX_HC02197_e_ -> EX_HC02197_e\n",
      "Updating reaction ID: EX_HC02198_e_ -> EX_HC02198_e\n",
      "Updating reaction ID: EX_HC02199_e_ -> EX_HC02199_e\n",
      "Updating reaction ID: EX_HC02200_e_ -> EX_HC02200_e\n",
      "Updating reaction ID: EX_HC02201_e_ -> EX_HC02201_e\n",
      "Updating reaction ID: EX_HC02202_e_ -> EX_HC02202_e\n",
      "Updating reaction ID: EX_HC02203_e_ -> EX_HC02203_e\n",
      "Updating reaction ID: EX_HC02204_e_ -> EX_HC02204_e\n",
      "Updating reaction ID: EX_HC02205_e_ -> EX_HC02205_e\n",
      "Updating reaction ID: EX_HC02206_e_ -> EX_HC02206_e\n",
      "Updating reaction ID: EX_HC02207_e_ -> EX_HC02207_e\n",
      "Updating reaction ID: EX_HC02208_e_ -> EX_HC02208_e\n",
      "Updating reaction ID: EX_HC02210_e_ -> EX_HC02210_e\n",
      "Updating reaction ID: EX_HC02213_e_ -> EX_HC02213_e\n",
      "Updating reaction ID: EX_HC02214_e_ -> EX_HC02214_e\n",
      "Updating reaction ID: EX_HC02216_e_ -> EX_HC02216_e\n",
      "Updating reaction ID: EX_HC02217_e_ -> EX_HC02217_e\n",
      "Updating reaction ID: EX_HC02220_e_ -> EX_HC02220_e\n",
      "Updating reaction ID: EX_hco3_e_ -> EX_hco3_e\n",
      "Updating reaction ID: EX_hcoumarin_e_ -> EX_hcoumarin_e\n",
      "Updating reaction ID: EX_hdca_e_ -> EX_hdca_e\n",
      "Updating reaction ID: EX_hdcea_e_ -> EX_hdcea_e\n",
      "Updating reaction ID: EX_hestratriol_e_ -> EX_hestratriol_e\n",
      "Updating reaction ID: EX_hexc_e_ -> EX_hexc_e\n",
      "Updating reaction ID: EX_his_L_e_ -> EX_his_L_e\n",
      "Updating reaction ID: EX_hista_e_ -> EX_hista_e\n",
      "Updating reaction ID: EX_hom_L_e_ -> EX_hom_L_e\n",
      "Updating reaction ID: EX_hpdca_e_ -> EX_hpdca_e\n",
      "Updating reaction ID: EX_chopg_e_ -> EX_chopg_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_htaxol_e_ -> EX_htaxol_e\n",
      "Updating reaction ID: EX_hxan_e_ -> EX_hxan_e\n",
      "Updating reaction ID: EX_hyptaur_e_ -> EX_hyptaur_e\n",
      "Updating reaction ID: EX_i_e_ -> EX_i_e\n",
      "Updating reaction ID: EX_idour_e_ -> EX_idour_e\n",
      "Updating reaction ID: EX_idp_e_ -> EX_idp_e\n",
      "Updating reaction ID: EX_ile_L_e_ -> EX_ile_L_e\n",
      "Updating reaction ID: EX_imp_e_ -> EX_imp_e\n",
      "Updating reaction ID: EX_inost_e_ -> EX_inost_e\n",
      "Updating reaction ID: EX_ins_e_ -> EX_ins_e\n",
      "Updating reaction ID: EX_k_e_ -> EX_k_e\n",
      "Updating reaction ID: EX_ksi_deg1_e_ -> EX_ksi_deg1_e\n",
      "Updating reaction ID: EX_ksi_e_ -> EX_ksi_e\n",
      "Updating reaction ID: EX_ksii_core2_e_ -> EX_ksii_core2_e\n",
      "Updating reaction ID: EX_ksii_core4_e_ -> EX_ksii_core4_e\n",
      "Updating reaction ID: EX_lac_D_e_ -> EX_lac_D_e\n",
      "Updating reaction ID: EX_lac_L_e_ -> EX_lac_L_e\n",
      "Updating reaction ID: EX_lcts_e_ -> EX_lcts_e\n",
      "Updating reaction ID: EX_Lcystin_e_ -> EX_Lcystin_e\n",
      "Updating reaction ID: EX_leu_L_e_ -> EX_leu_L_e\n",
      "Updating reaction ID: EX_leugly_e_ -> EX_leugly_e\n",
      "Updating reaction ID: EX_leuktrA4_e_ -> EX_leuktrA4_e\n",
      "Updating reaction ID: EX_leuktrB4_e_ -> EX_leuktrB4_e\n",
      "Updating reaction ID: EX_leuktrC4_e_ -> EX_leuktrC4_e\n",
      "Updating reaction ID: EX_leuktrD4_e_ -> EX_leuktrD4_e\n",
      "Updating reaction ID: EX_leuktrE4_e_ -> EX_leuktrE4_e\n",
      "Updating reaction ID: EX_leuktrF4_e_ -> EX_leuktrF4_e\n",
      "Updating reaction ID: EX_leuleu_e_ -> EX_leuleu_e\n",
      "Updating reaction ID: EX_lgnc_e_ -> EX_lgnc_e\n",
      "Updating reaction ID: EX_limnen_e_ -> EX_limnen_e\n",
      "Updating reaction ID: EX_lipoate_e_ -> EX_lipoate_e\n",
      "Updating reaction ID: EX_lneldc_e_ -> EX_lneldc_e\n",
      "Updating reaction ID: EX_lnlc_e_ -> EX_lnlc_e\n",
      "Updating reaction ID: EX_lnlnca_e_ -> EX_lnlnca_e\n",
      "Updating reaction ID: EX_lnlncg_e_ -> EX_lnlncg_e\n",
      "Updating reaction ID: EX_lpchol_cho_e_ -> EX_lpchol_cho_e\n",
      "Updating reaction ID: EX_lys_L_e_ -> EX_lys_L_e\n",
      "Updating reaction ID: EX_mag_cho_e_ -> EX_mag_cho_e\n",
      "Updating reaction ID: EX_mal_L_e_ -> EX_mal_L_e\n",
      "Updating reaction ID: EX_malcoa_e_ -> EX_malcoa_e\n",
      "Updating reaction ID: EX_malt_e_ -> EX_malt_e\n",
      "Updating reaction ID: EX_malthp_e_ -> EX_malthp_e\n",
      "Updating reaction ID: EX_malthx_e_ -> EX_malthx_e\n",
      "Updating reaction ID: EX_maltpt_e_ -> EX_maltpt_e\n",
      "Updating reaction ID: EX_malttr_e_ -> EX_malttr_e\n",
      "Updating reaction ID: EX_maltttr_e_ -> EX_maltttr_e\n",
      "Updating reaction ID: EX_man_e_ -> EX_man_e\n",
      "Updating reaction ID: EX_meoh_e_ -> EX_meoh_e\n",
      "Updating reaction ID: EX_mepi_e_ -> EX_mepi_e\n",
      "Updating reaction ID: EX_mercplaccys_e_ -> EX_mercplaccys_e\n",
      "Updating reaction ID: EX_met_L_e_ -> EX_met_L_e\n",
      "Updating reaction ID: EX_mthgxl_e_ -> EX_mthgxl_e\n",
      "Updating reaction ID: EX_n2m2nmasn_e_ -> EX_n2m2nmasn_e\n",
      "Updating reaction ID: EX_na1_e_ -> EX_na1_e\n",
      "Updating reaction ID: EX_nac_e_ -> EX_nac_e\n",
      "Updating reaction ID: EX_nad_e_ -> EX_nad_e\n",
      "Updating reaction ID: EX_nadp_e_ -> EX_nadp_e\n",
      "Updating reaction ID: EX_ncam_e_ -> EX_ncam_e\n",
      "Updating reaction ID: EX_nh4_e_ -> EX_nh4_e\n",
      "Updating reaction ID: EX_nifedipine_e_ -> EX_nifedipine_e\n",
      "Updating reaction ID: EX_no_e_ -> EX_no_e\n",
      "Updating reaction ID: EX_no2_e_ -> EX_no2_e\n",
      "Updating reaction ID: EX_npthl_e_ -> EX_npthl_e\n",
      "Updating reaction ID: EX_nrpphr_e_ -> EX_nrpphr_e\n",
      "Updating reaction ID: EX_nrpphrsf_e_ -> EX_nrpphrsf_e\n",
      "Updating reaction ID: EX_nrvnc_e_ -> EX_nrvnc_e\n",
      "Updating reaction ID: EX_o2_e_ -> EX_o2_e\n",
      "Updating reaction ID: EX_o2s_e_ -> EX_o2s_e\n",
      "Updating reaction ID: EX_oagd3_cho_e_ -> EX_oagd3_cho_e\n",
      "Updating reaction ID: EX_oagt3_cho_e_ -> EX_oagt3_cho_e\n",
      "Updating reaction ID: EX_ocdca_e_ -> EX_ocdca_e\n",
      "Updating reaction ID: EX_ocdcea_e_ -> EX_ocdcea_e\n",
      "Updating reaction ID: EX_octa_e_ -> EX_octa_e\n",
      "Updating reaction ID: EX_oh1_e_ -> EX_oh1_e\n",
      "Updating reaction ID: EX_omeprazole_e_ -> EX_omeprazole_e\n",
      "Updating reaction ID: EX_onpthl_e_ -> EX_onpthl_e\n",
      "Updating reaction ID: EX_orn_e_ -> EX_orn_e\n",
      "Updating reaction ID: EX_orot_e_ -> EX_orot_e\n",
      "Updating reaction ID: EX_oxa_e_ -> EX_oxa_e\n",
      "Updating reaction ID: EX_paf_cho_e_ -> EX_paf_cho_e\n",
      "Updating reaction ID: EX_pan4p_e_ -> EX_pan4p_e\n",
      "Updating reaction ID: EX_pchol_cho_e_ -> EX_pchol_cho_e\n",
      "Updating reaction ID: EX_pe_cho_e_ -> EX_pe_cho_e\n",
      "Updating reaction ID: EX_pect_e_ -> EX_pect_e\n",
      "Updating reaction ID: EX_pectindchac_e_ -> EX_pectindchac_e\n",
      "Updating reaction ID: EX_pectingchol_e_ -> EX_pectingchol_e\n",
      "Updating reaction ID: EX_pectintchol_e_ -> EX_pectintchol_e\n",
      "Updating reaction ID: EX_peplys_e_ -> EX_peplys_e\n",
      "Updating reaction ID: EX_perillyl_e_ -> EX_perillyl_e\n",
      "Updating reaction ID: EX_pglyc_cho_e_ -> EX_pglyc_cho_e\n",
      "Updating reaction ID: EX_phe_L_e_ -> EX_phe_L_e\n",
      "Updating reaction ID: EX_pheacgln_e_ -> EX_pheacgln_e\n",
      "Updating reaction ID: EX_pheme_e_ -> EX_pheme_e\n",
      "Updating reaction ID: EX_phyQ_e_ -> EX_phyQ_e\n",
      "Updating reaction ID: EX_phyt_e_ -> EX_phyt_e\n",
      "Updating reaction ID: EX_pi_e_ -> EX_pi_e\n",
      "Updating reaction ID: EX_pnto_R_e_ -> EX_pnto_R_e\n",
      "Updating reaction ID: EX_ppa_e_ -> EX_ppa_e\n",
      "Updating reaction ID: EX_ppi_e_ -> EX_ppi_e\n",
      "Updating reaction ID: EX_prgstrn_e_ -> EX_prgstrn_e\n",
      "Updating reaction ID: EX_pro_D_e_ -> EX_pro_D_e\n",
      "Updating reaction ID: EX_pro_L_e_ -> EX_pro_L_e\n",
      "Updating reaction ID: EX_progly_e_ -> EX_progly_e\n",
      "Updating reaction ID: EX_prostgd2_e_ -> EX_prostgd2_e\n",
      "Updating reaction ID: EX_prostge1_e_ -> EX_prostge1_e\n",
      "Updating reaction ID: EX_prostge2_e_ -> EX_prostge2_e\n",
      "Updating reaction ID: EX_prostgf2_e_ -> EX_prostgf2_e\n",
      "Updating reaction ID: EX_prostgh2_e_ -> EX_prostgh2_e\n",
      "Updating reaction ID: EX_prostgi2_e_ -> EX_prostgi2_e\n",
      "Updating reaction ID: EX_prpp_e_ -> EX_prpp_e\n",
      "Updating reaction ID: EX_ps_cho_e_ -> EX_ps_cho_e\n",
      "Updating reaction ID: EX_psyl_e_ -> EX_psyl_e\n",
      "Updating reaction ID: EX_psylchol_e_ -> EX_psylchol_e\n",
      "Updating reaction ID: EX_psyltchol_e_ -> EX_psyltchol_e\n",
      "Updating reaction ID: EX_psyltdechol_e_ -> EX_psyltdechol_e\n",
      "Updating reaction ID: EX_ptdca_e_ -> EX_ptdca_e\n",
      "Updating reaction ID: EX_ptrc_e_ -> EX_ptrc_e\n",
      "Updating reaction ID: EX_ptth_e_ -> EX_ptth_e\n",
      "Updating reaction ID: EX_pydam_e_ -> EX_pydam_e\n",
      "Updating reaction ID: EX_pydx_e_ -> EX_pydx_e\n",
      "Updating reaction ID: EX_pydx5p_e_ -> EX_pydx5p_e\n",
      "Updating reaction ID: EX_pydxn_e_ -> EX_pydxn_e\n",
      "Updating reaction ID: EX_pyr_e_ -> EX_pyr_e\n",
      "Updating reaction ID: EX_q10_e_ -> EX_q10_e\n",
      "Updating reaction ID: EX_q10h2_e_ -> EX_q10h2_e\n",
      "Updating reaction ID: EX_rbt_e_ -> EX_rbt_e\n",
      "Updating reaction ID: EX_retfa_e_ -> EX_retfa_e\n",
      "Updating reaction ID: EX_retinol_9_cis_e_ -> EX_retinol_9_cis_e\n",
      "Updating reaction ID: EX_retinol_cis_11_e_ -> EX_retinol_cis_11_e\n",
      "Updating reaction ID: EX_retinol_e_ -> EX_retinol_e\n",
      "Updating reaction ID: EX_retn_e_ -> EX_retn_e\n",
      "Updating reaction ID: EX_retnglc_e_ -> EX_retnglc_e\n",
      "Updating reaction ID: EX_rib_D_e_ -> EX_rib_D_e\n",
      "Updating reaction ID: EX_ribflv_e_ -> EX_ribflv_e\n",
      "Updating reaction ID: EX_Rtotal_e_ -> EX_Rtotal_e\n",
      "Updating reaction ID: EX_Rtotal2_e_ -> EX_Rtotal2_e\n",
      "Updating reaction ID: EX_Rtotal3_e_ -> EX_Rtotal3_e\n",
      "Updating reaction ID: EX_s2l2fn2m2masn_e_ -> EX_s2l2fn2m2masn_e\n",
      "Updating reaction ID: EX_s2l2n2m2masn_e_ -> EX_s2l2n2m2masn_e\n",
      "Updating reaction ID: EX_sarcs_e_ -> EX_sarcs_e\n",
      "Updating reaction ID: EX_sbt_d_e_ -> EX_sbt_d_e\n",
      "Updating reaction ID: EX_sel_e_ -> EX_sel_e\n",
      "Updating reaction ID: EX_ser_D_e_ -> EX_ser_D_e\n",
      "Updating reaction ID: EX_ser_L_e_ -> EX_ser_L_e\n",
      "Updating reaction ID: EX_sfcys_e_ -> EX_sfcys_e\n",
      "Updating reaction ID: EX_sl_L_e_ -> EX_sl_L_e\n",
      "Updating reaction ID: EX_so4_e_ -> EX_so4_e\n",
      "Updating reaction ID: EX_spc_cho_e_ -> EX_spc_cho_e\n",
      "Updating reaction ID: EX_sph1p_e_ -> EX_sph1p_e\n",
      "Updating reaction ID: EX_sphs1p_e_ -> EX_sphs1p_e\n",
      "Updating reaction ID: EX_spmd_e_ -> EX_spmd_e\n",
      "Updating reaction ID: EX_sprm_e_ -> EX_sprm_e\n",
      "Updating reaction ID: EX_srtn_e_ -> EX_srtn_e\n",
      "Updating reaction ID: EX_strch1_e_ -> EX_strch1_e\n",
      "Updating reaction ID: EX_strch2_e_ -> EX_strch2_e\n",
      "Updating reaction ID: EX_strdnc_e_ -> EX_strdnc_e\n",
      "Updating reaction ID: EX_succ_e_ -> EX_succ_e\n",
      "Updating reaction ID: EX_sucr_e_ -> EX_sucr_e\n",
      "Updating reaction ID: EX_tag_cho_e_ -> EX_tag_cho_e\n",
      "Updating reaction ID: EX_tagat_D_e_ -> EX_tagat_D_e\n",
      "Updating reaction ID: EX_taur_e_ -> EX_taur_e\n",
      "Updating reaction ID: EX_taxol_e_ -> EX_taxol_e\n",
      "Updating reaction ID: EX_tchola_e_ -> EX_tchola_e\n",
      "Updating reaction ID: EX_tcynt_e_ -> EX_tcynt_e\n",
      "Updating reaction ID: EX_tdchola_e_ -> EX_tdchola_e\n",
      "Updating reaction ID: EX_tdechola_e_ -> EX_tdechola_e\n",
      "Updating reaction ID: EX_tethex3_e_ -> EX_tethex3_e\n",
      "Updating reaction ID: EX_tetpent3_e_ -> EX_tetpent3_e\n",
      "Updating reaction ID: EX_tetpent6_e_ -> EX_tetpent6_e\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_tettet6_e_ -> EX_tettet6_e\n",
      "Updating reaction ID: EX_thf_e_ -> EX_thf_e\n",
      "Updating reaction ID: EX_thm_e_ -> EX_thm_e\n",
      "Updating reaction ID: EX_thmmp_e_ -> EX_thmmp_e\n",
      "Updating reaction ID: EX_thmtp_e_ -> EX_thmtp_e\n",
      "Updating reaction ID: EX_thr_L_e_ -> EX_thr_L_e\n",
      "Updating reaction ID: EX_thym_e_ -> EX_thym_e\n",
      "Updating reaction ID: EX_thymd_e_ -> EX_thymd_e\n",
      "Updating reaction ID: EX_thyox_L_e_ -> EX_thyox_L_e\n",
      "Updating reaction ID: EX_tmndnc_e_ -> EX_tmndnc_e\n",
      "Updating reaction ID: EX_tolbutamide_e_ -> EX_tolbutamide_e\n",
      "Updating reaction ID: EX_tre_e_ -> EX_tre_e\n",
      "Updating reaction ID: EX_triodthy_e_ -> EX_triodthy_e\n",
      "Updating reaction ID: EX_triodthysuf_e_ -> EX_triodthysuf_e\n",
      "Updating reaction ID: EX_trp_L_e_ -> EX_trp_L_e\n",
      "Updating reaction ID: EX_tststerone_e_ -> EX_tststerone_e\n",
      "Updating reaction ID: EX_tststeroneglc_e_ -> EX_tststeroneglc_e\n",
      "Updating reaction ID: EX_tststerones_e_ -> EX_tststerones_e\n",
      "Updating reaction ID: EX_tsul_e_ -> EX_tsul_e\n",
      "Updating reaction ID: EX_ttdca_e_ -> EX_ttdca_e\n",
      "Updating reaction ID: EX_txa2_e_ -> EX_txa2_e\n",
      "Updating reaction ID: EX_tymsf_e_ -> EX_tymsf_e\n",
      "Updating reaction ID: EX_Tyr_ggn_e_ -> EX_Tyr_ggn_e\n",
      "Updating reaction ID: EX_tyr_L_e_ -> EX_tyr_L_e\n",
      "Updating reaction ID: EX_udp_e_ -> EX_udp_e\n",
      "Updating reaction ID: EX_udpg_e_ -> EX_udpg_e\n",
      "Updating reaction ID: EX_udpgal_e_ -> EX_udpgal_e\n",
      "Updating reaction ID: EX_ump_e_ -> EX_ump_e\n",
      "Updating reaction ID: EX_ura_e_ -> EX_ura_e\n",
      "Updating reaction ID: EX_urate_e_ -> EX_urate_e\n",
      "Updating reaction ID: EX_urea_e_ -> EX_urea_e\n",
      "Updating reaction ID: EX_uri_e_ -> EX_uri_e\n",
      "Updating reaction ID: EX_utp_e_ -> EX_utp_e\n",
      "Updating reaction ID: EX_vacc_e_ -> EX_vacc_e\n",
      "Updating reaction ID: EX_val_L_e_ -> EX_val_L_e\n",
      "Updating reaction ID: EX_vitd2_e_ -> EX_vitd2_e\n",
      "Updating reaction ID: EX_vitd3_e_ -> EX_vitd3_e\n",
      "Updating reaction ID: EX_whddca_e_ -> EX_whddca_e\n",
      "Updating reaction ID: EX_whhdca_e_ -> EX_whhdca_e\n",
      "Updating reaction ID: EX_whtststerone_e_ -> EX_whtststerone_e\n",
      "Updating reaction ID: EX_whttdca_e_ -> EX_whttdca_e\n",
      "Updating reaction ID: EX_xolest_cho_e_ -> EX_xolest_cho_e\n",
      "Updating reaction ID: EX_xolest2_cho_e_ -> EX_xolest2_cho_e\n",
      "Updating reaction ID: EX_xoltri24_e_ -> EX_xoltri24_e\n",
      "Updating reaction ID: EX_xoltri25_e_ -> EX_xoltri25_e\n",
      "Updating reaction ID: EX_xoltri27_e_ -> EX_xoltri27_e\n",
      "Updating reaction ID: EX_xyl_D_e_ -> EX_xyl_D_e\n",
      "Updating reaction ID: EX_xylt_e_ -> EX_xylt_e\n",
      "Updating reaction ID: EX_yvite_e_ -> EX_yvite_e\n"
     ]
    }
   ],
   "source": [
    "iCHO2291_changed = update_reaction_ids_specific(iCHO2291)\n",
    "iCHO1766_changed = update_reaction_ids_specific(iCHO1766)\n",
    "iCHO2101_changed = update_reaction_ids_specific(iCHO2101)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reaction ID: 3KLGALHGm_e, Reaction Name: 3-Ketolactose galactohydrolase Galactose metabolism EC:3.2.1.23\n",
      "Reaction ID: OXYODPCm_e, Reaction Name: Fe(II):oxygen oxidoreductase Porphyrin and chlorophyll metabolism EC:1.16.3.1\n",
      "Reaction ID: EX_HC01610_e, Reaction Name: Exchange of CoproporphyrinogenI\n",
      "Reaction ID: URPHYRCLPCm_e, Reaction Name: Uroporphyrinogen I carboxy-lyase Porphyrin and chlorophyll metabolism EC:4.1.1.37\n",
      "Reaction ID: EX_HC02179_e, Reaction Name: 20-hydroxy-arachidonate exchange\n",
      "Reaction ID: EX_C05965_e, Reaction Name: 12(S)-HPETE(1-) exchange\n",
      "Reaction ID: EX_HC02176_e, Reaction Name: laurate exchange\n",
      "Reaction ID: DOLGPP_er, Reaction Name: Dolichyl-beta-D-glucosyl-phosphate dolichylphosphohydrolase (uterus)\n",
      "Reaction ID: DOLPH_er, Reaction Name: Dolichol-phosphate phosphohydrolase, human (uterus)\n",
      "Reaction ID: DOLPMT2_er, Reaction Name: dolichyl-phosphate-mannose-glycolipid alpha-mannosyltransferase (uterus)\n",
      "Reaction ID: DOLPMT4_er, Reaction Name: dolichyl-phosphate-mannose-glycolipid alpha-mannosyltransferase (uterus)\n",
      "Reaction ID: EX_10fthf_e, Reaction Name: 10-Formyltetrahydrofolate exchange\n",
      "Reaction ID: EX_10fthf5glu_e, Reaction Name: 10-formyltetrahydrofolate-[Glu](5) exchange\n",
      "Reaction ID: EX_10fthf6glu_e, Reaction Name: 10-formyltetrahydrofolate-[Glu](6) exchange\n",
      "Reaction ID: EX_10fthf7glu_e, Reaction Name: 10-formyltetrahydrofolate-[Glu](7) exchange\n",
      "Reaction ID: EX_11_cis_retfa_e, Reaction Name: fatty acid 11-cis-retinol exchange\n",
      "Reaction ID: EX_13_cis_retnglc_e, Reaction Name: 13-cis-retinoyl glucuronide exchange\n",
      "Reaction ID: EX_1glyc_cho_e, Reaction Name: 1 acyl phosphoglycerol exchange\n",
      "Reaction ID: EX_1mncam_e, Reaction Name: 1-Methylnicotinamide exchange\n",
      "Reaction ID: EX_23cump_e, Reaction Name: 2,3-cyclic UMP(1-) exchange\n",
      "Reaction ID: EX_2425dhvitd2_e, Reaction Name: 24R,25-Dihyoxyvitamin D2 exchange\n",
      "Reaction ID: EX_2425dhvitd3_e, Reaction Name: 24R,25-Dihydroxyvitamin D3 exchange\n",
      "Reaction ID: EX_24nph_e, Reaction Name: 2,4 dihydroxy nitrophenol exchange\n",
      "Reaction ID: EX_25hvitd2_e, Reaction Name: 25-Hydroxyvitamin D2 exchange\n",
      "Reaction ID: EX_25hvitd3_e, Reaction Name: 25-Hydroxyvitamin D3 exchange\n",
      "Reaction ID: EX_2hb_e, Reaction Name: 2-Hydroxybutyrate exchange\n",
      "Reaction ID: EX_2mcit_e, Reaction Name: 2-Methylcitrate exchange\n",
      "Reaction ID: EX_34dhoxpeg_e, Reaction Name: 3,4-Dihydroxyphenylethyleneglycol exchange\n",
      "Reaction ID: EX_34dhphe_e, Reaction Name: 3,4-Dihydroxy-L-phenylalanine exchange\n",
      "Reaction ID: EX_35cgmp_e, Reaction Name: 3,5-Cyclic GMP exchange\n",
      "Reaction ID: EX_3aib_D_e, Reaction Name: D-3-Amino-isobutanoate exchange\n",
      "Reaction ID: EX_3aib_e, Reaction Name: L-3-Amino-isobutanoate exchange\n",
      "Reaction ID: EX_3mlda_e, Reaction Name: 3-Methylimidazoleacetic acid exchange\n",
      "Reaction ID: EX_3mob_e, Reaction Name: Exchange of 3-methyl-2-oxobutanoate\n",
      "Reaction ID: EX_3mop_e, Reaction Name: Exchange of 3-methyl-2-oxopentanoate\n",
      "Reaction ID: EX_3ump_e, Reaction Name: 3-UMP(2-) exchange\n",
      "Reaction ID: EX_4abut_e, Reaction Name: 4-Aminobutanoate exchange\n",
      "Reaction ID: EX_4abutn_e, Reaction Name: Exchange of 4-ammoniobutanal\n",
      "Reaction ID: EX_4hdebrisoquine_e, Reaction Name: 4 hydroxy debrisoquine exchange\n",
      "Reaction ID: EX_4hphac_e, Reaction Name: 4-Hydroxyphenylacetate exchange\n",
      "Reaction ID: EX_4hpro_e, Reaction Name: exchange reaction for hydroxy proline\n",
      "Reaction ID: EX_4mop_e, Reaction Name: Exchange of 4-methyl-2-oxopentanoate\n",
      "Reaction ID: EX_4mptnl_e, Reaction Name: 4-methylpentanal Exchange\n",
      "Reaction ID: EX_4mtolbutamide_e, Reaction Name: 4 hydroxy tolbutamide exchange\n",
      "Reaction ID: EX_4nph_e, Reaction Name: 4-Nitrophenol exchange\n",
      "Reaction ID: EX_4nphsf_e, Reaction Name: 4-Nitrophenyl sulfate exchange\n",
      "Reaction ID: EX_4pyrdx_e, Reaction Name: 4-Pyridoxate exchange\n",
      "Reaction ID: EX_5adtststerone_e, Reaction Name: 5alpha-Dihydrotestosterone exchange\n",
      "Reaction ID: EX_5adtststeroneglc_e, Reaction Name: 5alpha-Dihydrotestosterone glucuronide exchange\n",
      "Reaction ID: EX_5adtststerones_e, Reaction Name: 5alpha-Dihydrotestosterone sulfate exchange\n",
      "Reaction ID: EX_5dhf_e, Reaction Name: pentaglutamyl folate (DHF) exchange\n",
      "Reaction ID: EX_5fthf_e, Reaction Name: 5-Formyltetrahydrofolate exchange\n",
      "Reaction ID: EX_5homeprazole_e, Reaction Name: 5 hydroxy omeprazole exchange\n",
      "Reaction ID: EX_5htrp_e, Reaction Name: 5-Hydroxy-L-tryptophan exchange\n",
      "Reaction ID: EX_5mta_e, Reaction Name: Exchange of 5-S-methyl-5-thioadenosine\n",
      "Reaction ID: EX_5mthf_e, Reaction Name: exchange reaction for 5-Methyltetrahydrofolate\n",
      "Reaction ID: EX_5oxpro_e, Reaction Name: Exchange of 5-oxoprolinate\n",
      "Reaction ID: EX_5thf_e, Reaction Name: pentaglutamyl folate (THF) exchange\n",
      "Reaction ID: EX_6dhf_e, Reaction Name: haxglutamyl folate (DHF) exchange\n",
      "Reaction ID: EX_6htststerone_e, Reaction Name: 6 beta hydroxy testosterone exchange\n",
      "Reaction ID: EX_6thf_e, Reaction Name: hexaglutamyl folate (THF) exchange\n",
      "Reaction ID: EX_7dhf_e, Reaction Name: heptaglutamyl folate (DHF) exchange\n",
      "Reaction ID: EX_7thf_e, Reaction Name: heptaglutamyl folate (THF) exchange\n",
      "Reaction ID: EX_9_cis_retfa_e, Reaction Name: fatty acid 9-cis-retinol exchange\n",
      "Reaction ID: EX_abt_e, Reaction Name: L-Arabinitol exchange\n",
      "Reaction ID: EX_ac_e, Reaction Name: Acetate exchange\n",
      "Reaction ID: EX_acac_e, Reaction Name: Acetoacetate exchange\n",
      "Reaction ID: EX_acald_e, Reaction Name: Acetaldehyde exchange\n",
      "Reaction ID: EX_acetone_e, Reaction Name: Acetone exchange\n",
      "Reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e, Reaction Name: Type IIIA glycolipid exchange\n",
      "Reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e, Reaction Name: Type IIIAb exchange\n",
      "Reaction ID: EX_acgam_e, Reaction Name: N-Acetyl-D-glucosamine exchange\n",
      "Reaction ID: EX_ach_e, Reaction Name: Acetylcholine exchange\n",
      "Reaction ID: EX_acmana_e, Reaction Name: Exchange of N-acetyl-D-mannosamine\n",
      "Reaction ID: EX_acn13acngalgbside_cho_e, Reaction Name: sialyl (1,3) sialyl (2,6) galactosylgloboside (cho) exchange\n",
      "Reaction ID: EX_acn23acngalgbside_cho_e, Reaction Name: sialyl (2,3) sialyl (2,6) galactosylgloboside (cho) exchange\n",
      "Reaction ID: EX_acnacngal14acglcgalgluside_cho_e, Reaction Name: 3,8-LD1 exchange\n",
      "Reaction ID: EX_acnacngalgbside_cho_e, Reaction Name: disialyl galactosylgloboside (cho) exchange\n",
      "Reaction ID: EX_acngalacglcgal14acglcgalgluside_cho_e, Reaction Name: VI3NeuAc-nLc6Cer exchange\n",
      "Reaction ID: EX_ade_e, Reaction Name: Adenine exchange\n",
      "Reaction ID: EX_adn_e, Reaction Name: exchange reaction for Adenosine\n",
      "Reaction ID: EX_adpcbl_e, Reaction Name: exchange reaction for Adenosylcobalamin\n",
      "Reaction ID: EX_adprbp_e, Reaction Name: ADPribose 2-phosphate exchange\n",
      "Reaction ID: EX_adprib_e, Reaction Name: ADPribose exchange\n",
      "Reaction ID: EX_adrn_e, Reaction Name: adrenic acid exchange\n",
      "Reaction ID: EX_adrnl_e, Reaction Name: Adrenaline exchange\n",
      "Reaction ID: EX_aflatoxin_e, Reaction Name: aflatoxin B1 exchange\n",
      "Reaction ID: EX_ahandrostanglc_e, Reaction Name: Etiocholan-3alpha-ol-17-one 3-glucuronide exchange\n",
      "Reaction ID: EX_ahcys_e, Reaction Name: Exchange of S-adenosyl-L-homocysteine\n",
      "Reaction ID: EX_ahdt_e, Reaction Name: Exchange of 7,8-dihydroneopterin 3-triphosphate(4-)\n",
      "Reaction ID: EX_aicar_e, Reaction Name: Exchange of 5-amino-1-(5-phospho-D-ribosyl)imidazole-4-carboxamide(2-)\n",
      "Reaction ID: EX_ak2lgchol_cho_e, Reaction Name: 1-alkyl 2-lysoglycerol 3-phosphocholine exchange\n",
      "Reaction ID: EX_akg_e, Reaction Name: 2-Oxoglutarate exchange\n",
      "Reaction ID: EX_ala_B_e, Reaction Name: beta-Alanine exchange\n",
      "Reaction ID: EX_ala_D_e, Reaction Name: D-Alanine exchange\n",
      "Reaction ID: EX_ala_L_e, Reaction Name: exchange reaction for L-alanine\n",
      "Reaction ID: EX_alaala_e, Reaction Name: exchange reaction for alanylalanine\n",
      "Reaction ID: EX_aldstrn_e, Reaction Name: Aldosterone exchange\n",
      "Reaction ID: EX_amp_e, Reaction Name: AMP exchange\n",
      "Reaction ID: EX_andrstrn_e, Reaction Name: Androsterone exchange\n",
      "Reaction ID: EX_andrstrnglc_e, Reaction Name: androsterone glucuronide exchange\n",
      "Reaction ID: EX_anth_e, Reaction Name: Exchange of anthranilate\n",
      "Reaction ID: EX_antipyrene_e, Reaction Name: antipyrene exchange\n",
      "Reaction ID: EX_apnnox_e, Reaction Name: alpha-Pinene-oxide exchange\n",
      "Reaction ID: EX_appnn_e, Reaction Name: (+)-alpha-Pinene exchange\n",
      "Reaction ID: EX_aprgstrn_e, Reaction Name: 20alpha-Hydroxyprogesterone exchange\n",
      "Reaction ID: EX_aqcobal_e, Reaction Name: Aquacob(III)alamin exchange\n",
      "Reaction ID: EX_arab_L_e, Reaction Name: L-Arabinose exchange\n",
      "Reaction ID: EX_arach_e, Reaction Name: exchange reaction for arachidate\n",
      "Reaction ID: EX_arachcoa_e, Reaction Name: arachidyl coenzyme A exchange\n",
      "Reaction ID: EX_arachd_e, Reaction Name: nC20:4 exchange\n",
      "Reaction ID: EX_arg_L_e, Reaction Name: L-Arginine exchange\n",
      "Reaction ID: EX_ascb_L_e, Reaction Name: L-Ascorbate exchange\n",
      "Reaction ID: EX_asn_L_e, Reaction Name: exchange reaction for L-asparagine\n",
      "Reaction ID: EX_asp_D_e, Reaction Name: D-Aspartate exchange\n",
      "Reaction ID: EX_asp_L_e, Reaction Name: L-Aspartate exchange\n",
      "Reaction ID: EX_atp_e, Reaction Name: ATP exchange\n",
      "Reaction ID: EX_avite1_e, Reaction Name: alpha-Tocopherol exchange\n",
      "Reaction ID: EX_avite2_e, Reaction Name: alpha-Tocotrienol exchange\n",
      "Reaction ID: EX_bglc_e, Reaction Name: Exchange reaction for beta glucans\n",
      "Reaction ID: EX_bhb_e, Reaction Name: (R)-3-Hydroxybutanoate transport via H+ symport\n",
      "Reaction ID: EX_bildglcur_e, Reaction Name: Bilirubin beta-diglucuronide exchange\n",
      "Reaction ID: EX_bilglcur_e, Reaction Name: exchange reaction for blirubin mono-glucuronide\n",
      "Reaction ID: EX_bilirub_e, Reaction Name: Bilirubin exchange\n",
      "Reaction ID: EX_biocyt_e, Reaction Name: Biocytin exchange\n",
      "Reaction ID: EX_btn_e, Reaction Name: Biotin exchange\n",
      "Reaction ID: EX_but_e, Reaction Name: Butyrate (n-C4:0) exchange\n",
      "Reaction ID: EX_bvite_e, Reaction Name: beta-tocopherol exchange\n",
      "Reaction ID: EX_bz_e, Reaction Name: Benzoate exchange\n",
      "Reaction ID: EX_xtra_e, Reaction Name: Xanthurenic acid exchange\n",
      "Reaction ID: EX_C02528_e, Reaction Name: chenodeoxycholate exchange\n",
      "Reaction ID: EX_12e8hdxt_e, Reaction Name: (5Z,9E,14Z)-(8xi,11R,12S)-11,12-epoxy-8-hydroxyicosa-5,9,14-trienoate exchange\n",
      "Reaction ID: EX_ca2_e, Reaction Name: Calcium exchange\n",
      "Reaction ID: EX_camp_e, Reaction Name: cAMP exchange\n",
      "Reaction ID: EX_carn_e, Reaction Name: exchange reaction for carnosine\n",
      "Reaction ID: EX_caro_e, Reaction Name: beta-Carotene exchange\n",
      "Reaction ID: EX_carveol_e, Reaction Name: (E)-carveol exchange\n",
      "Reaction ID: EX_cbasp_e, Reaction Name: Exchange of N-carbamoyl-L-aspartate\n",
      "Reaction ID: EX_cca_d3_e, Reaction Name: Calcitroic acid (D3) exchange\n",
      "Reaction ID: EX_cdp_e, Reaction Name: CDP(3-) exchange\n",
      "Reaction ID: EX_CE0074_e, Reaction Name: alloxan exchange\n",
      "Reaction ID: EX_CE1925_e, Reaction Name: 3-carboxy-alpha-chromanol exchange\n",
      "Reaction ID: EX_CE1926_e, Reaction Name: gama-carboxyethyl-hydroxychroman exchange\n",
      "Reaction ID: EX_CE1935_e, Reaction Name: spermine monoaldehyde exchange\n",
      "Reaction ID: EX_CE1936_e, Reaction Name: spermine dialdehyde exchange\n",
      "Reaction ID: EX_CE1939_e, Reaction Name: spermidine monoaldehyde 1 exchange\n",
      "Reaction ID: EX_CE1940_e, Reaction Name: spermidine monoaldehyde 2 exchange\n",
      "Reaction ID: EX_CE1943_e, Reaction Name: spermidine dialdehyde exchange\n",
      "Reaction ID: EX_CE2011_e, Reaction Name: hypothiocyanite exchange\n",
      "Reaction ID: EX_CE2250_e, Reaction Name: 3-oxodocosanoyl-CoA exchange\n",
      "Reaction ID: EX_CE2838_e, Reaction Name: maltononaose exchange\n",
      "Reaction ID: EX_CE2839_e, Reaction Name: maltodecaose exchange\n",
      "Reaction ID: EX_CE2915_e, Reaction Name: beta-casomorphin exchange\n",
      "Reaction ID: EX_CE2916_e, Reaction Name: neocasomorphin exchange\n",
      "Reaction ID: EX_CE2917_e, Reaction Name: apelin-13 exchange\n",
      "Reaction ID: EX_CE4633_e, Reaction Name: hypochlorous acid exchange\n",
      "Reaction ID: EX_CE4722_e, Reaction Name: beta-casomorphin (1-6) exchange\n",
      "Reaction ID: EX_CE4723_e, Reaction Name: neocasomorphin (1-5) exchange\n",
      "Reaction ID: EX_CE4724_e, Reaction Name: apelin (1-12) exchange\n",
      "Reaction ID: EX_CE4881_e, Reaction Name: nitryl chloride exchange\n",
      "Reaction ID: EX_CE5786_e, Reaction Name: kinetensin exchange\n",
      "Reaction ID: EX_CE5788_e, Reaction Name: kinetensin 1-7 exchange\n",
      "Reaction ID: EX_CE5789_e, Reaction Name: kinetensin 1-8 exchange\n",
      "Reaction ID: EX_CE5797_e, Reaction Name: neuromedin N exchange\n",
      "Reaction ID: EX_CE5798_e, Reaction Name: neuromedin N (1-4) exchange\n",
      "Reaction ID: EX_CE5853_e, Reaction Name: alpha-CEHC-glucuronide exchange\n",
      "Reaction ID: EX_CE5854_e, Reaction Name: gama-CEHC-glucuronide exchange\n",
      "Reaction ID: EX_CE5867_e, Reaction Name: N-acetyl-seryl-aspartyl-lysyl-proline exchange\n",
      "Reaction ID: EX_CE5868_e, Reaction Name: N-acetyl-seryl-aspartate exchange\n",
      "Reaction ID: EX_CE5869_e, Reaction Name: lysyl-proline exchange\n",
      "Reaction ID: EX_cgly_e, Reaction Name: Cys-Gly exchange\n",
      "Reaction ID: EX_chol_e, Reaction Name: exchange reaction for Choline\n",
      "Reaction ID: EX_cholate_e, Reaction Name: exchange reaction for cholate\n",
      "Reaction ID: EX_chsterol_e, Reaction Name: exchange reaction for cholesterol\n",
      "Reaction ID: EX_chtn_e, Reaction Name: chitin exchange\n",
      "Reaction ID: EX_cit_e, Reaction Name: Citrate exchange\n",
      "Reaction ID: EX_citr_L_e, Reaction Name: L-citrulline exchange\n",
      "Reaction ID: EX_cl_e, Reaction Name: exchange reaction for Chloride\n",
      "Reaction ID: EX_CLPND_e, Reaction Name: clupanodonic acid (docosapentaenoic (n-3)) exchange\n",
      "Reaction ID: EX_cmp_e, Reaction Name: CMP exchange\n",
      "Reaction ID: EX_co_e, Reaction Name: Carbon monoxide exchange\n",
      "Reaction ID: EX_co2_e, Reaction Name: CO2 exchange\n",
      "Reaction ID: EX_coa_e, Reaction Name: exchange reaction for coa\n",
      "Reaction ID: EX_coumarin_e, Reaction Name: coumarin exchange\n",
      "Reaction ID: EX_creat_e, Reaction Name: Creatine exchange\n",
      "Reaction ID: EX_crm_cho_e, Reaction Name: N-acylsphingosine exchange\n",
      "Reaction ID: EX_crmp_cho_e, Reaction Name: Ceramide 1-phosphate exchange\n",
      "Reaction ID: EX_crn_e, Reaction Name: exchange reaction for L-Carnitine\n",
      "Reaction ID: EX_crtsl_e, Reaction Name: cortisol Exchange\n",
      "Reaction ID: EX_crtstrn_e, Reaction Name: corticosterone Exchange\n",
      "Reaction ID: EX_crvnc_e, Reaction Name: nC22:6 exchange\n",
      "Reaction ID: EX_csn_e, Reaction Name: Cytosine exchange\n",
      "Reaction ID: EX_cspg_a_e, Reaction Name: chondroitin sulfate A (GalNAc4S-GlcA) proteoglycan exchange\n",
      "Reaction ID: EX_cspg_b_e, Reaction Name: chondroitin sulfate B / dermatan sulfate (IdoA2S-GalNAc4S) proteoglycan exchange\n",
      "Reaction ID: EX_cspg_c_e, Reaction Name: chondroitin sulfate C (GalNAc6S-GlcA) proteoglycan exchange\n",
      "Reaction ID: EX_cspg_d_e, Reaction Name: chondroitin sulfate D (GlcNAc6S-GlcA2S) proteoglycan exchange\n",
      "Reaction ID: EX_cspg_ee_, Reaction Name: chondroitin sulfate E (GalNAc4,6diS-GlcA) proteoglycan exchange\n",
      "Reaction ID: EX_ctp_e, Reaction Name: Exchange of CTP(4-)\n",
      "Reaction ID: EX_cyan_e, Reaction Name: Hydrogen cyanide exchange\n",
      "Reaction ID: EX_cys_L_e, Reaction Name: exchange reaction for L-cysteine\n",
      "Reaction ID: EX_cysam_e, Reaction Name: exchange reaction for cysam\n",
      "Reaction ID: EX_cytd_e, Reaction Name: exchange reaction for Cytidine\n",
      "Reaction ID: EX_dad_2_e, Reaction Name: exchange reaction for 2-deoxyadenosine\n",
      "Reaction ID: EX_dag_cho_e, Reaction Name: exchange reaction for diglyceride\n",
      "Reaction ID: EX_dca_e, Reaction Name: Decanoate (n-C10:0) exchange\n",
      "Reaction ID: EX_dchac_e, Reaction Name: exchange reaction for deoxycholic acid\n",
      "Reaction ID: EX_dcsptn1_e, Reaction Name: docosa-4,7,10,13,16-pentaenoic acid (n-6) exchange\n",
      "Reaction ID: EX_dcyt_e, Reaction Name: exchange reaction for Deoxycytidine\n",
      "Reaction ID: EX_debrisoquine_e, Reaction Name: debrisoquine exchange\n",
      "Reaction ID: EX_dgchol_e, Reaction Name: Chenodeoxyglycocholate exchange\n",
      "Reaction ID: EX_dgmp_e, Reaction Name: Exchange of dGMP(2-)\n",
      "Reaction ID: EX_dgsn_e, Reaction Name: exchange reaction for Deoxyguanosine\n",
      "Reaction ID: EX_dgtp_e, Reaction Name: Exchange of dGTP\n",
      "Reaction ID: EX_dha_e, Reaction Name: Exchange of glycerone\n",
      "Reaction ID: EX_dhap_e, Reaction Name: Exchange of glycerone phosphate(2-)\n",
      "Reaction ID: EX_dhdascb_e, Reaction Name: exchange reaction for dehydroascorbide(1-)\n",
      "Reaction ID: EX_dheas_e, Reaction Name: Dehydroepiandrosterone sulfate exchange\n",
      "Reaction ID: EX_dhf_e, Reaction Name: 7,8-Dihydrofolate exchange\n",
      "Reaction ID: EX_digalsgalside_cho_e, Reaction Name: Digalactosylceramidesulfate exchange\n",
      "Reaction ID: EX_din_e, Reaction Name: exchange reaction for Deoxyinosine\n",
      "Reaction ID: EX_dlnlcg_e, Reaction Name: dihomo-gamma-linolenic acid (n-6) exchange\n",
      "Reaction ID: EX_dmantipyrine_e, Reaction Name: demethylated antipyrine exchange\n",
      "Reaction ID: EX_dmhptcrn_e, Reaction Name: 2,6 dimethylheptanoyl carnitine exchange\n",
      "Reaction ID: EX_dopa_e, Reaction Name: Dopamine exchange\n",
      "Reaction ID: EX_dopasf_e, Reaction Name: Dopamine 3-O-sulfate exchange\n",
      "Reaction ID: EX_dpcoa_e, Reaction Name: exchange reaction for dpcoa\n",
      "Reaction ID: EX_drib_e, Reaction Name: Deoxyribose exchange\n",
      "Reaction ID: EX_dtdp_e, Reaction Name: ({[(2R,3S,5R)-3-hydroxy-5-(5-methyl-2,4-dioxo-1,2,3,4-tetrahydropyrimidin-1-yl)oxolan-2-yl]methyl phosphonato}oxy)phosphonate exchange\n",
      "Reaction ID: EX_dtmp_e, Reaction Name: Exchange of dTMP(2-)\n",
      "Reaction ID: EX_dttp_e, Reaction Name: Exchange of dTTP(4-)\n",
      "Reaction ID: EX_duri_e, Reaction Name: exchange reaction for Deoxyuridine\n",
      "Reaction ID: EX_eaflatoxin_e, Reaction Name: 8,9 epxoy aflatoxin B1 exchange\n",
      "Reaction ID: EX_ebastine_e, Reaction Name: ebastine exchange\n",
      "Reaction ID: EX_ebastineoh_e, Reaction Name: hydroxylated ebastine exchange\n",
      "Reaction ID: EX_eicostet_e, Reaction Name: eicosatetranoic acid exchange\n",
      "Reaction ID: EX_elaid_e, Reaction Name: elaidic acid exchange\n",
      "Reaction ID: EX_estradiol_e, Reaction Name: estradiol exchange\n",
      "Reaction ID: EX_estradiolglc_e, Reaction Name: estradiol glucuronide exchange\n",
      "Reaction ID: EX_estriolglc_e, Reaction Name: 16-Glucuronide-estriol exchange\n",
      "Reaction ID: EX_estroneglc_e, Reaction Name: estrone glucuronide exchange\n",
      "Reaction ID: EX_estrones_e, Reaction Name: Estrone 3-sulfate exchange\n",
      "Reaction ID: EX_etoh_e, Reaction Name: Ethanol exchange\n",
      "Reaction ID: EX_fad_e, Reaction Name: Exchange of Flavin adenine dinucleotide oxidized\n",
      "Reaction ID: EX_fald_e, Reaction Name: Exchange of formaldehyde\n",
      "Reaction ID: EX_fe2_e, Reaction Name: EX fe2(u)\n",
      "Reaction ID: EX_fe3_e, Reaction Name: Fe3+ exchange\n",
      "Reaction ID: EX_fmn_e, Reaction Name: exchange reaction for FMN\n",
      "Reaction ID: EX_fol_e, Reaction Name: exchange reaction for Folate\n",
      "Reaction ID: EX_for_e, Reaction Name: Formate exchange\n",
      "Reaction ID: EX_fru_e, Reaction Name: exchange reaction for D-Fructose\n",
      "Reaction ID: EX_fuc_L_e, Reaction Name: L-Fucose exchange\n",
      "Reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho_e, Reaction Name: III3Fuc-nLc6Cer exchange\n",
      "Reaction ID: EX_fuc14galacglcgalgluside_cho_e, Reaction Name: Lea glycolipid exchange\n",
      "Reaction ID: EX_fucacgalfucgalacglcgalgluside_cho_e, Reaction Name: (Gal)2 (GalNAc)1 (Glc)1 (GlcNAc)1 (LFuc)2 (Cer)1 exchange\n",
      "Reaction ID: EX_fucacngal14acglcgalgluside_cho_e, Reaction Name: IV3-a-NeuAc,III3-a-Fuc-nLc4Cer exchange\n",
      "Reaction ID: EX_fucacngalacglcgalgluside_cho_e, Reaction Name: IV3-a-Neu5Ac,III4-a-Fuc-Lc4Cer exchange\n",
      "Reaction ID: EX_fucfuc12gal14acglcgalgluside_cho_e, Reaction Name: Ley glycolipid exchange\n",
      "Reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho_e, Reaction Name: V3Fuc,III3Fuc-nLc6Cer exchange\n",
      "Reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e, Reaction Name: (Gal)4 (Glc)1 (GlcNAc)3 (LFuc)3 (Cer)1 exchange\n",
      "Reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e, Reaction Name: (Gal)3 (Glc)1 (GlcNAc)2 (LFuc)3 (Cer)1 exchange\n",
      "Reaction ID: EX_fucfucgalacglcgalgluside_cho_e, Reaction Name: Leb glycolipid exchange\n",
      "Reaction ID: EX_fucgal14acglcgalgluside_cho_e, Reaction Name: Lacto-N-fucopentaosyl III ceramide exchange\n",
      "Reaction ID: EX_fucgalfucgalacglcgalgluside_cho_e, Reaction Name: (Gal)3 (Glc)1 (GlcNAc)1 (LFuc)2 (Cer)1 exchange\n",
      "Reaction ID: EX_fucgalgbside_cho_e, Reaction Name: fucosyl galactosylgloboside (cho) exchange\n",
      "Reaction ID: EX_g1p_e, Reaction Name: Exchange of alpha-D-glucose 1-phosphate(2-)\n",
      "Reaction ID: EX_gal_e, Reaction Name: exchange reaction for D-Galactose\n",
      "Reaction ID: EX_galacglcgalgbside_cho_e, Reaction Name: Gal-GlcNAc-Gal globoside (cho) exchange\n",
      "Reaction ID: EX_galfuc12gal14acglcgalgluside_cho_e, Reaction Name: (Gal)3 (Glc)1 (GlcNAc)1 (LFuc)1 (Cer)1 exchange\n",
      "Reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho_e, Reaction Name: (Gal)4 (Glc)1 (GlcNAc)2 (LFuc)1 (Cer)1 exchange\n",
      "Reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e, Reaction Name: (Gal)6 (Glc)1 (GlcNAc)3 (LFuc)2 (Cer)1 exchange\n",
      "Reaction ID: EX_galgalgalthcrm_cho_e, Reaction Name: Gal-Gal-Gal-Gal-Gal-Glc-Cer (cho) exchange\n",
      "Reaction ID: EX_galside_cho_e, Reaction Name: D-galactosyl-N-acylsphingosine exchange\n",
      "Reaction ID: EX_gam_e, Reaction Name: exchange reaction for D-Glucosamine\n",
      "Reaction ID: EX_gbside_cho_e, Reaction Name: globoside (cho) exchange\n",
      "Reaction ID: EX_gchola_e, Reaction Name: Glycocholate exchange\n",
      "Reaction ID: EX_gd1b2_cho_e, Reaction Name: GD1beta (cho) exchange\n",
      "Reaction ID: EX_gd1c_cho_e, Reaction Name: GD1c (cho) exchange\n",
      "Reaction ID: EX_gdp_e, Reaction Name: GDP exchange\n",
      "Reaction ID: EX_glc_e, Reaction Name: D-Glucose exchange\n",
      "Reaction ID: EX_glcur_e, Reaction Name: D-glucuronate exchange\n",
      "Reaction ID: EX_glgchlo_e, Reaction Name: exchange reaction for beta glucan-glycocholate complex\n",
      "Reaction ID: EX_gln_L_e, Reaction Name: exchange reaction for L-glutamine\n",
      "Reaction ID: EX_gltcho_e, Reaction Name: exchange reaction for beta glucan-taurocholic acid complex\n",
      "Reaction ID: EX_gltdechol_e, Reaction Name: exchange reaction for beta glucan-taurodeoxycholic acid complex\n",
      "Reaction ID: EX_glu_L_e, Reaction Name: L-Glutamate exchange\n",
      "Reaction ID: EX_gluala_e, Reaction Name: 5-L-Glutamyl-L-alanine exchange\n",
      "Reaction ID: EX_gly_e, Reaction Name: exchange reaction for Glycine\n",
      "Reaction ID: EX_glyb_e, Reaction Name: Glycine betaine exchange\n",
      "Reaction ID: EX_glyc_e, Reaction Name: Glycerol exchange\n",
      "Reaction ID: EX_glyc_S_e, Reaction Name: (S)-Glycerate exchange\n",
      "Reaction ID: EX_glygly_e, Reaction Name: Exchange reaction for glycylglycine\n",
      "Reaction ID: EX_glygn2_e, Reaction Name: glycogen, structure 2 (glycogenin-1,6-{7[1,4-Glc], 4[1,4-Glc]}) exchange\n",
      "Reaction ID: EX_glygn4_e, Reaction Name: exchange reaction for glycogen, structure 4 (glycogenin-1,6-{2[1,4-Glc], [1,4-Glc]})\n",
      "Reaction ID: EX_glygn5_e, Reaction Name: exchange reaction for glycogen, structure 5 (glycogenin-2[1,4-Glc])\n",
      "Reaction ID: EX_glyleu_e, Reaction Name: Glycylleucine exchange\n",
      "Reaction ID: EX_glyphe_e, Reaction Name: exchange reaction for glycylphenylalaine\n",
      "Reaction ID: EX_glypro_e, Reaction Name: Glycylproline exchange\n",
      "Reaction ID: EX_glysar_e, Reaction Name: exchange reaction for glycylsarcosine\n",
      "Reaction ID: EX_gmp_e, Reaction Name: GMP exchange\n",
      "Reaction ID: EX_gp1c_cho_e, Reaction Name: GP1c (cho) exchange\n",
      "Reaction ID: EX_gp1calpha_cho_e, Reaction Name: GP1c alpha (cho) exchange\n",
      "Reaction ID: EX_gq1b_cho_e, Reaction Name: GQ1b (cho) exchange\n",
      "Reaction ID: EX_gq1balpha_cho_e, Reaction Name: GQ1balpha (cho) exchange\n",
      "Reaction ID: EX_gsn_e, Reaction Name: exchange reaction for Guanosine\n",
      "Reaction ID: EX_gt1a_cho_e, Reaction Name: GT1a (cho) exchange\n",
      "Reaction ID: EX_gthox_e, Reaction Name: Oxidized glutathione exchange\n",
      "Reaction ID: EX_gthrd_e, Reaction Name: Reduced glutathione exchange\n",
      "Reaction ID: EX_gtp_e, Reaction Name: GTP exchange\n",
      "Reaction ID: EX_gua_e, Reaction Name: Guanine exchange\n",
      "Reaction ID: EX_gum_e, Reaction Name: Exchange reaction for guar gums\n",
      "Reaction ID: EX_gumdchac_e, Reaction Name: Exchange reaction for guar gum-deoxyxholic acid complex\n",
      "Reaction ID: EX_gumgchol_e, Reaction Name: exchange reaction for guar gum-glycocholate complex\n",
      "Reaction ID: EX_gumtchol_e, Reaction Name: exchange reaction for guar gum-taurocholic acid complex\n",
      "Reaction ID: EX_h_e, Reaction Name: exchange reaction for proton\n",
      "Reaction ID: EX_h2o_e, Reaction Name: H2O exchange\n",
      "Reaction ID: EX_h2o2_e, Reaction Name: Hydrogen peroxide exchange\n",
      "Reaction ID: EX_ha_e, Reaction Name: hyaluronan exchange\n",
      "Reaction ID: EX_ha_pre1_e, Reaction Name: hyaluronan biosynthesis, precursor 1 exchange\n",
      "Reaction ID: EX_HC00004_e, Reaction Name: ApoA1 exchange\n",
      "Reaction ID: EX_HC00229_e, Reaction Name: Exchange of Isomaltose\n",
      "Reaction ID: EX_HC00250_e, Reaction Name: Exchange of hydrosulfide\n",
      "Reaction ID: EX_HC00822_e, Reaction Name: Chitobiose exchange\n",
      "Reaction ID: EX_HC00955_e, Reaction Name: L-3-Cyanoalanine exchange\n",
      "Reaction ID: EX_HC01104_e, Reaction Name: Exchange of 4-nitrophenyl phosphate(2-)\n",
      "Reaction ID: EX_HC01361_e, Reaction Name: Exchange of Dihydroneopterin\n",
      "Reaction ID: EX_HC01440_e, Reaction Name: Exchange of 3-Keto-beta-D-galactose\n",
      "Reaction ID: EX_HC01441_e, Reaction Name: Exchange of Lactose-6P\n",
      "Reaction ID: EX_HC01444_e, Reaction Name: Exchange of Galactosylglycerol\n",
      "Reaction ID: EX_HC01446_e, Reaction Name: Exchange of 3-Ketolactose\n",
      "Reaction ID: EX_HC01577_e, Reaction Name: Exchange of gamma-Glutamyl-beta-cyanoalanine\n",
      "Reaction ID: EX_HC01609_e, Reaction Name: Exchange of UroporphyrinogenI\n",
      "Reaction ID: EX_HC01700_e, Reaction Name: Exchange of gamma-Glutamyl-3-aminopropiononitrile\n",
      "Reaction ID: EX_HC01787_e, Reaction Name: Lepidimoide exchange\n",
      "Reaction ID: EX_HC02154_e, Reaction Name: GM4-pool exchange\n",
      "Reaction ID: EX_HC02160_e, Reaction Name: Exchange of GM2-pool\n",
      "Reaction ID: EX_HC02161_e, Reaction Name: Exchange of GM1-pool\n",
      "Reaction ID: EX_HC02172_e, Reaction Name: Zinc exchange\n",
      "Reaction ID: EX_HC02180_e, Reaction Name: thromboxane-b2 exchange\n",
      "Reaction ID: EX_HC02187_e, Reaction Name: reverse-triiodthyronine exchange\n",
      "Reaction ID: EX_HC02191_e, Reaction Name: lithocholate exchange\n",
      "Reaction ID: EX_HC02192_e, Reaction Name: taurolithocholate exchange\n",
      "Reaction ID: EX_HC02193_e, Reaction Name: glycolithocholate exchange\n",
      "Reaction ID: EX_HC02194_e, Reaction Name: ursodeoxycholate exchange\n",
      "Reaction ID: EX_HC02195_e, Reaction Name: tauroursodeoxycholate exchange\n",
      "Reaction ID: EX_HC02196_e, Reaction Name: glycoursodeoxycholate exchange\n",
      "Reaction ID: EX_HC02197_e, Reaction Name: sulfoglycolithocholate(2-) exchange\n",
      "Reaction ID: EX_HC02198_e, Reaction Name: 2-[(4R)-4-[(1S,2S,5R,7R,10R,11S,14R,15R)-2,15-dimethyl-5-(sulfonatooxy)tetracyclo[8.7.0.0^{2,7}.0^{11,15}]heptadecan-14-yl]pentanamido]ethane-1-sulfonate exchange\n",
      "Reaction ID: EX_HC02199_e, Reaction Name: glutathionyl-leuc4 exchange\n",
      "Reaction ID: EX_HC02200_e, Reaction Name: S-glutathionyl-2-4-dinitrobenzene exchange\n",
      "Reaction ID: EX_HC02201_e, Reaction Name: S-glutathionyl-ethacrynic-acid exchange\n",
      "Reaction ID: EX_HC02202_e, Reaction Name: prostaglandin A1(1-) exchange\n",
      "Reaction ID: EX_HC02203_e, Reaction Name: prostaglandin-a2 exchange\n",
      "Reaction ID: EX_HC02204_e, Reaction Name: prostaglandin-b1 exchange\n",
      "Reaction ID: EX_HC02205_e, Reaction Name: prostaglandin-b2 exchange\n",
      "Reaction ID: EX_HC02206_e, Reaction Name: prostaglandin C1(1-) exchange\n",
      "Reaction ID: EX_HC02207_e, Reaction Name: prostaglandin-c2 exchange\n",
      "Reaction ID: EX_HC02208_e, Reaction Name: prostaglandin-d1 exchange\n",
      "Reaction ID: EX_HC02210_e, Reaction Name: prostaglandin-d3 exchange\n",
      "Reaction ID: EX_HC02213_e, Reaction Name: prostaglandin-e3 exchange\n",
      "Reaction ID: EX_HC02214_e, Reaction Name: prostaglandin-f1alpha exchange\n",
      "Reaction ID: EX_HC02216_e, Reaction Name: prostaglandin-f2beta exchange\n",
      "Reaction ID: EX_HC02217_e, Reaction Name: prostaglandin-g2 exchange\n",
      "Reaction ID: EX_HC02220_e, Reaction Name: sulfochenodeoxycholate exchange\n",
      "Reaction ID: EX_hco3_e, Reaction Name: Bicarbonate exchange\n",
      "Reaction ID: EX_hcoumarin_e, Reaction Name: hydroxy coumarin exchange\n",
      "Reaction ID: EX_hdca_e, Reaction Name: Hexadecanoate (n-C16:0) exchange\n",
      "Reaction ID: EX_hdcea_e, Reaction Name: exchange reaction for Hexadecenoate (n-C16:1)\n",
      "Reaction ID: EX_hestratriol_e, Reaction Name: 4,17 dihydroxy estradiol exchange\n",
      "Reaction ID: EX_hexc_e, Reaction Name: hexacosanoate (n-C26:0) exchange\n",
      "Reaction ID: EX_his_L_e, Reaction Name: exchange reaction for L-histidine\n",
      "Reaction ID: EX_hista_e, Reaction Name: Histamine exchange\n",
      "Reaction ID: EX_hom_L_e, Reaction Name: L-Homoserine exchange\n",
      "Reaction ID: EX_hpdca_e, Reaction Name: heptadecanoate exchange\n",
      "Reaction ID: EX_chopg_e, Reaction Name: heparan sulfate proteoglycan exchange\n",
      "Reaction ID: EX_htaxol_e, Reaction Name: hydroxylated taxol exchange\n",
      "Reaction ID: EX_hxan_e, Reaction Name: Hypoxanthine exchange\n",
      "Reaction ID: EX_hyptaur_e, Reaction Name: Exchange of hypotaurine\n",
      "Reaction ID: EX_i_e, Reaction Name: Iodide exchange\n",
      "Reaction ID: EX_idour_e, Reaction Name: Exchange of L-iduronate\n",
      "Reaction ID: EX_idp_e, Reaction Name: IDP exchange\n",
      "Reaction ID: EX_ile_L_e, Reaction Name: L-Isoleucine exchange\n",
      "Reaction ID: EX_imp_e, Reaction Name: IMP exchange\n",
      "Reaction ID: EX_inost_e, Reaction Name: myo-Inositol exchange\n",
      "Reaction ID: EX_ins_e, Reaction Name: Inosine exchange\n",
      "Reaction ID: EX_k_e, Reaction Name: K+ exchange\n",
      "Reaction ID: EX_ksi_deg1_e, Reaction Name: keratan sulfate I, degradation product 1 exchange\n",
      "Reaction ID: EX_ksi_e, Reaction Name: keratan sulfate I exchange\n",
      "Reaction ID: EX_ksii_core2_e, Reaction Name: keratan sulfate II (core 2-linked) exchange\n",
      "Reaction ID: EX_ksii_core4_e, Reaction Name: keratan sulfate II (core 4-linked) exchange\n",
      "Reaction ID: EX_lac_D_e, Reaction Name: D-lactate exchange\n",
      "Reaction ID: EX_lac_L_e, Reaction Name: L-Lactate exchange\n",
      "Reaction ID: EX_lcts_e, Reaction Name: Lactose exchange\n",
      "Reaction ID: EX_Lcystin_e, Reaction Name: L-Cystine exchange\n",
      "Reaction ID: EX_leu_L_e, Reaction Name: L-Leucine exchange\n",
      "Reaction ID: EX_leugly_e, Reaction Name: exchange reaction for leucylglycine\n",
      "Reaction ID: EX_leuktrA4_e, Reaction Name: leukotriene A4 exchange\n",
      "Reaction ID: EX_leuktrB4_e, Reaction Name: leukotriene B4 exchange\n",
      "Reaction ID: EX_leuktrC4_e, Reaction Name: Leukotriene C4 exchange\n",
      "Reaction ID: EX_leuktrD4_e, Reaction Name: leukotriene D4 exchange\n",
      "Reaction ID: EX_leuktrE4_e, Reaction Name: leukotriene E4 exchange\n",
      "Reaction ID: EX_leuktrF4_e, Reaction Name: leukotriene F4 exchange\n",
      "Reaction ID: EX_leuleu_e, Reaction Name: exchange reaction for leucylleucine\n",
      "Reaction ID: EX_lgnc_e, Reaction Name: lignoceric acid exchange\n",
      "Reaction ID: EX_limnen_e, Reaction Name: Limonene exchange\n",
      "Reaction ID: EX_lipoate_e, Reaction Name: Lipoate exchange\n",
      "Reaction ID: EX_lneldc_e, Reaction Name: linoelaidic acid exchange\n",
      "Reaction ID: EX_lnlc_e, Reaction Name: linoleic acid (all cis C18:2) exchange\n",
      "Reaction ID: EX_lnlnca_e, Reaction Name: alpha-Linolenic acid exchange\n",
      "Reaction ID: EX_lnlncg_e, Reaction Name: gamma-linolenic acid exchange\n",
      "Reaction ID: EX_lpchol_cho_e, Reaction Name: exchange reaction for lysophosphatidylcholine\n",
      "Reaction ID: EX_lys_L_e, Reaction Name: L-Lysine exchange\n",
      "Reaction ID: EX_mag_cho_e, Reaction Name: Monoacylglycerol 2 (cho) exchange\n",
      "Reaction ID: EX_mal_L_e, Reaction Name: Exchange of L-malate\n",
      "Reaction ID: EX_malcoa_e, Reaction Name: malonyl-CoA(4-) exchange\n",
      "Reaction ID: EX_malt_e, Reaction Name: Maltose exchange\n",
      "Reaction ID: EX_malthp_e, Reaction Name: maltoheptaose exchange\n",
      "Reaction ID: EX_malthx_e, Reaction Name: Maltohexaose exchange\n",
      "Reaction ID: EX_maltpt_e, Reaction Name: maltopentaose exchange\n",
      "Reaction ID: EX_malttr_e, Reaction Name: Maltotriose exchange\n",
      "Reaction ID: EX_maltttr_e, Reaction Name: Maltotetraose exchange\n",
      "Reaction ID: EX_man_e, Reaction Name: exchange reaction for D-Mannose\n",
      "Reaction ID: EX_meoh_e, Reaction Name: methanol exchange\n",
      "Reaction ID: EX_mepi_e, Reaction Name: Metanephrine exchange\n",
      "Reaction ID: EX_mercplaccys_e, Reaction Name: 3-mercaptolactate-cysteine disulfide exchange\n",
      "Reaction ID: EX_met_L_e, Reaction Name: L-Methionine exchange\n",
      "Reaction ID: EX_mthgxl_e, Reaction Name: Methylglyoxal exchange\n",
      "Reaction ID: EX_n2m2nmasn_e, Reaction Name: N-Acetyl-beta-D-glucosaminyl-1,2-alpha-D-mannosyl-1,3-(N-acetyl-beta-D-glucosaminyl-1,2-alpha-D-mannosyl-1,6)-(N-acetyl-beta-D-glucosaminyl-1,4)-beta-D-mannosyl-1,4-N-acetyl-beta-D-glucosaminyl-R exchange\n",
      "Reaction ID: EX_na1_e, Reaction Name: exchange reaction for Sodium\n",
      "Reaction ID: EX_nac_e, Reaction Name: Nicotinate exchange\n",
      "Reaction ID: EX_nad_e, Reaction Name: Nicotinamide adenine dinucleotide exchange\n",
      "Reaction ID: EX_nadp_e, Reaction Name: Nicotinamide adenine dinucleotide phosphate exchange\n",
      "Reaction ID: EX_ncam_e, Reaction Name: Nicotinamide exchange\n",
      "Reaction ID: EX_nh4_e, Reaction Name: Ammonia exchange\n",
      "Reaction ID: EX_nifedipine_e, Reaction Name: nifedipine exchange\n",
      "Reaction ID: EX_no_e, Reaction Name: exchange reaction for NO\n",
      "Reaction ID: EX_no2_e, Reaction Name: Nitrite exchange\n",
      "Reaction ID: EX_npthl_e, Reaction Name: Naphthalene exchange\n",
      "Reaction ID: EX_nrpphr_e, Reaction Name: Norepinephrine exchange\n",
      "Reaction ID: EX_nrpphrsf_e, Reaction Name: Sulfate derivative of norepinephrine exchange\n",
      "Reaction ID: EX_nrvnc_e, Reaction Name: nervonic acid exchange\n",
      "Reaction ID: EX_o2_e, Reaction Name: exchange reaction for oxugen\n",
      "Reaction ID: EX_o2s_e, Reaction Name: Superoxide anion exchange\n",
      "Reaction ID: EX_oagd3_cho_e, Reaction Name: 9-O-Acetylated GD3 (cho) exchange\n",
      "Reaction ID: EX_oagt3_cho_e, Reaction Name: 9-O-Acetylated GT3 (cho) exchange\n",
      "Reaction ID: EX_ocdca_e, Reaction Name: echange reaction for octadecanoate (n-C18:0)\n",
      "Reaction ID: EX_ocdcea_e, Reaction Name: octadecenoate (n-C18:1) exchange\n",
      "Reaction ID: EX_octa_e, Reaction Name: exchange reaction for octanoate (n-C8:0)\n",
      "Reaction ID: EX_oh1_e, Reaction Name: exchange reaction for hydroxide ion\n",
      "Reaction ID: EX_omeprazole_e, Reaction Name: omeprazole exchange\n",
      "Reaction ID: EX_onpthl_e, Reaction Name: naphthalene epoxide exchange\n",
      "Reaction ID: EX_orn_e, Reaction Name: Ornithine exchange\n",
      "Reaction ID: EX_orot_e, Reaction Name: Exchange of orotate\n",
      "Reaction ID: EX_oxa_e, Reaction Name: Oxalate exchange\n",
      "Reaction ID: EX_paf_cho_e, Reaction Name: 1-alkyl 2-acteylglycerol 3-phosphocholine (cho) exchange\n",
      "Reaction ID: EX_pan4p_e, Reaction Name: exchange reaction for pan4p\n",
      "Reaction ID: EX_pchol_cho_e, Reaction Name: Phosphatidylcholine (cho) exchange\n",
      "Reaction ID: EX_pe_cho_e, Reaction Name: exchange reaction for phosphatidylethanolamine\n",
      "Reaction ID: EX_pect_e, Reaction Name: exchange reaction for pectins\n",
      "Reaction ID: EX_pectindchac_e, Reaction Name: exchange reaction for pectin-deoxycholic acid complex\n",
      "Reaction ID: EX_pectingchol_e, Reaction Name: exchange reaction for pectin-glycocholate complex\n",
      "Reaction ID: EX_pectintchol_e, Reaction Name: exchange reaction for pectin-taurocholic acid complex\n",
      "Reaction ID: EX_peplys_e, Reaction Name: Peptidyl-L-lysine exchange\n",
      "Reaction ID: EX_perillyl_e, Reaction Name: Perillyl alcohol exchange\n",
      "Reaction ID: EX_pglyc_cho_e, Reaction Name: phosphatidylglycerol (cho) exchange\n",
      "Reaction ID: EX_phe_L_e, Reaction Name: exchange reaction for L-phenylalanine\n",
      "Reaction ID: EX_pheacgln_e, Reaction Name: alpha-N-Phenylacetyl-L-glutamine exchange\n",
      "Reaction ID: EX_pheme_e, Reaction Name: exchange reaction for heme\n",
      "Reaction ID: EX_phyQ_e, Reaction Name: Phylloquinone exchange\n",
      "Reaction ID: EX_phyt_e, Reaction Name: phytanic acid exchange\n",
      "Reaction ID: EX_pi_e, Reaction Name: Phosphate exchange\n",
      "Reaction ID: EX_pnto_R_e, Reaction Name: (R)-Pantothenate exchange\n",
      "Reaction ID: EX_ppa_e, Reaction Name: Propionate exchange\n",
      "Reaction ID: EX_ppi_e, Reaction Name: exchange reaction for ppi\n",
      "Reaction ID: EX_prgstrn_e, Reaction Name: Progesterone exchange\n",
      "Reaction ID: EX_pro_D_e, Reaction Name: D-Proline exchange\n",
      "Reaction ID: EX_pro_L_e, Reaction Name: L-Proline exchange\n",
      "Reaction ID: EX_progly_e, Reaction Name: exchange reaction for prolyl-glycine\n",
      "Reaction ID: EX_prostgd2_e, Reaction Name: Prostaglandin D2 exchange\n",
      "Reaction ID: EX_prostge1_e, Reaction Name: Prostaglandin E1 exchange\n",
      "Reaction ID: EX_prostge2_e, Reaction Name: Prostaglandin E2 exchange\n",
      "Reaction ID: EX_prostgf2_e, Reaction Name: Prostaglandin F2alpha exchange\n",
      "Reaction ID: EX_prostgh2_e, Reaction Name: prostaglandin H2(1-) exchange\n",
      "Reaction ID: EX_prostgi2_e, Reaction Name: prostaglandin I2(1-) exchange\n",
      "Reaction ID: EX_prpp_e, Reaction Name: Exchange of 5-O-phosphonato-alpha-D-ribofuranosyl diphosphate(5-)\n",
      "Reaction ID: EX_ps_cho_e, Reaction Name: phosphatidylserine (cho) exchange\n",
      "Reaction ID: EX_psyl_e, Reaction Name: exchange reaction for psyllium\n",
      "Reaction ID: EX_psylchol_e, Reaction Name: exchange reaction for psyllium-glycocholic acid complex\n",
      "Reaction ID: EX_psyltchol_e, Reaction Name: exchange reaction for psyllium-taurocholic acid complex\n",
      "Reaction ID: EX_psyltdechol_e, Reaction Name: exchange reaction for psyllium-taurodeoxycholic acid complex\n",
      "Reaction ID: EX_ptdca_e, Reaction Name: pentadecanoate exchange\n",
      "Reaction ID: EX_ptrc_e, Reaction Name: Exchange of 1,4-butanediammonium\n",
      "Reaction ID: EX_ptth_e, Reaction Name: exchange reaction for ptth\n",
      "Reaction ID: EX_pydam_e, Reaction Name: exchange reaction for Pyridoxamine\n",
      "Reaction ID: EX_pydx_e, Reaction Name: exchange reaction for Pyridoxal\n",
      "Reaction ID: EX_pydx5p_e, Reaction Name: Exchange of pyridoxal 5-phosphate(2-)\n",
      "Reaction ID: EX_pydxn_e, Reaction Name: exchange reaction for Pyridoxine\n",
      "Reaction ID: EX_pyr_e, Reaction Name: Pyruvate exchange\n",
      "Reaction ID: EX_q10_e, Reaction Name: exchange reaction for ubiquinone\n",
      "Reaction ID: EX_q10h2_e, Reaction Name: exchange reaction for ubiquinol\n",
      "Reaction ID: EX_rbt_e, Reaction Name: Ribitol exchange\n",
      "Reaction ID: EX_retfa_e, Reaction Name: fatty acid retinol exchange\n",
      "Reaction ID: EX_retinol_9_cis_e, Reaction Name: 9-cis-retinol exchange\n",
      "Reaction ID: EX_retinol_cis_11_e, Reaction Name: cis-11-retinol exchange\n",
      "Reaction ID: EX_retinol_e, Reaction Name: Retinol exchange\n",
      "Reaction ID: EX_retn_e, Reaction Name: Retinoate exchange\n",
      "Reaction ID: EX_retnglc_e, Reaction Name: retinoyl glucuronide exchange\n",
      "Reaction ID: EX_rib_D_e, Reaction Name: exchange reaction for ribose\n",
      "Reaction ID: EX_ribflv_e, Reaction Name: exchange reaction for Riboflavin\n",
      "Reaction ID: EX_Rtotal_e, Reaction Name: R total exchange\n",
      "Reaction ID: EX_Rtotal2_e, Reaction Name: R total 2 position exchange\n",
      "Reaction ID: EX_Rtotal3_e, Reaction Name: R total 3 position exchange\n",
      "Reaction ID: EX_s2l2fn2m2masn_e, Reaction Name: PA6 exchange\n",
      "Reaction ID: EX_s2l2n2m2masn_e, Reaction Name: de-Fuc form of PA6 exchange\n",
      "Reaction ID: EX_sarcs_e, Reaction Name: Sarcosine exchange\n",
      "Reaction ID: EX_sbt_d_e, Reaction Name: exchange reaction for D-Sorbitol\n",
      "Reaction ID: EX_sel_e, Reaction Name: Selenate exchange\n",
      "Reaction ID: EX_ser_D_e, Reaction Name: D-Serine exchange\n",
      "Reaction ID: EX_ser_L_e, Reaction Name: exchange reaction for L-serine\n",
      "Reaction ID: EX_sfcys_e, Reaction Name: exchange reaction for sulfocysteine\n",
      "Reaction ID: EX_sl_L_e, Reaction Name: L-sulfolactate exchange\n",
      "Reaction ID: EX_so4_e, Reaction Name: Sulfate exchange\n",
      "Reaction ID: EX_spc_cho_e, Reaction Name: sphingosylphosphorylcholine (cho) exchange\n",
      "Reaction ID: EX_sph1p_e, Reaction Name: Sphinganine 1-phosphate exchange\n",
      "Reaction ID: EX_sphs1p_e, Reaction Name: Sphingosine 1-phosphate exchange\n",
      "Reaction ID: EX_spmd_e, Reaction Name: Exchange of spermidine(3+)\n",
      "Reaction ID: EX_sprm_e, Reaction Name: EX sprm(e)\n",
      "Reaction ID: EX_srtn_e, Reaction Name: Serotonin exchange\n",
      "Reaction ID: EX_strch1_e, Reaction Name: starch, structure 1 (1,6-{7[1,4-Glc], 4[1,4-Glc]}) exchange\n",
      "Reaction ID: EX_strch2_e, Reaction Name: exchange reaction for starch, structure 2 (1,6-{2[1,4-Glc], [1,4-Glc]})\n",
      "Reaction ID: EX_strdnc_e, Reaction Name: stearidonic acid exchange\n",
      "Reaction ID: EX_succ_e, Reaction Name: Succinate exchange\n",
      "Reaction ID: EX_sucr_e, Reaction Name: Sucrose exchange\n",
      "Reaction ID: EX_tag_cho_e, Reaction Name: triacylglycerol (cho) exchange\n",
      "Reaction ID: EX_tagat_D_e, Reaction Name: D-Tagatose exchange\n",
      "Reaction ID: EX_taur_e, Reaction Name: Taurine exchange\n",
      "Reaction ID: EX_taxol_e, Reaction Name: paclitaxel exchange\n",
      "Reaction ID: EX_tchola_e, Reaction Name: Taurocholic acid exchange\n",
      "Reaction ID: EX_tcynt_e, Reaction Name: Thiocyanate exchange\n",
      "Reaction ID: EX_tdchola_e, Reaction Name: taurochenodeoxycholate exchange\n",
      "Reaction ID: EX_tdechola_e, Reaction Name: exchange reaction for Taurodeoxycholic acid\n",
      "Reaction ID: EX_tethex3_e, Reaction Name: tetracosahexaenoic acid, n-3 exchange\n",
      "Reaction ID: EX_tetpent3_e, Reaction Name: tetracosapentaenoic acid, n-3 exchange\n",
      "Reaction ID: EX_tetpent6_e, Reaction Name: tetracosapentaenoic acid, n-6 exchange\n",
      "Reaction ID: EX_tettet6_e, Reaction Name: tetracosatetraenoic acid n-6 exchange\n",
      "Reaction ID: EX_thf_e, Reaction Name: 5,6,7,8-Tetrahydrofolate exchange\n",
      "Reaction ID: EX_thm_e, Reaction Name: exchange reaction for Thiamin\n",
      "Reaction ID: EX_thmmp_e, Reaction Name: Thiamin monophosphate exchange\n",
      "Reaction ID: EX_thmtp_e, Reaction Name: Thiamin triphosphate exchange\n",
      "Reaction ID: EX_thr_L_e, Reaction Name: L-Threonine exchange\n",
      "Reaction ID: EX_thym_e, Reaction Name: Thymine exchange\n",
      "Reaction ID: EX_thymd_e, Reaction Name: Thymidine exchange\n",
      "Reaction ID: EX_thyox_L_e, Reaction Name: L-Thyroxine exchange\n",
      "Reaction ID: EX_tmndnc_e, Reaction Name: timnodonic acid exchange\n",
      "Reaction ID: EX_tolbutamide_e, Reaction Name: tolbutamide exchange\n",
      "Reaction ID: EX_tre_e, Reaction Name: Trehalose exchange\n",
      "Reaction ID: EX_triodthy_e, Reaction Name: Triiodothyronine exchange\n",
      "Reaction ID: EX_triodthysuf_e, Reaction Name: Triiodothyronine sulfate exchange\n",
      "Reaction ID: EX_trp_L_e, Reaction Name: L-Tryptophan exchange\n",
      "Reaction ID: EX_tststerone_e, Reaction Name: Testosterone exchange\n",
      "Reaction ID: EX_tststeroneglc_e, Reaction Name: testosterone glucuronide exchange\n",
      "Reaction ID: EX_tststerones_e, Reaction Name: testosterone sulfate exchange\n",
      "Reaction ID: EX_tsul_e, Reaction Name: Thiosulfate exchange\n",
      "Reaction ID: EX_ttdca_e, Reaction Name: exchange reaction for tetradecanoate (n-C14:0)\n",
      "Reaction ID: EX_txa2_e, Reaction Name: Thromboxane A2 exchange\n",
      "Reaction ID: EX_tymsf_e, Reaction Name: Tyramine O-sulfate exchange\n",
      "Reaction ID: EX_Tyr_ggn_e, Reaction Name: Tyr-194 of apo-glycogenin protein (primer for glycogen synthesis) exchange\n",
      "Reaction ID: EX_tyr_L_e, Reaction Name: L-Tyrosine exchange\n",
      "Reaction ID: EX_udp_e, Reaction Name: UDP exchange\n",
      "Reaction ID: EX_udpg_e, Reaction Name: Exchange of [(2R,3S,4R,5R)-5-(2,4-dioxo-1,2,3,4-tetrahydropyrimidin-1-yl)-3,4-dihydroxyoxolan-2-yl]methyl {[(3R,4S,5S,6R)-3,4,5-trihydroxy-6-(hydroxymethyl)oxan-2-yl phosphonato]oxy}phosphonate\n",
      "Reaction ID: EX_udpgal_e, Reaction Name: UDP-D-galactose(2-) exchange\n",
      "Reaction ID: EX_ump_e, Reaction Name: UMP exchange\n",
      "Reaction ID: EX_ura_e, Reaction Name: Uracil exchange\n",
      "Reaction ID: EX_urate_e, Reaction Name: Urate exchange\n",
      "Reaction ID: EX_urea_e, Reaction Name: Urea exchange\n",
      "Reaction ID: EX_uri_e, Reaction Name: exchange reaction for Uridine\n",
      "Reaction ID: EX_utp_e, Reaction Name: UTP exchange\n",
      "Reaction ID: EX_vacc_e, Reaction Name: vaccenic acid exchange\n",
      "Reaction ID: EX_val_L_e, Reaction Name: L-Valine exchange\n",
      "Reaction ID: EX_vitd2_e, Reaction Name: Vitamin D2 exchange\n",
      "Reaction ID: EX_vitd3_e, Reaction Name: exchange reaction for calctriol\n",
      "Reaction ID: EX_whddca_e, Reaction Name: omega hydroxy dodecanoate (n-C12:0) exchange\n",
      "Reaction ID: EX_whhdca_e, Reaction Name: omega hydroxy hexadecanoate (n-C16:0) exchange\n",
      "Reaction ID: EX_whtststerone_e, Reaction Name: w hydroxy testosterone exchange\n",
      "Reaction ID: EX_whttdca_e, Reaction Name: omega hydroxy tetradecanoate (n-C14:0) exchange\n",
      "Reaction ID: EX_xolest_cho_e, Reaction Name: cholesterol ester exchange\n",
      "Reaction ID: EX_xolest2_cho_e, Reaction Name: cholesterol ester (from FULLR2) exchange\n",
      "Reaction ID: EX_xoltri24_e, Reaction Name: 7-alpha,24(S)-Dihydroxycholesterol exchange\n",
      "Reaction ID: EX_xoltri25_e, Reaction Name: 7-alpha,25-Dihydroxycholesterol exchange\n",
      "Reaction ID: EX_xoltri27_e, Reaction Name: 7-alpha,27-Dihydroxycholesterol exchange\n",
      "Reaction ID: EX_xyl_D_e, Reaction Name: D-Xylose exchange\n",
      "Reaction ID: EX_xylt_e, Reaction Name: exchange reaction for xylitol\n",
      "Reaction ID: EX_yvite_e, Reaction Name: gamma-Tocopherol exchange\n",
      "Reaction ID: AMPPPHRTPm_e, Reaction Name: AMP:pyrophosphate phosphoribosyltransferase Purine metabolism EC:2.4.2.7\n",
      "Reaction ID: UDPGLPPM_e, Reaction Name: UDPglucose pyrophosphohydrolase Pyrimidine metabolism / Starch and sucrose metabolism EC:3.6.1.45\n",
      "Reaction ID: GAMGLUGm2_e, Reaction Name: gamma-glutamyltranspeptidase Glutathione metabolism EC:3.4.11.4\n",
      "Reaction ID: CDPDIPPM_e, Reaction Name: CDP diphosphohydrolase Pyrimidine metabolism EC:3.6.1.5\n",
      "Reaction ID: CTPDIPPm_e, Reaction Name: CTP diphosphohydrolase Pyrimidine metabolism EC:3.6.1.5\n",
      "Reaction ID: URITRIPYRPm_e, Reaction Name: Uridine triphosphate pyrophosphohydrolase Pyrimidine metabolism EC:3.6.1.19\n",
      "Reaction ID: GLYCPP_e, Reaction Name: Glycerone phosphate phosphohydrolase (alkaline optimum) EC:3.1.3.1\n",
      "Reaction ID: PUTOOU_e, Reaction Name: Putrescine:oxygen oxidoreductase (deaminating) Urea cycle and metabolism of amino groups EC:1.4.3.6\n",
      "Reaction ID: LCTSGTYGm_e, Reaction Name: Lactose 6-phosphate galactohydrolase Galactose metabolism EC:3.2.1.23\n",
      "Reaction ID: ISOLSDGLUm_e, Reaction Name: Isomaltose 6-alpha-D-glucanohydrolase Starch and sucrose metabolism EC:3.2.1.10\n",
      "Reaction ID: DGTPdGMPm_e, Reaction Name: 2-Deoxyguanosine 5-triphosphate pyrophosphohydrolase Purine metabolism EC:3.6.1.19\n",
      "Reaction ID: DTDPDTMPm_e, Reaction Name: dTDP diphosphohydrolase Pyrimidine metabolism EC:3.6.1.5\n",
      "Reaction ID: NTP9Pm_e, Reaction Name: dTTP nucleotidohydrolase Pyrimidine metabolism EC:3.6.1.39\n",
      "Reaction ID: 4NIPPGHD_e, Reaction Name: 4-Nitrophenyl phosphate phosphohydrolase gamma-Hexachlorocyclohexane degradation EC:3.1.3.2 EC:3.1.3.41 EC:3.1.3.1\n",
      "Reaction ID: GALGALGLYCm_e, Reaction Name: Galactosylglycerol galactohydrolase Galactose metabolism EC:3.2.1.22 EC:3.2.1.23\n",
      "Reaction ID: 5GLTPEGLTCYA1_e, Reaction Name: (5-Glutamyl)-peptide:amino-acid 5-glutamyltransferase Cyanoamino acid metabolism EC:2.3.2.2\n",
      "Reaction ID: 5GLTPEGLTCYA2_e, Reaction Name: (5-Glutamyl)-peptide:amino-acid 5-glutamyltransferase Cyanoamino acid metabolism EC:2.3.2.2\n",
      "Reaction ID: AHDTDIPTRIPFB_e, Reaction Name: 2-Amino-4-hydroxy-6-(erythro-1,2,3-trihydroxypropyl) dihydropteridine triphosphate phosphohydrolase (alkaline optimum) Folate biosynthesis EC:3.1.3.1\n",
      "Reaction ID: 2CHTNH2O_e, Reaction Name: EC:3.2.1.14\n",
      "Reaction ID: GM1-P_GM2-P_e, Reaction Name: D-Galactosyl-N-acetyl-D-galactosaminyl-(N-acetylneuraminyl)-D- galactosyl-D-glucosylceramide galactohydrolase EC:3.2.1.23\n",
      "Reaction ID: gthox_export, Reaction Name: gthox export\n",
      "Reaction ID: DM_ethamp[r], Reaction Name: DM_ethamp[r]\n"
     ]
    }
   ],
   "source": [
    "for reaction in iCHO2101_changed.reactions:\n",
    "    if \"_e\"in reaction.id:\n",
    "        print(f\"Reaction ID: {reaction.id}, Reaction Name: {reaction.name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpkkuukbas.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Optimized objective (growth rate): 0.08204526758059716\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpu2i17l35.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp3ezb6hry.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpnyjtrs6x.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp_ywzblk6.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpfs254g27.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmphpjthc0m.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp3iwmkgw0.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100753319             Eif3i              Eif3i\n",
      "1        100689093              Pisd               Pisd\n",
      "2        100750895              Atic               Atic\n",
      "3        100752118             Eif3a              Eif3a\n",
      "4        100758991             Paics              Paics\n",
      "5        100751998             Eif3h              Eif3h\n",
      "6        100760149              Ppat               Ppat\n",
      "7        100760956              Pfas               Pfas\n",
      "8        100689232               Cad                Cad\n",
      "9        100689326            Sptlc1             Sptlc1\n",
      "10       100756632             Dhodh              Dhodh\n",
      "11       100756786              Umps               Umps\n",
      "12       100761591              Rrm1               Rrm1\n",
      "13       100758493             Pola1              Pola1\n",
      "14       100762815             Pold2              Pold2\n",
      "15       100774890             Rev3l              Rev3l\n",
      "16       100751096              Poll               Poll\n",
      "17       100773935             Pold1              Pold1\n",
      "18       100753174              Rfc5               Rfc5\n",
      "19       100756090             Pold3              Pold3\n",
      "20       100774671              Pole               Pole\n",
      "21       100768307             Pola2              Pola2\n",
      "22       100763640            Ptpmt1             Ptpmt1\n",
      "23       100763004             Eif3g              Eif3g\n",
      "24       100689060              Adsl               Adsl\n",
      "25       100753359             Eif3l              Eif3l\n",
      "26       100769028             Eif3f              Eif3f\n",
      "27       100774578              Kdsr               Kdsr\n",
      "28       100689310            Eef1e1             Eef1e1\n",
      "29       100689051              Eef2               Eef2\n",
      "30       100762904             Acaca              Acaca\n",
      "31       100771190             Eif3e              Eif3e\n",
      "32       100753545             Eef1g              Eef1g\n",
      "33       100766600             Eif3k              Eif3k\n",
      "34       100762979             Eif3b              Eif3b\n",
      "35       100768213             Eif3m              Eif3m\n",
      "36       100756126              Gmps               Gmps\n",
      "37       100767580           Hsd17b7            Hsd17b7\n",
      "38       100756809              Agps               Agps\n",
      "39       100766541               Mvd                Mvd\n",
      "40       100765639               Ebp                Ebp\n",
      "41       100773070             Cept1              Cept1\n",
      "42       100756363             Hmgcr              Hmgcr\n",
      "43       100689449              Pgs1               Pgs1\n",
      "44       100753498               Lss                Lss\n",
      "45       100762189      LOC100762189       LOC100762189\n",
      "46       100751653               Mvk                Mvk\n",
      "47       100759992              Pmvk               Pmvk\n",
      "48       100689192             Fdft1              Fdft1\n",
      "49       100764834              Sqle               Sqle\n",
      "50       100765024             Gnpat              Gnpat\n",
      "51       100689446              Gart               Gart\n",
      "Total matched metabolic genes: 367\n",
      "Matched essential genes: 52/367\n"
     ]
    }
   ],
   "source": [
    "model = load_json_model('iCHO3595_unblocked.json')\n",
    "GM_constrained, _, _ = constraint_model_feg(model, time='P2 to P4', exp_dict=temp_dict_zela)\n",
    "matched_df,len_3k,len_3k_ess = find_and_match_essential_genes(GM_constrained, df_loaded, save_csv=False, print_df=True, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp6h98g4d2.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Optimized objective (growth rate): 215.53766987183332\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpp9_zvm0v.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp39b0g_q4.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp2euazrii.lp\n",
      "Reading time = 0.11 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp80mx4uyf.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp3t0dsyhn.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpq4yjrfyj.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmppx40hkf_.lp\n",
      "Reading time = 0.17 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100753545             Eef1g              Eef1g\n",
      "1        100766600             Eif3k              Eif3k\n",
      "2        100762979             Eif3b              Eif3b\n",
      "3        100768213             Eif3m              Eif3m\n",
      "4        100753319             Eif3i              Eif3i\n",
      "5        100689093              Pisd               Pisd\n",
      "6        100752118             Eif3a              Eif3a\n",
      "7        100751998             Eif3h              Eif3h\n",
      "8        100756126              Gmps               Gmps\n",
      "9        100756809              Agps               Agps\n",
      "10       100773070             Cept1              Cept1\n",
      "11       100689449              Pgs1               Pgs1\n",
      "12       100761591              Rrm1               Rrm1\n",
      "13       100758493             Pola1              Pola1\n",
      "14       100762815             Pold2              Pold2\n",
      "15       100774890             Rev3l              Rev3l\n",
      "16       100751096              Poll               Poll\n",
      "17       100773935             Pold1              Pold1\n",
      "18       100765024             Gnpat              Gnpat\n",
      "19       100753174              Rfc5               Rfc5\n",
      "20       100756090             Pold3              Pold3\n",
      "21       100774671              Pole               Pole\n",
      "22       100768307             Pola2              Pola2\n",
      "23       100763640            Ptpmt1             Ptpmt1\n",
      "24       100763004             Eif3g              Eif3g\n",
      "25       100689060              Adsl               Adsl\n",
      "26       100753359             Eif3l              Eif3l\n",
      "27       100769028             Eif3f              Eif3f\n",
      "28       100689310            Eef1e1             Eef1e1\n",
      "29       100689446              Gart               Gart\n",
      "30       100689051              Eef2               Eef2\n",
      "31       100771190             Eif3e              Eif3e\n",
      "32       100750895              Atic               Atic\n",
      "33       100758991             Paics              Paics\n",
      "34       100760149              Ppat               Ppat\n",
      "35       100760956              Pfas               Pfas\n",
      "36       100689232               Cad                Cad\n",
      "37       100689326            Sptlc1             Sptlc1\n",
      "38       100756632             Dhodh              Dhodh\n",
      "39       100767580           Hsd17b7            Hsd17b7\n",
      "40       100756786              Umps               Umps\n",
      "41       100766541               Mvd                Mvd\n",
      "42       100765639               Ebp                Ebp\n",
      "43       100756363             Hmgcr              Hmgcr\n",
      "44       100753498               Lss                Lss\n",
      "45       100762189      LOC100762189       LOC100762189\n",
      "46       100751653               Mvk                Mvk\n",
      "47       100759992              Pmvk               Pmvk\n",
      "48       100689192             Fdft1              Fdft1\n",
      "49       100764834              Sqle               Sqle\n",
      "50       100774578              Kdsr               Kdsr\n",
      "51       100762904             Acaca              Acaca\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpuu16201c.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp1hm_i9v7.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpkonnxaw_.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp2x3vt7p_.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp4paujz9z.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpgn6q2pq2.lp\n",
      "Reading time = 0.09 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpprl9kukc.lp\n",
      "Reading time = 0.11 seconds\n",
      ": 4764 rows, 17114 columns, 72396 nonzeros\n",
      "Total matched metabolic genes: 367\n",
      "Matched essential genes: 52/367\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "too many values to unpack (expected 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/ipykernel_36436/1439938275.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mload_json_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'iCHO3595_unblocked.json'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mGM_constrained\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGM_unconstrained\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconstraint_model_feg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'P2 to P4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexp_dict\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtemp_dict_zela\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmake_GM\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mmatched_df\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlen_3k\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlen_3k_ess\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfind_and_match_essential_genes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGM_constrained\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_loaded\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msave_csv\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprint_df\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodel_GM\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mGM_unconstrained\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m: too many values to unpack (expected 3)"
     ]
    }
   ],
   "source": [
    "model = load_json_model('iCHO3595_unblocked.json')\n",
    "GM_constrained, _, _, _, GM_unconstrained = constraint_model_feg(model, time='P2 to P4', exp_dict=temp_dict_zela, make_GM = True)\n",
    "matched_df,len_3k,len_3k_ess, len_3k_ess_GM = find_and_match_essential_genes(GM_constrained, df_loaded, save_csv=False, print_df=True, model_GM=GM_unconstrained)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmprlihb5nh.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Optimized objective (growth rate): 0.07008336960609673\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpenn264wd.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp0ghx2zrz.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpjuhc3mj1.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp85_yca33.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpzy5_8f36.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp_bubieut.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpoei0w3el.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100750801                                Cdipt\n",
      "1        100774578                                 Kdsr\n",
      "2        100754276                                 Tyms\n",
      "3        100767580                              Hsd17b7\n",
      "4        100765639                                  Ebp\n",
      "5        100689325                                 Ugp2\n",
      "6        100689446                                 Gart\n",
      "7        100689060                                 Adsl\n",
      "8        100689415                               Sptlc2\n",
      "9        100689326                               Sptlc1\n",
      "10       100750895                                 Atic\n",
      "11       100760956                                 Pfas\n",
      "12       100758991                                Paics\n",
      "13       100756786                                 Umps\n",
      "14       100752327                                Mat2a\n",
      "15       100762189                         LOC100762189\n",
      "16       100770439                                 Guk1\n",
      "17       100771700                                Dhcr7\n",
      "18       100756632                                Dhodh\n",
      "19       100753498                                  Lss\n",
      "20       100754883                                 Fdps\n",
      "21       100752960                                 Sc5d\n",
      "22       100689093                                 Pisd\n",
      "23       100760149                                 Ppat\n",
      "24       100771594                               Txnrd1\n",
      "25       100764834                                 Sqle\n",
      "26       100689192                                Fdft1\n",
      "27       100766541                                  Mvd\n",
      "28       100689232                                  Cad\n",
      "29       100761591                                 Rrm1\n",
      "30       100756363                                Hmgcr\n",
      "31       100751653                                  Mvk\n",
      "32       100759992                                 Pmvk\n",
      "33       100754784                                Msmo1\n",
      "34       100689449                                 Pgs1\n",
      "Total matched metabolic genes: 229\n",
      "Matched essential genes: 35/229\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_1766, _, _ = constraint_model_feg(iCHO1766_changed, time='P2 to P4', exp_dict=temp_dict_zela, objective = \"biomass_cho\")\n",
    "matched_df_1766, len_1766, len_1766_ess = find_and_match_essential_genes(GM_constrained_1766, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmplp0e500b.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Optimized objective (growth rate): 0.0\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpzwzhfl61.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpa6vg7225.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpdg3m7d70.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpbhmhvjj5.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmph3x4wp4q.lp\n",
      "Reading time = 0.09 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp0qxcvp3y.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpvn7j0uqr.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Empty DataFrame\n",
      "Columns: [gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol]\n",
      "Index: []\n",
      "Total matched metabolic genes: 254\n",
      "Matched essential genes: 0/254\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_2101, _, _ = constraint_model_feg(iCHO2101_changed, time='P2 to P4', exp_dict=temp_dict_zela, objective = \"biomass_cho\")\n",
    "matched_df_2101, len_2101, len_2101_ess = find_and_match_essential_genes(GM_constrained_2101, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpo39adjx_.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Optimized objective (growth rate): 0.0\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp2e9mqq8a.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpiibytreg.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmprlxv4jm4.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpdn_e82tb.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpl2ps8x7x.lp\n",
      "Reading time = 0.14 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmprnlvtq7r.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpko_29my1.lp\n",
      "Reading time = 0.12 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Empty DataFrame\n",
      "Columns: [gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol]\n",
      "Index: []\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmptust07z6.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpguct__n9.lp\n",
      "Reading time = 0.09 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp0a9ddpa2.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpup5noo65.lp\n",
      "Reading time = 0.11 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpim13fqdc.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpv4q0d3m8.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmprc2bvyzh.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Total matched metabolic genes_GM: 254\n",
      "Matched essential genes_GM: 0/254\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_2101, _, _, _, GM_unconstrained_2101 = constraint_model_feg(iCHO2101_changed, time='P2 to P4', exp_dict=temp_dict_zela, make_GM = True, objective = \"biomass_cho\")\n",
    "matched_df_2101,len_2101,len_2101_ess, len_2101_ess_GM = find_and_match_essential_genes(GM_constrained_2101, df_loaded, save_csv=False, print_df=True, model_GM=GM_unconstrained_2101)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpm054x9su.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Optimized objective (growth rate): 0.0\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmph98zgzht.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp97vy8m7n.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpe454_5a2.lp\n",
      "Reading time = 0.09 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpycbxvrwg.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpaythv_w0.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmp022ae9te.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-09-16\n",
      "Read LP format model from file /var/folders/8_/lxtrgg8n30b2czkwbgcxncmh0000gp/T/tmpcn3d0kjx.lp\n",
      "Reading time = 0.10 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "  gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0       100752019                                 Gclc\n",
      "1       100752327                                Mat2a\n",
      "Total matched metabolic genes: 340\n",
      "Matched essential genes: 2/340\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_2291, _, _ = constraint_model_feg(iCHO2291_changed, time='P2 to P4', exp_dict=temp_dict_zela, objective = \"biomass_cho\")\n",
    "matched_df_2291, len_2291, len_2291_ess = find_and_match_essential_genes(GM_constrained_2291, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(iCHO2101.genes)\n",
    "entrez_dict = df_loaded.set_index('entrez_gene_id')['gene'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_gene = []    \n",
    "for gene in iCHO2101.genes:\n",
    "    # Check if the gene.id matches any Entrez Gene ID in the reverse dictionary\n",
    "    if gene.id in entrez_dict:\n",
    "        entrez_gene_symbol = entrez_dict[gene.id]\n",
    "        tmp_gene.append((gene.id, gene.name, entrez_gene_symbol))\n",
    "print(len(tmp_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List to store essential gene IDs\n",
    "essential_gene_ids = [gene.id for gene in essential_genes_tmp_2101]\n",
    "\n",
    "# List to store matching essential genes\n",
    "matched_essential_genes = []\n",
    "\n",
    "for gene in tmp_gene:\n",
    "    gene_id = gene[0]  # The gene.id is the first element in each tuple in tmp_gene\n",
    "    if gene_id in essential_gene_ids:\n",
    "        matched_essential_genes.append(gene)\n",
    "\n",
    "# Print the matched essential genes\n",
    "print(\"Matched essential genes:\")\n",
    "for gene in matched_essential_genes:\n",
    "    print(f\"Gene ID: {gene[0]}, Gene Name: {gene[1]}, Entrez Gene Symbol: {gene[2]}\")\n",
    "len(matched_essential_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(matched_essential_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "matched_zela = []\n",
    "matched_wt = []\n",
    "\n",
    "objective = 'biomass_cho'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "time_points_dict = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "time_points_zela = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8'}\n",
    "time_points_wt = {'P2 to P4':'P4', 'P4 to P6':'P6'}\n",
    "\n",
    "for time_key in time_points_zela.keys():\n",
    "    print(f\"Processing time point: Zela {time_key}\")\n",
    "    #Zela\n",
    "    GM_constrained, pfba_obj, pfba_solution = constraint_model_feg(\n",
    "        iCHO2291_changed, \n",
    "        time=time_key, \n",
    "        exp_dict=temp_dict_zela,\n",
    "        objective = \"biomass_cho\"\n",
    "    )\n",
    "    # Find and match essential genes\n",
    "    matched_df_zela = find_and_match_essential_genes(GM_constrained, df_loaded, save_csv=False, print_df=True)\n",
    "    \n",
    "    # Append the matched DataFrame for Zela to the list\n",
    "    matched_zela.append(matched_df_zela)\n",
    "    \n",
    "for time_key in time_points_wt.keys():\n",
    "    print(f\"Processing time point: WT {time_key}\")\n",
    "    # WT\n",
    "    GM_constrained, pfba_obj, pfba_solution = constraint_model_feg(\n",
    "        iCHO2291_changed, \n",
    "        time=time_key, \n",
    "        exp_dict=temp_dict_wt,\n",
    "        objective = \"biomass_cho\"\n",
    "    )\n",
    "    # Find and match essential genes\n",
    "    matched_df_wt = find_and_match_essential_genes(GM_constrained, df_loaded, save_csv=False, print_df=True)\n",
    "    \n",
    "    # Append the matched DataFrame for WT to the list\n",
    "    matched_wt.append(matched_df_wt)\n",
    "\n",
    "# Concatenate matched DataFrames into one and remove duplicates\n",
    "if matched_zela:\n",
    "    combined_matched_zela = pd.concat(matched_zela, ignore_index=True).drop_duplicates()\n",
    "    print(\"Combined Matched Zela DataFrame:\")\n",
    "    print(combined_matched_zela)\n",
    "\n",
    "if matched_wt:\n",
    "    combined_matched_wt = pd.concat(matched_wt, ignore_index=True).drop_duplicates()\n",
    "    print(\"Combined Matched WT DataFrame:\")\n",
    "    print(combined_matched_wt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_zela"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine the two combined DataFrames and remove duplicates\n",
    "if matched_zela and matched_wt:\n",
    "    combined_all = pd.concat([combined_matched_zela, combined_matched_wt], ignore_index=True).drop_duplicates()\n",
    "    print(\"Combined All (WT + Zela) DataFrame without duplicates:\")\n",
    "    print(len(combined_all))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(combined_matched_zela)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(combined_matched_wt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# Data for the bar chart\n",
    "data = {\n",
    "    'iCHO1766': [len_1766, len_1766_ess],\n",
    "    'iCHO2101': [len_2101, len_2101_ess],\n",
    "    'iCHO2291': [len_2291, len_2291_ess],\n",
    "    'iCHO3595': [len_3k, len_3k_ess]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a DataFrame from the data\n",
    "counts_df = pd.DataFrame(data, index=['Matched Genes', 'Matched Essential Genes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA90AAAJOCAYAAACqS2TfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAACaWElEQVR4nOzdd3gU5d7G8XvTQ0hCCkkIBAy9hN6xEHoLRUCqSsejiCAgTcGgAgrSpIgiEqTrUUBUQJSigEgT6QqSIC1BAgktBAjz/sHLnGwKJJAlJH4/17WX7swzM7+Z7C577zzzjMUwDEMAAAAAACDL2WV3AQAAAAAA5FaEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbuARExUVJYvFYj42btyYbbU89thjZh3h4eHZVgdsKzQ01Pw7d+/ePbvLyTWy+rgm/1yIiIh44PUBthQeHm6+Xh977LHsLgdIV0REhNXn64PitY+0ELqRo23cuNHqg9Jisahjx45ptp07d26qtlkRJLt3726uLzQ09IHXl5ucOnVKY8aMUd26deXv7y9nZ2d5eXmpdOnSaty4sd577z3t378/u8vEfUjrvZfWI7eG+Jz4Q8Wff/6poUOHqlatWsqfP7+cnJzk6+urcuXKKSwsTB988IH++uuv7C4z2928eVOLFy9W+/bt9dhjj8nNzU3u7u4qU6aM2rZtq8WLF+vKlSvZXSay0bVr1+Tl5WX1Wffkk09md1mPrJQnEywWi2rWrJlm2x9//PFf8+8I/l0csrsAIKt99dVXOnXqlAoWLGg1fcaMGdlU0b+PYRgaP368wsPDdePGDat5169fV1xcnP744w+tW7dOw4cPl2EY2VQpYDsvvviiwsLCJEkhISHZVkdiYqKGDRumDz74INV7LTY2VrGxsTp48KC+/fZbTZ48WVFRUdlT6CNg79696tixow4fPpxq3uHDh3X48GEtX75cU6ZM0cCBAx9+gTlI48aNlTdvXkmSp6dnNleTtZYvX664uDiraZs3b9bRo0dVvHjx7Ckqh9m+fbu2b9+uGjVqWE3nuxpyK0I3cp2bN29q9uzZevvtt81pmzdv1p49e7KvqH+ZgQMH6oMPPjCfOzg4qGnTpqpcubJcXV119uxZ7d69W1u3btXNmzezsVJklY4dO6patWqppmdn2Mxu6fW6eZhu3bqlzp07a/ny5eY0V1dXhYWFqVy5cnJwcNCZM2e0Y8cO7dy5MxsrzX6HDh1S3bp1rcJU+fLl1bRpU3l5eSkmJkYbN27U77//nn1F5gCXLl2Su7u76tSpozp16mR3OTaR3uUd8+fPt/ruYQt3jm9uMH36dC1YsMB8fvz4ca1atSobKwJsyABysA0bNhiSzIednZ0hyfDz8zOuXbtmtuvQoYPV/DuPN99802p9Y8eONVq1amUUL17c8PLyMhwcHIx8+fIZNWrUMMaOHWtcvnzZbDtv3jyrdaX12LBhg9n+1q1bxtKlS40WLVoYAQEBhqOjo+Ht7W1Uq1bNGDp0qNkuMjIy1To+//xzo3r16oaLi4vh4+NjdOvWzYiNjU33mDzzzDNGoUKFDCcnJ8PDw8N44oknjE8++cRISkpKc5mPP/7YCAkJMZydnY2CBQsagwYNMi5evGgUKVIk3WOVnnXr1lnVX6xYMePw4cNptj1//rzx/vvvP/B+pHXMFixYYFStWtUmx+ynn34y2rRpYwQGBhqOjo6Gm5ubUaRIEaNp06bGm2++acTFxWXoWH3++edGly5djHLlyhn58+c311W2bFnj5ZdfNiIjI1MtU7duXXM/u3XrZhw+fNho37694eXlZbi4uBi1atWyet0l99VXX5mvIz8/P6Nnz55GTExMqnVmRMr33rx58zK0XGaO3eXLl40xY8YYlStXNvLmzWs4ODgY+fPnNypWrGj07t3bWL16dar1HzlyxHjppZeMUqVKGa6uroarq6sREhJijB49Os2/S8rX+K+//mo0bdrUcHd3N9zc3IyGDRsav//+u9n+zTffvOf7/s7f7W7HNTOfNXfcz/H++OOPrZarXr26cerUqTTbnj592pgxY0aa85YvX26EhYWZn11eXl5Gw4YNjS+//DJV25SvjaNHjxrTpk0zypUrZzg5ORkFChQwBgwYYCQkJDzwtgzDMFauXGk0adLE8PPzMxwcHAx3d3ejaNGiRuvWrY1x48al+7mXUq1atazqHjdunHHr1q1U7bZs2WKsXbvWatqVK1eMSZMmGbVr1zY8PT0NR0dHIyAgwGjZsqXxzTffpFpHyn8/4uLijP79+xsBAQFGnjx5jNDQUOPXX381DOP251v79u2NfPnyGXnz5jWaNGli7Nu3z2p9aX0GfvbZZ0aVKlUMFxcXI3/+/EavXr2MmJgYq+USEhKMkSNHGk2aNDGCg4MNDw8Pw8HBwfDx8TGefPJJY/r06caNGzfuuq3169cbM2fONP8NqVu3rmEY1u+VIkWKWK0jKirK6Nu3r1G8eHHDxcXFcHZ2NgIDA406deoYr776qnHw4MFUx2zdunVG27Ztzc8ODw8Po0aNGsb48eONixcvpmqf8v2yevVq48knnzTy5MljeHp6Gk8//bQRFRWVarl7OXnypNV3iZIlS5r/X7hw4TRfM3esWbPGaNeunREUFGQ4Ozsbnp6eRvny5Y3+/ftb7UPK2pcuXWpUr17dyJMnT6pjmdnjsnfvXqNr165GkSJFDCcnJ8PFxcUICgoy6tWrZwwfPtw4efKk2fbGjRvGlClTjFq1ahmenp6Gvb294e3tbZQtW9Z47rnnjCVLlmTomKV8zdw5fk5OTkZ0dLTZbujQoWYbe3v7u/67dPLkSWPw4MFGuXLlDDc3N8PZ2dkoWrSo0bNnT2Pv3r1p1hEVFWV06tTJ8PLyMvLkyWM8+eSTxrp161K9H1NKSEgwpk2bZjzxxBOGl5eX4ejoaAQGBhqdO3c2du/enap9Vr/2kTsQupGjpfxy17p1a/P/P/vsM8MwDOPUqVOGg4ODIclo06bNXUO3m5vbXb9Mly9f3rh06ZJhGJkL3VevXjWaNm1617Z3pPzHqXHjxmm2f/zxx1Mdj2HDht11Gy1atDCuX79utczw4cPTbFutWjXD398/06E75X7u3Lkzg3/N+9+PlMfs8ccft9kx++GHH6y+DKT1OHToUIb2s0WLFnddj4eHR6ovD8mDXIUKFYy8efOmWs7JycnYv3+/1XKzZ89OcxvBwcFG2bJl7/rlJi33E7oze+xCQ0Pv2rZjx45W6//yyy8NV1fXdNsXK1bMOH78uNUyyUN3jRo1zM+K5A9vb2/zi2FWhe7MfNbckdnjbRiGUbp0aXMZFxcXqy/UGZGUlGR06dLlrrX27dvXapmUr4303o9dunR54G1l5HM4vXCf3C+//GK1TMuWLTN8jM6cOWOUK1furjW88MILd627atWqqZZxcXExvv76a8PHxyfVPB8fH+Ps2bPm+lJ+BtavXz/NOooXL26cO3fOXO6ff/655/Fr2LChcfPmzXS3lfLve6/QHRMTY+TPn/+u2/zwww+tjtegQYPu2r5EiRKp3tvJ59epUyfdz4SMvD6SGzdunLl8njx5jE2bNlmt84cffki1TFJSktGjR48MfW6krD3l8U1+LDN7XA4cOGDkyZPnrssk/zGzW7dud21bs2bNDB2zlK+Z5N/V3nrrLcMwbn9PuvNar1SpktVnc8rPz02bNhn58uVLty5HR0cjIiIiVQ0BAQGp2losFqNZs2ZW05KLiYkxypcvn+62HBwcjPnz51stk5WvfeQedC9HrtK1a1f99NNPunDhgmbMmKHnnntOH374odmFuX///lqxYkW6yxcuXFghISEqXLiwvLy8ZBiGIiMjtWzZMl25ckX79u3TrFmzNHToUFWvXl0TJ07UsmXLzG6ZRYsW1Ysvvmiur1ixYpKkQYMGac2aNeb0xx57TK1bt5a7u7v27t2rb7/9Nt2avv/+e9WuXVsNGjTQN998Y3aT37Jli3755RfVrl1bkrR48WK999575nItWrRQrVq1dOrUKc2fP18JCQn69ttv9eabb2rcuHGSpB07dlgtExAQoOeff16XL1/W3LlzlZiYmJHDbrp165bVaOsVK1ZU1apVM7WO+9mPlLZs2WKzY/bxxx8rKSlJklS6dGk988wzcnBw0N9//609e/Zo9+7dGd5XLy8vNW3aVKVKlZKXl5ecnJwUExOjr776SidOnNDFixc1bNgwfffdd2kuv3fvXvn6+uo///mPYmJizG56169f1wcffKCPPvpIknTy5Emr60/d3d3Vq1cv2dnZ6dNPP1VkZGSGa07PmjVrdO7cuVTTO3bsqKCgIEmZO3aHDh0yX0t2dnZ6/vnnVbJkSZ07d06RkZGpRvU/duyYunbtqmvXrkmSKlSooDZt2uj69etasGCBTp06pb/++kudO3fWli1b0tyH7du3q0iRIurcubMOHDhgdnM8f/68Pv30U40YMcK8TvXDDz/UsWPHJEnVqlWz6kru7e19z+OVmc+a+3XmzBmra5ObNGmSaqyLe3n33Xe1ePFiSbf/Ds8884xCQkJ05MgRLVq0SElJSfr4449VtWpV9e3bN811bNmyRU2aNFH16tW1ePFi87gtWbJEEyZMMGu6n219+OGH5naqV6+usLAw3bx5UydOnNCvv/6qQ4cOZWg/169fb/W8Z8+eGT5GXbt21YEDB8znHTt2VMmSJfXtt9+ar+mPPvpIlSpV0n/+85801/Hbb7+pd+/ecnd31/Tp03Xz5k1du3ZNrVq1Up48eTRw4EBduHBB8+fPl3T7Wvy5c+dq+PDh6e5PvXr19OSTT2rLli368ccfJUlHjx7VsGHD9Mknn0i6PSJ+8eLFVbNmTQUGBsrLy0s3btzQ4cOH9cUXX+jmzZv64Ycf9OWXX6pDhw5pbmvLli0qWrSo2rZtKxcXF129evWux+vLL7/UP//8I+n2Z2CPHj3k4+Oj06dP6/Dhw/r555+t2n/22WeaPHmy+bxChQpq1aqVoqKitGjRIhmGoSNHjqhDhw7atm1bmtvcunWrQkJC1Lp1a/3888/66aefJEl//fWXli9frs6dO9+15uTu/A0kKSwsTE899ZSKFy+uo0ePSrrd9bxBgwZWy0ycOFHz5s0zn/v6+qpDhw7Knz+/Dh06pK+//jrd7W3ZskX+/v7q2LGjvL29zc/r+zku8+fPN/8+hQoV0rPPPis3NzedPHlS+/fvtzp+ly9f1sKFC83n7dq1U5UqVRQfH6/jx49r06ZNGT5mKTVs2FB//PGHDh8+rNmzZ2vEiBFavHixYmNjJd3+rvbWW2+luWxcXJyefvpp8zIQNzc39ezZU66urlqwYIHOnDmjGzduqHfv3qpSpYrKly8vSXr55ZcVHR1trqdly5aqXLmyVq9erdWrV6db67PPPqt9+/ZJuj02QdeuXRUQEKBNmzbpxx9/1M2bN9W7d29VrVpV5cqVu+t+Z/a1j1wmezM/8GBSnlFZtWqVMWTIEPP5Tz/9ZJ6tLVeunGEY1r8gp3X2Ni4uzvjuu++M2bNnG5MmTTImTpxoPPXUU+Yy9evXt2qf/JfgO7/wJxcbG2t19qxq1aqpuo7+9ddf5v+n/EW4Vq1aZve+2NhYqzOFH3zwgblc5cqVzekpzwYlP8uZN29eIzEx0TAMw3jhhRfM6fb29sYff/xhLrNo0aJ7HquUzp49a7VMhw4drOan/HvdeQwePPiB9uNhHrNWrVqZ09PqWnfmzBnjypUr9zxWd1y/ft346aefjLlz5xpTpkwxJk6caHVGxNnZ2epMe/Kzp3Z2dlZdn5P35KhSpYo5ffz48VbHJ/mZmC1btljNu98z3ek9knd1z8yx2717t9m2TJkyqbps3rx506pr6Kuvvmq2L1++vPn3MgzDOHz4sFVNW7ZsMeclP5uSN29e48yZM+a85K+Ptm3bWm0/I13y79Ums581yfchI2e6t2/fbrVM8stYDCP9s8TTp083DOP22bnkZ1nHjRtntXzyXjIlSpQwp6d8bbRv396ct2fPHqt5X3/99QNtq0KFCub0X375JdUxiIyMzFD38pdeesmqroz2Vvntt9+slhsxYoQ5LzEx0ShTpow5r3jx4ua8lMf+nXfeMed16tTJal7y90r16tXTfE2m1UPqznvm1q1bVj2mnJ2dU31GxcTEGCtXrjRmzZplvP/++8bEiRONkJAQc5mePXumu60SJUoY8fHxqY5Nemf7Jk+ebE5P2QPAMG5fVpK8y3HFihXN9sHBwVZnpt966y2rWjZv3mzOSz69SJEi5r+7169fN/z8/Mx5gwYNSlVDerZu3Wq13uXLlxuGYRivv/66OS1PnjxW3bqTkpIMX19fc35QUJDxzz//WK03Ojra6m+SfBv58uVL85KQ+zkur7zyijlt/PjxqdZ5/vx54/z58+b/32nr4eFh9ZlqGLdfV8eOHcvQcUv5mpk+fboxY8YM8/myZcuMSpUqGdLtXhwJCQnpnumeMmWK1bqSX+rx119/GY6Ojua83r17G4Zx+9IZi8ViTn/22WfNZa5fv56qp8odv//+u9X0rVu3Wu1/7dq1zXl9+vQx52XVax+5C6EbOVpaofvYsWPm9UIFCxY0582ePdswjPRDd1JSkvHaa68ZTk5OaX4RvfMoWbKkVQ33Ct3fffed1fKff/75Xfcp5T9Oc+bMsZqfvMv3mDFjDMO4fT1h8n9Q7vXYsWOHYRiGVZfGlN3Ebt68afWPV0ZCd0xMjNV2Unb/vVfovt/9eJjHbOLEieY0Z2dnIzQ01Ojbt68xadIkY9u2bXe9ni+lhQsXWn0ZS+9x+vRpc5nkQS5ld/nkXeWDg4PN6e3atTOn+/v7p6ojODg4zS83d3M/oTszxy4hIcEqhBUtWtRo166dMWLECGPJkiWpvrTWqFEjw3/LiRMnmssl/2LXtWtXq3V27NjRnFevXj2reQ8Suu/3syb5vIyE7l9//dVqmWHDhlnNv1foPnjwYIaPqSTzb5LytbFu3TpzmwkJCVbz7nTLvN9t9evXz5yWN29eo1GjRsZLL71kzJgxI93rOtPy4osvWq0/o6F75syZVsulvB5zzJgxVvPvdAlPeeyTdy1O/gODo6Oj1TXVXbt2TfM1mfIz8M7lVXfMnz/fav6d68WvXr1qdO/ePdV4JykfjRs3TndbU6ZMSfPYpBc8fv31V/Oz12KxGFWqVDGeffZZ4+233zZWr15tNR7L5cuXrbaV8oej48ePW82fMGGCOS/59Ndff91quZo1a5rzevTokWb9aenbt6+5nKenp1nrvn37rLb3ySefmMukfG0nrzE9ydsPGDAg1fz7PS5ffPGFOc3e3t6oU6eO0aNHD+Pdd981NmzYYHUZgWEYVmE0MDDQaN26tTFkyBBj/vz5mbpUJa3QfenSJcPDw8OQrL+rDR8+3DAMI93Q/cwzz5jT/fz8Um2rXr165vyyZcsahmEYq1atstp+yvFAUr5P75g1a1aGP5PunNgxjKx57SP3oXs5cp3g4GC1aNFCq1at0qlTpyRJ+fLl07PPPnvX5T744ANNnDjxnuvPbJfr8+fPWz1/7LHHMrV8kSJFrJ47Ozub/3/r1i1J0oULFzJ126073ZuSj9Lr5+dn1cbe3l4+Pj5W3bHuxdfXVy4uLmYX3z/++MNqfrFixcxj/Nprr6Va/n73IyVbHrOBAwdq7969Wrx4sRITE7Vx40arrs4hISFat26dAgIC7rq+3bt36/nnnzfruZv0XnMZ2U/p7n9nSfL393/gLubz5s27571UM3PsXFxc9Pnnn6tHjx76+++/dezYMbNbsiQ5OTnp3Xff1auvviop9fvsbh7kdZMVbPVZk1KhQoWsnqd8P965RObChQtpXqqRmWMq3T6uvr6+qaYnP67Jj6n0v+N6v9saN26cjh07ptWrV+vy5ctat26d1q1bZ7arW7euvvvuO+XJk+eu60t5rA4fPqzSpUvfs44LFy5YPU/5/vL390/VPn/+/KnWk7zbf/Jj5OfnJweH/31VS/7/d3tNZqQOSRoxYkS6I3End7fXYsmSJe+5fHI1atTQ5MmTNWrUKF2+fFm7d++2urTE19dXX3zxhUJDQ1Pdliuj+5VSVry3r127ps8//9x83qZNG3M9ISEhKleunHmZQUREhHr16iXpwb8DpHV87/e4tG/fXkOGDNH06dOVmJiorVu3auvWrWa7IkWK6NtvvzW7SS9evFidO3fWwYMHdfr0aa1cudJsa2dnpwEDBlh1cc+MvHnzqlu3bpo+fbr5Xc3e3t7qEr20JP8bp/fvWcq2mT1ed2TFvyvJZea1j9yH0I1cqX///la3nejZs6fc3NzuusyyZcvM/w8JCdHixYtVunRpOTo6aujQoRn6kpyWlNd3RkVFqXr16hle3tHR0eq5xWJJ1SZfvnxWz9u2bWtet5yWUqVKpVru7NmzVm2SkpLM66syys7OTqGhoeb163v27NGePXtUqVIlSVJQUJCGDBkiKe3Qfb/7kZItj5mDg4M+++wzTZo0SVu3btUff/yhP/74Q8uXL9eFCxe0f/9+DR8+/J5fZL/44gvzy56bm5v++9//qm7dunJ1ddV3332nFi1a3HX5jO6ndPe/syTFxMTcc1tZIbPHrn79+oqMjNTu3bu1Z88eHT16VFu3btXPP/+s69eva8iQIWrVqpWKFSsmLy8vczsVK1a8649s6b3/Mno8H5StPmtSCgwMVOnSpc3ruteuXavo6GjzB6Fy5cqpXLlyioqKSjN0Jz+mktS7d+9033NS2l+AJevjmt4xvd9teXh46LvvvtPJkye1bds2/fnnnzp48KCWL1+uq1evatOmTZowYYLCw8PTXZd0+7WWXEREhNq0aXPXZdKq++zZs/Lx8TGfp3xvpWx/R8rX3h3JQ3ZmpHyfp6zjzmdC8tdivXr19PHHHys4OFj29vbq0KGDvvjii3tu614/aKRl4MCB6tu3r7Zt26YDBw7oyJEjWrNmjY4cOaJz586pe/fuioqKSvU5fa/9yujxvZ/3dsp7c8+fP9/q+u7kkt+zO63vAJmR1vF9kOMyceJEvfHGG9q6dasOHz6sP//8U19//bVOnz6t48ePq1+/fuaPoRUqVNCBAwe0b98+7d69W0eOHNHu3bu1evVq3bp1S1OmTFGrVq3uOyS+/PLLmjFjhvkDeOvWrVW4cOG7LpN8X+7179mdtpk9XmltS5LGjRuX7ns1o++DjL72kfsQupErNWzY0PyyaWdnp379+t1zmeQBs169eubgGwkJCXcd5CT5B3BaA8jUrFlTDg4O5mBuEydOVFhYmFxdXc02x48fT/VLfGa4ubmpYsWK5v1jL1y4oFdffVX29vZW7f755x9z0Bvp9gBQu3btkiTt3LlTf/75p/mr+rJly3Tjxo1M1zJgwACrQeM6d+6s7777TsHBwTbbj/txv9v6448/FBQUpPz586t169Zmu5CQEA0aNEiSzGN6N8lfb0WLFlXTpk3N50uXLr3v/UpLtWrV9OWXX0q6/eXixx9/NAf62bp1a5YMpJYRmTl2165dU2RkpMqUKaNq1aqZ9wA3DENeXl6Kj4/XrVu3tGfPHhUrVkx16tTRjh07JN0eQOzZZ59N1dvg2rVr+uKLL1S3bt0H3pd7ve/v5n4/a+7HgAEDzDNHCQkJat++vVasWJHmGemUSpcuLR8fH7PexMRE80ez5P7++28dOnTIKmxm1v1ua//+/SpVqpQKFSqk9u3bm+0GDBigDz74QFLG3o+1atVSzZo19euvv0qSVq5cqYkTJ6b54+Avv/yiy5cvq1GjRqnuQ71gwQLzB4zr169bvZeLFy+e5lluW1i4cKGeffZZWSwWGYahRYsWmfOcnJzM11zy12JYWJiKFy8u6XYo2bBhg01qO336tOzt7eXv76/69eubP3j89ttvqlKliqTb/ybGxsbKx8fH6nP6v//9r8aMGSMXFxdJShV6bXlf8Iz0CEjuzj27S5UqJV9fX3OgyRkzZqhnz55W75d//vlHbm5uGQ5uKf/9yuhxiYyMlJeXl/Lly6dmzZqpWbNmkqTGjRurbdu2kqzfL3d+NC9fvrz5mpFu/7C5d+9es/39hu6SJUuqcePGWrt2raTbJ0zupU6dOuaPQWfPntX333+vxo0bS7o9oObmzZtT7XeVKlXM94IkLVq0yPw398aNG1Y9GFJuK7mAgAD16NEjVbvt27en6sWTlsy+9pG7ELqRK1ksFn3++ef666+/5O7unqFwVqpUKR05ckSSNGfOHFksFnl4eOiLL75I1S0zueTdAnft2qUBAwYoKChITk5OeuWVV+Tt7a1evXqZI0nv2LFD5cqVU5s2beTh4aGDBw9q5cqVD9yVdMiQIXruueckSRs2bFDFihUVFhYmT09PnT17Vjt37tQvv/yiJ554wjyD07NnT3388ccyDENJSUmqW7euunXrpkuXLmnu3Ln3VUfTpk3Vt29fffzxx5Jud9MsU6aMWrVqpfLly8tisZijvGbVftyv+9nWlClTtGDBAjVo0EDBwcHy9/fX+fPn9dlnn5nrTfmrelqSn8Xbt2+fOnbsqJCQEG3cuDHVSMoPqmvXrgoPDzdfY08//bR69+4ti8WiTz/9NEu2kd7o5X5+fnr++eclZe7YxcXFqWzZsipXrpxq1KihwMBAubq6avPmzYqPj0/Vvn///po9e7YSExN19uxZVaxYUR06dFBgYKAuXryoffv2adOmTbp8+bL5N38Qyd/33377rYYPHy5fX1/5+vres5v9/X7W3I8+ffro66+/Nkfn3bJli4oVK6bWrVurVKlSSkpK0m+//ZbmsnZ2dho4cKBGjRol6XagPHLkiOrXry83NzedPn1a27ZtMy+VaNKkyX3Xeb/bGjJkiLZv364GDRqYP+icPn3aaqTojLwfJWnu3Ll6/PHHzdfX0KFDzS/n+fLlU3R0tDZt2qQ9e/ZoypQpatSokSpVqqTQ0FDzzOD48eMVGRmpkiVL6ptvvrEaPf3OpRAPw/fff68GDRroqaee0ubNm83Ry6Xbnwd3wl2pUqW0f/9+SdI777yjmJgYWSwWLViwIM33c1b46aef1LVrVz3xxBMqU6aMAgMDlZSUpK+++sps4+TkZP4w/eqrr5rvqWPHjqlmzZpq3bq1IiMjrX5MqFGjhh5//HGb1Hzq1Cn98MMP5vPy5curbNmyqdpt3bpVJ06ckHR7dPG33npLdnZ2Gjx4sEaMGCHp9g9HZcqUUceOHZU/f34dPXpUy5cv1759+zLV9fx+jsuyZcv05ptvKjQ0VCVKlFCBAgV05coVLVmyxGyf/P1Sq1YtBQYG6sknn1RgYKA8PDz0+++/m4E7Zfv7MXPmTO3bt0/29vYZCu/dunXT22+/bXb9btu2rdXo5XdOFjg4OJghPjAwUM2aNTPvBLJw4UJdvHhRlSpV0urVq63uPpBcpUqV1KBBA/P906dPH61atcrsvRcZGalNmzYpMjJS8+bNU8WKFe9ae2Zf+8hlsu9ycuDBpTWQ2r0kb598cLCff/45zXv05s2b12jbtm2ag2IYxu3Ra9MahMbNzc1sc/XqVaNJkyZ3HYTjjpQDjiQfiMowrAcXSTm42WuvvXbPwT5SDvaW3jLlypWzGuQro/fpNozbg7CNHDnynoPz3HmkHKk4s/vxMI9Z8hHf03rY2dmZI9reTWxsrBEYGJjmOlLeGzX5IEt3G8ArvcFbDMOwGik2+SMwMNAoUaJEuutMT0YHUqtYseJ9HbszZ87cc901atSwGmTqv//9713v053yvWYYd39t3G2QxJUrV6b7vrnX3+p+P2uSt83ofboN4/aggb169crQ38tisRgLFy40l71586bRuXPney6XfP9SvjaSv37vth/3s617fa66uLiYg4ZlxO7du42SJUves4bkg4edOnXK6n7oaT169eplNVBgyoHUkrvb+zi912TKz8AWLVqkWUfRokWtBiFcsmRJmu0KFChgNGrUKEPbSvl5e6/9SG+byR8pRxRPPup2evuV0deZYWRsIMTkkt+bW5Kxbdu2NNulHFjvzp0ikpKSjO7du991H9K7T/fd3uuZPS4p72SR1mPatGlme2dn57u2DQ4ONuLi4u55/NIaSO1e7naf7vXr1xuenp7p1uXg4GDMnTvXapljx45ZjVqf/JH89ZDy/RgdHX3X+3Sn9XfKytc+cg87AZAkPfHEE1q7dq3q1KkjZ2dneXp6qnnz5tq6datVt6qUKlWqpCVLlqhKlSpm166UXF1dtXr1ai1ZskTNmzeXv7+/HB0d5enpqUqVKmnw4MFZsg8TJkzQpk2b1KlTJxUuXFjOzs7y8PBQ6dKl1bp1a82ZMydVN6oJEyZo9uzZKlu2rJycnFSgQAH169dPP//88z2vg0+Pvb29xo4dqz///FPDhg1TzZo15ePjIwcHB+XJk0fBwcFq3ry5xo8fr4MHD5pnAB5kP+5XZrfVq1cvDRs2TE899ZSCgoLk4uIiJycnBQUF6ZlnntGmTZsydAbe29tbmzdvVtu2beXh4SFXV1dVr15dX3311T3PlN6Pfv366b///a+qVq0qZ2dn+fr66rnnntOvv/6qwMDALN9eWjJz7Ly8vDRjxgx17txZZcuWlbe3t+zt7eXh4aFq1arp7bff1o8//mh1zWu7du20b98+vfLKKypbtqzc3Nzk4uKiokWLql69eho/frzVfasfRKtWrTRjxgyVKVMm3Wv80nO/nzX3K0+ePPrkk0+0Z88e9e/fX5UqVVK+fPlkb2+vvHnzqmTJkmrTpo2mTJmiyMhIde3a1VzW3t5eixcv1sqVK9W6dWsFBgbK0dFRXl5eCgkJUceOHbVo0SJNmzbtgeu8n2299tprGjBggGrVqqWCBQvKyclJzs7OKlq0qLp166bt27erRo0aGa6hcuXK2r9/vxYsWKCnn35ahQsXlqurq/LmzatSpUqpTZs2WrhwoXr37m0uExgYqJ07d2rChAmqWbOmPDw85ODgID8/P4WFhWnlypX65JNPbDZGQFqGDBmiJUuWqGrVqnJxcZGvr6969uyprVu3Wl1a0KlTJ33++eeqWLGiHB0d5ePjo44dO2rbtm02+1x44oknNHbsWLVo0ULFihWTu7u7HBwclD9/fjVo0EARERF6//33rZaZNm2a1qxZozZt2qhAgQJycHBQ3rx5Va1aNb3zzjv67bffMj1AWWYk765dtmxZ1axZM812nTt3tvoecKdLup2dnebNm6fVq1erXbt25mvV3d1dZcqU0UsvvXRf3Ykze1zatGmj0aNHq2HDhnrssceUJ08eOTg4qECBAmrRooW+/vprvfLKK2b7Dz/8UD169FCFChWUP39+c/0VKlTQ0KFD9euvv8rT0zPTdT+oevXqad++fRo4cKDKlCkjV1dXOTs767HHHlP37t21c+dO9ezZ02qZ4OBgbdu2TR06dFC+fPnk6uqq2rVra9WqVXf9N9ff31/bt2/X9OnTVbduXXl7e8vBwUEBAQGqWrWqXnzxRa1du9bqczM99/PaR+5hMYxMDN8LAACAR0pUVJTVuBkbNmxgBGQAeIRwphsAAAAAABshdAMAAAAAYCOEbgAAAAAAbIRrugEAAAAAsBHOdAMAAAAAYCOEbgAAAAAAbMTh3k3+HW7duqXTp0/L3d39od5LEwAAAACQ8xiGoUuXLikwMFB2dumfzyZ0/7/Tp08rKCgou8sAAAAAAOQgJ06cUKFChdKdT+j+f+7u7pJuHzAPD49srgYAAAAA8Ci7ePGigoKCzCyZHkL3/7vTpdzDw4PQDQAAAADIkHtdnsxAagAAAAAA2AihGwAAAAAAGyF0AwAAAMgV9uzZoxYtWqhw4cJydXWVt7e3ateurYULF6Zqe+PGDU2ePFnly5eXq6ur8uXLpzp16mjr1q1mm4iICFkslnQf77777sPcPeRQXNMNAAAAIFeIi4tTUFCQOnfurIIFC+rKlStatGiRnnvuOUVFRemNN96QJCUlJenpp5/W5s2bNXToUNWpU0dXrlzRrl27dOXKFXN9LVq00C+//JJqO6NHj9a6dev09NNPP7R9Q85lMQzDyO4iHgUXL16Up6en4uPjGUgNAAAAyEVq1aql06dP6++//5YkTZ06VYMHD9aWLVtUq1atTK3rypUrCggIUKVKlfTzzz/bolzkEBnNkJzpBgAAwL9KUlKSbty4kd1l4CEqXbq0HBwcdO3aNUnSF198oXbt2qlSpUrmtIxasWKFfHx89NJLL2V6WeQMDg4Osre3v+eo5BnFme7/x5luAACA3M0wDEVHRysuLi67S4GN3Yk4t27d0tWrV3X+/Hl5e3vL3d1dN2/e1KlTp+Tu7i6LxaLLly/r1q1bcnR0lIeHh/LmzXvXdUdHR+v69esqVKiQ7OwYIiu3sre3l5+fnzw9PdMN35zpBgAAAJK5E7j9/PyUJ0+eLDuLhUfPqVOndOHCBUm3w1P58uXl4+Mj6Xb38MTERNnZ2cnBwUFFixaVvb29Lly4YIYob2/vNNd77do1XblyRV5eXipYsOBD2x88PIZh6ObNm7p48aLOnDmjhIQEFShQ4IHWSegGAABArpeUlGQG7jvhC7lXoUKFFBAQoJs3byouLk5nzpyRvb29OU26Ha5KlSolZ2dnSZKvr68OHTqkc+fOKTAwMM31/vPPP5KkgIAAubi4PJydQbZwd3eXs7Ozzp07Jz8/P9nb29/3uugPAQAAgFzvzjXcefLkyeZK8DA4OzvLzc1Nnp6eKlKkiPLnz69Tp07pxo0bcnC4fd7RxcXFDNySZLFY5OHhoevXr6d5zf+tW7cUGxsrV1dXubm5PbR9QfZxc3OTYRgPPAYEoRsAAAD/GnQp/3e6E54SExPl7Ox8X9dix8fH6+bNm8qfP78NKsSjKKs+LwjdAAAAAHK1S5cuSbp9BtxisShfvny6du2aEhMTzTaGYejixYtydnaWo6NjqnWcO3dOFosl3eu9gfQQugEAAIAcLCIiQhaLRRaLRRs3bkw13zAMFS9eXBaLRaGhofe1jVmzZikiIuK+lo2KipLFYtH7779/X8tnxvvvvy+LxaI9e/bo0qVLunDhgv766y/FxsYqICDADNOBgYGys7PTsmXL1K5dOwUFBcnZ2VlVq1bVM888o8GDB+vw4cPmeq9fv674+Hh5eXmZ3dOBjOIVAwAAgH+1pm066GTMuewuQ4X8fbVmxef3vby7u7vmzp2bKlhv2rRJf/31l9zd3e973bNmzZKvr6+6d+9+3+t4GO5co3369GklJSXJzs5OefLkUXBwsNUAei4uLlq6dKkmTJigChUq6Pnnn1eJEiXk6empY8eOaf78+Zo8ebJu3rwpe3t7xcbGSro92BqQWYRuAAAA/KudjDmny3UHZ3cZOrlp0gMt37FjRy1atEgzZ860umfw3LlzVbt2bV28ePFBS3zk3flhoWzZsnrsscfSbbdkyRJNmDBB//nPfzRr1qxU1+4OGjRIs2bNMp8XKFDggW8bhX8vupcDAAAAuUDnzp0l3Q6Ud8THx+vLL79Uz54901xmzJgxqlmzpry9veXh4aEqVapo7ty5MgzDbPPYY4/pwIED2rRpk9mNPXmgjYuL0+DBg1W0aFE5OzvLz89PzZs3t+qefcfkyZMVHBysvHnzqnbt2tq2bVuqNjt37lSrVq3k7e0tFxcXVa5cWZ9/nroHwLZt2/T444/LxcVFgYGBGjFiRIZHmX7nnXfk6+urKVOmpDlYlsViUb9+/VLdJuqHH35QgwYN5OHhoTx58ujxxx/Xjz/+aNUmPDxcFotFBw4cUOfOneXp6Sl/f3/17NlT8fHxVm0Nw9CsWbNUqVIlubq6ysvLS+3bt9exY8es2v32228KCwuTn5+fnJ2dFRgYqBYtWujkyZMZ2l9kL0I3AAAAkAt4eHioffv2+vTTT81pS5YskZ2dnTp27JjmMlFRUXrhhRf0+eef66uvvlLbtm3Vv39/vf3222ab5cuXq2jRoqpcubJ++eUX/fLLL1q+fLmk2wOUPfHEE/roo4/Uo0cPrVq1SrNnz1bJkiV15swZq23NnDlT69at09SpU7Vo0SJduXJFzZs3twqiGzZs0OOPP664uDjNnj1bK1euVKVKldSxY0era8oPHjyoBg0aKC4uThEREZo9e7Z+++03vfPOO/c8TqdPn9bBgwfVqFGjTN1re+HChWrcuLE8PDw0f/58ff755/L29laTJk1SBW9JateunUqWLKkvv/xSw4cP1+LFi/Xqq69atXnhhRc0cOBANWzYUCtWrNCsWbN04MAB1alTRzExMZKkK1euqFGjRoqJibE6hoULFzYHiMOjje7lAAAAQC7Rs2dP1atXTwcOHFC5cuX06aef6plnnkn3eu558+aZ/3/r1i2FhobKMAxNmzZNo0aNksViUeXKleXq6ioPDw/VqlXLavmpU6fqwIEDWrdunRo2bGhOb9u2baptubu765tvvjHPHgcGBqpGjRpavXq1OnXqJEl66aWXVK5cOa1fv94csKxJkyY6d+6cRo4cqeeff152dnZ66623ZBiG1q9fL39/f0lSixYtFBIScs9jdOLECUlSkSJFUs1LSkqyOstvb28vi8Wiq1evasCAAQoLCzN/cJCk5s2bq0qVKho5cqR+/fVXq3X16tVLr732miSpYcOGOnr0qD799FPNnTtXFotF27Zt05w5czRp0iQNGjTIXO7JJ59UyZIlNXnyZL333ns6fPiwYmNjNXfuXLVu3dps16FDh3vuKx4NnOkGAAAAcom6deuqWLFi+vTTT7Vv3z7t2LEj3a7lkrR+/Xo1bNhQnp6esre3l6Ojo0aPHq3Y2FidPXv2nttbvXq1SpYsaRW409OiRQur7toVKlSQJB0/flySdPToUR0+fFhdu3aVJN28edN8NG/eXGfOnNEff/wh6fYZ8QYNGpiBW7odkNM7o59RPj4+cnR0NB9ffvmlJGnr1q06f/68unXrZlXXrVu31LRpU+3YsUNXrlyxWlerVq2snleoUEHXrl0zj+s333wji8WiZ5991mqdAQEBqlixojkSffHixeXl5aVhw4Zp9uzZOnjw4APtIx4+znQDAAAAuYTFYlGPHj30wQcf6Nq1aypZsqSefPLJNNtu375djRs3VmhoqObMmaNChQrJyclJK1as0NixY5WQkHDP7f3zzz8qXLhwhmpLPnq49L+Rxu9s50536iFDhmjIkCFpruPcudujzN+5BVhKaU1LKSgoSNL/wn5yGzdu1M2bN7Vr1y795z//Maffqa19+/bprvf8+fNyc3Mzn2dkfw3DsPrhILmiRYtKkjw9PbVp0yaNHTtWI0eO1IULF1SgQAH16dNHb7zxRpr3FMejhdANAAAA5CLdu3fX6NGjNXv2bI0dOzbddkuXLpWjo6O++eYbq2ubV6xYkeFt5c+fP8sG87pzO64RI0ak2T1dkkqVKiXpdqCNjo5ONT+taSkFBgaqXLlyWrduna5du2a175UqVZIkXb58Oc3apk+fnqqL/R3phef0+Pr6ymKx6OeffzYDeXLJp5UvX15Lly6VYRjau3evIiIi9NZbb8nV1VXDhw/P1Hbx8BG6AQAAgFykYMGCeu2113T48GF169Yt3XYWi0UODg5WXb4TEhK0YMGCVG2dnZ3TPPPdrFkzjR49WuvXr1f9+vUfqO5SpUqpRIkS+v3339W9Zy/duHEzVZu/T56WJFWpWk0//PCDNv28xQzESUlJ+uz/a//z6F+6kpCY7ra6de+hoa8NUfcePfXGqNGpRjCPPP63JOnEyVM6cOgPefvkl4eHhzZv2ap6DRqluc4jf0VKks7+c/ts/OE/j8rrn1hz/qnTZ6xqq1CpsgzD0KlTpzJ8fbbFYlHFihU1ZcoURUREaPfu3RlaDtmL0A0AAADkMu++++4927Ro0UKTJ09Wly5d1LdvX8XGxur9999P86zrnTOty5YtU9GiReXi4qLy5ctr4MCBWrZsmVq3bq3hw4erRo0aSkhI0KZNmxQWFqZ69eplqu6PPvpIzZo1U4/u3dSqU3f5BRRQfFycIo/+oUP79+r92RGSpN6DXtfGDRvUs3dvvTDgNbm4umrZ/LlKuHY7aCe55VeSe/pnnpt06qUjf5/RnOmT9MfRv9TqmS4qHFxMt27dUszpU/rmq2WSJBffQkpy95ezuzTs7Qka9epLirtyTY1atJa3j68unI/VHwf360LsOb0xfrIkyXDOe7uGvPmV5P6/Lua3XDytaqtQoaL69u2rHj16aOfOnXrqqafk5uamM2fOaPPmzSpfvrxefPFFffPNN5o1a5batGmjokWLyjAMffXVV4qLi1OjRmn/AIBHC6EbAAAA+BeqX7++Pv30U7333ntq2bKlChYsqD59+sjPz0+9evWyajtmzBidOXNGffr00aVLl1SkSBFFRUXJ3d1dmzdvVnh4uD7++GONGTNGXl5eql69uvr27ZvpmurVq6ft27dr6LDhmjBmpC7Gxymfl7eKliilxmFtzHYlSpfVR0uWa9LbozRq0Evy8MynsLYd1LB5S701bGCGtvXy0DdUp259LZ3/iT6aOkGx5/6Rg4OjAoMKq1rNOho4IlxlK1Qy24e17agCgYU0b/YHenv4q7py5bK8ffKrVLkQtX6mS6b3Vbr9I0OtWrX00UcfadasWbp165YCAwP1+OOPq0aNGrf3tUQJ5cuXTxMmTNDp06fl5OSkUqVKKSIi4q49GfDosBjJx8T/F7t48aI8PT0VHx8vDw+P7C4HAAAAWejatWuKjIxUcHBwqnszN23TQSdjzmVTZf9TyN9Xa1Z8nt1lPBIOHPrjrmeqcwP7SzEqV6ZUdpeBu7jb54aU8QzJmW4AAAD8qxF0AdgS9+kGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANiIQ3YXAAAAAGSn555uoviYv7O7DHn6F9aC5WszvVxERIR69OghSdqwYYNCQ0Ot5huGoRIlSuivv/5S3bp1tXHjxkxvY9asWcqTJ4+6d++e6WWjoqIUHBysiRMnasiQIZlePjNWfr5Yowf303dbf1fBoMLptvtw8ruaPeW9dOffa/lHQXp/kzvHe968eZn+e2V22ZiYGE2bNk3fffedjh07pmvXrsnPz081atRQ9+7d1aJFC9nb22eqhtyI0A0AAHCf9uzZo9dff1379u3TP//8I1dXV5UqVUr9+vXTs88+m+YyhmGobt26+vnnn9WvXz/NmDEjVZvp06dr5syZioyMVGBgoLp3766RI0fK0dHR1rv0rxQf87e+bnw6u8tQq+8fbHl3d3fNnTs3VejetGmT/vrrL7m7u9/3umfNmiVfX9/7Ct2PslkL/it3D49U0/P7+WdDNZmT3t+kQIEC+uWXX1SsWDGbbn/btm1q1aqVDMPQiy++qFq1ailv3rz6+++/tWrVKrVt21YfffSRevXqZdM6cgJCNwAAwH2Ki4tTUFCQOnfurIIFC+rKlStatGiRnnvuOUVFRemNN95ItczMmTN19OjRdNc5duxYjRo1SsOHD1fjxo21Y8cOvfHGGzp16pQ+/vhjW+4OcriOHTtq0aJFmjlzpjySBcm5c+eqdu3aunjxYjZW92gqW6GSvLx9sruMLOXs7KxatWrZdBtxcXFq06aN8ubNqy1btqhAgQJW85999lnt3btXsbGxNq0jp8j2a7o//PBDVahQQR4eHvLw8FDt2rW1evVqc3737t1lsVisHilfRImJierfv798fX3l5uamVq1a6eTJkw97VwAAwL9MaGioZs+erWeffVb16tVTWFiYlixZopo1a6YZkKOiojRixAjNnDkzzfXFxsbqnXfeUZ8+fTRu3DiFhobqtdde05tvvqlPPvlEBw8etPUuIQfr3LmzJGnJkiXmtPj4eH355Zfq2bNnmsuMGTNGNWvWlLe3tzw8PFSlShXNnTtXhmGYbR577DEdOHBAmzZtMr+PP/bYY+b8uLg4DR48WEWLFpWzs7P8/PzUvHlzHT58ONX2Jk+erODgYOXNm1e1a9fWtm3bUrXZuXOnXu73op4MCVb14gHq0PQprV21PFW7vbt3qNvTTVS9eIAaVi2jae+O0c2bNzJ8vDLq88/m6pnGT6hWqUKqXTpIrUNr6IN33zLnJyRc1aS3R6lZnYqqXjxAT4YEq3Pzelq94r9W6znw+296pUdnc7/at31an3/+uVWbiIgIWSwWbdiwQS+++KJ8fX3l4+Ojtm3b6vTp//XGuNvfJCoqShaLRREREWb7o0ePqkePHipRooTy5MmjggULqmXLltq3b999HZM5c+YoJiZGEyZMSBW476hQoYLq1atnNS06OlovvPCCChUqJCcnJwUHB2vMmDG6efOm2eZO/e+//36GXy+tWrWSt7e3XFxcVLly5VTH9erVqxoyZIiCg4Pl4uIib29vVatWzeq9YkvZfqa7UKFCevfdd1W8eHFJ0vz589W6dWv99ttvKleunCSpadOmmjdvnrmMk5OT1ToGDhyoVatWaenSpfLx8dHgwYMVFhamXbt2cQ0BAAB46Hx9fXX27NlU0/v27atGjRrp6aefTnO5NWvW6Nq1a+b1uXf06NFDr7/+ulasWKGyZcvapGbkfB4eHmrfvr0+/fRTvfDCC5JuB3A7Ozt17NhRU6dOTbVMVFSUXnjhBRUufPv65W3btql///46deqURo8eLUlavny52rdvL09PT82aNUvS7bOpknTp0iU98cQTioqK0rBhw1SzZk1dvnxZP/30k86cOaPSpUub25o5c6ZKly5t1jFq1Cg1b95ckZGR8vT0lHT7mvSmTZuqfPkKemP8ZOV199Car7/S0Jd66lpCglp36CJJ+uvPw+rTqY0CgwrrrUkz5erqqmWffZoq6N7LraQkq8AnSRaLxcwQq1d+qbGvD1HnHn016I23ZWdnp7+jjunYkT/M9u+/9bq++fJzvfza6yodUkEJV6/o6B+HFBd33myzfevPeum59ipfqaq5X99/uUgdO3bU1atXU3UR7927t1q0aKHFixfrxIkTeu211/Tss89q/fr19/ybpOX06dPy8fHRu+++q/z58+v8+fOaP3++atasqd9++02lSpXK1HFbt26d7O3t1bx58wwvEx0drRo1asjOzk6jR49WsWLF9Msvv+idd95RVFSUVd6TMvd6qVmzpmbPni1PT08tXbo01XEdNGiQFixYoHfeeUeVK1fWlStXtH///od2Jj7bQ3fLli2tno8dO1Yffvihtm3bZoZuZ2dnBQQEpLl8fHy85s6dqwULFqhhw4aSpIULFyooKEg//PCDmjRpYtsdAAAA/3q3bt3SrVu3dOHCBX3xxRdau3Ztqmu1P/nkE23fvv2uZ6v3798vSSpfvrzV9AIFCsjX19ecD6SnZ8+eqlevng4cOKBy5crp008/1TPPPJPu9dzJg86tW7cUGhoqwzA0bdo0jRo1ShaLRZUrV5arq6s8PDxS9TidOnWqDhw4oHXr1pnfxSWpbdu2qbbl7u6ub775xgy0gYGBqlGjhlavXq1OnTpJkl566SWVK1dOc+dFyOJVUJL0eGgDxV2I1fT33lbL9p1kZ2enj6ZOkAxDnyxdKZ/8fpKkJxs0UbuGdTJ1vOpXSR02g4oE65vNuyVJe3b+KndPTw1/63+DrtV8oq5V+z07flXtp+rpuT4vmdOeamCdQca9PkTFSpbWnGVfy8HhdgR7qmqIbt68oZEjR+r555+Xnd3/OiE3bdpUH3zwgfn8/PnzGjp0qKKjoxUQEHDXv0lannrqKT311FPm86SkJLVo0ULlypXTRx99pMmTJ99zHcmdOHFC+fPnV548eaym3/ksvMPOzs7cr/DwcF24cEEHDhwwf+Rp0KCBXF1dNWTIEL322mtWPypm5vWyfv1687g2adJE586dszquW7ZsUePGjfXqq6+a62/RokWm9vlBZHv38uSSkpK0dOlSXblyRbVr1zanb9y4UX5+fipZsqT69Olj9cvxrl27dOPGDTVu3NicFhgYqJCQEG3dujXdbSUmJurixYtWDwAAgPvx0ksvydHRUX5+fnr11Vf1wQcfmGcaJenUqVMaMmSIJkyYoMDAwHTXExsbK2dnZ7m5uaWa5+3tzfWRuKe6deuqWLFi+vTTT7Vv3z7t2LEj3a7lkrR+/Xo1bNhQnp6esre3l6Ojo0aPHq3Y2Ng0e2uktHr1apUsWdIqcKcn5UjWFSpUkCQdP35c0u0u0IcPH1bXrl0lSTdv3jQfT9ZrpH/ORivqryOSpB2/bFaNJ54yA7ck2dvbq0nLtHuRpOfjJSu0+Jv1Vo8pnyw054dUqqJL8fEa1q+XNqz9ThfOp34PhlSqqi0bf9DU8eHa8ctmXUtIsJr/d+QxRR79U83bPJNqv5o3b64zZ87ojz/+sFqmVatWVs9THqvMunnzpsaNG6eyZcvKyclJDg4OcnJy0pEjR3To0KH7WmdaBg0aJEdHR/ORfD+++eYb1atXT4GBgVbHoFmzZpJuD/iX3IO8XlIe1zthffjw4dq4caMSUvyNbC3bz3RL0r59+1S7dm1du3ZNefPm1fLly81fOZo1a6ZnnnlGRYoUUWRkpEaNGqX69etr165dcnZ2VnR0tJycnOTl5WW1Tn9/f0VHR6e7zfHjx2vMmDE23S8AAPDvMHLkSPXu3Vtnz57VqlWr9PLLL+vKlSvm7ZH+85//qGLFiurTp88912WxWO5rHiDdfo306NFDH3zwga5du6aSJUvqySefTLPt9u3b1bhxY4WGhmrOnDnmdbYrVqzQ2LFjMxRM/vnnH/Os5b34+FgPWHanO/Sd7cTExEjS/79v0r61WNz/h974C+flmz/1COPJQ3hGlCwbcteB1Fq266Skm0n6asl8DX7hed26dUvlKlbRy6+9rtpP3b5eedhb78q/QKDWrlquebOmydnZRXXq1terb7ylIsHFFHvu9o8Xk98ZpcnvjEpzO+fOnbPej3scq8waNGiQZs6cqWHDhqlu3bry8vKSnZ2devfufV/rLFy4sI4cOaKrV69ane0ePHiweeeGlD8cxMTEaNWqVenehSGzxyD56yW9W9HdWecHH3ygQoUKadmyZXrvvffk4uKiJk2aaOLEiSpRokSG9vlBPBKhu1SpUtqzZ4/i4uL05Zdfqlu3btq0aZPKli2rjh07mu1CQkJUrVo1FSlSRN9++22a3VbuMAzjrv8wjRgxQoMGDTKfX7x4UUFBQVmzQwAA4F+lcOHCZvC4c43jiBEjzO80a9as0ebNmxUfH2+13PXr1xUXFyc3Nzc5OjrKx8dH165dS/VFVrrdvbRq1aoPZ4eQo3Xv3l2jR4/W7NmzNXbs2HTbLV26VI6Ojvrmm2/k4uJiTl+xYkWGt5U/f/4sG8DY19dX0u33TsXKVXQrT+ow/Fix2+NAeXp569w/Manmx/5z77PzmdWmY1e16dhVV69e0e5ft2rWpHfVv0cnfb1phwILFVaePG56afAIvTR4hGL/OavNG37QtHfH6JUenbVy43Yz1Pfq96oaNPvfpbV2V2NVLPgxScr0NdWZtXDhQj3//PMaN26c1fRz584pX758mV5fo0aN9P333+u7775T+/btzelBQUFmpko5Dpevr68qVKiQ7mvybr2A0pL89ZJeLrxzXN3c3DRmzBiNGTNGMTEx5lnvli1bpjngX1Z7JEK3k5OTOZBatWrVtGPHDk2bNk0fffRRqrYFChRQkSJFdOTI7a4lAQEBun79ui5cuGB1tvvs2bOqUyf9azqcnZ3vOtgAAADA/apRo4Zmz56tY8eOaf/+/bp582aa113OmTNHc+bM0fLly9WmTRvzWu59+/apZs2aZrvo6GidO3dOISEhD20fkHMVLFhQr732mg4fPqxu3bql285iscjBwcGqC29CQoIWLFiQqq2zs3OaZ0SbNWum0aNHa/369apfv/4D1V2qVCmVKFFCv//+u7o+101J7unfK7t67Se0ad0axf5z1jy7nZSUlOYo51klTx43PVGvkW5cv6GBvbvqrz8PK7CQ9Vl+n/x+at2hi/48uF8L536ohISreqxYCRUOLqY/D+3XK8NHm23tL8WoXJn7D9vp/U3SYrFYUmWfb7/9VqdOnTJzWGb07t1b77//voYOHarHH3883RHMkwsLC9N3332nYsWKpeqlfD+Sv15S/phwN/7+/urevbt+//13TZ06Nc0fObPaIxG6UzIMQ4mJiWnOi42N1YkTJ8w/bNWqVeXo6Kh169apQ4cOkqQzZ85o//79mjBhwkOrGQAA4I4NGzbIzs5ORYsWVffu3RUaGpqqTb169dSmTRsNGDDADNNNmzaVi4uLIiIirEL3ndsItWnT5iHtAXK6d999955tWrRoocmTJ6tLly7q27evYmNj9f7776d5Yqp8+fJaunSpli1bpqJFi8rFxUXly5fXwIEDtWzZMrVu3VrDhw9XjRo1lJCQoE2bNiksLCzVLaPu5aOPPlKzZs3Ut08vterUXX4BBRQfF6fIo3/o0P69en92hCSpzytDtGndGvXu1FovDHhNLq6uWjZ/rhKuXsnU9g7u3SP3ZPc0v6NoiVLK6+6hMUMHyNnFRZWq1VR+P3+d++es5s6cIncPD5WrWEWS1LVlQz3VoIlKliknD09PHTv6p775apkqVq0uV9fbYW7Uu1PU7/ln9J+u7dT6mc7yCyigS2ei9PWKr7R792598cUXmapbSv9vkpawsDBFRESodOnSqlChgnbt2qWJEyeqUKFCmd6uJOXLl08rVqxQy5YtVbFiRb344ouqVauW8ubNq9jYWP3000+Kjo62Ogn61ltvad26dapTp45eeeUVlSpVSteuXVNUVJS+++47zZ49O9P13Hm9NGnSRN27d1fBggV1/vx5HTp0yOq41qxZU2FhYapQoYK8vLx06NAhLViwQLVr17Z54JYegdA9cuRINWvWTEFBQbp06ZKWLl2qjRs3as2aNbp8+bLCw8PVrl07FShQQFFRURo5cqR8fX3NW214enqqV69eGjx4sHx8fOTt7a0hQ4aofPnyGRrQAQAA4H717dtXHh4eqlGjhvz9/XXu3Dl98cUXWrZsmV577TXlz59f+fPnt7qncXIFCxa0CuTe3t564403NGrUKHl7e6tx48basWOHwsPD1bt3b24XhixVv359ffrpp3rvvffUsmVLFSxYUH369JGfn5969epl1XbMmDE6c+aM+vTpo0uXLqlIkSKKioqSu7u7Nm/erPDwcH388ccaM2aMvLy8VL16dfXt2zfTNdWrV0/bt2/X0GHDNWHMSF2Mj1M+L28VLVFKjcPamO1KlC6rj5Ys16S3R2nUoJfk4ZlPYW07qGHzlnpr2MAMb++l59qnOf2jxctV68lQValRWyu/WKzvv1nx/7X4qHL1Who75UN5+9zu3lzj8Se1ad1qLfxklq5dS5CffwGFteukPq8MNtdXo86TWrTqB835YPL/9ssznypUKG+eOMys9P4maZk2bZocHR01fvx4Xb58WVWqVNFXX32lN9544762LUm1atXS/v37NW3aNK1YsUKTJk3S9evXlT9/flWtWlVz5swx7x0v3e6xvHPnTr399tuaOHGiTp48KXd3dwUHB6tp06b3dfb7zutl7NixGjhwoC5cuCAfHx+VLVvW6rjWr19fX3/9taZMmaKrV6+qYMGCev755/X666/f9/5nhsUwDOOhbCkdvXr10o8//qgzZ87I09NTFSpU0LBhw9SoUSMlJCSoTZs2+u233xQXF6cCBQqoXr16evvtt62uv7527Zpee+01LV68WAkJCWrQoIFmzZqVqWu0L168KE9PT8XHx8sjjV+7AAAAUpo3b57mzZunQ4cOKS4uTnnz5lXFihXVu3dvczCh9FgsFvXr1y/VrcWk24P+zJw5U1FRUQoICDDv053eAES4t2vXrikyMlLBwcFW1y9L0nNPN1F8zN/ZVNn/ePoX1oLla7O7jEfCgUN/3LV7eW7woN3LYXt3+9yQMp4hsz10PyoI3QAAALnXvb4849FC6MajIKtC9yN1n24AAAAAAHITQjcAAAAAADZC6AYAAAAAwEYI3QAAAAAA2AihGwAAAP8ajCEMIKOy6vOC0A0AAIBc787t1q5evZrNlQDIKa5cuSKLxfLAt2t0yKJ6AAAAcrSmbTroZMy57C7Dpgr5+2rNis+zu4xsYW9vr3z58uns2bOSpDx58shisWRzVUjPrVu3ZNy8nt1l2NStW7d07dq17C4DKRiGoZs3b+rixYu6ePGi8uXLJ3t7+wdaJ6EbAABA0smYc7pcd3B2l2FTJzdNyu4SslVAQIAkmcEbj66zZ2N0yzkhu8uwKbvEi3Kg3/Ejy97eXgUKFJCnp+cDr4vQDQAAgH8Fi8WiAgUKyM/PTzdu3MjucnAXr4wI19XqPbK7DJvKs2OeVi2dn91lIA0ODg6yt7fPst4whG4AAAD8q9jb2z9wd1HY1pmYf3T5UlJ2l2FTeWP+kYuLS3aXgYeADg0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbCTbQ/eHH36oChUqyMPDQx4eHqpdu7ZWr15tzjcMQ+Hh4QoMDJSrq6tCQ0N14MABq3UkJiaqf//+8vX1lZubm1q1aqWTJ08+7F0BAAAAAMBKtofuQoUK6d1339XOnTu1c+dO1a9fX61btzaD9YQJEzR58mTNmDFDO3bsUEBAgBo1aqRLly6Z6xg4cKCWL1+upUuXavPmzbp8+bLCwsKUlJSUXbsFAAAAAED2h+6WLVuqefPmKlmypEqWLKmxY8cqb9682rZtmwzD0NSpU/X666+rbdu2CgkJ0fz583X16lUtXrxYkhQfH6+5c+dq0qRJatiwoSpXrqyFCxdq3759+uGHH7J57wAAAAAA/2bZHrqTS0pK0tKlS3XlyhXVrl1bkZGRio6OVuPGjc02zs7Oqlu3rrZu3SpJ2rVrl27cuGHVJjAwUCEhIWYbAAAAAACyg0N2FyBJ+/btU+3atXXt2jXlzZtXy5cvV9myZc3Q7O/vb9Xe399fx48flyRFR0fLyclJXl5eqdpER0enu83ExEQlJiaazy9evJhVuwMAAAAAgKRH5Ex3qVKltGfPHm3btk0vvviiunXrpoMHD5rzLRaLVXvDMFJNS+lebcaPHy9PT0/zERQU9GA7AQAAAABACo9E6HZyclLx4sVVrVo1jR8/XhUrVtS0adMUEBAgSanOWJ89e9Y8+x0QEKDr16/rwoUL6bZJy4gRIxQfH28+Tpw4kcV7BQAAAAD4t3skQndKhmEoMTFRwcHBCggI0Lp168x5169f16ZNm1SnTh1JUtWqVeXo6GjV5syZM9q/f7/ZJi3Ozs7mbcruPAAAAAAAyErZfk33yJEj1axZMwUFBenSpUtaunSpNm7cqDVr1shisWjgwIEaN26cSpQooRIlSmjcuHHKkyePunTpIkny9PRUr169NHjwYPn4+Mjb21tDhgxR+fLl1bBhw2zeOwAAAADAv1m2h+6YmBg999xzOnPmjDw9PVWhQgWtWbNGjRo1kiQNHTpUCQkJeumll3ThwgXVrFlT33//vdzd3c11TJkyRQ4ODurQoYMSEhLUoEEDRUREyN7ePrt2CwAAAACA7A/dc+fOvet8i8Wi8PBwhYeHp9vGxcVF06dP1/Tp07O4OgAAAAAA7t8jeU03AAAAAAC5AaEbAAAAAAAbIXQDAAAAAGAjhG4AAAAAAGyE0A0AAAAAgI0QugEAAAAAsBFCNwAAAAAANkLoBgAAAADARgjdAAAAAADYCKEbAAAAAAAbIXQDAAAAAGAjhG4AAAAAAGyE0A0AAAAAgI0QugEAAAAAsBFCNwAAAAAANkLoBgAAAADARgjdAAAAAADYCKEbAAAAAAAbIXQDAAAAAGAjhG4AAAAAAGyE0A0AAAAAgI0QugEAAAAAsBFCNwAAAAAANkLoBgAAAADARgjdAAAAAADYCKEbAAAAAAAbIXQDAAAAAGAjhG4AAAAAAGyE0A0AAAAAgI0QugEAAAAAsBFCNwAAAAAANkLoBgAAAADARgjdAAAAAADYCKEbAAAAAAAbIXQDQA6zfv169ezZU6VLl5abm5sKFiyo1q1ba9euXVbtunfvLovFkupRunTpu67/4MGDcnZ2lsVi0c6dO225KwAAALmeQ3YXAADInA8//FCxsbEaMGCAypYtq3/++UeTJk1SrVq1tHbtWtWvX99s6+rqqvXr11st7+rqmu66k5KS1LNnT/n6+ur06dM22wcAAIB/C0I3AOQwM2fOlJ+fn9W0pk2bqnjx4ho3bpxV6Lazs1OtWrUyvO4pU6bo5MmTGjZsmAYMGJBlNQMAAPxbEboBIIdJGbglKW/evCpbtqxOnDhx3+s9cuSIRo8eraVLl+r8+fMPUiIAAAD+H9d0A0AuEB8fr927d6tcuXJW0xMSEhQQECB7e3sVKlRIL7/8cpqB2jAM9e7dW2FhYWrVqtXDKhsAACDX40w3AOQC/fr105UrV/T666+b0ypWrKiKFSsqJCREkrRp0yZNmTJFP/74o3bs2KG8efOabWfOnKl9+/bp888/f+i1AwAA5GaEbgDI4UaNGqVFixZp+vTpqlq1qjn91VdftWrXqFEjVa5cWe3bt9ecOXPM+cePH9eIESM0depU+fv7P9TaAQAAcjtCNwDkYGPGjNE777yjsWPH6uWXX75n+6efflpubm7atm2bOa1fv34KCQlRu3btFBcXJ0m6evWqJOny5cuKj4+Xp6enTeoHAADI7QjdAJBDjRkzRuHh4QoPD9fIkSMzvJxhGLKz+9+QHvv379fx48fl5eWVqm29evXk6elphnEAAABkDqEbAHKgt99+W+Hh4XrjjTf05ptvZni5//73v7p69arVbcSWLl2qa9euWbVbs2aN3nvvPc2ePTvV4GwAAADIOEI3AOQwkyZN0ujRo9W0aVO1aNHCqqu4JNWqVUvHjx9Xly5d1KlTJxUvXlwWi0WbNm3S1KlTVa5cOfXu3duqfUpRUVGSpKpVq6patWo23R8AAIDcjNANADnMqlWrJN0+G71mzZpU8w3DkIeHh/z9/TV58mTFxMQoKSlJRYoU0SuvvKKRI0fKzc3tYZcNAADwr0ToBoAcZuPGjfds4+Xlpa+++uq+t9G9e3d17979vpcHAADAbXb3bgIAAAAAAO4HoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2ku336R4/fry++uorHT58WK6urqpTp47ee+89lSpVymzTvXt3zZ8/32q5mjVratu2bebzxMREDRkyREuWLFFCQoIaNGigWbNmqVChQg9tXwAgM5q26aCTMeeyuwybKuTvqzUrPs/uMgAAALJNtofuTZs2qV+/fqpevbpu3ryp119/XY0bN9bBgwfl5uZmtmvatKnmzZtnPndycrJaz8CBA7Vq1SotXbpUPj4+Gjx4sMLCwrRr1y7Z29s/tP0BgIw6GXNOl+sOzu4ybOrkpknZXQIAAEC2yvbQvWbNGqvn8+bNk5+fn3bt2qWnnnrKnO7s7KyAgIA01xEfH6+5c+dqwYIFatiwoSRp4cKFCgoK0g8//KAmTZrYbgcAAAAAAEjHI3dNd3x8vCTJ29vbavrGjRvl5+enkiVLqk+fPjp79qw5b9euXbpx44YaN25sTgsMDFRISIi2bt36cAoHAAAAACCFbD/TnZxhGBo0aJCeeOIJhYSEmNObNWumZ555RkWKFFFkZKRGjRql+vXra9euXXJ2dlZ0dLScnJzk5eVltT5/f39FR0enua3ExEQlJiaazy9evGibnQIAAAAA/Gs9UqH75Zdf1t69e7V582ar6R07djT/PyQkRNWqVVORIkX07bffqm3btumuzzAMWSyWNOeNHz9eY8aMyZrCAQAAAABIwyPTvbx///76+uuvtWHDhnuOOF6gQAEVKVJER44ckSQFBATo+vXrunDhglW7s2fPyt/fP811jBgxQvHx8ebjxIkTWbMjAAAAAAD8v2wP3YZh6OWXX9ZXX32l9evXKzg4+J7LxMbG6sSJEypQoIAkqWrVqnJ0dNS6devMNmfOnNH+/ftVp06dNNfh7OwsDw8PqwcAAAAAAFkpS7qXnzhxQgcOHFD16tXl4+OTqWX79eunxYsXa+XKlXJ3dzevwfb09JSrq6suX76s8PBwtWvXTgUKFFBUVJRGjhwpX19fPf3002bbXr16afDgwfLx8ZG3t7eGDBmi8uXLm6OZAwAAAADwsGU6dL/xxhu6cuWKpkyZIkn64Ycf1LJlS12/fl358uXTTz/9pHLlymV4fR9++KEkKTQ01Gr6vHnz1L17d9nb22vfvn367LPPFBcXpwIFCqhevXpatmyZ3N3dzfZTpkyRg4ODOnTooISEBDVo0EARERHcoxsAAAAAkG0yHbq//PJLDRo0yHz+xhtvqEKFCnr99dcVHh6ud955R0uWLMnw+gzDuOt8V1dXrV279p7rcXFx0fTp0zV9+vQMbxsAAAAAAFvKdOg+deqUihcvLun2tdU7duzQd999pyZNmujatWsaPHhwlhcJAAAAAEBOlOmB1AzD0K1btyRJW7Zskb29vZ566ilJt0cVP3fuXNZWCAAAAABADpXp0F2sWDF98803kqSlS5eqRo0acnV1lXR7xHAvL6+srRCPhPXr16tnz54qXbq03NzcVLBgQbVu3Vq7du0y2yQlJWny5Mlq2rSpChUqpDx58qhMmTIaPny44uLiUq3zzJkz6t69u/z8/OTi4qIKFSpo7ty5D3GvAAAAAMC2Mt29/IUXXlC/fv3Mgc0+/fRTc96WLVtUtmzZLC0Qj4YPP/xQsbGxGjBggMqWLat//vlHkyZNUq1atbR27VrVr19fCQkJCg8PV+fOndW7d2/5+vpq9+7deuedd7Rq1Srt3LnT/IEmPj5eTzzxhK5fv64JEyaoQIECWrJkiXr37q34+HircQMAAAAAIKfKdOh+8cUX5eXlpa1bt6pGjRp69tlnzXkJCQnq3r17VtaHR8TMmTPl5+dnNa1p06YqXry4xo0bp/r168vV1VWRkZFWt40LDQ1V4cKF9cwzz+jLL780Xy8ffvihjh07pp07d6pq1aqSpCZNmujMmTMaPXq0evbsqXz58j20/QMAAAAAW7iv+3R36tRJnTp1SjX9448/fuCC8GhKGbglKW/evCpbtqxOnDghSbK3t0/zPu01atSQJLOddLtXhL+/vxm47wgLC9P333+vNWvWpPkaAwAAAICcJNPXdN9x+PBhffTRRxo7dqyio6MlSadPn1ZCQkKWFYdHW3x8vHbv3n3P+7KvX79ekqzaXb9+Xc7Ozqna3pm2d+/eLKwUAAAAALJHpkN3UlKSevXqpXLlyunFF1/U6NGjdfr0aUm3r/ceP358lheJR1O/fv105coVvf766+m2OXXqlIYPH65q1aopLCzMnF62bFmdPHlSf//9t1X7zZs3S7p9OzoAAAAAyOkyHbrHjh2rxYsXa+LEidq/f78MwzDnNWvWTGvWrMnSAvFoGjVqlBYtWqQpU6ak6iJ+x/nz59W8eXMZhqFly5bJzu5/L7e+ffvK0dFRXbt21YEDBxQbG6uZM2dq2bJlkmTVFgAAAAByqkwnm4iICI0aNUqDBg1SqVKlrOYFBwcrMjIyy4rDo2nMmDF65513NHbsWL388stptrlw4YIaNWqkU6dOad26dSpatKjV/DJlymj58uU6fvy4QkJC5Ovrq/fee0+TJk2SJBUsWNDm+wEAAAAAtpbp0H3q1CnVrl07zXkuLi66dOnSAxeFR9eYMWMUHh6u8PBwjRw5Ms02Fy5cUMOGDRUZGal169apQoUKabZr1qyZjh8/rj///FMHDx60Gvn8qaeestk+AAAAAMDDkunQ7efnp2PHjqU5748//lChQoUeuCg8mt5++22Fh4frjTfe0JtvvplmmzuB+9ixY/r+++9VuXLlu67TYrGoRIkSKlOmjJKSkjRt2jRVqlSJ0A0AAAAgV8j0LcOaN2+usWPHqmnTpgoICJB0OzjFx8frgw8+UMuWLbO8SGS/SZMmafTo0WratKlatGihbdu2Wc2vVauWEhIS1KRJE/3222+aOnWqbt68adUuf/78KlasmPm8f//+Cg0NlY+Pj44dO6YPPvhAJ0+e1KZNmx7afgEAAACALWU6dL/11ltavXq1ypYtq3r16slisWjkyJHav3+/HB0dNWrUKFvUiWy2atUqSdKaNWvSHCzPMAzFxMRox44dkqQBAwakatOtWzdFRESYz0+cOKH+/fvr3Llz8vHxUdOmTbVy5UoVKVLENjsBAAAAAA9ZpkO3v7+/duzYoTfffFPffvut7O3t9fvvvyssLExvvfWWvL29bVEnstnGjRvv2eaxxx6zGs3+XlasWHH/BQEAAABADpDp0C3dDt6zZ8/O6loAAAAAAMhVuBkyAAAAAAA2cl9nujdv3qzFixfr+PHjSkhIsJpnsVj0448/ZklxAAAAAADkZJkO3fPmzVOvXr3k7e2tkiVLytnZ2Wp+Zq7pBQAAAAAgN8t06J4wYYI6dOig+fPnpwrcAAAAAADgfzJ9Tffx48fVu3dvAjcAAAAAAPeQ6TPdZcqUUUxMjC1qwUPQtE0HnYw5l91l2FQhf1+tWfF5dpcBAAAAAJkP3ePGjdOQIUMUGhqqggUL2qIm2NDJmHO6XHdwdpdhUyc3TcruEgAAAABA0n2E7pkzZyo+Pl4lS5ZUpUqV5OPjYzXfYrFo5cqVWVYgAAAAAAA5VaZD9969e2Vvby8/Pz+dPn1ap0+ftppvsViyrDgAAAAAAHKyTIfuqKgoG5QBAAAAAEDuk+nRywEAAAAAQMbcV+hOTEzURx99pM6dO6tRo0Y6cuSIJGnlypU6duxYlhYIAAAAAEBOlenu5efOnVO9evV04MABBQQEKCYmRpcuXZIkrVixQmvXrtWsWbOyvFAAAAAAAHKaTJ/pHjp0qOLi4rRz5079/fffMgzDnFevXj1t2rQpSwsEAAAAACCnyvSZ7m+++UbvvfeeqlSpoqSkJKt5hQoV0smTJ7OsOAAAAAAAcrJMn+m+ePGiihQpkua8Gzdu6ObNmw9cFAAAAAAAuUGmQ3dwcLB++eWXNOdt375dpUqVeuCiAAAAAADIDTIdurt27ar33ntPK1euNK/ntlgs2rFjh6ZNm6bnnnsuy4sEAAAAACAnyvQ13cOGDdOWLVv09NNPy8vLS5LUpEkTxcbGqmnTphowYECWFwkAAAAAQE6U6dDt6Oio7777TsuWLdO3336rmJgY+fr6KiwsTJ06dZKd3X3d+hsAAAAAgFwn06Fbut2dvFOnTurUqVNW1wMAAAAAQK7BaWkAAAAAAGwkQ6H78uXL6tatmxYvXmxOu3Xrljw8PKweQUFBiouLs1WtAAAAAADkKBkK3YsXL9bXX3+tZs2amdMMw9Dly5cVGhqqdu3aqV27dpKkTz75xDaVAgAAAACQw2Tomu4lS5aoW7du5mjlyYWHh6tKlSqSpPfee0/ffPONhgwZkrVVAgAAAACQA2XoTPe+fftUv379e7arUKGCDhw48MBFAQAAAACQG2QodF+8eFE+Pj5W0+zt7bV8+XIVL17cnObh4aGLFy9mbYUAAAAAAORQGepe7uHhobNnz6aa3rp1a6vnZ8+eVd68ebOmMgAAAAAAcrgMnekOCQnR2rVr79lu7dq1CgkJeeCiAAAAAADIDTIUup955hnNnz9fW7ZsSbfN5s2bNX/+fHXs2DHLigMAAAAAICfLUPfyPn366NNPP1WDBg30wgsvqGXLlipSpIgk6fjx4/r66681Z84clStXTn369LFpwQAAAAAA5BQZCt1OTk5au3atunbtqunTp2vGjBlW8w3DUOPGjbVw4UI5OjrapFAAAAAAAHKaDIVuSfL19dXatWu1c+dOrVu3TidOnJAkFS5cWI0aNVLVqlVtViQAAAAAADlRhkP3HdWqVVO1atVsUQsAAAAAALlKhgZSAwAAAAAAmUfoBgAAAADARgjdAAAAAADYCKEbAAAAAAAbyVDoHjRokDla+d9//60bN25kWQHjx49X9erV5e7uLj8/P7Vp00Z//PGHVRvDMBQeHq7AwEC5uroqNDRUBw4csGqTmJio/v37y9fXV25ubmrVqpVOnjyZZXUCAAAAAJBZGQrdU6dO1ZkzZyRJwcHB+u2337KsgE2bNqlfv37atm2b1q1bp5s3b6px48a6cuWK2WbChAmaPHmyZsyYoR07diggIECNGjXSpUuXzDYDBw7U8uXLtXTpUm3evFmXL19WWFiYkpKSsqxWAAAAAAAyI0O3DPPy8lJMTIyk22edLRZLlhWwZs0aq+fz5s2Tn5+fdu3apaeeekqGYWjq1Kl6/fXX1bZtW0nS/Pnz5e/vr8WLF+uFF15QfHy85s6dqwULFqhhw4aSpIULFyooKEg//PCDmjRpkmX1AgAAAACQURkK3bVq1VKvXr1Uo0YNSdLgwYOVL1++NNtaLBatXLnyvguKj4+XJHl7e0uSIiMjFR0drcaNG5ttnJ2dVbduXW3dulUvvPCCdu3apRs3bli1CQwMVEhIiLZu3UroBgAAAABkiwyF7lmzZmngwIE6cOCALBaLjh49Kmdn5zTbPshZcMMwNGjQID3xxBMKCQmRJEVHR0uS/P39rdr6+/vr+PHjZhsnJyd5eXmlanNn+ZQSExOVmJhoPr948eJ91w0AAAAAQFoyFLqLFCmi5cuXS5Ls7Oy0YsUK86x3Vnr55Ze1d+9ebd68OdW8lGE+I93c79Zm/PjxGjNmzP0XCwAAAADAPWT6lmEbNmxQ2bJls7yQ/v376+uvv9aGDRtUqFAhc3pAQIAkpTpjffbsWfPsd0BAgK5fv64LFy6k2yalESNGKD4+3nzcGZ0dAAAAAICskunQXbduXeXNm1dHjx7Vxx9/rPHjx2vOnDk6evTofRVgGIZefvllffXVV1q/fr2Cg4Ot5gcHBysgIEDr1q0zp12/fl2bNm1SnTp1JElVq1aVo6OjVZszZ85o//79ZpuUnJ2d5eHhYfUAAAAAACArZah7eXKGYah///6aPXu2bt26ZU63s7PTSy+9pA8++CBT6+vXr58WL16slStXyt3d3Tyj7enpKVdXV1ksFg0cOFDjxo1TiRIlVKJECY0bN0558uRRly5dzLa9evXS4MGD5ePjI29vbw0ZMkTly5c3RzMHAAAAAOBhy3TonjJlimbNmqUXX3xR3bt3V2BgoE6fPq358+dr1qxZCg4O1quvvprh9X344YeSpNDQUKvp8+bNU/fu3SVJQ4cOVUJCgl566SVduHBBNWvW1Pfffy93d3eruhwcHNShQwclJCSoQYMGioiIkL29fWZ3EQAAAACALJHp0P3JJ5+of//+mjZtmjmtYMGCql69uuzt7TVnzpxMhW7DMO7ZxmKxKDw8XOHh4em2cXFx0fTp0zV9+vQMbxsAAAAAAFvK9DXdx44dU1hYWJrzwsLCdOzYsQcuCgAAAACA3CDTodvT09O8P3ZKx48fZ0AyAAAAAAD+X6ZDd6NGjfTGG29o165dVtP37NmjN998U02aNMmy4gAAAAAAyMkyHbrHjx8vBwcH1ahRQ+XLl1fjxo1Vvnx5Va1aVXZ2dho/frwt6gQAAAAAIMfJdOgOCgrSnj17NHToULm5uSkyMlJubm4aPny4fvvtNxUqVMgWdQIAAAAAkONkevRySfL19eWMNgAAAAAA95DpM90AAAAAACBjCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOZDt3Xr1+XYRi2qAUAAAAAgFwlU6H72rVrcnV11YoVK2xUDgAAAAAAuUemQreLi4t8fHzk5uZmq3oAAAAAAMg1Mt29vGXLllq+fLktagEAAAAAIFdxyOwCnTp1Uq9evdSzZ0+1bdtWBQoUkMVisWpTpUqVLCsQAAAAAICcKtOhu0mTJpKkiIgIzZ8/32qeYRiyWCxKSkrKmuoAAAAAAMjBMh26582bZ4s6AAAAAADIdTIdurt162aLOgAAAAAAyHUyPZBacn/88Ye2bNmiK1euZFU9AAAAAADkGvcVuj/77DMVKlRIZcuW1VNPPaU//vhDktShQwfNmTMnSwsEAAAAACCnynTo/uKLL9S9e3dVqVJFM2bMkGEY5rwqVaro888/z9ICAQAAAADIqTIdusePH68ePXro66+/Vt++fa3mlSlTRgcPHsyy4gAAAAAAyMkyHboPHTqkTp06pTnP29tbsbGxD1wUAAAAAAC5QaZDd548eRQfH5/mvFOnTsnLy+uBiwIAAAAAIDfIdOh+/PHHU13LfUdERIRCQ0Ozoi4AAAAAAHK8TN+ne/To0XriiSdUo0YNdenSRRaLRV999ZXefPNN/fTTT9q+fbst6gQAAAAAIMfJ9JnuatWqafXq1bp8+bIGDx4swzA0btw4/fnnn/ruu+8UEhJiizoBAAAAAMhxMn2mW5Lq1aunQ4cO6a+//lJMTIx8fX1VsmTJrK4NAAAAAIAc7b5C9x3FihVTsWLFsqoWAAAAAABylUx3L5ekqKgovfDCCypZsqR8fHxUsmRJvfDCC4qMjMzq+gAAAAAAyLEyHbr37NmjypUrKyIiQgULFlTjxo1VsGBBRUREqHLlytqzZ48NygQAAAAAIOfJdPfygQMHKn/+/Prhhx9UuHBhc/rx48fVqFEjvfrqq9qwYUOWFgkAAAAAQE6U6TPd27dv15gxY6wCtyQVKVJE4eHh+vXXX7OsOAAAAAAAcrJMh25PT095enqmOS9fvnzy8PB44KIAAAAAAMgNMh26u3Tpok8++STNeXPmzFHnzp0fuCgAAAAAAHKDDF3T/dVXX5n/X7VqVf33v/9VjRo11LlzZwUEBCg6OlpLlizR2bNn9cwzz9isWAAAAAAAcpIMhe727dvLYrHIMAzzvydOnNDOnTtTtX3uuefUpUuXLC8UAAAAAICcJkOhm9HIAQAAAADIvAyF7rp169q6DgAAAAAAcp1MD6QGAAAAAAAyJkNnulNasWKFFi1apOPHj+vatWtW8ywWi37//fcsKQ4AAAAAgJws06F74sSJGjZsmPLnz6/ixYvLzc3NFnUBAAAAAJDjZTp0z5o1Sz179tRHH30ke3t7W9QEAAAAAECukOlrumNjY9WlSxcCNwAAAAAA95Dp0P3444/r0KFDtqgFAAAAAIBcJdPdy6dOnaqnn35aQUFBatq0qZycnGxRFwAAAAAAOV6mQ3fx4sXVsGFDPf3007JYLMqTJ4/VfIvFovj4+CwrEAAAAACAnCrToXvo0KGaMWOGKlWqpDJlynCmGwAAAACAdGQ6dEdERGjYsGEaP368LeoBAAAAACDXyPRAaklJSWrUqJEtagEAAAAAIFfJdOhu3Lixtm3bZotaAAAAAADIVTLdvXzUqFHq2LGj3Nzc1KJFC3l7e6dqk9Y0AAAAAAD+bTIduitWrChJGjRokAYNGpRmm6SkpAerCgAAAACAXCDToXv06NGyWCy2qAUAAAAAgFwl06E7PDzcBmUAAAAAAJD7ZHogtaz2008/qWXLlgoMDJTFYtGKFSus5nfv3l0Wi8XqUatWLas2iYmJ6t+/v3x9feXm5qZWrVrp5MmTD3EvAAAAAABILdNnut966627zrdYLBo1alSG13flyhVVrFhRPXr0ULt27dJs07RpU82bN8987uTkZDV/4MCBWrVqlZYuXSofHx8NHjxYYWFh2rVrl+zt7TNcCwAAAAAAWSnLu5dnNnQ3a9ZMzZo1u2sbZ2dnBQQEpDkvPj5ec+fO1YIFC9SwYUNJ0sKFCxUUFKQffvhBTZo0yXAtAAAAAABkpUx3L79161aqx7lz5/TJJ58oJCREUVFRWV7kxo0b5efnp5IlS6pPnz46e/asOW/Xrl26ceOGGjdubE4LDAxUSEiItm7dmu46ExMTdfHiRasHAAAAAABZKUuu6fb29lbPnj3VpUsXvfLKK1mxSlOzZs20aNEirV+/XpMmTdKOHTtUv359JSYmSpKio6Pl5OQkLy8vq+X8/f0VHR2d7nrHjx8vT09P8xEUFJSldQMAAAAAkKUDqdWoUUM//vhjVq5SHTt2VIsWLRQSEqKWLVtq9erV+vPPP/Xtt9/edTnDMO56a7MRI0YoPj7efJw4cSJL6wYAAAAAIEtD9++//668efNm5SpTKVCggIoUKaIjR45IkgICAnT9+nVduHDBqt3Zs2fl7++f7nqcnZ3l4eFh9QAAAAAAICtleiC1zz77LNW0xMRE7d27V59++qmeffbZLCksPbGxsTpx4oQKFCggSapataocHR21bt06dejQQZJ05swZ7d+/XxMmTLBpLQAAAAAA3E2mQ3f37t3TnO7i4qJnn31W77//fqbWd/nyZR09etR8HhkZqT179sjb21ve3t4KDw9Xu3btVKBAAUVFRWnkyJHy9fXV008/LUny9PRUr169NHjwYPn4+Mjb21tDhgxR+fLlzdHMAQAAAADIDpkO3ZGRkammubi43LUr993s3LlT9erVM58PGjRIktStWzd9+OGH2rdvnz777DPFxcWpQIECqlevnpYtWyZ3d3dzmSlTpsjBwUEdOnRQQkKCGjRooIiICO7RDQAAAADIVpkO3UWKFMnSAkJDQ2UYRrrz165de891uLi4aPr06Zo+fXpWlgYAAAAAwAPJ0oHUAAAAAADA/2ToTHeFChUyvEKLxaLff//9vgsCAAAAACC3yFDo9vb2vus9r6XbA6Lt2rXrnu0AAAAAAPi3yFDo3rhxY7rzbt68qY8//lhvvfWWLBaLunTpklW1AQAAAACQoz3QNd1ffPGFypYtq/79+6tixYratWuXFixYkFW1AQAAAACQo91X6N64caNq1qypjh07ysPDQ99//73Wrl2rSpUqZXF5AAAAAADkXJkK3fv27VPz5s3VoEEDxcbGavHixdq5c6caNGhgq/oAAAAAAMixMhS6T5w4oW7duqlKlSratWuXpk6dqkOHDqlTp062rg8AAAAAgBwrQwOplSxZUtevX1fTpk01dOhQubu7a9++fem2r1KlSpYVCAAAAABATpWh0J2YmChJWr16tdasWZNuO8MwZLFYlJSUlDXVAQAAAACQg2UodM+bN8/WdQAAAAAAkOtkKHR369bN1nUAAAAAAJDrPNB9ugEAAAAAQPoI3QAAAAAA2AihGwAAAAAAGyF0AwAAAABgI4RuAAAAAABshNANAAAAAICNELoBAAAAALARQjcAAAAAADZC6AYAAAAAwEYI3QAAAAAA2AihGwAAAAAAGyF0AwAAAABgI4RuAAAAAABshNANAAAAAICNELoBAAAAALARQjcAAAAAADZC6AYAAAAAwEYI3QAAAAAA2AihGwAAAAAAGyF0AwAAAABgI4RuAAAAAABshNANAAAAAICNELoBAAAAALARQjcAAAAAADZC6AYAAAAAwEYI3QAAAAAA2AihGwAAAAAAGyF0AwAAAABgI4RuAAAAAABshNANAAAAAICNELoBAAAAALARQjcAAAAAADZC6AYAAAAAwEYI3QAAAAAA2AihGwAAAAAAGyF0AwAAAABgI4RuAAAAAABshNANAAAAAICNELoBAAAAALARQjcAAAAAADZC6AYAAAAAwEYI3QAAAAAA2Ei2h+6ffvpJLVu2VGBgoCwWi1asWGE13zAMhYeHKzAwUK6urgoNDdWBAwes2iQmJqp///7y9fWVm5ubWrVqpZMnTz7EvQAAAAAAILVsD91XrlxRxYoVNWPGjDTnT5gwQZMnT9aMGTO0Y8cOBQQEqFGjRrp06ZLZZuDAgVq+fLmWLl2qzZs36/LlywoLC1NSUtLD2g0AAAAAAFJxyO4CmjVrpmbNmqU5zzAMTZ06Va+//rratm0rSZo/f778/f21ePFivfDCC4qPj9fcuXO1YMECNWzYUJK0cOFCBQUF6YcfflCTJk0e2r4AAAAAAJBctp/pvpvIyEhFR0ercePG5jRnZ2fVrVtXW7dulSTt2rVLN27csGoTGBiokJAQs01aEhMTdfHiRasHAAAAAABZ6ZEO3dHR0ZIkf39/q+n+/v7mvOjoaDk5OcnLyyvdNmkZP368PD09zUdQUFAWVw8AAAAA+Ld7pEP3HRaLxeq5YRippqV0rzYjRoxQfHy8+Thx4kSW1AoAAAAAwB2PdOgOCAiQpFRnrM+ePWue/Q4ICND169d14cKFdNukxdnZWR4eHlYPAAAAAACy0iMduoODgxUQEKB169aZ065fv65NmzapTp06kqSqVavK0dHRqs2ZM2e0f/9+sw0AAAAAANkh20cvv3z5so4ePWo+j4yM1J49e+Tt7a3ChQtr4MCBGjdunEqUKKESJUpo3LhxypMnj7p06SJJ8vT0VK9evTR48GD5+PjI29tbQ4YMUfny5c3RzAEAAAAAyA7ZHrp37typevXqmc8HDRokSerWrZsiIiI0dOhQJSQk6KWXXtKFCxdUs2ZNff/993J3dzeXmTJlihwcHNShQwclJCSoQYMGioiIkL29/UPfHwAAAAAA7sj20B0aGirDMNKdb7FYFB4ervDw8HTbuLi4aPr06Zo+fboNKgQAAAAA4P480td0AwAAAACQkxG6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAABshdAMAAAAAYCOEbgAAAAAAbITQDQAAAACAjRC6AQAAAACwEUI3AAAAAAA2QugGAAAAAMBGCN0AAAAAANgIoRsAAAAAcrCNGzfKYrGk+di2bZskKSkpSZMnT1bTpk1VqFAh5cmTR2XKlNHw4cMVFxeXvTuQyzlkdwEAAAAAgAc3btw41atXz2paSEiIJCkhIUHh4eHq3LmzevfuLV9fX+3evVvvvPOOVq1apZ07d8rV1TU7ys71CN0AAAAAkAuUKFFCtWrVSnOeq6urIiMj5ePjY04LDQ1V4cKF9cwzz+jLL7/Us88++7BK/VehezkAAAAA5HL29vZWgfuOGjVqSJJOnDjxsEv61yB0AwAAAEAu0K9fPzk4OMjDw0NNmjTR5s2b77nM+vXrJUnlypWzdXn/WoRuAAAAAMjBPD09NWDAAH300UfasGGDpk2bphMnTig0NFRr165Nd7lTp05p+PDhqlatmsLCwh5ixf8uXNMNAAAAADlY5cqVVblyZfP5k08+qaefflrly5fX0KFD1aRJk1TLnD9/Xs2bN5dhGFq2bJns7DgfayscWQAAAADIZfLly6ewsDDt3btXCQkJVvMuXLigRo0a6dSpU1q3bp2KFi2aTVX+O3CmGwAAAAByIcMwJEkWi8WcduHCBTVs2FCRkZH68ccfVaFChewq71+D0A0AAAAAucyFCxf0zTffqFKlSnJxcTGnNWzYUMeOHdO6deusuqTDdgjdAAAAAJCDdenSRYULF1a1atXk6+urI0eOaNKkSYqJiVFERIQkKSEhQU2aNNFvv/2mqVOn6ubNm9q2bZu5jvz586tYsWLZtAe5G6EbAAAAAHKwChUqaNmyZZo9e7YuX74sb29vPfHEE1qwYIGqV68uSYqJidGOHTskSQMGDEi1jm7dupkBHVmL0A0AAAAAOdjw4cM1fPjwu7Z57LHHzGu88XAxejkAAAAAADZC6AYAAAAAwEYI3QAAAAAA2AihGwAAAAAAG3nkQ3d4eLgsFovVIyAgwJxvGIbCw8MVGBgoV1dXhYaG6sCBA9lYMQAAAAAAtz3yoVuSypUrpzNnzpiPffv2mfMmTJigyZMna8aMGdqxY4cCAgLUqFEjXbp0KRsrBgAAAAAgh9wyzMHBwers9h2GYWjq1Kl6/fXX1bZtW0nS/Pnz5e/vr8WLF+uFF1542KUCAAAAwD0l/P27WtUpk91l2JSnf2EtWL42u8vIdjkidB85ckSBgYFydnZWzZo1NW7cOBUtWlSRkZGKjo5W48aNzbbOzs6qW7eutm7dSugGAAAA8EhysyTq68ans7sMm2r1fXZX8Gh45EN3zZo19dlnn6lkyZKKiYnRO++8ozp16ujAgQOKjo6WJPn7+1st4+/vr+PHj991vYmJiUpMTDSfX7x4MeuLBwAAAAD8qz3y13Q3a9ZM7dq1U/ny5dWwYUN9++23km53I7/DYrFYLWMYRqppKY0fP16enp7mIygoKOuLR463Z88etWjRQoULF5arq6u8vb1Vu3ZtLVy40Kpd9+7dUw34Z7FYVLp06WyqHAAAAMCj4JE/052Sm5ubypcvryNHjqhNmzaSpOjoaBUoUMBsc/bs2VRnv1MaMWKEBg0aZD6/ePEiwRupxMXFKSgoSJ07d1bBggV15coVLVq0SM8995yioqL0xhtvmG1dXV21fv16q+VdXV0fdskAAAAAHiE5LnQnJibq0KFDevLJJxUcHKyAgACtW7dOlStXliRdv35dmzZt0nvvvXfX9Tg7O8vZ2flhlIwcLDQ0VKGhoVbTwsLCFBkZqY8//tgqdNvZ2alWrVoPuUIAAAAAj7JHvnv5kCFDtGnTJkVGRurXX39V+/btdfHiRXXr1k0Wi0UDBw7UuHHjtHz5cu3fv1/du3dXnjx51KVLl+wuHbmYr6+vHBxy3G9WAAAAAB6yRz41nDx5Up07d9a5c+eUP39+1ar1f+3de1RVdfrH8c/xAiQ3FQS8YRcyUFNLS81fKY0Q2oxmo6k4iJhaOYaNVpZooOZlTVNqo6UVqZmRmdrKcjBUuqhh2vGUl8lrqKOiAgZeEEH2748WZ80JNHA47gO8X2udtTzf/ZzvfvaOJ87DvnVVRkaGWrVqJUl6/vnnVVBQoDFjxujs2bPq0qWLvvjiC3l7e5ucOWqSkpISlZSU6OzZs1q5cqXWr1+v+fPnO8QUFBQoKChIZ86cUdOmTfXII49o2rRpaty4sUlZAwAAADCbyzfdH3744TWXWywWJSUlKSkp6cYkhFppzJgxWrRokSTJzc1Nr7/+usMj6Tp06KAOHTqoXbt2kqSvvvpKc+bM0caNG7V9+3Z5eXmZkjcAAAAAc7l80w24gkmTJmnkyJE6ffq01q5dq7Fjx+rChQt69tlnJUl/+9vfHOIjIiJ01113acCAAXr77bfLLAcAAABQO9B0AxUQHBys4OBgSVKfPn0k/XoH/NjYWDVp0qTcz/Tv31+enp7KyMi4YXkCAAAAcC0ufyM1wBXde++9Ki4u1uHDh68ZZxiG6tShzAAAAIDaim4AuA7p6emqU6eObr311qvGfPzxx7p48SKPEQMAAABqMU4vB65h9OjR8vHx0b333qvAwEBlZ2dr5cqVWrFihZ577jk1adJER44cUXR0tAYPHqyQkBBZLBZ99dVXmjt3rtq2bauRI0eavRkAAAAATELTDVxDt27dtHjxYi1dulS//PKLvLy81KFDBy1btkx/+ctfJEk+Pj4KDAzUa6+9plOnTunKlStq1aqV4uPjNWnSJHl6epq8FQAAAADMQtMNXENcXJzi4uKuGdOoUSOtXr36BmUEAAAAoDrhmm4AAAAAAJyEphsAAAAAACeh6QYAAAAAwElougEAAAAAcBKabgAAAAAAnISmGwAAAAAAJ+GRYahxCo7+oL73hZmdhlP5BgZr2Zr1ZqcBAAAA4HfQdKPG8bQU6tPIE2an4VR9vzA7AwAAAAAVwenlAAAAAAA4CU03AAAAAABOQtMNAAAAAICT0HQDAAAAAOAkNN0AAAAAADgJTTcAAAAAAE5C0w0AAAAAgJPQdAMAAAAA4CQ03QAAAAAAOAlNNwAAAAAATkLTDQAAAACAk9B0AwAAAADgJDTdAAAAAAA4CU03AAAAAABOQtMNAAAAAICT0HQDAAAAAOAkNN0AAAAAADgJTTcAAAAAAE5C0w0AAAAAgJPQdAMAAAAA4CQ03QAAAAAAOAlNNwAAAAAATkLTDQAAgCqxadMmjRgxQqGhofL09FTz5s3Vr18/ff/992anBgCmoekGAABAlXjzzTeVmZmpcePGad26dZo3b55Onz6trl27atOmTWanBwCmqGd2AgAAAKgZFixYoICAAIexqKgohYSEaObMmXrwwQdNygwAzMORbgBAtXf+/Hk988wzatasmTw8PNSxY0d9+OGHZqcF1Dq/bbglycvLS23atNGxY8dMyAgAzMeRbgBAtffoo49q+/btmj17tlq3bq0PPvhAQ4YMUUlJiaKjo81OD6jV8vLyZLVaOcoNoNai6QYAVGvr1q1TWlqavdGWpPDwcB05ckTPPfecBg0apLp165qcJVB7/fWvf9WFCxeUkJBgdioAYApOLwcAVGtr1qyRl5eXBg4c6DAeFxenEydOaNu2bSZlBmDKlClavny55syZo06dOpmdDgCYgqYbAFCt7d69W2FhYapXz/Hkrfbt29uXA7jxpk6dqpdfflkzZszQ2LFjzU4HAExD0w0AqNZycnLUuHHjMuOlYzk5OTc6JaDWmzp1qpKSkpSUlKRJkyaZnQ4AmIqmGwBQ7VkslutaBqDqTZ8+XUlJSZo8ebISExPNTgcATMeN1AAA1Zqfn1+5R7Nzc3Mlqdyj4ACc49VXX9VLL72kqKgoPfzww8rIyHBY3rVrV5MyAwDz0HQDAKq1O++8UykpKSouLna4rnvXrl2SpHbt2pmVGlDrrF27VpKUmpqq1NTUMssNw7jRKQGA6Ti9HABQrfXv31/nz5/XqlWrHMaXLl2qZs2aqUuXLiZlBtQ+X375pQzDuOoLAGojjnQDAKq13r17KyIiQk899ZTy8/MVEhKilJQUpaam6v333+cZ3QAAwFQ03QCAam/16tVKSEjQSy+9pNzcXIWGhiolJUWDBw82OzUAAFDL0XQDAKo9Ly8vzZs3T/PmzTM7FQAAAAdc0w0AAAAAgJPUqKb7jTfe0C233CIPDw916tRJ33zzjdkpAQAAAABqsRrTdK9YsULPPPOMEhIStHPnTt1///3q3bu3jh49anZqAAAAAIBaqsZc0/3aa6/p8ccf18iRIyVJc+fO1fr16/Xmm29q1qxZJmcHAABgvoKjP6jvfWFmp+FUvoHBWrZmvdlpAIBdjWi6L1++rO+//14vvPCCw3hkZKS2bt1qUlYAAACuxdNSqE8jT5idhlP1/cLsDADAUY1ourOzs3XlyhUFBgY6jAcGBiorK6vczxQWFqqwsND+Pi8vT5KUn5/vvERdwJXiYpUUXjQ7Dae6UmIov9AwOw2nKiq+UuN/VmuDWlGPxcX8rKLaqBU1ye9IVBPUY81Q0+uxdNsM49r/HWtE013KYrE4vDcMo8xYqVmzZmnq1Kllxlu2bOmU3FzKjpp/gznf2WZn4Gzn5Ovra3YSqAq1oR75WUV1Uhtqkt+RqC6oxxqgdtTjuXPX3s4a0XT7+/urbt26ZY5qnz59uszR71Ivvviixo8fb39fUlKi3Nxc+fn5XbVRh+vLz89Xy5YtdezYMfn4+JidDlCrUY+Aa6EmAddBPdYMhmHo3Llzatas2TXjakTT7ebmpk6dOiktLU39+/e3j6elpalfv37lfsbd3V3u7u4OYw0bNnRmmriBfHx8+B8Y4CKoR8C1UJOA66Aeq7+KHMmvEU23JI0fP14xMTHq3LmzunXrprfeektHjx7Vk08+aXZqAAAAAIBaqsY03YMGDVJOTo6mTZumkydPql27dlq3bp1atWpldmoAAAAAgFqqxjTdkjRmzBiNGTPG7DRgInd3dyUmJpa5dADAjUc9Aq6FmgRcB/VYu1iM37u/OQAAAAAAuC51zE4AAAAAAICaiqYbAAAAAAAnoelGlenZs6eeeeYZs9MAIOoRcDXUJOA6qEfcaDTdqDKrV6/W9OnT7e8PHjyouLg4tWjRQu7u7rrllls0ZMgQ7dixwx5jsVj0ySeflJlr+PDheuSRRxzGjh07pscff1zNmjWTm5ubWrVqpXHjxiknJ6dMHg899JD8/f1lsVhks9kclmdmZspisZT7WrlypUPs559/ri5duuimm26Sv7+/Hn300TK5LlmyRO3bt5eHh4eCgoI0duzYCu4xwHlcoR6Lioo0ceJE3XnnnfL09FSzZs00bNgwnThxwmGut956Sz179pSPj48sFot++eWXMjmcPXtWMTEx8vX1la+vr2JiYsrEjRs3Tp06dZK7u7s6duxY4X0F3AjVpSZzc3P19NNP64477lCDBg0UHBys+Ph45eXlOazParUqIiJCDRs2lJ+fn0aPHq3z5887xFCTcFWuUI+SlJSUpNDQUHl6eqpRo0bq1auXtm3b5hDTs2fPMt9XBw8e7BBTkXos73vvwoULK7Pb8D+g6UaVady4sby9vSVJO3bsUKdOnbR//34tWrRIe/fu1Zo1axQaGqoJEyZUeu7Dhw+rc+fO2r9/v1JSUnTw4EEtXLhQGzduVLdu3ZSbm2uPvXDhgrp3767Zs2eXO1fLli118uRJh9fUqVPl6emp3r172+NWrVqlmJgYxcXF6YcfftCWLVsUHR3tMNdrr72mhIQEvfDCC9qzZ482btyohx56qNLbB1Q1V6jHixcvymq1asqUKbJarVq9erX279+vvn37Osx38eJFRUVFadKkSVddZ3R0tGw2m1JTU5WamiqbzaaYmBiHGMMwNGLECA0aNKjS2wQ4W3WpyRMnTujEiRP6xz/+oV27dmnJkiVKTU3V448/7hDTq1cvhYSEaNu2bUpNTdWePXs0fPhwh7yoSbgqV6hHSWrdurXmz5+vXbt2afPmzbr55psVGRmpM2fOOMw5atQoh++tixYtsi+raD1K0uLFix3miY2NrfT24ToZQBXp0aOHMW7cOKOkpMRo27at0alTJ+PKlStl4s6ePWv/tyRjzZo1ZWJiY2ONfv362d9HRUUZLVq0MC5evOgQd/LkSaNBgwbGk08+WWaOn3/+2ZBk7Ny583dz79ixozFixAj7+6KiIqN58+bGO++8c9XP5ObmGjfddJOxYcOG350fuNFcrR5Lfffdd4Yk48iRI2WWpaenG5IccjIMw9i7d68hycjIyLCPffvtt4Yk46effiozT2JiotGhQ4er5gCYoTrWZKmPPvrIcHNzM4qKigzDMIxFixYZAQEBDvnv3LnTkGQcOHCgzOepSbgaV63HvLw8Q5LDd8vSXK+movV4tfxxY3CkG1XOZrNpz549mjBhgurUKfsj1rBhw0rNl5ubq/Xr12vMmDG66aabHJYFBQVp6NChWrFihYzrfPrd999/L5vN5vBXfKvVquPHj6tOnTq666671LRpU/Xu3Vt79uyxx6SlpamkpETHjx9XWFiYWrRooccee0zHjh27rjwAZ3C1eszLy5PFYqnUer/99lv5+vqqS5cu9rGuXbvK19dXW7durVT+gNmqY03m5eXJx8dH9erVkyQVFhbKzc3NIf/SdW/evLlS+QNmcqV6vHz5st566y35+vqqQ4cODsuWL18uf39/tW3bVs8++6zOnTtnX1aZehw7dqz8/f11zz33aOHChSopKanU9uH60XSjyh04cECSFBoaWqH4IUOGyMvLy+G1fPlyh/kMw1BYWFi5nw8LC9PZs2fLnIpTUcnJyQoLC9N9991nHzt8+LCkX6+1mTx5sj777DM1atRIPXr0sJ8WdPjwYZWUlGjmzJmaO3euPv74Y+Xm5ioiIkKXL1++rlyAquZK9Xjp0iW98MILio6Olo+PT4W3ISsrSwEBAWXGAwIClJWVVeF5AFdQ3WoyJydH06dP1xNPPGEfe/DBB5WVlaVXXnlFly9f1tmzZ+2Xh5w8ebJC2wW4Aleox88++0xeXl7y8PDQnDlzlJaWJn9/f/vyoUOHKiUlRV9++aWmTJmiVatWOdxjqKL1OH36dK1cuVIbNmzQ4MGDNWHCBM2cObNC243/XT2zE0DNU/rXO4vFUqH4OXPmqFevXg5jEydO1JUrV5yyvv9WUFCgDz74QFOmTHEYL/3LX0JCgv785z9L+vU6mBYtWmjlypV64oknVFJSoqKiIr3++uuKjIyUJKWkpCgoKEjp6elc2w2X4Cr1WFRUpMGDB6ukpERvvPFGheb6b+XlbxjGddU9YKbqVJP5+fl6+OGH1aZNGyUmJtrH27Ztq6VLl2r8+PF68cUXVbduXcXHxyswMFB169atUF6AK3CFegwPD5fNZlN2drbefvttPfbYY9q2bZv9j82jRo2yx7Zr10633367OnfuLKvVqrvvvrvC9Th58mT7v0tvbDht2jSHcTgPR7pR5Vq3bi1J+ve//12h+KCgIIWEhDi8Sm9uIUkhISGyWCzau3dvuZ//6aef1KhRI4e/ClbUxx9/rIsXL2rYsGEO402bNpUktWnTxj7m7u6uW2+9VUePHr1qTJMmTeTv72+PAczmCvVYVFSkxx57TD///LPS0tIqdZS7NKdTp06VGT9z5owCAwMrNRdgtupSk+fOnVNUVJS8vLy0Zs0a1a9f32F5dHS0srKydPz4ceXk5CgpKUlnzpzRLbfcUqHtAlyBK9Sjp6enQkJC1LVrVyUnJ6tevXpKTk6+ag5333236tevbz9KL11fPXbt2lX5+fnl/n5F1aPpRpXr2LGj2rRpo1dffbXca0XKexzQtfj5+SkiIkJvvPGGCgoKHJZlZWVp+fLlGjRo0HUd8UpOTlbfvn3VpEkTh/HSR5zs27fPPlZUVKTMzEy1atVKktS9e3dJcojJzc1Vdna2PQYwm9n1WPrl/sCBA9qwYYP8/PwqvQ3dunVTXl6evvvuO/vYtm3blJeX53BZCFAdVIeazM/PV2RkpNzc3PTpp5/Kw8PjqusPDAyUl5eXVqxYIQ8PD0VERFQqf8BMZtdjeQzDUGFh4VWX79mzR0VFRfaDP/+tMvW4c+dOeXh4VPq6dVwfmm5UOYvFosWLF2v//v164IEHtG7dOh0+fFg//vijZsyYoX79+lV6zvnz56uwsFAPPfSQvv76ax07dkypqamKiIhQ8+bNNWPGDHtsbm6ubDab/a+M+/btk81mK3Pt58GDB/X1119r5MiRZdbn4+OjJ598UomJifriiy+0b98+PfXUU5KkgQMHSvr1r6P9+vXTuHHjtHXrVu3evVuxsbEKDQ1VeHh4pbcRcAYz67G4uFgDBgzQjh07tHz5cl25ckVZWVnKyspyuO9BVlaWbDabDh48KEnatWuXbDab/f4JYWFhioqK0qhRo5SRkaGMjAyNGjVKf/zjH3XHHXfY5zl48KC91gsKCmSz2WSz2bjHAlyKq9fkuXPnFBkZqQsXLig5OVn5+fn2mP8+hXb+/PmyWq3av3+/FixYoLFjx2rWrFkOX+CpSbg6M+vxwoULmjRpkjIyMnTkyBFZrVaNHDlS//nPf+zfNQ8dOqRp06Zpx44dyszM1Lp16zRw4EDddddd9oM/peu8Vj2uXbtWb7/9tnbv3q1Dhw7pnXfeUUJCgkaPHi13d/f/fUfi993o26Wj5vrtIw327dtnDBs2zGjWrJnh5uZmtGrVyhgyZIhhtVrtMarg4xcMwzAyMzON4cOHG0FBQUb9+vWNli1bGk8//bSRnZ3tELd48WJDUplXYmKiQ9yLL75otGjRotxHRBiGYVy+fNmYMGGCERAQYHh7exu9evUydu/e7RCTl5dnjBgxwmjYsKHRuHFjo3///sbRo0d/f2cBTuYK9Vj62L7yXunp6fa4xMTEcmMWL15sj8nJyTGGDh1qeHt7G97e3sbQoUPLPFqsR48e5c7z888/X8ceBKpWdanJ0kf3/V4txcTEGI0bNzbc3NyM9u3bG++9916520xNwhW5Qj0WFBQY/fv3t6+zadOmRt++fY3vvvvOHnP06FHjgQcesNfabbfdZsTHxxs5OTkO6/u9evzXv/5ldOzY0fDy8jIaNGhgtGvXzpg7d679MYBwPothXOdzlgAAAAAAwDVxejkAAAAAAE5C0w0AAAAAgJPQdAMAAAAA4CQ03QAAAAAAOAlNNwAAAAAATkLTDQAAAACAk9B0AwAAAADgJDTdAAAAAAA4CU03AADV0JIlS2SxWGSxWPTll1+WWW4YhkJCQmSxWNSzZ88qW6/FYlFSUlKlP5eZmSmLxaIlS5ZUWS4AAFQHNN0AAFRj3t7eSk5OLjP+1Vdf6dChQ/L29jYhKwAAUIqmGwCAamzQoEFatWqV8vPzHcaTk5PVrVs3BQcHm5QZAACQaLoBAKjWhgwZIklKSUmxj+Xl5WnVqlUaMWJEmfjc3FyNGTNGzZs3l5ubm2699VYlJCSosLDQIS4/P1+jRo2Sn5+fvLy8FBUVpf3795ebw4EDBxQdHa2AgAC5u7srLCxMCxYs+N3cz5w5o9GjR6tly5Zyd3dXkyZN1L17d23YsKEyuwAAAJdWz+wEAADA9fPx8dGAAQP07rvv6oknnpD0awNep04dDRo0SHPnzrXHXrp0SeHh4Tp06JCmTp2q9u3b65tvvtGsWbNks9n0+eefS/r1evBHHnlEW7du1UsvvaR77rlHW7ZsUe/evcusf+/evbrvvvsUHBysV199VUFBQVq/fr3i4+OVnZ2txMTEq+YeExMjq9WqGTNmqHXr1vrll19ktVqVk5NTtTsJAAAT0XQDAFDNjRgxQuHh4dqzZ4/atm2rd999VwMHDixzPffSpUv1448/6qOPPtLAgQMlSREREfLy8tLEiROVlpamiIgIrV+/Xunp6Zo3b57i4+PtcW5ubkpISHCYc/z48fL29tbmzZvl4+Njjy0sLNTs2bMVHx+vRo0alZv3li1bNHLkSI0aNco+1q9fvyrbLwAAuAJOLwcAoJrr0aOHbrvtNr377rvatWuXtm/fXu6p5Zs2bZKnp6cGDBjgMD58+HBJ0saNGyVJ6enpkqShQ4c6xEVHRzu8v3TpkjZu3Kj+/furQYMGKi4utr/69OmjS5cuKSMj46p533vvvVqyZIlefvllZWRkqKioqNLbDgCAq6PpBgCgmrNYLIqLi9P777+vhQsXqnXr1rr//vvLxOXk5CgoKEgWi8VhPCAgQPXq1bOf1p2Tk6N69erJz8/PIS4oKKjMfMXFxfrnP/+p+vXrO7z69OkjScrOzr5q3itWrFBsbKzeeecddevWTY0bN9awYcOUlZV1XfsBAABXRNMNAEANMHz4cGVnZ2vhwoWKi4srN8bPz0+nTp2SYRgO46dPn1ZxcbH8/f3tccXFxWWurf5tM9yoUSPVrVtXw4cP1/bt28t9lTbf5fH399fcuXOVmZmpI0eOaNasWVq9erX9yDsAADUBTTcAADVA8+bN9dxzz+lPf/qTYmNjy435wx/+oPPnz+uTTz5xGH/vvffsyyUpPDxckrR8+XKHuA8++MDhfYMGDRQeHq6dO3eqffv26ty5c5nXb4+WX01wcLDGjh2riIgIWa3WCn0GAIDqgBupAQBQQ8yePfuay4cNG6YFCxYoNjZWmZmZuvPOO7V582bNnDlTffr0Ua9evSRJkZGReuCBB/T888/rwoUL6ty5s7Zs2aJly5aVmXPevHn6v//7P91///166qmndPPNN+vcuXM6ePCg1q5dq02bNpWbS15ensLDwxUdHa3Q0FB5e3tr+/btSk1N1aOPPvq/7wwAAFwETTcAALWEh4eH0tPTlZCQoFdeeUVnzpxR8+bN9eyzzzo82qtOnTr69NNPNX78eP3973/X5cuX1b17d61bt06hoaEOc7Zp00ZWq1XTp0/X5MmTdfr0aTVs2FC33377NU8t9/DwUJcuXbRs2TJlZmaqqKhIwcHBmjhxop5//nmn7QMAAG40i/HbC7sAAAAAAECV4JpuAAAAAACchKYbAAAAAAAnoekGAAAAAMBJaLoBAAAAAHASmm4AAAAAAJyEphsAAAAAACeh6QYAAAAAwElougEAAAAAcBKabgAAAAAAnISmGwAAAAAAJ6HpBgAAAADASWi6AQAAAABwkv8HageBYKu/c1QAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1000x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plotting the bar chart\n",
    "ax = counts_df.T.plot(kind='bar', figsize=(10, 6), color=['#1f77b4', '#ff7f0e'], edgecolor='black', linewidth=0.5)\n",
    "\n",
    "# Annotate bars with their values\n",
    "for p in ax.patches:\n",
    "    ax.annotate(f'{p.get_height()}', (p.get_x() + p.get_width() / 2., p.get_height()), \n",
    "                ha='center', va='bottom', xytext=(0, 3), textcoords='offset points', fontsize=12)\n",
    "\n",
    "# Set titles and labels\n",
    "ax.set_title('Matched Genes and Essential Genes Comparison Across Models', fontsize=14, fontweight='bold')\n",
    "ax.set_ylabel('Number of Genes', fontsize=12)\n",
    "ax.set_xlabel('Models', fontsize=12)\n",
    "\n",
    "# Add legend and show plot\n",
    "plt.legend(['Matched Genes', 'Matched Essential Genes'], loc='upper right', fontsize=12)\n",
    "plt.xticks(rotation=0)  # Ensure x-axis labels are horizontal\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Comparing the list of essential genes iCHO2291 vs iCHO3K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Memote"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install memote\n",
    "!pip install --upgrade typing_extensions\n",
    "# !pip install --upgrade pip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "import os\n",
    "import xlrd\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "from memote import test_model, snapshot_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mem_return, mem_result = test_model(model, results=True)\n",
    "with open('iOpol_MemoteReport.html', 'w') as f:\n",
    "    f.write(snapshot_report(mem_result))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_deletion_results = single_gene_deletion(model)\n",
    "essentiality_threshold = 0.01\n",
    "essential_genes = gene_deletion_results[gene_deletion_results['growth'] < essentiality_threshold]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming 'df' contains 'gene' and 'entrez_gene_id'\n",
    "# Create a reverse lookup dictionary where the key is the Entrez Gene ID and the value is the gene symbol\n",
    "entrez_dict = df.set_index('entrez_gene_id')['gene'].to_dict()\n",
    "\n",
    "# Print the essential genes and match them with Entrez Gene IDs\n",
    "print(f\"Number of essential genes for biomass: {len(essential_genes)}\")\n",
    "print(\"Essential genes for biomass production with matching Entrez IDs:\")\n",
    "\n",
    "# Create an empty list to store the results\n",
    "matched_genes = []\n",
    "\n",
    "for gene_idx in essential_genes.index:\n",
    "    gene_exp_ess = model.genes[gene_idx]\n",
    "    \n",
    "    # Check if the gene.id matches any Entrez Gene ID (value) in the reverse dictionary\n",
    "    if gene_exp_ess.id in entrez_dict:\n",
    "        # If it matches, retrieve the corresponding gene symbol\n",
    "        entrez_gene_symbol = entrez_dict[gene_exp_ess.id]\n",
    "        print(f\"{gene_exp_ess.id}: {gene_exp_ess.name}, Entrez Gene Symbol: {entrez_gene_symbol}\")\n",
    "        \n",
    "        # Append the matched result to the list\n",
    "        matched_genes.append((gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol))\n",
    "    else:\n",
    "        print(f\"{gene_exp_ess.id}: {gene_exp_ess.name}, No Entrez Gene Match\")\n",
    "\n",
    "# Convert the matched list into a DataFrame if needed\n",
    "matched_genes_essential_df = pd.DataFrame(matched_genes, columns=['gene_exp_ess.id', 'gene_exp_ess.name', 'entrez_gene_symbol'])\n",
    "\n",
    "# Save the result to a CSV file if desired\n",
    "matched_genes_essential_df.to_csv(\"matched_gene_entrez_ids.csv\", index=False)\n",
    "\n",
    "# Print or return the DataFrame for further use\n",
    "print(matched_genes_df)\n",
    "\n",
    "# Create an empty list to store the results\n",
    "matched_genes_model = []\n",
    "\n",
    "# Iterate through all genes in the model\n",
    "for gene in model.genes:\n",
    "    # Check if the gene.id (or gene.name) matches any Entrez Gene ID in the dictionary\n",
    "    if gene.id in entrez_dict:  # Here we're assuming gene.id matches the entrez_gene_id\n",
    "        entrez_gene_symbol = entrez_dict[gene.id]  # Get the corresponding gene symbol\n",
    "        print(f\"Model gene.id: {gene.id}, Gene name: {gene.name}, Matched Entrez Gene Symbol: {entrez_gene_symbol}\")\n",
    "        \n",
    "        # Append the matched gene information to the list\n",
    "        matched_genes_model.append((gene.id, gene.name, entrez_gene_symbol))\n",
    "    else:\n",
    "        print(f\"Model gene.id: {gene.id}, Gene name: {gene.name}, No match found.\")\n",
    "\n",
    "# Convert the matched genes list to a DataFrame\n",
    "matched_genes_df = pd.DataFrame(matched_genes_model, columns=['gene.id', 'gene.name', 'entrez_gene_symbol'])\n",
    "\n",
    "# Save the result to a CSV file if needed\n",
    "matched_genes_df.to_csv(\"matched_model_genes.csv\", index=False)\n",
    "\n",
    "# Print or return the matched genes DataFrame\n",
    "print(matched_genes_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(entrez_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "essential_genes_model = []\n",
    "for g in model.genes:\n",
    "    if g.name in all_gene_names:\n",
    "        essential_genes_model.append(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set1 = set(sim_essential_genes)\n",
    "set2 = set(essential_genes_model)\n",
    "\n",
    "# Find shared elements\n",
    "shared_elements = set1.intersection(set2)\n",
    "\n",
    "# Find unique elements\n",
    "unique_in_list1 = set1.difference(set2)\n",
    "unique_in_list2 = set2.difference(set1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Shared genes between Exp Validated Essential Genes and Simulated Essential Genes:\", len(shared_elements))\n",
    "print(\"Unique genes in Simulated Essential Genes:\", len(unique_in_list1))\n",
    "print(\"Unique genes in Exp Validated Essentail Genes:\", len(unique_in_list2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Essential for growth genes and comparison to (An optimised genome-wide, virus free CRISPR screen for mammalian cells)\n",
    "https://pubmed.ncbi.nlm.nih.gov/34935002/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Context-specific Model Generation <a id='context_specific'></a>\n",
    "\n",
    "Here we use a matrix generated with rmf_CADRE to generate each context_specific model for each one of the conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from scipy.io import loadmat\n",
    "from time import process_time\n",
    "\n",
    "import pymCADRE\n",
    "from pymCADRE.rank import rank_reactions\n",
    "from pymCADRE.check import check_model_function\n",
    "from pymCADRE.prune import prune_model\n",
    "\n",
    "import cobra\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Intervals for Context-specific model generation\n",
    "Here we define the set of constraints in the uptake and secretion rates that we want to incorporate to our context specific models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Model Loading and Constraining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load generic model\n",
    "model = load_json_model('iCHO3595_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_dict = uptsec_intrvl_zela\n",
    "time = 'P2 to P4'\n",
    "\n",
    "with model as model:\n",
    "    model.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "    for rxn in model.boundary:\n",
    "        \n",
    "        # Models that are forced to secrete ethanol are not feasible\n",
    "        if rxn.id == 'EX_etoh_e':\n",
    "            rxn.bounds = (-0.1,0.1)\n",
    "            continue\n",
    "\n",
    "        # Replace the lower and upper bound with experimental data\n",
    "        if rxn.id in temp_dict.keys():\n",
    "            rxn.bounds = temp_dict[rxn.id][time]\n",
    "            continue\n",
    "        \n",
    "        # Keep boundaries open for essential metabolites\n",
    "        if rxn.id == 'EX_h2o_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_h_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_o2_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_hco3_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_so4_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_pi_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "\n",
    "        # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "        if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Tyr_ggn_c':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Ser_Thr_g':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_pre_prot_r':\n",
    "            rxn.bounds = (-0.0006,1000)\n",
    "            continue\n",
    "\n",
    "        # Close uptake rates for the rest of the boundaries\n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "        \n",
    "    pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "    print(pfba_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading Ubiquity and Confidence Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the .mat file\n",
    "mat_data = loadmat('../Data/Context_specific_models/ubiData.mat')\n",
    "\n",
    "# Access the ubiData structure\n",
    "ubiData = mat_data['ubiData']\n",
    "\n",
    "# Accessing individual fields within ubiData\n",
    "ubiScores = ubiData['ubiScores'][0, 0]\n",
    "uScore = ubiData['uScore'][0, 0]\n",
    "rxns = ubiData['rxns'][0, 0]\n",
    "Condition = ubiData['Condition'][0, 0]\n",
    "\n",
    "# Print or process the data as needed\n",
    "print(ubiScores.shape)\n",
    "print(uScore.shape)\n",
    "print(rxns.shape)\n",
    "print(Condition.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,cond in enumerate(Condition[0]):\n",
    "    if \"WT_P4\" in cond[0]:\n",
    "        print(cond[0], i)\n",
    "    elif \"ZeLa_P4\" in cond[0]:\n",
    "        print(cond[0], i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Condition[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ubiquity score list\n",
    "U = list(ubiScores[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conf Score List\n",
    "conf_scores = pd.read_csv('../Data/Context_specific_models/confidence_scores.csv', header=None).iloc[:, 0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Protected Reations\n",
    "high_confidence_reactions = ['biomass_cho_s', 'DNAsyn', 'LipidSyn_cho_s', 'PROTsyn_cho_s', 'RNAsyn_cho_s', 'EX_bhb_e', 'EX_nh4_e', 'EX_ac_e', 'EX_ala_L_e',\n",
    "    'EX_arg_L_e', 'EX_asn_L_e', 'EX_asp_L_e', 'EX_2hb_e', 'EX_cit_e', 'EX_cys_L_e', 'EX_etoh_e', 'EX_for_e', 'EX_fum_e', 'EX_glc_e', 'EX_glu_L_e',\n",
    "    'EX_gln_L_e', 'EX_gly_e', 'EX_his_L_e', 'EX_4hpro_e', 'EX_ile_L_e', 'EX_lac_L_e', 'EX_leu_L_e', 'EX_lys_L_e', 'EX_mal_L_e', 'EX_met_L_e',\n",
    "    'EX_phe_L_e', 'EX_pro_L_e', 'EX_5oxpro_e', 'EX_pyr_e', 'EX_ser_L_e', 'EX_thr_L_e', 'EX_trp_L_e', 'EX_tyr_L_e', 'EX_val_L_e', 'EX_h2o_e',\n",
    "    'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e', 'SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reduced model generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################\n",
    "# Rank reactions\n",
    "##############################################\n",
    "\n",
    "print('Processing inputs and ranking reactions...')\n",
    "GM, C, NC, P, Z, model_C = rank_reactions(model, U, conf_scores, high_confidence_reactions, method=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output the results\n",
    "print(\"Generic Model:\", GM)\n",
    "print(\"Core Reactions:\", len(C))\n",
    "print(\"Non-Core Reactions:\", len(NC))\n",
    "print(\"Ranked Non-Core Reactions:\", len(P))\n",
    "print(\"Zero-Expression Reactions:\", len(Z))\n",
    "print(\"Core Reactions in Original Model:\", len(model_C))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.util import create_stoichiometric_matrix\n",
    "\n",
    "def retrieve_precursor_metabolites(model, protected_rxns):\n",
    "    \"\"\"\n",
    "    Retrieve precursor metabolites from a specified list of reactions in a metabolic model.\n",
    "\n",
    "    Args:\n",
    "        model: cobra.Model - The metabolic model.\n",
    "        protected_rxns: list - A list of reaction IDs for which precursor metabolites need to be retrieved.\n",
    "\n",
    "    Returns:\n",
    "        precursor_mets: list - A list of precursor metabolites.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a list of all the reactions in the model\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Ensure that the protected reactions list is not empty\n",
    "    if protected_rxns:\n",
    "        \n",
    "        #Define a list to store the indices of the protected reactions\n",
    "        rxn_indices = []\n",
    "        \n",
    "        for rxn in protected_rxns:\n",
    "            # Ensure that the protected reaction is included in the model\n",
    "            if rxn in model_reactions:\n",
    "                # Get the indices of the protected reactions in the model, only if the reaction exists\n",
    "                rxn_indices.append(model.reactions.index(model.reactions.get_by_id(rxn)))\n",
    "\n",
    "        # Create stoichiometric matrix S\n",
    "        S = create_stoichiometric_matrix(model)\n",
    "        \n",
    "        # Retrieve precursor metabolites: metabolites with a negative coefficient in the stoichiometric matrix\n",
    "        precursor_mets = [\n",
    "            model.metabolites[i] \n",
    "            for i in range(len(model.metabolites)) \n",
    "            if any(S[i, idx] < 0 for idx in rxn_indices)\n",
    "        ]\n",
    "\n",
    "        return precursor_mets\n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precursor_mets = retrieve_precursor_metabolites(model, high_confidence_reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# Define inputs to the model pruning step\n",
    "##################################################\n",
    "\n",
    "# Define core vs. non-core ratio threshold for removing reactions\n",
    "eta = 1 / 3\n",
    "\n",
    "# Check functionality of generic model\n",
    "genericStatus = check_model_function(GM, 'required_mets', precursor_mets)[0]\n",
    "print(genericStatus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constraint the model with exp data\n",
    "\n",
    "time = 'P2 to P4'\n",
    "\n",
    "GM.solver = 'gurobi'\n",
    "objective = 'biomass_cho_s' # 'biomass_cho'\n",
    "GM.objective = objective\n",
    "\n",
    "GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "for rxn in GM.boundary:\n",
    "    \n",
    "    # Models that are forced to secrete ethanol are not feasible\n",
    "    if rxn.id == 'EX_etoh_e':\n",
    "        rxn.bounds = (-0.1,0.1)\n",
    "        continue\n",
    "\n",
    "    # Replace the lower and upper bound with experimental data\n",
    "    if rxn.id in temp_dict.keys():\n",
    "        rxn.bounds = temp_dict[rxn.id][time]\n",
    "        continue\n",
    "    \n",
    "    # Keep boundaries open for essential metabolites\n",
    "    if rxn.id == 'EX_h2o_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_h_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_o2_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_hco3_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_so4_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_pi_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "\n",
    "    # Boundaries from Sink reactions on iCHO_v1 (100 times lower)\n",
    "    if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Tyr_ggn_c':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Ser_Thr_g':\n",
    "        rxn.bounds = (-0.001,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_pre_prot_r':\n",
    "        rxn.bounds = (-0.0006,1000)\n",
    "        continue\n",
    "\n",
    "    # Close uptake rates for the rest of the boundaries\n",
    "    if rxn.id.startswith(\"EX_\"):\n",
    "        rxn.bounds = (0,1000) \n",
    "    if rxn.id.startswith(\"SK_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    if rxn.id.startswith(\"DM_\"):\n",
    "        rxn.bounds = (0,1000)\n",
    "    \n",
    "pfba_solution = cobra.flux_analysis.pfba(GM)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(pfba_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if genericStatus:\n",
    "    print('Generic model passed precursor metabolites test')\n",
    "\n",
    "    ##############################################################\n",
    "    # If generic functionality test is passed, prune reactions\n",
    "    ###############################################################\n",
    "    print('Pruning reactions...')\n",
    "    t0 = process_time()\n",
    "    PM, cRes = prune_model(GM, P, C, Z, eta, precursor_mets, salvage_check=0, method=1)\n",
    "    # Stop the stopwatch / counter\n",
    "    t_stop = process_time()\n",
    "    # compute elapsed time\n",
    "    pruneTime = t_stop - t0\n",
    "\n",
    "else:\n",
    "    print('Generic model failed precursor metabolites test!!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ATP Generation loops removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import chain\n",
    "import pickle\n",
    "import numpy as np\n",
    "from optlang.symbolics import Zero\n",
    "\n",
    "import cobra\n",
    "from cobra.util import create_stoichiometric_matrix\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading and Constraining Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = load_json_model('iCHO3595_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)\n",
    "\n",
    "for rxn in model.boundary:\n",
    "    if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "        rxn.bounds = (-1000, 1000)\n",
    "    elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "        rxn.bounds = (-0.001, 1000)\n",
    "    elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "        rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "for reaction in model.reactions:\n",
    "    if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "        continue\n",
    "    elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "        reaction.lower_bound = 0\n",
    "    for r,v in uptsec_zela.items():\n",
    "        if reaction.id == r:\n",
    "            reaction.upper_bound = 1000\n",
    "            reaction.lower_bound = v[('U6','P2 to P4')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ATP loops removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################\n",
    "# Thanasis' original version\n",
    "#########################################\n",
    "\n",
    "\n",
    "# Get the maximum flux through the oxidatitve phosphoryaltion pathway. Get the maximum flux of the objectives, biomass and IgG. \n",
    "# If a reaction that consumes ATP (negative stoichiometric coefficient), has a flux value below a cutoff in both maximisation of biomass\n",
    "# IgG and ox. pho. pathway, then the reaction doesn't provide any useable ATP (cyclic)\n",
    "\n",
    "step6_safe_rxn_rmv = []\n",
    "\n",
    "protected_subsystems = [ # Why protecting these subsystems\n",
    "    'BIOMASS', 'Product Formation', 'OXIDATIVE PHOSPHORYLATION', 'CITRIC ACID CYCLE', 'GLYCOLYSIS/GLUCONEOGENESIS',\n",
    "    'PYRUVATE METABOLISM', 'PENTOSE PHOSPHATE PATHWAY', 'GLUTAMATE METABOLISM', 'NAD METABOLISM',\n",
    "    'TRANSPORT, MITOCHONDRIAL', 'TRANSPORT, ENDOPLASMIC RETICULAR', 'ARGININE AND PROLINE METABOLISM',\n",
    "    'UBIQUINONE SYNTHESIS', 'VALINE, LEUCINE, AND ISOLEUCINE METABOLISM', 'TRYPTOPHAN METABOLISM',\n",
    "    'METHIONINE AND CYSTEINE METABOLISM', 'TYROSINE METABOLISM', 'GLYCINE, SERINE, AND THREONINE METABOLISM',\n",
    "    'HISTIDINE METABOLISM', 'ALANINE AND ASPARTATE METABOLISM', 'PHENYLALANINE METABOLISM',\n",
    "    'LYSINE METABOLISM', 'TYROSINE METABOLISM', 'N-GLYCAN BIOSYNTHESIS', 'PROTEIN PRODUCTION', 'SECRETORY PATHWAY'\n",
    "]\n",
    "\n",
    "\n",
    "# biomass and igg reactions of the model; we want to ensure their production after removals of any ATP cycles\n",
    "biomass_rxn = 'biomass_cho_s'\n",
    "igg_rxn = 'igg_formation'\n",
    "\n",
    "model.objective = biomass_rxn\n",
    "biomass_control = model.optimize('maximise')\n",
    "model.objective = igg_rxn\n",
    "igg_control = model.optimize('maximise')\n",
    "print(\"Original: Biomass: \", round(biomass_control.objective_value, 5), \"IgG: \", round(igg_control.objective_value, 5))\n",
    "\n",
    "model_temp = model.copy()\n",
    "atp_related_mets = ['atp_c', 'atp_n', 'atp_m', 'atp_x'] # What about atp_l, atp_g, atp_r, atp_e\n",
    "atp_met_indices = [model_temp.metabolites.get_by_id(met) for met in atp_related_mets]\n",
    "\n",
    "# Find the reactions associated with these metabolites\n",
    "atp_rxns_indices = []\n",
    "for met in atp_met_indices:\n",
    "    for reac in met.reactions:\n",
    "        atp_rxns_indices.append(reac.id)\n",
    "atp_rxns_indices = list(set(atp_rxns_indices))\n",
    "\n",
    "# Determine which reactions are consuming ATP (negative stoichiometry for ATP)\n",
    "atp_consumption_idx = []\n",
    "stoich_matrix = create_stoichiometric_matrix(model_temp)\n",
    "for atp_rxn in atp_rxns_indices:\n",
    "    reaction = model_temp.reactions.get_by_id(atp_rxn)\n",
    "    if not reaction.reversibility:  # Only irreversible reactions\n",
    "        for met in atp_met_indices:\n",
    "            if stoich_matrix[model_temp.metabolites.index(met), model_temp.reactions.index(reaction)] < 0:\n",
    "                atp_consumption_idx.append(atp_rxn)\n",
    "                break\n",
    "\n",
    "# Perform FBA to get maximum ATP production\n",
    "model_temp.objective = 0\n",
    "oxpho_rxns = [r for r in model_temp.reactions if 'OXIDATIVE PHOSPHORYLATION' in r.subsystem]\n",
    "reaction_variables = ((rxn.forward_variable, rxn.reverse_variable) for rxn in oxpho_rxns)\n",
    "variables = chain(*reaction_variables)\n",
    "model_temp.objective = model_temp.problem.Objective( Zero, direction=\"max\", sloppy=True, name=\"maxATP\" )\n",
    "model_temp.objective.set_linear_coefficients({v: 1.0 for v in variables})\n",
    "print(\"objective function\", model_temp.objective.expression)\n",
    "solution_maxATP = model_temp.optimize()\n",
    "\n",
    "for obj in [biomass_rxn, igg_rxn]:\n",
    "    model_temp.objective = obj\n",
    "    sol = model_temp.optimize('maximize')\n",
    "    atp_rate = np.vstack([sol[atp_consumption_idx], solution_maxATP[atp_consumption_idx]])\n",
    "\n",
    "    # Find ATP consuming reactions with low rates\n",
    "    atp_rxn_lowFlux = [id for id, rate in zip(atp_consumption_idx, np.sum(atp_rate, axis=0)) if rate <= 2e-12]\n",
    "\n",
    "    atp_loop_rxn_rmv = []\n",
    "    for rxn in atp_rxn_lowFlux:\n",
    "        if model_temp.reactions.get_by_id(rxn).subsystem not in protected_subsystems:\n",
    "            atp_loop_rxn_rmv.append(rxn)\n",
    "\n",
    "    # Filter out exchange reactions\n",
    "    exchange_reactions = [r.id for r in model_temp.exchanges]\n",
    "    atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id not in exchange_reactions]\n",
    "\n",
    "    # Ensure ATP maintenance reaction is not deleted\n",
    "    id_atp_maint = model_temp.reactions.get_by_id('ATPM').id\n",
    "    atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id != id_atp_maint]\n",
    "\n",
    "    if obj == biomass_rxn:\n",
    "        atp_loop_rxns_bio = atp_loop_rxn_rmv\n",
    "    elif obj == igg_rxn:\n",
    "        atp_loop_rxns_igg = atp_loop_rxn_rmv\n",
    "\n",
    "# Print the reactions to be deleted\n",
    "atp_loop_rxns = list(set(atp_loop_rxns_igg).intersection(atp_loop_rxns_bio))\n",
    "print(\"Reactions to be deleted:\", len(atp_loop_rxns))\n",
    "\n",
    "# Remove reactions from the model and check that Biomass and IgG are produced\n",
    "model_temp.objective = biomass_rxn\n",
    "sol_bio_control = model_temp.optimize('maximize')\n",
    "model_temp.objective = igg_rxn\n",
    "sol_igg_control = model_temp.optimize('maximize')\n",
    "model_temp_temp = model_temp.copy()\n",
    "\n",
    "for rxn_to_rmv in atp_loop_rxns:\n",
    "    model_temp_temp.remove_reactions(rxn_to_rmv)\n",
    "    model_temp_temp.objective = biomass_rxn\n",
    "    sol_bio = model_temp_temp.optimize('maximize')\n",
    "    model_temp_temp.objective = igg_rxn\n",
    "    sol_igg = model_temp_temp.optimize('maximize')\n",
    "    if round(sol_bio.objective_value, 5) == round(sol_bio_control.objective_value, 5) and round(sol_igg.objective_value, 5) == round(sol_igg_control.objective_value, 5):\n",
    "        step6_safe_rxn_rmv.append(rxn_to_rmv)\n",
    "\n",
    "model_temp.remove_reactions(step6_safe_rxn_rmv)\n",
    "print(\"Reactions consuming ATP in a cyclic fashion AND are safe to remove: \", len(step6_safe_rxn_rmv))\n",
    "print(\"Model size after the removal of ATP cyclic reactions: Reactions: \", len(model_temp.reactions), \"Metabolites: \", len(model_temp.metabolites))\n",
    "model_temp.objective = biomass_rxn\n",
    "sol_bio = model_temp.optimize('maximize')\n",
    "model_temp.objective = igg_rxn\n",
    "sol_igg = model_temp.optimize('maximize')\n",
    "print(\"Old Bio: \", round(sol_bio_control.objective_value, 3), \"New Bio: \", round(sol_bio.objective_value, 3), \"Old IgG: \", round(sol_igg_control.objective_value, 5), \"New IgG: \", round(sol_igg.objective_value, 5))\n",
    "model = model_temp.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#########################################\n",
    "# Pablo's modification\n",
    "#########################################\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "def get_atp_related_reactions(model, atp_mets):\n",
    "    \"\"\" Get reactions associated with ATP and related metabolites like GTP, GDP, etc. \"\"\"\n",
    "    atp_rxns_indices = set()  # Use set to avoid duplicates\n",
    "    for met in atp_mets:\n",
    "        for reac in met.reactions:\n",
    "            if any(x in reac.reactants + reac.products for x in atp_mets):\n",
    "                atp_rxns_indices.add(reac.id)\n",
    "    return list(atp_rxns_indices)\n",
    "\n",
    "def find_low_flux_reactions(model, atp_rxns_indices, atp_met_indices, stoich_matrix, consumption_cutoff=2e-12):\n",
    "    \"\"\" Identify reactions with low ATP or related metabolite consumption flux. \"\"\"\n",
    "    atp_consumption_idx = []\n",
    "    for atp_rxn in atp_rxns_indices:\n",
    "        reaction = model.reactions.get_by_id(atp_rxn)\n",
    "        for met in atp_met_indices:\n",
    "            # Check for both ATP and GTP-related consumption (stoichiometric coefficient < 0)\n",
    "            if stoich_matrix[model.metabolites.index(met), model.reactions.index(reaction)] < 0:\n",
    "                atp_consumption_idx.append(atp_rxn)\n",
    "                break\n",
    "    return atp_consumption_idx\n",
    "\n",
    "def perform_fba_for_max_atp(model, oxpho_rxns):\n",
    "    \"\"\" Set FBA objective to maximize ATP production. \"\"\"\n",
    "    reaction_variables = ((rxn.forward_variable, rxn.reverse_variable) for rxn in oxpho_rxns)\n",
    "    variables = chain(*reaction_variables)\n",
    "    model.objective = model.problem.Objective(Zero, direction=\"max\", sloppy=True, name=\"maxATP\")\n",
    "    model.objective.set_linear_coefficients({v: 1.0 for v in variables})\n",
    "    return model.optimize()\n",
    "\n",
    "def remove_cyclic_atp_reactions(model, atp_rxns_to_remove, biomass_rxn):\n",
    "    \"\"\" Remove reactions consuming ATP in a cyclic fashion and check for safety. \"\"\"\n",
    "    \n",
    "    safe_rxn_irrev = []\n",
    "\n",
    "    # Copy the original model\n",
    "    temp_model = model.copy()\n",
    "\n",
    "    # Set the objective for the temporary model\n",
    "    temp_model.objective = biomass_rxn\n",
    "    sol_bio_control = temp_model.optimize()\n",
    "\n",
    "    # Cumulative list of safe reactions to remove or modify\n",
    "    candidate_safe_rmv = []\n",
    "    candidate_safe_irrev = []\n",
    "\n",
    "    for rxn_id in tqdm(atp_rxns_to_remove, desc=\"Processing reactions\"):\n",
    "        with temp_model:\n",
    "            # Try to make the reaction irreversible\n",
    "            if temp_model.reactions.get_by_id(rxn_id).lower_bound != 0:\n",
    "                temp_model.reactions.get_by_id(rxn_id).bounds = (0, 1000)\n",
    "                sol_bio_temp = temp_model.optimize()\n",
    "                \n",
    "                if (round(sol_bio_temp.objective_value, 5) == round(sol_bio_control.objective_value, 5)):\n",
    "                    candidate_safe_irrev.append(rxn_id)\n",
    "                    \n",
    "                    # Test cumulative effect of all candidate modifications\n",
    "                    for rxn in candidate_safe_irrev:\n",
    "                        temp_model.reactions.get_by_id(rxn).bounds = (0, 1000)\n",
    "                    sol_bio_cumulative = temp_model.optimize()\n",
    "\n",
    "                    # If the cumulative effect is still safe, add to the final modification list\n",
    "                    if round(sol_bio_cumulative.objective_value, 5) == round(sol_bio_control.objective_value, 5):\n",
    "                        safe_rxn_irrev = list(candidate_safe_irrev)\n",
    "                    else:\n",
    "                        # If not safe, undo the last added irreversible modification\n",
    "                        candidate_safe_irrev.remove(rxn_id)\n",
    "                        print(f\"Making {rxn_id} irreversible is not safe, reverting.\")\n",
    "\n",
    "    # Make irreversible reactions that can be safely modified\n",
    "    for rxn in safe_rxn_irrev:\n",
    "        if model.reactions.get_by_id(rxn).lower_bound != 0:\n",
    "            model.reactions.get_by_id(rxn).bounds = (0, 1000)\n",
    "    \n",
    "    return model, safe_rxn_irrev\n",
    "\n",
    "# Set objectives\n",
    "biomass_rxn = 'biomass_cho_s'\n",
    "model.objective = biomass_rxn\n",
    "biomass_control = model.optimize('maximize')\n",
    "\n",
    "# Find ATP related reactions\n",
    "atp_related_mets = ['atp_c', 'atp_n', 'atp_m', 'atp_x', 'atp_l', 'atp_g', 'atp_r', 'atp_e', 'datp_c', 'dadp_c', 'gtp_c', 'gdp_c',\n",
    "                    'utp_c', 'udp_c', 'ctp_c', 'cdp_c', 'ump_c', 'mlthf_c', 'mlthf_m', 'dgtp_c', 'dgdp_c', 'adxrd_c', 'adxrd_m',\n",
    "                    'adxo_2_2_m', 'adxo_2_2_c', 'udpg_c', 'udpgal_c', 'succ_m', 'succoa_m', '4met2obut_c', 'q10h2_m', 'q10_m']\n",
    "atp_met_indices = [model.metabolites.get_by_id(met) for met in atp_related_mets]\n",
    "atp_rxns_indices = get_atp_related_reactions(model, atp_met_indices)\n",
    "\n",
    "# Create stoichiometric matrix\n",
    "stoich_matrix = create_stoichiometric_matrix(model)\n",
    "atp_consumption_idx = find_low_flux_reactions(model, atp_rxns_indices, atp_met_indices, stoich_matrix)\n",
    "\n",
    "# Maximize ATP production\n",
    "oxpho_rxns = [r for r in model.reactions if 'OXIDATIVE PHOSPHORYLATION' in r.subsystem]\n",
    "solution_maxATP = perform_fba_for_max_atp(model, oxpho_rxns)\n",
    "\n",
    "# Check flux of ATP consuming reactions\n",
    "atp_loop_rxn_rmv = []\n",
    "\n",
    "model.objective = biomass_rxn\n",
    "sol = model.optimize('maximize')\n",
    "atp_rate = np.vstack([sol[atp_consumption_idx], solution_maxATP[atp_consumption_idx]])\n",
    "\n",
    "# Find low flux ATP reactions\n",
    "atp_rxn_lowFlux = [rxn_id for rxn_id, rate in zip(atp_consumption_idx, np.sum(atp_rate, axis=0)) if rate <= 2e-12]\n",
    "\n",
    "# Filter protected subsystems and exchanges\n",
    "protected_subsystems = ['BIOMASS', 'OXIDATIVE PHOSPHORYLATION']\n",
    "atp_loop_rxn_rmv += [rxn for rxn in atp_rxn_lowFlux if model.reactions.get_by_id(rxn).subsystem not in protected_subsystems]\n",
    "\n",
    "# Ensure ATPM and exchange reactions are not removed\n",
    "atp_loop_rxn_rmv = [r_id for r_id in atp_loop_rxn_rmv if r_id != 'ATPM' and not r_id.startswith(\"EX_\")]\n",
    "\n",
    "# Remove duplicates in the list\n",
    "atp_loop_rxn_rmv = set(atp_loop_rxn_rmv)\n",
    "\n",
    "# Remove cyclic ATP reactions and check safety\n",
    "pruned_model, safe_rxn_irrev = remove_cyclic_atp_reactions(model, atp_loop_rxn_rmv, biomass_rxn)\n",
    "\n",
    "print(f\"Reactions consuming ATP in a cyclic fashion and safe to make irreversible: {len(safe_rxn_irrev)}\")\n",
    "\n",
    "# Final biomass\n",
    "pruned_model.objective = biomass_rxn\n",
    "sol_bio = pruned_model.optimize('maximize')\n",
    "\n",
    "print(f\"Old Bio: {biomass_control.objective_value:.3f}, New Bio: {sol_bio.objective_value:.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pruned_model as model:\n",
    "    objective = 'biomass_cho_s'\n",
    "    model.objective = objective\n",
    "    '''\n",
    "    model.reactions.SCP22x.knock_out()\n",
    "    model.reactions.TMNDNCCOAtx.knock_out()\n",
    "    model.reactions.OCCOAtx.knock_out()\n",
    "    model.reactions.TMNDNCCOAtx.knock_out()\n",
    "    model.reactions.r0391.knock_out()\n",
    "    model.reactions.BiGGRxn67.knock_out()\n",
    "    model.reactions.r2247.knock_out()\n",
    "    model.reactions.r2280.knock_out()\n",
    "    model.reactions.r2246.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.r2279.knock_out()\n",
    "    model.reactions.r2245.knock_out()\n",
    "    model.reactions.r2305.knock_out()\n",
    "    model.reactions.r2317.knock_out()\n",
    "    model.reactions.r2335.knock_out()\n",
    "    model.reactions.HMR_0293.knock_out()\n",
    "    model.reactions.HMR_7741.knock_out()\n",
    "    model.reactions.r0509.knock_out()\n",
    "    model.reactions.r1453.knock_out()\n",
    "    model.reactions.HMR_4343.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.ACONTm.knock_out()\n",
    "    model.reactions.PDHm.knock_out()\n",
    "    model.reactions.r0426.knock_out()\n",
    "    model.reactions.r0383.knock_out()\n",
    "    model.reactions.r0555.knock_out()\n",
    "    model.reactions.r1393.knock_out()\n",
    "    model.reactions.NICRNS.knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R00648').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R03326').knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R08726').knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.RE2915M.knock_out()\n",
    "    model.reactions.HMR_3288.knock_out()\n",
    "    model.reactions.HMR_1325.knock_out()\n",
    "    model.reactions.HMR_7599.knock_out()\n",
    "    model.reactions.r1431.knock_out()\n",
    "    model.reactions.r1433.knock_out()\n",
    "    model.reactions.RE2439C.knock_out()\n",
    "    model.reactions.r0082.knock_out()\n",
    "    model.reactions.r0791.knock_out()\n",
    "    model.reactions.r1450.knock_out()\n",
    "\n",
    "    \n",
    "    model.reactions.get_by_id('GAUGE-R00270').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R02285').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R04283').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R06127').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R06128').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R06238').knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R00524').knock_out()\n",
    "    model.reactions.RE3477C.knock_out()\n",
    "\n",
    "    model.reactions.AAPSAS.knock_out()\n",
    "    model.reactions.RE3347C.knock_out()\n",
    "    model.reactions.HMR_0960.knock_out()\n",
    "    model.reactions.HMR_0980.knock_out()\n",
    "    model.reactions.RE3476C.knock_out()\n",
    "    model.reactions.r0708.knock_out()\n",
    "    model.reactions.r0777.knock_out()\n",
    "    model.reactions.r0084.knock_out()\n",
    "    model.reactions.r0424.knock_out()\n",
    "    model.reactions.r0698.knock_out()\n",
    "\n",
    "    model.reactions.get_by_id('3HDH260p').knock_out()\n",
    "    model.reactions.HMR_3272.knock_out()\n",
    "    model.reactions.ACOAD183n3m.knock_out()\n",
    "    model.reactions.HMR_1996.knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R01463').knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R04807').knock_out()\n",
    "    model.reactions.r1468.knock_out()\n",
    "    model.reactions.r2435.knock_out()\n",
    "    model.reactions.r0655.knock_out()\n",
    "    model.reactions.r0603.knock_out()\n",
    "\n",
    "    model.reactions.r0541.knock_out()\n",
    "    model.reactions.RE0383C.knock_out()\n",
    "    model.reactions.HMR_1329.knock_out()\n",
    "    model.reactions.TYRA.knock_out()\n",
    "    model.reactions.NRPPHRt_2H.knock_out()\n",
    "    model.reactions.get_by_id('GAUGE-R07364').knock_out()\n",
    "    model.reactions.get_by_id('GapFill-R03599').knock_out()\n",
    "    model.reactions.ARD.knock_out()\n",
    "    model.reactions.RE3095C.knock_out()\n",
    "    model.reactions.RE3104C.knock_out()\n",
    "    model.reactions.RE3104R.knock_out()\n",
    "\n",
    "    model.reactions.ACONT.knock_out()\n",
    "    model.reactions.ACONTm.knock_out()\n",
    "    model.reactions.ICDHxm.knock_out()\n",
    "    model.reactions.ICDHy.knock_out()\n",
    "    model.reactions.AKGDm.knock_out()\n",
    "    model.reactions.r0083.knock_out()\n",
    "    model.reactions.r0425.knock_out()\n",
    "    model.reactions.r0556.knock_out()\n",
    "    model.reactions.NH4t4r.knock_out()\n",
    "    model.reactions.PROPAT4te.knock_out()\n",
    "    \n",
    "    model.reactions.r0085.knock_out()\n",
    "    model.reactions.r0156.knock_out()\n",
    "    model.reactions.r0464.knock_out()\n",
    "    model.reactions.ABUTDm.knock_out()\n",
    "    model.reactions.OIVD1m.knock_out()\n",
    "    model.reactions.OIVD2m.knock_out()\n",
    "    model.reactions.OIVD3m.knock_out()\n",
    "    model.reactions.r2194.knock_out()\n",
    "    model.reactions.r2202.knock_out()\n",
    "    model.reactions.HMR_9617.knock_out()\n",
    "\n",
    "    model.reactions.r2197.knock_out()\n",
    "    model.reactions.r2195.knock_out()\n",
    "    model.reactions.get_by_id('2OXOADOXm').knock_out()\n",
    "    model.reactions.r2328.knock_out()\n",
    "    model.reactions.r0386.knock_out()\n",
    "    model.reactions.r0451.knock_out()\n",
    "    model.reactions.FAS100COA.knock_out()\n",
    "    model.reactions.FAS120COA.knock_out()\n",
    "    model.reactions.FAS140COA.knock_out()\n",
    "    model.reactions.FAS80COA_L.knock_out()\n",
    "\n",
    "    model.reactions.r0604.knock_out()\n",
    "    model.reactions.r0670.knock_out()\n",
    "    model.reactions.r2334.knock_out()\n",
    "    model.reactions.r0193.knock_out()\n",
    "    model.reactions.r0595.knock_out()\n",
    "    model.reactions.r0795.knock_out()\n",
    "    model.reactions.GLYCLm.knock_out()\n",
    "    model.reactions.MACACI.knock_out()\n",
    "    model.reactions.r2193.knock_out()\n",
    "    model.reactions.r0779.knock_out()\n",
    "    model.reactions.r0669.knock_out()\n",
    "    model.reactions.UDCHOLt.knock_out()\n",
    "    model.reactions.r2146.knock_out()\n",
    "    model.reactions.r2139.knock_out()\n",
    "    model.reactions.r0391.knock_out()\n",
    "    '''\n",
    "    for rxn in model.boundary:\n",
    "        if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "            rxn.bounds = (-1000, 1000)\n",
    "        elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "            rxn.bounds = (-0.001, 1000)\n",
    "        elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "            rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "            continue\n",
    "        elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "            reaction.lower_bound = 1000\n",
    "        for r,v in uptsec_zela.items():\n",
    "            if reaction.id == r:\n",
    "                reaction.upper_bound = 1000\n",
    "                reaction.lower_bound = v[('U7','P2 to P4')]\n",
    "    pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "    pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "    print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "    print(f'Simulated growth rate is: {pfba_obj}')\n",
    "    \n",
    "    for i,f in pfba_solution.fluxes.items():\n",
    "        if abs(f) > 1:\n",
    "            print(i,f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Biomass prediction using exp. data <a id='biomass'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "import cobra\n",
    "from cobra import Reaction\n",
    "from cobra.io import load_json_model, read_sbml_model, load_matlab_model\n",
    "from cobra.exceptions import Infeasible\n",
    "from cobra.sampling import ACHRSampler\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory containing context-specific models models\n",
    "model_directory = '../Data/Context_specific_models/'\n",
    "\n",
    "# Model Names\n",
    "wt_model_names = ['WT_P4_Bio141', 'WT_P4_Bio142', 'WT_P4_Bio143',\n",
    "                  'WT_P6_Bio141', 'WT_P6_Bio142', 'WT_P6_Bio143']\n",
    "\n",
    "zela_model_names = ['ZeLa_P2_Bio145', 'ZeLa_P2_Bio146', 'ZeLa_P2_Bio147', 'ZeLa_P2_Bio148',\n",
    "                    'ZeLa_P4_Bio144', 'ZeLa_P4_Bio145', 'ZeLa_P4_Bio146', 'ZeLa_P4_Bio147', 'ZeLa_P4_Bio148',\n",
    "                    'ZeLa_P6_Bio144', 'ZeLa_P6_Bio146', 'ZeLa_P6_Bio147', 'ZeLa_P6_Bio148',\n",
    "                    'ZeLa_P8_Bio144', 'ZeLa_P8_Bio145', 'ZeLa_P8_Bio146', 'ZeLa_P8_Bio147', 'ZeLa_P8_Bio148']\n",
    "\n",
    "# Dictionaries to store the loaded models\n",
    "wt_models = {}\n",
    "zela_models = {}\n",
    "\n",
    "# List all .mat files in the directory\n",
    "model_files = [f for f in os.listdir(model_directory) if f.endswith('.mat')]\n",
    "\n",
    "# Iterate through the .mat files and load only the models that match predefined strings\n",
    "for model_file in model_files:\n",
    "    # Full path to the model file\n",
    "    model_path = os.path.join(model_directory, model_file)\n",
    "    \n",
    "    # Check if the file corresponds to a WT model\n",
    "    for model_name in wt_model_names:\n",
    "        if model_name in model_file:\n",
    "            wt_models[model_name] = load_matlab_model(model_path)\n",
    "            wt_models[model_name].id = model_name\n",
    "            print(f\"Loaded WT model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed\n",
    "    \n",
    "    # Check if the file corresponds to a ZeLa model\n",
    "    for model_name in zela_model_names:\n",
    "        if model_name in model_file:\n",
    "            zela_models[model_name] = load_matlab_model(model_path)\n",
    "            zela_models[model_name].id = model_name\n",
    "            print(f\"Loaded ZeLa model: {model_name}\")\n",
    "            break  # Stop checking other names since this file is already processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATP loop generating reactions\n",
    "\n",
    "atp_reactions_list = [\n",
    "    'SCP22x', 'TMNDNCCOAtx', 'OCCOAtx', 'r0391', 'BiGGRxn67', 'r2247', 'r2280', 'r2246', 'r2279', 'r2245', \n",
    "    'r2305', 'r2317', 'r2335', 'HMR_0293', 'HMR_7741', 'r0509', 'r1453', 'HMR_4343', 'ACONTm', 'PDHm', \n",
    "    'r0426', 'r0383', 'r0555', 'r1393', 'NICRNS', 'GAUGE-R00648', 'GapFill-R08726', \n",
    "    'RE2915M', 'HMR_3288', 'HMR_1325', 'HMR_7599', 'r1431', 'r1433', 'RE2439C', 'r0082', 'r0791', 'r1450', \n",
    "    'GAUGE-R00270', 'GAUGE-R02285', 'GAUGE-R04283', 'GAUGE-R06127', 'GAUGE-R06128', 'GAUGE-R06238', \n",
    "    'GAUGE-R00524', 'GAUGE-R03326', 'RE3477C', 'AAPSAS', 'RE3347C', 'HMR_0960', 'HMR_0980', 'RE3476C', \n",
    "    'r0708', 'r0777', 'r0084', 'r0424', 'r0698', '3HDH260p', 'HMR_3272', 'ACOAD183n3m', 'HMR_1996', \n",
    "    'GapFill-R01463', 'GapFill-R04807', 'r1468', 'r2435', 'r0655', 'r0603', 'r0541', 'RE0383C', 'HMR_1329', \n",
    "    'TYRA', 'NRPPHRt_2H', 'GAUGE-R07364', 'GapFill-R03599', 'ARD', 'RE3095C', 'RE3104C', 'RE3104R', 'ACONT', \n",
    "    'ACONTm', 'ICDHxm', 'ICDHy', 'AKGDm', 'r0083', 'r0425', 'r0556', 'NH4t4r', 'PROPAT4te', 'r0085', \n",
    "    'r0156', 'r0464', 'ABUTDm', 'OIVD1m', 'OIVD2m', 'OIVD3m', 'r2194', 'r2202', 'HMR_9617', 'r2197', \n",
    "    'r2195', '2OXOADOXm', 'r2328', 'r0386', 'r0451', 'FAS100COA', 'FAS120COA', 'FAS140COA', 'FAS80COA_L', \n",
    "    'r0604', 'r0670', 'r2334', 'r0193', 'r0595', 'r0795', 'GLYCLm', 'MACACI', 'r2193', 'r0779', 'r0669', \n",
    "    'UDCHOLt', 'r2146', 'r2139'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_raw_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_zela = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Sanity Check. Make sure all the exchange reactions in the experimental data are present in the model and that the models ar growing\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('')\n",
    "    print(f'GR1: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    \n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_reactions_list:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            try:\n",
    "                new_gr = model.slim_optimize()\n",
    "                print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "                \n",
    "                # Check if the model's growth rate is infeasible or close to zero\n",
    "                if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                    print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                    reaction.bounds = original_bounds  # Revert the bounds if infeasible\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f'Error optimizing model after knocking out {atp_reaction}: {e}')\n",
    "                # Revert the reaction bounds if an error occurs\n",
    "                reaction.bounds = original_bounds\n",
    "\n",
    "    print('')\n",
    "    print('Missing reactions from exp data:')\n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')\n",
    "    print('')\n",
    "\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print('---------------------------------------')\n",
    "    print(f'------------{model.id}------------')\n",
    "    print('')\n",
    "    print(f'Objective:{model.objective}')\n",
    "    print('---------------------------------------')\n",
    "    print(f'GR: {model.slim_optimize()}')\n",
    "    exchanges = list(uptsec_wt.keys())\n",
    "    model_ex = [r.id for r in model.boundary]\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    \n",
    "    # Remove ATP loop reactions\n",
    "    for atp_reaction in atp_reactions_list:\n",
    "        if atp_reaction in model_reactions:\n",
    "            # Knock out the reaction temporarily\n",
    "            reaction = model.reactions.get_by_id(atp_reaction)\n",
    "            original_bounds = reaction.bounds  # Store the original bounds to revert later\n",
    "            reaction.knock_out()\n",
    "            \n",
    "            # Test if the model remains feasible after knocking out the reaction\n",
    "            try:\n",
    "                new_gr = model.slim_optimize()\n",
    "                print(f'GR2 (After knocking out {atp_reaction}): {new_gr}')\n",
    "                \n",
    "                # Check if the model's growth rate is infeasible or close to zero\n",
    "                if math.isnan(new_gr) or new_gr < 1e-6:\n",
    "                    print(f'{atp_reaction} causes infeasibility. Reverting knockout.')\n",
    "                    reaction.bounds = original_bounds  # Revert the bounds if infeasible\n",
    "                else:\n",
    "                    print(f'{atp_reaction} successfully knocked out without infeasibility.')\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f'Error optimizing model after knocking out {atp_reaction}: {e}')\n",
    "                # Revert the reaction bounds if an error occurs\n",
    "                reaction.bounds = original_bounds\n",
    "\n",
    "    \n",
    "    for ex in exchanges:\n",
    "        if ex not in model_ex:\n",
    "            print(ex)\n",
    "    print('---------------------------------------')\n",
    "    print('---------------------------------------')\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfba_solutions_fluxes = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# WT\n",
    "\n",
    "results = []\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "replicates = {'U1':'Bio141', 'U2':'Bio142', 'U3':'Bio143'}\n",
    "\n",
    "for name,model in wt_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "\n",
    "                     with model as modified_model:\n",
    "\n",
    "                        # Open the bounds for the biomass reaction \n",
    "                        modified_model.reactions.biomass_cho_s.bounds = (0,1000)\n",
    "                         \n",
    "                        # Create a copy of the modified_model before making changes\n",
    "                        pre_modification_model = modified_model.copy()\n",
    "                         \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "                        \n",
    "                        print(f'Calculating Growth Rate for WT Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_wt['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                reaction.lower_bound = 8\n",
    "                            for r,v in uptsec_wt.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "            \n",
    "                            # If pFBA succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = modified_model\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')\n",
    "                            print('Reverting to the original model and retrying pFBA...')\n",
    "\n",
    "                            for rxn in pre_modification_model.boundary:\n",
    "                            # Keep boundaries open for essential metabolites\n",
    "                                if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                    rxn.bounds = (-1000, 1000)\n",
    "                                elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                    rxn.bounds = (-0.001, 1000)\n",
    "                            \n",
    "                            pfba_solution = cobra.flux_analysis.pfba(pre_modification_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate with original model is: {pfba_obj}')\n",
    "\n",
    "                            # If pFBA with original model succeeds, proceed to sampling\n",
    "                            pfba_model_to_sample = pre_modification_model\n",
    "\n",
    "                        results.append([rep_suffix, interval_key, exp_gr, pfba_obj])\n",
    "                        # Save the solution with metadata\n",
    "                        pfba_solutions_fluxes.append({\n",
    "                            \"model\": pre_modification_model.id,\n",
    "                            \"condition\": interval_key,\n",
    "                            \"fluxes\": pfba_solution.fluxes\n",
    "                        })\n",
    "\n",
    "                        \n",
    "                         # Perform flux sampling using ACHRSampler\n",
    "                        #print(\"Performing flux sampling...\")\n",
    "                        #achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                        #samples = 5000 \n",
    "                        #sampled_fluxes = achr_sampler.sample(samples)\n",
    "                        \n",
    "                        # Validate and save the sampled fluxes\n",
    "                        #if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                        #    print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                        #else:\n",
    "                        #    f_name = f\"../Simulations/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                        #    with open(f_name, 'wb') as file:\n",
    "                        #        pickle.dump(sampled_fluxes, file)\n",
    "                        \n",
    "# Creating a DataFrame\n",
    "df_wt = pd.DataFrame(results, columns=['Hue', 'Category', 'X Axis', 'Y Axis'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df_wt\n",
    "# Define markers for categories\n",
    "markers = {\n",
    "    'P0 to P2': 'p',  # Circle\n",
    "    'P2 to P4': 'o',  # Circle\n",
    "    'P4 to P6': 's',   # Square\n",
    "}\n",
    "\n",
    "# Set up the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# Plot each group with different colors for hue and different shapes for categories\n",
    "hues = df['Hue'].unique()\n",
    "colors = plt.colormaps.get_cmap('tab10')\n",
    "\n",
    "for hue_idx, hue in enumerate(hues):\n",
    "    hue_subset = df[df['Hue'] == hue]\n",
    "    for category, marker in markers.items():\n",
    "        subset = hue_subset[hue_subset['Category'] == category]\n",
    "        plt.scatter(subset['X Axis'], subset['Y Axis'], label=f'{hue} - {category}', marker=marker, s=100, color=colors(hue_idx))\n",
    "\n",
    "# Add the identity line y = x\n",
    "plt.plot([0.00, 0.055], [0.00, 0.055], color='red', linestyle='--')\n",
    "\n",
    "# Set evenly distributed ticks from 0.01 to 0.08\n",
    "ticks = [i / 100.0 for i in range(1, 11)]\n",
    "plt.xticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "plt.yticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "\n",
    "# Set limits for both axes\n",
    "plt.xlim(0.0, 0.055)\n",
    "plt.ylim(0.0, 0.055)\n",
    "\n",
    "# Setting the same scale for both axes\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Adding labels and title\n",
    "plt.xlabel('Experimental Growth Rate')\n",
    "plt.ylabel('Simulated Growth Rate')\n",
    "plt.title('WT')\n",
    "plt.legend(title='Hue - Category', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Simulations/growth_rate_pred/bar_plot_growth_rate_wt.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ZeLa\n",
    "\n",
    "results = []\n",
    "\n",
    "objective = 'biomass_cho_s'\n",
    "\n",
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "intervals = {'P0 to P2':'P2', 'P2 to P4':'P4', 'P4 to P6':'P6', 'P6 to P8':'P8', 'P8 to P12':'P12', 'P12 to P14':'P14'}\n",
    "replicates = {'U4':'Bio144', 'U5':'Bio145', 'U6':'Bio146', 'U7':'Bio147', 'U8':'Bio148'}\n",
    "\n",
    "for name,model in zela_models.items():\n",
    "    print(f\"Processing model: {name}\")\n",
    "    # Set lower bounds of the reactions according to the experimental data\n",
    "    for interval_key, interval_model_time in intervals.items():\n",
    "        if interval_model_time in name:  # Match model name with interval\n",
    "            for rep_key, rep_suffix in replicates.items():\n",
    "                if rep_suffix in name:  # Match replicate with model name\n",
    "                    \n",
    "                     with model as modified_model:\n",
    "\n",
    "                        # Open the bounds for the biomass reaction \n",
    "                        modified_model.reactions.biomass_cho_s.bounds = (0,1000)\n",
    "                         \n",
    "                        # Create a copy of the modified_model before making changes\n",
    "                        pre_modification_model = modified_model.copy()\n",
    " \n",
    "                        for rxn in modified_model.boundary:\n",
    "                            if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                rxn.bounds = (-1000, 1000)\n",
    "                            elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                rxn.bounds = (-0.001, 1000)\n",
    "                            elif rxn.id.startswith((\"EX_\", \"SK_\", \"DM_\")):\n",
    "                                rxn.bounds = (0, 1000)  # Close uptake rates for others\n",
    "\n",
    "                        print(f'Calculating Growth Rate for ZeLa Condition:{rep_key,interval_key}')\n",
    "                        exp_gr = uptsec_zela['exp_growth_rate'][(rep_key,interval_key)]\n",
    "                        print(f'Experimental growth rate is: {exp_gr}')\n",
    "                        for reaction in modified_model.reactions:\n",
    "                            if reaction.id == 'EX_etoh_e': #Model creates infeasible solutions when secreting etoh\n",
    "                                continue\n",
    "                            elif reaction.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                reaction.lower_bound = 8\n",
    "                            for r,v in uptsec_zela.items():\n",
    "                                if reaction.id == r:\n",
    "                                    reaction.upper_bound = 1000\n",
    "                                    reaction.lower_bound = v[(rep_key,interval_key)]\n",
    "                                    \n",
    "                        try:\n",
    "                            pfba_solution = cobra.flux_analysis.pfba(modified_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate is: {pfba_obj}')\n",
    "                            \n",
    "                        except Infeasible:\n",
    "                            print(f'Infeasible solution for replicate {rep_key} interval {interval_key}')\n",
    "                            print('Reverting to the original model and retrying pFBA...')\n",
    "\n",
    "                            for rxn in pre_modification_model.reactions:\n",
    "                                if rxn.id in ['EX_h2o_e', 'EX_h_e', 'EX_o2_e', 'EX_hco3_e', 'EX_so4_e', 'EX_pi_e']:\n",
    "                                    rxn.bounds = (-1000, 1000)\n",
    "                                elif rxn.id in ['SK_Asn_X_Ser_Thr_r', 'SK_Tyr_ggn_c', 'SK_Ser_Thr_g', 'SK_pre_prot_r']:\n",
    "                                    rxn.bounds = (-0.001, 1000)\n",
    "                                elif rxn.id == 'ATPM': # Add ATP Maintenance Cost\n",
    "                                    rxn.lower_bound = 8\n",
    "                                \n",
    "                            pfba_solution = cobra.flux_analysis.pfba(pre_modification_model)\n",
    "                            pfba_obj = pfba_solution.fluxes[objective]\n",
    "                            pfba_atp = pfba_solution.fluxes['ATPM']\n",
    "                            print(f'Simulated ATP Maintencance is: {pfba_atp}')\n",
    "                            print(f'Simulated growth rate with original model is: {pfba_obj}')\n",
    "                            \n",
    "                        results.append([rep_suffix, interval_key, exp_gr, pfba_obj])\n",
    "\n",
    "                        # Save the solution with metadata\n",
    "                        pfba_solutions_fluxes.append({\n",
    "                            \"model\": pre_modification_model.id,\n",
    "                            \"condition\": interval_key,\n",
    "                            \"fluxes\": pfba_solution.fluxes\n",
    "                        })\n",
    "                         \n",
    "                        '''\n",
    "                        # Perform flux sampling using ACHRSampler\n",
    "                        print(\"Performing flux sampling...\")\n",
    "                        achr_sampler = ACHRSampler(pfba_model_to_sample)\n",
    "                        samples = 1 \n",
    "                        sampled_fluxes = achr_sampler.sample(samples)\n",
    "                        \n",
    "                        # Validate and save the sampled fluxes\n",
    "                        if 'v' not in achr_sampler.validate(sampled_fluxes).flatten():\n",
    "                            print(f\"Error in sampling for {rep_key}, {interval_key}\")\n",
    "                        else:\n",
    "                            f_name = f\"../Simulations/flux_sampling/Sampling_{name}_{samples}.pkl\"\n",
    "                            with open(f_name, 'wb') as file:\n",
    "                                pickle.dump(sampled_fluxes, file)\n",
    "                        '''\n",
    "    \n",
    "# Creating a DataFrame\n",
    "df_zela = pd.DataFrame(results, columns=['Hue', 'Category', 'X Axis', 'Y Axis'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df_zela\n",
    "# Define markers for categories\n",
    "markers = {\n",
    "    'P0 to P2': 'p',\n",
    "    'P2 to P4': 'o',  # Circle\n",
    "    'P4 to P6': 's',   # Square\n",
    "    'P6 to P8': 'v',\n",
    "}\n",
    "\n",
    "# Set up the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# Plot each group with different colors for hue and different shapes for categories\n",
    "hues = df['Hue'].unique()\n",
    "colors = plt.colormaps.get_cmap('tab10')\n",
    "\n",
    "for hue_idx, hue in enumerate(hues):\n",
    "    hue_subset = df[df['Hue'] == hue]\n",
    "    for category, marker in markers.items():\n",
    "        subset = hue_subset[hue_subset['Category'] == category]\n",
    "        plt.scatter(subset['X Axis'], subset['Y Axis'], label=f'{hue} - {category}', marker=marker, s=100, color=colors(hue_idx))\n",
    "\n",
    "# Add the identity line y = x\n",
    "plt.plot([0.00, 0.055], [0.00, 0.055], color='red', linestyle='--')\n",
    "\n",
    "# Set evenly distributed ticks from 0.01 to 0.08\n",
    "ticks = [i / 100.0 for i in range(1,11)]\n",
    "plt.xticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "plt.yticks(ticks=ticks, labels=[f'{i/100.0:.2f}' for i in range(1, 11)])\n",
    "\n",
    "# Set limits for both axes\n",
    "plt.xlim(0.0, 0.055)\n",
    "plt.ylim(0.0, 0.055)\n",
    "\n",
    "# Setting the same scale for both axes\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Adding labels and title\n",
    "plt.xlabel('Experimental Growth Rate')\n",
    "plt.ylabel('Simulated Growth Rate')\n",
    "plt.title('ZeLa')\n",
    "plt.legend(title='Hue - Category', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Simulations/growth_rate_pred/bar_plot_growth_rate_zela.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for reaction, values in uptsec_zela.items():\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    \n",
    "    # Organize data for plotting\n",
    "    batches = sorted(set(k[0] for k in values.keys()))\n",
    "    timepoints = ['P0 to P2', 'P2 to P4', 'P4 to P6', 'P6 to P8', 'P8 to P12', 'P12 to P14']\n",
    "    \n",
    "    for batch in batches:\n",
    "        batch_data = [values.get((batch, tp), 0) for tp in timepoints]\n",
    "        plt.plot(timepoints, batch_data, marker='o', label=batch)\n",
    "    \n",
    "    plt.title(f'Changes in {reaction} over time')\n",
    "    plt.xlabel('Timepoints')\n",
    "    plt.ylabel('Values')\n",
    "    plt.legend()\n",
    "    plt.xticks(rotation=45)\n",
    "    plt.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for reaction, values in uptsec_wt.items():\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    \n",
    "    # Organize data for plotting\n",
    "    batches = sorted(set(k[0] for k in values.keys()))\n",
    "    timepoints = ['P0 to P2', 'P2 to P4', 'P4 to P6', 'P6 to P8', 'P8 to P12', 'P12 to P14']\n",
    "    \n",
    "    for batch in batches:\n",
    "        batch_data = [values.get((batch, tp), 0) for tp in timepoints]\n",
    "        plt.plot(timepoints, batch_data, marker='o', label=batch)\n",
    "    \n",
    "    plt.title(f'Changes in {reaction} over time')\n",
    "    plt.xlabel('Timepoints')\n",
    "    plt.ylabel('Values')\n",
    "    plt.legend()\n",
    "    plt.xticks(rotation=45)\n",
    "    plt.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Flux Enrichment Analysis <a id='fea'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 1\n",
    "# Use the fluxes from the growth rate calculations to filter reactions with fluxes\n",
    "# Define a cut off for the flux ??? # Plot to visualization of the flux distribution\n",
    "# Run Flux Enrichment Analysis on those reactions to see pathways enriched when optimized by biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 2\n",
    "# Transcriptomic data from cell batches / biorreactiors\n",
    "# Overlay this into the recons and extract the reactions associated to the genes\n",
    "# Extract a reaction vector / \n",
    "# Run Flux Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import hypergeom\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def flux_enrichment_analysis(generic_model, rxn_list, attribute='subsystem'):\n",
    "    \n",
    "    # Check if the attribute exists in the first reaction as a proxy for all\n",
    "    if not hasattr(model.reactions[0], attribute):\n",
    "        raise ValueError(f'Attribute {attribute} not found in model reactions')\n",
    "\n",
    "    # Extract attribute information for all reactions\n",
    "    attribute_values = [getattr(rxn, attribute, 'None') for rxn in model.reactions]\n",
    "    unique_attributes = set(attribute_values)\n",
    "    \n",
    "    # Count occurrences in the model and in the reaction set\n",
    "    model_counts = {attr: attribute_values.count(attr) for attr in unique_attributes}\n",
    "    rxn_set_counts = {attr: 0 for attr in unique_attributes}\n",
    "    for rxn in rxn_list:\n",
    "        rxn_attr = getattr(model.reactions.get_by_id(rxn), attribute, 'None')\n",
    "        rxn_set_counts[rxn_attr] += 1\n",
    "\n",
    "    # Calculate p-values using hypergeometric test\n",
    "    M = len(model.reactions)  # Total number of reactions\n",
    "    n = len(rxn_list)  # Size of reaction set\n",
    "    p_values = []\n",
    "    for attr in unique_attributes:\n",
    "        N = model_counts[attr]  # Total reactions in group\n",
    "        x = rxn_set_counts[attr]  # Reactions in group and in set\n",
    "        p_value = hypergeom.sf(x-1, M, N, n)\n",
    "        p_values.append(p_value)\n",
    "\n",
    "    # Adjust p-values for multiple testing\n",
    "    _, adj_p_values, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "    # Compile results\n",
    "    results = pd.DataFrame({\n",
    "        'Group': list(unique_attributes),\n",
    "        'P-value': p_values,\n",
    "        'Adjusted P-value': adj_p_values,\n",
    "        'Enriched set size': [rxn_set_counts[attr] for attr in unique_attributes],\n",
    "        'Total set size': [model_counts[attr] for attr in unique_attributes],\n",
    "    }).sort_values(by='Adjusted P-value')\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test 1: FEA on reactions active during growth rate optimization\n",
    "Use the fluxes from the growth rate calculations to filter reactions with fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate a vector of active reactions for each one of the conditions\n",
    "\n",
    "active_reactions = []\n",
    "\n",
    "for sol in pfba_solutions_fluxes:\n",
    "    rxns_fluxes = []\n",
    "    for n,f in sol['fluxes'].items():\n",
    "        if f != 0:\n",
    "            rxns_fluxes.append(n)\n",
    "    \n",
    "    active_reactions.append({\n",
    "        \"batch\": sol['model'],\n",
    "        \"condition\": sol['condition'],\n",
    "        \"flux_vector\": rxns_fluxes\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load generic model\n",
    "model = load_json_model('iCHO3595_unblocked.json')\n",
    "\n",
    "# Add atp demand reaction\n",
    "DM_atp = Reaction('DM_atp_c')\n",
    "model.add_reactions([DM_atp])\n",
    "model.reactions.DM_atp_c.build_reaction_from_string('atp_c -->')\n",
    "model.reactions.DM_atp_c.bounds = (0.01, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate FEA results for each one of the vectors generated for each condition\n",
    "\n",
    "fea_results = []\n",
    "\n",
    "# Load generic model\n",
    "#model = load_json_model('iCHO3595_unblocked.json')\n",
    "\n",
    "for fluxes in active_reactions:\n",
    "    results = flux_enrichment_analysis(model, fluxes['flux_vector'], 'subsystem')\n",
    "    fea_results.append({\n",
    "        \"Batch\": fluxes['batch'],\n",
    "        \"Condition\": fluxes['condition'],\n",
    "        \"Results\": results\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Transform p-values to -log10\n",
    "\n",
    "all_data = pd.DataFrame()\n",
    "\n",
    "for result in fea_results:\n",
    "    batch = result['Batch']\n",
    "    temp_df = result['Results'][['Group', 'P-value']].copy()\n",
    "    temp_df.columns = ['Group', f'P-value_{batch}']\n",
    "    if all_data.empty:\n",
    "        all_data = temp_df\n",
    "    else:\n",
    "        all_data = pd.merge(all_data, temp_df, on='Group', how='outer')\n",
    "\n",
    "# Remove groups with a 0 value in all conditions (assuming a '0' value indicates non-significance)\n",
    "significant_filter = (all_data.drop(columns='Group') < 0.05).any(axis=1)\n",
    "filtered_data = all_data[significant_filter]\n",
    "filtered_data.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# Transform the p-values\n",
    "for col in filtered_data.columns:\n",
    "    if col.startswith('P-value'):\n",
    "        filtered_data[f'-log10_{col}'] = -np.log10(filtered_data[col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import zscore\n",
    "\n",
    "# Extract relevant columns for heatmap\n",
    "heatmap_data = filtered_data[[col for col in filtered_data.columns if col.startswith('-log10')]]\n",
    "\n",
    "# Clean column names\n",
    "heatmap_data.columns = heatmap_data.columns.str.replace(r\"-log10_P-value_\", \"\", regex=True)\n",
    "\n",
    "# Normalize the data (z-score)\n",
    "normalized_data = heatmap_data.apply(zscore, axis=1)\n",
    "\n",
    "# Prepare heatmap data\n",
    "normalized_data.index = filtered_data['Group']\n",
    "\n",
    "# Reorder columns: Group P2, P4, P6, and P8 together\n",
    "columns_order = []\n",
    "for phase in ['P2', 'P4', 'P6', 'P8']:\n",
    "    columns_order.extend([col for col in normalized_data.columns if phase in col])\n",
    "\n",
    "# Reorder the DataFrame\n",
    "normalized_data = normalized_data[columns_order]\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(20, 20))\n",
    "sns.heatmap(normalized_data, annot=True, cmap=\"viridis\", cbar_kws={'label': 'Z-score of -log10(P-value)'}, vmin=-1.5, vmax=1.5)\n",
    "plt.title('Heatmap of Z-score Normalized -log10(P-values) Across Conditions', fontsize=20)\n",
    "plt.xlabel('Conditions', fontsize=15)\n",
    "plt.ylabel('Groups', fontsize=15)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Simulations/flux_enrichment_analysis/FEA_heatmap_all.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the columns based on the presence of \"P2 to P4\" or \"P4 to P6\"\n",
    "subset_df = heatmap_data.filter(regex='_P4_|_P6_')\n",
    "\n",
    "# Normalize the data (z-score)\n",
    "normalized_data = subset_df.apply(zscore, axis=1)\n",
    "\n",
    "# Prepare heatmap data\n",
    "normalized_data.index = filtered_data['Group']\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(20, 20))\n",
    "sns.heatmap(normalized_data, annot=True, cmap=\"viridis\", cbar_kws={'label': 'Z-score of -log10(P-value)'}, vmin=-1.5, vmax=1.5)\n",
    "plt.title('Heatmap of Z-score Normalized -log10(P-values) Across Conditions', fontsize=20)\n",
    "plt.xlabel('Conditions', fontsize=15)\n",
    "plt.ylabel('Groups', fontsize=15)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Simulations/flux_enrichment_analysis/FEA_heatmap_p2_p6.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
