{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b4c0343d-7431-4db9-9bc8-e354a3a70b4e",
   "metadata": {},
   "source": [
    "## Comparison of Metabolic Reconstructions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f793376b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "from collections import Counter\n",
    "from functools import reduce\n",
    "from cobra.io import read_sbml_model, load_json_model, load_matlab_model\n",
    "from google_sheet import GoogleSheet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1b218fa-7f54-4547-818d-bb9f1bf8a902",
   "metadata": {},
   "source": [
    "## 1. Subsystems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d8b6c09",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate models for comparison\n",
    "icho1766 = read_sbml_model('../Data/reconciliation/models/iCHOv1_final.xml')\n",
    "icho2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "#icho2291_DC = read_sbml_model('../Data/Reconciliation/models/iCHO2291_DC.xml')\n",
    "icho2101 = load_json_model('../Data/Reconciliation/models/iCHO2101.json')\n",
    "recon3d = load_matlab_model('../Data/Reconciliation/models/Recon3D_301.mat')\n",
    "iCHO3000 = load_json_model('iCHO3595.json')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c063085",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate datasets from Google Sheet ----- #####\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "# Read data from the Google Sheet\n",
    "sheet_rxns = 'Rxns'\n",
    "rxns = sheet.read_google_sheet(sheet_rxns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5ccb98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Conf. Score distribution throughout the dataset --- ####\n",
    "data_numeric = pd.to_numeric(rxns['Conf. Score'], errors='coerce')\n",
    "data_numeric = data_numeric.dropna()\n",
    "\n",
    "# Plotting the histogram\n",
    "plt.figure(figsize=(10, 4))\n",
    "plt.hist(data_numeric, bins='auto', color='steelblue', edgecolor='black')\n",
    "plt.grid(color='gray', linestyle='--', linewidth=0.5)\n",
    "plt.xticks(data_numeric.unique())\n",
    "plt.xlabel('Values', fontsize=12, fontweight='bold')\n",
    "plt.ylabel('Frequency', fontsize=12, fontweight='bold')\n",
    "plt.title('Confidence Score Distribution', fontsize=14, fontweight='bold')\n",
    "\n",
    "plt.savefig('../Analyses/conf_score_distribution.png', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67a59651",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Calculate the amount of reactions per Subsystem ---- ####\n",
    "\n",
    "#iCHO 1766\n",
    "subs = []\n",
    "for reaction in icho1766.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho1766_subs = Counter(subs)\n",
    "\n",
    "#iCHO 2291\n",
    "subs = []\n",
    "for reaction in icho2291.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho2291_subs = Counter(subs)\n",
    "\n",
    "# iCHO 2101\n",
    "subs = []\n",
    "for reaction in icho2101.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho2101_subs = Counter(subs)\n",
    "\n",
    "# Recon3D\n",
    "subs = []\n",
    "for reaction in recon3d.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "recon3d_subs = Counter(subs)\n",
    "\n",
    "#iCHO 3644\n",
    "subs = []\n",
    "for reaction in iCHO3000.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "iCHO3000_subs = Counter(subs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874ec2d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping of subsystems to categories\n",
    "category_mapping = {\n",
    "    'Amino Acid Metabolism and Derivatives': [\n",
    "        'ARGININE AND PROLINE METABOLISM', 'METHIONINE METABOLISM', 'AMINO ACID DERIVATIVE', \n",
    "        'GLYCINE, SERINE, AND THREONINE METABOLISM', 'HISTIDINE METABOLISM', 'LYSINE METABOLISM', \n",
    "        'METHIONINE AND CYSTEINE METABOLISM', 'TYROSINE METABOLISM', 'VALINE, LEUCINE, AND ISOLEUCINE METABOLISM',\n",
    "        'ALANINE AND ASPARTATE METABOLISM', 'AMINOSUGAR METABOLISM', 'SELENOAMINO ACID METABOLISM',\n",
    "        'GLYCINE, SERINE, ALANINE AND THREONINE METABOLISM'\n",
    "    ],\n",
    "    'Carbohydrate and Energy Metabolism': [\n",
    "        'GLYCOLYSIS/GLUCONEOGENESIS', 'FRUCTOSE AND MANNOSE METABOLISM', 'CENTRAL CARBON AND ENERGY METABOLISM',\n",
    "        'GALACTOSE METABOLISM', 'PENTOSE AND GLUCURONATE INTERCONVERSIONS', 'STARCH AND SUCROSE METABOLISM',\n",
    "        'PENTOSE PHOSPHATE PATHWAY', 'PYRUVATE METABOLISM'\n",
    "    ],\n",
    "    'Lipid and Fatty Acid Metabolism': [\n",
    "        'FATTY ACID ELONGATION', 'FATTY ACID METABOLISM', 'CHOLESTEROL METABOLISM', 'EICOSANOID METABOLISM',\n",
    "        'ARACHIDONIC ACID METABOLISM', 'BILE ACID SYNTHESIS', 'FATTY ACID SYNTHESIS', 'GLYCEROLIPID METABOLISM',\n",
    "        'SPHINGOLIPID METABOLISM', 'SQUALENE AND CHOLESTEROL SYNTHESIS', 'STEROID METABOLISM', 'TRIACYLGLYCEROL SYNTHESIS',\n",
    "        'VITAMIN A METABOLISM', 'VITAMIN D METABOLISM', 'LINOLEATE METABOLISM', 'FATTY ACID OXIDATION'\n",
    "    ],\n",
    "    'Nucleotide and Nucleic Acid Metabolism': [\n",
    "        'PURINE CATABOLISM', 'PYRIMIDINE BIOSYNTHESIS', 'NUCLEOTIDE SALVAGE PATHWAY', 'NUCLEOTIDES',\n",
    "        'NUCLEOTIDE INTERCONVERSION', 'PURINE SYNTHESIS', 'PYRIMIDINE SYNTHESIS', 'PYRIMIDINE CATABOLISM',\n",
    "        'NUCLEOTIDE SUGAR METABOLISM'\n",
    "    ],\n",
    "    'Vitamin and Cofactor Metabolism': [\n",
    "        'VITAMIN B12 METABOLISM', 'VITAMIN C METABOLISM', 'VITAMIN D', 'VITAMIN A METABOLISM',\n",
    "        'VITAMIN E METABOLISM', 'VITAMIN B2 METABOLISM', 'VITAMIN B6 METABOLISM', 'FOLATE METABOLISM',\n",
    "        'BIOTIN METABOLISM', 'THIAMINE METABOLISM', 'TETRAHYDROBIOPTERIN METABOLISM', 'VITAMIN K METABOLISM',\n",
    "        'VITAMIN AND COFACTOR METABOLISM'\n",
    "    ],\n",
    "    'Glycan, Glycosaminoglycan, and Glycolipid Metabolism': [\n",
    "        'N-GLYCAN BIOSYNTHESIS', 'GLYCOSPHINGOLIPID METABOLISM', 'CHONDROITIN SYNTHESIS', 'KERATAN SULFATE DEGRADATION',\n",
    "        'HEPARAN SULFATE DEGRADATION', 'HYALURONAN METABOLISM', 'N-GLYCAN DEGRADATION', 'O-GLYCAN SYNTHESIS',\n",
    "        'GLYCOSPHINGOLIPID METABOLISM', 'KERATAN SULFATE SYNTHESIS', 'N-GLYCAN SYNTHESIS', 'N-GLYCAN METABOLISM',\n",
    "        'O-GLYCAN METABOLISM'\n",
    "    ],\n",
    "    'Protein and Peptide Metabolism': [\n",
    "        'PROTEIN PRODUCTION', 'PROTEIN ASSEMBLY', 'PROTEIN DEGRADATION', 'PROTEIN MODIFICATION',\n",
    "        'PEPTIDE METABOLISM', 'UBIQUINONE SYNTHESIS', 'HEME SYNTHESIS', 'PROTEIN PRODUCTION'\n",
    "    ],\n",
    "    'Xenobiotic and Drug Metabolism': [\n",
    "        'CYP METABOLISM', 'CYTOCHROME METABOLISM', 'XENOBIOTICS METABOLISM', 'ROS DETOXIFICATION',\n",
    "        'DRUG METABOLISM', 'ALKALOID SYNTHESIS'\n",
    "    ],\n",
    "    'Miscellaneous and Complex Pathways': [\n",
    "        'BIOMASS SYNTHESIS', 'R GROUP SYNTHESIS', 'MISCELLANEOUS', 'BIOMASS AND MAINTENANCE FUNCTIONS',\n",
    "        'UNASSIGNED', 'OTHER CARBON METABOLISM'\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "090f720b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# List to store reactions that were not changed along with their subsystems in icho2291_tmp\n",
    "not_changed_reactions = []\n",
    "icho2291_tmp = icho2291.copy()\n",
    "# Replace the subsystem in iCHO2291_tmp with the corresponding subsystem from iCHO2291_DC_tmp based on reaction names\n",
    "for reaction in icho2291_tmp.reactions:\n",
    "    try:\n",
    "        matching_reaction = icho2291_DC.reactions.get_by_id(reaction.id)\n",
    "        # If a matching reaction is found, replace the subsystem\n",
    "        if matching_reaction:\n",
    "            print(f\"Reaction: {reaction.subsystem}\")\n",
    "            reaction.subsystem = matching_reaction.subsystem\n",
    "            print(f\" has been changed to {matching_reaction.subsystem}\")\n",
    "    except:\n",
    "        print(f\"Reaction: {reaction} not changed\")\n",
    "        # Add the reaction to the list of reactions not changed\n",
    "        not_changed_reactions.append({\n",
    "            'Reaction ID': reaction.id,\n",
    "            'Reaction Name': reaction.name,\n",
    "            'Subsystem in icho2291_tmp': reaction.subsystem\n",
    "        })\n",
    "\n",
    "# Convert the list of not changed reactions to a DataFrame\n",
    "not_changed_df = pd.DataFrame(not_changed_reactions)\n",
    "\n",
    "# Display the DataFrame\n",
    "print(not_changed_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f2182c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through all reactions in the icho2291_tmp model\n",
    "for reaction in icho2291_tmp.reactions:\n",
    "    # 1) Replace \"GLYCINE, SERINE AND THREONINE METABOLISM\" with \"GLYCINE, SERINE, AND THREONINE METABOLISM\"\n",
    "    if reaction.subsystem == \"GLYCINE, SERINE AND THREONINE METABOLISM\":\n",
    "        reaction.subsystem = \"GLYCINE, SERINE, ALANINE AND THREONINE METABOLISM\"\n",
    "        print(f\"Updated subsystem for {reaction.id}: {reaction.subsystem}\")\n",
    "    \n",
    "    # 2) Replace \"N-GLYCAN SYNTHESIS\" with \"N-GLYCAN METABOLISM\"\n",
    "    elif reaction.subsystem == \"N-GLYCAN SYNTHESIS\":\n",
    "        reaction.subsystem = \"N-GLYCAN METABOLISM\"\n",
    "        print(f\"Updated subsystem for {reaction.id}: {reaction.subsystem}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c437ecd4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tmp = not_changed_df['Subsystem in icho2291_tmp']\n",
    "\n",
    "# Loop through tmp with enumerate to get both index and subsys\n",
    "for idx, subsys in enumerate(tmp):\n",
    "    if subsys == \"Exchange/demand/sink reaction\":\n",
    "        continue\n",
    "    else:\n",
    "        print(idx, subsys)\n",
    "        print(not_changed_df['Reaction ID'].iloc[idx])\n",
    "        tmp_name = not_changed_df['Reaction ID'].iloc[idx]\n",
    "        tmp_reaction = icho2291.reactions.get_by_id(tmp_name)\n",
    "        print(tmp_reaction)\n",
    "        \n",
    "        # If the reaction name matches \"Asn_X_Ser/Thrtr\", update only that reaction's subsystem\n",
    "        if tmp_name == \"Asn_X_Ser/Thrtr\":\n",
    "            # Get the specific reaction in icho2291_tmp using the reaction ID\n",
    "            target_reaction = icho2291_tmp.reactions.get_by_id(tmp_name)\n",
    "            target_reaction.subsystem = \"TRANSPORT, ENDOPLASMIC RETICULAR\"\n",
    "            print(f\"Updated subsystem for {tmp_name}: {target_reaction.subsystem}\")\n",
    "        if tmp_name == \"Ser/Thrtg\":\n",
    "            # Get the specific reaction in icho2291_tmp using the reaction ID\n",
    "            target_reaction = icho2291_tmp.reactions.get_by_id(tmp_name)\n",
    "            target_reaction.subsystem = \"TRANSPORT, GOLGI APPARATUS\"\n",
    "            print(f\"Updated subsystem for {tmp_name}: {target_reaction.subsystem}\")\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6ad8aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the updated subsystem to verify the change\n",
    "updated_reaction = icho2291_tmp.reactions.get_by_id(\"Asn_X_Ser/Thrtr\")\n",
    "print(f\"Updated subsystem for {updated_reaction.id}: {updated_reaction.subsystem}\")\n",
    "updated_reaction = icho2291_tmp.reactions.get_by_id(\"Ser/Thrtg\")\n",
    "print(f\"Updated subsystem for {updated_reaction.id}: {updated_reaction.subsystem}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28c55638",
   "metadata": {},
   "outputs": [],
   "source": [
    "#iCHO 2291_tmp\n",
    "subs = []\n",
    "for reaction in icho2291_tmp.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho2291_tmp_subs = Counter(subs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a469032",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert Counters to DataFrames\n",
    "df_icho1766 = pd.DataFrame.from_dict(icho1766_subs, orient='index', columns=['iCHO1766'])\n",
    "df_icho2291 = pd.DataFrame.from_dict(icho2291_tmp_subs, orient='index', columns=['iCHO2291'])\n",
    "df_icho2101 = pd.DataFrame.from_dict(icho2101_subs, orient='index', columns=['iCHO2101'])\n",
    "# df_recon3d = pd.DataFrame.from_dict(recon3d_subs, orient='index', columns=['Recon3D'])\n",
    "df_iCHO3000 = pd.DataFrame.from_dict(iCHO3000_subs, orient='index', columns=['iCHO3K'])\n",
    "\n",
    "# Combine the DataFrames\n",
    "# combined_df = pd.concat([df_icho1766, df_icho2101, df_icho2291, df_recon3d, df_iCHO3000], axis=1).fillna(0)\n",
    "combined_df = pd.concat([df_icho1766, df_icho2101, df_icho2291, df_iCHO3000], axis=1).fillna(0)\n",
    "\n",
    "# Create a mask where each row is True if all its values are non-zero\n",
    "#non_zero_mask = (combined_df != 0).all(axis=1)\n",
    "\n",
    "# Apply the mask to filter out rows with any zero values\n",
    "#filtered_df = combined_df[non_zero_mask]\n",
    "\n",
    "# Create a mask where each row is True if the index does not start with \"Transport\"\n",
    "mask = ~combined_df.index.str.startswith('TRANSPORT') & ~combined_df.index.str.startswith(\"EXCHANGE\") & ~combined_df.index.str.startswith(\"SINK\") & ~combined_df.index.str.startswith(\"DEMAND\") & ~combined_df.index.str.startswith(\"INTRACELLULAR\") & ~combined_df.index.str.startswith(\"EXTRACELLULAR\")\n",
    "\n",
    "\n",
    "# Apply the mask to the DataFrame to keep only rows that do not start with \"Transport\"\n",
    "filtered_df = combined_df[mask]\n",
    "\n",
    "# Plot \n",
    "ax = filtered_df.plot(kind='barh', figsize=(20, 25), width=1.8) \n",
    "\n",
    "# Title and labels\n",
    "plt.title('Subsystem Reaction Counts Across Different CHO Recons', fontsize=28)\n",
    "plt.ylabel('Subsystem', fontsize=26)  # This was previously 'xlabel'\n",
    "plt.xlabel('Reaction Count', fontsize=26)  # This was previously 'ylabel'\n",
    "\n",
    "# Adjust the layout\n",
    "plt.tight_layout()  \n",
    "\n",
    "# Customize legend\n",
    "plt.legend(fontsize=24)\n",
    "\n",
    "# Save the plot to a file\n",
    "plt.savefig('../Analyses/recons_comparisons/subsystem_reaction_counts.png', dpi=300)  \n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99560a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71da257e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the filtered DataFrame to a CSV file\n",
    "csv_file_path = '../Analyses/recons_comparisons/subsystem_reaction_counts.csv'\n",
    "\n",
    "# Save to CSV\n",
    "filtered_df.to_csv(csv_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe3c08d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the list of subsystems related to 'Amino Acid Metabolism and Derivatives'\n",
    "amino_acid_subsystems = category_mapping['Amino Acid Metabolism and Derivatives']\n",
    "\n",
    "# Filter the combined DataFrame for rows where the index matches any of the amino acid subsystems\n",
    "filtered_amino_acid_df = combined_df[combined_df.index.isin(amino_acid_subsystems)]\n",
    "\n",
    "# Plot the filtered DataFrame\n",
    "ax = filtered_amino_acid_df.plot(kind='barh', figsize=(10, 8), width=0.8) \n",
    "\n",
    "# Title and labels\n",
    "plt.title('Amino Acid Metabolism and Derivatives Reaction Counts', fontsize=24)\n",
    "plt.ylabel('Subsystem', fontsize=18)\n",
    "plt.xlabel('Reaction Count', fontsize=18)\n",
    "\n",
    "# Adjust the layout\n",
    "plt.tight_layout()  \n",
    "\n",
    "# Customize legend\n",
    "plt.legend(fontsize=14)\n",
    "\n",
    "# Save the plot to a file\n",
    "plt.savefig('../Analyses/recons_comparisons/amino_acid_subsystems_reaction_counts.png', dpi=300)  \n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "315c7344",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the list of subsystems related to 'Amino Acid Metabolism and Derivatives'\n",
    "amino_acid_subsystems = category_mapping['Amino Acid Metabolism and Derivatives']\n",
    "\n",
    "# Filter the combined DataFrame for rows where the index matches any of the amino acid subsystems\n",
    "filtered_amino_acid_df = combined_df[combined_df.index.isin(amino_acid_subsystems)]\n",
    "\n",
    "# Set figure size to a 16:9 aspect ratio (e.g., 16 inches wide, 9 inches tall)\n",
    "fig, ax = plt.subplots(figsize=(12, 6.5))\n",
    "\n",
    "# Plot the filtered DataFrame\n",
    "filtered_amino_acid_df.plot(kind='barh', ax=ax, width=0.8)\n",
    "ax.tick_params(axis='y', labelsize=13)  \n",
    "# Title and labels\n",
    "plt.title('Amino Acid Metabolism and Derivatives\\nReaction Counts', fontsize=20, loc='center')  # Title in the center\n",
    "plt.ylabel('Subsystem', fontsize=18)\n",
    "plt.xlabel('Reaction Count', fontsize=18)\n",
    "\n",
    "# Adjust the layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Customize legend\n",
    "plt.legend(fontsize=14)\n",
    "\n",
    "# Save the plot to a file\n",
    "plt.savefig('../Analyses/recons_comparisons/amino_acid_subsystems_reaction_counts.png', dpi=300)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bc269fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the list of subsystems related to 'Amino Acid Metabolism and Derivatives'\n",
    "amino_acid_subsystems = category_mapping['Amino Acid Metabolism and Derivatives']\n",
    "\n",
    "# Filter the combined DataFrame for rows where the index matches any of the amino acid subsystems\n",
    "filtered_amino_acid_df = combined_df[combined_df.index.isin(amino_acid_subsystems)]\n",
    "\n",
    "# Set figure size to a 16:9 aspect ratio (e.g., 12 inches wide, 6.5 inches tall)\n",
    "fig, ax = plt.subplots(figsize=(12, 6.5))\n",
    "\n",
    "# Plot the filtered DataFrame\n",
    "filtered_amino_acid_df.plot(kind='barh', ax=ax, width=0.8)\n",
    "\n",
    "# Move the y-axis to the right\n",
    "ax.yaxis.tick_right()  # Move y-axis ticks and labels to the right\n",
    "ax.yaxis.set_label_position(\"right\")  # Move y-axis label to the right\n",
    "\n",
    "# Invert x-axis so that bars start from the right side\n",
    "ax.invert_xaxis()  # Make the x-axis go from right to left\n",
    "\n",
    "# Title and labels\n",
    "plt.title('Amino Acid Metabolism and Derivatives Reaction Counts', fontsize=20, loc='left', x=0.15)  # Title in the center\n",
    "plt.ylabel('Subsystem', fontsize=18)\n",
    "plt.xlabel('Reaction Count', fontsize=18)\n",
    "\n",
    "# Increase font size for y-axis labels (on the right side)\n",
    "ax.tick_params(axis='y', labelsize=13)\n",
    "\n",
    "# Adjust the layout to avoid overlap\n",
    "plt.tight_layout()\n",
    "\n",
    "# Customize the legend\n",
    "plt.legend(fontsize=14)\n",
    "\n",
    "# Save the plot to a file\n",
    "plt.savefig('../Analyses/recons_comparisons/amino_acid_subsystems_reaction_counts_right_y_axis.png', dpi=300)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b6009a3-7b1c-4a45-aa23-2a71c189a9d3",
   "metadata": {},
   "source": [
    "## 2. Reactions, Genes and Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5bc92ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reaction counts for each recon\n",
    "icho1766_rxns = len(icho1766.reactions)\n",
    "icho2291_rxns = len(icho2291.reactions)\n",
    "icho2101_rxns = len(icho2101.reactions)\n",
    "recon3d_rxns = len(recon3d.reactions)\n",
    "icho3000_rxns = len(iCHO3000.reactions)\n",
    "\n",
    "#Metabolite counts for each recon\n",
    "icho1766_mets = len(icho1766.metabolites)\n",
    "icho2291_mets = len(icho2291.metabolites)\n",
    "icho2101_mets = len(icho2101.metabolites)\n",
    "recon3d_mets = len(recon3d.metabolites)\n",
    "icho3000_mets = len(iCHO3000.metabolites)\n",
    "\n",
    "# Gene counts for each recon\n",
    "icho1766_gens = len(icho1766.genes)\n",
    "icho2291_gens = len(icho2291.genes)\n",
    "icho2101_gens = len(icho2101.genes)\n",
    "recon3d_gens = len(recon3d.genes)\n",
    "icho3000_gens = len(iCHO3000.genes)\n",
    "\n",
    "# Subsystem count for each recon\n",
    "icho1766_subs_count = len(icho1766_subs)\n",
    "icho2291_subs_count = len(icho2291_subs)\n",
    "icho2101_subs_count = len(icho2101_subs)\n",
    "recon3d_subs_count = len(recon3d_subs)\n",
    "icho3000_subs_count = len(iCHO3000_subs)\n",
    "\n",
    "\n",
    "# Creating a DataFrame from the counts\n",
    "data = {\n",
    "    'iCHO1766': [icho1766_rxns, icho1766_mets, icho1766_gens],# icho1766_subs_count],\n",
    "    'iCHO2101': [icho2101_rxns, icho2101_mets, icho2101_gens],# icho2101_subs_count],\n",
    "    'iCHO2291': [icho2291_rxns, icho2291_mets, icho2291_gens],# icho2291_subs_count],\n",
    "    'iCHO3K': [icho3000_rxns, icho3000_mets, icho3000_gens],# icho3000_subs_count],\n",
    "    #'Recon3D': [recon3d_rxns, recon3d_mets, recon3d_gens, recon3d_subs_count]\n",
    "}\n",
    "\n",
    "counts_df = pd.DataFrame(data, index=['Reactions', 'Metabolites', 'Genes'])\n",
    "\n",
    "# Improved aesthetics for a research publication plot\n",
    "def create_research_plot(counts_df):\n",
    "    # Set a professional color palette\n",
    "    colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728']\n",
    "    model_names = ['iCHO1766', 'iCHO2101', 'iCHO2291', 'iCHO3K']\n",
    "\n",
    "    # Create subplots\n",
    "    fig, axes = plt.subplots(nrows=2, ncols=2, figsize=(7, 6))\n",
    "    \n",
    "    # Turn off the first subplot\n",
    "    ax1, ax2, ax3, ax4 = axes.flatten()\n",
    "    ax1.axis('off')\n",
    "    \n",
    "    # Define the axes for the categories\n",
    "    axes_mapping = {0: ax2, 1: ax3, 2: ax4}\n",
    "    categories = ['Reactions', 'Metabolites', 'Genes']\n",
    "\n",
    "    # A function to annotate the bars with their values\n",
    "    def annotate_bars(ax):\n",
    "        for p in ax.patches:\n",
    "            ax.annotate(f'{p.get_height()}', (p.get_x() + p.get_width() / 2., p.get_height()), \n",
    "                        ha='center', va='bottom', xytext=(0, 3), textcoords='offset points', fontweight='light', fontsize=12)\n",
    "\n",
    "    # Plotting and annotating each category\n",
    "    for index, category in enumerate(categories):\n",
    "        ax = axes_mapping[index]\n",
    "        counts_df.loc[category].plot(kind='bar', ax=ax, title=category, color=colors, edgecolor='black', linewidth=.3, width=0.9, legend=False)\n",
    "        annotate_bars(ax)\n",
    "        ax.set_title(category, fontsize=15, fontweight='light', pad=10)\n",
    "        ax.grid(True, linestyle='-', linewidth=0.5, axis='y')\n",
    "        ax.set_xticks([]) # Remove x-axis ticks\n",
    "        ax.set_xticklabels([])  # Remove x-axis labels\n",
    "        ax.tick_params(axis='y', labelsize=8)\n",
    "\n",
    "    # Remove border of the plots\n",
    "    for ax in [ax2, ax3, ax4]:\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "    # Create custom handles for the legend\n",
    "    custom_handles = [mpatches.Patch(color=color, label=model_name) for color, model_name in zip(colors, model_names)]\n",
    "    fig.legend(handles=custom_handles, loc='upper left',bbox_to_anchor=(0.05, 0.9), fontsize=13, frameon=False, fancybox=False, shadow=False)\n",
    "    \n",
    "    # Adjust spacing between the subplots\n",
    "    plt.subplots_adjust(wspace=0.3, hspace=0.3)\n",
    "    \n",
    "    plt.savefig('../Analyses/recons_comparisons/metabolic_models_comparison.png', dpi=300, bbox_inches='tight')\n",
    "    plt.show()\n",
    "    \n",
    "create_research_plot(counts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "151676d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "list1 = [re.sub(r'\\[.*?\\]', '', m.id) for m in recon3d.metabolites]\n",
    "set1 = set(list1)\n",
    "print(len(list1))\n",
    "print(len(set1))\n",
    "print('----------------------------------------')\n",
    "list2 = [m.id[:-2] for m in icho1766.metabolites]\n",
    "set2 = set(list2)\n",
    "print(len(list2))\n",
    "print(len(set2))\n",
    "print('----------------------------------------')\n",
    "list3 = [m.id[:-2] for m in icho2101.metabolites]\n",
    "set3 = set(list3)\n",
    "print(len(list3))\n",
    "print(len(set3))\n",
    "print('----------------------------------------')\n",
    "list4 = [m.id[:-2] for m in icho2291.metabolites]\n",
    "set4 = set(list4)\n",
    "print(len(list4))\n",
    "print(len(set4))\n",
    "print('----------------------------------------')\n",
    "list5 = [m.id[:-2] for m in iCHO3000.metabolites]\n",
    "set5 = set(list5)\n",
    "print(len(list5))\n",
    "print(len(set5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63261e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "icho3000_subs_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a44a2404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a dictionary to store the counts\n",
    "counts = {\n",
    "    'Model': [],\n",
    "    'Reactions with GPR': [],\n",
    "    'Reactions without GPR': [],\n",
    "    'Total Reactions': []\n",
    "}\n",
    "\n",
    "# Loop through the models and calculate counts\n",
    "models = [icho1766, icho2291, icho2101, iCHO3000]#, recon3d]\n",
    "model_names = ['iCHO1766', 'iCHO2291', 'iCHO2101', 'iCHO3644']#, 'Recon3D']\n",
    "\n",
    "for model, name in zip(models, model_names):\n",
    "    rxn_with_gpr = sum(1 for rxn in model.reactions if str(rxn.gpr) not in ['', 'nan'])\n",
    "    rxn_without_gpr = sum(1 for rxn in model.reactions if str(rxn.gpr) in ['', 'nan'])\n",
    "    total = len(model.reactions)\n",
    "\n",
    "    counts['Model'].append(name)\n",
    "    counts['Reactions with GPR'].append(rxn_with_gpr)\n",
    "    counts['Reactions without GPR'].append(rxn_without_gpr)\n",
    "    counts['Total Reactions'].append(total)\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "counts_df = pd.DataFrame(counts)\n",
    "counts_df['%'] = (counts_df['Reactions with GPR']/counts_df['Total Reactions'])*100\n",
    "\n",
    "# Setting the index to the model names\n",
    "counts_df.set_index('Model', inplace=True)\n",
    "\n",
    "def plot_gpr_coverage(counts_df):\n",
    "    # Create a plot with the right size and stacked bars directly\n",
    "    ax = counts_df[['Reactions with GPR', 'Reactions without GPR']].plot(\n",
    "        kind='bar', stacked=True, color=['#1f77b4', '#ff7f0e'], edgecolor='black', linewidth=.3, width=0.8, legend=False, figsize=(6, 6))\n",
    "    plt.title('GPR Coverage', fontsize=20, pad=10)\n",
    "\n",
    "\n",
    "    # Label configuration\n",
    "    ax.tick_params(axis='x', rotation=0)  # Set rotation for x ticks if needed\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=45)  # Rotate labels if needed\n",
    "    ax.set_xlabel('')\n",
    "\n",
    "    # Grid and axis adjustments\n",
    "    ax.grid(True, linestyle='-', linewidth=0.5, axis='y')\n",
    "    ax.set_axisbelow(True)  # Ensure grid is behind the bars\n",
    "\n",
    "    # Hide x-axis ticks\n",
    "    ax.tick_params(axis='x', which='both', length=0)\n",
    "    ax.tick_params(axis='y', labelsize=8)\n",
    "\n",
    "    # Remove plot frame borders\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(False)\n",
    "        \n",
    "    # Legend configuration\n",
    "    ax.legend(loc='center left',bbox_to_anchor=(0.1, 1.4), fontsize=16, frameon=False, fancybox=False, shadow=False)\n",
    "\n",
    "\n",
    "    # Optimize layout\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Save the figure\n",
    "    plt.savefig('../Analyses/recons_comparisons/gpr_coverage.png', dpi=300, bbox_inches='tight')\n",
    "\n",
    "    # Show the plot\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "# Call the function with the DataFrame\n",
    "plot_gpr_coverage(counts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7291d20-7112-41e3-a462-78fde3114615",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1c382b9",
   "metadata": {},
   "source": [
    "## 3. Gene Essentiality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bae4b2eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from scipy.io import loadmat\n",
    "from time import process_time\n",
    "import cobra\n",
    "from cobra.io import load_json_model, read_sbml_model\n",
    "from cobra.flux_analysis.variability import find_essential_genes, find_essential_reactions\n",
    "\n",
    "iCHO3K = load_json_model('iCHO3595.json')\n",
    "iCHO1766 = read_sbml_model('../Data/Reconciliation/models/iCHOv1_final.xml')\n",
    "iCHO2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "iCHO2101 = load_json_model('../Data/Reconciliation/models/iCHO2101.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b7178d28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset for the experimentally validated essential genes\n",
    "\n",
    "exp_essential_genes = pd.read_csv('../Data/Gene_Essentiality/cho_essential_genes.csv', sep='\\t')\n",
    "exp_essential_genes['gene'] = exp_essential_genes['gene'].str.replace('__1$', '', regex=True)\n",
    "# Extract all gene names\n",
    "all_gene_names = set()\n",
    "for names in exp_essential_genes['gene']:\n",
    "    for name in names.split('|'): #Extract the names of the genes with two aliases\n",
    "        all_gene_names.add(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "64fefc52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "' In case of retrieving the gene ID please activate this code\\n# Generation gene ID of exp_essential_genes with biopython\\nfrom Bio import Entrez\\nimport pandas as pd\\n\\n# Set your email (this is required by NCBI for accessing their services)\\nEntrez.email = \"dh.choi@orcid\"\\n\\n# Your dataset with gene symbols (replace with actual dataframe)\\ngene_symbols = exp_essential_genes.gene  # Example gene symbols\\n\\n# Function to fetch gene ID using Entrez API\\ndef get_gene_id(gene_symbol):\\n    search_handle = Entrez.esearch(db=\"gene\", term=f\"{gene_symbol}[Gene] AND Cricetulus griseus[Organism]\")\\n    record = Entrez.read(search_handle)\\n    search_handle.close()\\n    if record[\"IdList\"]:\\n        return record[\"IdList\"][0]  # Return the first gene ID found\\n    else:\\n        return None\\n\\n# Create a DataFrame to store gene symbols and their corresponding gene IDs\\ndf = pd.DataFrame({\\'gene\\': gene_symbols})\\ndf[\\'entrez_gene_id\\'] = df[\\'gene\\'].apply(get_gene_id)\\n\\n# Display the result\\nprint(df)\\n\\n# Save the DataFrame to a pickle file\\nwith open(\"gene_id_mapping.pkl\", \"wb\") as f:\\n    pickle.dump(df, f)\\n\\n# Save the result to a CSV file as well, if needed\\ndf.to_csv(\"gene_id_mapping.csv\", index=False)\\n\\n'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''' In case of retrieving the gene ID please activate this code\n",
    "# Generation gene ID of exp_essential_genes with biopython\n",
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "\n",
    "# Set your email (this is required by NCBI for accessing their services)\n",
    "Entrez.email = \"dh.choi@orcid\"\n",
    "\n",
    "# Your dataset with gene symbols (replace with actual dataframe)\n",
    "gene_symbols = exp_essential_genes.gene  # Example gene symbols\n",
    "\n",
    "# Function to fetch gene ID using Entrez API\n",
    "def get_gene_id(gene_symbol):\n",
    "    search_handle = Entrez.esearch(db=\"gene\", term=f\"{gene_symbol}[Gene] AND Cricetulus griseus[Organism]\")\n",
    "    record = Entrez.read(search_handle)\n",
    "    search_handle.close()\n",
    "    if record[\"IdList\"]:\n",
    "        return record[\"IdList\"][0]  # Return the first gene ID found\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Create a DataFrame to store gene symbols and their corresponding gene IDs\n",
    "df = pd.DataFrame({'gene': gene_symbols})\n",
    "df['entrez_gene_id'] = df['gene'].apply(get_gene_id)\n",
    "\n",
    "# Display the result\n",
    "print(df)\n",
    "\n",
    "# Save the DataFrame to a pickle file\n",
    "with open(\"gene_id_mapping.pkl\", \"wb\") as f:\n",
    "    pickle.dump(df, f)\n",
    "\n",
    "# Save the result to a CSV file as well, if needed\n",
    "df.to_csv(\"gene_id_mapping.csv\", index=False)\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8d02bbe8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          gene entrez_gene_id\n",
      "0          Ubc      100689267\n",
      "1       Ncaph2      100756862\n",
      "2        Huwe1      100757403\n",
      "3      Isg20l2      100752909\n",
      "4        Coq8b      100770066\n",
      "...        ...            ...\n",
      "1975     Sprtn      100764217\n",
      "1976     Perm1      100773289\n",
      "1977   Tmem147      100751744\n",
      "1978    Rnf216      100758416\n",
      "1979  Rnaseh2b      100751948\n",
      "\n",
      "[1980 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load the DataFrame from the pickle file\n",
    "with open(\"gene_id_mapping.pkl\", \"rb\") as f:\n",
    "    df_loaded = pickle.load(f)\n",
    "\n",
    "# Display the loaded DataFrame\n",
    "print(df_loaded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6385072e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate \"Intervals dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_wt = pickle.load(file)\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_intrvl_zela_dict.pkl', 'rb') as file:\n",
    "    uptsec_intrvl_zela = pickle.load(file)\n",
    "    \n",
    "temp_dict_zela = uptsec_intrvl_zela\n",
    "temp_dict_wt = uptsec_intrvl_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "532b99c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def constraint_model_feg(model, time, exp_dict, objective=None, solver=None, make_GM=None):\n",
    "    \"\"\"\n",
    "    Constrain the model with experimental data and calculate the objective function value using pFBA.\n",
    "    \n",
    "    Parameters:\n",
    "    - model: COBRA model to be constrained.\n",
    "    - time: Time interval for applying the experimental constraints. Default is 'P2 to P4'.\n",
    "    - objective: The objective reaction. Default is 'biomass_cho_s'.\n",
    "    - solver: Solver to be used. Default is 'gurobi'.\n",
    "    \n",
    "    Returns:\n",
    "    - pfba_obj: The optimized objective function value (growth rate or other objective).\n",
    "    - pfba_solution: The full pFBA solution (flux distribution).\n",
    "    \"\"\"\n",
    "    \n",
    "    GM = model.copy()  # Copy the model to avoid modifying the original\n",
    "    temp_dict = exp_dict\n",
    "    \n",
    "    # Default values if no argument is provided\n",
    "    if objective is None:\n",
    "        objective = 'biomass_cho_s'\n",
    "    if solver is None:\n",
    "        solver = 'gurobi'\n",
    "\n",
    "    # Set the solver and the objective reaction\n",
    "    GM.solver = solver\n",
    "    GM.objective = objective\n",
    "\n",
    "    # Constraint the model with experimental data\n",
    "    if objective is None:\n",
    "        GM.reactions.biomass_cho_s.bounds = temp_dict['exp_growth_rate'][time]\n",
    "    \n",
    "    for rxn in GM.boundary:\n",
    "        # Models that are forced to secrete ethanol are not feasible\n",
    "        if rxn.id == 'EX_etoh_e':\n",
    "            rxn.bounds = (-0.1,0.1)\n",
    "            continue\n",
    "\n",
    "        # Replace the lower and upper bound with experimental data\n",
    "        if rxn.id in temp_dict.keys():\n",
    "            rxn.bounds = temp_dict[rxn.id][time]\n",
    "            continue\n",
    "\n",
    "        # Keep boundaries open for essential metabolites\n",
    "        if rxn.id == 'EX_h2o_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_h_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_o2_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_hco3_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_so4_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "        if rxn.id == 'EX_pi_e':\n",
    "            rxn.bounds = (-1000,1000)\n",
    "            continue\n",
    "\n",
    "        # Boundaries from Sink reactions on iCHO_v1\n",
    "        if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Tyr_ggn_c':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_Ser_Thr_g':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "        if rxn.id == 'SK_pre_prot_r':\n",
    "            rxn.bounds = (-0.001,1000)\n",
    "            continue\n",
    "\n",
    "        # Close uptake rates for the rest of the boundaries\n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (-0.00001,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)\n",
    "    model_exp_constrained = GM \n",
    "    \n",
    "    if make_GM == True:\n",
    "        for rxn in GM.boundary:\n",
    "            if rxn.id in temp_dict.keys():\n",
    "                rxn.bounds = (-1000,1000)\n",
    "                print(rxn.bounds)\n",
    "                print(rxn)\n",
    "                continue\n",
    "        pfba_solution_GM = cobra.flux_analysis.pfba(GM)\n",
    "        pfba_obj_GM = pfba_solution_GM.fluxes[objective] \n",
    "        print(f\"Optimized objective w/ opened bounds (growth rate): {pfba_obj_GM}\")\n",
    "    else: \n",
    "        pfba_obj_GM =[]\n",
    "        GM = []\n",
    "            \n",
    "            \n",
    "    # Perform pFBA (parsimonious FBA) to minimize the total flux\n",
    "    pfba_solution = cobra.flux_analysis.pfba(model_exp_constrained)\n",
    "    pfba_obj = pfba_solution.fluxes[objective]\n",
    "\n",
    "    # Print and return the objective value and the full pFBA solution\n",
    "    print(f\"Optimized objective w/ exp constraints (growth rate): {pfba_obj}\")\n",
    "    \n",
    "    # Return the objective value and the full pFBA solution\n",
    "    return model_exp_constrained, pfba_obj, pfba_solution, pfba_obj_GM, GM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7e4d877",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def find_and_match_essential_genes(model, df_exp_essential, save_csv=True, print_df=False, model_GM = None):\n",
    "    \"\"\"\n",
    "    This function finds and matches essential genes in the model with experimental Entrez gene data.\n",
    "    It returns the number of total matched genes (len_mat) and the number of essential matched genes (len_mat_ess).\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if the necessary columns 'entrez_gene_id' and 'gene' exist in the dataframe\n",
    "    required_columns = ['entrez_gene_id', 'gene']\n",
    "    for col in required_columns:\n",
    "        if col not in df_exp_essential.columns:\n",
    "            raise KeyError(f\"Column '{col}' is missing from the dataframe.\")\n",
    "\n",
    "    # Create a reverse lookup dictionary where the key is the Entrez Gene ID and the value is the gene symbol\n",
    "    entrez_dict = df_exp_essential.set_index('entrez_gene_id')['gene'].to_dict()\n",
    "\n",
    "    # Matching metabolic genes\n",
    "    tmp_gene = []    \n",
    "    for gene in model.genes:\n",
    "        # Check if the gene.id matches any Entrez Gene ID in the reverse dictionary\n",
    "        if gene.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene.id]\n",
    "            tmp_gene.append((gene.id, gene.name, entrez_gene_symbol))\n",
    "    \n",
    "    # Number of total matched genes\n",
    "    len_mat = len(tmp_gene)\n",
    "\n",
    "    # Find the essential genes for the given model\n",
    "    essential_genes = find_essential_genes(model)\n",
    "    \n",
    "    # Create a list to store the matched essential genes\n",
    "    matched_genes = []\n",
    "\n",
    "    # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "    for gene in essential_genes:\n",
    "        gene_exp_ess = model.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "        if gene_exp_ess.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene_exp_ess.id]\n",
    "            matched_genes.append((gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol))\n",
    "\n",
    "    # Convert the matched essential genes into a DataFrame\n",
    "    matched_genes_essential_df = pd.DataFrame(matched_genes, columns=['gene_exp_ess.id', 'gene_exp_ess.name', 'entrez_gene_symbol'])\n",
    "\n",
    "    # Save the result to a CSV file if save_csv is True\n",
    "    if save_csv:\n",
    "        matched_genes_essential_df.to_csv(\"'../Data/Gene_Essentiality/matched_gene_entrez_ids_feg.csv\", index=False)\n",
    "\n",
    "    # Print the DataFrame if print_df is True\n",
    "    if print_df:\n",
    "        print(matched_genes_essential_df)\n",
    "\n",
    "    # Number of essential matched genes\n",
    "    len_mat_ess = len(matched_genes_essential_df)\n",
    "    \n",
    "    if model_GM is None:\n",
    "        # Print the summary information\n",
    "        print(f\"Total matched metabolic genes: {len_mat}\")\n",
    "        print(f\"Matched essential genes: {len_mat_ess}/{len_mat}\")\n",
    "        len_mat_GM_ess = []\n",
    "    else: \n",
    "        \n",
    "        essential_genes = find_essential_genes(model_GM)\n",
    "\n",
    "        # Create a list to store the matched essential genes\n",
    "        matched_genes_GM = []\n",
    "\n",
    "        # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "        for gene in essential_genes:\n",
    "            gene_exp_ess_GM = model_GM.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "            if gene_exp_ess_GM.id in entrez_dict:\n",
    "                entrez_gene_symbol = entrez_dict[gene_exp_ess_GM.id]\n",
    "                matched_genes_GM.append((gene_exp_ess_GM.id, gene_exp_ess_GM.name, entrez_gene_symbol))\n",
    "\n",
    "        # Convert the matched essential genes into a DataFrame\n",
    "        matched_genes_essential_GM_df = pd.DataFrame(matched_genes_GM, columns=['gene_exp_ess_GM.id', 'gene_exp_ess_GM.name', 'entrez_gene_symbol'])\n",
    "        len_mat_GM_ess = len(matched_genes_essential_GM_df)\n",
    "        # Print the summary information\n",
    "        print(f\"Total matched metabolic genes_GM: {len_mat}\")\n",
    "        print(f\"Matched essential genes_GM: {len_mat_GM_ess}/{len_mat}\")\n",
    "        \n",
    "\n",
    "\n",
    "    # Return the matched DataFrame and gene counts\n",
    "    return matched_genes_essential_df, len_mat, len_mat_ess, len_mat_GM_ess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8d67e573-e174-4e5c-b57e-8e227a81f6c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Pablo's modifification\n",
    "\n",
    "\n",
    "def find_and_match_essential_genes(model, df_exp_essential, save_csv=True, print_df=False, model_GM = None):\n",
    "    \"\"\"\n",
    "    This function finds and matches essential genes in the model with experimental Entrez gene data.\n",
    "    It returns the number of total matched genes (len_mat) and the number of essential matched genes (len_mat_ess).\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if the necessary columns 'entrez_gene_id' and 'gene' exist in the dataframe\n",
    "    required_columns = ['entrez_gene_id', 'gene']\n",
    "    for col in required_columns:\n",
    "        if col not in df_exp_essential.columns:\n",
    "            raise KeyError(f\"Column '{col}' is missing from the dataframe.\")\n",
    "\n",
    "    # Create a reverse lookup dictionary where the key is the Entrez Gene ID and the value is the gene symbol\n",
    "    entrez_dict = df_exp_essential.set_index('entrez_gene_id')['gene'].to_dict()\n",
    "\n",
    "    # Matching metabolic genes\n",
    "    tmp_gene = []    \n",
    "    for gene in model.genes:\n",
    "        # Check if the gene.id matches any Entrez Gene ID in the reverse dictionary\n",
    "        if gene.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene.id]\n",
    "            tmp_gene.append((gene.id, gene.name, entrez_gene_symbol))\n",
    "    \n",
    "    # Number of total matched genes\n",
    "    len_mat = len(tmp_gene)\n",
    "\n",
    "    # Find the essential reactions for the given model\n",
    "    threshold = model.slim_optimize(error_value=None) * 0.9\n",
    "    essential_reactions = find_essential_reactions(model, threshold=threshold)\n",
    "\n",
    "    # Remove tRNA biosynthesis reactions (specifically from iCHO2291\n",
    "    essential_reactions = [r for r in essential_reactions if not (r.subsystem == 'Biomass synthesis' and r.id.endswith('TRS'))]\n",
    "\n",
    "\n",
    "    # Extract genes from the essential reactions list\n",
    "    essential_genes = [g for r in essential_reactions for g in r.genes]\n",
    "        \n",
    "    # Create a list to store the matched essential genes\n",
    "    matched_genes = []\n",
    "\n",
    "    # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "    for gene in essential_genes:\n",
    "        gene_exp_ess = model.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "        if gene_exp_ess.id in entrez_dict:\n",
    "            entrez_gene_symbol = entrez_dict[gene_exp_ess.id]\n",
    "            matched_genes.append((gene_exp_ess.id, gene_exp_ess.name, entrez_gene_symbol))\n",
    "\n",
    "    # Convert the matched essential genes into a DataFrame\n",
    "    matched_genes_essential_df = pd.DataFrame(matched_genes, columns=['gene_exp_ess.id', 'gene_exp_ess.name', 'entrez_gene_symbol'])\n",
    "\n",
    "    # Save the result to a CSV file if save_csv is True\n",
    "    if save_csv:\n",
    "        matched_genes_essential_df.to_csv(\"'../Data/Gene_Essentiality/matched_gene_entrez_ids_feg.csv\", index=False)\n",
    "\n",
    "    # Print the DataFrame if print_df is True\n",
    "    if print_df:\n",
    "        print(matched_genes_essential_df)\n",
    "\n",
    "    # Number of essential matched genes\n",
    "    len_mat_ess = len(matched_genes_essential_df)\n",
    "    \n",
    "    if model_GM is None:\n",
    "        # Print the summary information\n",
    "        print(f\"Total matched metabolic genes: {len_mat}\")\n",
    "        print(f\"Matched essential genes: {len_mat_ess}/{len_mat}\")\n",
    "        len_mat_GM_ess = []\n",
    "    else: \n",
    "        \n",
    "        essential_genes = find_essential_genes(model_GM)\n",
    "\n",
    "        # Create a list to store the matched essential genes\n",
    "        matched_genes_GM = []\n",
    "\n",
    "        # Iterate over the essential genes and check if they match with the Entrez Gene IDs\n",
    "        for gene in essential_genes:\n",
    "            gene_exp_ess_GM = model_GM.genes.get_by_id(gene.id)  # Get the gene from the model\n",
    "\n",
    "            if gene_exp_ess_GM.id in entrez_dict:\n",
    "                entrez_gene_symbol = entrez_dict[gene_exp_ess_GM.id]\n",
    "                matched_genes_GM.append((gene_exp_ess_GM.id, gene_exp_ess_GM.name, entrez_gene_symbol))\n",
    "\n",
    "        # Convert the matched essential genes into a DataFrame\n",
    "        matched_genes_essential_GM_df = pd.DataFrame(matched_genes_GM, columns=['gene_exp_ess_GM.id', 'gene_exp_ess_GM.name', 'entrez_gene_symbol'])\n",
    "        len_mat_GM_ess = len(matched_genes_essential_GM_df)\n",
    "        # Print the summary information\n",
    "        print(f\"Total matched metabolic genes_GM: {len_mat}\")\n",
    "        print(f\"Matched essential genes_GM: {len_mat_GM_ess}/{len_mat}\")\n",
    "        \n",
    "\n",
    "\n",
    "    # Return the matched DataFrame and gene counts\n",
    "    return matched_genes_essential_df, len_mat, len_mat_ess, len_mat_GM_ess, essential_reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3f6c094b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_reaction_ids_specific(model):\n",
    "    # Iterate over each reaction in the model\n",
    "    for rxn in model.reactions:\n",
    "        # Check if the reaction ID contains \"(e)\"\n",
    "        if \"(e)\" in rxn.id:\n",
    "            # Replace \"(e)\" with \"_e\"\n",
    "            new_id = rxn.id.replace(\"(e)\", \"_e\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(c)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(c)\", \"_c\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(x)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(x)\", \"_x\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(r)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(r)\", \"_r\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"(g)\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"(g)\", \"_g\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_e_\" in rxn.id:\n",
    "            # Replace \"(e)\" with \"_e\"\n",
    "            new_id = rxn.id.replace(\"_e_\", \"_e\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_c_\" in rxn.id:\n",
    "            # Replace \"_c_\" with \"_c\"\n",
    "            new_id = rxn.id.replace(\"_c_\", \"_c\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_x_\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"_x_\", \"_x\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_r_\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"_r_\", \"_r\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "        \n",
    "        if \"_g_\" in rxn.id:\n",
    "            new_id = rxn.id.replace(\"_g_\", \"_g\")\n",
    "            print(f\"Updating reaction ID: {rxn.id} -> {new_id}\")\n",
    "            # Update the reaction ID with the new one\n",
    "            rxn.id = new_id\n",
    "            continue\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "11b2d09a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating reaction ID: EX_10fthf(e) -> EX_10fthf_e\n",
      "Updating reaction ID: EX_10fthf5glu(e) -> EX_10fthf5glu_e\n",
      "Updating reaction ID: EX_10fthf6glu(e) -> EX_10fthf6glu_e\n",
      "Updating reaction ID: EX_10fthf7glu(e) -> EX_10fthf7glu_e\n",
      "Updating reaction ID: EX_11_cis_retfa(e) -> EX_11_cis_retfa_e\n",
      "Updating reaction ID: EX_13_cis_retnglc(e) -> EX_13_cis_retnglc_e\n",
      "Updating reaction ID: EX_1glyc_cho(e) -> EX_1glyc_cho_e\n",
      "Updating reaction ID: EX_1mncam(e) -> EX_1mncam_e\n",
      "Updating reaction ID: EX_23cump(e) -> EX_23cump_e\n",
      "Updating reaction ID: EX_2425dhvitd2(e) -> EX_2425dhvitd2_e\n",
      "Updating reaction ID: EX_2425dhvitd3(e) -> EX_2425dhvitd3_e\n",
      "Updating reaction ID: EX_24nph(e) -> EX_24nph_e\n",
      "Updating reaction ID: EX_25hvitd2(e) -> EX_25hvitd2_e\n",
      "Updating reaction ID: EX_25hvitd3(e) -> EX_25hvitd3_e\n",
      "Updating reaction ID: EX_2hb(e) -> EX_2hb_e\n",
      "Updating reaction ID: EX_2mcit(e) -> EX_2mcit_e\n",
      "Updating reaction ID: EX_34dhoxpeg(e) -> EX_34dhoxpeg_e\n",
      "Updating reaction ID: EX_34dhphe(e) -> EX_34dhphe_e\n",
      "Updating reaction ID: EX_35cgmp(e) -> EX_35cgmp_e\n",
      "Updating reaction ID: EX_3aib_D(e) -> EX_3aib_D_e\n",
      "Updating reaction ID: EX_3aib(e) -> EX_3aib_e\n",
      "Updating reaction ID: EX_3mlda(e) -> EX_3mlda_e\n",
      "Updating reaction ID: EX_3mob(e) -> EX_3mob_e\n",
      "Updating reaction ID: EX_3mop(e) -> EX_3mop_e\n",
      "Updating reaction ID: EX_3ump(e) -> EX_3ump_e\n",
      "Updating reaction ID: EX_4abut(e) -> EX_4abut_e\n",
      "Updating reaction ID: EX_4abutn(e) -> EX_4abutn_e\n",
      "Updating reaction ID: EX_4hdebrisoquine(e) -> EX_4hdebrisoquine_e\n",
      "Updating reaction ID: EX_4hphac(e) -> EX_4hphac_e\n",
      "Updating reaction ID: EX_4hpro(e) -> EX_4hpro_e\n",
      "Updating reaction ID: EX_4mop(e) -> EX_4mop_e\n",
      "Updating reaction ID: EX_4mptnl(e) -> EX_4mptnl_e\n",
      "Updating reaction ID: EX_4mtolbutamide(e) -> EX_4mtolbutamide_e\n",
      "Updating reaction ID: EX_4nph(e) -> EX_4nph_e\n",
      "Updating reaction ID: EX_4nphsf(e) -> EX_4nphsf_e\n",
      "Updating reaction ID: EX_4pyrdx(e) -> EX_4pyrdx_e\n",
      "Updating reaction ID: EX_5adtststerone(e) -> EX_5adtststerone_e\n",
      "Updating reaction ID: EX_5adtststeroneglc(e) -> EX_5adtststeroneglc_e\n",
      "Updating reaction ID: EX_5adtststerones(e) -> EX_5adtststerones_e\n",
      "Updating reaction ID: EX_5dhf(e) -> EX_5dhf_e\n",
      "Updating reaction ID: EX_5fthf(e) -> EX_5fthf_e\n",
      "Updating reaction ID: EX_5homeprazole(e) -> EX_5homeprazole_e\n",
      "Updating reaction ID: EX_5htrp(e) -> EX_5htrp_e\n",
      "Updating reaction ID: EX_5mta(e) -> EX_5mta_e\n",
      "Updating reaction ID: EX_5mthf(e) -> EX_5mthf_e\n",
      "Updating reaction ID: EX_5oxpro(e) -> EX_5oxpro_e\n",
      "Updating reaction ID: EX_5thf(e) -> EX_5thf_e\n",
      "Updating reaction ID: EX_6dhf(e) -> EX_6dhf_e\n",
      "Updating reaction ID: EX_6htststerone(e) -> EX_6htststerone_e\n",
      "Updating reaction ID: EX_6thf(e) -> EX_6thf_e\n",
      "Updating reaction ID: EX_7dhf(e) -> EX_7dhf_e\n",
      "Updating reaction ID: EX_7thf(e) -> EX_7thf_e\n",
      "Updating reaction ID: EX_9_cis_retfa(e) -> EX_9_cis_retfa_e\n",
      "Updating reaction ID: EX_abt(e) -> EX_abt_e\n",
      "Updating reaction ID: EX_ac(e) -> EX_ac_e\n",
      "Updating reaction ID: EX_acac(e) -> EX_acac_e\n",
      "Updating reaction ID: EX_acald(e) -> EX_acald_e\n",
      "Updating reaction ID: EX_acetone(e) -> EX_acetone_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho(e) -> EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho(e) -> EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgam(e) -> EX_acgam_e\n",
      "Updating reaction ID: EX_ach(e) -> EX_ach_e\n",
      "Updating reaction ID: EX_acmana(e) -> EX_acmana_e\n",
      "Updating reaction ID: EX_acn13acngalgbside_cho(e) -> EX_acn13acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acn23acngalgbside_cho(e) -> EX_acn23acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acnacngal14acglcgalgluside_cho(e) -> EX_acnacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acnacngalgbside_cho(e) -> EX_acnacngalgbside_cho_e\n",
      "Updating reaction ID: EX_acngalacglcgal14acglcgalgluside_cho(e) -> EX_acngalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_ade(e) -> EX_ade_e\n",
      "Updating reaction ID: EX_adn(e) -> EX_adn_e\n",
      "Updating reaction ID: EX_adpcbl(e) -> EX_adpcbl_e\n",
      "Updating reaction ID: EX_adprbp(e) -> EX_adprbp_e\n",
      "Updating reaction ID: EX_adprib(e) -> EX_adprib_e\n",
      "Updating reaction ID: EX_adrn(e) -> EX_adrn_e\n",
      "Updating reaction ID: EX_adrnl(e) -> EX_adrnl_e\n",
      "Updating reaction ID: EX_aflatoxin(e) -> EX_aflatoxin_e\n",
      "Updating reaction ID: EX_ahandrostanglc(e) -> EX_ahandrostanglc_e\n",
      "Updating reaction ID: EX_ahcys(e) -> EX_ahcys_e\n",
      "Updating reaction ID: EX_ahdt(e) -> EX_ahdt_e\n",
      "Updating reaction ID: EX_aicar(e) -> EX_aicar_e\n",
      "Updating reaction ID: EX_ak2lgchol_cho(e) -> EX_ak2lgchol_cho_e\n",
      "Updating reaction ID: EX_akg(e) -> EX_akg_e\n",
      "Updating reaction ID: EX_ala_B(e) -> EX_ala_B_e\n",
      "Updating reaction ID: EX_ala_D(e) -> EX_ala_D_e\n",
      "Updating reaction ID: EX_ala_L(e) -> EX_ala_L_e\n",
      "Updating reaction ID: EX_alaala(e) -> EX_alaala_e\n",
      "Updating reaction ID: EX_aldstrn(e) -> EX_aldstrn_e\n",
      "Updating reaction ID: EX_amp(e) -> EX_amp_e\n",
      "Updating reaction ID: EX_andrstrn(e) -> EX_andrstrn_e\n",
      "Updating reaction ID: EX_andrstrnglc(e) -> EX_andrstrnglc_e\n",
      "Updating reaction ID: EX_anth(e) -> EX_anth_e\n",
      "Updating reaction ID: EX_antipyrene(e) -> EX_antipyrene_e\n",
      "Updating reaction ID: EX_apnnox(e) -> EX_apnnox_e\n",
      "Updating reaction ID: EX_appnn(e) -> EX_appnn_e\n",
      "Updating reaction ID: EX_aprgstrn(e) -> EX_aprgstrn_e\n",
      "Updating reaction ID: EX_aqcobal(e) -> EX_aqcobal_e\n",
      "Updating reaction ID: EX_arab_L(e) -> EX_arab_L_e\n",
      "Updating reaction ID: EX_arach(e) -> EX_arach_e\n",
      "Updating reaction ID: EX_arachd(e) -> EX_arachd_e\n",
      "Updating reaction ID: EX_arg_L(e) -> EX_arg_L_e\n",
      "Updating reaction ID: EX_ascb_L(e) -> EX_ascb_L_e\n",
      "Updating reaction ID: EX_asn_L(e) -> EX_asn_L_e\n",
      "Updating reaction ID: EX_asp_D(e) -> EX_asp_D_e\n",
      "Updating reaction ID: EX_asp_L(e) -> EX_asp_L_e\n",
      "Updating reaction ID: EX_atp(e) -> EX_atp_e\n",
      "Updating reaction ID: EX_avite1(e) -> EX_avite1_e\n",
      "Updating reaction ID: EX_bhb(e) -> EX_bhb_e\n",
      "Updating reaction ID: EX_bildglcur(e) -> EX_bildglcur_e\n",
      "Updating reaction ID: EX_bilglcur(e) -> EX_bilglcur_e\n",
      "Updating reaction ID: EX_bilirub(e) -> EX_bilirub_e\n",
      "Updating reaction ID: EX_biocyt(e) -> EX_biocyt_e\n",
      "Updating reaction ID: EX_btn(e) -> EX_btn_e\n",
      "Updating reaction ID: EX_but(e) -> EX_but_e\n",
      "Updating reaction ID: EX_bz(e) -> EX_bz_e\n",
      "Updating reaction ID: EX_xanta(e) -> EX_xanta_e\n",
      "Updating reaction ID: EX_cdchac(e) -> EX_cdchac_e\n",
      "Updating reaction ID: EX_ca2(e) -> EX_ca2_e\n",
      "Updating reaction ID: EX_camp(e) -> EX_camp_e\n",
      "Updating reaction ID: EX_carn(e) -> EX_carn_e\n",
      "Updating reaction ID: EX_caro(e) -> EX_caro_e\n",
      "Updating reaction ID: EX_carveol(e) -> EX_carveol_e\n",
      "Updating reaction ID: EX_cbasp(e) -> EX_cbasp_e\n",
      "Updating reaction ID: EX_cca_d3(e) -> EX_cca_d3_e\n",
      "Updating reaction ID: EX_cdp(e) -> EX_cdp_e\n",
      "Updating reaction ID: EX_cgly(e) -> EX_cgly_e\n",
      "Updating reaction ID: EX_chol(e) -> EX_chol_e\n",
      "Updating reaction ID: EX_cholate(e) -> EX_cholate_e\n",
      "Updating reaction ID: EX_chopg(e) -> EX_chopg_e\n",
      "Updating reaction ID: EX_chsterol(e) -> EX_chsterol_e\n",
      "Updating reaction ID: EX_chtn(e) -> EX_chtn_e\n",
      "Updating reaction ID: EX_cit(e) -> EX_cit_e\n",
      "Updating reaction ID: EX_citr_L(e) -> EX_citr_L_e\n",
      "Updating reaction ID: EX_cl(e) -> EX_cl_e\n",
      "Updating reaction ID: EX_clpnd(e) -> EX_clpnd_e\n",
      "Updating reaction ID: EX_cmp(e) -> EX_cmp_e\n",
      "Updating reaction ID: EX_co(e) -> EX_co_e\n",
      "Updating reaction ID: EX_co2(e) -> EX_co2_e\n",
      "Updating reaction ID: EX_coumarin(e) -> EX_coumarin_e\n",
      "Updating reaction ID: EX_creat(e) -> EX_creat_e\n",
      "Updating reaction ID: EX_crm_cho(e) -> EX_crm_cho_e\n",
      "Updating reaction ID: EX_crmp_cho(e) -> EX_crmp_cho_e\n",
      "Updating reaction ID: EX_crn(e) -> EX_crn_e\n",
      "Updating reaction ID: EX_crtsl(e) -> EX_crtsl_e\n",
      "Updating reaction ID: EX_crtstrn(e) -> EX_crtstrn_e\n",
      "Updating reaction ID: EX_crvnc(e) -> EX_crvnc_e\n",
      "Updating reaction ID: EX_csn(e) -> EX_csn_e\n",
      "Updating reaction ID: EX_cspg_a(e) -> EX_cspg_a_e\n",
      "Updating reaction ID: EX_cspg_b(e) -> EX_cspg_b_e\n",
      "Updating reaction ID: EX_cspg_c(e) -> EX_cspg_c_e\n",
      "Updating reaction ID: EX_cspg_d(e) -> EX_cspg_d_e\n",
      "Updating reaction ID: EX_cspg(e)e_ -> EX_cspg_ee_\n",
      "Updating reaction ID: EX_ctp(e) -> EX_ctp_e\n",
      "Updating reaction ID: EX_cyan(e) -> EX_cyan_e\n",
      "Updating reaction ID: EX_cys_L(e) -> EX_cys_L_e\n",
      "Updating reaction ID: EX_cytd(e) -> EX_cytd_e\n",
      "Updating reaction ID: EX_dad_2(e) -> EX_dad_2_e\n",
      "Updating reaction ID: EX_dag_cho(e) -> EX_dag_cho_e\n",
      "Updating reaction ID: EX_dca(e) -> EX_dca_e\n",
      "Updating reaction ID: EX_dcsptn1(e) -> EX_dcsptn1_e\n",
      "Updating reaction ID: EX_dcyt(e) -> EX_dcyt_e\n",
      "Updating reaction ID: EX_debrisoquine(e) -> EX_debrisoquine_e\n",
      "Updating reaction ID: EX_dgchol(e) -> EX_dgchol_e\n",
      "Updating reaction ID: EX_dgmp(e) -> EX_dgmp_e\n",
      "Updating reaction ID: EX_dgsn(e) -> EX_dgsn_e\n",
      "Updating reaction ID: EX_dgtp(e) -> EX_dgtp_e\n",
      "Updating reaction ID: EX_dha(e) -> EX_dha_e\n",
      "Updating reaction ID: EX_dhap(e) -> EX_dhap_e\n",
      "Updating reaction ID: EX_dhdascb(e) -> EX_dhdascb_e\n",
      "Updating reaction ID: EX_dheas(e) -> EX_dheas_e\n",
      "Updating reaction ID: EX_dhf(e) -> EX_dhf_e\n",
      "Updating reaction ID: EX_digalsgalside_cho(e) -> EX_digalsgalside_cho_e\n",
      "Updating reaction ID: EX_din(e) -> EX_din_e\n",
      "Updating reaction ID: EX_dlnlcg(e) -> EX_dlnlcg_e\n",
      "Updating reaction ID: EX_dmantipyrine(e) -> EX_dmantipyrine_e\n",
      "Updating reaction ID: EX_dmhptcrn(e) -> EX_dmhptcrn_e\n",
      "Updating reaction ID: EX_dopa(e) -> EX_dopa_e\n",
      "Updating reaction ID: EX_dopasf(e) -> EX_dopasf_e\n",
      "Updating reaction ID: EX_drib(e) -> EX_drib_e\n",
      "Updating reaction ID: EX_dtdp(e) -> EX_dtdp_e\n",
      "Updating reaction ID: EX_dtmp(e) -> EX_dtmp_e\n",
      "Updating reaction ID: EX_dttp(e) -> EX_dttp_e\n",
      "Updating reaction ID: EX_duri(e) -> EX_duri_e\n",
      "Updating reaction ID: EX_eaflatoxin(e) -> EX_eaflatoxin_e\n",
      "Updating reaction ID: EX_ebastine(e) -> EX_ebastine_e\n",
      "Updating reaction ID: EX_ebastineoh(e) -> EX_ebastineoh_e\n",
      "Updating reaction ID: EX_eicostet(e) -> EX_eicostet_e\n",
      "Updating reaction ID: EX_elaid(e) -> EX_elaid_e\n",
      "Updating reaction ID: EX_estradiol(e) -> EX_estradiol_e\n",
      "Updating reaction ID: EX_estradiolglc(e) -> EX_estradiolglc_e\n",
      "Updating reaction ID: EX_estriolglc(e) -> EX_estriolglc_e\n",
      "Updating reaction ID: EX_estroneglc(e) -> EX_estroneglc_e\n",
      "Updating reaction ID: EX_estrones(e) -> EX_estrones_e\n",
      "Updating reaction ID: EX_etoh(e) -> EX_etoh_e\n",
      "Updating reaction ID: EX_fad(e) -> EX_fad_e\n",
      "Updating reaction ID: EX_fald(e) -> EX_fald_e\n",
      "Updating reaction ID: EX_fe2(e) -> EX_fe2_e\n",
      "Updating reaction ID: EX_fe3(e) -> EX_fe3_e\n",
      "Updating reaction ID: EX_fmn(e) -> EX_fmn_e\n",
      "Updating reaction ID: EX_fol(e) -> EX_fol_e\n",
      "Updating reaction ID: EX_for(e) -> EX_for_e\n",
      "Updating reaction ID: EX_fru(e) -> EX_fru_e\n",
      "Updating reaction ID: EX_fuc_L(e) -> EX_fuc_L_e\n",
      "Updating reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho(e) -> EX_fuc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fuc14galacglcgalgluside_cho(e) -> EX_fuc14galacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacgalfucgalacglcgalgluside_cho(e) -> EX_fucacgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngal14acglcgalgluside_cho(e) -> EX_fucacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngalacglcgalgluside_cho(e) -> EX_fucacngalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc12gal14acglcgalgluside_cho(e) -> EX_fucfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho(e) -> EX_fucfuc132galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho(e) -> EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho(e) -> EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucgalacglcgalgluside_cho(e) -> EX_fucfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgal14acglcgalgluside_cho(e) -> EX_fucgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalfucgalacglcgalgluside_cho(e) -> EX_fucgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalgbside_cho(e) -> EX_fucgalgbside_cho_e\n",
      "Updating reaction ID: EX_g1p(e) -> EX_g1p_e\n",
      "Updating reaction ID: EX_gal(e) -> EX_gal_e\n",
      "Updating reaction ID: EX_galacglcgalgbside_cho(e) -> EX_galacglcgalgbside_cho_e\n",
      "Updating reaction ID: EX_galfuc12gal14acglcgalgluside_cho(e) -> EX_galfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho(e) -> EX_galfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho(e) -> EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalgalthcrm_cho(e) -> EX_galgalgalthcrm_cho_e\n",
      "Updating reaction ID: EX_gam(e) -> EX_gam_e\n",
      "Updating reaction ID: EX_gbside_cho(e) -> EX_gbside_cho_e\n",
      "Updating reaction ID: EX_gchola(e) -> EX_gchola_e\n",
      "Updating reaction ID: EX_gd1b2_cho(e) -> EX_gd1b2_cho_e\n",
      "Updating reaction ID: EX_gd1c_cho(e) -> EX_gd1c_cho_e\n",
      "Updating reaction ID: EX_gdp(e) -> EX_gdp_e\n",
      "Updating reaction ID: EX_glc(e) -> EX_glc_e\n",
      "Updating reaction ID: EX_glcur(e) -> EX_glcur_e\n",
      "Updating reaction ID: EX_gln_L(e) -> EX_gln_L_e\n",
      "Updating reaction ID: EX_glu_L(e) -> EX_glu_L_e\n",
      "Updating reaction ID: EX_gluala(e) -> EX_gluala_e\n",
      "Updating reaction ID: EX_gly(e) -> EX_gly_e\n",
      "Updating reaction ID: EX_glyb(e) -> EX_glyb_e\n",
      "Updating reaction ID: EX_glyc(e) -> EX_glyc_e\n",
      "Updating reaction ID: EX_glyc_S(e) -> EX_glyc_S_e\n",
      "Updating reaction ID: EX_glygly(e) -> EX_glygly_e\n",
      "Updating reaction ID: EX_glygn2(e) -> EX_glygn2_e\n",
      "Updating reaction ID: EX_glygn4(e) -> EX_glygn4_e\n",
      "Updating reaction ID: EX_glygn5(e) -> EX_glygn5_e\n",
      "Updating reaction ID: EX_glyleu(e) -> EX_glyleu_e\n",
      "Updating reaction ID: EX_glyphe(e) -> EX_glyphe_e\n",
      "Updating reaction ID: EX_glypro(e) -> EX_glypro_e\n",
      "Updating reaction ID: EX_glysar(e) -> EX_glysar_e\n",
      "Updating reaction ID: EX_gmp(e) -> EX_gmp_e\n",
      "Updating reaction ID: EX_gp1c_cho(e) -> EX_gp1c_cho_e\n",
      "Updating reaction ID: EX_gp1calpha_cho(e) -> EX_gp1calpha_cho_e\n",
      "Updating reaction ID: EX_gq1b_cho(e) -> EX_gq1b_cho_e\n",
      "Updating reaction ID: EX_gq1balpha_cho(e) -> EX_gq1balpha_cho_e\n",
      "Updating reaction ID: EX_gsn(e) -> EX_gsn_e\n",
      "Updating reaction ID: EX_gt1a_cho(e) -> EX_gt1a_cho_e\n",
      "Updating reaction ID: EX_gthox(e) -> EX_gthox_e\n",
      "Updating reaction ID: EX_gthrd(e) -> EX_gthrd_e\n",
      "Updating reaction ID: EX_gtp(e) -> EX_gtp_e\n",
      "Updating reaction ID: EX_gua(e) -> EX_gua_e\n",
      "Updating reaction ID: EX_h(e) -> EX_h_e\n",
      "Updating reaction ID: EX_h2o(e) -> EX_h2o_e\n",
      "Updating reaction ID: EX_h2o2(e) -> EX_h2o2_e\n",
      "Updating reaction ID: EX_ha(e) -> EX_ha_e\n",
      "Updating reaction ID: EX_ha_pre1(e) -> EX_ha_pre1_e\n",
      "Updating reaction ID: EX_ApoA1(e) -> EX_ApoA1_e\n",
      "Updating reaction ID: EX_isomal(e) -> EX_isomal_e\n",
      "Updating reaction ID: EX_h2s(e) -> EX_h2s_e\n",
      "Updating reaction ID: EX_chtbs(e) -> EX_chtbs_e\n",
      "Updating reaction ID: EX_cyala(e) -> EX_cyala_e\n",
      "Updating reaction ID: EX_npphos(e) -> EX_npphos_e\n",
      "Updating reaction ID: EX_dhnpt(e) -> EX_dhnpt_e\n",
      "Updating reaction ID: EX_3kgal(e) -> EX_3kgal_e\n",
      "Updating reaction ID: EX_lcts6p(e) -> EX_lcts6p_e\n",
      "Updating reaction ID: EX_ggl(e) -> EX_ggl_e\n",
      "Updating reaction ID: EX_3klcts(e) -> EX_3klcts_e\n",
      "Updating reaction ID: EX_glucyala(e) -> EX_glucyala_e\n",
      "Updating reaction ID: EX_uppg1(e) -> EX_uppg1_e\n",
      "Updating reaction ID: EX_cpppg1(e) -> EX_cpppg1_e\n",
      "Updating reaction ID: EX_gluampn(e) -> EX_gluampn_e\n",
      "Updating reaction ID: EX_lpdmd(e) -> EX_lpdmd_e\n",
      "Updating reaction ID: EX_gm4p(e) -> EX_gm4p_e\n",
      "Updating reaction ID: EX_zn2(e) -> EX_zn2_e\n",
      "Updating reaction ID: EX_ddca(e) -> EX_ddca_e\n",
      "Updating reaction ID: EX_wharachd(e) -> EX_wharachd_e\n",
      "Updating reaction ID: EX_txb2(e) -> EX_txb2_e\n",
      "Updating reaction ID: EX_335trithyr(e) -> EX_335trithyr_e\n",
      "Updating reaction ID: EX_gthleuc4(e) -> EX_gthleuc4_e\n",
      "Updating reaction ID: EX_gth24dnb(e) -> EX_gth24dnb_e\n",
      "Updating reaction ID: EX_gthetcrya(e) -> EX_gthetcrya_e\n",
      "Updating reaction ID: EX_prostga1(e) -> EX_prostga1_e\n",
      "Updating reaction ID: EX_prostga2(e) -> EX_prostga2_e\n",
      "Updating reaction ID: EX_prostgb1(e) -> EX_prostgb1_e\n",
      "Updating reaction ID: EX_prostgb2(e) -> EX_prostgb2_e\n",
      "Updating reaction ID: EX_prostgc1(e) -> EX_prostgc1_e\n",
      "Updating reaction ID: EX_prostgc2(e) -> EX_prostgc2_e\n",
      "Updating reaction ID: EX_prostge3(e) -> EX_prostge3_e\n",
      "Updating reaction ID: EX_hco3(e) -> EX_hco3_e\n",
      "Updating reaction ID: EX_hcoumarin(e) -> EX_hcoumarin_e\n",
      "Updating reaction ID: EX_hdca(e) -> EX_hdca_e\n",
      "Updating reaction ID: EX_hdcea(e) -> EX_hdcea_e\n",
      "Updating reaction ID: EX_hestratriol(e) -> EX_hestratriol_e\n",
      "Updating reaction ID: EX_hexc(e) -> EX_hexc_e\n",
      "Updating reaction ID: EX_his_L(e) -> EX_his_L_e\n",
      "Updating reaction ID: EX_hista(e) -> EX_hista_e\n",
      "Updating reaction ID: EX_hom_L(e) -> EX_hom_L_e\n",
      "Updating reaction ID: EX_hpdca(e) -> EX_hpdca_e\n",
      "Updating reaction ID: EX_htaxol(e) -> EX_htaxol_e\n",
      "Updating reaction ID: EX_hxan(e) -> EX_hxan_e\n",
      "Updating reaction ID: EX_i(e) -> EX_i_e\n",
      "Updating reaction ID: EX_idour(e) -> EX_idour_e\n",
      "Updating reaction ID: EX_idp(e) -> EX_idp_e\n",
      "Updating reaction ID: EX_ile_L(e) -> EX_ile_L_e\n",
      "Updating reaction ID: EX_imp(e) -> EX_imp_e\n",
      "Updating reaction ID: EX_inost(e) -> EX_inost_e\n",
      "Updating reaction ID: EX_ins(e) -> EX_ins_e\n",
      "Updating reaction ID: EX_k(e) -> EX_k_e\n",
      "Updating reaction ID: EX_ksi_deg1(e) -> EX_ksi_deg1_e\n",
      "Updating reaction ID: EX_ksi(e) -> EX_ksi_e\n",
      "Updating reaction ID: EX_ksii_core2(e) -> EX_ksii_core2_e\n",
      "Updating reaction ID: EX_ksii_core4(e) -> EX_ksii_core4_e\n",
      "Updating reaction ID: EX_lac_D(e) -> EX_lac_D_e\n",
      "Updating reaction ID: EX_lac_L(e) -> EX_lac_L_e\n",
      "Updating reaction ID: EX_lcts(e) -> EX_lcts_e\n",
      "Updating reaction ID: EX_Lcystin(e) -> EX_Lcystin_e\n",
      "Updating reaction ID: EX_leu_L(e) -> EX_leu_L_e\n",
      "Updating reaction ID: EX_leugly(e) -> EX_leugly_e\n",
      "Updating reaction ID: EX_leuktrA4(e) -> EX_leuktrA4_e\n",
      "Updating reaction ID: EX_leuktrB4(e) -> EX_leuktrB4_e\n",
      "Updating reaction ID: EX_leuktrC4(e) -> EX_leuktrC4_e\n",
      "Updating reaction ID: EX_leuktrD4(e) -> EX_leuktrD4_e\n",
      "Updating reaction ID: EX_leuktrE4(e) -> EX_leuktrE4_e\n",
      "Updating reaction ID: EX_leuktrF4(e) -> EX_leuktrF4_e\n",
      "Updating reaction ID: EX_leuleu(e) -> EX_leuleu_e\n",
      "Updating reaction ID: EX_lgnc(e) -> EX_lgnc_e\n",
      "Updating reaction ID: EX_limnen(e) -> EX_limnen_e\n",
      "Updating reaction ID: EX_lipoate(e) -> EX_lipoate_e\n",
      "Updating reaction ID: EX_lneldc(e) -> EX_lneldc_e\n",
      "Updating reaction ID: EX_lnlc(e) -> EX_lnlc_e\n",
      "Updating reaction ID: EX_lnlnca(e) -> EX_lnlnca_e\n",
      "Updating reaction ID: EX_lnlncg(e) -> EX_lnlncg_e\n",
      "Updating reaction ID: EX_lpchol_cho(e) -> EX_lpchol_cho_e\n",
      "Updating reaction ID: EX_lys_L(e) -> EX_lys_L_e\n",
      "Updating reaction ID: EX_mag_cho(e) -> EX_mag_cho_e\n",
      "Updating reaction ID: EX_mal_L(e) -> EX_mal_L_e\n",
      "Updating reaction ID: EX_malt(e) -> EX_malt_e\n",
      "Updating reaction ID: EX_malttr(e) -> EX_malttr_e\n",
      "Updating reaction ID: EX_man(e) -> EX_man_e\n",
      "Updating reaction ID: EX_meoh(e) -> EX_meoh_e\n",
      "Updating reaction ID: EX_mepi(e) -> EX_mepi_e\n",
      "Updating reaction ID: EX_mercplaccys(e) -> EX_mercplaccys_e\n",
      "Updating reaction ID: EX_met_L(e) -> EX_met_L_e\n",
      "Updating reaction ID: EX_mthgxl(e) -> EX_mthgxl_e\n",
      "Updating reaction ID: EX_n2m2nmasn(e) -> EX_n2m2nmasn_e\n",
      "Updating reaction ID: EX_na1(e) -> EX_na1_e\n",
      "Updating reaction ID: EX_nac(e) -> EX_nac_e\n",
      "Updating reaction ID: EX_nad(e) -> EX_nad_e\n",
      "Updating reaction ID: EX_nadp(e) -> EX_nadp_e\n",
      "Updating reaction ID: EX_ncam(e) -> EX_ncam_e\n",
      "Updating reaction ID: EX_nh4(e) -> EX_nh4_e\n",
      "Updating reaction ID: EX_nifedipine(e) -> EX_nifedipine_e\n",
      "Updating reaction ID: EX_no(e) -> EX_no_e\n",
      "Updating reaction ID: EX_npthl(e) -> EX_npthl_e\n",
      "Updating reaction ID: EX_nrpphr(e) -> EX_nrpphr_e\n",
      "Updating reaction ID: EX_nrpphrsf(e) -> EX_nrpphrsf_e\n",
      "Updating reaction ID: EX_nrvnc(e) -> EX_nrvnc_e\n",
      "Updating reaction ID: EX_o2(e) -> EX_o2_e\n",
      "Updating reaction ID: EX_o2s(e) -> EX_o2s_e\n",
      "Updating reaction ID: EX_oagd3_cho(e) -> EX_oagd3_cho_e\n",
      "Updating reaction ID: EX_oagt3_cho(e) -> EX_oagt3_cho_e\n",
      "Updating reaction ID: EX_ocdca(e) -> EX_ocdca_e\n",
      "Updating reaction ID: EX_ocdcea(e) -> EX_ocdcea_e\n",
      "Updating reaction ID: EX_octa(e) -> EX_octa_e\n",
      "Updating reaction ID: EX_oh1(e) -> EX_oh1_e\n",
      "Updating reaction ID: EX_omeprazole(e) -> EX_omeprazole_e\n",
      "Updating reaction ID: EX_onpthl(e) -> EX_onpthl_e\n",
      "Updating reaction ID: EX_orn(e) -> EX_orn_e\n",
      "Updating reaction ID: EX_orot(e) -> EX_orot_e\n",
      "Updating reaction ID: EX_oxa(e) -> EX_oxa_e\n",
      "Updating reaction ID: EX_paf_cho(e) -> EX_paf_cho_e\n",
      "Updating reaction ID: EX_pchol_cho(e) -> EX_pchol_cho_e\n",
      "Updating reaction ID: EX_pe_cho(e) -> EX_pe_cho_e\n",
      "Updating reaction ID: EX_peplys(e) -> EX_peplys_e\n",
      "Updating reaction ID: EX_perillyl(e) -> EX_perillyl_e\n",
      "Updating reaction ID: EX_pglyc_cho(e) -> EX_pglyc_cho_e\n",
      "Updating reaction ID: EX_phe_L(e) -> EX_phe_L_e\n",
      "Updating reaction ID: EX_pheacgln(e) -> EX_pheacgln_e\n",
      "Updating reaction ID: EX_pheme(e) -> EX_pheme_e\n",
      "Updating reaction ID: EX_phyQ(e) -> EX_phyQ_e\n",
      "Updating reaction ID: EX_phyt(e) -> EX_phyt_e\n",
      "Updating reaction ID: EX_pi(e) -> EX_pi_e\n",
      "Updating reaction ID: EX_pnto_R(e) -> EX_pnto_R_e\n",
      "Updating reaction ID: EX_ppa(e) -> EX_ppa_e\n",
      "Updating reaction ID: EX_ppi(e) -> EX_ppi_e\n",
      "Updating reaction ID: EX_prgstrn(e) -> EX_prgstrn_e\n",
      "Updating reaction ID: EX_pro_D(e) -> EX_pro_D_e\n",
      "Updating reaction ID: EX_pro_L(e) -> EX_pro_L_e\n",
      "Updating reaction ID: EX_progly(e) -> EX_progly_e\n",
      "Updating reaction ID: EX_prostgd2(e) -> EX_prostgd2_e\n",
      "Updating reaction ID: EX_prostge1(e) -> EX_prostge1_e\n",
      "Updating reaction ID: EX_prostge2(e) -> EX_prostge2_e\n",
      "Updating reaction ID: EX_prostgf2(e) -> EX_prostgf2_e\n",
      "Updating reaction ID: EX_prostgh2(e) -> EX_prostgh2_e\n",
      "Updating reaction ID: EX_prostgi2(e) -> EX_prostgi2_e\n",
      "Updating reaction ID: EX_prpp(e) -> EX_prpp_e\n",
      "Updating reaction ID: EX_ps_cho(e) -> EX_ps_cho_e\n",
      "Updating reaction ID: EX_ptdca(e) -> EX_ptdca_e\n",
      "Updating reaction ID: EX_ptrc(e) -> EX_ptrc_e\n",
      "Updating reaction ID: EX_pydam(e) -> EX_pydam_e\n",
      "Updating reaction ID: EX_pydx(e) -> EX_pydx_e\n",
      "Updating reaction ID: EX_pydx5p(e) -> EX_pydx5p_e\n",
      "Updating reaction ID: EX_pydxn(e) -> EX_pydxn_e\n",
      "Updating reaction ID: EX_pyr(e) -> EX_pyr_e\n",
      "Updating reaction ID: EX_q10(e) -> EX_q10_e\n",
      "Updating reaction ID: EX_q10h2(e) -> EX_q10h2_e\n",
      "Updating reaction ID: EX_rbt(e) -> EX_rbt_e\n",
      "Updating reaction ID: EX_retfa(e) -> EX_retfa_e\n",
      "Updating reaction ID: EX_retinol(e) -> EX_retinol_e\n",
      "Updating reaction ID: EX_retn(e) -> EX_retn_e\n",
      "Updating reaction ID: EX_retnglc(e) -> EX_retnglc_e\n",
      "Updating reaction ID: EX_rib_D(e) -> EX_rib_D_e\n",
      "Updating reaction ID: EX_ribflv(e) -> EX_ribflv_e\n",
      "Updating reaction ID: EX_s2l2fn2m2masn(e) -> EX_s2l2fn2m2masn_e\n",
      "Updating reaction ID: EX_s2l2n2m2masn(e) -> EX_s2l2n2m2masn_e\n",
      "Updating reaction ID: EX_sarcs(e) -> EX_sarcs_e\n",
      "Updating reaction ID: EX_sbt_d(e) -> EX_sbt_d_e\n",
      "Updating reaction ID: EX_sel(e) -> EX_sel_e\n",
      "Updating reaction ID: EX_ser_D(e) -> EX_ser_D_e\n",
      "Updating reaction ID: EX_ser_L(e) -> EX_ser_L_e\n",
      "Updating reaction ID: EX_sfcys(e) -> EX_sfcys_e\n",
      "Updating reaction ID: EX_sl_L(e) -> EX_sl_L_e\n",
      "Updating reaction ID: EX_so4(e) -> EX_so4_e\n",
      "Updating reaction ID: EX_spc_cho(e) -> EX_spc_cho_e\n",
      "Updating reaction ID: EX_sph1p(e) -> EX_sph1p_e\n",
      "Updating reaction ID: EX_sphs1p(e) -> EX_sphs1p_e\n",
      "Updating reaction ID: EX_spmd(e) -> EX_spmd_e\n",
      "Updating reaction ID: EX_sprm(e) -> EX_sprm_e\n",
      "Updating reaction ID: EX_srtn(e) -> EX_srtn_e\n",
      "Updating reaction ID: EX_strch1(e) -> EX_strch1_e\n",
      "Updating reaction ID: EX_strch2(e) -> EX_strch2_e\n",
      "Updating reaction ID: EX_strdnc(e) -> EX_strdnc_e\n",
      "Updating reaction ID: EX_succ(e) -> EX_succ_e\n",
      "Updating reaction ID: EX_sucr(e) -> EX_sucr_e\n",
      "Updating reaction ID: EX_tag_cho(e) -> EX_tag_cho_e\n",
      "Updating reaction ID: EX_tagat_D(e) -> EX_tagat_D_e\n",
      "Updating reaction ID: EX_taur(e) -> EX_taur_e\n",
      "Updating reaction ID: EX_taxol(e) -> EX_taxol_e\n",
      "Updating reaction ID: EX_tchola(e) -> EX_tchola_e\n",
      "Updating reaction ID: EX_tcynt(e) -> EX_tcynt_e\n",
      "Updating reaction ID: EX_tdchola(e) -> EX_tdchola_e\n",
      "Updating reaction ID: EX_tethex3(e) -> EX_tethex3_e\n",
      "Updating reaction ID: EX_tetpent3(e) -> EX_tetpent3_e\n",
      "Updating reaction ID: EX_tetpent6(e) -> EX_tetpent6_e\n",
      "Updating reaction ID: EX_tettet6(e) -> EX_tettet6_e\n",
      "Updating reaction ID: EX_thf(e) -> EX_thf_e\n",
      "Updating reaction ID: EX_thm(e) -> EX_thm_e\n",
      "Updating reaction ID: EX_thmmp(e) -> EX_thmmp_e\n",
      "Updating reaction ID: EX_thmtp(e) -> EX_thmtp_e\n",
      "Updating reaction ID: EX_thr_L(e) -> EX_thr_L_e\n",
      "Updating reaction ID: EX_thym(e) -> EX_thym_e\n",
      "Updating reaction ID: EX_thymd(e) -> EX_thymd_e\n",
      "Updating reaction ID: EX_thyox_L(e) -> EX_thyox_L_e\n",
      "Updating reaction ID: EX_tmndnc(e) -> EX_tmndnc_e\n",
      "Updating reaction ID: EX_tolbutamide(e) -> EX_tolbutamide_e\n",
      "Updating reaction ID: EX_tre(e) -> EX_tre_e\n",
      "Updating reaction ID: EX_triodthy(e) -> EX_triodthy_e\n",
      "Updating reaction ID: EX_triodthysuf(e) -> EX_triodthysuf_e\n",
      "Updating reaction ID: EX_trp_L(e) -> EX_trp_L_e\n",
      "Updating reaction ID: EX_tststerone(e) -> EX_tststerone_e\n",
      "Updating reaction ID: EX_tststeroneglc(e) -> EX_tststeroneglc_e\n",
      "Updating reaction ID: EX_tststerones(e) -> EX_tststerones_e\n",
      "Updating reaction ID: EX_tsul(e) -> EX_tsul_e\n",
      "Updating reaction ID: EX_ttdca(e) -> EX_ttdca_e\n",
      "Updating reaction ID: EX_txa2(e) -> EX_txa2_e\n",
      "Updating reaction ID: EX_tymsf(e) -> EX_tymsf_e\n",
      "Updating reaction ID: EX_Tyr_ggn(e) -> EX_Tyr_ggn_e\n",
      "Updating reaction ID: EX_tyr_L(e) -> EX_tyr_L_e\n",
      "Updating reaction ID: EX_udp(e) -> EX_udp_e\n",
      "Updating reaction ID: EX_udpg(e) -> EX_udpg_e\n",
      "Updating reaction ID: EX_udpgal(e) -> EX_udpgal_e\n",
      "Updating reaction ID: EX_ump(e) -> EX_ump_e\n",
      "Updating reaction ID: EX_ura(e) -> EX_ura_e\n",
      "Updating reaction ID: EX_urate(e) -> EX_urate_e\n",
      "Updating reaction ID: EX_urea(e) -> EX_urea_e\n",
      "Updating reaction ID: EX_uri(e) -> EX_uri_e\n",
      "Updating reaction ID: EX_utp(e) -> EX_utp_e\n",
      "Updating reaction ID: EX_vacc(e) -> EX_vacc_e\n",
      "Updating reaction ID: EX_val_L(e) -> EX_val_L_e\n",
      "Updating reaction ID: EX_vitd2(e) -> EX_vitd2_e\n",
      "Updating reaction ID: EX_vitd3(e) -> EX_vitd3_e\n",
      "Updating reaction ID: EX_whddca(e) -> EX_whddca_e\n",
      "Updating reaction ID: EX_whhdca(e) -> EX_whhdca_e\n",
      "Updating reaction ID: EX_whtststerone(e) -> EX_whtststerone_e\n",
      "Updating reaction ID: EX_whttdca(e) -> EX_whttdca_e\n",
      "Updating reaction ID: EX_xolest2_cho(e) -> EX_xolest2_cho_e\n",
      "Updating reaction ID: EX_xoltri24(e) -> EX_xoltri24_e\n",
      "Updating reaction ID: EX_xoltri25(e) -> EX_xoltri25_e\n",
      "Updating reaction ID: EX_xoltri27(e) -> EX_xoltri27_e\n",
      "Updating reaction ID: EX_xyl_D(e) -> EX_xyl_D_e\n",
      "Updating reaction ID: EX_xylt(e) -> EX_xylt_e\n",
      "Updating reaction ID: EX_glcn(e) -> EX_glcn_e\n",
      "Updating reaction ID: EX_retinal(e) -> EX_retinal_e\n",
      "Updating reaction ID: EX_crtn(e) -> EX_crtn_e\n",
      "Updating reaction ID: EX_trypta(e) -> EX_trypta_e\n",
      "Updating reaction ID: EX_tym(e) -> EX_tym_e\n",
      "Updating reaction ID: EX_doco13ac(e) -> EX_doco13ac_e\n",
      "Updating reaction ID: EX_dhcholestanate(e) -> EX_dhcholestanate_e\n",
      "Updating reaction ID: EX_2hyoxplac(e) -> EX_2hyoxplac_e\n",
      "Updating reaction ID: EX_pheacgly(e) -> EX_pheacgly_e\n",
      "Updating reaction ID: EX_acglu(e) -> EX_acglu_e\n",
      "Updating reaction ID: EX_acorn(e) -> EX_acorn_e\n",
      "Updating reaction ID: EX_coke(e) -> EX_coke_e\n",
      "Updating reaction ID: EX_galt(e) -> EX_galt_e\n",
      "Updating reaction ID: EX_Lkynr(e) -> EX_Lkynr_e\n",
      "Updating reaction ID: EX_hLkynr(e) -> EX_hLkynr_e\n",
      "Updating reaction ID: EX_kynate(e) -> EX_kynate_e\n",
      "Updating reaction ID: EX_1a2425thvitd2(e) -> EX_1a2425thvitd2_e\n",
      "Updating reaction ID: EX_34dhoxmand(e) -> EX_34dhoxmand_e\n",
      "Updating reaction ID: EX_3mox4hoxm(e) -> EX_3mox4hoxm_e\n",
      "Updating reaction ID: EX_gncore1(e) -> EX_gncore1_e\n",
      "Updating reaction ID: EX_ind3ac(e) -> EX_ind3ac_e\n",
      "Updating reaction ID: EX_12HPET(e) -> EX_12HPET_e\n",
      "Updating reaction ID: EX_15HPET(e) -> EX_15HPET_e\n",
      "Updating reaction ID: EX_arg_D(e) -> EX_arg_D_e\n",
      "Updating reaction ID: DM_oak2g_hs(c) -> DM_oak2g_hs_c\n",
      "Updating reaction ID: EX_mg2(e) -> EX_mg2_e\n",
      "Updating reaction ID: EX_cu2(e) -> EX_cu2_e\n",
      "Updating reaction ID: EX_eicostri(e) -> EX_eicostri_e\n",
      "Updating reaction ID: EX_docastri(e) -> EX_docastri_e\n",
      "Updating reaction ID: EX_Rtotal(e) -> EX_Rtotal_e\n",
      "Updating reaction ID: EX_Rtotal2(e) -> EX_Rtotal2_e\n",
      "Updating reaction ID: EX_Rtotal3(e) -> EX_Rtotal3_e\n",
      "Updating reaction ID: EX_HC01610_e_ -> EX_HC01610_e\n",
      "Updating reaction ID: EX_HC02179_e_ -> EX_HC02179_e\n",
      "Updating reaction ID: EX_C05965_e_ -> EX_C05965_e\n",
      "Updating reaction ID: EX_HC02176_e_ -> EX_HC02176_e\n",
      "Updating reaction ID: EX_10fthf_e_ -> EX_10fthf_e\n",
      "Updating reaction ID: EX_10fthf5glu_e_ -> EX_10fthf5glu_e\n",
      "Updating reaction ID: EX_10fthf6glu_e_ -> EX_10fthf6glu_e\n",
      "Updating reaction ID: EX_10fthf7glu_e_ -> EX_10fthf7glu_e\n",
      "Updating reaction ID: EX_11_cis_retfa_e_ -> EX_11_cis_retfa_e\n",
      "Updating reaction ID: EX_13_cis_retnglc_e_ -> EX_13_cis_retnglc_e\n",
      "Updating reaction ID: EX_1glyc_cho_e_ -> EX_1glyc_cho_e\n",
      "Updating reaction ID: EX_1mncam_e_ -> EX_1mncam_e\n",
      "Updating reaction ID: EX_23cump_e_ -> EX_23cump_e\n",
      "Updating reaction ID: EX_2425dhvitd2_e_ -> EX_2425dhvitd2_e\n",
      "Updating reaction ID: EX_2425dhvitd3_e_ -> EX_2425dhvitd3_e\n",
      "Updating reaction ID: EX_24nph_e_ -> EX_24nph_e\n",
      "Updating reaction ID: EX_25hvitd2_e_ -> EX_25hvitd2_e\n",
      "Updating reaction ID: EX_25hvitd3_e_ -> EX_25hvitd3_e\n",
      "Updating reaction ID: EX_2hb_e_ -> EX_2hb_e\n",
      "Updating reaction ID: EX_2mcit_e_ -> EX_2mcit_e\n",
      "Updating reaction ID: EX_34dhoxpeg_e_ -> EX_34dhoxpeg_e\n",
      "Updating reaction ID: EX_34dhphe_e_ -> EX_34dhphe_e\n",
      "Updating reaction ID: EX_35cgmp_e_ -> EX_35cgmp_e\n",
      "Updating reaction ID: EX_3aib_D_e_ -> EX_3aib_D_e\n",
      "Updating reaction ID: EX_3aib_e_ -> EX_3aib_e\n",
      "Updating reaction ID: EX_3mlda_e_ -> EX_3mlda_e\n",
      "Updating reaction ID: EX_3mob_e_ -> EX_3mob_e\n",
      "Updating reaction ID: EX_3mop_e_ -> EX_3mop_e\n",
      "Updating reaction ID: EX_3ump_e_ -> EX_3ump_e\n",
      "Updating reaction ID: EX_4abut_e_ -> EX_4abut_e\n",
      "Updating reaction ID: EX_4abutn_e_ -> EX_4abutn_e\n",
      "Updating reaction ID: EX_4hdebrisoquine_e_ -> EX_4hdebrisoquine_e\n",
      "Updating reaction ID: EX_4hphac_e_ -> EX_4hphac_e\n",
      "Updating reaction ID: EX_4hpro_e_ -> EX_4hpro_e\n",
      "Updating reaction ID: EX_4mop_e_ -> EX_4mop_e\n",
      "Updating reaction ID: EX_4mptnl_e_ -> EX_4mptnl_e\n",
      "Updating reaction ID: EX_4mtolbutamide_e_ -> EX_4mtolbutamide_e\n",
      "Updating reaction ID: EX_4nph_e_ -> EX_4nph_e\n",
      "Updating reaction ID: EX_4nphsf_e_ -> EX_4nphsf_e\n",
      "Updating reaction ID: EX_4pyrdx_e_ -> EX_4pyrdx_e\n",
      "Updating reaction ID: EX_5adtststerone_e_ -> EX_5adtststerone_e\n",
      "Updating reaction ID: EX_5adtststeroneglc_e_ -> EX_5adtststeroneglc_e\n",
      "Updating reaction ID: EX_5adtststerones_e_ -> EX_5adtststerones_e\n",
      "Updating reaction ID: EX_5dhf_e_ -> EX_5dhf_e\n",
      "Updating reaction ID: EX_5fthf_e_ -> EX_5fthf_e\n",
      "Updating reaction ID: EX_5homeprazole_e_ -> EX_5homeprazole_e\n",
      "Updating reaction ID: EX_5htrp_e_ -> EX_5htrp_e\n",
      "Updating reaction ID: EX_5mta_e_ -> EX_5mta_e\n",
      "Updating reaction ID: EX_5mthf_e_ -> EX_5mthf_e\n",
      "Updating reaction ID: EX_5oxpro_e_ -> EX_5oxpro_e\n",
      "Updating reaction ID: EX_5thf_e_ -> EX_5thf_e\n",
      "Updating reaction ID: EX_6dhf_e_ -> EX_6dhf_e\n",
      "Updating reaction ID: EX_6htststerone_e_ -> EX_6htststerone_e\n",
      "Updating reaction ID: EX_6thf_e_ -> EX_6thf_e\n",
      "Updating reaction ID: EX_7dhf_e_ -> EX_7dhf_e\n",
      "Updating reaction ID: EX_7thf_e_ -> EX_7thf_e\n",
      "Updating reaction ID: EX_9_cis_retfa_e_ -> EX_9_cis_retfa_e\n",
      "Updating reaction ID: EX_abt_e_ -> EX_abt_e\n",
      "Updating reaction ID: EX_ac_e_ -> EX_ac_e\n",
      "Updating reaction ID: EX_acac_e_ -> EX_acac_e\n",
      "Updating reaction ID: EX_acald_e_ -> EX_acald_e\n",
      "Updating reaction ID: EX_acetone_e_ -> EX_acetone_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgam_e_ -> EX_acgam_e\n",
      "Updating reaction ID: EX_ach_e_ -> EX_ach_e\n",
      "Updating reaction ID: EX_acmana_e_ -> EX_acmana_e\n",
      "Updating reaction ID: EX_acn13acngalgbside_cho_e_ -> EX_acn13acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acn23acngalgbside_cho_e_ -> EX_acn23acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acnacngal14acglcgalgluside_cho_e_ -> EX_acnacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acnacngalgbside_cho_e_ -> EX_acnacngalgbside_cho_e\n",
      "Updating reaction ID: EX_acngalacglcgal14acglcgalgluside_cho_e_ -> EX_acngalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_ade_e_ -> EX_ade_e\n",
      "Updating reaction ID: EX_adn_e_ -> EX_adn_e\n",
      "Updating reaction ID: EX_adpcbl_e_ -> EX_adpcbl_e\n",
      "Updating reaction ID: EX_adprbp_e_ -> EX_adprbp_e\n",
      "Updating reaction ID: EX_adprib_e_ -> EX_adprib_e\n",
      "Updating reaction ID: EX_adrn_e_ -> EX_adrn_e\n",
      "Updating reaction ID: EX_adrnl_e_ -> EX_adrnl_e\n",
      "Updating reaction ID: EX_aflatoxin_e_ -> EX_aflatoxin_e\n",
      "Updating reaction ID: EX_ahandrostanglc_e_ -> EX_ahandrostanglc_e\n",
      "Updating reaction ID: EX_ahcys_e_ -> EX_ahcys_e\n",
      "Updating reaction ID: EX_ahdt_e_ -> EX_ahdt_e\n",
      "Updating reaction ID: EX_aicar_e_ -> EX_aicar_e\n",
      "Updating reaction ID: EX_ak2lgchol_cho_e_ -> EX_ak2lgchol_cho_e\n",
      "Updating reaction ID: EX_akg_e_ -> EX_akg_e\n",
      "Updating reaction ID: EX_ala_B_e_ -> EX_ala_B_e\n",
      "Updating reaction ID: EX_ala_D_e_ -> EX_ala_D_e\n",
      "Updating reaction ID: EX_ala_L_e_ -> EX_ala_L_e\n",
      "Updating reaction ID: EX_alaala_e_ -> EX_alaala_e\n",
      "Updating reaction ID: EX_aldstrn_e_ -> EX_aldstrn_e\n",
      "Updating reaction ID: EX_amp_e_ -> EX_amp_e\n",
      "Updating reaction ID: EX_andrstrn_e_ -> EX_andrstrn_e\n",
      "Updating reaction ID: EX_andrstrnglc_e_ -> EX_andrstrnglc_e\n",
      "Updating reaction ID: EX_anth_e_ -> EX_anth_e\n",
      "Updating reaction ID: EX_antipyrene_e_ -> EX_antipyrene_e\n",
      "Updating reaction ID: EX_apnnox_e_ -> EX_apnnox_e\n",
      "Updating reaction ID: EX_appnn_e_ -> EX_appnn_e\n",
      "Updating reaction ID: EX_aprgstrn_e_ -> EX_aprgstrn_e\n",
      "Updating reaction ID: EX_aqcobal_e_ -> EX_aqcobal_e\n",
      "Updating reaction ID: EX_arab_L_e_ -> EX_arab_L_e\n",
      "Updating reaction ID: EX_arach_e_ -> EX_arach_e\n",
      "Updating reaction ID: EX_arachcoa_e_ -> EX_arachcoa_e\n",
      "Updating reaction ID: EX_arachd_e_ -> EX_arachd_e\n",
      "Updating reaction ID: EX_arg_L_e_ -> EX_arg_L_e\n",
      "Updating reaction ID: EX_ascb_L_e_ -> EX_ascb_L_e\n",
      "Updating reaction ID: EX_asn_L_e_ -> EX_asn_L_e\n",
      "Updating reaction ID: EX_asp_D_e_ -> EX_asp_D_e\n",
      "Updating reaction ID: EX_asp_L_e_ -> EX_asp_L_e\n",
      "Updating reaction ID: EX_atp_e_ -> EX_atp_e\n",
      "Updating reaction ID: EX_avite1_e_ -> EX_avite1_e\n",
      "Updating reaction ID: EX_avite2_e_ -> EX_avite2_e\n",
      "Updating reaction ID: EX_bglc_e_ -> EX_bglc_e\n",
      "Updating reaction ID: EX_bhb_e_ -> EX_bhb_e\n",
      "Updating reaction ID: EX_bildglcur_e_ -> EX_bildglcur_e\n",
      "Updating reaction ID: EX_bilglcur_e_ -> EX_bilglcur_e\n",
      "Updating reaction ID: EX_bilirub_e_ -> EX_bilirub_e\n",
      "Updating reaction ID: EX_biocyt_e_ -> EX_biocyt_e\n",
      "Updating reaction ID: EX_btn_e_ -> EX_btn_e\n",
      "Updating reaction ID: EX_but_e_ -> EX_but_e\n",
      "Updating reaction ID: EX_bvite_e_ -> EX_bvite_e\n",
      "Updating reaction ID: EX_bz_e_ -> EX_bz_e\n",
      "Updating reaction ID: EX_C02470_e_ -> EX_C02470_e\n",
      "Updating reaction ID: EX_C02528_e_ -> EX_C02528_e\n",
      "Updating reaction ID: EX_C04849_e_ -> EX_C04849_e\n",
      "Updating reaction ID: EX_ca2_e_ -> EX_ca2_e\n",
      "Updating reaction ID: EX_camp_e_ -> EX_camp_e\n",
      "Updating reaction ID: EX_carn_e_ -> EX_carn_e\n",
      "Updating reaction ID: EX_caro_e_ -> EX_caro_e\n",
      "Updating reaction ID: EX_carveol_e_ -> EX_carveol_e\n",
      "Updating reaction ID: EX_cbasp_e_ -> EX_cbasp_e\n",
      "Updating reaction ID: EX_cca_d3_e_ -> EX_cca_d3_e\n",
      "Updating reaction ID: EX_cdp_e_ -> EX_cdp_e\n",
      "Updating reaction ID: EX_CE0074_e_ -> EX_CE0074_e\n",
      "Updating reaction ID: EX_CE1925_e_ -> EX_CE1925_e\n",
      "Updating reaction ID: EX_CE1926_e_ -> EX_CE1926_e\n",
      "Updating reaction ID: EX_CE1935_e_ -> EX_CE1935_e\n",
      "Updating reaction ID: EX_CE1936_e_ -> EX_CE1936_e\n",
      "Updating reaction ID: EX_CE1939_e_ -> EX_CE1939_e\n",
      "Updating reaction ID: EX_CE1940_e_ -> EX_CE1940_e\n",
      "Updating reaction ID: EX_CE1943_e_ -> EX_CE1943_e\n",
      "Updating reaction ID: EX_CE2011_e_ -> EX_CE2011_e\n",
      "Updating reaction ID: EX_CE2250_e_ -> EX_CE2250_e\n",
      "Updating reaction ID: EX_CE2838_e_ -> EX_CE2838_e\n",
      "Updating reaction ID: EX_CE2839_e_ -> EX_CE2839_e\n",
      "Updating reaction ID: EX_CE2915_e_ -> EX_CE2915_e\n",
      "Updating reaction ID: EX_CE2916_e_ -> EX_CE2916_e\n",
      "Updating reaction ID: EX_CE2917_e_ -> EX_CE2917_e\n",
      "Updating reaction ID: EX_CE4633_e_ -> EX_CE4633_e\n",
      "Updating reaction ID: EX_CE4722_e_ -> EX_CE4722_e\n",
      "Updating reaction ID: EX_CE4723_e_ -> EX_CE4723_e\n",
      "Updating reaction ID: EX_CE4724_e_ -> EX_CE4724_e\n",
      "Updating reaction ID: EX_CE4881_e_ -> EX_CE4881_e\n",
      "Updating reaction ID: EX_CE5786_e_ -> EX_CE5786_e\n",
      "Updating reaction ID: EX_CE5788_e_ -> EX_CE5788_e\n",
      "Updating reaction ID: EX_CE5789_e_ -> EX_CE5789_e\n",
      "Updating reaction ID: EX_CE5797_e_ -> EX_CE5797_e\n",
      "Updating reaction ID: EX_CE5798_e_ -> EX_CE5798_e\n",
      "Updating reaction ID: EX_CE5853_e_ -> EX_CE5853_e\n",
      "Updating reaction ID: EX_CE5854_e_ -> EX_CE5854_e\n",
      "Updating reaction ID: EX_CE5867_e_ -> EX_CE5867_e\n",
      "Updating reaction ID: EX_CE5868_e_ -> EX_CE5868_e\n",
      "Updating reaction ID: EX_CE5869_e_ -> EX_CE5869_e\n",
      "Updating reaction ID: EX_cgly_e_ -> EX_cgly_e\n",
      "Updating reaction ID: EX_chol_e_ -> EX_chol_e\n",
      "Updating reaction ID: EX_cholate_e_ -> EX_cholate_e\n",
      "Updating reaction ID: EX_chsterol_e_ -> EX_chsterol_e\n",
      "Updating reaction ID: EX_chtn_e_ -> EX_chtn_e\n",
      "Updating reaction ID: EX_cit_e_ -> EX_cit_e\n",
      "Updating reaction ID: EX_citr_L_e_ -> EX_citr_L_e\n",
      "Updating reaction ID: EX_cl_e_ -> EX_cl_e\n",
      "Updating reaction ID: EX_CLPND_e_ -> EX_CLPND_e\n",
      "Updating reaction ID: EX_cmp_e_ -> EX_cmp_e\n",
      "Updating reaction ID: EX_co_e_ -> EX_co_e\n",
      "Updating reaction ID: EX_co2_e_ -> EX_co2_e\n",
      "Updating reaction ID: EX_coa_e_ -> EX_coa_e\n",
      "Updating reaction ID: EX_coumarin_e_ -> EX_coumarin_e\n",
      "Updating reaction ID: EX_creat_e_ -> EX_creat_e\n",
      "Updating reaction ID: EX_crm_cho_e_ -> EX_crm_cho_e\n",
      "Updating reaction ID: EX_crmp_cho_e_ -> EX_crmp_cho_e\n",
      "Updating reaction ID: EX_crn_e_ -> EX_crn_e\n",
      "Updating reaction ID: EX_crtsl_e_ -> EX_crtsl_e\n",
      "Updating reaction ID: EX_crtstrn_e_ -> EX_crtstrn_e\n",
      "Updating reaction ID: EX_crvnc_e_ -> EX_crvnc_e\n",
      "Updating reaction ID: EX_csn_e_ -> EX_csn_e\n",
      "Updating reaction ID: EX_cspg_a_e_ -> EX_cspg_a_e\n",
      "Updating reaction ID: EX_cspg_b_e_ -> EX_cspg_b_e\n",
      "Updating reaction ID: EX_cspg_c_e_ -> EX_cspg_c_e\n",
      "Updating reaction ID: EX_cspg_d_e_ -> EX_cspg_d_e\n",
      "Updating reaction ID: EX_cspg_e_e_ -> EX_cspg_ee_\n",
      "Updating reaction ID: EX_ctp_e_ -> EX_ctp_e\n",
      "Updating reaction ID: EX_cyan_e_ -> EX_cyan_e\n",
      "Updating reaction ID: EX_cys_L_e_ -> EX_cys_L_e\n",
      "Updating reaction ID: EX_cysam_e_ -> EX_cysam_e\n",
      "Updating reaction ID: EX_cytd_e_ -> EX_cytd_e\n",
      "Updating reaction ID: EX_dad_2_e_ -> EX_dad_2_e\n",
      "Updating reaction ID: EX_dag_cho_e_ -> EX_dag_cho_e\n",
      "Updating reaction ID: EX_dca_e_ -> EX_dca_e\n",
      "Updating reaction ID: EX_dchac_e_ -> EX_dchac_e\n",
      "Updating reaction ID: EX_dcsptn1_e_ -> EX_dcsptn1_e\n",
      "Updating reaction ID: EX_dcyt_e_ -> EX_dcyt_e\n",
      "Updating reaction ID: EX_debrisoquine_e_ -> EX_debrisoquine_e\n",
      "Updating reaction ID: EX_dgchol_e_ -> EX_dgchol_e\n",
      "Updating reaction ID: EX_dgmp_e_ -> EX_dgmp_e\n",
      "Updating reaction ID: EX_dgsn_e_ -> EX_dgsn_e\n",
      "Updating reaction ID: EX_dgtp_e_ -> EX_dgtp_e\n",
      "Updating reaction ID: EX_dha_e_ -> EX_dha_e\n",
      "Updating reaction ID: EX_dhap_e_ -> EX_dhap_e\n",
      "Updating reaction ID: EX_dhdascb_e_ -> EX_dhdascb_e\n",
      "Updating reaction ID: EX_dheas_e_ -> EX_dheas_e\n",
      "Updating reaction ID: EX_dhf_e_ -> EX_dhf_e\n",
      "Updating reaction ID: EX_digalsgalside_cho_e_ -> EX_digalsgalside_cho_e\n",
      "Updating reaction ID: EX_din_e_ -> EX_din_e\n",
      "Updating reaction ID: EX_dlnlcg_e_ -> EX_dlnlcg_e\n",
      "Updating reaction ID: EX_dmantipyrine_e_ -> EX_dmantipyrine_e\n",
      "Updating reaction ID: EX_dmhptcrn_e_ -> EX_dmhptcrn_e\n",
      "Updating reaction ID: EX_dopa_e_ -> EX_dopa_e\n",
      "Updating reaction ID: EX_dopasf_e_ -> EX_dopasf_e\n",
      "Updating reaction ID: EX_dpcoa_e_ -> EX_dpcoa_e\n",
      "Updating reaction ID: EX_drib_e_ -> EX_drib_e\n",
      "Updating reaction ID: EX_dtdp_e_ -> EX_dtdp_e\n",
      "Updating reaction ID: EX_dtmp_e_ -> EX_dtmp_e\n",
      "Updating reaction ID: EX_dttp_e_ -> EX_dttp_e\n",
      "Updating reaction ID: EX_duri_e_ -> EX_duri_e\n",
      "Updating reaction ID: EX_eaflatoxin_e_ -> EX_eaflatoxin_e\n",
      "Updating reaction ID: EX_ebastine_e_ -> EX_ebastine_e\n",
      "Updating reaction ID: EX_ebastineoh_e_ -> EX_ebastineoh_e\n",
      "Updating reaction ID: EX_eicostet_e_ -> EX_eicostet_e\n",
      "Updating reaction ID: EX_elaid_e_ -> EX_elaid_e\n",
      "Updating reaction ID: EX_estradiol_e_ -> EX_estradiol_e\n",
      "Updating reaction ID: EX_estradiolglc_e_ -> EX_estradiolglc_e\n",
      "Updating reaction ID: EX_estriolglc_e_ -> EX_estriolglc_e\n",
      "Updating reaction ID: EX_estroneglc_e_ -> EX_estroneglc_e\n",
      "Updating reaction ID: EX_estrones_e_ -> EX_estrones_e\n",
      "Updating reaction ID: EX_etoh_e_ -> EX_etoh_e\n",
      "Updating reaction ID: EX_fad_e_ -> EX_fad_e\n",
      "Updating reaction ID: EX_fald_e_ -> EX_fald_e\n",
      "Updating reaction ID: EX_fe2_e_ -> EX_fe2_e\n",
      "Updating reaction ID: EX_fe3_e_ -> EX_fe3_e\n",
      "Updating reaction ID: EX_fmn_e_ -> EX_fmn_e\n",
      "Updating reaction ID: EX_fol_e_ -> EX_fol_e\n",
      "Updating reaction ID: EX_for_e_ -> EX_for_e\n",
      "Updating reaction ID: EX_fru_e_ -> EX_fru_e\n",
      "Updating reaction ID: EX_fuc_L_e_ -> EX_fuc_L_e\n",
      "Updating reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fuc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fuc14galacglcgalgluside_cho_e_ -> EX_fuc14galacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacgalfucgalacglcgalgluside_cho_e_ -> EX_fucacgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngal14acglcgalgluside_cho_e_ -> EX_fucacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngalacglcgalgluside_cho_e_ -> EX_fucacngalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc12gal14acglcgalgluside_cho_e_ -> EX_fucfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfuc132galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucgalacglcgalgluside_cho_e_ -> EX_fucfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgal14acglcgalgluside_cho_e_ -> EX_fucgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalfucgalacglcgalgluside_cho_e_ -> EX_fucgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalgbside_cho_e_ -> EX_fucgalgbside_cho_e\n",
      "Updating reaction ID: EX_g1p_e_ -> EX_g1p_e\n",
      "Updating reaction ID: EX_gal_e_ -> EX_gal_e\n",
      "Updating reaction ID: EX_galacglcgalgbside_cho_e_ -> EX_galacglcgalgbside_cho_e\n",
      "Updating reaction ID: EX_galfuc12gal14acglcgalgluside_cho_e_ -> EX_galfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_galfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e_ -> EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalgalthcrm_cho_e_ -> EX_galgalgalthcrm_cho_e\n",
      "Updating reaction ID: EX_galside_cho_e_ -> EX_galside_cho_e\n",
      "Updating reaction ID: EX_gam_e_ -> EX_gam_e\n",
      "Updating reaction ID: EX_gbside_cho_e_ -> EX_gbside_cho_e\n",
      "Updating reaction ID: EX_gchola_e_ -> EX_gchola_e\n",
      "Updating reaction ID: EX_gd1b2_cho_e_ -> EX_gd1b2_cho_e\n",
      "Updating reaction ID: EX_gd1c_cho_e_ -> EX_gd1c_cho_e\n",
      "Updating reaction ID: EX_gdp_e_ -> EX_gdp_e\n",
      "Updating reaction ID: EX_glc_e_ -> EX_glc_e\n",
      "Updating reaction ID: EX_glcur_e_ -> EX_glcur_e\n",
      "Updating reaction ID: EX_glgchlo_e_ -> EX_glgchlo_e\n",
      "Updating reaction ID: EX_gln_L_e_ -> EX_gln_L_e\n",
      "Updating reaction ID: EX_gltcho_e_ -> EX_gltcho_e\n",
      "Updating reaction ID: EX_gltdechol_e_ -> EX_gltdechol_e\n",
      "Updating reaction ID: EX_glu_L_e_ -> EX_glu_L_e\n",
      "Updating reaction ID: EX_gluala_e_ -> EX_gluala_e\n",
      "Updating reaction ID: EX_gly_e_ -> EX_gly_e\n",
      "Updating reaction ID: EX_glyb_e_ -> EX_glyb_e\n",
      "Updating reaction ID: EX_glyc_e_ -> EX_glyc_e\n",
      "Updating reaction ID: EX_glyc_S_e_ -> EX_glyc_S_e\n",
      "Updating reaction ID: EX_glygly_e_ -> EX_glygly_e\n",
      "Updating reaction ID: EX_glygn2_e_ -> EX_glygn2_e\n",
      "Updating reaction ID: EX_glygn4_e_ -> EX_glygn4_e\n",
      "Updating reaction ID: EX_glygn5_e_ -> EX_glygn5_e\n",
      "Updating reaction ID: EX_glyleu_e_ -> EX_glyleu_e\n",
      "Updating reaction ID: EX_glyphe_e_ -> EX_glyphe_e\n",
      "Updating reaction ID: EX_glypro_e_ -> EX_glypro_e\n",
      "Updating reaction ID: EX_glysar_e_ -> EX_glysar_e\n",
      "Updating reaction ID: EX_gmp_e_ -> EX_gmp_e\n",
      "Updating reaction ID: EX_gp1c_cho_e_ -> EX_gp1c_cho_e\n",
      "Updating reaction ID: EX_gp1calpha_cho_e_ -> EX_gp1calpha_cho_e\n",
      "Updating reaction ID: EX_gq1b_cho_e_ -> EX_gq1b_cho_e\n",
      "Updating reaction ID: EX_gq1balpha_cho_e_ -> EX_gq1balpha_cho_e\n",
      "Updating reaction ID: EX_gsn_e_ -> EX_gsn_e\n",
      "Updating reaction ID: EX_gt1a_cho_e_ -> EX_gt1a_cho_e\n",
      "Updating reaction ID: EX_gthox_e_ -> EX_gthox_e\n",
      "Updating reaction ID: EX_gthrd_e_ -> EX_gthrd_e\n",
      "Updating reaction ID: EX_gtp_e_ -> EX_gtp_e\n",
      "Updating reaction ID: EX_gua_e_ -> EX_gua_e\n",
      "Updating reaction ID: EX_gum_e_ -> EX_gum_e\n",
      "Updating reaction ID: EX_gumdchac_e_ -> EX_gumdchac_e\n",
      "Updating reaction ID: EX_gumgchol_e_ -> EX_gumgchol_e\n",
      "Updating reaction ID: EX_gumtchol_e_ -> EX_gumtchol_e\n",
      "Updating reaction ID: EX_h_e_ -> EX_h_e\n",
      "Updating reaction ID: EX_h2o_e_ -> EX_h2o_e\n",
      "Updating reaction ID: EX_h2o2_e_ -> EX_h2o2_e\n",
      "Updating reaction ID: EX_ha_e_ -> EX_ha_e\n",
      "Updating reaction ID: EX_ha_pre1_e_ -> EX_ha_pre1_e\n",
      "Updating reaction ID: EX_HC00004_e_ -> EX_HC00004_e\n",
      "Updating reaction ID: EX_HC00229_e_ -> EX_HC00229_e\n",
      "Updating reaction ID: EX_HC00250_e_ -> EX_HC00250_e\n",
      "Updating reaction ID: EX_HC00822_e_ -> EX_HC00822_e\n",
      "Updating reaction ID: EX_HC00955_e_ -> EX_HC00955_e\n",
      "Updating reaction ID: EX_HC01104_e_ -> EX_HC01104_e\n",
      "Updating reaction ID: EX_HC01361_e_ -> EX_HC01361_e\n",
      "Updating reaction ID: EX_HC01440_e_ -> EX_HC01440_e\n",
      "Updating reaction ID: EX_HC01441_e_ -> EX_HC01441_e\n",
      "Updating reaction ID: EX_HC01444_e_ -> EX_HC01444_e\n",
      "Updating reaction ID: EX_HC01446_e_ -> EX_HC01446_e\n",
      "Updating reaction ID: EX_HC01577_e_ -> EX_HC01577_e\n",
      "Updating reaction ID: EX_HC01609_e_ -> EX_HC01609_e\n",
      "Updating reaction ID: EX_HC01700_e_ -> EX_HC01700_e\n",
      "Updating reaction ID: EX_HC01787_e_ -> EX_HC01787_e\n",
      "Updating reaction ID: EX_HC02154_e_ -> EX_HC02154_e\n",
      "Updating reaction ID: EX_HC02160_e_ -> EX_HC02160_e\n",
      "Updating reaction ID: EX_HC02161_e_ -> EX_HC02161_e\n",
      "Updating reaction ID: EX_HC02172_e_ -> EX_HC02172_e\n",
      "Updating reaction ID: EX_HC02180_e_ -> EX_HC02180_e\n",
      "Updating reaction ID: EX_HC02187_e_ -> EX_HC02187_e\n",
      "Updating reaction ID: EX_HC02191_e_ -> EX_HC02191_e\n",
      "Updating reaction ID: EX_HC02192_e_ -> EX_HC02192_e\n",
      "Updating reaction ID: EX_HC02193_e_ -> EX_HC02193_e\n",
      "Updating reaction ID: EX_HC02194_e_ -> EX_HC02194_e\n",
      "Updating reaction ID: EX_HC02195_e_ -> EX_HC02195_e\n",
      "Updating reaction ID: EX_HC02196_e_ -> EX_HC02196_e\n",
      "Updating reaction ID: EX_HC02197_e_ -> EX_HC02197_e\n",
      "Updating reaction ID: EX_HC02198_e_ -> EX_HC02198_e\n",
      "Updating reaction ID: EX_HC02199_e_ -> EX_HC02199_e\n",
      "Updating reaction ID: EX_HC02200_e_ -> EX_HC02200_e\n",
      "Updating reaction ID: EX_HC02201_e_ -> EX_HC02201_e\n",
      "Updating reaction ID: EX_HC02202_e_ -> EX_HC02202_e\n",
      "Updating reaction ID: EX_HC02203_e_ -> EX_HC02203_e\n",
      "Updating reaction ID: EX_HC02204_e_ -> EX_HC02204_e\n",
      "Updating reaction ID: EX_HC02205_e_ -> EX_HC02205_e\n",
      "Updating reaction ID: EX_HC02206_e_ -> EX_HC02206_e\n",
      "Updating reaction ID: EX_HC02207_e_ -> EX_HC02207_e\n",
      "Updating reaction ID: EX_HC02208_e_ -> EX_HC02208_e\n",
      "Updating reaction ID: EX_HC02210_e_ -> EX_HC02210_e\n",
      "Updating reaction ID: EX_HC02213_e_ -> EX_HC02213_e\n",
      "Updating reaction ID: EX_HC02214_e_ -> EX_HC02214_e\n",
      "Updating reaction ID: EX_HC02216_e_ -> EX_HC02216_e\n",
      "Updating reaction ID: EX_HC02217_e_ -> EX_HC02217_e\n",
      "Updating reaction ID: EX_HC02220_e_ -> EX_HC02220_e\n",
      "Updating reaction ID: EX_hco3_e_ -> EX_hco3_e\n",
      "Updating reaction ID: EX_hcoumarin_e_ -> EX_hcoumarin_e\n",
      "Updating reaction ID: EX_hdca_e_ -> EX_hdca_e\n",
      "Updating reaction ID: EX_hdcea_e_ -> EX_hdcea_e\n",
      "Updating reaction ID: EX_hestratriol_e_ -> EX_hestratriol_e\n",
      "Updating reaction ID: EX_hexc_e_ -> EX_hexc_e\n",
      "Updating reaction ID: EX_his_L_e_ -> EX_his_L_e\n",
      "Updating reaction ID: EX_hista_e_ -> EX_hista_e\n",
      "Updating reaction ID: EX_hom_L_e_ -> EX_hom_L_e\n",
      "Updating reaction ID: EX_hpdca_e_ -> EX_hpdca_e\n",
      "Updating reaction ID: EX_chopg_e_ -> EX_chopg_e\n",
      "Updating reaction ID: EX_htaxol_e_ -> EX_htaxol_e\n",
      "Updating reaction ID: EX_hxan_e_ -> EX_hxan_e\n",
      "Updating reaction ID: EX_hyptaur_e_ -> EX_hyptaur_e\n",
      "Updating reaction ID: EX_i_e_ -> EX_i_e\n",
      "Updating reaction ID: EX_idour_e_ -> EX_idour_e\n",
      "Updating reaction ID: EX_idp_e_ -> EX_idp_e\n",
      "Updating reaction ID: EX_ile_L_e_ -> EX_ile_L_e\n",
      "Updating reaction ID: EX_imp_e_ -> EX_imp_e\n",
      "Updating reaction ID: EX_inost_e_ -> EX_inost_e\n",
      "Updating reaction ID: EX_ins_e_ -> EX_ins_e\n",
      "Updating reaction ID: EX_k_e_ -> EX_k_e\n",
      "Updating reaction ID: EX_ksi_deg1_e_ -> EX_ksi_deg1_e\n",
      "Updating reaction ID: EX_ksi_e_ -> EX_ksi_e\n",
      "Updating reaction ID: EX_ksii_core2_e_ -> EX_ksii_core2_e\n",
      "Updating reaction ID: EX_ksii_core4_e_ -> EX_ksii_core4_e\n",
      "Updating reaction ID: EX_lac_D_e_ -> EX_lac_D_e\n",
      "Updating reaction ID: EX_lac_L_e_ -> EX_lac_L_e\n",
      "Updating reaction ID: EX_lcts_e_ -> EX_lcts_e\n",
      "Updating reaction ID: EX_Lcystin_e_ -> EX_Lcystin_e\n",
      "Updating reaction ID: EX_leu_L_e_ -> EX_leu_L_e\n",
      "Updating reaction ID: EX_leugly_e_ -> EX_leugly_e\n",
      "Updating reaction ID: EX_leuktrA4_e_ -> EX_leuktrA4_e\n",
      "Updating reaction ID: EX_leuktrB4_e_ -> EX_leuktrB4_e\n",
      "Updating reaction ID: EX_leuktrC4_e_ -> EX_leuktrC4_e\n",
      "Updating reaction ID: EX_leuktrD4_e_ -> EX_leuktrD4_e\n",
      "Updating reaction ID: EX_leuktrE4_e_ -> EX_leuktrE4_e\n",
      "Updating reaction ID: EX_leuktrF4_e_ -> EX_leuktrF4_e\n",
      "Updating reaction ID: EX_leuleu_e_ -> EX_leuleu_e\n",
      "Updating reaction ID: EX_lgnc_e_ -> EX_lgnc_e\n",
      "Updating reaction ID: EX_limnen_e_ -> EX_limnen_e\n",
      "Updating reaction ID: EX_lipoate_e_ -> EX_lipoate_e\n",
      "Updating reaction ID: EX_lneldc_e_ -> EX_lneldc_e\n",
      "Updating reaction ID: EX_lnlc_e_ -> EX_lnlc_e\n",
      "Updating reaction ID: EX_lnlnca_e_ -> EX_lnlnca_e\n",
      "Updating reaction ID: EX_lnlncg_e_ -> EX_lnlncg_e\n",
      "Updating reaction ID: EX_lpchol_cho_e_ -> EX_lpchol_cho_e\n",
      "Updating reaction ID: EX_lys_L_e_ -> EX_lys_L_e\n",
      "Updating reaction ID: EX_mag_cho_e_ -> EX_mag_cho_e\n",
      "Updating reaction ID: EX_mal_L_e_ -> EX_mal_L_e\n",
      "Updating reaction ID: EX_malcoa_e_ -> EX_malcoa_e\n",
      "Updating reaction ID: EX_malt_e_ -> EX_malt_e\n",
      "Updating reaction ID: EX_malthp_e_ -> EX_malthp_e\n",
      "Updating reaction ID: EX_malthx_e_ -> EX_malthx_e\n",
      "Updating reaction ID: EX_maltpt_e_ -> EX_maltpt_e\n",
      "Updating reaction ID: EX_malttr_e_ -> EX_malttr_e\n",
      "Updating reaction ID: EX_maltttr_e_ -> EX_maltttr_e\n",
      "Updating reaction ID: EX_man_e_ -> EX_man_e\n",
      "Updating reaction ID: EX_meoh_e_ -> EX_meoh_e\n",
      "Updating reaction ID: EX_mepi_e_ -> EX_mepi_e\n",
      "Updating reaction ID: EX_mercplaccys_e_ -> EX_mercplaccys_e\n",
      "Updating reaction ID: EX_met_L_e_ -> EX_met_L_e\n",
      "Updating reaction ID: EX_mthgxl_e_ -> EX_mthgxl_e\n",
      "Updating reaction ID: EX_n2m2nmasn_e_ -> EX_n2m2nmasn_e\n",
      "Updating reaction ID: EX_na1_e_ -> EX_na1_e\n",
      "Updating reaction ID: EX_nac_e_ -> EX_nac_e\n",
      "Updating reaction ID: EX_nad_e_ -> EX_nad_e\n",
      "Updating reaction ID: EX_nadp_e_ -> EX_nadp_e\n",
      "Updating reaction ID: EX_ncam_e_ -> EX_ncam_e\n",
      "Updating reaction ID: EX_nh4_e_ -> EX_nh4_e\n",
      "Updating reaction ID: EX_nifedipine_e_ -> EX_nifedipine_e\n",
      "Updating reaction ID: EX_no_e_ -> EX_no_e\n",
      "Updating reaction ID: EX_no2_e_ -> EX_no2_e\n",
      "Updating reaction ID: EX_npthl_e_ -> EX_npthl_e\n",
      "Updating reaction ID: EX_nrpphr_e_ -> EX_nrpphr_e\n",
      "Updating reaction ID: EX_nrpphrsf_e_ -> EX_nrpphrsf_e\n",
      "Updating reaction ID: EX_nrvnc_e_ -> EX_nrvnc_e\n",
      "Updating reaction ID: EX_o2_e_ -> EX_o2_e\n",
      "Updating reaction ID: EX_o2s_e_ -> EX_o2s_e\n",
      "Updating reaction ID: EX_oagd3_cho_e_ -> EX_oagd3_cho_e\n",
      "Updating reaction ID: EX_oagt3_cho_e_ -> EX_oagt3_cho_e\n",
      "Updating reaction ID: EX_ocdca_e_ -> EX_ocdca_e\n",
      "Updating reaction ID: EX_ocdcea_e_ -> EX_ocdcea_e\n",
      "Updating reaction ID: EX_octa_e_ -> EX_octa_e\n",
      "Updating reaction ID: EX_oh1_e_ -> EX_oh1_e\n",
      "Updating reaction ID: EX_omeprazole_e_ -> EX_omeprazole_e\n",
      "Updating reaction ID: EX_onpthl_e_ -> EX_onpthl_e\n",
      "Updating reaction ID: EX_orn_e_ -> EX_orn_e\n",
      "Updating reaction ID: EX_orot_e_ -> EX_orot_e\n",
      "Updating reaction ID: EX_oxa_e_ -> EX_oxa_e\n",
      "Updating reaction ID: EX_paf_cho_e_ -> EX_paf_cho_e\n",
      "Updating reaction ID: EX_pan4p_e_ -> EX_pan4p_e\n",
      "Updating reaction ID: EX_pchol_cho_e_ -> EX_pchol_cho_e\n",
      "Updating reaction ID: EX_pe_cho_e_ -> EX_pe_cho_e\n",
      "Updating reaction ID: EX_pect_e_ -> EX_pect_e\n",
      "Updating reaction ID: EX_pectindchac_e_ -> EX_pectindchac_e\n",
      "Updating reaction ID: EX_pectingchol_e_ -> EX_pectingchol_e\n",
      "Updating reaction ID: EX_pectintchol_e_ -> EX_pectintchol_e\n",
      "Updating reaction ID: EX_peplys_e_ -> EX_peplys_e\n",
      "Updating reaction ID: EX_perillyl_e_ -> EX_perillyl_e\n",
      "Updating reaction ID: EX_pglyc_cho_e_ -> EX_pglyc_cho_e\n",
      "Updating reaction ID: EX_phe_L_e_ -> EX_phe_L_e\n",
      "Updating reaction ID: EX_pheacgln_e_ -> EX_pheacgln_e\n",
      "Updating reaction ID: EX_pheme_e_ -> EX_pheme_e\n",
      "Updating reaction ID: EX_phyQ_e_ -> EX_phyQ_e\n",
      "Updating reaction ID: EX_phyt_e_ -> EX_phyt_e\n",
      "Updating reaction ID: EX_pi_e_ -> EX_pi_e\n",
      "Updating reaction ID: EX_pnto_R_e_ -> EX_pnto_R_e\n",
      "Updating reaction ID: EX_ppa_e_ -> EX_ppa_e\n",
      "Updating reaction ID: EX_ppi_e_ -> EX_ppi_e\n",
      "Updating reaction ID: EX_prgstrn_e_ -> EX_prgstrn_e\n",
      "Updating reaction ID: EX_pro_D_e_ -> EX_pro_D_e\n",
      "Updating reaction ID: EX_pro_L_e_ -> EX_pro_L_e\n",
      "Updating reaction ID: EX_progly_e_ -> EX_progly_e\n",
      "Updating reaction ID: EX_prostgd2_e_ -> EX_prostgd2_e\n",
      "Updating reaction ID: EX_prostge1_e_ -> EX_prostge1_e\n",
      "Updating reaction ID: EX_prostge2_e_ -> EX_prostge2_e\n",
      "Updating reaction ID: EX_prostgf2_e_ -> EX_prostgf2_e\n",
      "Updating reaction ID: EX_prostgh2_e_ -> EX_prostgh2_e\n",
      "Updating reaction ID: EX_prostgi2_e_ -> EX_prostgi2_e\n",
      "Updating reaction ID: EX_prpp_e_ -> EX_prpp_e\n",
      "Updating reaction ID: EX_ps_cho_e_ -> EX_ps_cho_e\n",
      "Updating reaction ID: EX_psyl_e_ -> EX_psyl_e\n",
      "Updating reaction ID: EX_psylchol_e_ -> EX_psylchol_e\n",
      "Updating reaction ID: EX_psyltchol_e_ -> EX_psyltchol_e\n",
      "Updating reaction ID: EX_psyltdechol_e_ -> EX_psyltdechol_e\n",
      "Updating reaction ID: EX_ptdca_e_ -> EX_ptdca_e\n",
      "Updating reaction ID: EX_ptrc_e_ -> EX_ptrc_e\n",
      "Updating reaction ID: EX_ptth_e_ -> EX_ptth_e\n",
      "Updating reaction ID: EX_pydam_e_ -> EX_pydam_e\n",
      "Updating reaction ID: EX_pydx_e_ -> EX_pydx_e\n",
      "Updating reaction ID: EX_pydx5p_e_ -> EX_pydx5p_e\n",
      "Updating reaction ID: EX_pydxn_e_ -> EX_pydxn_e\n",
      "Updating reaction ID: EX_pyr_e_ -> EX_pyr_e\n",
      "Updating reaction ID: EX_q10_e_ -> EX_q10_e\n",
      "Updating reaction ID: EX_q10h2_e_ -> EX_q10h2_e\n",
      "Updating reaction ID: EX_rbt_e_ -> EX_rbt_e\n",
      "Updating reaction ID: EX_retfa_e_ -> EX_retfa_e\n",
      "Updating reaction ID: EX_retinol_9_cis_e_ -> EX_retinol_9_cis_e\n",
      "Updating reaction ID: EX_retinol_cis_11_e_ -> EX_retinol_cis_11_e\n",
      "Updating reaction ID: EX_retinol_e_ -> EX_retinol_e\n",
      "Updating reaction ID: EX_retn_e_ -> EX_retn_e\n",
      "Updating reaction ID: EX_retnglc_e_ -> EX_retnglc_e\n",
      "Updating reaction ID: EX_rib_D_e_ -> EX_rib_D_e\n",
      "Updating reaction ID: EX_ribflv_e_ -> EX_ribflv_e\n",
      "Updating reaction ID: EX_Rtotal_e_ -> EX_Rtotal_e\n",
      "Updating reaction ID: EX_Rtotal2_e_ -> EX_Rtotal2_e\n",
      "Updating reaction ID: EX_Rtotal3_e_ -> EX_Rtotal3_e\n",
      "Updating reaction ID: EX_s2l2fn2m2masn_e_ -> EX_s2l2fn2m2masn_e\n",
      "Updating reaction ID: EX_s2l2n2m2masn_e_ -> EX_s2l2n2m2masn_e\n",
      "Updating reaction ID: EX_sarcs_e_ -> EX_sarcs_e\n",
      "Updating reaction ID: EX_sbt_d_e_ -> EX_sbt_d_e\n",
      "Updating reaction ID: EX_sel_e_ -> EX_sel_e\n",
      "Updating reaction ID: EX_ser_D_e_ -> EX_ser_D_e\n",
      "Updating reaction ID: EX_ser_L_e_ -> EX_ser_L_e\n",
      "Updating reaction ID: EX_sfcys_e_ -> EX_sfcys_e\n",
      "Updating reaction ID: EX_sl_L_e_ -> EX_sl_L_e\n",
      "Updating reaction ID: EX_so4_e_ -> EX_so4_e\n",
      "Updating reaction ID: EX_spc_cho_e_ -> EX_spc_cho_e\n",
      "Updating reaction ID: EX_sph1p_e_ -> EX_sph1p_e\n",
      "Updating reaction ID: EX_sphs1p_e_ -> EX_sphs1p_e\n",
      "Updating reaction ID: EX_spmd_e_ -> EX_spmd_e\n",
      "Updating reaction ID: EX_sprm_e_ -> EX_sprm_e\n",
      "Updating reaction ID: EX_srtn_e_ -> EX_srtn_e\n",
      "Updating reaction ID: EX_strch1_e_ -> EX_strch1_e\n",
      "Updating reaction ID: EX_strch2_e_ -> EX_strch2_e\n",
      "Updating reaction ID: EX_strdnc_e_ -> EX_strdnc_e\n",
      "Updating reaction ID: EX_succ_e_ -> EX_succ_e\n",
      "Updating reaction ID: EX_sucr_e_ -> EX_sucr_e\n",
      "Updating reaction ID: EX_tag_cho_e_ -> EX_tag_cho_e\n",
      "Updating reaction ID: EX_tagat_D_e_ -> EX_tagat_D_e\n",
      "Updating reaction ID: EX_taur_e_ -> EX_taur_e\n",
      "Updating reaction ID: EX_taxol_e_ -> EX_taxol_e\n",
      "Updating reaction ID: EX_tchola_e_ -> EX_tchola_e\n",
      "Updating reaction ID: EX_tcynt_e_ -> EX_tcynt_e\n",
      "Updating reaction ID: EX_tdchola_e_ -> EX_tdchola_e\n",
      "Updating reaction ID: EX_tdechola_e_ -> EX_tdechola_e\n",
      "Updating reaction ID: EX_tethex3_e_ -> EX_tethex3_e\n",
      "Updating reaction ID: EX_tetpent3_e_ -> EX_tetpent3_e\n",
      "Updating reaction ID: EX_tetpent6_e_ -> EX_tetpent6_e\n",
      "Updating reaction ID: EX_tettet6_e_ -> EX_tettet6_e\n",
      "Updating reaction ID: EX_thf_e_ -> EX_thf_e\n",
      "Updating reaction ID: EX_thm_e_ -> EX_thm_e\n",
      "Updating reaction ID: EX_thmmp_e_ -> EX_thmmp_e\n",
      "Updating reaction ID: EX_thmtp_e_ -> EX_thmtp_e\n",
      "Updating reaction ID: EX_thr_L_e_ -> EX_thr_L_e\n",
      "Updating reaction ID: EX_thym_e_ -> EX_thym_e\n",
      "Updating reaction ID: EX_thymd_e_ -> EX_thymd_e\n",
      "Updating reaction ID: EX_thyox_L_e_ -> EX_thyox_L_e\n",
      "Updating reaction ID: EX_tmndnc_e_ -> EX_tmndnc_e\n",
      "Updating reaction ID: EX_tolbutamide_e_ -> EX_tolbutamide_e\n",
      "Updating reaction ID: EX_tre_e_ -> EX_tre_e\n",
      "Updating reaction ID: EX_triodthy_e_ -> EX_triodthy_e\n",
      "Updating reaction ID: EX_triodthysuf_e_ -> EX_triodthysuf_e\n",
      "Updating reaction ID: EX_trp_L_e_ -> EX_trp_L_e\n",
      "Updating reaction ID: EX_tststerone_e_ -> EX_tststerone_e\n",
      "Updating reaction ID: EX_tststeroneglc_e_ -> EX_tststeroneglc_e\n",
      "Updating reaction ID: EX_tststerones_e_ -> EX_tststerones_e\n",
      "Updating reaction ID: EX_tsul_e_ -> EX_tsul_e\n",
      "Updating reaction ID: EX_ttdca_e_ -> EX_ttdca_e\n",
      "Updating reaction ID: EX_txa2_e_ -> EX_txa2_e\n",
      "Updating reaction ID: EX_tymsf_e_ -> EX_tymsf_e\n",
      "Updating reaction ID: EX_Tyr_ggn_e_ -> EX_Tyr_ggn_e\n",
      "Updating reaction ID: EX_tyr_L_e_ -> EX_tyr_L_e\n",
      "Updating reaction ID: EX_udp_e_ -> EX_udp_e\n",
      "Updating reaction ID: EX_udpg_e_ -> EX_udpg_e\n",
      "Updating reaction ID: EX_udpgal_e_ -> EX_udpgal_e\n",
      "Updating reaction ID: EX_ump_e_ -> EX_ump_e\n",
      "Updating reaction ID: EX_ura_e_ -> EX_ura_e\n",
      "Updating reaction ID: EX_urate_e_ -> EX_urate_e\n",
      "Updating reaction ID: EX_urea_e_ -> EX_urea_e\n",
      "Updating reaction ID: EX_uri_e_ -> EX_uri_e\n",
      "Updating reaction ID: EX_utp_e_ -> EX_utp_e\n",
      "Updating reaction ID: EX_vacc_e_ -> EX_vacc_e\n",
      "Updating reaction ID: EX_val_L_e_ -> EX_val_L_e\n",
      "Updating reaction ID: EX_vitd2_e_ -> EX_vitd2_e\n",
      "Updating reaction ID: EX_vitd3_e_ -> EX_vitd3_e\n",
      "Updating reaction ID: EX_whddca_e_ -> EX_whddca_e\n",
      "Updating reaction ID: EX_whhdca_e_ -> EX_whhdca_e\n",
      "Updating reaction ID: EX_whtststerone_e_ -> EX_whtststerone_e\n",
      "Updating reaction ID: EX_whttdca_e_ -> EX_whttdca_e\n",
      "Updating reaction ID: EX_xolest_cho_e_ -> EX_xolest_cho_e\n",
      "Updating reaction ID: EX_xolest2_cho_e_ -> EX_xolest2_cho_e\n",
      "Updating reaction ID: EX_xoltri24_e_ -> EX_xoltri24_e\n",
      "Updating reaction ID: EX_xoltri25_e_ -> EX_xoltri25_e\n",
      "Updating reaction ID: EX_xoltri27_e_ -> EX_xoltri27_e\n",
      "Updating reaction ID: EX_xyl_D_e_ -> EX_xyl_D_e\n",
      "Updating reaction ID: EX_xylt_e_ -> EX_xylt_e\n",
      "Updating reaction ID: EX_yvite_e_ -> EX_yvite_e\n",
      "Updating reaction ID: SK_Asn_X_Ser_Thr_r_ -> SK_Asn_X_Ser_Thr_r\n",
      "Updating reaction ID: SK_Ser_Thr_g_ -> SK_Ser_Thr_g\n",
      "Updating reaction ID: SK_Tyr_ggn_c_ -> SK_Tyr_ggn_c\n",
      "Updating reaction ID: DM_epo_g_ -> DM_epo_g\n",
      "Updating reaction ID: DM_igg_g_ -> DM_igg_g\n",
      "Updating reaction ID: DM_T_antigen_g_ -> DM_T_antigen_g\n",
      "Updating reaction ID: DM_avite1_c_ -> DM_avite1_c\n",
      "Updating reaction ID: DM_avite2_c_ -> DM_avite2_c\n",
      "Updating reaction ID: DM_bvite_c_ -> DM_bvite_c\n",
      "Updating reaction ID: DM_core5_g_ -> DM_core5_g\n",
      "Updating reaction ID: DM_core7_g_ -> DM_core7_g\n",
      "Updating reaction ID: DM_core8_g_ -> DM_core8_g\n",
      "Updating reaction ID: DM_dem2emgacpail_prot_cho_r_ -> DM_dem2emgacpail_prot_cho_r\n",
      "Updating reaction ID: DM_dgpi_prot_cho_r_ -> DM_dgpi_prot_cho_r\n",
      "Updating reaction ID: DM_dsT_antigen_g_ -> DM_dsT_antigen_g\n",
      "Updating reaction ID: DM_ethamp_r_ -> DM_ethamp_r\n",
      "Updating reaction ID: DM_gncore2_g_ -> DM_gncore2_g\n",
      "Updating reaction ID: DM_kdn_c_ -> DM_kdn_c\n",
      "Updating reaction ID: DM_n5m2masn_g_ -> DM_n5m2masn_g\n",
      "Updating reaction ID: DM_sTn_antigen_g_ -> DM_sTn_antigen_g\n",
      "Updating reaction ID: DM_sprm_c_ -> DM_sprm_c\n",
      "Updating reaction ID: DM_yvite_c_ -> DM_yvite_c\n",
      "Updating reaction ID: DM_56iqcrbxlt_c_ -> DM_56iqcrbxlt_c\n",
      "Updating reaction ID: DM_atp_c_ -> DM_atp_c\n",
      "Updating reaction ID: SK_pre_prot_r_ -> SK_pre_prot_r\n",
      "Updating reaction ID: EX_HC01610_e_ -> EX_HC01610_e\n",
      "Updating reaction ID: EX_HC02179_e_ -> EX_HC02179_e\n",
      "Updating reaction ID: EX_C05965_e_ -> EX_C05965_e\n",
      "Updating reaction ID: EX_HC02176_e_ -> EX_HC02176_e\n",
      "Updating reaction ID: EX_10fthf_e_ -> EX_10fthf_e\n",
      "Updating reaction ID: EX_10fthf5glu_e_ -> EX_10fthf5glu_e\n",
      "Updating reaction ID: EX_10fthf6glu_e_ -> EX_10fthf6glu_e\n",
      "Updating reaction ID: EX_10fthf7glu_e_ -> EX_10fthf7glu_e\n",
      "Updating reaction ID: EX_11_cis_retfa_e_ -> EX_11_cis_retfa_e\n",
      "Updating reaction ID: EX_13_cis_retnglc_e_ -> EX_13_cis_retnglc_e\n",
      "Updating reaction ID: EX_1glyc_cho_e_ -> EX_1glyc_cho_e\n",
      "Updating reaction ID: EX_1mncam_e_ -> EX_1mncam_e\n",
      "Updating reaction ID: EX_23cump_e_ -> EX_23cump_e\n",
      "Updating reaction ID: EX_2425dhvitd2_e_ -> EX_2425dhvitd2_e\n",
      "Updating reaction ID: EX_2425dhvitd3_e_ -> EX_2425dhvitd3_e\n",
      "Updating reaction ID: EX_24nph_e_ -> EX_24nph_e\n",
      "Updating reaction ID: EX_25hvitd2_e_ -> EX_25hvitd2_e\n",
      "Updating reaction ID: EX_25hvitd3_e_ -> EX_25hvitd3_e\n",
      "Updating reaction ID: EX_2hb_e_ -> EX_2hb_e\n",
      "Updating reaction ID: EX_2mcit_e_ -> EX_2mcit_e\n",
      "Updating reaction ID: EX_34dhoxpeg_e_ -> EX_34dhoxpeg_e\n",
      "Updating reaction ID: EX_34dhphe_e_ -> EX_34dhphe_e\n",
      "Updating reaction ID: EX_35cgmp_e_ -> EX_35cgmp_e\n",
      "Updating reaction ID: EX_3aib_D_e_ -> EX_3aib_D_e\n",
      "Updating reaction ID: EX_3aib_e_ -> EX_3aib_e\n",
      "Updating reaction ID: EX_3mlda_e_ -> EX_3mlda_e\n",
      "Updating reaction ID: EX_3mob_e_ -> EX_3mob_e\n",
      "Updating reaction ID: EX_3mop_e_ -> EX_3mop_e\n",
      "Updating reaction ID: EX_3ump_e_ -> EX_3ump_e\n",
      "Updating reaction ID: EX_4abut_e_ -> EX_4abut_e\n",
      "Updating reaction ID: EX_4abutn_e_ -> EX_4abutn_e\n",
      "Updating reaction ID: EX_4hdebrisoquine_e_ -> EX_4hdebrisoquine_e\n",
      "Updating reaction ID: EX_4hphac_e_ -> EX_4hphac_e\n",
      "Updating reaction ID: EX_4hpro_e_ -> EX_4hpro_e\n",
      "Updating reaction ID: EX_4mop_e_ -> EX_4mop_e\n",
      "Updating reaction ID: EX_4mptnl_e_ -> EX_4mptnl_e\n",
      "Updating reaction ID: EX_4mtolbutamide_e_ -> EX_4mtolbutamide_e\n",
      "Updating reaction ID: EX_4nph_e_ -> EX_4nph_e\n",
      "Updating reaction ID: EX_4nphsf_e_ -> EX_4nphsf_e\n",
      "Updating reaction ID: EX_4pyrdx_e_ -> EX_4pyrdx_e\n",
      "Updating reaction ID: EX_5adtststerone_e_ -> EX_5adtststerone_e\n",
      "Updating reaction ID: EX_5adtststeroneglc_e_ -> EX_5adtststeroneglc_e\n",
      "Updating reaction ID: EX_5adtststerones_e_ -> EX_5adtststerones_e\n",
      "Updating reaction ID: EX_5dhf_e_ -> EX_5dhf_e\n",
      "Updating reaction ID: EX_5fthf_e_ -> EX_5fthf_e\n",
      "Updating reaction ID: EX_5homeprazole_e_ -> EX_5homeprazole_e\n",
      "Updating reaction ID: EX_5htrp_e_ -> EX_5htrp_e\n",
      "Updating reaction ID: EX_5mta_e_ -> EX_5mta_e\n",
      "Updating reaction ID: EX_5mthf_e_ -> EX_5mthf_e\n",
      "Updating reaction ID: EX_5oxpro_e_ -> EX_5oxpro_e\n",
      "Updating reaction ID: EX_5thf_e_ -> EX_5thf_e\n",
      "Updating reaction ID: EX_6dhf_e_ -> EX_6dhf_e\n",
      "Updating reaction ID: EX_6htststerone_e_ -> EX_6htststerone_e\n",
      "Updating reaction ID: EX_6thf_e_ -> EX_6thf_e\n",
      "Updating reaction ID: EX_7dhf_e_ -> EX_7dhf_e\n",
      "Updating reaction ID: EX_7thf_e_ -> EX_7thf_e\n",
      "Updating reaction ID: EX_9_cis_retfa_e_ -> EX_9_cis_retfa_e\n",
      "Updating reaction ID: EX_abt_e_ -> EX_abt_e\n",
      "Updating reaction ID: EX_ac_e_ -> EX_ac_e\n",
      "Updating reaction ID: EX_acac_e_ -> EX_acac_e\n",
      "Updating reaction ID: EX_acald_e_ -> EX_acald_e\n",
      "Updating reaction ID: EX_acetone_e_ -> EX_acetone_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_acgalfucgalacgalfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acgam_e_ -> EX_acgam_e\n",
      "Updating reaction ID: EX_ach_e_ -> EX_ach_e\n",
      "Updating reaction ID: EX_acmana_e_ -> EX_acmana_e\n",
      "Updating reaction ID: EX_acn13acngalgbside_cho_e_ -> EX_acn13acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acn23acngalgbside_cho_e_ -> EX_acn23acngalgbside_cho_e\n",
      "Updating reaction ID: EX_acnacngal14acglcgalgluside_cho_e_ -> EX_acnacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_acnacngalgbside_cho_e_ -> EX_acnacngalgbside_cho_e\n",
      "Updating reaction ID: EX_acngalacglcgal14acglcgalgluside_cho_e_ -> EX_acngalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_ade_e_ -> EX_ade_e\n",
      "Updating reaction ID: EX_adn_e_ -> EX_adn_e\n",
      "Updating reaction ID: EX_adpcbl_e_ -> EX_adpcbl_e\n",
      "Updating reaction ID: EX_adprbp_e_ -> EX_adprbp_e\n",
      "Updating reaction ID: EX_adprib_e_ -> EX_adprib_e\n",
      "Updating reaction ID: EX_adrn_e_ -> EX_adrn_e\n",
      "Updating reaction ID: EX_adrnl_e_ -> EX_adrnl_e\n",
      "Updating reaction ID: EX_aflatoxin_e_ -> EX_aflatoxin_e\n",
      "Updating reaction ID: EX_ahandrostanglc_e_ -> EX_ahandrostanglc_e\n",
      "Updating reaction ID: EX_ahcys_e_ -> EX_ahcys_e\n",
      "Updating reaction ID: EX_ahdt_e_ -> EX_ahdt_e\n",
      "Updating reaction ID: EX_aicar_e_ -> EX_aicar_e\n",
      "Updating reaction ID: EX_ak2lgchol_cho_e_ -> EX_ak2lgchol_cho_e\n",
      "Updating reaction ID: EX_akg_e_ -> EX_akg_e\n",
      "Updating reaction ID: EX_ala_B_e_ -> EX_ala_B_e\n",
      "Updating reaction ID: EX_ala_D_e_ -> EX_ala_D_e\n",
      "Updating reaction ID: EX_ala_L_e_ -> EX_ala_L_e\n",
      "Updating reaction ID: EX_alaala_e_ -> EX_alaala_e\n",
      "Updating reaction ID: EX_aldstrn_e_ -> EX_aldstrn_e\n",
      "Updating reaction ID: EX_amp_e_ -> EX_amp_e\n",
      "Updating reaction ID: EX_andrstrn_e_ -> EX_andrstrn_e\n",
      "Updating reaction ID: EX_andrstrnglc_e_ -> EX_andrstrnglc_e\n",
      "Updating reaction ID: EX_anth_e_ -> EX_anth_e\n",
      "Updating reaction ID: EX_antipyrene_e_ -> EX_antipyrene_e\n",
      "Updating reaction ID: EX_apnnox_e_ -> EX_apnnox_e\n",
      "Updating reaction ID: EX_appnn_e_ -> EX_appnn_e\n",
      "Updating reaction ID: EX_aprgstrn_e_ -> EX_aprgstrn_e\n",
      "Updating reaction ID: EX_aqcobal_e_ -> EX_aqcobal_e\n",
      "Updating reaction ID: EX_arab_L_e_ -> EX_arab_L_e\n",
      "Updating reaction ID: EX_arach_e_ -> EX_arach_e\n",
      "Updating reaction ID: EX_arachcoa_e_ -> EX_arachcoa_e\n",
      "Updating reaction ID: EX_arachd_e_ -> EX_arachd_e\n",
      "Updating reaction ID: EX_arg_L_e_ -> EX_arg_L_e\n",
      "Updating reaction ID: EX_ascb_L_e_ -> EX_ascb_L_e\n",
      "Updating reaction ID: EX_asn_L_e_ -> EX_asn_L_e\n",
      "Updating reaction ID: EX_asp_D_e_ -> EX_asp_D_e\n",
      "Updating reaction ID: EX_asp_L_e_ -> EX_asp_L_e\n",
      "Updating reaction ID: EX_atp_e_ -> EX_atp_e\n",
      "Updating reaction ID: EX_avite1_e_ -> EX_avite1_e\n",
      "Updating reaction ID: EX_avite2_e_ -> EX_avite2_e\n",
      "Updating reaction ID: EX_bglc_e_ -> EX_bglc_e\n",
      "Updating reaction ID: EX_bhb_e_ -> EX_bhb_e\n",
      "Updating reaction ID: EX_bildglcur_e_ -> EX_bildglcur_e\n",
      "Updating reaction ID: EX_bilglcur_e_ -> EX_bilglcur_e\n",
      "Updating reaction ID: EX_bilirub_e_ -> EX_bilirub_e\n",
      "Updating reaction ID: EX_biocyt_e_ -> EX_biocyt_e\n",
      "Updating reaction ID: EX_btn_e_ -> EX_btn_e\n",
      "Updating reaction ID: EX_but_e_ -> EX_but_e\n",
      "Updating reaction ID: EX_bvite_e_ -> EX_bvite_e\n",
      "Updating reaction ID: EX_bz_e_ -> EX_bz_e\n",
      "Updating reaction ID: EX_xtra_e_ -> EX_xtra_e\n",
      "Updating reaction ID: EX_C02528_e_ -> EX_C02528_e\n",
      "Updating reaction ID: EX_12e8hdxt_e_ -> EX_12e8hdxt_e\n",
      "Updating reaction ID: EX_ca2_e_ -> EX_ca2_e\n",
      "Updating reaction ID: EX_camp_e_ -> EX_camp_e\n",
      "Updating reaction ID: EX_carn_e_ -> EX_carn_e\n",
      "Updating reaction ID: EX_caro_e_ -> EX_caro_e\n",
      "Updating reaction ID: EX_carveol_e_ -> EX_carveol_e\n",
      "Updating reaction ID: EX_cbasp_e_ -> EX_cbasp_e\n",
      "Updating reaction ID: EX_cca_d3_e_ -> EX_cca_d3_e\n",
      "Updating reaction ID: EX_cdp_e_ -> EX_cdp_e\n",
      "Updating reaction ID: EX_CE0074_e_ -> EX_CE0074_e\n",
      "Updating reaction ID: EX_CE1925_e_ -> EX_CE1925_e\n",
      "Updating reaction ID: EX_CE1926_e_ -> EX_CE1926_e\n",
      "Updating reaction ID: EX_CE1935_e_ -> EX_CE1935_e\n",
      "Updating reaction ID: EX_CE1936_e_ -> EX_CE1936_e\n",
      "Updating reaction ID: EX_CE1939_e_ -> EX_CE1939_e\n",
      "Updating reaction ID: EX_CE1940_e_ -> EX_CE1940_e\n",
      "Updating reaction ID: EX_CE1943_e_ -> EX_CE1943_e\n",
      "Updating reaction ID: EX_CE2011_e_ -> EX_CE2011_e\n",
      "Updating reaction ID: EX_CE2250_e_ -> EX_CE2250_e\n",
      "Updating reaction ID: EX_CE2838_e_ -> EX_CE2838_e\n",
      "Updating reaction ID: EX_CE2839_e_ -> EX_CE2839_e\n",
      "Updating reaction ID: EX_CE2915_e_ -> EX_CE2915_e\n",
      "Updating reaction ID: EX_CE2916_e_ -> EX_CE2916_e\n",
      "Updating reaction ID: EX_CE2917_e_ -> EX_CE2917_e\n",
      "Updating reaction ID: EX_CE4633_e_ -> EX_CE4633_e\n",
      "Updating reaction ID: EX_CE4722_e_ -> EX_CE4722_e\n",
      "Updating reaction ID: EX_CE4723_e_ -> EX_CE4723_e\n",
      "Updating reaction ID: EX_CE4724_e_ -> EX_CE4724_e\n",
      "Updating reaction ID: EX_CE4881_e_ -> EX_CE4881_e\n",
      "Updating reaction ID: EX_CE5786_e_ -> EX_CE5786_e\n",
      "Updating reaction ID: EX_CE5788_e_ -> EX_CE5788_e\n",
      "Updating reaction ID: EX_CE5789_e_ -> EX_CE5789_e\n",
      "Updating reaction ID: EX_CE5797_e_ -> EX_CE5797_e\n",
      "Updating reaction ID: EX_CE5798_e_ -> EX_CE5798_e\n",
      "Updating reaction ID: EX_CE5853_e_ -> EX_CE5853_e\n",
      "Updating reaction ID: EX_CE5854_e_ -> EX_CE5854_e\n",
      "Updating reaction ID: EX_CE5867_e_ -> EX_CE5867_e\n",
      "Updating reaction ID: EX_CE5868_e_ -> EX_CE5868_e\n",
      "Updating reaction ID: EX_CE5869_e_ -> EX_CE5869_e\n",
      "Updating reaction ID: EX_cgly_e_ -> EX_cgly_e\n",
      "Updating reaction ID: EX_chol_e_ -> EX_chol_e\n",
      "Updating reaction ID: EX_cholate_e_ -> EX_cholate_e\n",
      "Updating reaction ID: EX_chsterol_e_ -> EX_chsterol_e\n",
      "Updating reaction ID: EX_chtn_e_ -> EX_chtn_e\n",
      "Updating reaction ID: EX_cit_e_ -> EX_cit_e\n",
      "Updating reaction ID: EX_citr_L_e_ -> EX_citr_L_e\n",
      "Updating reaction ID: EX_cl_e_ -> EX_cl_e\n",
      "Updating reaction ID: EX_CLPND_e_ -> EX_CLPND_e\n",
      "Updating reaction ID: EX_cmp_e_ -> EX_cmp_e\n",
      "Updating reaction ID: EX_co_e_ -> EX_co_e\n",
      "Updating reaction ID: EX_co2_e_ -> EX_co2_e\n",
      "Updating reaction ID: EX_coa_e_ -> EX_coa_e\n",
      "Updating reaction ID: EX_coumarin_e_ -> EX_coumarin_e\n",
      "Updating reaction ID: EX_creat_e_ -> EX_creat_e\n",
      "Updating reaction ID: EX_crm_cho_e_ -> EX_crm_cho_e\n",
      "Updating reaction ID: EX_crmp_cho_e_ -> EX_crmp_cho_e\n",
      "Updating reaction ID: EX_crn_e_ -> EX_crn_e\n",
      "Updating reaction ID: EX_crtsl_e_ -> EX_crtsl_e\n",
      "Updating reaction ID: EX_crtstrn_e_ -> EX_crtstrn_e\n",
      "Updating reaction ID: EX_crvnc_e_ -> EX_crvnc_e\n",
      "Updating reaction ID: EX_csn_e_ -> EX_csn_e\n",
      "Updating reaction ID: EX_cspg_a_e_ -> EX_cspg_a_e\n",
      "Updating reaction ID: EX_cspg_b_e_ -> EX_cspg_b_e\n",
      "Updating reaction ID: EX_cspg_c_e_ -> EX_cspg_c_e\n",
      "Updating reaction ID: EX_cspg_d_e_ -> EX_cspg_d_e\n",
      "Updating reaction ID: EX_cspg_e_e_ -> EX_cspg_ee_\n",
      "Updating reaction ID: EX_ctp_e_ -> EX_ctp_e\n",
      "Updating reaction ID: EX_cyan_e_ -> EX_cyan_e\n",
      "Updating reaction ID: EX_cys_L_e_ -> EX_cys_L_e\n",
      "Updating reaction ID: EX_cysam_e_ -> EX_cysam_e\n",
      "Updating reaction ID: EX_cytd_e_ -> EX_cytd_e\n",
      "Updating reaction ID: EX_dad_2_e_ -> EX_dad_2_e\n",
      "Updating reaction ID: EX_dag_cho_e_ -> EX_dag_cho_e\n",
      "Updating reaction ID: EX_dca_e_ -> EX_dca_e\n",
      "Updating reaction ID: EX_dchac_e_ -> EX_dchac_e\n",
      "Updating reaction ID: EX_dcsptn1_e_ -> EX_dcsptn1_e\n",
      "Updating reaction ID: EX_dcyt_e_ -> EX_dcyt_e\n",
      "Updating reaction ID: EX_debrisoquine_e_ -> EX_debrisoquine_e\n",
      "Updating reaction ID: EX_dgchol_e_ -> EX_dgchol_e\n",
      "Updating reaction ID: EX_dgmp_e_ -> EX_dgmp_e\n",
      "Updating reaction ID: EX_dgsn_e_ -> EX_dgsn_e\n",
      "Updating reaction ID: EX_dgtp_e_ -> EX_dgtp_e\n",
      "Updating reaction ID: EX_dha_e_ -> EX_dha_e\n",
      "Updating reaction ID: EX_dhap_e_ -> EX_dhap_e\n",
      "Updating reaction ID: EX_dhdascb_e_ -> EX_dhdascb_e\n",
      "Updating reaction ID: EX_dheas_e_ -> EX_dheas_e\n",
      "Updating reaction ID: EX_dhf_e_ -> EX_dhf_e\n",
      "Updating reaction ID: EX_digalsgalside_cho_e_ -> EX_digalsgalside_cho_e\n",
      "Updating reaction ID: EX_din_e_ -> EX_din_e\n",
      "Updating reaction ID: EX_dlnlcg_e_ -> EX_dlnlcg_e\n",
      "Updating reaction ID: EX_dmantipyrine_e_ -> EX_dmantipyrine_e\n",
      "Updating reaction ID: EX_dmhptcrn_e_ -> EX_dmhptcrn_e\n",
      "Updating reaction ID: EX_dopa_e_ -> EX_dopa_e\n",
      "Updating reaction ID: EX_dopasf_e_ -> EX_dopasf_e\n",
      "Updating reaction ID: EX_dpcoa_e_ -> EX_dpcoa_e\n",
      "Updating reaction ID: EX_drib_e_ -> EX_drib_e\n",
      "Updating reaction ID: EX_dtdp_e_ -> EX_dtdp_e\n",
      "Updating reaction ID: EX_dtmp_e_ -> EX_dtmp_e\n",
      "Updating reaction ID: EX_dttp_e_ -> EX_dttp_e\n",
      "Updating reaction ID: EX_duri_e_ -> EX_duri_e\n",
      "Updating reaction ID: EX_eaflatoxin_e_ -> EX_eaflatoxin_e\n",
      "Updating reaction ID: EX_ebastine_e_ -> EX_ebastine_e\n",
      "Updating reaction ID: EX_ebastineoh_e_ -> EX_ebastineoh_e\n",
      "Updating reaction ID: EX_eicostet_e_ -> EX_eicostet_e\n",
      "Updating reaction ID: EX_elaid_e_ -> EX_elaid_e\n",
      "Updating reaction ID: EX_estradiol_e_ -> EX_estradiol_e\n",
      "Updating reaction ID: EX_estradiolglc_e_ -> EX_estradiolglc_e\n",
      "Updating reaction ID: EX_estriolglc_e_ -> EX_estriolglc_e\n",
      "Updating reaction ID: EX_estroneglc_e_ -> EX_estroneglc_e\n",
      "Updating reaction ID: EX_estrones_e_ -> EX_estrones_e\n",
      "Updating reaction ID: EX_etoh_e_ -> EX_etoh_e\n",
      "Updating reaction ID: EX_fad_e_ -> EX_fad_e\n",
      "Updating reaction ID: EX_fald_e_ -> EX_fald_e\n",
      "Updating reaction ID: EX_fe2_e_ -> EX_fe2_e\n",
      "Updating reaction ID: EX_fe3_e_ -> EX_fe3_e\n",
      "Updating reaction ID: EX_fmn_e_ -> EX_fmn_e\n",
      "Updating reaction ID: EX_fol_e_ -> EX_fol_e\n",
      "Updating reaction ID: EX_for_e_ -> EX_for_e\n",
      "Updating reaction ID: EX_fru_e_ -> EX_fru_e\n",
      "Updating reaction ID: EX_fuc_L_e_ -> EX_fuc_L_e\n",
      "Updating reaction ID: EX_fuc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fuc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fuc14galacglcgalgluside_cho_e_ -> EX_fuc14galacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacgalfucgalacglcgalgluside_cho_e_ -> EX_fucacgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngal14acglcgalgluside_cho_e_ -> EX_fucacngal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucacngalacglcgalgluside_cho_e_ -> EX_fucacngalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc12gal14acglcgalgluside_cho_e_ -> EX_fucfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfuc132galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfuc132galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglc13galacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_fucfucfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucfucgalacglcgalgluside_cho_e_ -> EX_fucfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgal14acglcgalgluside_cho_e_ -> EX_fucgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalfucgalacglcgalgluside_cho_e_ -> EX_fucgalfucgalacglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_fucgalgbside_cho_e_ -> EX_fucgalgbside_cho_e\n",
      "Updating reaction ID: EX_g1p_e_ -> EX_g1p_e\n",
      "Updating reaction ID: EX_gal_e_ -> EX_gal_e\n",
      "Updating reaction ID: EX_galacglcgalgbside_cho_e_ -> EX_galacglcgalgbside_cho_e\n",
      "Updating reaction ID: EX_galfuc12gal14acglcgalgluside_cho_e_ -> EX_galfuc12gal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galfucgalacglcgal14acglcgalgluside_cho_e_ -> EX_galfucgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e_ -> EX_galgalfucfucgalacglcgalacglcgal14acglcgalgluside_cho_e\n",
      "Updating reaction ID: EX_galgalgalthcrm_cho_e_ -> EX_galgalgalthcrm_cho_e\n",
      "Updating reaction ID: EX_galside_cho_e_ -> EX_galside_cho_e\n",
      "Updating reaction ID: EX_gam_e_ -> EX_gam_e\n",
      "Updating reaction ID: EX_gbside_cho_e_ -> EX_gbside_cho_e\n",
      "Updating reaction ID: EX_gchola_e_ -> EX_gchola_e\n",
      "Updating reaction ID: EX_gd1b2_cho_e_ -> EX_gd1b2_cho_e\n",
      "Updating reaction ID: EX_gd1c_cho_e_ -> EX_gd1c_cho_e\n",
      "Updating reaction ID: EX_gdp_e_ -> EX_gdp_e\n",
      "Updating reaction ID: EX_glc_e_ -> EX_glc_e\n",
      "Updating reaction ID: EX_glcur_e_ -> EX_glcur_e\n",
      "Updating reaction ID: EX_glgchlo_e_ -> EX_glgchlo_e\n",
      "Updating reaction ID: EX_gln_L_e_ -> EX_gln_L_e\n",
      "Updating reaction ID: EX_gltcho_e_ -> EX_gltcho_e\n",
      "Updating reaction ID: EX_gltdechol_e_ -> EX_gltdechol_e\n",
      "Updating reaction ID: EX_glu_L_e_ -> EX_glu_L_e\n",
      "Updating reaction ID: EX_gluala_e_ -> EX_gluala_e\n",
      "Updating reaction ID: EX_gly_e_ -> EX_gly_e\n",
      "Updating reaction ID: EX_glyb_e_ -> EX_glyb_e\n",
      "Updating reaction ID: EX_glyc_e_ -> EX_glyc_e\n",
      "Updating reaction ID: EX_glyc_S_e_ -> EX_glyc_S_e\n",
      "Updating reaction ID: EX_glygly_e_ -> EX_glygly_e\n",
      "Updating reaction ID: EX_glygn2_e_ -> EX_glygn2_e\n",
      "Updating reaction ID: EX_glygn4_e_ -> EX_glygn4_e\n",
      "Updating reaction ID: EX_glygn5_e_ -> EX_glygn5_e\n",
      "Updating reaction ID: EX_glyleu_e_ -> EX_glyleu_e\n",
      "Updating reaction ID: EX_glyphe_e_ -> EX_glyphe_e\n",
      "Updating reaction ID: EX_glypro_e_ -> EX_glypro_e\n",
      "Updating reaction ID: EX_glysar_e_ -> EX_glysar_e\n",
      "Updating reaction ID: EX_gmp_e_ -> EX_gmp_e\n",
      "Updating reaction ID: EX_gp1c_cho_e_ -> EX_gp1c_cho_e\n",
      "Updating reaction ID: EX_gp1calpha_cho_e_ -> EX_gp1calpha_cho_e\n",
      "Updating reaction ID: EX_gq1b_cho_e_ -> EX_gq1b_cho_e\n",
      "Updating reaction ID: EX_gq1balpha_cho_e_ -> EX_gq1balpha_cho_e\n",
      "Updating reaction ID: EX_gsn_e_ -> EX_gsn_e\n",
      "Updating reaction ID: EX_gt1a_cho_e_ -> EX_gt1a_cho_e\n",
      "Updating reaction ID: EX_gthox_e_ -> EX_gthox_e\n",
      "Updating reaction ID: EX_gthrd_e_ -> EX_gthrd_e\n",
      "Updating reaction ID: EX_gtp_e_ -> EX_gtp_e\n",
      "Updating reaction ID: EX_gua_e_ -> EX_gua_e\n",
      "Updating reaction ID: EX_gum_e_ -> EX_gum_e\n",
      "Updating reaction ID: EX_gumdchac_e_ -> EX_gumdchac_e\n",
      "Updating reaction ID: EX_gumgchol_e_ -> EX_gumgchol_e\n",
      "Updating reaction ID: EX_gumtchol_e_ -> EX_gumtchol_e\n",
      "Updating reaction ID: EX_h_e_ -> EX_h_e\n",
      "Updating reaction ID: EX_h2o_e_ -> EX_h2o_e\n",
      "Updating reaction ID: EX_h2o2_e_ -> EX_h2o2_e\n",
      "Updating reaction ID: EX_ha_e_ -> EX_ha_e\n",
      "Updating reaction ID: EX_ha_pre1_e_ -> EX_ha_pre1_e\n",
      "Updating reaction ID: EX_HC00004_e_ -> EX_HC00004_e\n",
      "Updating reaction ID: EX_HC00229_e_ -> EX_HC00229_e\n",
      "Updating reaction ID: EX_HC00250_e_ -> EX_HC00250_e\n",
      "Updating reaction ID: EX_HC00822_e_ -> EX_HC00822_e\n",
      "Updating reaction ID: EX_HC00955_e_ -> EX_HC00955_e\n",
      "Updating reaction ID: EX_HC01104_e_ -> EX_HC01104_e\n",
      "Updating reaction ID: EX_HC01361_e_ -> EX_HC01361_e\n",
      "Updating reaction ID: EX_HC01440_e_ -> EX_HC01440_e\n",
      "Updating reaction ID: EX_HC01441_e_ -> EX_HC01441_e\n",
      "Updating reaction ID: EX_HC01444_e_ -> EX_HC01444_e\n",
      "Updating reaction ID: EX_HC01446_e_ -> EX_HC01446_e\n",
      "Updating reaction ID: EX_HC01577_e_ -> EX_HC01577_e\n",
      "Updating reaction ID: EX_HC01609_e_ -> EX_HC01609_e\n",
      "Updating reaction ID: EX_HC01700_e_ -> EX_HC01700_e\n",
      "Updating reaction ID: EX_HC01787_e_ -> EX_HC01787_e\n",
      "Updating reaction ID: EX_HC02154_e_ -> EX_HC02154_e\n",
      "Updating reaction ID: EX_HC02160_e_ -> EX_HC02160_e\n",
      "Updating reaction ID: EX_HC02161_e_ -> EX_HC02161_e\n",
      "Updating reaction ID: EX_HC02172_e_ -> EX_HC02172_e\n",
      "Updating reaction ID: EX_HC02180_e_ -> EX_HC02180_e\n",
      "Updating reaction ID: EX_HC02187_e_ -> EX_HC02187_e\n",
      "Updating reaction ID: EX_HC02191_e_ -> EX_HC02191_e\n",
      "Updating reaction ID: EX_HC02192_e_ -> EX_HC02192_e\n",
      "Updating reaction ID: EX_HC02193_e_ -> EX_HC02193_e\n",
      "Updating reaction ID: EX_HC02194_e_ -> EX_HC02194_e\n",
      "Updating reaction ID: EX_HC02195_e_ -> EX_HC02195_e\n",
      "Updating reaction ID: EX_HC02196_e_ -> EX_HC02196_e\n",
      "Updating reaction ID: EX_HC02197_e_ -> EX_HC02197_e\n",
      "Updating reaction ID: EX_HC02198_e_ -> EX_HC02198_e\n",
      "Updating reaction ID: EX_HC02199_e_ -> EX_HC02199_e\n",
      "Updating reaction ID: EX_HC02200_e_ -> EX_HC02200_e\n",
      "Updating reaction ID: EX_HC02201_e_ -> EX_HC02201_e\n",
      "Updating reaction ID: EX_HC02202_e_ -> EX_HC02202_e\n",
      "Updating reaction ID: EX_HC02203_e_ -> EX_HC02203_e\n",
      "Updating reaction ID: EX_HC02204_e_ -> EX_HC02204_e\n",
      "Updating reaction ID: EX_HC02205_e_ -> EX_HC02205_e\n",
      "Updating reaction ID: EX_HC02206_e_ -> EX_HC02206_e\n",
      "Updating reaction ID: EX_HC02207_e_ -> EX_HC02207_e\n",
      "Updating reaction ID: EX_HC02208_e_ -> EX_HC02208_e\n",
      "Updating reaction ID: EX_HC02210_e_ -> EX_HC02210_e\n",
      "Updating reaction ID: EX_HC02213_e_ -> EX_HC02213_e\n",
      "Updating reaction ID: EX_HC02214_e_ -> EX_HC02214_e\n",
      "Updating reaction ID: EX_HC02216_e_ -> EX_HC02216_e\n",
      "Updating reaction ID: EX_HC02217_e_ -> EX_HC02217_e\n",
      "Updating reaction ID: EX_HC02220_e_ -> EX_HC02220_e\n",
      "Updating reaction ID: EX_hco3_e_ -> EX_hco3_e\n",
      "Updating reaction ID: EX_hcoumarin_e_ -> EX_hcoumarin_e\n",
      "Updating reaction ID: EX_hdca_e_ -> EX_hdca_e\n",
      "Updating reaction ID: EX_hdcea_e_ -> EX_hdcea_e\n",
      "Updating reaction ID: EX_hestratriol_e_ -> EX_hestratriol_e\n",
      "Updating reaction ID: EX_hexc_e_ -> EX_hexc_e\n",
      "Updating reaction ID: EX_his_L_e_ -> EX_his_L_e\n",
      "Updating reaction ID: EX_hista_e_ -> EX_hista_e\n",
      "Updating reaction ID: EX_hom_L_e_ -> EX_hom_L_e\n",
      "Updating reaction ID: EX_hpdca_e_ -> EX_hpdca_e\n",
      "Updating reaction ID: EX_chopg_e_ -> EX_chopg_e\n",
      "Updating reaction ID: EX_htaxol_e_ -> EX_htaxol_e\n",
      "Updating reaction ID: EX_hxan_e_ -> EX_hxan_e\n",
      "Updating reaction ID: EX_hyptaur_e_ -> EX_hyptaur_e\n",
      "Updating reaction ID: EX_i_e_ -> EX_i_e\n",
      "Updating reaction ID: EX_idour_e_ -> EX_idour_e\n",
      "Updating reaction ID: EX_idp_e_ -> EX_idp_e\n",
      "Updating reaction ID: EX_ile_L_e_ -> EX_ile_L_e\n",
      "Updating reaction ID: EX_imp_e_ -> EX_imp_e\n",
      "Updating reaction ID: EX_inost_e_ -> EX_inost_e\n",
      "Updating reaction ID: EX_ins_e_ -> EX_ins_e\n",
      "Updating reaction ID: EX_k_e_ -> EX_k_e\n",
      "Updating reaction ID: EX_ksi_deg1_e_ -> EX_ksi_deg1_e\n",
      "Updating reaction ID: EX_ksi_e_ -> EX_ksi_e\n",
      "Updating reaction ID: EX_ksii_core2_e_ -> EX_ksii_core2_e\n",
      "Updating reaction ID: EX_ksii_core4_e_ -> EX_ksii_core4_e\n",
      "Updating reaction ID: EX_lac_D_e_ -> EX_lac_D_e\n",
      "Updating reaction ID: EX_lac_L_e_ -> EX_lac_L_e\n",
      "Updating reaction ID: EX_lcts_e_ -> EX_lcts_e\n",
      "Updating reaction ID: EX_Lcystin_e_ -> EX_Lcystin_e\n",
      "Updating reaction ID: EX_leu_L_e_ -> EX_leu_L_e\n",
      "Updating reaction ID: EX_leugly_e_ -> EX_leugly_e\n",
      "Updating reaction ID: EX_leuktrA4_e_ -> EX_leuktrA4_e\n",
      "Updating reaction ID: EX_leuktrB4_e_ -> EX_leuktrB4_e\n",
      "Updating reaction ID: EX_leuktrC4_e_ -> EX_leuktrC4_e\n",
      "Updating reaction ID: EX_leuktrD4_e_ -> EX_leuktrD4_e\n",
      "Updating reaction ID: EX_leuktrE4_e_ -> EX_leuktrE4_e\n",
      "Updating reaction ID: EX_leuktrF4_e_ -> EX_leuktrF4_e\n",
      "Updating reaction ID: EX_leuleu_e_ -> EX_leuleu_e\n",
      "Updating reaction ID: EX_lgnc_e_ -> EX_lgnc_e\n",
      "Updating reaction ID: EX_limnen_e_ -> EX_limnen_e\n",
      "Updating reaction ID: EX_lipoate_e_ -> EX_lipoate_e\n",
      "Updating reaction ID: EX_lneldc_e_ -> EX_lneldc_e\n",
      "Updating reaction ID: EX_lnlc_e_ -> EX_lnlc_e\n",
      "Updating reaction ID: EX_lnlnca_e_ -> EX_lnlnca_e\n",
      "Updating reaction ID: EX_lnlncg_e_ -> EX_lnlncg_e\n",
      "Updating reaction ID: EX_lpchol_cho_e_ -> EX_lpchol_cho_e\n",
      "Updating reaction ID: EX_lys_L_e_ -> EX_lys_L_e\n",
      "Updating reaction ID: EX_mag_cho_e_ -> EX_mag_cho_e\n",
      "Updating reaction ID: EX_mal_L_e_ -> EX_mal_L_e\n",
      "Updating reaction ID: EX_malcoa_e_ -> EX_malcoa_e\n",
      "Updating reaction ID: EX_malt_e_ -> EX_malt_e\n",
      "Updating reaction ID: EX_malthp_e_ -> EX_malthp_e\n",
      "Updating reaction ID: EX_malthx_e_ -> EX_malthx_e\n",
      "Updating reaction ID: EX_maltpt_e_ -> EX_maltpt_e\n",
      "Updating reaction ID: EX_malttr_e_ -> EX_malttr_e\n",
      "Updating reaction ID: EX_maltttr_e_ -> EX_maltttr_e\n",
      "Updating reaction ID: EX_man_e_ -> EX_man_e\n",
      "Updating reaction ID: EX_meoh_e_ -> EX_meoh_e\n",
      "Updating reaction ID: EX_mepi_e_ -> EX_mepi_e\n",
      "Updating reaction ID: EX_mercplaccys_e_ -> EX_mercplaccys_e\n",
      "Updating reaction ID: EX_met_L_e_ -> EX_met_L_e\n",
      "Updating reaction ID: EX_mthgxl_e_ -> EX_mthgxl_e\n",
      "Updating reaction ID: EX_n2m2nmasn_e_ -> EX_n2m2nmasn_e\n",
      "Updating reaction ID: EX_na1_e_ -> EX_na1_e\n",
      "Updating reaction ID: EX_nac_e_ -> EX_nac_e\n",
      "Updating reaction ID: EX_nad_e_ -> EX_nad_e\n",
      "Updating reaction ID: EX_nadp_e_ -> EX_nadp_e\n",
      "Updating reaction ID: EX_ncam_e_ -> EX_ncam_e\n",
      "Updating reaction ID: EX_nh4_e_ -> EX_nh4_e\n",
      "Updating reaction ID: EX_nifedipine_e_ -> EX_nifedipine_e\n",
      "Updating reaction ID: EX_no_e_ -> EX_no_e\n",
      "Updating reaction ID: EX_no2_e_ -> EX_no2_e\n",
      "Updating reaction ID: EX_npthl_e_ -> EX_npthl_e\n",
      "Updating reaction ID: EX_nrpphr_e_ -> EX_nrpphr_e\n",
      "Updating reaction ID: EX_nrpphrsf_e_ -> EX_nrpphrsf_e\n",
      "Updating reaction ID: EX_nrvnc_e_ -> EX_nrvnc_e\n",
      "Updating reaction ID: EX_o2_e_ -> EX_o2_e\n",
      "Updating reaction ID: EX_o2s_e_ -> EX_o2s_e\n",
      "Updating reaction ID: EX_oagd3_cho_e_ -> EX_oagd3_cho_e\n",
      "Updating reaction ID: EX_oagt3_cho_e_ -> EX_oagt3_cho_e\n",
      "Updating reaction ID: EX_ocdca_e_ -> EX_ocdca_e\n",
      "Updating reaction ID: EX_ocdcea_e_ -> EX_ocdcea_e\n",
      "Updating reaction ID: EX_octa_e_ -> EX_octa_e\n",
      "Updating reaction ID: EX_oh1_e_ -> EX_oh1_e\n",
      "Updating reaction ID: EX_omeprazole_e_ -> EX_omeprazole_e\n",
      "Updating reaction ID: EX_onpthl_e_ -> EX_onpthl_e\n",
      "Updating reaction ID: EX_orn_e_ -> EX_orn_e\n",
      "Updating reaction ID: EX_orot_e_ -> EX_orot_e\n",
      "Updating reaction ID: EX_oxa_e_ -> EX_oxa_e\n",
      "Updating reaction ID: EX_paf_cho_e_ -> EX_paf_cho_e\n",
      "Updating reaction ID: EX_pan4p_e_ -> EX_pan4p_e\n",
      "Updating reaction ID: EX_pchol_cho_e_ -> EX_pchol_cho_e\n",
      "Updating reaction ID: EX_pe_cho_e_ -> EX_pe_cho_e\n",
      "Updating reaction ID: EX_pect_e_ -> EX_pect_e\n",
      "Updating reaction ID: EX_pectindchac_e_ -> EX_pectindchac_e\n",
      "Updating reaction ID: EX_pectingchol_e_ -> EX_pectingchol_e\n",
      "Updating reaction ID: EX_pectintchol_e_ -> EX_pectintchol_e\n",
      "Updating reaction ID: EX_peplys_e_ -> EX_peplys_e\n",
      "Updating reaction ID: EX_perillyl_e_ -> EX_perillyl_e\n",
      "Updating reaction ID: EX_pglyc_cho_e_ -> EX_pglyc_cho_e\n",
      "Updating reaction ID: EX_phe_L_e_ -> EX_phe_L_e\n",
      "Updating reaction ID: EX_pheacgln_e_ -> EX_pheacgln_e\n",
      "Updating reaction ID: EX_pheme_e_ -> EX_pheme_e\n",
      "Updating reaction ID: EX_phyQ_e_ -> EX_phyQ_e\n",
      "Updating reaction ID: EX_phyt_e_ -> EX_phyt_e\n",
      "Updating reaction ID: EX_pi_e_ -> EX_pi_e\n",
      "Updating reaction ID: EX_pnto_R_e_ -> EX_pnto_R_e\n",
      "Updating reaction ID: EX_ppa_e_ -> EX_ppa_e\n",
      "Updating reaction ID: EX_ppi_e_ -> EX_ppi_e\n",
      "Updating reaction ID: EX_prgstrn_e_ -> EX_prgstrn_e\n",
      "Updating reaction ID: EX_pro_D_e_ -> EX_pro_D_e\n",
      "Updating reaction ID: EX_pro_L_e_ -> EX_pro_L_e\n",
      "Updating reaction ID: EX_progly_e_ -> EX_progly_e\n",
      "Updating reaction ID: EX_prostgd2_e_ -> EX_prostgd2_e\n",
      "Updating reaction ID: EX_prostge1_e_ -> EX_prostge1_e\n",
      "Updating reaction ID: EX_prostge2_e_ -> EX_prostge2_e\n",
      "Updating reaction ID: EX_prostgf2_e_ -> EX_prostgf2_e\n",
      "Updating reaction ID: EX_prostgh2_e_ -> EX_prostgh2_e\n",
      "Updating reaction ID: EX_prostgi2_e_ -> EX_prostgi2_e\n",
      "Updating reaction ID: EX_prpp_e_ -> EX_prpp_e\n",
      "Updating reaction ID: EX_ps_cho_e_ -> EX_ps_cho_e\n",
      "Updating reaction ID: EX_psyl_e_ -> EX_psyl_e\n",
      "Updating reaction ID: EX_psylchol_e_ -> EX_psylchol_e\n",
      "Updating reaction ID: EX_psyltchol_e_ -> EX_psyltchol_e\n",
      "Updating reaction ID: EX_psyltdechol_e_ -> EX_psyltdechol_e\n",
      "Updating reaction ID: EX_ptdca_e_ -> EX_ptdca_e\n",
      "Updating reaction ID: EX_ptrc_e_ -> EX_ptrc_e\n",
      "Updating reaction ID: EX_ptth_e_ -> EX_ptth_e\n",
      "Updating reaction ID: EX_pydam_e_ -> EX_pydam_e\n",
      "Updating reaction ID: EX_pydx_e_ -> EX_pydx_e\n",
      "Updating reaction ID: EX_pydx5p_e_ -> EX_pydx5p_e\n",
      "Updating reaction ID: EX_pydxn_e_ -> EX_pydxn_e\n",
      "Updating reaction ID: EX_pyr_e_ -> EX_pyr_e\n",
      "Updating reaction ID: EX_q10_e_ -> EX_q10_e\n",
      "Updating reaction ID: EX_q10h2_e_ -> EX_q10h2_e\n",
      "Updating reaction ID: EX_rbt_e_ -> EX_rbt_e\n",
      "Updating reaction ID: EX_retfa_e_ -> EX_retfa_e\n",
      "Updating reaction ID: EX_retinol_9_cis_e_ -> EX_retinol_9_cis_e\n",
      "Updating reaction ID: EX_retinol_cis_11_e_ -> EX_retinol_cis_11_e\n",
      "Updating reaction ID: EX_retinol_e_ -> EX_retinol_e\n",
      "Updating reaction ID: EX_retn_e_ -> EX_retn_e\n",
      "Updating reaction ID: EX_retnglc_e_ -> EX_retnglc_e\n",
      "Updating reaction ID: EX_rib_D_e_ -> EX_rib_D_e\n",
      "Updating reaction ID: EX_ribflv_e_ -> EX_ribflv_e\n",
      "Updating reaction ID: EX_Rtotal_e_ -> EX_Rtotal_e\n",
      "Updating reaction ID: EX_Rtotal2_e_ -> EX_Rtotal2_e\n",
      "Updating reaction ID: EX_Rtotal3_e_ -> EX_Rtotal3_e\n",
      "Updating reaction ID: EX_s2l2fn2m2masn_e_ -> EX_s2l2fn2m2masn_e\n",
      "Updating reaction ID: EX_s2l2n2m2masn_e_ -> EX_s2l2n2m2masn_e\n",
      "Updating reaction ID: EX_sarcs_e_ -> EX_sarcs_e\n",
      "Updating reaction ID: EX_sbt_d_e_ -> EX_sbt_d_e\n",
      "Updating reaction ID: EX_sel_e_ -> EX_sel_e\n",
      "Updating reaction ID: EX_ser_D_e_ -> EX_ser_D_e\n",
      "Updating reaction ID: EX_ser_L_e_ -> EX_ser_L_e\n",
      "Updating reaction ID: EX_sfcys_e_ -> EX_sfcys_e\n",
      "Updating reaction ID: EX_sl_L_e_ -> EX_sl_L_e\n",
      "Updating reaction ID: EX_so4_e_ -> EX_so4_e\n",
      "Updating reaction ID: EX_spc_cho_e_ -> EX_spc_cho_e\n",
      "Updating reaction ID: EX_sph1p_e_ -> EX_sph1p_e\n",
      "Updating reaction ID: EX_sphs1p_e_ -> EX_sphs1p_e\n",
      "Updating reaction ID: EX_spmd_e_ -> EX_spmd_e\n",
      "Updating reaction ID: EX_sprm_e_ -> EX_sprm_e\n",
      "Updating reaction ID: EX_srtn_e_ -> EX_srtn_e\n",
      "Updating reaction ID: EX_strch1_e_ -> EX_strch1_e\n",
      "Updating reaction ID: EX_strch2_e_ -> EX_strch2_e\n",
      "Updating reaction ID: EX_strdnc_e_ -> EX_strdnc_e\n",
      "Updating reaction ID: EX_succ_e_ -> EX_succ_e\n",
      "Updating reaction ID: EX_sucr_e_ -> EX_sucr_e\n",
      "Updating reaction ID: EX_tag_cho_e_ -> EX_tag_cho_e\n",
      "Updating reaction ID: EX_tagat_D_e_ -> EX_tagat_D_e\n",
      "Updating reaction ID: EX_taur_e_ -> EX_taur_e\n",
      "Updating reaction ID: EX_taxol_e_ -> EX_taxol_e\n",
      "Updating reaction ID: EX_tchola_e_ -> EX_tchola_e\n",
      "Updating reaction ID: EX_tcynt_e_ -> EX_tcynt_e\n",
      "Updating reaction ID: EX_tdchola_e_ -> EX_tdchola_e\n",
      "Updating reaction ID: EX_tdechola_e_ -> EX_tdechola_e\n",
      "Updating reaction ID: EX_tethex3_e_ -> EX_tethex3_e\n",
      "Updating reaction ID: EX_tetpent3_e_ -> EX_tetpent3_e\n",
      "Updating reaction ID: EX_tetpent6_e_ -> EX_tetpent6_e\n",
      "Updating reaction ID: EX_tettet6_e_ -> EX_tettet6_e\n",
      "Updating reaction ID: EX_thf_e_ -> EX_thf_e\n",
      "Updating reaction ID: EX_thm_e_ -> EX_thm_e\n",
      "Updating reaction ID: EX_thmmp_e_ -> EX_thmmp_e\n",
      "Updating reaction ID: EX_thmtp_e_ -> EX_thmtp_e\n",
      "Updating reaction ID: EX_thr_L_e_ -> EX_thr_L_e\n",
      "Updating reaction ID: EX_thym_e_ -> EX_thym_e\n",
      "Updating reaction ID: EX_thymd_e_ -> EX_thymd_e\n",
      "Updating reaction ID: EX_thyox_L_e_ -> EX_thyox_L_e\n",
      "Updating reaction ID: EX_tmndnc_e_ -> EX_tmndnc_e\n",
      "Updating reaction ID: EX_tolbutamide_e_ -> EX_tolbutamide_e\n",
      "Updating reaction ID: EX_tre_e_ -> EX_tre_e\n",
      "Updating reaction ID: EX_triodthy_e_ -> EX_triodthy_e\n",
      "Updating reaction ID: EX_triodthysuf_e_ -> EX_triodthysuf_e\n",
      "Updating reaction ID: EX_trp_L_e_ -> EX_trp_L_e\n",
      "Updating reaction ID: EX_tststerone_e_ -> EX_tststerone_e\n",
      "Updating reaction ID: EX_tststeroneglc_e_ -> EX_tststeroneglc_e\n",
      "Updating reaction ID: EX_tststerones_e_ -> EX_tststerones_e\n",
      "Updating reaction ID: EX_tsul_e_ -> EX_tsul_e\n",
      "Updating reaction ID: EX_ttdca_e_ -> EX_ttdca_e\n",
      "Updating reaction ID: EX_txa2_e_ -> EX_txa2_e\n",
      "Updating reaction ID: EX_tymsf_e_ -> EX_tymsf_e\n",
      "Updating reaction ID: EX_Tyr_ggn_e_ -> EX_Tyr_ggn_e\n",
      "Updating reaction ID: EX_tyr_L_e_ -> EX_tyr_L_e\n",
      "Updating reaction ID: EX_udp_e_ -> EX_udp_e\n",
      "Updating reaction ID: EX_udpg_e_ -> EX_udpg_e\n",
      "Updating reaction ID: EX_udpgal_e_ -> EX_udpgal_e\n",
      "Updating reaction ID: EX_ump_e_ -> EX_ump_e\n",
      "Updating reaction ID: EX_ura_e_ -> EX_ura_e\n",
      "Updating reaction ID: EX_urate_e_ -> EX_urate_e\n",
      "Updating reaction ID: EX_urea_e_ -> EX_urea_e\n",
      "Updating reaction ID: EX_uri_e_ -> EX_uri_e\n",
      "Updating reaction ID: EX_utp_e_ -> EX_utp_e\n",
      "Updating reaction ID: EX_vacc_e_ -> EX_vacc_e\n",
      "Updating reaction ID: EX_val_L_e_ -> EX_val_L_e\n",
      "Updating reaction ID: EX_vitd2_e_ -> EX_vitd2_e\n",
      "Updating reaction ID: EX_vitd3_e_ -> EX_vitd3_e\n",
      "Updating reaction ID: EX_whddca_e_ -> EX_whddca_e\n",
      "Updating reaction ID: EX_whhdca_e_ -> EX_whhdca_e\n",
      "Updating reaction ID: EX_whtststerone_e_ -> EX_whtststerone_e\n",
      "Updating reaction ID: EX_whttdca_e_ -> EX_whttdca_e\n",
      "Updating reaction ID: EX_xolest_cho_e_ -> EX_xolest_cho_e\n",
      "Updating reaction ID: EX_xolest2_cho_e_ -> EX_xolest2_cho_e\n",
      "Updating reaction ID: EX_xoltri24_e_ -> EX_xoltri24_e\n",
      "Updating reaction ID: EX_xoltri25_e_ -> EX_xoltri25_e\n",
      "Updating reaction ID: EX_xoltri27_e_ -> EX_xoltri27_e\n",
      "Updating reaction ID: EX_xyl_D_e_ -> EX_xyl_D_e\n",
      "Updating reaction ID: EX_xylt_e_ -> EX_xylt_e\n",
      "Updating reaction ID: EX_yvite_e_ -> EX_yvite_e\n"
     ]
    }
   ],
   "source": [
    "iCHO2291_changed = update_reaction_ids_specific(iCHO2291)\n",
    "iCHO1766_changed = update_reaction_ids_specific(iCHO1766)\n",
    "iCHO2101_changed = update_reaction_ids_specific(iCHO2101)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab66e92e",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\"\"\"\"\"\"\n",
    "Temprary code\n",
    "\"\"\"\"\"\"\"\"\"\n",
    "iCHO2101_changed_tmp = iCHO2101_changed.copy()\n",
    "iCHO2101_changed_tmp.objective = 'biomass_cho'\n",
    "\n",
    "temp_dict = temp_dict_zela\n",
    "time = 'P2 to P4'\n",
    "solver = 'gurobi'\n",
    "\n",
    "iCHO2101_changed_tmp.solver = solver\n",
    "for rxn in iCHO2101_changed_tmp.boundary:\n",
    "    # Models that are forced to secrete ethanol are not feasible\n",
    "    if rxn.id == 'EX_etoh_e':\n",
    "        rxn.bounds = (-0.1,0.1)\n",
    "        continue\n",
    "\n",
    "    # Replace the lower and upper bound with experimental data\n",
    "    if rxn.id in temp_dict.keys():\n",
    "        rxn.bounds = temp_dict[rxn.id][time]\n",
    "        continue\n",
    "\n",
    "    # Keep boundaries open for essential metabolites\n",
    "    if rxn.id == 'EX_h2o_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_h_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_o2_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_hco3_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_so4_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "    if rxn.id == 'EX_pi_e':\n",
    "        rxn.bounds = (-1000,1000)\n",
    "        continue\n",
    "\n",
    "    # Boundaries from Sink reactions on iCHO_v1\n",
    "    if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "        rxn.bounds = (-0.1,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Tyr_ggn_c':\n",
    "        rxn.bounds = (-0.1,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_Ser_Thr_g':\n",
    "        rxn.bounds = (-0.1,1000)\n",
    "        continue\n",
    "    if rxn.id == 'SK_pre_prot_r':\n",
    "        rxn.bounds = (-0.1,1000)\n",
    "        continue\n",
    "\n",
    "#     # Close uptake rates for the rest of the boundaries\n",
    "#     if rxn.id.startswith(\"EX_\"):\n",
    "#         rxn.bounds = (0,1000) \n",
    "#     if rxn.id.startswith(\"SK_\"):\n",
    "#         rxn.bounds = (0,1000)\n",
    "#     if rxn.id.startswith(\"DM_\"):\n",
    "#         rxn.bounds = (0,1000)\n",
    "\n",
    "pfba_solution = cobra.flux_analysis.pfba(iCHO2101_changed_tmp)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "tmp_genes_ess = find_essential_genes(iCHO2101_changed_tmp)\n",
    "# with GM_constrained_2291 as model:\n",
    "#     objective = 'biomass_cho'\n",
    "#     pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "#     pfba_obj = pfba_solution.fluxes[objective]\n",
    "# #     tmp = model.reactions.get_by_id(objective)\n",
    "#     print(pfba_obj)\n",
    "#     make_GM = True\n",
    "#     GM = model.copy()  # Copy the model to avoid modifying the original\n",
    "#     temp_dict = temp_dict_zela\n",
    "#     time = 'P4 to P6'\n",
    "#     solver = 'gurobi'\n",
    "\n",
    "#     # Set the solver and the objective reaction\n",
    "#     GM.solver = solver\n",
    "#     GM.objective = objective\n",
    "\n",
    "#     for rxn in GM.boundary:\n",
    "#         # Models that are forced to secrete ethanol are not feasible\n",
    "#         if rxn.id == 'EX_etoh_e':\n",
    "#             rxn.bounds = (-0.1,0.1)\n",
    "#             continue\n",
    "\n",
    "#         # Replace the lower and upper bound with experimental data\n",
    "#         if rxn.id in temp_dict.keys():\n",
    "#             rxn.bounds = temp_dict[rxn.id][time]\n",
    "#             continue\n",
    "\n",
    "#         # Keep boundaries open for essential metabolites\n",
    "#         if rxn.id == 'EX_h2o_e':\n",
    "#             rxn.bounds = (-1000,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'EX_h_e':\n",
    "#             rxn.bounds = (-1000,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'EX_o2_e':\n",
    "#             rxn.bounds = (-1000,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'EX_hco3_e':\n",
    "#             rxn.bounds = (-1000,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'EX_so4_e':\n",
    "#             rxn.bounds = (-1000,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'EX_pi_e':\n",
    "#             rxn.bounds = (-1000,1000)\n",
    "#             continue\n",
    "\n",
    "#         # Boundaries from Sink reactions on iCHO_v1\n",
    "#         if rxn.id == 'SK_Asn_X_Ser_Thr_r':\n",
    "#             rxn.bounds = (-0.1,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'SK_Tyr_ggn_c':\n",
    "#             rxn.bounds = (-0.1,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'SK_Ser_Thr_g':\n",
    "#             rxn.bounds = (-0.1,1000)\n",
    "#             continue\n",
    "#         if rxn.id == 'SK_pre_prot_r':\n",
    "#             rxn.bounds = (-0.1,1000)\n",
    "#             continue\n",
    "\n",
    "#         # Close uptake rates for the rest of the boundaries\n",
    "#         if rxn.id.startswith(\"EX_\"):\n",
    "#             rxn.bounds = (0,1000) \n",
    "#         if rxn.id.startswith(\"SK_\"):\n",
    "#             rxn.bounds = (0,1000)\n",
    "#         if rxn.id.startswith(\"DM_\"):\n",
    "#             rxn.bounds = (0,1000)\n",
    "#     model_exp_constrained = GM \n",
    "    \n",
    "#     if make_GM == True:\n",
    "#         for rxn in GM.boundary:\n",
    "#             if rxn.id in temp_dict.keys():\n",
    "#                 rxn.bounds = (-1000,1000)\n",
    "#                 print(rxn)\n",
    "#                 continue\n",
    "#         pfba_solution_GM = cobra.flux_analysis.pfba(GM)\n",
    "#         pfba_obj_GM = pfba_solution_GM.fluxes[objective] \n",
    "#         print(f\"Optimized objective w/ opened bounds (growth rate): {pfba_obj_GM}\")\n",
    "            \n",
    "#     # Perform pFBA (parsimonious FBA) to minimize the total flux\n",
    "#     pfba_solution = cobra.flux_analysis.pfba(model_exp_constrained)\n",
    "#     pfba_obj = pfba_solution.fluxes[objective]\n",
    "\n",
    "#     # Print and return the objective value and the full pFBA solution\n",
    "#     print(f\"Optimized objective w/ exp constraints (growth rate): {pfba_obj}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2c4e13e3-e66f-44a4-8aa3-f6ddffd7b16b",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_time = 'P2 to P4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "326f17cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp5vzyxbb6.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Optimized objective w/ exp constraints (growth rate): 0.038931856248325444\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpuep1vtsk.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp5c6mdixi.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpa9lrovsg.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmphpqeg_38.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp0qbuzj0e.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp2wl8lnqw.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpkioq4i_s.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4456 rows, 13326 columns, 55296 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100756786                                 Umps\n",
      "1        100756786                                 Umps\n",
      "2        100770439                                 Guk1\n",
      "3        100767083                                Adss2\n",
      "4        100760149                                 Ppat\n",
      "5        100689449                                 Pgs1\n",
      "6        100754883                                 Fdps\n",
      "7        100767580                              Hsd17b7\n",
      "8        100756632                                Dhodh\n",
      "9        100774578                                 Kdsr\n",
      "10       100689232                                  Cad\n",
      "11       100762189                         LOC100762189\n",
      "12       100754883                                 Fdps\n",
      "13       100750969                                 Ahcy\n",
      "14       100752327                                Mat2a\n",
      "15       100689446                                 Gart\n",
      "16       100758991                                Paics\n",
      "17       100760956                                 Pfas\n",
      "18       100762904                                Acaca\n",
      "19       100750895                                 Atic\n",
      "20       100689398                               Impdh2\n",
      "21       100689446                                 Gart\n",
      "22       100752366                                 Idi1\n",
      "23       100689093                                 Pisd\n",
      "24       100689060                                 Adsl\n",
      "25       100753498                                  Lss\n",
      "26       100689060                                 Adsl\n",
      "27       100750895                                 Atic\n",
      "28       100758991                                Paics\n",
      "29       100761591                                 Rrm1\n",
      "30       100752542                                 Rrm2\n",
      "31       100770431                              Atp5f1a\n",
      "32       100760824                         LOC100760824\n",
      "33       100774535                               Atp5pb\n",
      "34       103160361                               Atp5mf\n",
      "35       100763024                              Atp5f1b\n",
      "36       100689101                                Ctps1\n",
      "37       100763416                                  Mtr\n",
      "38       100763175                         LOC100763175\n",
      "39       100766805                         LOC100766805\n",
      "40       100759002                         LOC100759002\n",
      "41       100766032                         LOC100766032\n",
      "42       100689232                                  Cad\n",
      "43       100689192                                Fdft1\n",
      "44       100764834                                 Sqle\n",
      "45       100689415                               Sptlc2\n",
      "46       100689326                               Sptlc1\n",
      "47       100766719                                Sgms1\n",
      "48       100754276                                 Tyms\n",
      "49       100754784                                Msmo1\n",
      "50       100771594                               Txnrd1\n",
      "51       100750801                                Cdipt\n",
      "52       100689232                                  Cad\n",
      "53       100689325                                 Ugp2\n",
      "54       100689446                                 Gart\n",
      "Total matched metabolic genes: 229\n",
      "Matched essential genes: 55/229\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_1766, _, _, _, GM_unconstrained_1766 = constraint_model_feg(iCHO1766_changed, time=exp_time, exp_dict=temp_dict_wt, objective = \"biomass_cho\")\n",
    "matched_df_1766,len_1766,len_1766_ess,len_1766_ess_GM, icho1766_essential_reactions = find_and_match_essential_genes(GM_constrained_1766, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ca244db9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpewv6av7e.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Optimized objective w/ exp constraints (growth rate): 0.00038931713774040337\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpm7cqp4d6.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp0gc5enj9.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpzuz30eq1.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpb8regbts.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpycw0w0vu.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpr8awn7_g.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp9lpeid6f.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 4527 rows, 14872 columns, 58208 nonzeros\n",
      "  gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0       100689449                                 Pgs1\n",
      "1       100766719                                Sgms1\n",
      "2       100770439                                 Guk1\n",
      "3       100754276                                 Tyms\n",
      "4       100689398                               Impdh2\n",
      "Total matched metabolic genes: 254\n",
      "Matched essential genes: 5/254\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_2101, _, _, _, GM_unconstrained_2101 = constraint_model_feg(iCHO2101_changed, time=exp_time, exp_dict=temp_dict_wt, objective = \"biomass_cho\")\n",
    "matched_df_2101,len_2101,len_2101_ess, len_2101_ess_GM, icho2101_essential_reactions = find_and_match_essential_genes(GM_constrained_2101, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e95984f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpqpgaazoz.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Optimized objective w/ exp constraints (growth rate): 0.00038931713774040337\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp1ygwqr60.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpgle14mqf.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpv8n0x5k4.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpo11_efi9.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp6sdom61f.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp9kx94948.lp\n",
      "Reading time = 0.08 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpqtxap8x4.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 3972 rows, 12472 columns, 49472 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100771594                               Txnrd1\n",
      "1        100689192                                Fdft1\n",
      "2        100750895                                 Atic\n",
      "3        100764834                                 Sqle\n",
      "4        100758991                                Paics\n",
      "..             ...               ...                ...\n",
      "88       100750895                                 Atic\n",
      "89       100689398                               Impdh2\n",
      "90       100763640                               Ptpmt1\n",
      "91       100689449                                 Pgs1\n",
      "92       100752366                                 Idi1\n",
      "\n",
      "[93 rows x 3 columns]\n",
      "Total matched metabolic genes: 340\n",
      "Matched essential genes: 93/340\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_2291, _, _, _, GM_unconstrained_2291 = constraint_model_feg(iCHO2291_changed, time=exp_time, exp_dict=temp_dict_wt, objective = \"biomass_cho\")\n",
    "matched_df_2291,len_2291,len_2291_ess, len_2291_ess_GM, icho2291_essential_reactions = find_and_match_essential_genes(GM_constrained_2291, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "804e2e8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp3oe2o0w_.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Optimized objective w/ exp constraints (growth rate): 0.046714678424891076\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpzsd3wiwz.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpq1v35trm.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpwwrb_ikm.lp\n",
      "Reading time = 0.09 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpnqlp6zju.lp\n",
      "Reading time = 0.10 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpjlrvbsvp.lp\n",
      "Reading time = 0.12 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpx0izhg34.lp\n",
      "Reading time = 0.12 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpxl28eota.lp\n",
      "Reading time = 0.12 seconds\n",
      ": 7377 rows, 22008 columns, 93108 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100764834              Sqle               Sqle\n",
      "1        100689446              Gart               Gart\n",
      "2        100689326            Sptlc1             Sptlc1\n",
      "3        100689415            Sptlc2             Sptlc2\n",
      "4        100760149              Ppat               Ppat\n",
      "..             ...               ...                ...\n",
      "69       100768213             Eif3m              Eif3m\n",
      "70       100753319             Eif3i              Eif3i\n",
      "71       100757133            Eif4a3             Eif4a3\n",
      "72       100752118             Eif3a              Eif3a\n",
      "73       100751998             Eif3h              Eif3h\n",
      "\n",
      "[74 rows x 3 columns]\n",
      "Total matched metabolic genes: 443\n",
      "Matched essential genes: 74/443\n"
     ]
    }
   ],
   "source": [
    "GM_constrained_3K, _, _, _, GM_unconstrained_3K = constraint_model_feg(iCHO3K, time=exp_time, exp_dict=temp_dict_wt, objective = \"biomass_cho\")\n",
    "matched_df_3K,len_3k,len_3k_ess, len_3k_ess_GM, icho3K_essential_reactions = find_and_match_essential_genes(GM_constrained_3K, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "01856159",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp55eu8bof.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Optimized objective w/ exp constraints (growth rate): 0.04671467842489109\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp18g85fnx.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp9zzd77x5.lp\n",
      "Reading time = 0.07 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpajxf44z_.lp\n",
      "Reading time = 0.09 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmpz6mfjhpq.lp\n",
      "Reading time = 0.13 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp2xq5u0_9.lp\n",
      "Reading time = 0.13 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp6xw62_gk.lp\n",
      "Reading time = 0.12 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n",
      "Read LP format model from file /var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/tmp97tqtmre.lp\n",
      "Reading time = 0.12 seconds\n",
      ": 7377 rows, 21908 columns, 92522 nonzeros\n",
      "   gene_exp_ess.id gene_exp_ess.name entrez_gene_symbol\n",
      "0        100758991             Paics              Paics\n",
      "1        100689326            Sptlc1             Sptlc1\n",
      "2        100689415            Sptlc2             Sptlc2\n",
      "3        100760956              Pfas               Pfas\n",
      "4        100689398            Impdh2             Impdh2\n",
      "..             ...               ...                ...\n",
      "73       100689060              Adsl               Adsl\n",
      "74       100750895              Atic               Atic\n",
      "75       100689446              Gart               Gart\n",
      "76       100773070             Cept1              Cept1\n",
      "77       100689446              Gart               Gart\n",
      "\n",
      "[78 rows x 3 columns]\n",
      "Total matched metabolic genes: 443\n",
      "Matched essential genes: 78/443\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Dong-Hyuk's modification -  To examine the differences before and after removing ATP and glucose loop reactions from the iCHO3K reconstruction\n",
    "\"\"\"\n",
    "\n",
    "with iCHO3K as model:\n",
    "    objective = 'biomass_cho_s'\n",
    "    model.objective = objective\n",
    "    \n",
    "    filtered_atp_loop_reactions = [\n",
    "    'SCP22x','TMNDNCCOAtx','OCCOAtx','r0391','BiGGRxn67','r2247','r2246','r2245','r2317','HMR_0293',\n",
    "    'HMR_7741','r1453','r1393','NICRNS','GapFill-R08726','RE2915M','HMR_3288','HMR_1325','RE2439C',\n",
    "     'r1450','RE3477C','AAPSAS','r0698','3HDH260p','HMR_3272','ACOAD183n3m','GapFill-R01463','r1468',\n",
    "     'r0655','r0603','r0541','HMR_1329','GapFill-R03599','AKGDm','r0556','OIVD1m','OIVD3m','2OXOADOXm',\n",
    "     'r0386','r0451','GLYCLm','MACACI'\n",
    "    ]\n",
    "    \n",
    "    glucloop_reactions = [\n",
    "    'GapFill-R01206', 'GAUGE-R00557', \n",
    "    'GAUGE-R00558', 'FNOR', 'GGH', 'r0741', 'r1479', 'XOLESTPOOL'\n",
    "    ]\n",
    "\n",
    "    model.remove_reactions(glucloop_reactions, remove_orphans=True)\n",
    "    model.remove_reactions(filtered_atp_loop_reactions, remove_orphans=True)\n",
    "\n",
    "\n",
    "    GM_constrained_3K, _, _, _, GM_unconstrained_3K = constraint_model_feg(model, time=exp_time, exp_dict=temp_dict_wt, objective = \"biomass_cho\")\n",
    "    matched_df_3K,len_3k,len_3k_ess, len_3k_ess_GM, icho3K_essential_reactions_2 = find_and_match_essential_genes(GM_constrained_3K, df_loaded, save_csv=False, print_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0e562282",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "# Data for the bar chart (len_ and len_ess for each model)\n",
    "data = {\n",
    "    'iCHO1766': [len_1766, len_1766_ess],\n",
    "    'iCHO2101': [len_2101, len_2101_ess],\n",
    "    'iCHO2291': [len_2291, len_2291_ess],\n",
    "    'iCHO3K': [len_3k, len_3k_ess]\n",
    "}\n",
    "\n",
    "# Create a DataFrame\n",
    "counts_df = pd.DataFrame(data, index=['Matched Genes', 'Matched Essential Genes'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7bc2be12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlcAAADaCAYAAACLimerAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAAA7HElEQVR4nO3deVyU1f4H8A8IjmyDiCggihJuIOIComJCGpKkJIqJC+5CZq4YZSKgpaalZWr+DCWXm3nNfc8KcLmaIimKek2u69VCpEFAZD+/P7wzOc6ADAwMy+f9es3rheec55nvGcbDd57nnDN6QggBIiIiItIKfV0HQERERFSXMLkiIiIi0iImV0RERERaxOSKiIiISIuYXBERERFpEZMrIiIiIi1ickVERESkRUyuiIiIiLSIyRURERGRFjG5IiV6enrQ09PTyXNv2rQJenp6GD9+vE6en4j+xrGgfkpISICenh68vb11HUqtxuRKy+QDkp6eHi5fvlxm2zfeeEPR9vbt29UTYB2VmZmJTz/9FF5eXmjWrBkaNmyIxo0bo02bNnj99dfx/vvv4/r167oOk+oRjgXVy9vbW+k1f/ExZMgQXYdY5eSvQUJCgk6e/+nTp1i7di18fX1hY2MDiUQCqVQKe3t79O3bFzNnzsS5c+d0Elt1M9B1AHXZP/7xDyxbtkxt3R9//IGff/5Za8+Vl5eHp0+fwsLCQmvnrC1+/PFHjBw5EjKZDG3atMGwYcNgY2OD7OxsXLt2DadPn8Yvv/wCZ2dntG/fXtfhUj3EsaD6BAYGonv37irl7dq100E02vfkyRMUFxdDKpWq1IWEhOCNN96Ag4NDtceVlJSEgIAA3Lt3D9bW1vDz80OrVq2Ql5eH69ev4/Tp0zh58iQsLCzQo0ePao+v2gnSKgBCX19fABB2dnaiuLhYbbvly5crtb1161aFn3PLli0CgBg3blyFzyEHQOjqbfHtt99q3I9Tp04JAwMDAUAsXrxYFBUVqbQpLi4Wp0+fFlevXtVitERl41hQcRUZC7y8vAQA8e2331ZZXLr28ccfCwAiKiqqyp4jPj5eABBeXl7lPub3338XUqlUABDvvfeeyM3NVdvu4sWLIjExUUuR1my8clUFWrZsCQMDA/znP//B8ePH8dprr6m02bJlC1555RUUFRXhzp07lXq+4uLiSh1fW5WUlGDChAkoKipCWFgYPvroI7Xt9PX10atXr2qOjohjAWlXUVGRrkNQ65133kFWVhaGDRuG1atXl9rO1dW1GqPSLc65qiKjR48G8Ox2wIuSkpKQkpKCcePGqT22pKQEW7ZswfDhw+Ho6AhjY2NYW1vjzTffxKVLl5Taent7Y8KECQCAzZs3K+YXREdHK7VLT0/Hhx9+iE6dOsHExASNGjVCy5Yty5yHcPDgQfTt2xdmZmZo2bIlPvjgAxQUFKhtGxsbCzc3NxgbG8PCwgJDhgzBtWvX1LZNT0/H1KlTYWtri0aNGsHZ2RmrV6+GEKLUWNQ5evQobty4ASMjI0RERGh0rFxKSgqGDRsGS0tLNGrUCK6urti8ebNKu9u3b0NPTw+tW7cG8OwPopubG0xMTODg4IDPP/9cbfxFRUVYsWIFnJ2d0ahRI1hZWSE4OBj3799XaVtYWIgVK1bAxcUFJiYmMDExQceOHRETE1OhvlHNwLGg6seCiti7dy9effVVxf99e3t7TJ06VaVddnY2IiMj0b59exgZGUEqlaJz5844ePCg2vOWt//R0dGK349MJsOsWbPQsmVLmJqawtPTU2VuUuvWrbFw4UIAwMKFCxW/302bNinalDbnKj4+Hu+88w46deoEc3NzNG7cGB4eHtizZ4+Gr5qqq1evIi4uDgCwePHiCp3jzp07GD9+PKytrSGRSNChQ4dSx9TnF1pU1ftS3tbc3BxGRkZwdHTExx9/rFmndHvhrO4BIOzt7cXvv/8uAAhzc3Px9OlTpTYzZswQenp64tatW8Le3l7lVkB2draQSCTCx8dHzJgxQ0RHR4uhQ4cKAMLW1lZkZWUp2n733XciMDBQABDdunUTS5cuFUuXLhUnT55UtPn3v/8tbG1tBQDRqlUrMWfOHBEZGSnGjRsn3N3dVeLH/26x6evri4CAABERESE6duwoAIj58+er9Pndd98VAMRrr70mIiIixKRJk4REIhGNGzcWN2/eVGqblpYmHBwcBAAxYMAAER0dLSZOnCikUqlo1aqVRrcC5s6dKwAIX1/fcrV/UUJCgjA2Nha2trZi9uzZ4oMPPhBt27YVAMQXX3yh1PbWrVuK321oaKho2LChGD16tPjoo4+EnZ2dACBiYmKUjikuLhaDBw8WAIS/v7+IjIwUI0eOFPr6+sLe3l7IZDKl9u+8844AIHr27CkiIyNFeHi4GDx4sFi8eHGF+ke6xbGg+sYCITS7Lbh9+3bFaxAeHi4iIiLEqFGjxLBhw1TavvHGG4pxJjo6WoSFhQkfHx+1z6NJ/6OiogQAMXnyZNGmTRvRqlUrMXfuXDF58mShr68vLCwsRFpamqL9119/Lfr37y8AiP79+yt+v5cuXVJ5DeLj45Wey93dXbi5uYnQ0FARHR0tQkNDhbGxsdDX1xdnz55VaqvpbcE1a9YIAKJ9+/blav+iq1eviqZNm4rGjRuLadOmiY8++kh069ZNABAzZ85UaV/V78tPP/1UABBOTk5i/vz5Yt68eSIwMFBMmzZNo34xudIy+YAqhBA9evQQAMSOHTsU9QUFBcLKykp4e3sLIYTaAbW4uFhkZ2ernHvAgAECgPj555+Vysuan1BSUiK6du0qAIjevXuLnJycl8YPQEilUhEXF6coT05OVuqb3IEDBwQAMWHCBKXyHTt2CABi0qRJSuVjxowRAERYWJhSeVpammLgKO+AGhAQIACIWbNmlav98/Ly8kSLFi2ElZWVSE9PV5RnZWWJ5s2bC6lUKp48eaIolydX8j9qzw9o+/btUzsYrV69WgAQCxcuVCqXz7H5+OOPlcrNzMyEVCoVBQUFGveHah6OBc9Ux1ggxN+JxbJly8SNGzeUHs8nKUIIxYeeFxOLFz169EgAEB07dnzp82vaf3lyBUAMHz5c5OXlKerkY9uLCZz8mNLmXJWWXD1+/Fil7ZIlSwQAERERoVSuaXI1e/ZsAUAMGTKkXO1f5ObmJho1aiSuX7+uKCsqKhIuLi5CX19f3L59W6l9Vb8vXVxcBACV94ymeFuwCqm7HXD48GGkp6eXehsAeDZHyNTUVPHvrKws3L17V1H23//+t9wx/Prrr7hw4QIAYM2aNTAxMSnXcd99953S/JAOHToAAO7evavU7uuvvwYA+Pj4IDU1VfFwdHQEAKVVUJmZmfjnP/8JAwMDzJ8/X+k8zZo1w5gxY8rdL+DZ5XoASq9Vee3btw/3799H3759kZmZqYg7LS0NnTt3RlZWltolww0bNsT+/fvh4uKiKHvZa+Pt7a302nTs2BEAVFaImZmZ4cmTJzhz5ozG/aGajWNB1Y4Fz/vggw/Qtm1bpUd4eLhSGzMzM5WY1DE2NoaBgQEePHiAq1evltlWk/4/r2fPnvjuu+8gkUgUZaW9xhX1/MrCvLw83L9/Hw0aNACg2XtIncqMw+fOncP58+fRpUsX6OvrK16zW7duoXv37igpKUF8fLzaY6vifQn8/d746aefNO7P8zihvQoFBQUhLCwMR44cQUZGBiwtLbF582YYGxsjMDCwzGPv3LmDZcuWYf/+/Srzc4QG8xHOnz8PALC1tUXXrl3LfdygQYOU/t2wYUO1zy1PQEaNGqX2PA8ePFCKpbCwEE5OTlpZJi7/zyz/z12ab775Bjdv3sTw4cMVS7Tlce/atQu7du16aexyNjY2Ksu85a9NSUmJoky+DQQAeHl5lev8UVFRCA0Nhbe3NwYNGoRp06bB19e3zL5R7cCxoGrHgudNmzZNZQNM+VxJublz52Lfvn2YP38+du3ahXfffRdjxoxRSnAAwMjICO+//z6WLl2KLl26YPjw4XjvvffULpDRpP/P8/X1haGhoVKZujGlMnJycrBq1Sp8//33uHbtmtJ5NXkPqVPecfiHH35AUlISXn/9dbz++usA/n7Nfv31V7Rt21btcaW9blXxvgSA+fPn46233sKYMWMQExODadOmYdiwYdDX1+xaFJOrKtSsWTP4+PjgyJEj2LFjB0aMGIFDhw5hxIgRZWb5Fy5cQL9+/ZCZmQkPDw9Mnz4ddnZ22LBhg8abw2VlZQEA7OzsKtOVUj1+/BgAEBMTg8aNG5fZ9uHDhwCAJk2aaOW55QPmlStXymy3bds2HD9+HB06dFAkRvK4R40ahYCAALXHubu7Vzg2+etuaGiIbdu2qW3z4pWDkJAQNG/eHPPnz8eBAwdw4MABuLm5YevWrYpPZVQ7cSxQpu2x4Hlubm4vTVi7du2KM2fOIDw8HEePHsXkyZMRGRmJdevWwd/fX6ntkiVL4OjoiEWLFmHbtm3Ytm0bBgwYgG+//Ra2traKdpr0vzplZWWhT58+uHz5Muzt7fHhhx+ibdu2OH/+PNauXVvp85d3HD506BA2b96MRo0aKZIr+Wvm4+ODkJAQtcc5OTlVKj5Nfy9+fn6Ii4vDBx98gOPHj+P48eNo27YtNm3ahN69e5f7eZlcVbExY8bgyJEj+Mc//oGSkhIUFBSUeRsAAGbPno3MzEx88sknSpfMK3KZsmnTpgCgdnWaNjRt2hR//vknHBwc0K9fvzLbyj+dyd/sleXl5YUvv/wSJ0+eRHp6OqysrMp9rPx1adCgwUsH4opo0qQJ9PX1UVhYiG7dupV7U7+33noLb731FhISErBo0SLEx8fDy8sL169fr1EDNmmOY8HftD0WVISLiwuOHDmC69evY+XKlYiJicGQIUNw8uRJeHp6KrWdOHEixo0bh8OHDyMqKgrHjh2Dj48PkpOTYWDw7M+oJv2vTitXrsTly5fh5+eHPXv2KK7waIv8ynxqaiouX76sNGXiZeTvyby8vCoZh+XPoenv5dVXX8Xp06eRlJSEpUuXYteuXejfvz+uXLlS7rGcc66q2JAhQ2BqaorTp09jxYoVaNmypdq9bp4nv4w5a9ascj2HfFlqYWGhSp38jX///n0kJydrEHn5yC+/f//99y9tK7/se+fOHeTn56vUqysry5tvvomWLVsiPz8f8+bN0+hYedwHDx5ETk6ORseWh5GRkWIX4vK8Ni/y9vbGL7/8gh49euDhw4c4fvy4tkOkasax4G/aHgsqo3379li/fj3CwsIghMDOnTvVtmvQoAEGDx6M06dPw87ODlevXlW6WqNJ/yuqrN9vaeTvoXfffVfriRUAdOnSBT179gQAxWtYXvL35OnTpyu9x1tpKvN76d69O3bu3InAwEDk5eXhwIED5T6WyVUVMzY2Vtx2unXrFoKDg19671Z+Beb57yO7ceMGTpw4AQD466+/lNq3aNECwLNbCC++sTt06KC4zD116lQ8efJEqb6wsLBS9/Xff/99GBgYIDY2Fvv371epf/4+vKurKxwcHJCVlYUNGzYotYuLi9M4QTI0NMT//d//QU9PDxs3bsSsWbOQm5tbrmMHDBiAbt26QSaTISQkRGVzvpKSEpXXSlPy/ixdulTt5PjnX5uCggKV758Tz1bzAnj2PqLajWNB1Y0Fmrpw4YJKX+X/fv7/WmZmJm7evKnU7vnX9fm2mvS/ouS/36SkpHIfo+49lJ2djR07dgBQfQ9VxNdff41GjRrhp59+QlBQULnP2aFDBwQEBKC4uBjjx49XO+ZW9nXT9Pei7rVV9954qUqtNSQVULMU9Mcff1QsH/33v/+tVKdu+bX8Kw6aNWsmPvzwQzFx4kRhYmIi/P39BQDRokUL8dVXXyna5+TkKPZa6tevn4iKilJaovrw4UPh7Ows8L+v4Zg1a5aIjIwUwcHBwtLSUmm/JXmcpfVNXd2GDRuEgYGB0NPTE35+fiIqKkrMnTtXDBgwQFhYWCi1PXTokGjQoIHQ19cXQUFBYsGCBeKNN94QEolETJw4UePl10I827PGxMREABBNmzYVwcHBIjIyUixYsECMGDFCmJqaql3WfPPmTcU+O23bthVhYWEiMjJSjBkzRjRv3lz88MMPirbP73P1orLqoqOjBQBhYGAghg8fLhYuXChmzZolXn31VdG5c2dFO5lMJgCIrl27infeeUdERkaKPn36KJbNFxYWavSakO5xLKjesUC+DUFgYKBiD6gXH5mZmUIIIVxdXUXLli3F2LFjRWRkpAgKChL6+vqiadOm4s6dO4pzXrhwQejp6QlPT0/x3nvviQULFojOnTsLAOLtt9+uVP/L2lahtLr//ve/ivEsICBAREZGigsXLqi8Bs9vxXD8+HGhp6cnDA0NxTvvvCPCwsKEra2t8PPzEw0bNhSGhobi/fffV2z5UZGvvxFCiLi4OGFlZSUACDMzMzF8+HAREREhoqKixNixY0XTpk3V9ikjI0Oxr1WLFi3E9OnTFXue2dvbi88++0ypfVW/L83NzUW7du3EpEmTRGRkpHjzzTcFAOHo6Kh2S4vSMLnSstL+yJZG3YBaVFQkPv74Y9GqVSthbGwsXn/9dXHixAlRVFQkhg0bJho1aiQ8PT2VznPp0iXRv39/IZVKhY2NjfjHP/6hVJ+VlSUWLFgg2rVrJxo2bCgaN24sunbtKkJDQ5X2V6nIG1cIIc6fPy9GjBghmjdvLgwNDUWzZs1E3759RXR0tErbn376SfTp00cYGRkJc3Nz8cYbb4hz585V6PvE5NLS0kR0dLTo3bu3sLS0FAYGBsLCwkK0b99eBAQEiKVLlyoNmnIymUyxAV2jRo2EmZmZcHZ2FpMnTxa///67ol1FkyshhPj555/FoEGDhKWlpTA0NBS2trbCx8dHrFmzRtEmPz9fzJs3T/Tq1UtYWloKIyMj4eTkJBYvXqy03xbVHhwLqncskCcWZT3kr+2GDRvEG2+8IWxtbYVEIhFt2rQRISEh4u7du0rnfPTokZg9e7bo3r27aNy4sTA1NRVdunQRa9euLfUDT3n7X5HkSohnyVLPnj2FiYmJaNWqlVLyXNo+V7t37xYuLi7CyMhIdOnSRWzcuFGUlJSI5cuXK84j30+qosmVEM/20/r8889Fv379RLNmzYShoaEwNzcXjo6Ows/PT0RHR4uUlBSV454+fSo+++wz0bVrV2FkZCRMTExEu3btxJgxY8S5c+eU2lb1+/LTTz8Vr732mmjWrJmQSCSibdu24v333xePHj3S6LXQ+19ARERERKQFnHNFREREpEVMroiIiIi0iMkVERERkRYxuSIiIiLSIiZXRERERFrE5IqIiIhIi5hcEREREWkRkysiIiIiLWJyRURERKRFTK6IapiUlBSYmJgovs29NOPGjYOenh4SEhKUynNycvD222+jVatWkEgkaNasGV5//XWVdkRUs9y7dw+jR4+GpaUljI2N0adPH8WXdMt9//33cHd3R+PGjWFsbIy2bdvigw8+wNOnT3UUNaljoOsAiOhvOTk5CAwMfOlAGRMTg61bt6qty8/Ph6mpKT777DPY2dnh7t27iIyMhJ+fH65fv46WLVtWRehEVAkPHz5Er1690LRpU8TExMDU1BSbNm2Cj48P4uLi4OnpCQDIyMjA22+/jS5dusDQ0BB79+7F8uXL8eTJE6xZs0bHvSA5frcgUQ0ycuRIXLx4Ec7Oznj06JHaq00XL15Er169MHfuXHzyySeIj49/6VWunTt3Yvjw4di1axeGDh1aNcETUYV99NFH+PTTT3H37l3Y2dkpyn18fCCTyXD+/PlSj+3UqRP09fVx6dKl6giVyoG3BYlqiDVr1uDgwYPYvXs3TE1N1bZ5/PgxAgMDMWrUKEyaNKnc53706BEAoG3btlqJlYi067fffoOVlZVSYgUAQ4cORVJSEm7duqX2uJKSEshkMrRr1646wqRyYnJFVAMkJiYiLCwMGzZsQMeOHUttN2HCBEilUqxdu/al5ywuLsYff/yBzZs3Y/78+Vi0aBFcXFy0GTYRaUmTJk3w6NEjZGRkKJXLpwikpKQolWdnZyMpKQkjRoyAiYkJPvvss2qLlV6OyRWRjslkMgwfPhxTp07FiBEjSm23cuVKJCQkYNeuXWjUqFGZ50xNTYWBgQFsbW0xceJEzJ49Gx988IG2QyciLQkNDQUAjB07Frdu3UJOTg527dqFFStWAACysrIUbceMGQOpVAo3Nzdcu3YNe/bsQZs2bXQSN6nH5IpIh4QQGDt2LKysrBAVFYWcnBzk5OSgqKgIxcXFip9Pnz6NDz/8EOvWrYOVlRVycnKQm5sL4NknW/nPcq1atcL58+dx+PBhzJs3D59//jk8PDxUPhUTUc3g5eWFPXv24NKlS3BwcICZmRm+/fZbzJ07FwBgbGysaLtkyRKcPHkSW7duhZWVFbp06YL169frKnRSgxPaiXTozp07aN26dZltvv32Wxw/fhybNm0qtY29vT1u375dav2xY8fg6+uL8PBwLFu2rGLBElG1uHv3LgwNDWFjY4OPP/4YkZGRSE5ORufOnVXaFhcXw9vbG4mJifjjjz9gYWGhg4jpRdyKgUiHrK2tcfLkSZXy6dOnAwBWr16Ndu3awdPTU2UC+4ULFzBjxgx89dVX6NWrV5nP069fP64mIqolWrVqpfg5Li4O1tbWcHZ2Vtu2QYMG6NevH06dOoXff/8dHh4e1RUmlYHJFZEOSSQS9OnTR6Xc3NwcABR1zZo1U1npV1RUBABwcXGBm5tbmc+TmJiIkpISlZVIRFRznTx5EgkJCViyZAkaNGhQaruzZ88CAP9/1yBMrojqmKSkJKxbtw4+Pj6wtbXFlStXsGjRIpiYmGDWrFm6Do+I1EhPT0dkZCT69+8PGxsbXLx4EQsWLED//v0V864A4P3330ezZs3QtWtXFBUVYcuWLfjxxx8xfvx4tGjRQoc9oOcxuSKqY0xMTHDjxg0cOHAAmZmZsLa2hre3N6KiotC+fXtdh0dEahgYGODq1avYsWMHcnJy0KZNG4SFhWHu3LkwNDRUtDMyMsLGjRtx7949GBgYwNHREevWrcPkyZN1GD29iBPaiYiIiLSIWzEQERERaRGTKyIiIiItYnJFREREpEVMroiIiIi0iMkVERERkRZxKwaiKlBQUIDk5GRdh1Eprq6uaNiwoa7DIKoV+H+ensfkiqgKJCcn4yd/fzg2lOg6lApJLcgH9u+Hu7u7rkMhqhWSk5MRsD4Akha18/98/v187Andw//zWsLkiqiKODaUwMXISNdhEFE1kbSQwNjBWNdhUA3AOVdEREREWsTkioiIiEiLmFwRERERaRGTKyIiIqoSCQkJ0NPTK/URHR2taHvnzh2MHTsWdnZ2MDExgaurKzZv3qy74CuBE9qJiIioSnTv3h2JiYkq5b/++iumT5+uWJ2Yk5OD/v37w9zcHOvWrYOJiQm+/vprjB8/HmZmZhg6dGh1h14pTK6IiIioSpiZmcHNzU2prLi4GCEhIXj77bfx5ptvAgBOnTqF//znPzh//jy6d+8OAPDy8kKTJk2QkJBQ65Ir3hYkIiKiarN27Vr85z//wapVqxRlEsmz/cGuX7+uKHvy5Any8/PRu3fvao+xsnjlioiIiKrFn3/+icjISCxcuBDW1taKcm9vbwwcOBChoaHQ19dHYGAg3n33XYwZMwYjRozQYcQVwytXREREVC3CwsLQvHlzTJs2TalcT08P27ZtQ+PGjTFlyhTY29sjPT0d69evh56eno6irbgKJ1e///47TE1NkZCQAABIS0uDr68vnJyc4OnpidTUVEXbsuqIiIio7ktISMC2bdvwySefwNDQUKkuIyMDffv2xdixY3H//n0MHz4cP//8M3r27Im0tDQdRVxxFUquioqKEBwcDBsbG0XZ9OnT4efnh6tXryI8PBzBwcHlqiMiIqK6rbi4GDNmzICzszMCAwNV6mfPng0AWLx4MaRSKb788kscOXIEycnJGD9+fDVHW3kVSq6ioqLg5eUFT09PAEBmZib27t2LSZMmAQD8/f1x8+ZNpKSklFmnTn5+PrKyspQe+fn5FQmTiIiIaoDNmzfj8uXLCAsLU3ubb//+/ejYsaNS2YABAzBw4EAcO3YMeXl51RWqVmg8of3UqVOIi4vDiRMnMGXKFADArVu3IJVKYWpqCuDZvVM7OzukpqaisLCw1LpOnTqpnH/p0qVYuHChUlloaCgmTJigceeIdEUmk8FgymTcb9BA16FUiEFxMTIyMnD27FmNj/Xw8KiCiIioNvv0008hlUpLnZxuaGiI3377DYWFhUq3DB88eAATExOV24g1nUbJVVZWFqZNm4Y9e/aodFQIofJveXZaVt2L5s2bhzlz5iiVSSQSxTJNqr9ycnIwceJE/Prrr0hLS4O5uTk6d+6MiIgIeHt7l3rcuHHjsGXLFsTHx6u0y8vLQ1RUFLZu3YrHjx+jR48e+PzzzxX7rFRUYmIismI2oIWRUaXOoyt/PX0KS19fxQZ/REQVdf78edy4cQMjR46EsbGx2jYzZsxAZGQkhgwZgvfeew8GBgbYvHkzzp8/j8WLF6NBLfugqlFydfjwYTx48AD9+vUDADx8+BBHjx7F+vXrkZOTg+zsbJiZmUEIgXv37sHR0RF2dnal1qnDRIpKk5+fD1NTU3z22Wews7PD3bt3ERkZCT8/P1y/fh0tW7ZUOSYmJgZbt24t9ZxBQUE4c+YMVqxYATs7O6xZswbe3t44ffo0XFxcqrI7RET1wsmTJwEAr732WqltFixYAAcHB6xevRojR46Evr4+nJycsHPnTgwbNqy6QtUajeZcBQUFIT09Hbdv38bt27fRo0cPbN++HW+99RYCAgIQGxsLANi9ezccHR3h7OwMc3PzUuuINGFpaYnY2FiMGDECnp6eGDlyJJYuXYqnT5+q/XqFixcvYsaMGZg/f77a8+3fvx/79u3D+vXrMWbMGHh7e2P79u1o0aIFZs6cWdXdISKqF2bPng0hhGIqUWlGjx6NX3/9FZmZmfjrr79w6tSpWplYAVrc52rVqlU4dOgQnJycsHLlSqWrBWXVEVXGo0ePAABt27ZVKn/8+DECAwMxatQoxWKKF23ZsgVSqRSDBw9WlBkYGCAoKAjx8fG4c+dO1QVONZb8vePu7o4OHTrgiy++UNRxWxkiKo9K7dAu3+MKAJo3b45jx46pbVdWHZGmiouL8fDhQxw7dgzz58/HokWLVG7hTZgwAVKpFGvXrsWff/6p9jznzp1Dhw4dVO7lOzk5Kert7e2rphNUYy1fvhwdOnTAzp07kZGRgdatW2PChAlo3LixYluZmTNnYt++fQgODsaZM2dUzpGfn6+yyplTHojqD379DdUqqampiqtU+vr6WLhwIcLDw5XarFy5EgkJCUhKSkKjRo1KPVdaWpoikXpekyZNFPVU/5iamuLChQsoKSlBfn4+OnbsCKlUqthWRj7Fwd/fHyEhIUhJSVFZ+cxVz/WPTCbDRIeJaGBSuyZeyxU7VHyFcF1U2VXPTK6oVmnVqhXOnz+Phw8f4l//+hc+//xz7Nq1Cz///DMsLS1x+vRpfPjhh9i6dSusrKyQk5OD3NxcAMDTp0+Rm5urtFpF3arVkpKSausP1TxhYWHw8vKCt7c3DA0NsXfvXujr65e55cyLyRVXPdc/iYmJiD0TC2MH9avharrcm7kY4DSAK4S1hMkV1SoNGzZUbJMwcOBA9O3bF76+vli+fDmWLVuGmJgYFBYWIigoSOVYPz8/2Nvb4/bt2wAAa2tryGQylXbysubNm1ddR6jGiomJQbdu3RAREYHQ0FCMGTMGP/74I4DybyvDRIrqioKCAiQnJ+s6jEpxdXVFw4YNq/U5mVxRrdavXz/o6+vj0qVLAICPPvpIZQL7hQsXMGPGDHz11Vfo1auXorxHjx746aefUFJSAn39v9d2yL89gJ/g6qfo6GgcPXoUNjY22Lt3Lzp27IiDBw+iX79+Gm0rQ1QXJCcn4yd/fzg2rJ0fFlIL8oH9+6t9PGdyRbVaYmIiSkpKYGdnB+DZqsEXVw4WFRUBAFxcXODm5qYoHzduHHbu3IlDhw4pVgwWFhZi+/btePXVV9G6devq6QTVKObm5jhz5gy6d+8OmUyGx48fw8bGRmlbmZkzZ3JbGao3HBtK4FJLN0TWFSZXVGskJSVh3bp18PHxga2tLa5cuYJFixbBxMQEs2bN0vh8gwYNUkxKXrlyJWxsbLBq1So8ePAAP/zwg/Y7QLXCtm3bMH36dGzcuBElJSVYsGABevbsCeDZtjLBwcFYv349LCwsuK0MEanF5IpqDRMTE9y4cQMHDhxAZmYmrK2t4e3tjaioKLRv375C59y+fTsiIiIwa9YsZGdnw8PDAwkJCejSpYt2g6dao0ePHqWumOK2MkRUHkyuqNbo0KEDjh8/rvFx3t7eKhOR5YyMjLBixQqsWLGisuEREREB0OIO7URERETE5IqIiIhIq5hcEREREWkRkysiIiIiLWJyRURERKRFXC1IOlUXvloB0M3XKxARUc3E5Ip0Kjk5GQHrAyBpUTu/WgEA8u/nY0/oHn5dDhERAWByRTWApIWk1n6TPBER0Ys454qIiIhIi5hcEREREWkRkysiIiIiLWJyRURERKRFTK6IiIiItIjJFRFRHeXt7Q09PT2Vx/bt2xVtEhIS1Lbp0KGDDiMnqt24FQMRUR02aNAgREVFKZW98sorKu32798PGxsbxb+NjIyqPDaiuorJFRFRHWZpaQk3N7eXtnNxcUHr1q2rPiCiekCj24Lh4eHo3r07XFxc4OHhgStXrgAA0tLS4OvrCycnJ3h6eiI1NVVxTFl1VDX++OMPhIeHo1OnTjAzM4O1tTWGDh2Kf//730rtKno7ICUlBSYmJvD29q7CXhAREdVOGiVXkZGRSEpKwuXLl+Hs7Iyvv/4aADB9+nT4+fnh6tWrCA8PR3BwsOKYsuqoavzrX//C8ePHMW3aNBw8eBDLli1DYmIiPDw8cOvWLZX2+/fvR2JiouKxa9euUs+dk5ODwMBAPH36tCq7QERasmXLFlhYWKBv377YvHkzhBBq27Vv3x4ODg4IDQ3F7du3qzdIojpGo+TK1NQUAJCRkYEbN25gwIAByMzMxN69ezFp0iQAgL+/P27evImUlJQy60qTn5+PrKwspUd+fn5F+1cvDR48GGfPnsXUqVPh5eWFcePGYePGjcjKykJsbKxKexcXF7i5uSkezs7OpZ57ypQp0NPTw9ChQ6uyC0SkBWvXrsWpU6ewbt06mJubY/z48Rg3bpxSm65du+LcuXPYu3cvJk6ciH379sHV1RUXL17UTdBEdYBGc64eP36MV199FdevX8fs2bMxePBgJCcnQyqVKhIvPT092NnZITU1FYWFhaXWderUSe1zLF26FAsXLlQqCw0NxYQJEyrSP/of+ZWmq1ev4uzZs4qfAeDChQtIS0t76Tl++OEH7N+/Hxs3bsTWrVuRlZWlOFdFyWQyTHSYiAYmDSp1Hl0qdihGRkaG0mshk8lgMGUy7jeonf0yKFbtU3l5eHhUQURUEfIPSr1790ZQUBBCQkIQExODCRMm4LXXXgMAmJubK750fODAgXj77bfRrVs3zJw5E8ePH9dZ7ES1mUbJlbm5OS5duoSMjAwEBwcjLCwMY8eOVbnMLISAnp6e4ufS6tSZN28e5syZo1QmkUggkUg0CZVe8N133wF4NnjK//jJE66goCC0aNECPj4+mDdvntpJrYmJiVi9ejW2bNmCESNG4OjRo3jy5Eml/5AmJiYi9kxsrf7i5tybuRjgNEDxBwp41q+smA1oUUtXXP319CksfX2V+kS137hx4xATE4O4uDhFcvWidu3aoVevXoiLi0NhYSEMDQ2rOUqi2q9C+1xZWlpiypQp+Omnn+Dg4ICcnBxkZ2cDeJY83bt3D46OjmXWlUYikUAqlSo9mFhVzqNHjzBv3jy4urpi9OjRivLy3g6QyWQYPnw4pk6dihEjRlRz9ETVKzc3F5MnT0bHjh3RqVMnBAUFAagbi3OMjZ99iMnLy3tpu5KSEhQUFFRHWER1jkbJ1Z9//gngWZJ0+PBhuLq6wtzcHAEBAYq5PLt374ajoyOcnZ3LrKPqkZaWhv79+8Pc3ByHDx9W2rtGfjtg4MCBiIiIwIkTJ1BcXIyZM2cq2gghMHbsWFhZWSEqKgo5OTnIyclBUVERiouLFT8T1RXh4eEwMjLCtWvXkJKSgq+++gpA3Vicc/DgQQBA9+7dS22TmZmJf/3rX2jXrh1MTEyqKzSiOkWj24KzZs3ClStXIIRA+/btsW7dOgDAqlWrEBwcjPXr18PCwgJbt25VHFNWHVWt27dvo3///mjZsiX27t2Lxo0bl9le3e2Au3fvKgbkJk2aqBxjZmaGb7/9FuPHj6+CHhBVrydPnmDjxo1Kq+WaNWumWJwj/6Do7++PkJAQpKSkqJ0/mp+fr7IQRxfTGyIiIuDh4QGpVIr4+HgsXboUPXr0QGBgoKLNsWPHcPPmTTg5OeHRo0f49NNPIZPJsGHDhmqNlagu0Si5ev4rE57XvHlzHDt2TOO6muCPP/7AF198gcOHD+POnTswMTFB7969sWTJEqX9nnbt2oXY2FhcvHgRMpkMDg4OCAkJwbRp09DghUnL9+7dw4cffoijR4/i6dOn6NatG5YsWYK+fftWW7/u3buHV199FZ07d8bu3bvLPag/fzvA0NAQ1tbWOHnypEq76dOnAwBWr16Ndu3aaTV2Il25ffs2JBIJdu/ejZ07dyIvLw/R0dFo2rSpRotzasrCnEOHDmH16tV4+vQprK2tMXLkSIwbNw5JSUmKNufOncP//d//IT09HRKJBM7Ozli9ejVsbGwqvWClPqnti3PULcwB6u/inMrOJ673O7Q/vyeUk5MTbt++rfi0d/HiRbRp0wYA8Pnnn8Pd3R2TJ0+GVCrF999/j5kzZ+LGjRtYvXq14nwPHz5Er1690LRpU8TExMDU1BSbNm2Cj48P4uLi4OnpWeV9ys/Ph7+/PwwMDPDPf/6z3ImVutsBEokEffr0UWlrbm4OAGrriGqr3NxcZGVloWXLlvjll18QFxcHf39/xMfHa7Q4p6YszLlw4cJL23h4eCAiIqIaoqnbavviHHULcwAuzqmoep9cDR48WOkSuZeXF2xsbODr64vY2Fh8/PHHAIDjx4+jYcOGinb9+/fHtWvX8M0332D58uWKuUxffvklHjx4gF9//RV2dnYAgAEDBsDHxwczZ87E+fPnq7xPq1evxsWLFzF9+nSVXdltbW1ha2sLgLcDiF5kb2+PRo0aYdCgQQCAfv36wdjYGLm5uYrFOWZmZi9dnMMVzkT1W4VWC9Yl6gZA+e3AjIwMRdnziZVc+/btUVBQoFgNCQC//fYbrKysFImV3NChQ5GUlKR2h3Rtk2/Sunr1ari7uys9vvnmG0U7mUyGzz77DAMGDMCkSZPQpEkT/PLLLxgyZEiVx0hUEzVr1gx9+vTBjh07AEDxFV/dunXj4hwiKrd6f+VKHfkco27dupXaRgiBU6dOwc7ODlZWVoryJk2a4NGjR8jIyIClpaWiXL6nVEpKiuJWY1XZtGkTNm3a9NJ2I0aMqPDWCgkJCRU6jqim++abbzBhwgQsWbIERkZG2L17N8zMzLg4h4jKjcnVC0rbE+pFy5cvx40bN7Bt2zaleRehoaH45z//ibFjx2LNmjWwsrLCjz/+iBUrVgAAsrKyqrwPRFRxrVu3Rnx8vEp5TV+cQ0Q1R72/Lfi8svaEet66deswf/58fPnllxg5cqRSnZeXF/bs2YNLly7BwcFBsVXB3LlzAfy9iR8RERHVTbxy9T/l3RNq0aJFWLJkCb777rtSb6n5+/vD398fd+/ehaGhIWxsbBQT41955ZWq6gIR1TEFBQVITk7WdRiV4urqqnbOKlFdxuQK5d8TKioqCsuWLcPBgwfx+uuvv/S8rVq1UvwcFxcHa2trToAlonJLTk7GwOhtMGza6uWNa6DCR3dxJBr8jkqqd+p9clXePaF27NiBRYsWYc2aNeVKrJ538uRJJCQkYMmSJSobjmqCn2KJ6h/Dpq0gseFGvUS1Sb1PrsqzJ1Rubi7ee+89mJiYwM3NTWWvKjc3N8XP6enpiIyMRP/+/WFjY4OLFy9iwYIF6N+/v2LeVUUlJyfjyuK+cLaqnTvlXkkvBuaf4KdYIiKq0+p9cvX8nlDP77QOPLsNGB0djYcPHyI9PR0A0LNnT5VzPL9zs4GBAa5evYodO3YgJycHbdq0QVhYGObOnQtDQ8NKx+ts1QDuLWpnckVERFQf1Pvkqjx7QrVu3Vrlqy9KY2FhgePHj2shMiIiIqqNuBUDERERkRYxuSIiIiLSIiZXRERERFrE5IqIiIhIi5hcEREREWlRnVwtyM02iYiISFfqZHLFr4wgIiIiXamTyRXAr4wgIiIi3eCcKyIiIiItYnJFREREpEVMroiIiIi0iMkVERERkRYxuSIiIiLSIo2Tq9mzZ6Nr165wcXGBn58f/vzzTwBAWloafH194eTkBE9PT6SmpiqOKauOiIiIqC7ROLkaNWoULly4gMuXL8PGxgYREREAgOnTp8PPzw9Xr15FeHg4goODFceUVfei/Px8ZGVlKT3y8/Mr0DUiIiKi6qfxPlfPb2zp6uqKI0eOIDMzE3v37kVsbCwAwN/fHyEhIUhJSYGdnV2pdZ06dVI5/9KlS7Fw4UKlstDQUEyYMKHcMcpkMkzt1Rz6kmJNu1cjlLRujoyMDJw9e1apXCaTAW5zcNZIT0eRVY7MRgAv9Esmk2Giw0Q0MGmgw8gqp9ihWOX3JZPJYDBlMu43qJ39MihW7VN5eXh4VEFERES1R4U3Ec3Ly0NMTAzmz5+PW7duQSqVwtTUFACgp6cHOzs7pKamorCwsNQ6dcnVvHnzMGfOHKUyiUQCiURS7tgSExOx7vtrkNiYV7R7OpX/RxoGdm+rskN7YmIisGsl3FvUzj/YifeLgW6DlfqVmJiI2DOxMHYw1mFklZN7MxcDnAao9CsrZgNaGBnpMLKK++vpU1j6+tbrbwn4/fff0a1bNxw8eBDe3t5IS0vD2LFjce/ePVhYWGDz5s1wdHTUdZhEVANVaEJ7dnY2/P39ERwcjKCgIACAEEKpjRACenp6L617kUQigVQqVXpoklgREVVWUVERgoODYWNjoyjj9AYiKi+Nr1zl5uZi8ODBmDJlCkaPHg0AcHBwQE5ODrKzs2FmZgYhBO7duwdHR0fY2dmVWkdEVBNFRUXBy8sLDx8+BIAypz5U1fQGoO5OcaiLavsUB3XTG4D6O8WhstMbNE6upk2bBnd3d0ViBQDm5uYICAhAbGwsZs6cid27d8PR0RHOzs4AUGYdEVFNcurUKcTFxeHEiROYMmUKAJQ59aGqpjcAdXeKQ11U26c4qJveAHCKQ0VplFxlZ2dj06ZNaNeuHeLj4wEAvXv3xldffYVVq1YhODgY69evh4WFBbZu3ao4rqw6IqKaIisrC9OmTcOePXtgaGioVKfp9AZOZyCqvzRKruS39dRp3rw5jh07pnEdEVFNcfjwYTx48AD9+vUDADx8+BBHjx7F+vXrOb2BiMqNO7QTEf1PUFAQ0tPTcfv2bdy+fRs9evTA9u3b8dZbbymmNwDg9AYiKlOFt2IgIqpPOL2BiMqLyRURUSkSEhIUP3N6AxGVF28LEhEREWkRkysiIiIiLWJyRURERKRFTK6IiIiItIjJFREREZEWMbkiIiIi0iImV0RERERaxOSKiIiISIuYXBERERFpEZMrIiIiIi1ickVERESkRUyuiIiIiLSIyRURERGRFjG5IiIiItIiJldEREREWsTkioiIiEiLmFwRERERaRGTKyIiIiItYnJFREREpEVMroiIiIi0qMLJVUFBAR4/fqzNWIiIiIhqPY2Tq3PnzqFfv36wsrLCDz/8oChPS0uDr68vnJyc4OnpidTU1HLVEREREdUlGidXbdq0wc6dO9G1a1el8unTp8PPzw9Xr15FeHg4goODy1X3ovz8fGRlZSk98vPzNQ2TiKjCZs+eja5du8LFxQV+fn74888/AfCDIhGVj4GmB1hZWamUZWZmYu/evYiNjQUA+Pv7IyQkBCkpKbCzsyu1rlOnTirnWrp0KRYuXKhUFhoaigkTJpQ7RplMhqm9mkNfUqxJ12qMktbNkZGRgbNnzyqVy2QywG0Ozhrp6SiyypHZCOCFfslkMkx0mIgGJg10GFnlFDsUq/y+ZDIZDKZMxv0GtbNfBsWqfSovDw+PKoioeo0aNQpffPEFAGDSpEmIiIjAhg0bFB8UZ86ciX379iE4OBhnzpzRcbREVNNonFypc+vWLUilUpiamgIA9PT0YGdnh9TUVBQWFpZapy65mjdvHubMmaNUJpFIIJFIyh1PYmIi1n1/DRIb80r0Snfy/0jDwO5t4e7urlSemJgI7FoJ9xa18w924v1ioNtgpX4lJiYi9kwsjB2MdRhZ5eTezMUApwEq/cqK2YAWRkY6jKzi/nr6FJa+virvwfri+X67urriyJEjZX6IfHEsy8/PV7niruk4RkS1l1aSKwAQQqj8W09P76V1L+IAREQ1RV5eHmJiYjB//vwyP0S+mFxp4wo8UHevwtdFtf0qvLor8ED9vQpf2SvwWkmuHBwckJOTg+zsbJiZmUEIgXv37sHR0RF2dnal1hER1VTZ2dkYNmwYgoODERQUhAsXLpT7g6I2rsADdfcqfF1U26/Cq7sCD/AqfEVpZZ8rc3NzBAQEKC6X7969G46OjnB2di6zjoioJsrNzcXgwYMxbtw4hIeHA1D+EAmgzA+KEokEUqlU6cEr8kT1h8bJ1ZEjR+Dm5oakpCR88skn6NmzJwBg1apVOHToEJycnLBy5Ups3bpVcUxZdURENc20adPg7u6O0aNHK8r4QZGIykvj24IDBw7EwIEDVcqbN2+OY8eOqT2mrDoiopokOzsbmzZtQrt27RAfHw8A6N27N7766iusWrUKwcHBWL9+PSwsLPhBkYjU0tqEdiKiukA+N1QdflCsuby9vXH8+HGV8u+//x5BQUE6iIjqMyZXRERUJwwaNAhRUVFKZa+88oqOoqH6jMkVERHVCZaWlnBzc9N1GETaWS1IRERERM8wuSIiojphy5YtsLCwQN++fbF58+ZS584RVTUmV0REVOutXbsWp06dwrp162Bubo7x48dj3Lhxug6L6ikmV0REVOs5Ozujd+/eCAoKwoEDBzBlyhRs3bpVsZ0GUXVickVERHWO/KpVXFycjiOh+ojJFRER1TnGxs++4y8vL0/HkVB9xOSKiIjqnIMHDwIAunfvruNIqD7iPldERFTrRUREwMPDA1KpFPHx8Vi6dCl69OiBwMBAXYdG9RCTKyIiqvVOnDiBdevWIScnB/b29ggLC8NHH30EAwP+maPqx3cdERHVeidOnNB1CEQKnHNFREREpEVMroiIiIi0iMkVERERkRYxuSIiIiLSIiZXRERERFrE1YJERFRtCgoKkJycrOswKs3V1RUNGzbUdRhUQzG5IiKiapOcnIwri/vC2aqBrkOpsCvpxcD8E3B3d9d1KFRDMbkiIqJq5WzVAO4tam9yRfQynHNFREREpEVMroiIiIi0iMkVERERkRZVW3KVlpYGX19fODk5wdPTE6mpqdX11EREWsFxjIjKo9qSq+nTp8PPzw9Xr15FeHg4goODq+upiYi0guMYEZVHtawWzMzMxN69exEbGwsA8Pf3R0hICFJSUtCpUyeltvn5+cjPz1cqk0gkkEgkGj1n4aO7lQtah8qK/Up6cTVGol1X0ovhrKY8/36+mtLao7T4Uwtqb79SC/LRRtdB1DC6GMeAujmW1eZxDKibY1lZsXMsqwBRDX777TdhaWmpVNatWzexZ88elbZRUVECgNIjKiqqOsIkIioVxzEiKq9q2+dKCKHybz09PZV28+bNw5w5c5TKKvJpj4hI2ziOEVF5VEty5eDggJycHGRnZ8PMzAxCCNy7dw+Ojo4qbSt66ZyIqCpxHCOi8qqWCe3m5uYICAhQzFXYvXs3HB0d4eys7q41EVHNw3GMiMpLT7x4nbuKpKWlITg4GP/9739hYWGBzZs3q/3ER0RUU3EcI6LyqLbkioiIiKg+4A7tRERERFrE5IqIiIhIi5hcEREREWkRkysiIiIiLWJyRURERKRFTK6IiIiItIjJFREREZEWMbkiIiIi0iImV0RERERaxOSKiIiISIv+H1MnBdAC1l5WAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 700x200 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Improved aesthetics for a research publication plot\n",
    "def create_research_plot(counts_df):\n",
    "    # Set a professional color palette\n",
    "    colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728']\n",
    "    model_names = ['iCHO1766', 'iCHO2101', 'iCHO2291', 'iCHO3K']\n",
    "\n",
    "    # Create subplots with 1 row and 2 columns\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(7,2))  \n",
    "    \n",
    "    # Define the axes mapping for two categories\n",
    "    axes_mapping = {0: axes[0], 1: axes[1]}  # Mapping two axes\n",
    "    categories = ['Matched Genes', 'Matched Essential Genes']\n",
    "\n",
    "    # A function to annotate the bars with their values\n",
    "    def annotate_bars(ax):\n",
    "        for p in ax.patches:\n",
    "            ax.annotate(f'{p.get_height()}', (p.get_x() + p.get_width() / 2., p.get_height()), \n",
    "                        ha='center', va='bottom', xytext=(0, 3), textcoords='offset points', fontweight='light', fontsize=12)\n",
    "\n",
    "    # Plotting and annotating each category\n",
    "    for index, category in enumerate(categories):\n",
    "        ax = axes_mapping[index]\n",
    "        counts_df.loc[category].plot(kind='bar', ax=ax, title=category, color=colors, edgecolor='black', linewidth=.3, width=0.9, legend=False)\n",
    "        annotate_bars(ax)\n",
    "        ax.set_title(category, fontsize=15, fontweight='light', pad=16)\n",
    "        ax.grid(True, linestyle='-', linewidth=0.5, axis='y')\n",
    "        ax.set_xticks([])  # Remove x-axis ticks\n",
    "        ax.set_xticklabels([])  # Remove x-axis labels\n",
    "        ax.tick_params(axis='y', labelsize=8)\n",
    "\n",
    "    # Remove border of the plots\n",
    "    for ax in axes:  # Applies to both axes\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "\n",
    "#     # Create custom handles for the legend\n",
    "#     custom_handles = [mpatches.Patch(color=color, label=model_name) for color, model_name in zip(colors, model_names)]\n",
    "#     fig.legend(handles=custom_handles, loc='upper left', bbox_to_anchor=(0.15, 0.9), fontsize=10, frameon=False, fancybox=False, shadow=False)\n",
    "    \n",
    "    # Adjust spacing between the subplots\n",
    "    plt.subplots_adjust(wspace=0.3, hspace=0.3)\n",
    "    \n",
    "    # Save the figure\n",
    "    plt.savefig('../Analyses/recons_comparisons/Essential_genes_coverage_comparison.png', dpi=300, bbox_inches='tight')\n",
    "    plt.show()\n",
    "\n",
    "# Call the function\n",
    "create_research_plot(counts_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b9cd571",
   "metadata": {},
   "source": [
    "#### Comparisions of Essential genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1973a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_unique_genes(df1, df2, df3):\n",
    "    \"\"\"\n",
    "    Find unique genes in each DataFrame compared to others.\n",
    "    \"\"\"\n",
    "    # Convert gene columns to sets for comparison\n",
    "    genes_1 = set(df1['gene_exp_ess.id'])\n",
    "    genes_2 = set(df2['gene_exp_ess.id'])\n",
    "    genes_3 = set(df3['gene_exp_ess.id'])\n",
    "\n",
    "    # Find differences\n",
    "    unique_1 = genes_1 - genes_2 - genes_3\n",
    "    unique_2 = genes_2 - genes_1 - genes_3\n",
    "    unique_3 = genes_3 - genes_2 - genes_1\n",
    "\n",
    "    return unique_1, unique_2, unique_3\n",
    "\n",
    "def find_gene_reactions(model, unique_genes):\n",
    "    \"\"\"\n",
    "    For each gene, find associated reactions in the given model.\n",
    "    \"\"\"\n",
    "    for gene_id in unique_genes:\n",
    "        try:\n",
    "            # Get the gene object from the model\n",
    "            gene = model.genes.get_by_id(gene_id)\n",
    "            print(f\"Gene ID: {gene.id}, Name: {gene.name}\")\n",
    "            print(\"Associated Reactions:\")\n",
    "            \n",
    "            # Print associated reactions\n",
    "            for reaction in gene.reactions:\n",
    "                print(f\" - {reaction.id}: {reaction.reaction}\")\n",
    "        except KeyError:\n",
    "            print(f\"Gene {gene_id} not found in the model.\")\n",
    "            \n",
    "def find_gene_reactions_to_excel(model, unique_genes, model_name):\n",
    "    \"\"\"\n",
    "    For each gene, find associated reactions in the given model and store results in a DataFrame.\n",
    "    \"\"\"\n",
    "    results = []\n",
    "\n",
    "    for gene_id in unique_genes:\n",
    "        try:\n",
    "            # Get the gene object from the model\n",
    "            gene = model.genes.get_by_id(gene_id)\n",
    "            \n",
    "            # Find associated reactions and append results\n",
    "            for reaction in gene.reactions:\n",
    "                results.append({\n",
    "                    'Gene ID': gene.id,\n",
    "                    'Gene Name': gene.name,\n",
    "                    'Reaction ID': reaction.id,\n",
    "                    'Reaction': reaction.reaction,\n",
    "                    'Model': model_name\n",
    "                })\n",
    "        except KeyError:\n",
    "            results.append({\n",
    "                'Gene ID': gene_id,\n",
    "                'Gene Name': 'Not Found',\n",
    "                'Reaction ID': 'N/A',\n",
    "                'Reaction': 'N/A',\n",
    "                'Model': model_name\n",
    "            })\n",
    "\n",
    "    # Convert to DataFrame\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "\n",
    "# Step 1: Find unique genes in each model\n",
    "unique_2101, unique_2291, unique_3K = find_unique_genes(matched_df_2101, matched_df_2291, matched_df_3K)\n",
    "\n",
    "# Step 2: Find reactions associated with the unique genes for each model\n",
    "print(\"\\nReactions for unique genes in iCHO2101:\")\n",
    "find_gene_reactions(iCHO2101_changed, unique_2101)\n",
    "\n",
    "print(\"\\nReactions for unique genes in iCHO2291:\")\n",
    "find_gene_reactions(iCHO2291_changed, unique_2291)\n",
    "\n",
    "print(\"\\nReactions for unique genes in iCHO3K:\")\n",
    "find_gene_reactions(iCHO3K, unique_3K)\n",
    "\n",
    "df_2101 = find_gene_reactions_to_excel(iCHO2101_changed, unique_2101, 'iCHO2101')\n",
    "df_2291 = find_gene_reactions_to_excel(iCHO2291_changed, unique_2291, 'iCHO2291')\n",
    "df_3K = find_gene_reactions_to_excel(iCHO3K, unique_3K, 'iCHO3K')\n",
    "\n",
    "\n",
    "combined_df = pd.concat([df_2101, df_2291, df_3K], ignore_index=True)\n",
    "\n",
    "combined_df.to_excel('unique_genes_reactions.xlsx', index=False)\n",
    "\n",
    "print(\"Results successfully saved to 'unique_genes_reactions.xlsx'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca81d234",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "##### ----- Generate datasets from Google Sheet ----- #####\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "# Read data from the Google Sheet\n",
    "sheet_rxns = 'Rxns'\n",
    "rxns = sheet.read_google_sheet(sheet_rxns)\n",
    "rxns_relevant_columns = rxns[['Reaction', 'GPR_yeo', 'GPR_final']]\n",
    "\n",
    "# Filter rows from rxns_relevant_columns that have Reaction IDs matching combined_df\n",
    "matching_rxns = rxns_relevant_columns[rxns_relevant_columns['Reaction'].isin(combined_df['Reaction ID'])]\n",
    "\n",
    "# Loop through the matching reactions and print 'GPR_yeo' and 'GPR_final' for each matching reaction\n",
    "for rxns_unique in combined_df['Reaction ID']:\n",
    "    # Find the row in rxns_relevant_columns that matches the rxns_unique\n",
    "    gpr_info = rxns_relevant_columns[rxns_relevant_columns['Reaction'] == rxns_unique]\n",
    "    \n",
    "    if not gpr_info.empty:\n",
    "        # Extract 'GPR_yeo' and 'GPR_final' for this reaction\n",
    "        gpr_yeo = gpr_info['GPR_yeo'].values[0]\n",
    "        gpr_final = gpr_info['GPR_final'].values[0]\n",
    "        \n",
    "        # Print or use the GPR information\n",
    "        print(f\"Reaction: {rxns_unique}\")\n",
    "        print(f\"GPR_yeo: {gpr_yeo}\")\n",
    "        print(f\"GPR_final: {gpr_final}\\n\")\n",
    "    else:\n",
    "        # If the reaction is not found, add it to the 'not_found_reactions' list\n",
    "        not_found_reactions.append(rxns_unique)\n",
    "\n",
    "# Print reactions not found in Google Sheet\n",
    "if not_found_reactions:\n",
    "    print(\"\\nReactions not found in the Google Sheet:\")\n",
    "    for reaction in not_found_reactions:\n",
    "        print(reaction)\n",
    "else:\n",
    "    print(\"\\nAll reactions found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5380e3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "gpr_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "120e6629",
   "metadata": {},
   "source": [
    "## MEMOTE for the different reconstructions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e42df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install memote\n",
    "# !pip install --upgrade typing_extensions\n",
    "# !pip install --upgrade pip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bb4d8c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "import os\n",
    "import xlrd\n",
    "import cobra\n",
    "from cobra.io import load_json_model\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "from memote import test_model, snapshot_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "343ce20a",
   "metadata": {},
   "outputs": [],
   "source": [
    "iCHO3K = load_json_model('iCHO3595.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a5a46de",
   "metadata": {},
   "outputs": [],
   "source": [
    "mem_return, mem_result = test_model(iCHO3K, results=True)\n",
    "with open('iOpol_MemoteReport.html', 'w') as f:\n",
    "    f.write(snapshot_report(mem_result))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87135296-4494-4d95-8b45-4db834938865",
   "metadata": {},
   "source": [
    "## Comparison of Secretory Pathway Recons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc952939",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '../Data/Sec_Recon_shared_genes/'\n",
    "\n",
    "with open(path+'common_jahirs_metrecon.txt', 'r') as file: common_jahirs_metrecon = [line.strip() for line in file]\n",
    "with open(path+'common_secrecon_jahirs.txt', 'r') as file: common_secrecon_jahirs = [line.strip() for line in file]\n",
    "with open(path+'common_secrecon_metrecon.txt', 'r') as file: common_secrecon_metrecon = [line.strip() for line in file]\n",
    "with open(path+'unique_secrecon_metrecon.txt', 'r') as file: unique_secrecon_metrecon = [line.strip() for line in file]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc79de3a-5336-48e7-8d87-8588a827e870",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Subsystems of genes shared between Sec Recon and Met Recon (iCHO3000) minus Jahir's\n",
    "\n",
    "rxns_secrecon_metrecon = []\n",
    "subs_secrecon_metrecon = []\n",
    "for g in iCHO3000.genes:\n",
    "    if str(g) in unique_secrecon_metrecon:\n",
    "        for rxn in g.reactions:\n",
    "            if rxn.id not in rxns_secrecon_metrecon:\n",
    "                rxns_secrecon_metrecon.append(rxn.id)\n",
    "                subs_secrecon_metrecon.append(rxn.subsystem)\n",
    "\n",
    "\n",
    "# Subsystems of genes shared between Jahir's Recon and Met Recon (iCHO3000)\n",
    "\n",
    "rxns_jahirs_metrecon = []\n",
    "subs_jahirs_metrecon = []\n",
    "for g in iCHO3000.genes:\n",
    "    if str(g) in common_jahirs_metrecon:\n",
    "        for rxn in g.reactions:\n",
    "            if rxn.id not in rxns_jahirs_metrecon:\n",
    "                rxns_jahirs_metrecon.append(rxn.id)\n",
    "                subs_jahirs_metrecon.append(rxn.subsystem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c934822c-d958-479c-82b9-70e94e199473",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pie Chart Subsystems of genes shared between Sec Recon and Met Recon (iCHO3000) minus Jahir's\n",
    "\n",
    "subsystem_counts = Counter(subs_secrecon_metrecon)\n",
    "labels = list(subsystem_counts.keys())\n",
    "sizes = list(subsystem_counts.values())\n",
    "\n",
    "# Plot the pie chart\n",
    "plt.figure(figsize=(10, 8)) \n",
    "plt.pie(sizes, labels=labels, autopct='%1.1f%%', startangle=140)\n",
    "plt.axis('equal')\n",
    "plt.title('Subsystem Distribution')\n",
    "plt.savefig('../Plots/pie_chart_subsystems_of_genes_shared_with_secrecon.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "539c0a29-9f68-4f2e-a957-ba17827a9019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pie Chart of subsystems of genes shared between Jahir's Recon and Met Recon (iCHO3000)\n",
    "\n",
    "subsystem_counts = Counter(subs_jahirs_metrecon)\n",
    "labels = list(subsystem_counts.keys())\n",
    "sizes = list(subsystem_counts.values())\n",
    "\n",
    "# Plot the pie chart\n",
    "plt.figure(figsize=(10, 8))\n",
    "plt.pie(sizes, labels=labels, autopct='%1.1f%%', startangle=140)\n",
    "plt.axis('equal') \n",
    "plt.title('Subsystem Distribution')\n",
    "plt.savefig('../Plots/pie_chart_subsystems_of_genes_shared_with_jahirs.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "483bd94f-d4f2-4da4-afae-f446e06df4d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count how many reactions from the rxns_secrecon_metrecon come from Recon3D\n",
    "\n",
    "c=0\n",
    "for i, row in rxns.iterrows():\n",
    "    if (row['iCHO1766']=='-') and (row['iCH02101']=='-') and (row['iCHO2291']=='-'): #Check if the reactions has been added from Recon3D or other databases\n",
    "        r = str(row['Reaction'])\n",
    "        if r in rxns_secrecon_metrecon:\n",
    "            c+=1\n",
    "print(len(rxns_secrecon_metrecon))\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1258b19e-69c1-4a66-80b3-95b614ba0a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra.io import load_matlab_model\n",
    "\n",
    "# Load the .mat file using COBRApy\n",
    "model = load_matlab_model('../Data/Reconciliation/models/model_iCHO2291.mat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63fcf1b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
