{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f793376b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter\n",
    "from functools import reduce\n",
    "from cobra.io import read_sbml_model, load_json_model, load_matlab_model\n",
    "from google_sheet import GoogleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d8b6c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate models for comparison\n",
    "icho1766 = read_sbml_model('../Data/reconciliation/models/iCHOv1_final.xml')\n",
    "icho2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "icho2101 = load_json_model('../Data/Reconciliation/models/iCHO2101.json')\n",
    "recon3d = load_matlab_model('../Data/Reconciliation/models/Recon3D_301.mat')\n",
    "iCHO3000 = load_json_model('iCHOv3_CHO_20022024.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c063085",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate datasets from Google Sheet ----- #####\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "# Read data from the Google Sheet\n",
    "sheet_rxns = 'Rxns'\n",
    "rxns = sheet.read_google_sheet(sheet_rxns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5ccb98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Conf. Score distribution throughout the dataset --- ####\n",
    "data_numeric = pd.to_numeric(rxns['Conf. Score'], errors='coerce')\n",
    "data_numeric = data_numeric.dropna()\n",
    "\n",
    "# Plotting the histogram\n",
    "plt.figure(figsize=(10, 4))\n",
    "plt.hist(data_numeric, bins='auto', color='steelblue', edgecolor='black')\n",
    "plt.grid(color='gray', linestyle='--', linewidth=0.5)\n",
    "plt.xticks(data_numeric.unique())\n",
    "plt.xlabel('Values', fontsize=12, fontweight='bold')\n",
    "plt.ylabel('Frequency', fontsize=12, fontweight='bold')\n",
    "plt.title('Confidence Score Distribution', fontsize=14, fontweight='bold')\n",
    "\n",
    "plt.savefig('../Comparison/conf_score_distribution.png', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67a59651",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Calculate the amount of reactions per Subsystem ---- ####\n",
    "\n",
    "#iCHO 1766\n",
    "subs = []\n",
    "for reaction in icho1766.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho1766_subs = Counter(subs)\n",
    "\n",
    "#iCHO 2291\n",
    "subs = []\n",
    "for reaction in icho2291.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho2291_subs = Counter(subs)\n",
    "\n",
    "# iCHO 2101\n",
    "subs = []\n",
    "for reaction in icho2101.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "icho2101_subs = Counter(subs)\n",
    "\n",
    "# Recon3D\n",
    "subs = []\n",
    "for reaction in recon3d.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "recon3d_subs = Counter(subs)\n",
    "\n",
    "#iCHO 3000\n",
    "subs = []\n",
    "for reaction in iCHO3000.reactions:\n",
    "    subs.append(reaction.subsystem.upper())\n",
    "    \n",
    "iCHO3000_subs = Counter(subs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874ec2d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping of subsystems to categories\n",
    "category_mapping = {\n",
    "    'Amino Acid Metabolism and Derivatives': [\n",
    "        'ARGININE AND PROLINE METABOLISM', 'METHIONINE METABOLISM', 'AMINO ACID DERIVATIVE', \n",
    "        'GLYCINE, SERINE, AND THREONINE METABOLISM', 'HISTIDINE METABOLISM', 'LYSINE METABOLISM', \n",
    "        'METHIONINE AND CYSTEINE METABOLISM', 'TYROSINE METABOLISM', 'VALINE, LEUCINE, AND ISOLEUCINE METABOLISM',\n",
    "        'ALANINE AND ASPARTATE METABOLISM', 'AMINOSUGAR METABOLISM', 'SELENOAMINO ACID METABOLISM',\n",
    "        'GLYCINE, SERINE, ALANINE, AND THREONINE METABOLISM'\n",
    "    ],\n",
    "    'Carbohydrate and Energy Metabolism': [\n",
    "        'GLYCOLYSIS/GLUCONEOGENESIS', 'FRUCTOSE AND MANNOSE METABOLISM', 'CENTRAL CARBON AND ENERGY METABOLISM',\n",
    "        'GALACTOSE METABOLISM', 'PENTOSE AND GLUCURONATE INTERCONVERSIONS', 'STARCH AND SUCROSE METABOLISM',\n",
    "        'PENTOSE PHOSPHATE PATHWAY', 'PYRUVATE METABOLISM'\n",
    "    ],\n",
    "    'Lipid and Fatty Acid Metabolism': [\n",
    "        'FATTY ACID ELONGATION', 'FATTY ACID METABOLISM', 'CHOLESTEROL METABOLISM', 'EICOSANOID METABOLISM',\n",
    "        'ARACHIDONIC ACID METABOLISM', 'BILE ACID SYNTHESIS', 'FATTY ACID SYNTHESIS', 'GLYCEROLIPID METABOLISM',\n",
    "        'SPHINGOLIPID METABOLISM', 'SQUALENE AND CHOLESTEROL SYNTHESIS', 'STEROID METABOLISM', 'TRIACYLGLYCEROL SYNTHESIS',\n",
    "        'VITAMIN A METABOLISM', 'VITAMIN D METABOLISM', 'LINOLEATE METABOLISM', 'FATTY ACID OXIDATION'\n",
    "    ],\n",
    "    'Nucleotide and Nucleic Acid Metabolism': [\n",
    "        'PURINE CATABOLISM', 'PYRIMIDINE BIOSYNTHESIS', 'NUCLEOTIDE SALVAGE PATHWAY', 'NUCLEOTIDES',\n",
    "        'NUCLEOTIDE INTERCONVERSION', 'PURINE SYNTHESIS', 'PYRIMIDINE SYNTHESIS', 'PYRIMIDINE CATABOLISM',\n",
    "        'NUCLEOTIDE SUGAR METABOLISM'\n",
    "    ],\n",
    "    'Vitamin and Cofactor Metabolism': [\n",
    "        'VITAMIN B12 METABOLISM', 'VITAMIN C METABOLISM', 'VITAMIN D', 'VITAMIN A METABOLISM',\n",
    "        'VITAMIN E METABOLISM', 'VITAMIN B2 METABOLISM', 'VITAMIN B6 METABOLISM', 'FOLATE METABOLISM',\n",
    "        'BIOTIN METABOLISM', 'THIAMINE METABOLISM', 'TETRAHYDROBIOPTERIN METABOLISM', 'VITAMIN K METABOLISM',\n",
    "        'VITAMIN AND COFACTOR METABOLISM'\n",
    "    ],\n",
    "    'Glycan, Glycosaminoglycan, and Glycolipid Metabolism': [\n",
    "        'N-GLYCAN BIOSYNTHESIS', 'GLYCOSPHINGOLIPID METABOLISM', 'CHONDROITIN SYNTHESIS', 'KERATAN SULFATE DEGRADATION',\n",
    "        'HEPARAN SULFATE DEGRADATION', 'HYALURONAN METABOLISM', 'N-GLYCAN DEGRADATION', 'O-GLYCAN SYNTHESIS',\n",
    "        'GLYCOSPHINGOLIPID METABOLISM', 'KERATAN SULFATE SYNTHESIS', 'N-GLYCAN SYNTHESIS', 'N-GLYCAN METABOLISM',\n",
    "        'O-GLYCAN METABOLISM'\n",
    "    ],\n",
    "    'Protein and Peptide Metabolism': [\n",
    "        'PROTEIN PRODUCTION', 'PROTEIN ASSEMBLY', 'PROTEIN DEGRADATION', 'PROTEIN MODIFICATION',\n",
    "        'PEPTIDE METABOLISM', 'UBIQUINONE SYNTHESIS', 'HEME SYNTHESIS', 'PROTEIN PRODUCTION'\n",
    "    ],\n",
    "    'Xenobiotic and Drug Metabolism': [\n",
    "        'CYP METABOLISM', 'CYTOCHROME METABOLISM', 'XENOBIOTICS METABOLISM', 'ROS DETOXIFICATION',\n",
    "        'DRUG METABOLISM', 'ALKALOID SYNTHESIS'\n",
    "    ],\n",
    "    'Miscellaneous and Complex Pathways': [\n",
    "        'BIOMASS SYNTHESIS', 'R GROUP SYNTHESIS', 'MISCELLANEOUS', 'BIOMASS AND MAINTENANCE FUNCTIONS',\n",
    "        'UNASSIGNED', 'OTHER CARBON METABOLISM'\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a469032",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert Counters to DataFrames\n",
    "df_icho1766 = pd.DataFrame.from_dict(icho1766_subs, orient='index', columns=['iCHO1766'])\n",
    "df_icho2291 = pd.DataFrame.from_dict(icho2291_subs, orient='index', columns=['iCHO2291'])\n",
    "df_icho2101 = pd.DataFrame.from_dict(icho2101_subs, orient='index', columns=['iCHO2101'])\n",
    "df_recon3d = pd.DataFrame.from_dict(recon3d_subs, orient='index', columns=['Recon3D'])\n",
    "df_iCHO3000 = pd.DataFrame.from_dict(iCHO3000_subs, orient='index', columns=['iCHO3000'])\n",
    "\n",
    "# Combine the DataFrames\n",
    "combined_df = pd.concat([df_icho1766, df_icho2101, df_icho2291, df_recon3d, df_iCHO3000], axis=1).fillna(0)\n",
    "\n",
    "# Create a mask where each row is True if all its values are non-zero\n",
    "#non_zero_mask = (combined_df != 0).all(axis=1)\n",
    "\n",
    "# Apply the mask to filter out rows with any zero values\n",
    "#filtered_df = combined_df[non_zero_mask]\n",
    "\n",
    "# Create a mask where each row is True if the index does not start with \"Transport\"\n",
    "mask = ~combined_df.index.str.startswith('TRANSPORT') & ~combined_df.index.str.startswith(\"EXCHANGE\") & ~combined_df.index.str.startswith(\"SINK\") & ~combined_df.index.str.startswith(\"DEMAND\") & ~combined_df.index.str.startswith(\"INTRACELLULAR\") & ~combined_df.index.str.startswith(\"EXTRACELLULAR\")\n",
    "\n",
    "\n",
    "# Apply the mask to the DataFrame to keep only rows that do not start with \"Transport\"\n",
    "filtered_df = combined_df[mask]\n",
    "\n",
    "# Plot \n",
    "ax = filtered_df.plot(kind='barh', figsize=(20, 25), width=1.8) \n",
    "\n",
    "# Title and labels\n",
    "plt.title('Subsystem Reaction Counts Across Different CHO Recons', fontsize=28)\n",
    "plt.ylabel('Subsystem', fontsize=26)  # This was previously 'xlabel'\n",
    "plt.xlabel('Reaction Count', fontsize=26)  # This was previously 'ylabel'\n",
    "\n",
    "# Adjust the layout\n",
    "plt.tight_layout()  \n",
    "\n",
    "# Customize legend\n",
    "plt.legend(fontsize=24)\n",
    "\n",
    "# Save the plot to a file\n",
    "plt.savefig('../Comparison/subsystem_reaction_counts.png', dpi=300)  \n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5bc92ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reaction counts for each recon\n",
    "icho1766_rxns = len(icho1766.reactions)\n",
    "icho2291_rxns = len(icho2291.reactions)\n",
    "icho2101_rxns = len(icho2101.reactions)\n",
    "recon3d_rxns = len(recon3d.reactions)\n",
    "icho3000_rxns = len(iCHO3000.reactions)\n",
    "\n",
    "#Metabolite counts for each recon\n",
    "icho1766_mets = len(icho1766.metabolites)\n",
    "icho2291_mets = len(icho2291.metabolites)\n",
    "icho2101_mets = len(icho2101.metabolites)\n",
    "recon3d_mets = len(recon3d.metabolites)\n",
    "icho3000_mets = len(iCHO3000.metabolites)\n",
    "\n",
    "# Gene counts for each recon\n",
    "icho1766_gens = len(icho1766.genes)\n",
    "icho2291_gens = len(icho2291.genes)\n",
    "icho2101_gens = len(icho2101.genes)\n",
    "recon3d_gens = len(recon3d.genes)\n",
    "icho3000_gens = len(iCHO3000.genes)\n",
    "\n",
    "# Subsystem count for each recon\n",
    "icho1766_subs_count = len(icho1766_subs)\n",
    "icho2291_subs_count = len(icho2291_subs)\n",
    "icho2101_subs_count = len(icho2101_subs)\n",
    "recon3d_subs_count = len(recon3d_subs)\n",
    "icho3000_subs_count = len(iCHO3000_subs)\n",
    "\n",
    "\n",
    "# Creating a DataFrame from the counts\n",
    "data = {\n",
    "    'iCHO1766': [icho1766_rxns, icho1766_mets, icho1766_gens, icho1766_subs_count],\n",
    "    'iCHO2101': [icho2101_rxns, icho2101_mets, icho2101_gens, icho2101_subs_count],\n",
    "    'iCHO2291': [icho2291_rxns, icho2291_mets, icho2291_gens, icho2291_subs_count],\n",
    "    'iCHO3000': [icho3000_rxns, icho3000_mets, icho3000_gens, icho3000_subs_count],\n",
    "    'Recon3D': [recon3d_rxns, recon3d_mets, recon3d_gens, recon3d_subs_count]\n",
    "}\n",
    "\n",
    "counts_df = pd.DataFrame(data, index=['Reactions', 'Metabolites', 'Genes', 'Subsystems'])\n",
    "\n",
    "# Defining colors for the bars\n",
    "colors = ['blue', 'orange', 'green', 'red','purple']\n",
    "\n",
    "# Creating subplots\n",
    "fig, axes = plt.subplots(nrows=2, ncols=2, figsize=(12, 11))\n",
    "\n",
    "# A function to annotate the bars with their values in bold and matching color\n",
    "def annotate_bars(ax, colors):\n",
    "    for i, p in enumerate(ax.patches):  # Loop over every bar, with index for color\n",
    "        ax.annotate(f'{p.get_height()}',  # Get the height of the bar (the data value)\n",
    "                    (p.get_x() + p.get_width() / 2., p.get_height()),  # Position for the text\n",
    "                    ha='center',  # Center align\n",
    "                    va='center',  # Center align vertically\n",
    "                    xytext=(0, 9),  # 9 points vertical offset\n",
    "                    textcoords='offset points',\n",
    "                    fontweight='bold',  # Make text bold\n",
    "                    color=colors[i % len(colors)])  # Match text color to the bar color\n",
    "\n",
    "# Plotting and annotating each category with bold labels and matching colors\n",
    "counts_df.loc['Reactions'].plot(kind='bar', ax=axes[0,0], title='Reactions', color=colors)\n",
    "annotate_bars(axes[0,0], colors)\n",
    "\n",
    "counts_df.loc['Metabolites'].plot(kind='bar', ax=axes[0,1], title='Metabolites', color=colors)\n",
    "annotate_bars(axes[0,1], colors)\n",
    "\n",
    "counts_df.loc['Genes'].plot(kind='bar', ax=axes[1,0], title='Genes', color=colors)\n",
    "annotate_bars(axes[1,0], colors)\n",
    "\n",
    "counts_df.loc['Subsystems'].plot(kind='bar', ax=axes[1,1], title='Subsystems', color=colors)\n",
    "annotate_bars(axes[1,1], colors)\n",
    "\n",
    "# Adjusting layout and rotating x-axis labels for all subplots uniformly\n",
    "for ax in axes.flatten():\n",
    "    ax.tick_params(axis='x', rotation=45)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Comparison/general_metrics.png', dpi=300)  \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "151676d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "list1 = [re.sub(r'\\[.*?\\]', '', m.id) for m in recon3d.metabolites]\n",
    "set1 = set(list1)\n",
    "print(len(list1))\n",
    "print(len(set1))\n",
    "print('----------------------------------------')\n",
    "list2 = [m.id[:-2] for m in icho1766.metabolites]\n",
    "set2 = set(list2)\n",
    "print(len(list2))\n",
    "print(len(set2))\n",
    "print('----------------------------------------')\n",
    "list3 = [m.id[:-2] for m in icho2101.metabolites]\n",
    "set3 = set(list3)\n",
    "print(len(list3))\n",
    "print(len(set3))\n",
    "print('----------------------------------------')\n",
    "list4 = [m.id[:-2] for m in icho2291.metabolites]\n",
    "set4 = set(list4)\n",
    "print(len(list4))\n",
    "print(len(set4))\n",
    "print('----------------------------------------')\n",
    "list5 = [m.id[:-2] for m in iCHO3000.metabolites]\n",
    "set5 = set(list5)\n",
    "print(len(list5))\n",
    "print(len(set5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b2f65f",
   "metadata": {},
   "outputs": [],
   "source": [
    "list2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63261e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "icho3000_subs_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a44a2404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a dictionary to store the counts\n",
    "counts = {\n",
    "    'Model': [],\n",
    "    'Reactions with GPR': [],\n",
    "    'Reactions without GPR': [],\n",
    "    'Total Reactions': []\n",
    "}\n",
    "\n",
    "# Loop through the models and calculate counts\n",
    "models = [icho1766, icho2291, icho2101, iCHO3000, recon3d]\n",
    "model_names = ['iCHO1766', 'iCHO2291', 'iCHO2101', 'iCHO3000', 'Recon3D']\n",
    "\n",
    "for model, name in zip(models, model_names):\n",
    "    rxn_with_gpr = sum(1 for rxn in model.reactions if str(rxn.gpr) not in ['', 'nan'])\n",
    "    rxn_without_gpr = sum(1 for rxn in model.reactions if str(rxn.gpr) in ['', 'nan'])\n",
    "    total = len(model.reactions)\n",
    "\n",
    "    counts['Model'].append(name)\n",
    "    counts['Reactions with GPR'].append(rxn_with_gpr)\n",
    "    counts['Reactions without GPR'].append(rxn_without_gpr)\n",
    "    counts['Total Reactions'].append(total)\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "counts_df = pd.DataFrame(counts)\n",
    "counts_df['%'] = (counts_df['Reactions with GPR']/counts_df['Total Reactions'])*100\n",
    "\n",
    "# Setting the index to the model names\n",
    "counts_df.set_index('Model', inplace=True)\n",
    "\n",
    "# Plotting stacked bar chart\n",
    "ax = counts_df[['Reactions with GPR','Reactions without GPR']].plot(kind='bar', stacked=True, figsize=(8, 6))\n",
    "\n",
    "# Adding titles and labels\n",
    "plt.title('GPR Coverage in each Recon')\n",
    "plt.ylabel('Reaction Count')\n",
    "\n",
    "ax.tick_params(axis='x', rotation=45)\n",
    "\n",
    "plt.savefig('../Comparison/gpr_coverage.png', dpi=300)  \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b9063c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc952939",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
