{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Tests\n",
    "\n",
    "This notebook is intended for running different kinds of analyses that would validate our reconstruction.\n",
    "\n",
    "[1. Subsystem Overview and Analysis](#subsystems) <br>\n",
    "[2. Context-specific Model Generation](#context_specific) <br>\n",
    "[3. Biomass prediction using exp. data](#biomass) <br>\n",
    "[4. Flux Enrichment Analysis](#fea) <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Subsystem Overview and Analysis <a id='subsystems'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from google_sheet import GoogleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate Subsystems datasets from Google Sheet ----- #####\n",
    "\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_subsystems = 'Summary Systems'\n",
    "subsystems = sheet.read_google_sheet(sheet_subsystems)\n",
    "\n",
    "# Remove the total count\n",
    "subsystems = subsystems.iloc[:-1, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### --- Pie Chart of the Amount of Reactions per System --- ####\n",
    "\n",
    "subsystems[\"Number of Reactions\"] = pd.to_numeric(subsystems[\"Number of Reactions\"])\n",
    "\n",
    "# Aggregate data by \"System\" and sum \"Number of Reactions\"\n",
    "system_reactions = subsystems.groupby(\"System\")[\"Number of Reactions\"].sum()\n",
    "\n",
    "# Set the figure and axes for more control\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "\n",
    "# Generate the pie chart again with the custom labels\n",
    "patches, texts, autotexts = ax.pie(system_reactions, labels=system_reactions.index, autopct=lambda p: '{:.1f}%'.format(p) if p > 0 else '', startangle=140, colors=plt.cm.tab20.colors)\n",
    "\n",
    "# Improve aesthetics\n",
    "for text in texts + autotexts:\n",
    "    text.set_color('black')\n",
    "ax.set_title(\"Pie Chart of Number of Reactions by System\", pad=30)  # Move the title upwards by increasing pad\n",
    "\n",
    "# Set equal aspect ratio\n",
    "ax.axis('equal')\n",
    "\n",
    "# Save the plot\n",
    "plt.savefig('../Plots/pie_chart_reactions_per_system.png')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the Sunburst Plot\n",
    "\n",
    "fig = px.sunburst(subsystems, path=['System', 'Subsystems', 'Kegg Pathway'],\n",
    "                  color='System')\n",
    "\n",
    "# Convert to a Graph Objects figure\n",
    "fig_go = go.Figure(fig)\n",
    "\n",
    "# Update layout\n",
    "fig_go.update_layout(width=1200, height=1000)\n",
    "\n",
    "# Define font sizes\n",
    "font_size_last_layer = 10\n",
    "font_size_previous_layer = 15\n",
    "default_font_size = 12\n",
    "\n",
    "# Create a list to store font sizes\n",
    "font_sizes = []\n",
    "\n",
    "# Update font size for each level\n",
    "for trace in fig_go.data:\n",
    "    if isinstance(trace, go.Sunburst):\n",
    "        for id in trace.ids:\n",
    "            level = id.count(\"/\")  # Determine level by the number of slashes in the id\n",
    "            if level == 2:  # Last layer (Kegg Pathway)\n",
    "                font_sizes.append(font_size_last_layer)\n",
    "            elif level == 1:  # Previous layer (Subsystems)\n",
    "                font_sizes.append(font_size_previous_layer)\n",
    "            else:\n",
    "                font_sizes.append(default_font_size)  # Default size for other layers\n",
    "\n",
    "# Apply the font sizes to the figure\n",
    "fig_go.update_traces(insidetextfont=dict(size=font_sizes))\n",
    "\n",
    "# Save the figure\n",
    "fig_go.write_html(\"../Plots/sunburst_subsystems.html\")  # Save as interactive HTML file\n",
    "fig_go.write_image(\"../Plots/sunburst_subsystems.png\", width=1200, height=1000, scale=2)  # Increase resolution by setting scale parameter\n",
    "\n",
    "# Show the plot\n",
    "fig_go.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Auxotrophies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.io import load_json_model\n",
    "iCHO_path = \"iCHOv3_CHO_04032024.json\"\n",
    "iCHO = load_json_model(iCHO_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids = {\n",
    "    \"arginine\": [\"EX_arg_L_e\", \"EX_arg_D_e\"],\n",
    "    \"asparagine\": [\"EX_asn_L_e\", \"EX_asn_D_e\"],\n",
    "    \"cysteine\": [\"EX_cys_L_e\", \"EX_cys_D_e\", \"EX_Lcystin_e\"],\n",
    "    \"histidine\": [\"EX_his_L_e\", \"EX_his_D_e\"],\n",
    "    \"isoleucine\": [\"EX_ile_L_e\", \"EX_ile_D_e\"],\n",
    "    \"leucine\": [\"EX_leu_L_e\", \"EX_leu_D_e\"],\n",
    "    \"lysine\": [\"EX_lys_L_e\", \"EX_lys_D_e\"],\n",
    "    \"methionine\": [\"EX_met_L_e\", \"EX_met_D_e\"],\n",
    "    \"phenylalanine\": [\"EX_phe_L_e\", \"EX_phe_D_e\"],\n",
    "    \"proline\": [\"EX_pro_L_e\", \"EX_pro_D_e\"],\n",
    "    \"threonine\": [\"EX_thr_L_e\", \"EX_thr_D_e\"],\n",
    "    \"tryptophan\": [\"EX_trp_L_e\", \"EX_trp_D_e\"],\n",
    "    \"valine\": [\"EX_val_L_e\", \"EX_val_D_e\"]\n",
    "}\n",
    "for amino_acid in amino_acids:\n",
    "    # ----- Setup initial bounds -----\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        exchange_reaction.bounds = -10, 10\n",
    "\n",
    "    # Arginine\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00557').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R00558').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GLYAMDTRc').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('GAUGE-R10107').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ARGSL').bounds = 0, 0\n",
    "\n",
    "    # Asparigine\n",
    "    iCHO.reactions.get_by_id('ASNS1').bounds = 0, 0\n",
    "\n",
    "    # Cysteine\n",
    "    iCHO.reactions.get_by_id('r0129').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_cgly_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECG').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('AMPTASECGe').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('CYSTGL').bounds = -10, 0\n",
    "    iCHO.reactions.get_by_id('EX_HC00250_e').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('EX_sfcys_e').bounds = 0, 0\n",
    "\n",
    "    # Histidine\n",
    "    iCHO.reactions.get_by_id('VALTRPVALr').bounds = 0,0 \n",
    "\n",
    "    # Isoleucine - DONE\n",
    "    iCHO.reactions.get_by_id('EX_CE2916_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2915_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('ILETA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('ILETAm').bounds = 0, 10\n",
    "\n",
    "    # Leucine\n",
    "    iCHO.reactions.get_by_id('LEULEULAPc').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_leugly_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('EX_glyleu_e').bounds = 0,0 \n",
    "    iCHO.reactions.get_by_id('LEUTA').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('LEUTAm').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('EX_CE5797_e').bounds = 0, 0\n",
    "\n",
    "    # Lysine\n",
    "    iCHO.reactions.get_by_id('EX_biocyt_e').bounds = 0,0 \n",
    "\n",
    "    # Methionine\n",
    "    iCHO.reactions.get_by_id('METS').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('BHMT').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('GAUGE-R00648').bounds = 0, 10 \n",
    "    iCHO.reactions.get_by_id('UNK2').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('UNK3').bounds = -10, 0 \n",
    "    iCHO.reactions.get_by_id('TYRA').bounds = -10, 0 \n",
    "    #iCHO.reactions.get_by_id('GAUGE-R06895').bounds = 0, 0 # Curated by MR with 1 score / Erased from the reconstruction\n",
    "\n",
    "    # Phenylalanine\n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_pheleu_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_glyphe_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE2917_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5786_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_CE5789_e').bounds = 0, 0 \n",
    "    iCHO.reactions.get_by_id('EX_phpyr_e').bounds = 0, 0 \n",
    "\n",
    "    # Proline\n",
    "    iCHO.reactions.get_by_id('EX_glypro_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_progly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('P5CR').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRxm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRx').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('P5CRm').bounds = 0, 0\n",
    "    iCHO.reactions.get_by_id('r1453').bounds = 0, 10\n",
    "\n",
    "    # Threonine\n",
    "    #iCHO.reactions.get_by_id('THRS').bounds = 0,0  # Erased from the reconstruction\n",
    "\n",
    "\n",
    "    # Tryptophan\n",
    "    iCHO.reactions.get_by_id('NBAHH_ir').bounds = 0,0  \n",
    "\n",
    "    # Valine\n",
    "    iCHO.reactions.get_by_id('EX_valarggly_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_vallystyr_e').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTA').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('VALTAm').bounds = 0, 10\n",
    "    iCHO.reactions.get_by_id('EX_valval_e').bounds = 0, 10\n",
    "    # -----------------------------------------\n",
    "\n",
    "    for exchange_reaction in iCHO.exchanges:\n",
    "        if exchange_reaction.id in amino_acids[amino_acid]:\n",
    "            exchange_reaction.bounds = 0, 10\n",
    "            # print(amino_acid, exchange_reaction.id, amino_acids[amino_acid])\n",
    "    sol = iCHO.optimize()\n",
    "    print(amino_acid, sol.objective_value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acids = {\n",
    "    \"alanine\": [\"EX_ala_B_e\", \"EX_ala_L_e\", \"EX_ala_D_e\"],\n",
    "    \"arginine\": [\"EX_arg_L_e\", \"EX_arg_D_e\"],\n",
    "    \"asparagine\": [\"EX_asn_L_e\", \"EX_asn_D_e\"],\n",
    "    \"aspartic acid\": [\"EX_asp_L_e\", \"EX_asp_D_e\"],\n",
    "    \"cysteine\": [\"EX_cys_L_e\", \"EX_cys_D_e\"],\n",
    "    \"glutamine\": [\"EX_gln_L_e\", \"EX_gln_D_e\"],\n",
    "    \"glutamic acid\": [\"EX_glu_L_e\", \"EX_glu_D_e\"],\n",
    "    \"glycine\": [\"EX_gly_e\"],\n",
    "    \"histidine\": [\"EX_his_L_e\", \"EX_his_D_e\"],\n",
    "    \"isoleucine\": [\"EX_ile_L_e\", \"EX_ile_D_e\"],\n",
    "    \"leucine\": [\"EX_leu_L_e\", \"EX_leu_D_e\"],\n",
    "    \"lysine\": [\"EX_lys_L_e\", \"EX_lys_D_e\"],\n",
    "    \"methionine\": [\"EX_met_L_e\", \"EX_met_D_e\"],\n",
    "    \"phenylalanine\": [\"EX_phe_L_e\", \"EX_phe_D_e\"],\n",
    "    \"proline\": [\"EX_pro_L_e\", \"EX_pro_D_e\"],\n",
    "    \"serine\": [\"EX_ser_L_e\", \"EX_ser_D_e\"],\n",
    "    \"threonine\": [\"EX_thr_L_e\", \"EX_thr_D_e\"],\n",
    "    \"tryptophan\": [\"EX_trp_L_e\", \"EX_trp_D_e\"],\n",
    "    \"tyrosine\": [\"EX_tyr_L_e\", \"EX_tyr_D_e\"],\n",
    "    \"valine\": [\"EX_val_L_e\", \"EX_val_D_e\"]\n",
    "}\n",
    "\n",
    "def test_auxotrophy(model, amino_acids):\n",
    "    # Iterate through each amino acid and its corresponding exchange reactions\n",
    "    for amino_acid, reactions in amino_acids.items():\n",
    "        # Set the bounds for each exchange reaction\n",
    "        for exchange in model.exchanges:\n",
    "            if exchange.id in reactions:\n",
    "                exchange.bounds = (0, 10)  # Allow secretion but not uptake\n",
    "            else:\n",
    "                exchange.bounds = (-10, 10)  # Allow uptake and secretion\n",
    "\n",
    "        # Optimize the model for the current setup\n",
    "        solution = model.optimize()\n",
    "        if solution.status == \"optimal\":\n",
    "            print(f\"{amino_acid} biomass production: {solution.objective_value}\")\n",
    "        else:\n",
    "            print(f\"{amino_acid} optimization failed: {solution.status}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_auxotrophy(iCHO, amino_acids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCHO.metabolites.leu_L_c.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matrix Sparsity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Deletion test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Essential for growth genes and comparison to (An optimised genome-wide, virus free CRISPR screen for mammalian cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Commparsion to other models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Context-specific Model Generation <a id='context_specific'></a>\n",
    "\n",
    "Here we use a matrix generated with rmf_CADRE to generate each context_specific model for each one of the conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.io\n",
    "from cobra.io import load_json_model, save_json_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the matrix\n",
    "matrix = scipy.io.loadmat('../Data/Context_specific_models/CHO_retained_reactions.mat')['retR']\n",
    "\n",
    "# Load conditions\n",
    "conditions = open('../Data/Context_specific_models/uniqueConditions.txt', 'r').read().strip().split(',')\n",
    "\n",
    "# Load model\n",
    "model = load_json_model('iCHO3644_unblocked.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over each condition to generate a context-specific model for each one\n",
    "\n",
    "for i, condition in enumerate(conditions):\n",
    "    \n",
    "    # Create a copy of the model for each condition\n",
    "    submodel = model.copy()\n",
    "    \n",
    "    # Assign a unique name based on the condition index\n",
    "    submodel.id = condition\n",
    "    \n",
    "    # List to collect reactions to remove\n",
    "    reactions_to_remove = [submodel.reactions[j].id for j in range(matrix.shape[0]) if matrix[j, i] == 0]\n",
    "    \n",
    "    # Remove the collected reactions from the submodel\n",
    "    submodel.remove_reactions(reactions_to_remove, remove_orphans=False)\n",
    "\n",
    "    # Save each model as JSON with a unique name based on the condition\n",
    "    model_name_json = f'../Data/Context_specific_models/iCHO_{condition}.json'\n",
    "    save_json_model(submodel, model_name_json)\n",
    "    \n",
    "    print(f'Model {condition} generated, {i}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Biomass prediction using exp. data <a id='biomass'></a>\n",
    "\n",
    "Here we incorporate uptake and secretion rates from (https://www.sciencedirect.com/science/article/pii/S2405471216303635?via%3Dihub) data and calculate the growth rate (Figure 3). Reproduce Hooman's analsys with iCHO_v1 and incorporate iCHO2291 and iCHO2441 into the analysis.\n",
    "\n",
    "To do:\n",
    "- Try ecFBA with enzyme trunover numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "import cobra\n",
    "from cobra.io import load_json_model\n",
    "from cobra.exceptions import Infeasible"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-04\n"
     ]
    }
   ],
   "source": [
    "# Load model\n",
    "zela_models_path = '../Data/Context_specific_models/'\n",
    "\n",
    "model_path = os.path.join(zela_models_path, 'iCHO_WT_P4.json')\n",
    "\n",
    "#model = load_json_model(model_path)\n",
    "model = load_json_model('iCHO3595.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCHO3595</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14c6e48c0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>7379</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>11107</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>3595</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*biomass_cho_s - 1.0*biomass_cho_s_reverse_dd74e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>m, x, c, r, e, n, l, im, g</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCHO3595 at 0x14c6e48c0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.solver = 'gurobi'\n",
    "objective = 'biomass_cho_s' # 'biomass_cho_s'\n",
    "model.objective = objective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rxn in model.boundary:\n",
    "    # IMPORTANT, all ex should be 0 - 1000, except those that we know are being uptaken by the cell\n",
    "\n",
    "    if rxn.id.startswith(\"EX_\"):\n",
    "        rxn.bounds = (0,1000) \n",
    "    if rxn.id.startswith(\"SK_\"):\n",
    "        rxn.bounds = (-1000,1000)\n",
    "    if rxn.id.startswith(\"DM_\"):\n",
    "        rxn.bounds = (0,1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load uptake and secretion rate dict\n",
    "\n",
    "with open('../Data/Uptake_Secretion_Rates/uptake_secretion_wt_dict.pkl', 'rb') as file:\n",
    "    uptsec_wt = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EX_bhb_e': {'P0 to P2': (0.005655231264380309, 0.01150591140447349),\n",
       "  'P12 to P14': (-9.098161773900422e-05, 0.001288544532797985),\n",
       "  'P2 to P4': (0.0032407909069092273, 0.004300936477247652),\n",
       "  'P4 to P6': (0.0003266460537807496, 0.0004614615248303923),\n",
       "  'P6 to P8': (-0.2665735327935187, 0.07757795799802666),\n",
       "  'P8 to P12': (0.0007852544859507166, 0.0015525178701479656)},\n",
       " 'EX_nh4_e': {'P0 to P2': (0.22222867666762844, 0.39132057600739256),\n",
       "  'P12 to P14': (-0.003800168451533614, 0.026463820759513784),\n",
       "  'P2 to P4': (0.09331646648603117, 0.11649394271188668),\n",
       "  'P4 to P6': (-0.006076639627319469, -0.0011890208837271985),\n",
       "  'P6 to P8': (-1.8795404696012334, 7.428980544633633),\n",
       "  'P8 to P12': (-0.010648232481031548, 0.0078826156522462)},\n",
       " 'EX_ac_e': {'P0 to P2': (0.00868399590224795, 0.022497025533987827),\n",
       "  'P12 to P14': (0.0506274555815184, 0.10982173713343438),\n",
       "  'P2 to P4': (0.0010842194486317125, 0.0015281127416154252),\n",
       "  'P4 to P6': (0.00536147580892445, 0.007227301085920544),\n",
       "  'P6 to P8': (-6.671733779415508, 26.22308317113306),\n",
       "  'P8 to P12': (0.059466820892660865, 0.09237169010692634)},\n",
       " 'EX_ala_L_e': {'P0 to P2': (0.10109706560929384, 0.1625346338852227),\n",
       "  'P12 to P14': (-0.014835039364822564, 0.012281727888971397),\n",
       "  'P2 to P4': (0.04798552563859218, 0.05653967493581927),\n",
       "  'P4 to P6': (0.0015052396619153982, 0.005891728840953991),\n",
       "  'P6 to P8': (-8.70574500806483, 2.2153806897774273),\n",
       "  'P8 to P12': (-0.004000467781606238, 0.014714748954061576)},\n",
       " 'EX_arg_L_e': {'P0 to P2': (-0.030205767530839003, -0.008142040856646594),\n",
       "  'P12 to P14': (0.00041352488305923, 0.0019312931487869022),\n",
       "  'P2 to P4': (-0.01884898170723418, -0.01529801058321137),\n",
       "  'P4 to P6': (-0.004448005742967942, -0.0028914141561779167),\n",
       "  'P6 to P8': (-1.256547206446597, 0.3220021149823486),\n",
       "  'P8 to P12': (-0.0005362804177535141, -0.0004202678211244772)},\n",
       " 'EX_asn_L_e': {'P0 to P2': (-0.2679529718873822, -0.09850346806118511),\n",
       "  'P12 to P14': (0.0, 0.0),\n",
       "  'P2 to P4': (-0.10843336015938503, -0.09549559643737511),\n",
       "  'P4 to P6': (-0.017499102788538792, -0.012452962482681467),\n",
       "  'P6 to P8': (-5.17686180508861, 1.3338464844464537),\n",
       "  'P8 to P12': (-0.00022237785450854135, -1.5820650177202585e-05)},\n",
       " 'EX_asp_L_e': {'P0 to P2': (0.02003875835541253, 0.03759929226132892),\n",
       "  'P12 to P14': (-0.004092478976382203, 0.0003392707572462785),\n",
       "  'P2 to P4': (0.01456951555293158, 0.02087295020206143),\n",
       "  'P4 to P6': (-0.0018625064200347664, -0.0011822377443128826),\n",
       "  'P6 to P8': (-3.3588995099160384, 0.8699390515531513),\n",
       "  'P8 to P12': (-0.0071251757424055885, -0.004905606215412872)},\n",
       " 'EX_2hb_e': {'P0 to P2': (0.007716904554183902, 0.01262195102342596),\n",
       "  'P12 to P14': (-0.000686152990636071, 0.0007103605379753307),\n",
       "  'P2 to P4': (0.0011330468691121624, 0.001651296609250617),\n",
       "  'P4 to P6': (0.0007219178221070849, 0.0014819973575084023),\n",
       "  'P6 to P8': (-0.3697613132699009, 1.435938797276438),\n",
       "  'P8 to P12': (0.0006925971777202027, 0.0012942032884971678)},\n",
       " 'EX_cit_e': {'P0 to P2': (0.0, 0.0),\n",
       "  'P12 to P14': (-0.0016108204198825274, 0.00016844207729431937),\n",
       "  'P2 to P4': (0.005959799128689479, 0.0075622343179195094),\n",
       "  'P4 to P6': (0.001751782413244022, 0.003026470057966634),\n",
       "  'P6 to P8': (-0.8051481751404139, 3.051668109831022),\n",
       "  'P8 to P12': (-0.0021204663436438496, -0.0008200244433440146)},\n",
       " 'EX_cys_L_e': {'P0 to P2': (-0.015130653276152045, -0.005133889755128801),\n",
       "  'P12 to P14': (0.0, 0.0),\n",
       "  'P2 to P4': (-0.003721885418836899, -0.0033091501094601413),\n",
       "  'P4 to P6': (-0.0003875855367568113, -0.0002713752478577518),\n",
       "  'P6 to P8': (-0.18326121499922746, 0.04716547916688882),\n",
       "  'P8 to P12': (0.0, 0.0)},\n",
       " 'EX_etoh_e': {'P0 to P2': (0.02694955502986522, 0.07149898700631213),\n",
       "  'P12 to P14': (-0.011864785484526179, -0.001728482964384384),\n",
       "  'P2 to P4': (0.016006042606443296, 0.07464951555114036),\n",
       "  'P4 to P6': (-0.015312380779900006, 0.026471966799530088),\n",
       "  'P6 to P8': (-23.88997520642353, 6.4305036465783605),\n",
       "  'P8 to P12': (-0.007448005494071619, 9.683280193895907e-05)},\n",
       " 'EX_for_e': {'P0 to P2': (0.05600483553443332, 0.10247819996427192),\n",
       "  'P12 to P14': (-0.003937463006791617, -0.0009459641908630791),\n",
       "  'P2 to P4': (0.027652376456539285, 0.030956496907215074),\n",
       "  'P4 to P6': (0.0065304796688588004, 0.010365972264627484),\n",
       "  'P6 to P8': (-0.23991863449349926, 0.9487181072651109),\n",
       "  'P8 to P12': (-0.0059773474606671425, -0.002749640625498923)},\n",
       " 'EX_fum_e': {'P0 to P2': (0.0, 0.0),\n",
       "  'P12 to P14': (-8.493786462642557e-06, 0.0007435982439755661),\n",
       "  'P2 to P4': (-0.00016066431702818767, 0.0005996073941125498),\n",
       "  'P4 to P6': (-0.0001232258923221273, 8.909881037937689e-05),\n",
       "  'P6 to P8': (-0.055295629083251946, 0.21196972641387316),\n",
       "  'P8 to P12': (0.0007719200562517781, 0.001335060183264195)},\n",
       " 'EX_glc_e': {'P0 to P2': (-0.8318325272622697, 0.31178475942430534),\n",
       "  'P12 to P14': (-0.22865934749647626, -0.09387716341800657),\n",
       "  'P2 to P4': (-0.5853348558421616, -0.4953355298825426),\n",
       "  'P4 to P6': (-0.14526279391260682, -0.1002207947163189),\n",
       "  'P6 to P8': (-87.55214917450274, 21.42663581736379),\n",
       "  'P8 to P12': (-0.29548368260042146, -0.2237217776618675)},\n",
       " 'EX_glu_L_e': {'P0 to P2': (0.009043232958865501, 0.03241492082077512),\n",
       "  'P12 to P14': (-0.0014012678829840594, -0.00047771792712913),\n",
       "  'P2 to P4': (0.009995190342042689, 0.015809435557194437),\n",
       "  'P4 to P6': (0.00028455104732210193, 0.0020303045849644724),\n",
       "  'P6 to P8': (-0.09996449832765389, 0.3591762498064759),\n",
       "  'P8 to P12': (-0.0005416177851294882, 0.004068101062076921)},\n",
       " 'EX_gln_L_e': {'P0 to P2': (-0.6764945248112806, -0.3292614760637263),\n",
       "  'P12 to P14': (-0.00041187470147167643, 0.000346480040940936),\n",
       "  'P2 to P4': (-0.11022278442041394, -0.10686085702946083),\n",
       "  'P4 to P6': (-0.005722665144154624, -0.0038519867751049867),\n",
       "  'P6 to P8': (-0.04363569848727934, 0.1628506437782829),\n",
       "  'P8 to P12': (0.00018127928252203044, 0.00025277044914864995)},\n",
       " 'EX_gly_e': {'P0 to P2': (0.017707140671783916, 0.03297828122848983),\n",
       "  'P12 to P14': (0.005337951451007323, 0.011954768915574215),\n",
       "  'P2 to P4': (0.011664762238946842, 0.015361887660412769),\n",
       "  'P4 to P6': (0.0042410756662695686, 0.006050612615114485),\n",
       "  'P6 to P8': (-0.23241724215785514, 0.9373000083439911),\n",
       "  'P8 to P12': (0.00022817120458928052, 0.005268858284933833)},\n",
       " 'EX_his_L_e': {'P0 to P2': (-0.011511873507846136, -0.00018226215201998502),\n",
       "  'P12 to P14': (-0.0008840695912452133, 0.0013606421810049927),\n",
       "  'P2 to P4': (-0.006798969568476408, -0.00596715768970421),\n",
       "  'P4 to P6': (-0.0019697207739715005, -0.0011967513553266171),\n",
       "  'P6 to P8': (-0.42018227476571685, 0.10544728693559785),\n",
       "  'P8 to P12': (-0.000154232243948385, -2.8444405536967628e-05)},\n",
       " 'EX_4hpro_e': {'P0 to P2': (-0.006640765660352492, -0.004809550762887639),\n",
       "  'P12 to P14': (0.0028364596512384433, 0.004999057270105377),\n",
       "  'P2 to P4': (-0.004224254317618865, -0.0006331999506674355),\n",
       "  'P4 to P6': (-0.0009221439123288048, -0.00020656095027660431),\n",
       "  'P6 to P8': (-0.0456134045635382, 0.015446392503012179),\n",
       "  'P8 to P12': (-0.0008966459649466831, 5.3021996389985106e-05)},\n",
       " 'EX_ile_L_e': {'P0 to P2': (-0.018294750931403085, -0.011377500173010442),\n",
       "  'P12 to P14': (-0.0001882029823187732, 0.0008574119444352229),\n",
       "  'P2 to P4': (-0.01610170921321085, -0.013334379454487769),\n",
       "  'P4 to P6': (-0.00501075189733004, -0.003442323203996545),\n",
       "  'P6 to P8': (-2.6282023886288544, 0.6813784291701158),\n",
       "  'P8 to P12': (-0.002463367421231854, -0.0006620843540461773)},\n",
       " 'EX_lac_L_e': {'P0 to P2': (1.081459798843718, 1.843911238054472),\n",
       "  'P12 to P14': (-0.0047621134947709726, 0.23134280417406494),\n",
       "  'P2 to P4': (0.7184475137252582, 0.8108292756320108),\n",
       "  'P4 to P6': (0.06918220805329268, 0.07441260173406637),\n",
       "  'P6 to P8': (-49.809207443868885, 14.644278321958172),\n",
       "  'P8 to P12': (0.14549861887715645, 0.30880996122525184)},\n",
       " 'EX_leu_L_e': {'P0 to P2': (-0.03352615341252578, -0.018207522707210756),\n",
       "  'P12 to P14': (0.00020145042741478379, 0.0010724266959048845),\n",
       "  'P2 to P4': (-0.026137250930531642, -0.02260684382913637),\n",
       "  'P4 to P6': (-0.007706937809535931, -0.005519727500593054),\n",
       "  'P6 to P8': (-5.017183929886168, 1.3067517013352734),\n",
       "  'P8 to P12': (-0.0032746849409815574, -0.00042835293809750843)},\n",
       " 'EX_lys_L_e': {'P0 to P2': (-0.044612715893601024, -0.009227447477468988),\n",
       "  'P12 to P14': (0.0002078745327130727, 0.0033284362686953456),\n",
       "  'P2 to P4': (-0.02649280604359986, -0.022178934192114745),\n",
       "  'P4 to P6': (-0.0056464771810930915, -0.004016867755843037),\n",
       "  'P6 to P8': (-2.627845182617604, 0.6871725849278313),\n",
       "  'P8 to P12': (-0.0012838945957174244, -3.246168429233581e-05)},\n",
       " 'EX_mal_L_e': {'P0 to P2': (0.0, 0.0),\n",
       "  'P12 to P14': (-0.00012172317004349959, 0.0017996159049952624),\n",
       "  'P2 to P4': (0.0, 0.0),\n",
       "  'P4 to P6': (0.0, 0.0),\n",
       "  'P6 to P8': (0.0, 0.0),\n",
       "  'P8 to P12': (0.002198442187310876, 0.004218499350484458)},\n",
       " 'EX_met_L_e': {'P0 to P2': (-0.014314063507190562, -0.006871380192055844),\n",
       "  'P12 to P14': (0.0003675077063700237, 0.0006486990761896689),\n",
       "  'P2 to P4': (-0.008176224574176813, -0.007576325570625968),\n",
       "  'P4 to P6': (-0.0019770251059460947, -0.0013251936729442456),\n",
       "  'P6 to P8': (-0.8891397682622355, 0.22994397454156484),\n",
       "  'P8 to P12': (-0.0007179269332194462, -0.0001831522795428832)},\n",
       " 'EX_phe_L_e': {'P0 to P2': (-0.003991960539982331, 0.006162214626350528),\n",
       "  'P12 to P14': (1.4996680209072584e-05, 0.00037916942487314253),\n",
       "  'P2 to P4': (-0.010231419976890541, -0.008844114415915405),\n",
       "  'P4 to P6': (-0.0030265931219659475, -0.0020621895019406595),\n",
       "  'P6 to P8': (-1.1241746820505358, 0.2891690799728963),\n",
       "  'P8 to P12': (-0.0007852516855924107, -0.0006406859135899807)},\n",
       " 'EX_pro_L_e': {'P0 to P2': (0.06332963748757964, 0.09335803769360115),\n",
       "  'P12 to P14': (0.001991658951964453, 0.005750923983995642),\n",
       "  'P2 to P4': (-0.042042741019464584, -0.032272756511549476),\n",
       "  'P4 to P6': (-0.0057069450573701535, -0.004352310947028974),\n",
       "  'P6 to P8': (-1.752603983571014, 0.44996608215095824),\n",
       "  'P8 to P12': (-0.0024564759729163504, -0.0006623111226624875)},\n",
       " 'EX_5oxpro_e': {'P0 to P2': (0.16215089945861158, 0.31527545117092215),\n",
       "  'P12 to P14': (-0.005874957243275055, 0.011300851835673335),\n",
       "  'P2 to P4': (0.004983102253668478, 0.015882230903914166),\n",
       "  'P4 to P6': (-0.00013605885040839463, 0.0024162716110602304),\n",
       "  'P6 to P8': (-0.4686900061749263, 1.7916032502814838),\n",
       "  'P8 to P12': (-3.750672992593792e-06, 0.005518746178583188)},\n",
       " 'EX_pyr_e': {'P0 to P2': (-0.08096212754742405, -0.04606071129531597),\n",
       "  'P12 to P14': (-0.0022058482797831158, 0.0024344148904222795),\n",
       "  'P2 to P4': (0.0024354922283552436, 0.005141826374258867),\n",
       "  'P4 to P6': (-0.005401861471831332, -0.003887614478065361),\n",
       "  'P6 to P8': (-0.9516894076082902, 0.24736447972938053),\n",
       "  'P8 to P12': (0.0007666696500161254, 0.0015911317258110358)},\n",
       " 'EX_ser_L_e': {'P0 to P2': (-0.19339304344385716, -0.060028229844815484),\n",
       "  'P12 to P14': (-0.000749080940628395, 0.001303883077887666),\n",
       "  'P2 to P4': (-0.06558888981974573, -0.05590732872151247),\n",
       "  'P4 to P6': (-0.019486471992335105, -0.012732823142995366),\n",
       "  'P6 to P8': (-5.536695881940392, 1.4249050792322961),\n",
       "  'P8 to P12': (-0.00028932833026156353, 0.0008084211812553097)},\n",
       " 'EX_thr_L_e': {'P0 to P2': (-0.024627682003573924, -0.005546914641901558),\n",
       "  'P12 to P14': (0.00037257521896944543, 0.002557964736355796),\n",
       "  'P2 to P4': (-0.02567684157357688, -0.012133621947408727),\n",
       "  'P4 to P6': (-0.004436476708188156, -0.0031935349495081774),\n",
       "  'P6 to P8': (-1.6905288529863298, 0.4488057802964207),\n",
       "  'P8 to P12': (-0.0027691890695883477, 0.0015628002121652056)},\n",
       " 'EX_trp_L_e': {'P0 to P2': (-0.010830433110398411, -0.003613722447754261),\n",
       "  'P12 to P14': (4.21662672036737e-06, 0.00011710046103440759),\n",
       "  'P2 to P4': (-0.0038542085241141163, -0.0032135015616940685),\n",
       "  'P4 to P6': (-0.001628276505462704, -0.0009961280258759414),\n",
       "  'P6 to P8': (-1.0211048771852496, 0.2614626280532331),\n",
       "  'P8 to P12': (-0.0015013097618872674, -0.0008655152414583245)},\n",
       " 'EX_tyr_L_e': {'P0 to P2': (0.0056735500873825884, 0.013638250169941248),\n",
       "  'P12 to P14': (-0.000639237910057818, 0.0006300825041016463),\n",
       "  'P2 to P4': (-0.011413145376258529, -0.010778724186150916),\n",
       "  'P4 to P6': (-0.0018461985662240772, -0.0011908301817429255),\n",
       "  'P6 to P8': (-0.9756605103661025, 0.2503002892834783),\n",
       "  'P8 to P12': (-1.717944830630025e-05, 0.00015151603245667838)},\n",
       " 'EX_val_L_e': {'P0 to P2': (-0.022192809248261395, -0.020061008250935503),\n",
       "  'P12 to P14': (-0.0001751152258958259, 0.0011751373119875542),\n",
       "  'P2 to P4': (-0.020946935588450368, -0.019404635742143652),\n",
       "  'P4 to P6': (-0.005780503308932539, -0.004008131270617485),\n",
       "  'P6 to P8': (-2.7877274646500707, 0.7232626216896592),\n",
       "  'P8 to P12': (-0.0019206861628747107, -0.0004745214990074508)},\n",
       " 'biomass_cho_s': {'P0 to P2': (0.028662138723042047, 0.038619214404480434),\n",
       "  'P12 to P14': (-0.008639228177407606, -0.00473021281370683),\n",
       "  'P2 to P4': (0.022376893560357457, 0.026887778353538873),\n",
       "  'P4 to P6': (0.003942997145049612, 0.0069023302748022425),\n",
       "  'P6 to P8': (-0.00506375928814193, -0.003765339618766975),\n",
       "  'P8 to P12': (-0.008559349173462663, -0.004725911030850367)}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uptsec_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set exp growth rate as upper bound and simulate uptake and secretion\n",
    "\n",
    "model.reactions.get_by_id(objective).upper_bound = 0.035\n",
    "pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(f'Simulated growth rate is: {pfba_obj}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparison of the experimental uptake and secretion values with simulated\n",
    "\n",
    "for rxn,flux in pfba_solution.fluxes.items():\n",
    "    for r,v in uptsec_p2_p4.items():\n",
    "        if rxn == r:\n",
    "            sim_bounds = flux\n",
    "            exp_bound = v\n",
    "            print('-----------------------------------')\n",
    "            print(f'Reaction {r}')\n",
    "            print(f'Exp bound: {exp_bound}')\n",
    "            print(f'Simulated bounds: {sim_bounds}')\n",
    "            print('-----------------------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before changing any bounds simulated growth rate is: 0.0\n",
      "Reaction EX_2hb_e setting bounds to (0.0011330468691121624, 0.001651296609250617)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_4hpro_e setting bounds to (-0.004224254317618865, -0.0006331999506674355)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_5oxpro_e setting bounds to (0.004983102253668478, 0.015882230903914166)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_ac_e setting bounds to (0.0010842194486317125, 0.0015281127416154252)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_ala_L_e setting bounds to (0.04798552563859218, 0.05653967493581927)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_arg_L_e setting bounds to (-0.01884898170723418, -0.01529801058321137)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_asn_L_e setting bounds to (-0.10843336015938503, -0.09549559643737511)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_asp_L_e setting bounds to (0.01456951555293158, 0.02087295020206143)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_bhb_e setting bounds to (0.0032407909069092273, 0.004300936477247652)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_cit_e setting bounds to (0.005959799128689479, 0.0075622343179195094)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_cys_L_e setting bounds to (-0.003721885418836899, -0.0033091501094601413)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_etoh_e setting bounds to (0.016006042606443296, 0.07464951555114036)\n",
      "Changing bounds for reaction EX_etoh_e to (0.016006042606443296, 0.07464951555114036) creates an infeasible solution\n",
      "Reaction EX_for_e setting bounds to (0.027652376456539285, 0.030956496907215074)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_glc_e setting bounds to (-0.5853348558421616, -0.4953355298825426)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_gln_L_e setting bounds to (-0.11022278442041394, -0.10686085702946083)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_glu_L_e setting bounds to (0.009995190342042689, 0.015809435557194437)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_gly_e setting bounds to (0.011664762238946842, 0.015361887660412769)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_his_L_e setting bounds to (-0.006798969568476408, -0.00596715768970421)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_ile_L_e setting bounds to (-0.01610170921321085, -0.013334379454487769)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_lac_L_e setting bounds to (0.7184475137252582, 0.8108292756320108)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_leu_L_e setting bounds to (-0.026137250930531642, -0.02260684382913637)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_lys_L_e setting bounds to (-0.02649280604359986, -0.022178934192114745)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_mal_L_e setting bounds to (0.0, 0.0)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_met_L_e setting bounds to (-0.008176224574176813, -0.007576325570625968)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_nh4_e setting bounds to (0.09331646648603117, 0.11649394271188668)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_phe_L_e setting bounds to (-0.010231419976890541, -0.008844114415915405)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_pro_L_e setting bounds to (-0.042042741019464584, -0.032272756511549476)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_pyr_e setting bounds to (0.0024354922283552436, 0.005141826374258867)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_ser_L_e setting bounds to (-0.06558888981974573, -0.05590732872151247)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_thr_L_e setting bounds to (-0.02567684157357688, -0.012133621947408727)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_trp_L_e setting bounds to (-0.0038542085241141163, -0.0032135015616940685)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_tyr_L_e setting bounds to (-0.011413145376258529, -0.010778724186150916)\n",
      "Simulated growth rate is: 0.0\n",
      "Reaction EX_val_L_e setting bounds to (-0.020946935588450368, -0.019404635742143652)\n",
      "Simulated growth rate is: 0.07382807235936599\n",
      "Reaction EX_fum_e setting bounds to (-0.00016066431702818767, 0.0005996073941125498)\n",
      "Simulated growth rate is: 0.07382807235936575\n"
     ]
    }
   ],
   "source": [
    "# Adjust the lower bound values according to the experimental growth rates in order to indetify bottlenecks\n",
    "cond = 'P2 to P4'\n",
    "\n",
    "\n",
    "pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "pfba_obj = pfba_solution.fluxes[objective]\n",
    "print(f'Before changing any bounds simulated growth rate is: {pfba_obj}')\n",
    "\n",
    "# Set lower bounds of the reactions according to the experimental data\n",
    "for reaction in model.reactions:\n",
    "    if reaction.id == 'biomass_cho_s':\n",
    "        continue\n",
    "    for r,v in uptsec_wt.items():\n",
    "        if reaction.id == r:\n",
    "            if not pd.isna(v[cond]):\n",
    "                reaction.bounds = v[cond]\n",
    "                #if reaction.id == 'EX_arg_L_e':\n",
    "                #    reaction.lower_bound = -0.0107\n",
    "                #if reaction.id == 'EX_his_L_e':\n",
    "                #    reaction.lower_bound = -0.00314\n",
    "                print(f'Reaction {r} setting bounds to {reaction.bounds}')\n",
    "                try:\n",
    "                    pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "                    pfba_obj = pfba_solution.fluxes[objective]\n",
    "                    print(f'Simulated growth rate is: {pfba_obj}')\n",
    "                except Infeasible:\n",
    "                    print(f'Changing bounds for reaction {r} to {reaction.bounds} creates an infeasible solution')\n",
    "                    reaction.lower_bound = -1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import cobra\n",
    "from cobra.io import load_json_model, read_sbml_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define models\n",
    "\n",
    "iCHO1766 = read_sbml_model('../Data/Reconciliation/models/iCHOv1_final.xml')\n",
    "iCHO2291 = read_sbml_model('../Data/Reconciliation/models/iCHO2291.xml')\n",
    "iCHO2441 = read_sbml_model('../Data/Reconciliation/models/iCHO2441.xml')\n",
    "iCHO3000 = load_json_model('iCHO3644.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Preprocess and standarization of iCHO1766\n",
    "iCHO1766.id = 'iCHO1766'\n",
    "iCHO1766.reactions.biomass_cho_producing.id = 'biomass_cho_prod'\n",
    "for r in iCHO1766.reactions:\n",
    "    if r.id.endswith(\"_\"):\n",
    "        r.id = r.id[:-1]\n",
    "\n",
    "# Preprocess and standarization of iCHO2101\n",
    "iCHO2441.id = 'iCHO2441'\n",
    "for r in iCHO2441.reactions:\n",
    "    if r.id.endswith(\"(e)\"):\n",
    "        r.id = r.id.replace(\"(e)\", \"_e\")\n",
    "    if r.id == \"SK_Asn_X_Ser/Thr[r]\":\n",
    "        r.id = \"SK_Asn_X_Ser_Thr_r\"\n",
    "    if r.id == \"SK_pre_prot[r]\":\n",
    "        r.id = \"SK_pre_prot_r\"\n",
    "    if r.id == \"SK_Ser/Thr[g]\":\n",
    "        r.id = \"SK_Ser_Thr_g\"\n",
    "    if r.id == \"SK_Tyr_ggn[c]\":\n",
    "        r.id = \"SK_Tyr_ggn_c\"\n",
    "    if r.id == \"DM_igg[g]\":\n",
    "        r.id = \"DM_igg_g\"\n",
    "\n",
    "# Preprocess and standarization of iCHO2291\n",
    "for r in iCHO2291.reactions:\n",
    "    if r.id.endswith(\"(e)\"):\n",
    "        r.id = r.id.replace(\"(e)\", \"_e\")\n",
    "    if r.id == \"SK_Asn_X_Ser/Thr[r]\":\n",
    "        r.id = \"SK_Asn_X_Ser_Thr_r\"\n",
    "    if r.id == \"SK_pre_prot[r]\":\n",
    "        r.id = \"SK_pre_prot_r\"\n",
    "    if r.id == \"SK_Ser/Thr[g]\":\n",
    "        r.id = \"SK_Ser_Thr_g\"\n",
    "    if r.id == \"SK_Tyr_ggn[c]\":\n",
    "        r.id = \"SK_Tyr_ggn_c\"\n",
    "    if r.id == \"DM_igg[g]\":\n",
    "        r.id = \"DM_igg_g\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#models = [iCHO1766, iCHO2291, iCHO2441, iCHO3000]\n",
    "models = [iCHO3000]\n",
    "\n",
    "for model in models:\n",
    "    model.reactions.biomass_cho.bounds = (0.0,10)\n",
    "    model.reactions.biomass_cho_prod.bounds = (0.0,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uptsec = pd.read_excel('../Data/Uptake_Secretion_Rates/mmc4.xlsx', header=2)\n",
    "\n",
    "biomass_bounds = uptsec[42:44]\n",
    "biomass_bounds = biomass_bounds.set_index(['Reaction'])\n",
    "\n",
    "uptsec = uptsec[:-5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "hp = uptsec.set_index('Reaction')['HP'].to_dict()\n",
    "hp_nabu = uptsec.set_index('Reaction')['HP+NaBu'].to_dict()\n",
    "lp = uptsec.set_index('Reaction')['LP'].to_dict()\n",
    "lp_nabu = uptsec.set_index('Reaction')['LP+NaBu'].to_dict()\n",
    "early_exp = uptsec.set_index('Reaction')['Early Exp'].to_dict()\n",
    "late_exp = uptsec.set_index('Reaction')['Late Exp'].to_dict()\n",
    "cold_1 = uptsec.set_index('Reaction')['Cold 1'].to_dict()\n",
    "cold_2 = uptsec.set_index('Reaction')['Cold 2'].to_dict()\n",
    "\n",
    "conditions = ['HP', 'HP+NaBu', 'LP', 'LP+NaBu', 'Early Exp', 'Late Exp', 'Cold 1', 'Cold 2']\n",
    "dicts = [hp, hp_nabu, lp, lp_nabu, early_exp, late_exp, cold_1, cold_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for model in models:\n",
    "    for rxn in model.boundary:\n",
    "    \n",
    "        # IMPORTANT, all ex should be 0 - 1000, except those that we know are being uptaken by the cell\n",
    "        \n",
    "        if rxn.id.startswith(\"EX_\"):\n",
    "            rxn.bounds = (0,1000) \n",
    "        if rxn.id.startswith(\"SK_\"):\n",
    "            rxn.bounds = (-1000,1000)\n",
    "        if rxn.id.startswith(\"DM_\"):\n",
    "            rxn.bounds = (0,1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "pfba_solutions_fluxes = []\n",
    "\n",
    "df = pd.DataFrame(columns=[\"Model\", \"Condition\", \"Biomass_CHO\", \"Biomass_CHO_Prod\"])\n",
    "\n",
    "for model in models:\n",
    "\n",
    "    model_data = []\n",
    "    \n",
    "    for cond,dic in zip(conditions,dicts):\n",
    "        \n",
    "        # Set lower bounds of the reactions according to the experimental data\n",
    "        for reaction in model.reactions:\n",
    "            for r,v in dic.items():\n",
    "                if reaction.id == r:\n",
    "                    if not pd.isna(v):\n",
    "                        reaction.lower_bound = v\n",
    "    \n",
    "        # Calculate growth rate for each condition\n",
    "        print('----------------------')\n",
    "        print(f' Condtion: {cond} ')\n",
    "        print('----------------------')\n",
    "\n",
    "        # Print experimental growth rate\n",
    "        exp_gr = biomass_bounds[cond]['Experimental growth rate']\n",
    "        print(f'Experimental Growth Rate: {exp_gr}')\n",
    "\n",
    "        # Calculate growth for biomass_cho\n",
    "        model.objective = model.reactions.biomass_cho\n",
    "        model.reactions.biomass_cho_prod.upper_bound = 0\n",
    "        model.reactions.biomass_cho.upper_bound = 1000\n",
    "        pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "        pfba_solution_biomass_cho = pfba_solution.fluxes['biomass_cho']\n",
    "        print(f'Model: {model.id}, Biomass: {pfba_solution_biomass_cho}')\n",
    "\n",
    "        # Calculate growth for biomass_cho_prod\n",
    "        model.objective = model.reactions.biomass_cho_prod\n",
    "        model.reactions.biomass_cho_prod.upper_bound = 1000\n",
    "        model.reactions.biomass_cho.upper_bound = 0\n",
    "        pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "        pfba_solution_biomass_cho_prod = pfba_solution.fluxes['biomass_cho_prod']\n",
    "        print(f'Model: {model.id}, Biomass Prod: {pfba_solution_biomass_cho_prod}')\n",
    "        print(' ')\n",
    "\n",
    "        # Save the solution with metadata\n",
    "        pfba_solutions_fluxes.append({\n",
    "            \"model_id\": model.id,\n",
    "            \"condition\": cond,\n",
    "            \"type\": \"biomass_cho_prod\",\n",
    "            \"fluxes\": pfba_solution.fluxes\n",
    "        })\n",
    "\n",
    "        # Append the results to the model_data list\n",
    "        model_data.append({\n",
    "            'Model': model.id,\n",
    "            'Condition': cond,\n",
    "            'Biomass_CHO': pfba_solution_biomass_cho,\n",
    "            'Biomass_CHO_Prod': pfba_solution_biomass_cho_prod,\n",
    "        })\n",
    "\n",
    "\n",
    "    # Convert the model_data list to a DataFrame and concatenate it with the main DataFrame\n",
    "    model_df = pd.DataFrame(model_data)\n",
    "    df = pd.concat([df, model_df], ignore_index=True)\n",
    "\n",
    "\n",
    "# Add the experimental growth rate values to the dataset\n",
    "exp_data = []\n",
    "\n",
    "for cond,dic in zip(conditions,dicts):\n",
    "    exp_gr = biomass_bounds[cond]['Experimental growth rate']\n",
    "\n",
    "    exp_data.append({\n",
    "            'Model': \"Exp\",\n",
    "            'Condition': cond,\n",
    "            'Biomass_CHO': exp_gr,\n",
    "            'Biomass_CHO_Prod': exp_gr,\n",
    "        })\n",
    "exp_df = pd.DataFrame(exp_data)\n",
    "df = pd.concat([df, exp_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"whitegrid\")\n",
    "sns.set_context(\"talk\") \n",
    "\n",
    "# Plotting for Biomass_CHO\n",
    "plt.figure(figsize=(12, 7))\n",
    "barplot_cho = sns.barplot(data=df, x=\"Condition\", y=\"Biomass_CHO\", hue=\"Model\", palette=\"viridis\")\n",
    "plt.title(\"Growth Rate Calculated with the Biomass CHO Reaction\")\n",
    "plt.ylabel(\"Growth Rate (hr-1)\")\n",
    "plt.xlabel(\"Condition\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.legend(title='Condition', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Plots/bar_plot_growth_rate_biomass_cho.png')\n",
    "plt.show()\n",
    "\n",
    "# Plotting for Biomass_CHO_Prod\n",
    "plt.figure(figsize=(12, 7))\n",
    "barplot_cho_prod = sns.barplot(data=df, x=\"Condition\", y=\"Biomass_CHO_Prod\", hue=\"Model\", palette=\"viridis\")\n",
    "plt.title(\"Growth Rate Calculated with the Biomass CHO Prod Reaction\")\n",
    "plt.ylabel(\"Growth Rate (hr-1)\")\n",
    "plt.xlabel(\"Condition\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.legend(title='Condition', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Plots/bar_plot_growth_rate_biomass_cho_prod.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Flux Enrichment Analysis <a id='fea'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 1\n",
    "# Use the fluxes from the growth rate calculations to filter reactions with fluxes\n",
    "# Define a cut off for the flux ??? # Plot to visualization of the flux distribution\n",
    "# Run Flux Enrichment Analysis on those reactions to see pathways enriched when optimized by biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTION 2\n",
    "# Transcriptomic data from cell batches / biorreactiors\n",
    "# Overlay this into the recons and extract the reactions associated to the genes\n",
    "# Extract a reaction vector / \n",
    "# Run Flux Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# R1 = g1 or (g2 and G3)\n",
    "# r1 = max(r1, (min(g2, g3))\n",
    "# r1 = sum(r1, min(g2, g3))\n",
    "# GIMME \n",
    "# Use mCADRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import hypergeom\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def flux_enrichment_analysis(model, rxn_indices, attribute='subsystem'):\n",
    "    # Ensure rxn_indices is a list of integers\n",
    "    if not isinstance(rxn_indices, list) or not all(isinstance(i, int) for i in rxn_indices):\n",
    "        raise ValueError('rxn_indices must be a list of integers')\n",
    "    \n",
    "    # Check if the attribute exists in the first reaction as a proxy for all\n",
    "    if not hasattr(model.reactions[0], attribute):\n",
    "        raise ValueError(f'Attribute {attribute} not found in model reactions')\n",
    "\n",
    "    # Extract attribute information for all reactions\n",
    "    attribute_values = [getattr(rxn, attribute, 'None') for rxn in model.reactions]\n",
    "    unique_attributes = set(attribute_values)\n",
    "    \n",
    "    # Count occurrences in the model and in the reaction set\n",
    "    model_counts = {attr: attribute_values.count(attr) for attr in unique_attributes}\n",
    "    rxn_set_counts = {attr: 0 for attr in unique_attributes}\n",
    "    for idx in rxn_indices:\n",
    "        rxn_attr = getattr(model.reactions[idx], attribute, 'None')\n",
    "        rxn_set_counts[rxn_attr] += 1\n",
    "\n",
    "    # Calculate p-values using hypergeometric test\n",
    "    M = len(model.reactions)  # Total number of reactions\n",
    "    n = len(rxn_indices)  # Size of reaction set\n",
    "    p_values = []\n",
    "    for attr in unique_attributes:\n",
    "        N = model_counts[attr]  # Total reactions in group\n",
    "        x = rxn_set_counts[attr]  # Reactions in group and in set\n",
    "        p_value = hypergeom.sf(x-1, M, N, n)\n",
    "        p_values.append(p_value)\n",
    "\n",
    "    # Adjust p-values for multiple testing\n",
    "    _, adj_p_values, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "    # Compile results\n",
    "    results = pd.DataFrame({\n",
    "        'Group': list(unique_attributes),\n",
    "        'P-value': p_values,\n",
    "        'Adjusted P-value': adj_p_values,\n",
    "        'Enriched set size': [rxn_set_counts[attr] for attr in unique_attributes],\n",
    "        'Total set size': [model_counts[attr] for attr in unique_attributes],\n",
    "    }).sort_values(by='Adjusted P-value')\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.io import load_json_model\n",
    "iCHO3000 = load_json_model('iCHOv3_CHO_05032024.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test 1: FEA on reactions active during growth rate optimization\n",
    "Use the fluxes from the growth rate calculations to filter reactions with fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate a vector of active reactions for each one of the conditions\n",
    "\n",
    "active_reactions = []\n",
    "\n",
    "for sol in pfba_solutions_fluxes:\n",
    "    if sol['model_id'] == 'iCHO3644':\n",
    "        rxns_fluxes = []\n",
    "        for i,(r,f) in enumerate(sol['fluxes'].items()):\n",
    "            if f != 0:\n",
    "                rxns_fluxes.append(i)\n",
    "        \n",
    "        active_reactions.append({\n",
    "            \"model_id\": sol['model_id'],\n",
    "            \"condition\": sol['condition'],\n",
    "            \"flux_vector\": rxns_fluxes\n",
    "        })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate FEA results for each one of the vectors generated for each condition\n",
    "\n",
    "fea_results = []\n",
    "\n",
    "for fluxes in active_reactions:\n",
    "    results = flux_enrichment_analysis(iCHO3000, fluxes['flux_vector'], 'subsystem')\n",
    "    fea_results.append({\n",
    "        \"Condition\": fluxes['condition'],\n",
    "        \"Results\": results\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform p-values to -log10\n",
    "\n",
    "all_data = pd.DataFrame()\n",
    "\n",
    "for result in fea_results:\n",
    "    condition = result['Condition']\n",
    "    temp_df = result['Results'][['Group', 'P-value']].copy()\n",
    "    temp_df.columns = ['Group', f'P-value_{condition}']\n",
    "    if all_data.empty:\n",
    "        all_data = temp_df\n",
    "    else:\n",
    "        all_data = pd.merge(all_data, temp_df, on='Group', how='outer')\n",
    "\n",
    "# Remove groups with a 0 value in all conditions (assuming a '0' value indicates non-significance)\n",
    "significant_filter = (all_data.drop(columns='Group') < 0.05).any(axis=1)\n",
    "filtered_data = all_data[significant_filter]\n",
    "filtered_data.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# Transform the p-values\n",
    "for col in filtered_data.columns:\n",
    "    if col.startswith('P-value'):\n",
    "        filtered_data[f'-log10({col})'] = -np.log10(filtered_data[col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Invert the data for plotting the yticks correctly\n",
    "filtered_data = filtered_data.iloc[::-1].reset_index(drop=True)\n",
    "\n",
    "# Adjusting the bar width and spacing\n",
    "bar_width = 0.11  # Decrease the bar width for more distinct separation\n",
    "spacing = 0  # Increase the spacing between bars of different conditions\n",
    "\n",
    "# Assuming 'filtered_data' is structured with 'Group' and transformed p-value columns\n",
    "# Example conditions derived from your DataFrame's columns\n",
    "conditions = [col.replace('-log10(P-value_', '').replace(')', '') for col in filtered_data if col.startswith('-log10')]\n",
    "conditions = conditions[::-1]\n",
    "\n",
    "# Define colors for each condition using a colormap\n",
    "colors = plt.cm.viridis(np.linspace(0, 1, len(conditions)))\n",
    "\n",
    "# Figure setup\n",
    "fig_height = len(filtered_data['Group']) * (bar_width + spacing) * len(conditions)\n",
    "plt.figure(figsize=(25, fig_height))\n",
    "\n",
    "# Initial y positions for each group's first condition bar\n",
    "initial_pos = np.arange(len(filtered_data['Group']))\n",
    "\n",
    "# Plotting\n",
    "for i, condition in enumerate(conditions):\n",
    "    position = [p + (bar_width + spacing) * i for p in initial_pos]\n",
    "    plt.barh(position, filtered_data[f'-log10(P-value_{condition})'], height=bar_width, color=colors[i], label=condition)\n",
    "\n",
    "plt.tick_params(axis='y', labelsize=25)\n",
    "plt.tick_params(axis='x', labelsize=15)\n",
    "plt.yticks(initial_pos + (bar_width + spacing) * (len(conditions) - 1) / 2, filtered_data['Group'])\n",
    "plt.xlabel('-log10(P-value)', fontsize=30)\n",
    "plt.legend(title='Condition', bbox_to_anchor=(1.05, 1), fontsize=20, loc='upper left', reverse=True)\n",
    "\n",
    "# Manually adjust the y-axis limits if necessary\n",
    "plt.ylim(-0.5, len(filtered_data['Group']) - 0.5 + (bar_width + spacing) * len(conditions))\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('../Plots/FEA_biomass_prod_iCHO3000.png', format='png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
