{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57a5e96e",
   "metadata": {},
   "source": [
    "# Genes\n",
    "The following notebook retrieves all the genes involved in the reactions from the **\"Rxns\" Sheet** in Google Sheet file. Then, information regarding the genes is retrieved from different databases and a **\"Genes\" Sheet** is generated and updated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08228a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from google_sheet import GoogleSheet\n",
    "from utils import get_gene_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac38d2d7",
   "metadata": {},
   "source": [
    "### 1. Generate \"rxns\" and \"genes\" datasets\n",
    "The \"rxns\" dataset contains all the reactions with their GPR associations. The \"genes\" dataset contains all the information of the genes/GPR involved in our reconstruction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19d8026b",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEY_FILE_PATH = 'credentials.json'\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_rxns = 'Rxns'\n",
    "sheet_genes = 'Genes'\n",
    "\n",
    "rxns = sheet.read_google_sheet(sheet_rxns)\n",
    "genes = sheet.read_google_sheet(sheet_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de13c811",
   "metadata": {},
   "source": [
    "### 2. Create the a \"gene_list\" with all the genes involved in our reconstruction\n",
    "We extract the gene IDs from the GPR annotations in the **\"rxns\" df**. This information is only retrieved from the **GPR_final** column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0f61d39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3045"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generation of gene_list from all the genes in the \"Whole Cell Network Reconstruction in CHO Cells\" dataset\n",
    "import re\n",
    "\n",
    "gene_list = []\n",
    "human_gene_list = []\n",
    "for index, row in rxns.iterrows():\n",
    "    if row['GPR_final'] != '':\n",
    "        gprs = str(row['GPR_final'])\n",
    "        gpr = re.findall(r'\\b[hH]?\\d+\\b', gprs)\n",
    "        for g in gpr:\n",
    "            if g.lower().startswith('h'):  # we convert to lower case to catch both 'h' and 'H'\n",
    "                human_gene_list.append(int(g[1:]))  # remove 'h' and convert to int before appending\n",
    "            else:\n",
    "                gene_list.append(int(g))  # convert to int before appending\n",
    "\n",
    "gene_list = list(set(gene_list))\n",
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c5c853",
   "metadata": {},
   "source": [
    "### 3. Replace remaining Human Gene IDs in our dataset with CHO Gene IDs\n",
    "Using the **cho2human_mapping** file, we replace the Human gene IDs with those of CHO, these IDs were not spotted when we first added the Recon3D reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd3c0d42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build a dictionary with human gene IDs as keys and CHO gene IDs as values\n",
    "orthologs = pd.read_csv('../Data/Orthologs/cho2human_mapping.tsv', sep='\\t')\n",
    "human_to_cho_dict = dict(zip(orthologs['HUMAN_ID'], orthologs['CHO_ID']))\n",
    "\n",
    "def replace_human_ids_with_cho_ids(s):\n",
    "    replaced_dict = {}\n",
    "    if isinstance(s, str):\n",
    "        genes = re.findall(r'\\b[hH]?\\d+\\b', s)\n",
    "        for gene in genes:\n",
    "            if gene.lower().startswith('h'):  # if gene ID starts with 'h' or 'H'\n",
    "                human_id = int(gene[1:])  # remove 'h' and convert to int\n",
    "                if human_id in human_to_cho_dict:  # if human gene ID is in the dictionary\n",
    "                    cho_id = human_to_cho_dict[human_id]  # get corresponding CHO gene ID\n",
    "                    s = s.replace(gene, str(cho_id))  # replace human gene ID with CHO gene ID in the string\n",
    "                    replaced_dict[gene] = cho_id\n",
    "    return s, replaced_dict\n",
    "\n",
    "# Initialize a dictionary to store replacements\n",
    "replacements = {}\n",
    "\n",
    "# Apply the function and update the replacements dictionary\n",
    "for i in range(len(rxns)):\n",
    "    rxns.at[i, 'GPR_final'], replacements_dict = replace_human_ids_with_cho_ids(rxns.at[i, 'GPR_final'])\n",
    "    replacements.update(replacements_dict)\n",
    "\n",
    "# Now, 'replacements' is a dictionary where keys are the original gene IDs and values are the replaced CHO gene IDs\n",
    "replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cc309360",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No changes to the original Rxns Sheet\n"
     ]
    }
   ],
   "source": [
    "######################################\n",
    "#### ---------------------------- ####\n",
    "#### ---- Update Rxns Sheet ----- ####\n",
    "#### ---------------------------- ####\n",
    "######################################\n",
    "if replacements:\n",
    "    sheet.update_google_sheet(sheet_rxns, rxns)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif not replacements:\n",
    "    print(\"No changes to the original Rxns Sheet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed00e36d",
   "metadata": {},
   "source": [
    "### 3. Add the genes from the \"gene_list\" to the genes df\n",
    "Using a list of all the genes included in the dataset we can retrieve information from the NIH database regarding Gene Symbol, Gene Name, Gene Ensembl ID, and mRNA ID and protein ID using the function get_gene_info()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "236d1799",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "55622e982a4a4132a248114e9e6ca772",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3045 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30100754813\n",
      "Gene 30100754813 does not have PICR Ensembl ID\n",
      "Gene (30100754813, 'ycf15') is a Vitis cinerea Gene\n",
      "A total of 0 Genes were added to the dataset\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "from urllib.error import HTTPError\n",
    "from Bio.Entrez.Parser import ValidationError\n",
    "\n",
    "max_retries = 5  # Set the maximum number of retries\n",
    "c = 0\n",
    "\n",
    "for g in tqdm(gene_list):\n",
    "    if str(g) not in list(genes['Gene Entrez ID']):\n",
    "        retry = True\n",
    "        retries = 0\n",
    "        while retry:\n",
    "            try:\n",
    "                print(g)\n",
    "                organism, gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "                if organism == 'Cricetulus griseus':\n",
    "                    new_row_data = {'Gene Entrez ID': g, 'Gene Symbol': gene_symbol, 'Gene Name': gene_name, 'Gene Description': gene_description,\n",
    "                                    'PICR Ensembl ID': picr_ensembl_id, 'CHOK1GS Ensembl ID': chok1gs_ensembl_id, 'Transcript ID': mRNA_ncbi_id,\n",
    "                                    'Protein ID': protein_ncbi_id, 'GO Terms': go_terms}\n",
    "                    new_row_df = pd.DataFrame(new_row_data, index=[len(genes)])\n",
    "                    genes = pd.concat([genes, new_row_df])\n",
    "                    c+=1\n",
    "                    retry = False\n",
    "                elif organism != 'Cricetulus griseus':\n",
    "                    print(f'Gene {g,gene_symbol} is a {organism} Gene')\n",
    "                    break\n",
    "            except ValidationError:\n",
    "                print(f'Gene {g} not found')\n",
    "                retry = False\n",
    "                continue\n",
    "            except HTTPError:\n",
    "                retries += 1\n",
    "                if retries > max_retries:\n",
    "                    print(f'HTTP Error, reached maximum retries ({max_retries}) for gene {g}')\n",
    "                    break\n",
    "                else:\n",
    "                    print(f'HTTP Error, retrying with gene {g}')\n",
    "                    time.sleep(1)\n",
    "                    \n",
    "print(f'A total of {c} Genes were added to the dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b2abe8ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No new genes added.\n",
      "No changes to the original Genes Sheet.\n"
     ]
    }
   ],
   "source": [
    "#######################################\n",
    "#### ----------------------------- ####\n",
    "#### ---- Update Genes Sheet ----- ####\n",
    "#### ----------------------------- ####\n",
    "#######################################\n",
    "if c > 0:\n",
    "    sheet.update_google_sheet(sheet_genes, genes)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif c == 0:\n",
    "    print(\"No new genes added.\")\n",
    "    print(\"No changes to the original Genes Sheet.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2218925",
   "metadata": {},
   "source": [
    "### 4. Eliminate unwanted genes\n",
    "We iterate over the entire dataset to spot **Human genes** in order to eliminate them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b810d3c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a copy of the DataFrame to avoid changing the original while iterating\n",
    "erased = 0\n",
    "genes_copy = genes.copy()\n",
    "\n",
    "for index, row in tqdm(genes.iterrows(), total=genes.shape[0]):\n",
    "    g = row['Gene Entrez ID']\n",
    "    retry = True\n",
    "    retries = 0\n",
    "    while retry:\n",
    "        try:\n",
    "            organism, gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "            if organism == 'Homo sapiens':\n",
    "                genes_copy = genes_copy.drop(index)\n",
    "                print(f'Gene {g,gene_symbol} erased from dataset')\n",
    "                erased+=1\n",
    "                retry = False\n",
    "            else:\n",
    "                retry = False\n",
    "        except ValidationError:\n",
    "            print(f'Gene {g} not found')\n",
    "            retry = False\n",
    "            continue\n",
    "        except HTTPError:\n",
    "            retries += 1\n",
    "            if retries > max_retries:\n",
    "                print(f'HTTP Error, reached maximum retries ({max_retries}) for gene {g}')\n",
    "                break\n",
    "            else:\n",
    "                print(f'HTTP Error, retrying with gene {g}')\n",
    "                time.sleep(1)\n",
    "\n",
    "# Replace the original DataFrame with the updated one\n",
    "genes = genes_copy\n",
    "print(f'A total of {erased} Genes were erased from the original dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a459ea77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################\n",
    "#### ----------------------------- ####\n",
    "#### ---- Update Genes Sheet ----- ####\n",
    "#### ----------------------------- ####\n",
    "#######################################\n",
    "if erased > 0:\n",
    "    sheet.update_google_sheet(sheet_genes, genes)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif erased == 0:\n",
    "    print(\"No erased genes.\")\n",
    "    print(\"No changes to the original Genes Sheet.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f3b506a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in tqdm(gene_list):\n",
    "    if g not in list(genes['Gene Entrez ID']):\n",
    "        try:\n",
    "            gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "            new_row_data = {'Gene Entrez ID': g, 'Gene Symbol': gene_symbol, 'Gene Name': gene_name, 'Gene Description': gene_description,\n",
    "                            'PICR Ensembl ID': picr_ensembl_id, 'CHOK1GS Ensembl ID': chok1gs_ensembl_id, 'Transcript ID': mRNA_ncbi_id,\n",
    "                            'Protein ID': protein_ncbi_id, 'GO Terms': go_terms}\n",
    "            new_row_df = pd.DataFrame(new_row_data, index=[len(genes)])\n",
    "            genes = pd.concat([genes, new_row_df])\n",
    "        except KeyError:\n",
    "            print(f'Gene {g} not found')\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dfcbfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch information from the NIH database\n",
    "\n",
    "\n",
    "# Complete null or blank information in the already generated \"Genes Sheet\" dataset\n",
    "for i,row in genes.iterrows():\n",
    "    if row['Gene Entrez ID'] == '':\n",
    "        for g in gene_list:\n",
    "            gene_sheet_list = [str(x) for x in genes['Gene Entrez ID']]\n",
    "            if g not in gene_sheet_list:\n",
    "                gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "                row['Gene Entrez ID'] = g\n",
    "                row['Gene Symbol'] = gene_symbol\n",
    "                row['Gene Name'] = gene_name\n",
    "                row['Gene Description'] = gene_description\n",
    "                row['PICR Ensembl ID'] = picr_ensembl_id\n",
    "                row['CHOK1GS Ensembl ID'] = chok1gs_ensembl_id\n",
    "                row['Transcript ID'] = mRNA_ncbi_id\n",
    "                row['Protein ID'] = protein_ncbi_id\n",
    "                row['GO Terms'] = go_terms\n",
    "                break          \n",
    "    elif row['Gene Entrez ID'] != '' and (row['Gene Symbol'] == '' or row['Gene Name'] == '' or row['PICR Ensembl ID'] == '' or row['Transcript ID'] == '' or row['Protein ID'] == ''):\n",
    "        row['Gene Symbol'] = gene_symbol\n",
    "        row['Gene Name'] = gene_name\n",
    "        row['Gene Description'] = gene_description\n",
    "        row['PICR Ensembl ID'] = picr_ensembl_id\n",
    "        row['CHOK1GS Ensembl ID'] = chok1gs_ensembl_id\n",
    "        row['Transcript ID'] = mRNA_ncbi_id\n",
    "        row['Protein ID'] = protein_ncbi_id\n",
    "        row['GO Terms'] = go_terms\n",
    "'''\n",
    "# Add genes from the gene_list that are not yet in the \"Genes Sheet\" dataset\n",
    "for g in gene_list:\n",
    "    # the first try/except is to avoid overwritting data in case there already some info in the dataset\n",
    "    try:\n",
    "        gene_sheet_list = [str(x) for x in df['Gene Entrez ID']]\n",
    "        id = max(df['Index']) + 2\n",
    "    except:\n",
    "        gene_sheet_list = []\n",
    "        id = 2\n",
    "    if gene not in gene_sheet_list:\n",
    "        try:\n",
    "            gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(gene)\n",
    "            print(id-1)\n",
    "            genes_sheet.update_cell(id,1,id-1)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,2,gene)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,3,gene_symbol)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,4,gene_name)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,5,gene_description)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,6,picr_ensembl_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,7,chok1gs_ensembl_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,8,mRNA_ncbi_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,9,protein_ncbi_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,10,go_terms)\n",
    "        except:\n",
    "            print('Google API quota exceeded')\n",
    "            time.sleep(5)\n",
    "            continue\n",
    "            \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b59d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = 'account1@theta-ocean-377718.iam.gserviceaccount.com'\n",
    "handle = Entrez.efetch(db='gene', id='3979190', retmode='xml')\n",
    "record = Entrez.read(handle)[0]\n",
    "\n",
    "#gene_name = record['Entrezgene_gene']['Gene-ref']['Gene-ref_desc']\n",
    "#gene_symbol = record['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']\n",
    "\n",
    "human_handle = Entrez.efetch(db='gene', id='3979190', retmode='xml')\n",
    "human_record = Entrez.read(human_handle)[0]\n",
    "if 'Entrezgene_comments' in human_record and 'Gene-commentary_comment' in human_record['Entrezgene_comments'][0]:\n",
    "    human_gene_description = human_record['Entrezgene_comments'][0]['Gene-commentary_comment'][0]['String']\n",
    "elif 'Entrezgene_summary' in human_record:\n",
    "    human_gene_description = human_record['Entrezgene_summary']\n",
    "else:\n",
    "    human_gene_description = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40d44cdb",
   "metadata": {},
   "source": [
    "#### Get the CHO genome and map the IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f04660af",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "import math\n",
    "import time\n",
    "import ssl\n",
    "from urllib.error import HTTPError\n",
    "import socket\n",
    "from Bio.Entrez.Parser import CorruptedXMLError\n",
    "import requests\n",
    "\n",
    "def get_uniprot_from_dict(dictionary, uniprotIDs, target_key, target_value1,  target_value2):\n",
    "    for key, value in dictionary.items():\n",
    "        if key == target_key and (value == target_value1 or value == target_value2):\n",
    "            uniprotIDs.append(dictionary.get('Dbtag_tag', {}).get('Object-id', {}).get('Object-id_str'))            \n",
    "        if isinstance(value, dict):\n",
    "            get_uniprot_from_dict(value, uniprotIDs, target_key, target_value1, target_value2)\n",
    "        elif isinstance(value, list):\n",
    "            for item in value:\n",
    "                if isinstance(item, dict):\n",
    "                    get_uniprot_from_dict(item, uniprotIDs, target_key, target_value1, target_value2)        \n",
    "    return uniprotIDs\n",
    "\n",
    "def get_human_mouse_orthologs(dictionary, orthologs, target_key, target_value1):\n",
    "    for key, value in dictionary.items():\n",
    "        if key == target_key and (value == target_value1):\n",
    "            orthologs.append(dictionary.get('Dbtag_tag', {}).get('Object-id', {}).get('Object-id_id'))        \n",
    "        if isinstance(value, dict):\n",
    "            get_human_mouse_orthologs(value, orthologs, target_key, target_value1)\n",
    "        elif isinstance(value, list):\n",
    "            for item in value:\n",
    "                if isinstance(item, dict):\n",
    "                    get_human_mouse_orthologs(item, orthologs, target_key, target_value1)        \n",
    "    return orthologs           \n",
    "\n",
    "\n",
    "\n",
    "# Entrez.email = \"athand01@gmail.com\"\n",
    "Entrez.email = \"a.antonakoudis@sartorius.com\"\n",
    "u_g = pd.read_excel('../CHO_Genome.xlsx')\n",
    "GO_terms = pd.read_csv('orthologs&GO.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "dd708214",
   "metadata": {},
   "outputs": [],
   "source": [
    "def UniProt_Data_Request(uniprot_id, session=None):\n",
    "\n",
    "    name = []\n",
    "    weight = []\n",
    "    length = []\n",
    "    annotation_score = []\n",
    "    existance = []\n",
    "    subcellular_locations_list = []\n",
    "    tissue_specificity = []\n",
    "    max_retries = 3\n",
    "    retries = 0\n",
    "    if session is None:\n",
    "        session = requests.Session()\n",
    "        certificate_path = '../../../../../../Downloads/Zscaler Root CA.crt'\n",
    "        session.verify = certificate_path\n",
    "\n",
    "    while retries < max_retries:\n",
    "        try:\n",
    "            url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.json\"\n",
    "            response = session.get(url, headers={\"Accept\": \"application/json\"}, timeout=150)\n",
    "            if response.status_code == 200:\n",
    "                entry = response.json()\n",
    "                if 'recommendedName' in entry['proteinDescription']:\n",
    "                    name  = entry['proteinDescription']['recommendedName']['fullName']['value']\n",
    "                elif 'submissionNames' in entry['proteinDescription']:\n",
    "                    name = entry['proteinDescription']['submissionNames'][0]['fullName']['value']\n",
    "                annotation_score = entry['annotationScore']\n",
    "                weight = entry['sequence']['molWeight']\n",
    "                length = entry['sequence']['length']\n",
    "                existance = entry['proteinExistence']              \n",
    "\n",
    "                for references in entry['references']:\n",
    "                    if 'referenceComments' in references:     \n",
    "                        for ref_comments in references['referenceComments']:\n",
    "                            if ref_comments['type'] == 'TISSUE':\n",
    "                                for ref_comments_tissues in ref_comments['type']:\n",
    "                                    tissue_specificity = ref_comments['value']\n",
    "\n",
    "                subcellular_locations_list = []\n",
    "                if 'comments' in entry:\n",
    "                    for i in entry['comments']:\n",
    "                        if 'subcellularLocations' in  i:\n",
    "                            for k in i['subcellularLocations']:\n",
    "                                subcellular_location = k['location']['value']\n",
    "                                subcellular_locations_list.append(subcellular_location)                \n",
    "                return name, weight, length, annotation_score, existance, subcellular_locations_list, tissue_specificity\n",
    "            else:\n",
    "                raise Exception(f\"API request failed for UniProt ID {uniprot_id}: Status code {response.status_code}\")\n",
    "\n",
    "        except requests.exceptions.RequestException as e:\n",
    "            # Handle request-related errors\n",
    "            print(f\"Request Error: {e}\")\n",
    "            retries += 1\n",
    "            if retries < max_retries:\n",
    "                print(\"Retrying after a delay...\")\n",
    "                time.sleep(1)  # Wait for 5 seconds before retrying\n",
    "\n",
    "        except Exception as e:\n",
    "            # Handle unexpected errors\n",
    "            print(f\"An unexpected error occurred: {e}\")\n",
    "            retries += 1\n",
    "            if retries < max_retries:\n",
    "                print(\"Retrying after a delay...\")\n",
    "                time.sleep(1)  # Wait for 5 seconds before retrying\n",
    "\n",
    "    return name, weight, length, annotation_score, existance, subcellular_locations_list, tissue_specificity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "e34657fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 |GENE:  Ercc2 ----- ['XPD', 'CXPD'] ----- 100689272 ----- ERCC excision repair 2, TFIIH core complex helicase subunit ----- The nucleotide excision r ... ----- ['GO:0005634', 'GO:000573 ... ----- [] ----- 17 ----- 9 ----- minus ----- 2068 ----- 13871\n",
      "ASSEMBLY SPECIFC GENE: S1, S2,  100689272 ----- NW_003614150 ----- 553134 ----- 561663 ----- XM_035457766 ----- Reference CriGri_1.0 Primary Assembly\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Gene-commentary_products'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\a.antonakoudis\\OneDrive - Sartorius Corporate Administration GmbH\\Documents\\Projects\\UCSD Project\\Whole-Cell-Network-Reconstruction-for-CHO-cells-2\\Notebooks\\Genes.ipynb Cell 22\u001b[0m line \u001b[0;36m2\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X30sZmlsZQ%3D%3D?line=293'>294</a>\u001b[0m                     \u001b[39mfor\u001b[39;00m d \u001b[39min\u001b[39;00m Entrez_Comments[\u001b[39m'\u001b[39m\u001b[39mGene-commentary_comment\u001b[39m\u001b[39m'\u001b[39m]:\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X30sZmlsZQ%3D%3D?line=294'>295</a>\u001b[0m                         \u001b[39mif\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mGene-commentary_comment\u001b[39m\u001b[39m'\u001b[39m \u001b[39min\u001b[39;00m d:\n\u001b[1;32m--> <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X30sZmlsZQ%3D%3D?line=295'>296</a>\u001b[0m                             \u001b[39mprint\u001b[39m(d[\u001b[39m'\u001b[39;49m\u001b[39mGene-commentary_products\u001b[39;49m\u001b[39m'\u001b[39;49m][\u001b[39m1\u001b[39m][\u001b[39m'\u001b[39m\u001b[39mGene-commentary_accession\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X30sZmlsZQ%3D%3D?line=297'>298</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m-----------------------------------------------------------------------------------------\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X30sZmlsZQ%3D%3D?line=298'>299</a>\u001b[0m counter \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Gene-commentary_products'"
     ]
    }
   ],
   "source": [
    "gene_symbol_list = []\n",
    "\n",
    "gene_synoyms = []\n",
    "ncbi_protein_NP_weirdName = []\n",
    "ncbi_geneID_list = []\n",
    "ncbi_gene_exon_size = []\n",
    "# ncbi_genomic_location = []\n",
    "uniprot_IDs = []\n",
    "orthologs_genes = []\n",
    "orthologs_organism = []\n",
    "aa_sequences = []\n",
    "gene_record = []\n",
    "\n",
    "counter = 4\n",
    "for gene_EntrezID in u_g['NCBI GeneID'][counter:5]:\n",
    "# if True:\n",
    "#     gene_EntrezID = 100774298\n",
    "    protein_NP_weirdName = []\n",
    "    if not math.isnan(gene_EntrezID):   \n",
    "        gene_EntrezID = int(gene_EntrezID)        \n",
    "        go_terms = []          \n",
    "        max_retries = 3\n",
    "        retries = 0\n",
    "        gene_found = True\n",
    "        while retries < max_retries and gene_found:\n",
    "\n",
    "            try:\n",
    "\n",
    "                search_results = Entrez.esearch(db=\"gene\", term = gene_EntrezID)            \n",
    "                record = Entrez.read(search_results)\n",
    "\n",
    "                # Check if any gene records were found\n",
    "                if int(record[\"Count\"]) > 0:\n",
    "\n",
    "                    # Get the Gene ID of the first result\n",
    "                    gene_id = record[\"IdList\"][0]\n",
    "\n",
    "                    # Fetch the gene information, maximum 3 tries\n",
    "                    fetch_tries = 0\n",
    "                    fetch_found = True\n",
    "                    while fetch_tries < 3 and fetch_found:\n",
    "\n",
    "                        gene_info = Entrez.efetch(db=\"gene\", id=gene_id, retmode=\"xml\")           \n",
    "                        fetch_tries += 1\n",
    "                        try:\n",
    "                            gene_record = Entrez.read(gene_info)\n",
    "                            fetch_found = False\n",
    "                        except CorruptedXMLError as xml_error:\n",
    "                            time.sleep(3)\n",
    "                            fetch_tries += 1\n",
    "\n",
    "                    # Get the gene name and gene NCBI ID\n",
    "                    if 'Gene-ref_locus' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "                        gene_symbol = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']                        \n",
    "                    else:\n",
    "                        gene_symbol = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_locus-tag']\n",
    "                \n",
    "                    # Get the Gene Synonyms\n",
    "                    if 'Gene-ref_syn' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "                        gene_synonyms = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_syn']\n",
    "                    else:\n",
    "                        gene_synonyms = []\n",
    "                    gene_ncbiID = gene_record[0]['Entrezgene_track-info']['Gene-track']['Gene-track_geneid']\n",
    "\n",
    "                    # Get the Gene Description\n",
    "\n",
    "                    # Get UniprotIDs\n",
    "                    gene_uniprotIDs = (get_uniprot_from_dict(gene_record[0], [], 'Dbtag_db', 'UniProtKB/TrEMBL', 'UniProtKB/Swiss-Prot'))\n",
    "\n",
    "                    # Get the Gene Name\n",
    "                    gene_name = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_desc']\n",
    "\n",
    "                    # Get the chromosome\n",
    "                    chromosome = int(gene_record[0]['Entrezgene_source']['BioSource']['BioSource_subtype'][0]['SubSource_name'])\n",
    "\n",
    "                    # Get number of Exons\n",
    "                    exon_count = int(gene_record[0]['Entrezgene_properties'][0]['Gene-commentary_text'])\n",
    "\n",
    "                    # Get the Strand\n",
    "                    strand = gene_record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_strand']['Na-strand'].attributes['value']\n",
    "                    \n",
    "                    # Get Human and Mouse Orthologs\n",
    "                    orthologs_organism_list = []\n",
    "                    orthologs = []     \n",
    "                    if 'Entrezgene_comments' in gene_record[0]:\n",
    "                        for Entrezgene_comments_LOOP in gene_record[0]['Entrezgene_comments']:\n",
    "                            if 'Gene-commentary_comment' in Entrezgene_comments_LOOP:\n",
    "                                for Gene_commentary_comment_LOOP in Entrezgene_comments_LOOP['Gene-commentary_comment']:\n",
    "                                    if 'Gene-commentary_source' in Gene_commentary_comment_LOOP:\n",
    "                                        if 'Object-id_id' in Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']:\n",
    "                                            orthologs.append(Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_id'])\n",
    "                                        else:\n",
    "                                            orthologs.append(Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str'])\n",
    "                                        orthologs_organism_list.append(Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_anchor'])                    \n",
    "                    if 'human' in orthologs_organism_list:\n",
    "                        human_ortholog_EntrezID = orthologs[orthologs_organism_list.index('human')]                                            \n",
    "                        search_results_human = Entrez.esearch(db=\"gene\", term=human_ortholog_EntrezID)     \n",
    "                        record_human = Entrez.read(search_results_human)\n",
    "                        gene_id_human = record_human[\"IdList\"][0]\n",
    "                        gene_info_human = Entrez.efetch(db=\"gene\", id=gene_id_human, retmode=\"xml\")           \n",
    "                        gene_record_human = Entrez.read(gene_info_human)\n",
    "                        gene_description = gene_record_human[0]['Entrezgene_summary']\n",
    "                    else:\n",
    "                        human_ortholog_EntrezID = []\n",
    "                        gene_description = []\n",
    "                    if 'mouse' in orthologs_organism_list:\n",
    "                        mouse_ortholog_EntrezID = orthologs[orthologs_organism_list.index('mouse')]\n",
    "                    else:\n",
    "                        mouse_ortholog_EntrezID = []\n",
    "                    \n",
    "\n",
    "\n",
    "                    # Get GO terms\n",
    "                    result = GO_terms[GO_terms['CHO GeneID'] == str(gene_EntrezID)]['GO_ids'].values\n",
    "                    if result.size > 0:\n",
    "                        go_terms = result[0]\n",
    "                    else:\n",
    "                        go_terms = []\n",
    "\n",
    "                    # Get the sequence                \n",
    "                    aa_bool = 'NOT Found AA_Seq'\n",
    "                    if protein_NP_weirdName != '-----':\n",
    "                        try:\n",
    "                            handle = Entrez.efetch(db=\"protein\", id=protein_NP_weirdName, rettype=\"fasta\")\n",
    "                            record = handle.read()\n",
    "                            amino_acid_sequence = ''.join(record.split('\\n')[1:])\n",
    "                            aa_bool = 'FOUND AA_Seq'\n",
    "                        except:\n",
    "                            handle = Entrez.efetch(db=\"nucleotide\", id=protein_NP_weirdName, rettype=\"fasta\")\n",
    "                            record = handle.read()\n",
    "                            amino_acid_sequence = ''.join(record.split('\\n')[1:])\n",
    "                            aa_bool = 'FOUND AA_Seq'\n",
    "                    # Get the COG terms\n",
    "                    cogs = []\n",
    "\n",
    "                # If there are no matches for the EntrezID at NCBI then return -----\n",
    "                else:\n",
    "                    gene_ncbiID = '-----'\n",
    "                    gene_symbol = '-----'\n",
    "                    gene_name_alsoknownas = '-----'\n",
    "                    protein_NP_weirdName = '-----'\n",
    "                    exon = '-----'\n",
    "                    # gene_location = '-----'                    \n",
    "                    uniprot_IDs = '-----'\n",
    "                    amino_acid_sequence = '-----'\n",
    "\n",
    "                gene_found = False\n",
    "                search_results.close()\n",
    "    \n",
    "            # Handle exceptions\n",
    "            except HTTPError as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "            except ssl.SSLError as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "            except ConnectionResetError as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "            except socket.error as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "    \n",
    "        # If the maximum fetch attempts reached, return ------\n",
    "        if retries >= max_retries:\n",
    "            gene_symbol = ncbi_gene\n",
    "            gene_synonyms = '-----'\n",
    "            gene_ncbiID = '-----'\n",
    "            protein_NP_weirdName = '-----'\n",
    "            exon = '-----'\n",
    "            # gene_location = '-----'\n",
    "            gene_uniprotIDs = '-----'\n",
    "            orthologs = '-----'\n",
    "            orthologs_organism_list = '-----'\n",
    "            amino_acid_sequence = '-----'\n",
    "    \n",
    "    # If for some reason the ncbi_gene is nto a number, then return the ncbi_gene for all cases\n",
    "    else:\n",
    "        gene_symbol = ncbi_gene\n",
    "        gene_synonyms = ncbi_gene\n",
    "        gene_ncbiID = ncbi_gene\n",
    "        protein_NP_weirdName = ncbi_gene\n",
    "        exon = ncbi_gene\n",
    "        # gene_location = ncbi_gene\n",
    "        amino_acid_sequence = ncbi_gene\n",
    "\n",
    "    # Append the lists with the data\n",
    "    # gene_symbol_list.append(gene_symbol)\n",
    "    # ncbi_geneName_alsoknownas_list.append(gene_name_alsoknownas)\n",
    "    # ncbi_geneID_list.append(gene_ncbiID)\n",
    "    # ncbi_protein_NP_weirdName.append(protein_NP_weirdName)\n",
    "    # ncbi_gene_exon_size.append(exon)\n",
    "    # # ncbi_genomic_location.append(gene_location)\n",
    "    # uniprot_IDs.append(gene_uniprotIDs)\n",
    "    # orthologs_genes.append(orthologs)\n",
    "    # orthologs_organism.append(orthologs_organism_list)\n",
    "    # aa_sequences.append(amino_acid_sequence)\n",
    "\n",
    "    # print(counter, gene_name, gene_name_alsoknownas, gene_ncbiID, exon, aa_bool, protein_NP_weirdName, gene_uniprotIDs, orthologs, orthologs_organism_list)\n",
    "    # print(gene_symbol, gene_synonyms, gene_EntrezID, gene_name, gene_description, go_terms, cogs, exon_count, chromosome, strand, human_ortholog_EntrezID, mouse_ortholog_EntrezID, assembly_specific_gene_symbol, gene_references)\n",
    "    \n",
    "    # Loop over the Genes\n",
    "    print(counter, \"|GENE: \", gene_symbol, \"-----\", gene_synonyms, \"-----\", gene_EntrezID, \"-----\", gene_name, \"-----\", gene_description[0:25], \"... -----\", go_terms[0:25], \"... -----\", cogs, \"-----\", exon_count, \"-----\", chromosome, \"-----\", strand, \"-----\", human_ortholog_EntrezID, \"-----\", mouse_ortholog_EntrezID)\n",
    "    \n",
    "    # Loop over the assembly specifics\n",
    "    for assembly_specific_info in gene_record[0]['Entrezgene_locus']:\n",
    "        assembly_id = assembly_specific_info['Gene-commentary_heading']\n",
    "        location_sequence = assembly_specific_info['Gene-commentary_accession']\n",
    "        start_range = assembly_specific_info['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_from']\n",
    "        end_range = assembly_specific_info['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_to']\n",
    "        \n",
    "        # Loop over the transcripts of the Assembly\n",
    "        for assembly_specific_transcript in assembly_specific_info['Gene-commentary_products']:\n",
    "            transcript_sequence_ID = assembly_specific_transcript['Gene-commentary_accession']\n",
    "            print(\"ASSEMBLY SPECIFC GENE: S1, S2, \", gene_EntrezID, \"-----\", location_sequence, \"-----\", start_range, \"-----\", end_range, \"-----\", transcript_sequence_ID, \"-----\", assembly_id)\n",
    "\n",
    "            # Loop over the Transcripts and proteins\n",
    "            # for transcripts in assembly_specific_gene['Gene-commentary_products']:\n",
    "            mRNA_handle = Entrez.efetch(db=\"nucleotide\", id=transcript_sequence_ID, rettype=\"fasta\")\n",
    "            mRNA_record = mRNA_handle.read()\n",
    "            mRNA_sequence = ''.join(mRNA_record.split('\\n')[1:])           \n",
    "            #\n",
    "            # Get Exon information from the gff file \n",
    "            # Downloadable from the assembly page\n",
    "            #\n",
    "            protein_IDs = []\n",
    "            for Entrez_Comments in gene_record[0]['Entrezgene_comments']:\n",
    "                if 'Gene-commentary_comment' in Entrez_Comments:\n",
    "                    if 'Gene-commentary_products' in Entrez_Comments['Gene-commentary_comment'][0]:\n",
    "                        for gene_products_protein in Entrez_Comments['Gene-commentary_comment'][0]['Gene-commentary_products']:\n",
    "                            uniprotID_list = []\n",
    "                            \n",
    "                            # gene_record[0]['Entrezgene_comments'][4]['Gene-commentary_comment'][1]\n",
    "                            \n",
    "                            # If there is only 1 transcript:\n",
    "                            if 'Gene-commentary_accession' in gene_products_protein:\n",
    "                                if gene_products_protein['Gene-commentary_accession'] == transcript_sequence_ID:\n",
    "                                    # Get the Protein name                             \n",
    "                                    uniprotID_list = []\n",
    "                                    if 'Gene-commentary_comment' in gene_products_protein['Gene-commentary_products'][0]:\n",
    "                                        for gene_products_protein_othername in gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_comment']:        \n",
    "                                            if 'Gene-commentary_comment' in gene_products_protein_othername:\n",
    "                                                for gene_products_protein_uniprotID in gene_products_protein_othername['Gene-commentary_comment'][0]['Gene-commentary_source']:\n",
    "                                                    if 'Other-source_src' in gene_products_protein_uniprotID:\n",
    "                                                        if 'Object-id_str' in gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']:\n",
    "                                                            protein_uniprot_id = gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                                            uniprotID_list.append(protein_uniprot_id)\n",
    "\n",
    "                                    protein_sequence_ID = gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_accession']\n",
    "                                    protein_handle = Entrez.efetch(db=\"protein\", id=protein_sequence_ID, rettype=\"fasta\")\n",
    "                                    protein_record = protein_handle.read()\n",
    "                                    protein_sequence = ''.join(protein_record.split('\\n')[1:])  \n",
    "                                    print(\"TRANSCRIPTS: \", transcript_sequence_ID, \"-----\", protein_sequence_ID, \"-----\", mRNA_sequence[0:25], '-----', assembly_id)\n",
    "                                    # Loop over the UniprotIDs and get information\n",
    "                                    subcellular_locations_list = []\n",
    "                                    for UniprotID in uniprotID_list:\n",
    "                                        name, weight, length, annotation_score, existance, subcellular_locations_list, tissue_specificity = UniProt_Data_Request(UniprotID)\n",
    "                                        print(\"PROTEINS: \", UniprotID, \"-----\", name, \"-----\", existance, \"-----\", annotation_score, \"-----\", length, \"-----\", protein_sequence[0:25], \"-----\", weight, \"kDa -----\", subcellular_locations_list, \"-----\", tissue_specificity, \"-----\", protein_sequence_ID)\n",
    "        \n",
    "\n",
    "                            # If there are multiple trancripts\n",
    "                            if 'Gene-commentary_products' in gene_products_protein:\n",
    "                                for gene_products_protein_perAssembly in gene_products_protein['Gene-commentary_products']:\n",
    "                                    if gene_products_protein_perAssembly['Gene-commentary_accession'] == transcript_sequence_ID:\n",
    "\n",
    "                                        # Get the Protein name                             \n",
    "                                        uniprotID_list = []\n",
    "                                        if 'Gene-commentary_comment' in gene_products_protein_perAssembly['Gene-commentary_products'][0]:\n",
    "                                            for gene_products_protein_othername in gene_products_protein_perAssembly['Gene-commentary_products'][0]['Gene-commentary_comment']:        \n",
    "                                                if 'Gene-commentary_comment' in gene_products_protein_othername:\n",
    "                                                    for gene_products_protein_uniprotID in gene_products_protein_othername['Gene-commentary_comment'][0]['Gene-commentary_source']:\n",
    "                                                        if 'Other-source_src' in gene_products_protein_uniprotID:\n",
    "                                                            if 'Object-id_str' in gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']:\n",
    "                                                                protein_uniprot_id = gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                                                uniprotID_list.append(protein_uniprot_id)\n",
    "                                                            \n",
    "                                        protein_sequence_ID = gene_products_protein_perAssembly['Gene-commentary_products'][0]['Gene-commentary_accession']\n",
    "                                        protein_handle = Entrez.efetch(db=\"protein\", id=protein_sequence_ID, rettype=\"fasta\")\n",
    "                                        protein_record = protein_handle.read()\n",
    "                                        protein_sequence = ''.join(protein_record.split('\\n')[1:])  \n",
    "                                        print(\"TRANSCRIPTS: \", transcript_sequence_ID, \"-----\", protein_sequence_ID, \"-----\", mRNA_sequence[0:25], '-----', assembly_id)\n",
    "                                        # Loop over the UniprotIDs and get information\n",
    "                                        subcellular_locations_list = []\n",
    "                                        for UniprotID in uniprotID_list:\n",
    "                                            name, weight, length, annotation_score, existance, subcellular_locations_list, tissue_specificity = UniProt_Data_Request(UniprotID)\n",
    "                                            print(\"PROTEINS: \", UniprotID, \"-----\", name, \"-----\", existance, \"-----\", annotation_score, \"-----\", length, \"-----\", protein_sequence[0:25], \"-----\", weight, \"kDa -----\", subcellular_locations_list, \"-----\", tissue_specificity, \"-----\", protein_sequence_ID)                         \n",
    "   \n",
    "\n",
    "                    else:\n",
    "                        for d in Entrez_Comments['Gene-commentary_comment']:\n",
    "                            if 'Gene-commentary_comment' in d:\n",
    "                                print(d['Gene-commentary_products'][1]['Gene-commentary_accession'])\n",
    "\n",
    "    print(\"-----------------------------------------------------------------------------------------\")\n",
    "    counter += 1\n",
    "    # if counter % 1000 == 0:\n",
    "    #     data = {\n",
    "    #         'NCBI_ID': ncbi_geneID_list,\n",
    "    #         'Gene_Name': ncbi_geneName_list,\n",
    "    #         'NP_Name': ncbi_protein_NP_weirdName,\n",
    "    #         'Uniprot_ID': uniprot_IDs,\n",
    "    #         'Ortholog_Genes': orthologs_genes,\n",
    "    #         'Ortholog_Organisms': orthologs_organism,\n",
    "    #         'Exon Number': ncbi_gene_exon_size,\n",
    "    #         'Sequence': aa_sequences\n",
    "    #     }\n",
    "    #     df = pd.DataFrame(data)\n",
    "    #     file_path = 'GeneIDs_' + str(counter) + '.xlsx'\n",
    "    #     df.to_excel(file_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "id": "1ecd560a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'RefSeq Status', 'Gene-commentary_label': 'MODEL'}\n",
      "{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'Annotation Information', 'Gene-commentary_properties': [{'Gene-commentary_type': StringElement('16', attributes={'value': 'property'}), 'Gene-commentary_label': 'Annotation category', 'Gene-commentary_text': 'partial on reference assembly'}]}\n",
      "{'Gene-commentary_type': StringElement('23', attributes={'value': 'gene-group'}), 'Gene-commentary_heading': 'Orthologs from Annotation Pipeline', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('23', attributes={'value': 'gene-group'}), 'Gene-commentary_label': 'Primary', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'GeneID', 'Dbtag_tag': {'Object-id': {'Object-id_id': '2068'}}}}, 'Other-source_pre-text': 'Ercc2', 'Other-source_anchor': 'human'}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'taxon', 'Dbtag_tag': {'Object-id': {'Object-id_id': '9606'}}}}}]}, {'Gene-commentary_type': StringElement('23', attributes={'value': 'gene-group'}), 'Gene-commentary_label': 'Additional', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'GeneID', 'Dbtag_tag': {'Object-id': {'Object-id_id': '13871'}}}}, 'Other-source_pre-text': 'Ercc2', 'Other-source_anchor': 'mouse'}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'taxon', 'Dbtag_tag': {'Object-id': {'Object-id_id': '10090'}}}}}]}]}\n",
      "{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_refs': [{'Pub_pmid': {'PubMedId': '11182546'}}, {'Pub_pmid': {'PubMedId': '8052270'}}, {'Pub_pmid': {'PubMedId': '7851887'}}, {'Pub_pmid': {'PubMedId': '7596368'}}, {'Pub_pmid': {'PubMedId': '7558017'}}]}\n",
      "{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'NCBI Reference Sequences (RefSeq)', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'Suppressed Reference Sequence(s)', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_text': 'NM_001244391.1: This RefSeq was removed because currently there is insufficient support for the transcript.', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '346986336'}}}}, 'Other-source_anchor': 'NM_001244391.1'}]}]}, {'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'RefSeqs of Annotated Genomes: Cricetulus griseus Annotation Release 104', 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Reference CriGri_1.0 Primary Assembly', 'Gene-commentary_accession': 'NW_003614150', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '351517399'}}}}}], 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '553134', 'Seq-interval_to': '561663', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '351517399'}}, 'Seq-interval_fuzz-to': {'Int-fuzz': {'Int-fuzz_lim': StringElement('', attributes={'value': 'gt'})}}}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('3', attributes={'value': 'mRNA'}), 'Gene-commentary_accession': 'XM_035457766', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1868062369'}}}}}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1868062369'}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'XP_035313657', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1868062370'}}}}, 'Other-source_anchor': 'XP_035313657', 'Other-source_post-text': 'general transcription and DNA repair factor IIH helicase subunit XPD isoform X1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1868062370'}}}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'Conserved Domains', 'Gene-commentary_source': [{'Other-source_pre-text': '(1)', 'Other-source_anchor': 'summary'}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'CDD', 'Dbtag_tag': {'Object-id': {'Object-id_id': '273169'}}}}, 'Other-source_anchor': 'TIGR00604: rad3; DNA repair helicase (rad3)'}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_text': 'Location: 7 - 586'}]}]}]}]}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Alternate CriGri-PICRH-1.0', 'Gene-commentary_accession': 'NC_048602', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1859291577'}}}}}], 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '27962398', 'Seq-interval_to': '27977164', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '1859291577'}}}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('3', attributes={'value': 'mRNA'}), 'Gene-commentary_accession': 'XM_027430909', 'Gene-commentary_version': '2', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1868365007'}}}}}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1868365007'}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'XP_027286710', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1537997429'}}}}, 'Other-source_anchor': 'XP_027286710', 'Other-source_post-text': 'general transcription and DNA repair factor IIH helicase subunit XPD isoform X2'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1537997429'}}}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'Related', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Ensembl', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'ENSCGRP00015009328.1'}}}}}]}, {'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'UniProtKB', 'Gene-commentary_source': [], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/TrEMBL', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'A0A8C2M382'}}}}, 'Other-source_anchor': 'A0A8C2M382'}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/TrEMBL', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'A0A9J7GHR4'}}}}, 'Other-source_anchor': 'A0A9J7GHR4'}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/Swiss-Prot', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'Q60452'}}}}, 'Other-source_anchor': 'Q60452'}]}]}]}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'Related', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Ensembl', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'ENSCGRT00015011478.2'}}}}}]}]}, {'Gene-commentary_type': StringElement('3', attributes={'value': 'mRNA'}), 'Gene-commentary_accession': 'XM_027430908', 'Gene-commentary_version': '2', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1868365006'}}}}}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1868365006'}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'XP_027286709', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1537997427'}}}}, 'Other-source_anchor': 'XP_027286709', 'Other-source_post-text': 'general transcription and DNA repair factor IIH helicase subunit XPD isoform X2'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1537997427'}}}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'Related', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Ensembl', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'ENSCGRP00015045438.1'}}}}}]}, {'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'UniProtKB', 'Gene-commentary_source': [], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/TrEMBL', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'A0A8C2M382'}}}}, 'Other-source_anchor': 'A0A8C2M382'}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/TrEMBL', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'A0A9J7GHR4'}}}}, 'Other-source_anchor': 'A0A9J7GHR4'}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/Swiss-Prot', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'Q60452'}}}}, 'Other-source_anchor': 'Q60452'}]}]}]}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_heading': 'Related', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Ensembl', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'ENSCGRT00015077223.1'}}}}}]}]}]}]}]}\n",
      "{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'Related Sequences', 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'AFTD01085842', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '342549610'}}}}, 'Other-source_anchor': 'AFTD01085842.1'}], 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '938', 'Seq-interval_to': '1234', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '342549610'}}}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'AFTD01085843', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '342549609'}}}}, 'Other-source_anchor': 'AFTD01085843.1'}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'AFTD01085844', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '342549608'}}}}, 'Other-source_anchor': 'AFTD01085844.1'}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'AFTD01085845', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '342549607'}}}}, 'Other-source_anchor': 'AFTD01085845.1'}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'AH006035', 'Gene-commentary_version': '3', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1036029482'}}}}, 'Other-source_anchor': 'AH006035.3'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '1036029482'}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'AAC13749', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '604908'}}}}, 'Other-source_anchor': 'AAC13749.1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '604908'}}}]}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'JH000571', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '344164494'}}}}, 'Other-source_anchor': 'JH000571.1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '344164494'}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'EGW05341', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '344249237'}}}}, 'Other-source_anchor': 'EGW05341.1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '344249237'}}}]}, {'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'EGW05343', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '344249239'}}}}, 'Other-source_anchor': 'EGW05343.1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '344249239'}}}]}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_heading': 'Genomic', 'Gene-commentary_accession': 'RAZU02000010', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_id': '1850227006'}}}}, 'Other-source_anchor': 'RAZU02000010.1'}], 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '27962398', 'Seq-interval_to': '27977164', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '1850227006'}}}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}], 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_label': 'Strain', 'Gene-commentary_text': '17A/GY'}]}, {'Gene-commentary_type': StringElement('3', attributes={'value': 'mRNA'}), 'Gene-commentary_heading': 'mRNA', 'Gene-commentary_accession': 'IACG01021056', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'IACG01021056.1'}}}}, 'Other-source_anchor': 'IACG01021056.1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_ddbj': {'Textseq-id': {'Textseq-id_accession': 'IACG01021056', 'Textseq-id_version': '1'}}}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}]}, {'Gene-commentary_type': StringElement('3', attributes={'value': 'mRNA'}), 'Gene-commentary_heading': 'mRNA', 'Gene-commentary_accession': 'IACG01021066', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Nucleotide', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'IACG01021066.1'}}}}, 'Other-source_anchor': 'IACG01021066.1'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_ddbj': {'Textseq-id': {'Textseq-id_accession': 'IACG01021066', 'Textseq-id_version': '1'}}}}}], 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_text': 'None'}]}, {'Gene-commentary_type': StringElement('255', attributes={'value': 'other'}), 'Gene-commentary_text': 'None', 'Gene-commentary_products': [{'Gene-commentary_type': StringElement('8', attributes={'value': 'peptide'}), 'Gene-commentary_accession': 'Q60452', 'Gene-commentary_version': '1', 'Gene-commentary_source': [{'Other-source_src': {'Dbtag': {'Dbtag_db': 'Protein', 'Dbtag_tag': {'Object-id': {'Object-id_id': '2495145'}}}}}, {'Other-source_src': {'Dbtag': {'Dbtag_db': 'UniProtKB/Swiss-Prot', 'Dbtag_tag': {'Object-id': {'Object-id_str': 'Q60452.1'}}}}, 'Other-source_anchor': 'UniProtKB/Swiss-Prot:Q60452', 'Other-source_url': 'https://www.uniprot.org/uniprot/Q60452'}], 'Gene-commentary_seqs': [{'Seq-loc_whole': {'Seq-id': {'Seq-id_gi': '2495145'}}}]}]}]}\n",
      "{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'Gene Location History', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'Cricetulus griseus Annotation Release 104', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('24', attributes={'value': 'assembly'}), 'Gene-commentary_heading': 'CriGri_1.0', 'Gene-commentary_label': 'Reference', 'Gene-commentary_text': 'current', 'Gene-commentary_accession': 'GCF_000223135', 'Gene-commentary_version': '1', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('25', attributes={'value': 'assembly-unit'}), 'Gene-commentary_heading': 'Primary Assembly', 'Gene-commentary_accession': 'GCF_000223145', 'Gene-commentary_version': '1', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_label': 'Unplaced Scaffold', 'Gene-commentary_accession': 'NW_003614150', 'Gene-commentary_version': '1', 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '553134', 'Seq-interval_to': '561663', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '351517399'}}, 'Seq-interval_fuzz-to': {'Int-fuzz': {'Int-fuzz_lim': StringElement('', attributes={'value': 'gt'})}}}}}]}]}]}, {'Gene-commentary_type': StringElement('24', attributes={'value': 'assembly'}), 'Gene-commentary_heading': 'CriGri-PICRH-1.0', 'Gene-commentary_label': 'Alternate', 'Gene-commentary_text': 'current', 'Gene-commentary_accession': 'GCF_003668045', 'Gene-commentary_version': '3', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('25', attributes={'value': 'assembly-unit'}), 'Gene-commentary_heading': 'Primary Assembly', 'Gene-commentary_accession': 'GCF_003668145', 'Gene-commentary_version': '2', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_label': 'Chromosome 9', 'Gene-commentary_accession': 'NC_048602', 'Gene-commentary_version': '1', 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '27962398', 'Seq-interval_to': '27977164', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '1859291577'}}}}}]}]}]}]}, {'Gene-commentary_type': StringElement('254', attributes={'value': 'comment'}), 'Gene-commentary_heading': 'Cricetulus griseus Annotation Release 103', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('24', attributes={'value': 'assembly'}), 'Gene-commentary_heading': 'CriGri_1.0', 'Gene-commentary_label': 'Reference', 'Gene-commentary_accession': 'GCF_000223135', 'Gene-commentary_version': '1', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('25', attributes={'value': 'assembly-unit'}), 'Gene-commentary_heading': 'Primary Assembly', 'Gene-commentary_accession': 'GCF_000223145', 'Gene-commentary_version': '1', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_label': 'Unplaced Scaffold', 'Gene-commentary_accession': 'NW_003614150', 'Gene-commentary_version': '1', 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '552205', 'Seq-interval_to': '559983', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '351517399'}}, 'Seq-interval_fuzz-to': {'Int-fuzz': {'Int-fuzz_lim': StringElement('', attributes={'value': 'gt'})}}}}}]}]}]}, {'Gene-commentary_type': StringElement('24', attributes={'value': 'assembly'}), 'Gene-commentary_heading': 'CriGri-PICR', 'Gene-commentary_label': 'Alternate', 'Gene-commentary_accession': 'GCF_003668045', 'Gene-commentary_version': '1', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('25', attributes={'value': 'assembly-unit'}), 'Gene-commentary_heading': 'Primary Assembly', 'Gene-commentary_accession': 'GCF_003668145', 'Gene-commentary_version': '1', 'Gene-commentary_comment': [{'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_label': 'Chromosome 9 Unlocalized Scaffold', 'Gene-commentary_accession': 'NW_020822646', 'Gene-commentary_version': '1', 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '603653', 'Seq-interval_to': '617830', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '1497152704'}}}}}]}, {'Gene-commentary_type': StringElement('1', attributes={'value': 'genomic'}), 'Gene-commentary_label': 'Unplaced Scaffold', 'Gene-commentary_accession': 'NW_020822787', 'Gene-commentary_version': '1', 'Gene-commentary_seqs': [{'Seq-loc_int': {'Seq-interval': {'Seq-interval_from': '21', 'Seq-interval_to': '4609', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'plus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '1497151473'}}}}}]}]}]}]}]}\n",
      "[3]\n"
     ]
    }
   ],
   "source": [
    "for i in gene_record[0]['Entrezgene_comments']:\n",
    "    print(i)\n",
    "print([3])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "c684caba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_048602'"
      ]
     },
     "execution_count": 307,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_record[0]['Entrezgene_comments'][4]['Gene-commentary_comment'][1]['Gene-commentary_products'][1]['Gene-commentary_accession']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "e9bf68be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['G3HKR2', 'P97260', 'Q2WEK9']\n"
     ]
    }
   ],
   "source": [
    "uniprotID_list = []\n",
    "if 'Gene-commentary_accession' in gene_products_protein:\n",
    "    if gene_products_protein['Gene-commentary_accession'] == transcript_sequence_ID:\n",
    "        # Get the Protein name                             \n",
    "        uniprotID_list = []\n",
    "        if 'Gene-commentary_comment' in gene_products_protein['Gene-commentary_products'][0]:\n",
    "            for gene_products_protein_othername in gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_comment']:        \n",
    "                if 'Gene-commentary_comment' in gene_products_protein_othername:\n",
    "                    for gene_products_protein_uniprotID in gene_products_protein_othername['Gene-commentary_comment'][0]['Gene-commentary_source']:\n",
    "                        if 'Other-source_src' in gene_products_protein_uniprotID:\n",
    "                            if 'Object-id_str' in gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']:\n",
    "                                protein_uniprot_id = gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                uniprotID_list.append(protein_uniprot_id)\n",
    "print(uniprotID_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "PermissionError",
     "evalue": "[WinError 32] The process cannot access the file because it is being used by another process: 'gene_model.db'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mPermissionError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\a.antonakoudis\\OneDrive - Sartorius Corporate Administration GmbH\\Documents\\Projects\\UCSD Project\\Whole-Cell-Network-Reconstruction-for-CHO-cells-2\\Notebooks\\Genes.ipynb Cell 24\u001b[0m line \u001b[0;36m7\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X32sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m gene_model_file \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m../../../../../../Downloads/ncbi_dataset/ncbi_dataset/data/GCF_000223135.1/genomic.gff\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X32sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m \u001b[39m# Create a GFF database\u001b[39;00m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X32sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m db \u001b[39m=\u001b[39m gffutils\u001b[39m.\u001b[39;49mcreate_db(gene_model_file, dbfn\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mgene_model.db\u001b[39;49m\u001b[39m'\u001b[39;49m, force\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X32sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m \u001b[39m# Define the transcript ID\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/a.antonakoudis/OneDrive%20-%20Sartorius%20Corporate%20Administration%20GmbH/Documents/Projects/UCSD%20Project/Whole-Cell-Network-Reconstruction-for-CHO-cells-2/Notebooks/Genes.ipynb#X32sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m transcript_id \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mXM_027427916.1\u001b[39m\u001b[39m\"\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\a.antonakoudis\\Anaconda3\\envs\\GEM\\lib\\site-packages\\gffutils\\create.py:1399\u001b[0m, in \u001b[0;36mcreate_db\u001b[1;34m(data, dbfn, id_spec, force, verbose, checklines, merge_strategy, transform, gtf_transcript_key, gtf_gene_key, gtf_subfeature, force_gff, force_dialect_check, from_string, keep_order, text_factory, force_merge_fields, pragmas, sort_attribute_values, dialect, _keep_tempfiles, infer_gene_extent, disable_infer_genes, disable_infer_transcripts, **kwargs)\u001b[0m\n\u001b[0;32m   1397\u001b[0m kwargs\u001b[39m.\u001b[39mupdate(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39madd_kwargs)\n\u001b[0;32m   1398\u001b[0m kwargs[\u001b[39m\"\u001b[39m\u001b[39mdialect\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m dialect\n\u001b[1;32m-> 1399\u001b[0m c \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39m(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m   1401\u001b[0m c\u001b[39m.\u001b[39mcreate()\n\u001b[0;32m   1402\u001b[0m \u001b[39mif\u001b[39;00m dbfn \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m:memory:\u001b[39m\u001b[39m\"\u001b[39m:\n",
      "File \u001b[1;32mc:\\Users\\a.antonakoudis\\Anaconda3\\envs\\GEM\\lib\\site-packages\\gffutils\\create.py:590\u001b[0m, in \u001b[0;36m_GFFDBCreator.__init__\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    584\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m    585\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    586\u001b[0m \u001b[39m    _DBCreator subclass specifically for working with GFF files.\u001b[39;00m\n\u001b[0;32m    587\u001b[0m \n\u001b[0;32m    588\u001b[0m \u001b[39m    create_db() delegates to this class -- see that function for docs\u001b[39;00m\n\u001b[0;32m    589\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 590\u001b[0m     \u001b[39msuper\u001b[39m(_GFFDBCreator, \u001b[39mself\u001b[39m)\u001b[39m.\u001b[39m\u001b[39m__init__\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\a.antonakoudis\\Anaconda3\\envs\\GEM\\lib\\site-packages\\gffutils\\create.py:119\u001b[0m, in \u001b[0;36m_DBCreator.__init__\u001b[1;34m(self, data, dbfn, force, verbose, id_spec, merge_strategy, checklines, transform, force_dialect_check, from_string, dialect, default_encoding, disable_infer_genes, disable_infer_transcripts, infer_gene_extent, force_merge_fields, text_factory, pragmas, _keep_tempfiles, directives, **kwargs)\u001b[0m\n\u001b[0;32m    117\u001b[0m \u001b[39mif\u001b[39;00m force:\n\u001b[0;32m    118\u001b[0m     \u001b[39mif\u001b[39;00m os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mexists(dbfn):\n\u001b[1;32m--> 119\u001b[0m         os\u001b[39m.\u001b[39;49munlink(dbfn)\n\u001b[0;32m    120\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdbfn \u001b[39m=\u001b[39m dbfn\n\u001b[0;32m    121\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mid_spec \u001b[39m=\u001b[39m id_spec\n",
      "\u001b[1;31mPermissionError\u001b[0m: [WinError 32] The process cannot access the file because it is being used by another process: 'gene_model.db'"
     ]
    }
   ],
   "source": [
    "import gffutils\n",
    "\n",
    "# Path to your GFF or GTF file\n",
    "gene_model_file = \"../../../../../../Downloads/ncbi_dataset/ncbi_dataset/data/GCF_000223135.1/genomic.gff\"\n",
    "\n",
    "# Create a GFF database\n",
    "db = gffutils.create_db(gene_model_file, dbfn='gene_model.db', force=True)\n",
    "\n",
    "# Define the transcript ID\n",
    "transcript_id = \"XM_027427916.1\"\n",
    "\n",
    "# Retrieve exon information for the given transcript\n",
    "exons = []\n",
    "for feature in db.features_of_type(\"exon\", order_by=\"start\"):\n",
    "    if transcript_id in feature.attributes.get('Parent', []):\n",
    "        exons.append(feature)\n",
    "\n",
    "# Print exon details\n",
    "for i, exon in enumerate(exons, 1):\n",
    "    print(f\"Exon {i}:\")\n",
    "    print(f\"Location: {exon.seqid}:{exon.start}-{exon.end}\")\n",
    "    print(f\"Strand: {exon.strand}\")\n",
    "    print(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9ef5677f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'LOCUS       XM_007608016            3801 bp    mRNA    linear   ROD 10-JUL-2020\\nDEFINITION  PREDICTED: Cricetulus griseus N-ethylmaleimide sensitive factor,\\n            vesicle fusing ATPase (Nsf), transcript variant X1, mRNA.\\nACCESSION   XM_007608016\\nVERSION     XM_007608016.2\\nDBLINK      BioProject: PRJNA72741\\nKEYWORDS    RefSeq; corrected model.\\nSOURCE      Cricetulus griseus (Chinese hamster)\\n  ORGANISM  Cricetulus griseus\\n            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\\n            Mammalia; Eutheria; Euarchontoglires; Glires; Rodentia; Myomorpha;\\n            Muroidea; Cricetidae; Cricetinae; Cricetulus.\\nCOMMENT     MODEL REFSEQ:  This record is predicted by automated computational\\n            analysis. This record is derived from a genomic sequence\\n            (NW_003622620.1) and transcript sequence (IACG01091083.1) annotated\\n            using gene prediction method: Gnomon.\\n            Also see:\\n                Documentation of NCBI\\'s Annotation Process\\n            \\n            On Jul 10, 2020 this sequence version replaced XM_007608016.1.\\n            \\n            ##Genome-Annotation-Data-START##\\n            Annotation Provider         :: NCBI\\n            Annotation Status           :: Full annotation\\n            Annotation Name             :: Cricetulus griseus Annotation\\n                                           Release 104\\n            Annotation Version          :: 104\\n            Annotation Pipeline         :: NCBI eukaryotic genome annotation\\n                                           pipeline\\n            Annotation Software Version :: 8.4\\n            Annotation Method           :: Best-placed RefSeq; Gnomon\\n            Features Annotated          :: Gene; mRNA; CDS; ncRNA\\n            ##Genome-Annotation-Data-END##\\n            \\n            ##RefSeq-Attributes-START##\\n            assembly gap :: added 2223 transcript bases to patch genome\\n                            assembly gap\\n            ##RefSeq-Attributes-END##\\nPRIMARY     REFSEQ_SPAN         PRIMARY_IDENTIFIER PRIMARY_SPAN        COMP\\n            1-2223              IACG01091083.1     1598-3820           c\\n            2224-2279           AFTD01265364.1     2549-2604\\n            2280-3801           AFTD01265364.1     3074-4595\\nFEATURES             Location/Qualifiers\\n     source          1..3801\\n                     /organism=\"Cricetulus griseus\"\\n                     /mol_type=\"mRNA\"\\n                     /db_xref=\"taxon:10029\"\\n                     /chromosome=\"Unknown\"\\n     gene            1..3801\\n                     /gene=\"Nsf\"\\n                     /note=\"The sequence of the model RefSeq transcript was\\n                     modified relative to its source genomic sequence to\\n                     represent the inferred CDS: added 2223 bases not found in\\n                     genome assembly; Derived by automated computational\\n                     analysis using gene prediction method: Gnomon. Supporting\\n                     evidence includes similarity to: 95% coverage of the\\n                     annotated genomic feature by RNAseq alignments, including\\n                     116 samples with support for all annotated introns\"\\n                     /db_xref=\"GeneID:100770898\"\\n     CDS             67..2301\\n                     /gene=\"Nsf\"\\n                     /note=\"The sequence of the model RefSeq protein was\\n                     modified relative to its source genomic sequence to\\n                     represent the inferred CDS: added 2157 bases not found in\\n                     genome assembly\"\\n                     /codon_start=1\\n                     /product=\"vesicle-fusing ATPase isoform X1\"\\n                     /protein_id=\"XP_007606206.1\"\\n                     /db_xref=\"GeneID:100770898\"\\n                     /translation=\"MAGRSMQAARCPTDELSLSNCAVVSEKDYQSGQHVIVRTSPNHK\\n                     YIFTLRTHPSVVPGSVAFSLPQRKWAGLSIGQEIEVALYSFDKAKQCIGTMTIEIDFL\\n                     QKKNIDSNPYDTDKMAAEFIQQFNNQAFSVGQQLVFSFNDKLFGLLVKDIEAMDPSIL\\n                     KGEPASGKRQKIEVGLVVGNSQVAFEKAENSSLNLIGKAKTKENRQSIINPDWNFEKM\\n                     GIGGLDKEFSDIFRRAFASRVFPPEIVEQMGCKHVKGILLYGPPGCGKTLLARQIGKM\\n                     LNAREPKVVNGPEILNKYVGESEANIRKLFADAEEEQRRLGANSGLHIIIFDEIDAIC\\n                     KQRGSMAGSTGVHDTVVNQLLSKIDGVEQLNNILVIGMTNRPDLIDEALLRPGRLEVK\\n                     MEIGLPDEKGRLQILHIHTARMRGHQLLSADVDIKELAVETKNFSGAELEGLVRAAQS\\n                     TAMNRHIKASTKVEVDMEKAESLQVTRGDFLASLENDIKPAFGTNQEDYASYIMNGII\\n                     KWGDPVTRVLDDGELLVQQTKNSDRTPLVSVLLEGPPHSGKTALAAKIAEESNFPFIK\\n                     ICSPDKMIGFSETAKCQAMKKIFDDAYKSQLSCVVVDDIERLLDYVPIGPRFSNLVLQ\\n                     ALLVLLKKAPPQGRKLLIIGTTSRKDVLQEMEMLNAFSTTIHVPNIATGEQLLEALEL\\n                     LGNFKDKERTTIAQQVKGKKVWIGIKKLLMLIEMSLQMDPEYRVRKFLALLREEGASP\\n                     LDFD\"\\nORIGIN      \\n        1 cgcgcggcgg gggcgggctc agccgagtgc ggagctgcag cctcggaacc ggacgtgtcc\\n       61 gcgaagatgg cgggccggag tatgcaagct gcgaggtgcc ctaccgatga attgtcctta\\n      121 agcaactgtg cagttgtgag cgagaaggat taccagtctg gacagcatgt gattgtgagg\\n      181 acctccccca atcacaaata catatttaca ctgaggaccc atccgtcggt ggttcccgga\\n      241 agcgttgcat tcagcttacc tcagcgaaaa tgggctggac tttctattgg gcaggaaata\\n      301 gaagtcgcct tgtattcatt tgacaaagcc aagcaatgca tcggcaccat gaccatcgag\\n      361 attgactttc tgcagaaaaa aaacatcgac tccaacccat atgataccga caagatggca\\n      421 gcagaattca tccagcagtt caacaaccag gccttctcag tgggacagca gcttgtcttt\\n      481 agctttaatg ataagctctt cggattactg gtgaaggaca ttgaagccat ggaccctagc\\n      541 atcctgaagg gagagcctgc atcaggcaaa aggcagaaga ttgaagtcgg actggttgtt\\n      601 ggaaatagtc aagttgcatt tgaaaaagca gaaaattcat cgctcaatct tattggcaaa\\n      661 gctaaaacca aggaaaatcg ccagtcaatc atcaatcctg actggaactt tgagaaaatg\\n      721 ggaatcggcg gtctggataa ggagttttca gatattttcc gacgagcatt tgcttctcgg\\n      781 gtgttccctc cggagattgt ggagcagatg ggttgtaaac atgttaaagg catcctgttg\\n      841 tacggccccc ctggttgtgg taagactctc ttggctcgac agattggcaa gatgctgaat\\n      901 gcgagagaac ccaaggtggt caatgggcca gaaatcctta acaaatatgt gggagaatcg\\n      961 gaggcgaaca ttcgtaaact ttttgctgat gccgaagagg agcaaaggag gcttggtgct\\n     1021 aacagtggct tacacatcat catctttgat gaaatcgatg ccatctgcaa gcagagaggg\\n     1081 agcatggctg gcagcaccgg agttcacgac accgttgtca accagctgct gtccaagatt\\n     1141 gatggcgttg agcagctgaa caacatcctg gttattggga tgaccaatag accagatttg\\n     1201 atagatgagg ctctccttcg acctggaaga ctggaagtta aaatggagat aggcttgcca\\n     1261 gatgagaagg gtcgactgca gatccttcat atccacacag caaggatgag aggacatcag\\n     1321 ttactgtctg cagatgtgga cattaaggaa ctggccgtgg agactaagaa tttcagtggc\\n     1381 gctgagctgg aaggtctggt gcgagcagca cagtcgacgg ccatgaaccg acacataaag\\n     1441 gccagtacga aagtggaagt agacatggag aaagcagaga gtctgcaggt gacaagagga\\n     1501 gacttcctgg cttctttgga gaatgacatc aaaccagctt ttggcacaaa ccaagaggat\\n     1561 tatgccagtt acattatgaa tggaatcatc aagtggggtg accctgtgac ccgtgtcctg\\n     1621 gatgacggag agctgctggt ccagcagact aaaaacagtg accgtacgcc tctggttagc\\n     1681 gtccttctgg aagggccccc tcacagcggg aagactgccc tagctgcaaa gatcgcagag\\n     1741 gaatccaact ttcccttcat caagatctgc tcccccgata agatgattgg cttttctgag\\n     1801 acagccaagt gtcaggccat gaagaagatc tttgatgatg cttacaaatc tcaactcagt\\n     1861 tgtgtggtgg tggatgacat cgaaaggcta cttgattatg tccccattgg cccccggttc\\n     1921 tccaacctgg tgttgcaagc tcttcttgtg ttactaaaga aggcacctcc tcagggccgc\\n     1981 aagcttctca tcatcgggac caccagccgc aaagatgtcc ttcaggaaat ggagatgctg\\n     2041 aatgccttca gcaccaccat ccacgtgccc aacattgcca caggagagca gctgctggaa\\n     2101 gctttggagc ttctgggcaa cttcaaagat aaggaacgca ccacaattgc tcagcaagtc\\n     2161 aaagggaaga aggtctggat aggaatcaag aagttactaa tgttgattga gatgtccctg\\n     2221 cagatggacc ctgagtatcg ggtgaggaaa ttcttggccc tcttgagaga agaaggcgct\\n     2281 agccccctgg actttgattg aaaatggatt gtttacagca gtgaccaagg gaagtgacca\\n     2341 aggtgaagat ggcctgggat cgtcactcat cgtcctcaag acaccagact aagtgaaaca\\n     2401 ttcaccgccc tccccacact tgctcagcat gactagcgca gtgaactccc tcgtgttttt\\n     2461 gagacatcct agtgtttccc ggaaggtgtt gacttgcttt agactgctct gctcaccctc\\n     2521 ccgtgctagc tcaggtgatc cctgcttatg tgggaggcat gctgcacctg tgcacattgc\\n     2581 tcctgccaca cccacccccc agaagacact tggcctagca gaaggtaggc ttcctgatgt\\n     2641 gaccaagaac aaaggcacaa gttccgaggg actcccttag cgttcctctg tgccaacagc\\n     2701 tgtgtgtctg gctaaaaggt tcttccatag aatgggctat tatctttgcc tagagaagca\\n     2761 ggctttgaat atccaagatg ctgttctcaa tcatattcag agatatttcc aagttgcaat\\n     2821 caggagagct ttcctaatcg gaagtccgat cattcctgtg gtggctgtga tatgccctga\\n     2881 gtgtaattag tttgctctgt ggtaagagat atgctaatca ttaccacttg gtagatgctg\\n     2941 ttaaaacctt atgaaagata catatggaaa gctgtacatc cctaaacaga aaaaaatatt\\n     3001 aaaataattt acaagcctct cagctctaat tagtgtccaa ctctaagtgg gtcagctcct\\n     3061 tgggatgacg tcagaggctt attagtgcca tattcttcct tttcctcagc tgaatggctt\\n     3121 ataatttact ctcctctatc tcaagtgatc tgtaatgtct aaagcacata agtaaaacac\\n     3181 tattatccag caaaacaagc cacctcctag tcctctccct caggccttga atatccattc\\n     3241 cagtgtactc agttcttcca aaacaggagg ggctcaccca gccctttgga gggtgatttg\\n     3301 tctaatcagt gggttagaag aaatttaatt agttcgtcat tctcggtgat gagtcttagg\\n     3361 gagcatgccc ctctccccgc tgctgctgaa aatacaaaga cagtgatcag aagtcagcca\\n     3421 ggaggcttga tgccgcaggg gcacctaggt atgatcctga cactgtcccc acagtagaag\\n     3481 gcaaatcatc tcctaatttg ctgtctgtat actttgtttt gggttctcat ccctggctta\\n     3541 ttttgtggct ttaagagtga attgtgttgt gaaactgtga tgcctgtggc cagtactgtg\\n     3601 gtccctgtcc atcaccatgg tccctgttgt ctgtggtcct gactttctcc tcccaacaaa\\n     3661 gccgtatatg gactaactta tgtttcttgt tcaatgaagt taatgtcttg tgtttttgcc\\n     3721 caaaactata ttgaaaaaga taaatattat attgtgtaat gcaaattgaa ggaatgcaat\\n     3781 aaagagtaaa tatacttgaa a\\n//\\n\\n'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1975cf6b",
   "metadata": {},
   "source": [
    "#### Combine different DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61015c42",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "All_Genes = pd.DataFrame()\n",
    "\n",
    "for i in range(1000, 35000, 1000):\n",
    "    Genes_temp = pd.read_excel('GeneIDs_' + str(i) + '.xlsx')\n",
    "    All_Genes = pd.concat([All_Genes, Genes_temp], ignore_index=True)\n",
    "All_Genes.to_excel('All_Genes.xlsx', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
